--- a/net/minecraft/entity/projectile/AbstractArrowEntity.java
+++ b/net/minecraft/entity/projectile/AbstractArrowEntity.java
@@ -17,6 +17,7 @@
 import net.minecraft.entity.IProjectile;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.Pose;
+import net.minecraft.entity.item.ItemEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.item.ItemStack;
@@ -47,6 +48,10 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_15_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.player.PlayerPickupArrowEvent;
 
 public abstract class AbstractArrowEntity extends Entity implements IProjectile {
    private static final DataParameter<Byte> field_184554_g = EntityDataManager.func_187226_a(AbstractArrowEntity.class, DataSerializers.field_187191_a);
@@ -54,19 +59,31 @@
    private static final DataParameter<Byte> field_213876_as = EntityDataManager.func_187226_a(AbstractArrowEntity.class, DataSerializers.field_187191_a);
    @Nullable
    private BlockState field_195056_av;
-   protected boolean field_70254_i;
+   public boolean field_70254_i; // protected->public CraftBukkit
    protected int field_184552_b;
    public AbstractArrowEntity.PickupStatus field_70251_a = AbstractArrowEntity.PickupStatus.DISALLOWED;
    public int field_70249_b;
    public UUID field_70250_c;
-   private int field_70252_j;
+   public int field_70252_j; // private->public CraftBukkit
    private int field_70257_an;
    private double field_70255_ao = 2.0D;
-   private int field_70256_ap;
+   public int field_70256_ap; // private->public CraftBukkit
    private SoundEvent field_213877_ay = this.func_213867_k();
    private IntOpenHashSet field_213878_az;
    private List<Entity> field_213875_aA;
 
+   // Spigot Start
+   @Override
+   public void inactiveTick()
+   {
+      if ( this.field_70254_i )
+      {
+         this.field_70252_j += 1;
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected AbstractArrowEntity(EntityType<? extends AbstractArrowEntity> p_i48546_1_, World p_i48546_2_) {
       super(p_i48546_1_, p_i48546_2_);
    }
@@ -160,7 +177,7 @@
 
       BlockPos blockpos = new BlockPos(this);
       BlockState blockstate = this.field_70170_p.func_180495_p(blockpos);
-      if (!blockstate.func_196958_f() && !flag) {
+      if (!blockstate.isAir(this.field_70170_p, blockpos) && !flag) {
          VoxelShape voxelshape = blockstate.func_196952_d(this.field_70170_p, blockpos);
          if (!voxelshape.func_197766_b()) {
             Vec3d vec3d1 = this.func_213303_ch();
@@ -218,7 +235,7 @@
                }
             }
 
-            if (raytraceresult != null && !flag) {
+            if (raytraceresult != null && raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
                this.func_184549_a(raytraceresult);
                this.field_70160_al = true;
             }
@@ -292,13 +309,14 @@
 
    protected void func_225516_i_() {
       ++this.field_70252_j;
-      if (this.field_70252_j >= 1200) {
+      if (this.field_70252_j >= ((this instanceof TridentEntity) ? field_70170_p.spigotConfig.tridentDespawnRate : field_70170_p.spigotConfig.arrowDespawnRate)) { // Spigot
          this.func_70106_y();
       }
 
    }
 
    protected void func_184549_a(RayTraceResult p_184549_1_) {
+      CraftEventFactory.callProjectileHitEvent(this, p_184549_1_); // CraftBukkit - Call event
       RayTraceResult.Type raytraceresult$type = p_184549_1_.func_216346_c();
       if (raytraceresult$type == RayTraceResult.Type.ENTITY) {
          this.func_213868_a((EntityRayTraceResult)p_184549_1_);
@@ -373,7 +391,13 @@
       boolean flag = entity.func_200600_R() == EntityType.field_200803_q;
       int j = entity.func_223314_ad();
       if (this.func_70027_ad() && !flag) {
-         entity.func_70015_d(5);
+         // CraftBukkit
+         EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+         Bukkit.getPluginManager().callEvent(combustEvent);
+         if(!combustEvent.isCancelled()){
+            entity.setOnFire(combustEvent.getDuration(), false);
+         }
+         // CraftBukkit
       }
 
       if (entity.func_70097_a(damagesource, (float)i)) {
@@ -510,6 +534,7 @@
 
    public void func_212361_a(@Nullable Entity p_212361_1_) {
       this.field_70250_c = p_212361_1_ == null ? null : p_212361_1_.func_110124_au();
+      this.projectileSource = p_212361_1_ == null ? null : (org.bukkit.entity.LivingEntity) p_212361_1_.getBukkitEntity(); // CraftBukkit
       if (p_212361_1_ instanceof PlayerEntity) {
          this.field_70251_a = ((PlayerEntity)p_212361_1_).field_71075_bZ.field_75098_d ? AbstractArrowEntity.PickupStatus.CREATIVE_ONLY : AbstractArrowEntity.PickupStatus.ALLOWED;
       }
@@ -523,8 +548,22 @@
 
    public void func_70100_b_(PlayerEntity p_70100_1_) {
       if (!this.field_70170_p.field_72995_K && (this.field_70254_i || this.func_203047_q()) && this.field_70249_b <= 0) {
+         // CraftBukkit start
+         ItemStack itemStack = this.func_184550_j();
+         if(this.field_70251_a == PickupStatus.ALLOWED && !itemStack.func_190926_b() && p_70100_1_.field_71071_by.canHold(itemStack) >0){
+            ItemEntity item = new ItemEntity(this.field_70170_p, this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), itemStack);
+            PlayerPickupArrowEvent event = new PlayerPickupArrowEvent((org.bukkit.entity.Player) p_70100_1_.getBukkitEntity(), new org.bukkit.craftbukkit.v1_15_R1.entity.CraftItem(this.field_70170_p.getServerCB(), this, item), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
+            this.field_70170_p.getServerCB().getPluginManager().callEvent(event);
+
+            if(event.isCancelled()){
+               return;
+            }
+            itemStack = item.func_92059_d();
+         }
+
          boolean flag = this.field_70251_a == AbstractArrowEntity.PickupStatus.ALLOWED || this.field_70251_a == AbstractArrowEntity.PickupStatus.CREATIVE_ONLY && p_70100_1_.field_71075_bZ.field_75098_d || this.func_203047_q() && this.func_212360_k().func_110124_au() == p_70100_1_.func_110124_au();
-         if (this.field_70251_a == AbstractArrowEntity.PickupStatus.ALLOWED && !p_70100_1_.field_71071_by.func_70441_a(this.func_184550_j())) {
+         if (this.field_70251_a == AbstractArrowEntity.PickupStatus.ALLOWED && !p_70100_1_.field_71071_by.func_70441_a(itemStack)) {
+            // CraftBukkit end
             flag = false;
          }
 
