--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityLargeFireball.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityLargeFireball.java
@@ -1,6 +1,5 @@
 package net.minecraft.entity.projectile;
 
-import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.DamageSource;
@@ -9,6 +8,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class EntityLargeFireball extends EntityFireball
 {
@@ -17,6 +17,8 @@
     public EntityLargeFireball(World p_i1767_1_)
     {
         super(p_i1767_1_);
+        // TODO: Maybe we should use net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent instead of this?
+        isIncendiary = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
     }
 
     @SideOnly(Side.CLIENT)
@@ -28,6 +30,8 @@
     public EntityLargeFireball(World p_i1769_1_, EntityLivingBase p_i1769_2_, double p_i1769_3_, double p_i1769_5_, double p_i1769_7_)
     {
         super(p_i1769_1_, p_i1769_2_, p_i1769_3_, p_i1769_5_, p_i1769_7_);
+        // TODO: Maybe we should use net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent instead of this?
+        isIncendiary = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
     }
 
     protected void func_70227_a(RayTraceResult p_70227_1_)
@@ -40,8 +44,17 @@
                 this.func_174815_a(this.field_70235_a, p_70227_1_.field_72308_g);
             }
 
-            boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
-            this.field_70170_p.func_72885_a((Entity)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_92057_e, flag, flag);
+            // TODO: Reimplement with correct `flag` usage below
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a);
+            // this.world.newExplosion((Entity)null, this.posX, this.posY, this.posZ, (float)this.explosionPower, flag, flag);
+            // CraftBukkit start - fire ExplosionPrimeEvent
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.v1_12_R1.entity.CraftEntity.getEntity(this.field_70170_p.getServer(), this));
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                // give 'this' instead of (Entity) null so we know what causes the damage
+                this.field_70170_p.func_72885_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, event.getRadius(), event.getFire(), isIncendiary);
+            }
+            // CraftBukkit end
             this.func_70106_y();
         }
     }
@@ -63,7 +76,8 @@
 
         if (p_70037_1_.func_150297_b("ExplosionPower", 99))
         {
-            this.field_92057_e = p_70037_1_.func_74762_e("ExplosionPower");
+            // CraftBukkit - set bukkitYield when setting explosionpower
+            bukkitYield = this.field_92057_e = p_70037_1_.func_74762_e("ExplosionPower");
         }
     }
 }
