--- a/net/minecraft/entity/projectile/WitherSkullEntity.java
+++ b/net/minecraft/entity/projectile/WitherSkullEntity.java
@@ -4,7 +4,6 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.LivingEntity;
-import net.minecraft.entity.boss.WitherEntity;
 import net.minecraft.fluid.IFluidState;
 import net.minecraft.network.datasync.DataParameter;
 import net.minecraft.network.datasync.DataSerializers;
@@ -17,11 +16,13 @@
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.Explosion;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class WitherSkullEntity extends DamagingProjectileEntity {
    private static final DataParameter<Boolean> field_184565_e = EntityDataManager.func_187226_a(WitherSkullEntity.class, DataSerializers.field_187198_h);
@@ -48,7 +49,7 @@
    }
 
    public float func_180428_a(Explosion p_180428_1_, IBlockReader p_180428_2_, BlockPos p_180428_3_, BlockState p_180428_4_, IFluidState p_180428_5_, float p_180428_6_) {
-      return this.func_82342_d() && WitherEntity.func_181033_a(p_180428_4_) ? Math.min(0.8F, p_180428_6_) : p_180428_6_;
+      return this.func_82342_d() && p_180428_4_.canEntityDestroy(p_180428_2_, p_180428_3_, this) ? Math.min(0.8F, p_180428_6_) : p_180428_6_;
    }
 
    protected void func_70227_a(RayTraceResult p_70227_1_) {
@@ -56,19 +57,23 @@
       if (!this.field_70170_p.field_72995_K) {
          if (p_70227_1_.func_216346_c() == RayTraceResult.Type.ENTITY) {
             Entity entity = ((EntityRayTraceResult)p_70227_1_).func_216348_a();
+            // Spigot start
+            boolean didDamage = false;
             if (this.field_70235_a != null) {
-               if (entity.func_70097_a(DamageSource.func_76358_a(this.field_70235_a), 8.0F)) {
+               didDamage = entity.func_70097_a(DamageSource.func_76356_a(this, field_70235_a), 8.0F);
+               if (didDamage) { // CraftBukkit
                   if (entity.func_70089_S()) {
                      this.func_174815_a(this.field_70235_a, entity);
                   } else {
-                     this.field_70235_a.func_70691_i(5.0F);
+                     this.field_70235_a.heal(5.0F, RegainReason.WITHER); // CraftBukkit
                   }
                }
             } else {
-               entity.func_70097_a(DamageSource.field_76376_m, 5.0F);
+               didDamage = entity.func_70097_a(DamageSource.field_76376_m, 5.0F);
             }
 
-            if (entity instanceof LivingEntity) {
+            if (didDamage && entity instanceof LivingEntity) {
+               // Spigot end
                int i = 0;
                if (this.field_70170_p.func_175659_aa() == Difficulty.NORMAL) {
                   i = 10;
@@ -77,13 +82,20 @@
                }
 
                if (i > 0) {
-                  ((LivingEntity)entity).func_195064_c(new EffectInstance(Effects.field_82731_v, 20 * i, 1));
+                  ((LivingEntity)entity).addPotionEffect(new EffectInstance(Effects.field_82731_v, 20 * i, 1), Cause.ATTACK); // CraftBukkit
                }
             }
          }
 
-         Explosion.Mode explosion$mode = this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223599_b) ? Explosion.Mode.DESTROY : Explosion.Mode.NONE;
-         this.field_70170_p.func_217398_a(this, this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), 1.0F, false, explosion$mode);
+         Explosion.Mode explosion$mode = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a) ? Explosion.Mode.DESTROY : Explosion.Mode.NONE;
+
+         // CraftBukkit start
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 1.0F, false);
+         this.field_70170_p.getServerCB().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            this.field_70170_p.func_217398_a(this, this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), event.getRadius(), event.getFire(), explosion$mode);
+         }
+         // CraftBukkit end
          this.func_70106_y();
       }
 
