--- a/net/minecraft/entity/player/ServerPlayerEntity.java
+++ b/net/minecraft/entity/player/ServerPlayerEntity.java
@@ -3,7 +3,9 @@
 import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.EnumSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.OptionalInt;
@@ -13,10 +15,13 @@
 import net.minecraft.advancements.PlayerAdvancements;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
+import net.minecraft.block.ChestBlock;
+import net.minecraft.block.pattern.BlockPattern;
 import net.minecraft.command.arguments.EntityAnchorArgument;
 import net.minecraft.crash.CrashReport;
 import net.minecraft.crash.CrashReportCategory;
 import net.minecraft.crash.ReportedException;
+import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.item.ItemEntity;
@@ -75,6 +80,7 @@
 import net.minecraft.scoreboard.Score;
 import net.minecraft.scoreboard.ScoreCriteria;
 import net.minecraft.scoreboard.ScorePlayerTeam;
+import net.minecraft.scoreboard.Scoreboard;
 import net.minecraft.scoreboard.Team;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.management.PlayerInteractionManager;
@@ -85,9 +91,11 @@
 import net.minecraft.tileentity.CommandBlockTileEntity;
 import net.minecraft.tileentity.SignTileEntity;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.CombatTracker;
 import net.minecraft.util.CooldownTracker;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.EntityDamageSource;
+import net.minecraft.util.FoodStats;
 import net.minecraft.util.Hand;
 import net.minecraft.util.HandSide;
 import net.minecraft.util.NonNullList;
@@ -110,19 +118,39 @@
 import net.minecraft.util.text.event.HoverEvent;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.GameType;
+import net.minecraft.world.World;
 import net.minecraft.world.dimension.DimensionType;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.storage.WorldInfo;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.util.BlockStateListPopulator;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.event.entity.PlayerDeathEvent;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.event.player.PlayerChangedMainHandEvent;
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerLocaleChangeEvent;
+import org.bukkit.event.player.PlayerPortalEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+import org.bukkit.event.world.PortalCreateEvent;
+import org.bukkit.inventory.MainHand;
 
 public class ServerPlayerEntity extends PlayerEntity implements IContainerListener {
    private static final Logger field_147102_bM = LogManager.getLogger();
-   private String field_71148_cg = "en_US";
+   public String field_71148_cg = "en_us"; // CraftBukkit - lowercase
    public ServerPlayNetHandler field_71135_a;
    public final MinecraftServer field_71133_b;
    public final PlayerInteractionManager field_71134_c;
-   private final List<Integer> field_71130_g = Lists.newLinkedList();
+   public final List<Integer> field_71130_g = Lists.newLinkedList();
    private final PlayerAdvancements field_192042_bX;
    private final ServerStatisticsManager field_147103_bO;
    private float field_130068_bO = Float.MIN_VALUE;
@@ -134,8 +162,8 @@
    private float field_71149_ch = -1.0E8F;
    private int field_71146_ci = -99999999;
    private boolean field_71147_cj = true;
-   private int field_71144_ck = -99999999;
-   private int field_147101_bU = 60;
+   public int field_71144_ck = -99999999;
+   public int field_147101_bU = 60;
    private ChatVisibility field_71143_cn;
    private boolean field_71140_co = true;
    private long field_143005_bX = Util.func_211177_b();
@@ -154,18 +182,67 @@
    public int field_71138_i;
    public boolean field_71136_j;
 
+   // CraftBukkit start
+   public String displayName;
+   public ITextComponent listName;
+   public org.bukkit.Location compassTarget;
+   public int newExp = 0;
+   public int newLevel = 0;
+   public int newTotalExp = 0;
+   public boolean keepLevel = false;
+   public double maxHealthCache;
+   public boolean joining = true;
+   public boolean sentListPacket = false;
+   public Integer clientViewDistance;
+   public long timeOffset;
+   public boolean relativeTime;
+   public WeatherType weather;
+   private float pluginRainPosition;
+   private float pluginRainPositionPrevious;
+   // CraftBukkit end
+
    public ServerPlayerEntity(MinecraftServer p_i45285_1_, ServerWorld p_i45285_2_, GameProfile p_i45285_3_, PlayerInteractionManager p_i45285_4_) {
       super(p_i45285_2_, p_i45285_3_);
       p_i45285_4_.field_73090_b = this;
       this.field_71134_c = p_i45285_4_;
       this.field_71133_b = p_i45285_1_;
       this.field_192041_cq = new ServerRecipeBook(p_i45285_1_.func_199529_aN());
-      this.field_147103_bO = p_i45285_1_.func_184103_al().func_152602_a(this);
+      this.field_147103_bO = p_i45285_1_.func_184103_al().getPlayerStats(this);
       this.field_192042_bX = p_i45285_1_.func_184103_al().func_192054_h(this);
       this.field_70138_W = 1.0F;
       this.func_205734_a(p_i45285_2_);
+      this.displayName = this.func_195047_I_();
+      this.canPickUpLoot = true;
+      this.maxHealthCache = this.func_110138_aP();
    }
 
+   public final BlockPos getSpawnPoint(final ServerWorld worldserver) {
+      BlockPos blockposition = worldserver.func_175694_M();
+      if (worldserver.field_73011_w.func_177495_o() && worldserver.func_72912_H().func_76077_q() != GameType.ADVENTURE) {
+         int i = Math.max(0, this.field_71133_b.func_184108_a(worldserver));
+         int j = MathHelper.func_76128_c(worldserver.func_175723_af().func_177729_b(blockposition.func_177958_n(), blockposition.func_177952_p()));
+         if (j < i) {
+            i = j;
+         }
+         if (j <= 1) {
+            i = 1;
+         }
+         int k = (i * 2 + 1) * (i * 2 + 1);
+         int l = this.func_205735_q(k);
+         int i2 = new Random().nextInt(k);
+         for (int j2 = 0; j2 < k; ++j2) {
+            int k2 = (i2 + l * j2) % k;
+            int l2 = k2 % (i * 2 + 1);
+            int i3 = k2 / (i * 2 + 1);
+            BlockPos blockposition2 = worldserver.func_201675_m().func_206921_a(blockposition.func_177958_n() + l2 - i, blockposition.func_177952_p() + i3 - i, false);
+            if (blockposition2 != null) {
+               return blockposition2;
+            }
+         }
+      }
+      return blockposition;
+   }
+
    private void func_205734_a(ServerWorld p_205734_1_) {
       BlockPos blockpos = p_205734_1_.func_175694_M();
       if (p_205734_1_.field_73011_w.func_191066_m() && p_205734_1_.func_72912_H().func_76077_q() != GameType.ADVENTURE) {
@@ -230,6 +307,7 @@
       if (p_70037_1_.func_150297_b("recipeBook", 10)) {
          this.field_192041_cq.func_192825_a(p_70037_1_.func_74775_l("recipeBook"));
       }
+      this.getBukkitEntity().readExtraData(p_70037_1_); // CraftBukkit
 
       if (this.func_70608_bn()) {
          this.func_213366_dy();
@@ -251,7 +329,16 @@
 
       Entity entity1 = this.func_184208_bv();
       Entity entity = this.func_184187_bx();
-      if (entity != null && entity1 != this && entity1.func_200601_bK()) {
+      boolean persistVehicle = true;
+      if (entity != null) {
+         for (Entity vehicle = entity; vehicle != null; vehicle = vehicle.func_184187_bx()) {
+            if (!vehicle.persist) {
+               persistVehicle = false;
+               break;
+            }
+         }
+      }
+      if (persistVehicle && entity != null && entity1 != this && entity1.func_200601_bK()) {
          CompoundNBT compoundnbt1 = new CompoundNBT();
          CompoundNBT compoundnbt2 = new CompoundNBT();
          entity1.func_70039_c(compoundnbt2);
@@ -263,6 +350,31 @@
       p_213281_1_.func_218657_a("recipeBook", this.field_192041_cq.func_192824_e());
    }
 
+   // CraftBukkit start - World fallback code, either respawn location or global spawn
+   public void func_70029_a(World world) {
+      super.func_70029_a(world);
+      if (world == null) {
+         this.field_70128_L = false;
+         Vec3d position = null;
+         if (this.spawnWorld != null && !this.spawnWorld.equals("")) {
+            final CraftWorld cworld = (CraftWorld) Bukkit.getServer().getWorld(this.spawnWorld);
+            if (cworld != null && this.func_180470_cg() != null) {
+               world = cworld.getHandle();
+               position = PlayerEntity.func_213822_a(cworld.getHandle(), this.func_180470_cg(), false).orElse(null);
+            }
+         }
+         if (world == null || position == null) {
+            world = ((CraftWorld) Bukkit.getServer().getWorlds().get(0)).getHandle();
+            position = new Vec3d(world.func_175694_M());
+         }
+         this.field_70170_p = world;
+         this.func_70107_b(position.func_82615_a(), position.func_82617_b(), position.func_82616_c());
+      }
+      this.field_71093_bK = ((ServerWorld)this.field_70170_p).func_201675_m().func_186058_p();
+      this.field_71134_c.func_73080_a((ServerWorld)world);
+   }
+   // CraftBukkit end
+
    public void func_195394_a(int p_195394_1_) {
       float f = (float)this.func_71050_bK();
       float f1 = (f - 1.0F) / f;
@@ -308,6 +420,11 @@
    }
 
    public void func_70071_h_() {
+      // CraftBukkit start
+      if (this.joining) {
+         this.joining = false;
+      }
+      // CraftBukkit end
       this.field_71134_c.func_73075_a();
       --this.field_147101_bU;
       if (this.field_70172_ad > 0) {
@@ -372,7 +489,7 @@
          }
 
          if (this.func_110143_aJ() != this.field_71149_ch || this.field_71146_ci != this.field_71100_bB.func_75116_a() || this.field_71100_bB.func_75115_e() == 0.0F != this.field_71147_cj) {
-            this.field_71135_a.func_147359_a(new SUpdateHealthPacket(this.func_110143_aJ(), this.field_71100_bB.func_75116_a(), this.field_71100_bB.func_75115_e()));
+            this.field_71135_a.func_147359_a(new SUpdateHealthPacket(this.getBukkitEntity().getScaledHealth(), this.field_71100_bB.func_75116_a(), this.field_71100_bB.func_75115_e()));
             this.field_71149_ch = this.func_110143_aJ();
             this.field_71146_ci = this.field_71100_bB.func_75116_a();
             this.field_71147_cj = this.field_71100_bB.func_75115_e() == 0.0F;
@@ -402,7 +519,11 @@
             this.field_184856_bZ = this.field_71067_cb;
             this.func_184849_a(ScoreCriteria.field_186701_k, MathHelper.func_76123_f((float)this.field_184856_bZ));
          }
-
+         // CraftBukkit start - Force max health updates
+         if (this.maxHealthCache != this.func_110138_aP()) {
+            this.getBukkitEntity().updateScaledHealth();
+         }
+         // CraftBukkit end
          if (this.field_71068_ca != this.field_184855_bY) {
             this.field_184855_bY = this.field_71068_ca;
             this.func_184849_a(ScoreCriteria.field_186702_l, MathHelper.func_76123_f((float)this.field_184855_bY));
@@ -416,7 +537,15 @@
          if (this.field_70173_aa % 20 == 0) {
             CriteriaTriggers.field_192135_o.func_192215_a(this);
          }
-
+         // CraftBukkit start - initialize oldLevel and fire PlayerLevelChangeEvent
+         if (this.oldLevel == -1) {
+            this.oldLevel = this.field_71068_ca;
+         }
+         if (this.oldLevel != this.field_71068_ca) {
+            CraftEventFactory.callPlayerLevelChangeEvent(this.field_70170_p.getServerCB().getPlayer(this), this.oldLevel, this.field_71068_ca);
+            this.oldLevel = this.field_71068_ca;
+         }
+         // CraftBukkit end
       } catch (Throwable throwable) {
          CrashReport crashreport = CrashReport.func_85055_a(throwable, "Ticking player");
          CrashReportCategory crashreportcategory = crashreport.func_85058_a("Player being ticked");
@@ -426,15 +555,46 @@
    }
 
    private void func_184849_a(ScoreCriteria p_184849_1_, int p_184849_2_) {
-      this.func_96123_co().func_197893_a(p_184849_1_, this.func_195047_I_(), (p_195397_1_) -> {
+      this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(p_184849_1_, this.func_195047_I_(), (p_195397_1_) -> {
          p_195397_1_.func_96647_c(p_184849_2_);
       });
    }
 
    public void func_70645_a(DamageSource p_70645_1_) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_70645_1_)) return;
       boolean flag = this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223609_l);
-      if (flag) {
-         ITextComponent itextcomponent = this.func_110142_aN().func_151521_b();
+      if (this.field_70128_L) {
+         return;
+      }
+      List<org.bukkit.inventory.ItemStack> loot = new ArrayList<org.bukkit.inventory.ItemStack>(this.field_71071_by.func_70302_i_());
+       boolean keepInventory = this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223600_c) || this.func_175149_v();
+      if (!keepInventory) {
+         for (final ItemStack item : this.field_71071_by.getContents()) {
+            if (!item.func_190926_b() && !EnchantmentHelper.func_190939_c(item)) {
+               loot.add(CraftItemStack.asCraftMirror(item));
+            }
+         }
+      }
+      this.func_213354_a(p_70645_1_, this.field_70718_bc > 0);
+      for (final org.bukkit.inventory.ItemStack item2 : this.drops) {
+         loot.add(item2);
+      }
+      this.drops.clear();
+      ITextComponent defaultMessage = this.func_110142_aN().func_151521_b();
+      String deathmessage = defaultMessage.getString();
+      PlayerDeathEvent event = CraftEventFactory.callPlayerDeathEvent(this, loot, deathmessage, keepInventory);
+      if (this.field_71070_bA != this.field_71069_bz) {
+         this.func_71053_j();
+      }
+      String deathMessage = event.getDeathMessage();
+      if (deathMessage != null && deathMessage.length() > 0 && flag) {
+         ITextComponent itextcomponent;
+         if (deathMessage.equals(deathmessage)) {
+            itextcomponent = this.func_110142_aN().func_151521_b();
+         }
+         else {
+            itextcomponent = CraftChatMessage.fromStringOrNull(deathMessage);
+         }
          this.field_71135_a.func_211148_a(new SCombatPacket(this.func_110142_aN(), SCombatPacket.Event.ENTITY_DIED, itextcomponent), (p_212356_2_) -> {
             if (!p_212356_2_.isSuccess()) {
                int i = 256;
@@ -462,11 +622,12 @@
       }
 
       this.func_192030_dh();
-      if (!this.func_175149_v()) {
-         this.func_213345_d(p_70645_1_);
+      this.func_226294_cV_();
+      if (!event.getKeepInventory()) {
+         this.field_71071_by.func_174888_l();
       }
-
-      this.func_96123_co().func_197893_a(ScoreCriteria.field_96642_c, this.func_195047_I_(), Score::func_96648_a);
+      this.func_175399_e(this);
+      this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(ScoreCriteria.field_96642_c, this.func_195047_I_(), Score::func_96648_a);
       LivingEntity livingentity = this.func_94060_bK();
       if (livingentity != null) {
          this.func_71029_a(Stats.field_199091_i.func_199076_b(livingentity.func_200600_R()));
@@ -489,10 +650,10 @@
          this.func_85039_t(p_191956_2_);
          String s = this.func_195047_I_();
          String s1 = p_191956_1_.func_195047_I_();
-         this.func_96123_co().func_197893_a(ScoreCriteria.field_96640_e, s, Score::func_96648_a);
+         this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(ScoreCriteria.field_96640_e, s, Score::func_96648_a);
          if (p_191956_1_ instanceof PlayerEntity) {
             this.func_195066_a(Stats.field_75932_A);
-            this.func_96123_co().func_197893_a(ScoreCriteria.field_96639_d, s, Score::func_96648_a);
+            this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(ScoreCriteria.field_96639_d, s, Score::func_96648_a);
          } else {
             this.func_195066_a(Stats.field_188070_B);
          }
@@ -508,7 +669,7 @@
       if (scoreplayerteam != null) {
          int i = scoreplayerteam.func_178775_l().func_175746_b();
          if (i >= 0 && i < p_195398_3_.length) {
-            this.func_96123_co().func_197893_a(p_195398_3_[i], p_195398_1_, Score::func_96648_a);
+            this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(p_195398_3_[i], p_195398_1_, Score::func_96648_a);
          }
       }
 
@@ -547,16 +708,32 @@
    }
 
    private boolean func_175400_cq() {
-      return this.field_71133_b.func_71219_W();
+      return this.field_70170_p.pvpMode;
    }
 
+   @Override
    @Nullable
-   public Entity func_212321_a(DimensionType p_212321_1_) {
-      this.field_184851_cj = true;
+   public Entity changeDimension(DimensionType p_212321_1_, net.minecraftforge.common.util.ITeleporter teleporter) {
+      return changeDimension(p_212321_1_, teleporter, PlayerTeleportEvent.TeleportCause.UNKNOWN);
+   }
+
+   @Nullable
+   public Entity changeDimension(DimensionType destination, PlayerTeleportEvent.TeleportCause cause) {
+      return changeDimension(destination, func_184102_h().func_71218_a(destination).func_85176_s(), cause);
+   }
+
+   @Nullable
+   public Entity changeDimension(DimensionType destination, net.minecraftforge.common.util.ITeleporter teleporter, PlayerTeleportEvent.TeleportCause cause) {
+      if (this.func_70608_bn()) {
+         return this;
+      }
+      if (!net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, destination)) return null;
+      // this.invulnerableDimensionChange = true;
       DimensionType dimensiontype = this.field_71093_bK;
-      if (dimensiontype == DimensionType.field_223229_c_ && p_212321_1_ == DimensionType.field_223227_a_) {
+      if (dimensiontype.getType() == DimensionType.field_223229_c_ && destination.getType() == DimensionType.field_223227_a_ && teleporter instanceof net.minecraft.world.Teleporter) { //Forge: Fix non-vanilla teleporters triggering end credits
+         this.field_184851_cj = true;
          this.func_213319_R();
-         this.func_71121_q().func_217434_e(this);
+         this.func_71121_q().removePlayer(this, true); //Forge: The player entity is cloned so keep the data until after cloning calls copyFrom
          if (!this.field_71136_j) {
             this.field_71136_j = true;
             this.field_71135_a.func_147359_a(new SChangeGameStatePacket(4, this.field_192040_cp ? 0.0F : 1.0F));
@@ -566,15 +743,19 @@
          return this;
       } else {
          ServerWorld serverworld = this.field_71133_b.func_71218_a(dimensiontype);
-         this.field_71093_bK = p_212321_1_;
-         ServerWorld serverworld1 = this.field_71133_b.func_71218_a(p_212321_1_);
-         WorldInfo worldinfo = serverworld1.func_72912_H();
-         this.field_71135_a.func_147359_a(new SRespawnPacket(p_212321_1_, WorldInfo.func_227498_c_(worldinfo.func_76063_b()), worldinfo.func_76067_t(), this.field_71134_c.func_73081_b()));
-         this.field_71135_a.func_147359_a(new SServerDifficultyPacket(worldinfo.func_176130_y(), worldinfo.func_176123_z()));
-         PlayerList playerlist = this.field_71133_b.func_184103_al();
-         playerlist.func_187243_f(this);
-         serverworld.func_217434_e(this);
-         this.field_70128_L = false;
+         // this.dimension = destination;
+         ServerWorld serverworld1 = this.field_71133_b.func_71218_a(destination);
+         /*
+         WorldInfo worldinfo = serverworld1.getWorldInfo();
+         net.minecraftforge.fml.network.NetworkHooks.sendDimensionDataPacket(this.connection.netManager, this);
+         this.connection.sendPacket(new SRespawnPacket(destination, WorldInfo.byHashing(worldinfo.getSeed()), worldinfo.getGenerator(), this.interactionManager.getGameType()));
+         this.connection.sendPacket(new SServerDifficultyPacket(worldinfo.getDifficulty(), worldinfo.isDifficultyLocked()));
+         PlayerList playerlist = this.server.getPlayerList();
+         playerlist.updatePermissionLevel(this);
+         serverworld.removeEntity(this, true); //Forge: the player entity is moved to the new world, NOT cloned. So keep the data alive with no matching invalidate call.
+         this.revive();
+         Entity e = teleporter.placeEntity(this, serverworld, serverworld1, this.rotationYaw, spawnPortal -> {//Forge: Start vanilla logic
+         */
          double d0 = this.func_226277_ct_();
          double d1 = this.func_226278_cu_();
          double d2 = this.func_226281_cx_();
@@ -583,23 +764,37 @@
          double d3 = 8.0D;
          float f2 = f1;
          serverworld.func_217381_Z().func_76320_a("moving");
-         if (dimensiontype == DimensionType.field_223227_a_ && p_212321_1_ == DimensionType.field_223228_b_) {
-            this.field_193110_cw = this.func_213303_ch();
-            d0 /= 8.0D;
-            d2 /= 8.0D;
-         } else if (dimensiontype == DimensionType.field_223228_b_ && p_212321_1_ == DimensionType.field_223227_a_) {
-            d0 *= 8.0D;
-            d2 *= 8.0D;
-         } else if (dimensiontype == DimensionType.field_223227_a_ && p_212321_1_ == DimensionType.field_223229_c_) {
-            BlockPos blockpos = serverworld1.func_180504_m();
-            d0 = (double)blockpos.func_177958_n();
-            d1 = (double)blockpos.func_177956_o();
-            d2 = (double)blockpos.func_177952_p();
-            f1 = 90.0F;
-            f = 0.0F;
-         }
+         double moveFactor = serverworld.func_201675_m().getMovementFactor() / serverworld1.func_201675_m().getMovementFactor();
+         d0 *= moveFactor;
+         d2 *= moveFactor;
+         if (serverworld1 == null) {
+         } else // CraftBukkit - empty to fall through to null to event
+            if (dimensiontype == DimensionType.field_223227_a_ && destination == DimensionType.field_223228_b_) {
+               this.field_193110_cw = this.func_213303_ch();
+            } else if (dimensiontype == DimensionType.field_223227_a_ && destination == DimensionType.field_223229_c_) {
+               BlockPos blockpos = serverworld1.func_180504_m();
+               d0 = (double) blockpos.func_177958_n();
+               d1 = (double) blockpos.func_177956_o();
+               d2 = (double) blockpos.func_177952_p();
+               f1 = 90.0F;
+               f = 0.0F;
+            }
 
-         this.func_70012_b(d0, d1, d2, f1, f);
+         Location enter = this.getBukkitEntity().getLocation();
+         Location exit = (serverworld1 == null) ? null : new Location(serverworld1.getWorldCB(), d0, d2, d3, f2, f);
+         final PlayerPortalEvent event = new PlayerPortalEvent(this.getBukkitEntity(), enter, exit, cause, 128, true, (destination.getType() == DimensionType.field_223229_c_) ? 0 : 16);
+         Bukkit.getServer().getPluginManager().callEvent(event);
+         if (event.isCancelled() || event.getTo() == null) {
+            return null;
+         }
+         exit = event.getTo();
+         if (exit == null) {
+            return null;
+         }
+         serverworld1 = ((CraftWorld) exit.getWorld()).getHandle();
+         d0 = exit.getX();
+         d2 = exit.getY();
+         d3 = exit.getZ();
          serverworld.func_217381_Z().func_76319_b();
          serverworld.func_217381_Z().func_76320_a("placing");
          double d7 = Math.min(-2.9999872E7D, serverworld1.func_175723_af().func_177726_b() + 16.0D);
@@ -608,44 +803,86 @@
          double d6 = Math.min(2.9999872E7D, serverworld1.func_175723_af().func_177733_e() - 16.0D);
          d0 = MathHelper.func_151237_a(d0, d7, d5);
          d2 = MathHelper.func_151237_a(d2, d4, d6);
-         this.func_70012_b(d0, d1, d2, f1, f);
-         if (p_212321_1_ == DimensionType.field_223229_c_) {
-            int i = MathHelper.func_76128_c(this.func_226277_ct_());
-            int j = MathHelper.func_76128_c(this.func_226278_cu_()) - 1;
-            int k = MathHelper.func_76128_c(this.func_226281_cx_());
-            int l = 1;
-            int i1 = 0;
-
-            for(int j1 = -2; j1 <= 2; ++j1) {
-               for(int k1 = -2; k1 <= 2; ++k1) {
-                  for(int l1 = -1; l1 < 3; ++l1) {
-                     int i2 = i + k1 * 1 + j1 * 0;
-                     int j2 = j + l1;
-                     int k2 = k + k1 * 0 - j1 * 1;
-                     boolean flag = l1 < 0;
-                     serverworld1.func_175656_a(new BlockPos(i2, j2, k2), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
+         Vec3d exitVelocity = Vec3d.field_186680_a;
+         final BlockPos exitPosition = new BlockPos(d0, d2, d3);
+         if (destination.getType() == DimensionType.field_223229_c_) {
+            int i = exitPosition.func_177958_n();
+            int j = exitPosition.func_177956_o() - 1;
+            int k = exitPosition.func_177952_p();
+            if (event.getCanCreatePortal()) {
+               int l = 1;
+               int i1 = 0;
+               BlockStateListPopulator blockList = new BlockStateListPopulator(serverworld1);
+               for (int j1 = -2; j1 <= 2; ++j1) {
+                  for (int k1 = -2; k1 <= 2; ++k1) {
+                     for (int l1 = -1; l1 < 3; ++l1) {
+                        int i2 = i + k1 * 1 + j1 * 0;
+                        int j2 = j + l1;
+                        int k2 = k + k1 * 0 - j1 * 1;
+                        boolean flag = l1 < 0;
+                        blockList.func_180501_a(new BlockPos(i2, j2, k2), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P(), 3);
+                     }
                   }
                }
-            }
 
-            this.func_70012_b((double)i, (double)j, (double)k, f1, 0.0F);
-            this.func_213317_d(Vec3d.field_186680_a);
-         } else if (!serverworld1.func_85176_s().func_222268_a(this, f2)) {
-            serverworld1.func_85176_s().func_85188_a(this);
-            serverworld1.func_85176_s().func_222268_a(this, f2);
+               org.bukkit.World bworld = serverworld1.getWorldCB();
+               PortalCreateEvent portalEvent = new PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), bworld, this.getBukkitEntity(), PortalCreateEvent.CreateReason.END_PLATFORM);
+               this.field_70170_p.getServerCB().getPluginManager().callEvent(portalEvent);
+               if (!portalEvent.isCancelled()) {
+                  blockList.updateList();
+               }
+            }
+            exit.setX(i);
+            exit.setY(j);
+            exit.setZ(k);
+            exitVelocity = Vec3d.field_186680_a;
+         } else {
+            BlockPattern.PortalInfo portalShape = serverworld1.func_85176_s().placeInPortal(this, exitPosition, f2, event.getSearchRadius(), true);
+            if (portalShape == null && event.getCanCreatePortal() && serverworld1.func_85176_s().makePortal(this, exitPosition, event.getCreationRadius())) {
+               portalShape = serverworld1.func_85176_s().placeInPortal(this, exitPosition, f2, event.getSearchRadius(), true);
+            }
+            if (portalShape == null) {
+               return null;
+            }
+            exitVelocity = portalShape.field_222506_b;
+            exit.setX(portalShape.field_222505_a.func_82615_a());
+            exit.setY(portalShape.field_222505_a.func_82617_b());
+            exit.setZ(portalShape.field_222505_a.func_82616_c());
+            exit.setYaw(f2 + portalShape.field_222507_c);
          }
 
          serverworld.func_217381_Z().func_76319_b();
+         PlayerTeleportEvent tpEvent = new PlayerTeleportEvent(this.getBukkitEntity(), enter, exit, cause);
+         Bukkit.getServer().getPluginManager().callEvent(tpEvent);
+         if (tpEvent.isCancelled() || tpEvent.getTo() == null) {
+            return null;
+         }
+         exit = tpEvent.getTo();
+         if (exit == null) {
+            return null;
+         }
+         serverworld = ((CraftWorld) exit.getWorld()).getHandle();
+         this.field_184851_cj = true;
+         destination = serverworld1.func_201675_m().func_186058_p();
+         this.field_71093_bK = destination;
+         this.field_71135_a.func_147359_a(new SRespawnPacket(serverworld1.field_73011_w.func_186058_p().getType(), WorldInfo.func_227498_c_(this.field_70170_p.func_72912_H().func_76063_b()), this.field_70170_p.func_72912_H().func_76067_t(), this.field_71134_c.func_73081_b()));
+         this.field_71135_a.func_147359_a(new SServerDifficultyPacket(this.field_70170_p.func_175659_aa(), this.field_70170_p.func_72912_H().func_176123_z()));
+         final PlayerList playerlist = this.field_71133_b.func_184103_al();
+         playerlist.func_187243_f(this);
+         serverworld.func_217434_e(this);
+         this.field_70128_L = false;
+         this.func_213317_d(exitVelocity);
          this.func_70029_a(serverworld1);
          serverworld1.func_217447_b(this);
          this.func_213846_b(serverworld);
-         this.field_71135_a.func_147364_a(this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_(), f1, f);
+         this.field_71135_a.teleport(exit);
+         this.field_71135_a.func_184342_d();
          this.field_71134_c.func_73080_a(serverworld1);
          this.field_71135_a.func_147359_a(new SPlayerAbilitiesPacket(this.field_71075_bZ));
          playerlist.func_72354_b(this, serverworld1);
          playerlist.func_72385_f(this);
 
-         for(EffectInstance effectinstance : this.func_70651_bq()) {
+         for (EffectInstance effectinstance : this.func_70651_bq()) {
             this.field_71135_a.func_147359_a(new SPlayEntityEffectPacket(this.func_145782_y(), effectinstance));
          }
 
@@ -653,11 +890,14 @@
          this.field_71144_ck = -1;
          this.field_71149_ch = -1.0F;
          this.field_71146_ci = -1;
+         net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerChangedDimensionEvent(this, dimensiontype, destination);
+         PlayerChangedWorldEvent changeEvent = new PlayerChangedWorldEvent(this.getBukkitEntity(), serverworld.getWorldCB());
+         this.field_70170_p.getServerCB().getPluginManager().callEvent(changeEvent);
          return this;
       }
    }
 
-   private void func_213846_b(ServerWorld p_213846_1_) {
+   public void func_213846_b(ServerWorld p_213846_1_) {
       DimensionType dimensiontype = p_213846_1_.field_73011_w.func_186058_p();
       DimensionType dimensiontype1 = this.field_70170_p.field_73011_w.func_186058_p();
       CriteriaTriggers.field_193134_u.func_193143_a(this, dimensiontype, dimensiontype1);
@@ -695,13 +935,21 @@
    }
 
    public Either<PlayerEntity.SleepResult, Unit> func_213819_a(BlockPos p_213819_1_) {
-      return super.func_213819_a(p_213819_1_).ifRight((p_213849_1_) -> {
+      // CraftBukkit start - add force parameter
+      return this.sleep(p_213819_1_, false);
+   }
+
+   @Override
+   public Either<PlayerEntity.SleepResult, Unit> sleep(BlockPos blockposition, boolean force) {
+      // CraftBukkit end
+      return super.sleep(blockposition, force).ifRight((p_213849_1_) -> {
          this.func_195066_a(Stats.field_188064_ad);
          CriteriaTriggers.field_192136_p.func_192215_a(this);
       });
    }
 
    public void func_225652_a_(boolean p_225652_1_, boolean p_225652_2_) {
+      if (!this.func_70608_bn()) return; // CraftBukkit - Can't leave bed if not in one!
       if (this.func_70608_bn()) {
          this.func_71121_q().func_72863_F().func_217216_a(this, new SAnimateHandPacket(this, 2));
       }
@@ -764,8 +1012,9 @@
       this.field_71135_a.func_147359_a(new SOpenSignMenuPacket(p_175141_1_.func_174877_v()));
    }
 
-   public void func_71117_bO() {
+   public int nextContainerCounter() { // CraftBukkit - void -> int
       this.field_71139_cq = this.field_71139_cq % 100 + 1;
+      return field_71139_cq; // CraftBukkit
    }
 
    public OptionalInt func_213829_a(@Nullable INamedContainerProvider p_213829_1_) {
@@ -776,8 +1025,22 @@
             this.func_71053_j();
          }
 
-         this.func_71117_bO();
+         this.nextContainerCounter();
          Container container = p_213829_1_.createMenu(this.field_71139_cq, this.field_71071_by, this);
+         if (container != null) {
+            container.setTitle(p_213829_1_.func_145748_c_());
+            final boolean cancelled = false;
+            container = CraftEventFactory.callInventoryOpenEvent(this, container, cancelled);
+            if (container == null && !cancelled) {
+               if (p_213829_1_ instanceof IInventory) {
+                  ((IInventory)p_213829_1_).func_174886_c(this);
+               }
+               else if (p_213829_1_ instanceof ChestBlock.DoubleInventory) {
+                  ((ChestBlock.DoubleInventory)p_213829_1_).inventorylargechest.func_174886_c(this);
+               }
+               return OptionalInt.empty();
+            }
+         }
          if (container == null) {
             if (this.func_175149_v()) {
                this.func_146105_b((new TranslationTextComponent("container.spectatorCantOpen")).func_211708_a(TextFormatting.RED), true);
@@ -785,9 +1048,10 @@
 
             return OptionalInt.empty();
          } else {
-            this.field_71135_a.func_147359_a(new SOpenWindowPacket(container.field_75152_c, container.func_216957_a(), p_213829_1_.func_145748_c_()));
+            this.field_71135_a.func_147359_a(new SOpenWindowPacket(container.field_75152_c, container.func_216957_a(), container.getTitle()));
             container.func_75132_a(this);
             this.field_71070_bA = container;
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
             return OptionalInt.of(this.field_71139_cq);
          }
       }
@@ -798,14 +1062,26 @@
    }
 
    public void func_184826_a(AbstractHorseEntity p_184826_1_, IInventory p_184826_2_) {
+      // CraftBukkit start - Inventory open hook
+      this.nextContainerCounter();
+      Container container = new HorseInventoryContainer(this.field_71139_cq, this.field_71071_by, p_184826_2_, p_184826_1_);
+      container.setTitle(p_184826_1_.func_145748_c_());
+      container = CraftEventFactory.callInventoryOpenEvent(this, container);
+      // CraftBukkit start - Inventory open hook
+      if (container == null) {
+         p_184826_2_.func_174886_c(this);
+         return;
+      }
+      // CraftBukkit end
       if (this.field_71070_bA != this.field_71069_bz) {
          this.func_71053_j();
       }
 
-      this.func_71117_bO();
+      // this.getNextWindowId();
       this.field_71135_a.func_147359_a(new SOpenHorseWindowPacket(this.field_71139_cq, p_184826_2_.func_70302_i_(), p_184826_1_.func_145782_y()));
-      this.field_71070_bA = new HorseInventoryContainer(this.field_71139_cq, this.field_71071_by, p_184826_2_, p_184826_1_);
+      this.field_71070_bA = container;
       this.field_71070_bA.func_75132_a(this);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
    }
 
    public void func_184814_a(ItemStack p_184814_1_, Hand p_184814_2_) {
@@ -844,6 +1120,11 @@
    public void func_71110_a(Container p_71110_1_, NonNullList<ItemStack> p_71110_2_) {
       this.field_71135_a.func_147359_a(new SWindowItemsPacket(p_71110_1_.field_75152_c, p_71110_2_));
       this.field_71135_a.func_147359_a(new SSetSlotPacket(-1, -1, this.field_71071_by.func_70445_o()));
+      // CraftBukkit start - Send a Set Slot to update the crafting result slot
+      if (EnumSet.of(InventoryType.CRAFTING, InventoryType.WORKBENCH).contains(field_71069_bz.getBukkitView().getType())) {
+         this.field_71135_a.func_147359_a(new SSetSlotPacket(field_71069_bz.field_75152_c, 0, field_71069_bz.func_75139_a(0).func_75211_c()));
+      }
+      // CraftBukkit end
    }
 
    public void func_71112_a(Container p_71112_1_, int p_71112_2_, int p_71112_3_) {
@@ -851,6 +1132,7 @@
    }
 
    public void func_71053_j() {
+      CraftEventFactory.handleInventoryCloseEvent(this); // CraftBukkit
       this.field_71135_a.func_147359_a(new SCloseWindowPacket(this.field_71070_bA.field_75152_c));
       this.func_71128_l();
    }
@@ -863,6 +1145,7 @@
 
    public void func_71128_l() {
       this.field_71070_bA.func_75134_a(this);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Close(this, this.field_71070_bA));
       this.field_71070_bA = this.field_71069_bz;
    }
 
@@ -884,14 +1167,14 @@
 
    public void func_71064_a(Stat<?> p_71064_1_, int p_71064_2_) {
       this.field_147103_bO.func_150871_b(this, p_71064_1_, p_71064_2_);
-      this.func_96123_co().func_197893_a(p_71064_1_, this.func_195047_I_(), (p_195396_1_) -> {
+      this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(p_71064_1_, this.func_195047_I_(), (p_195396_1_) -> {
          p_195396_1_.func_96649_a(p_71064_2_);
       });
    }
 
    public void func_175145_a(Stat<?> p_175145_1_) {
       this.field_147103_bO.func_150873_a(this, p_175145_1_, 0);
-      this.func_96123_co().func_197893_a(p_175145_1_, this.func_195047_I_(), Score::func_197891_c);
+      this.field_70170_p.getServerCB().getScoreboardManager().getScoreboardScores(p_175145_1_, this.func_195047_I_(), Score::func_197891_c);
    }
 
    public int func_195065_a(Collection<IRecipe<?>> p_195065_1_) {
@@ -932,8 +1215,17 @@
 
    public void func_71118_n() {
       this.field_71149_ch = -1.0E8F;
+      this.field_71144_ck = -1; // CraftBukkit - Added to reset
    }
 
+   // CraftBukkit start - Support multi-line messages
+   public void sendMessage(final ITextComponent[] ichatbasecomponent) {
+      for (ITextComponent component : ichatbasecomponent) {
+         this.func_145747_a(component);
+      }
+   }
+   // CraftBukkit end
+
    public void func_146105_b(ITextComponent p_146105_1_, boolean p_146105_2_) {
       this.field_71135_a.func_147359_a(new SChatPacket(p_146105_1_, p_146105_2_ ? ChatType.GAME_INFO : ChatType.CHAT));
    }
@@ -983,12 +1275,27 @@
       this.field_71144_ck = -1;
       this.field_71149_ch = -1.0F;
       this.field_71146_ci = -1;
-      this.field_192041_cq.func_193824_a(p_193104_1_.field_192041_cq);
+      // this.recipeBook.copyFrom(that.recipeBook);
       this.field_71130_g.addAll(p_193104_1_.field_71130_g);
       this.field_192040_cp = p_193104_1_.field_192040_cp;
       this.field_193110_cw = p_193104_1_.field_193110_cw;
       this.func_192029_h(p_193104_1_.func_192023_dk());
       this.func_192031_i(p_193104_1_.func_192025_dl());
+
+      this.spawnPosMap = p_193104_1_.spawnPosMap;
+      this.spawnForcedMap = p_193104_1_.spawnForcedMap;
+      if(p_193104_1_.field_71093_bK != DimensionType.field_223227_a_) {
+          this.field_71077_c = p_193104_1_.field_71077_c;
+          this.field_82248_d = p_193104_1_.field_82248_d;
+      }
+      this.field_213329_S = false; // SPIGOT-4767
+
+      //Copy over a section of the Entity Data from the old player.
+      //Allows mods to specify data that persists after players respawn.
+      CompoundNBT old = p_193104_1_.getPersistentData();
+      if (old.func_74764_b(PERSISTED_NBT_TAG))
+          getPersistentData().func_218657_a(PERSISTED_NBT_TAG, old.func_74781_a(PERSISTED_NBT_TAG));
+      net.minecraftforge.event.ForgeEventFactory.onPlayerClone(this, p_193104_1_, !p_193104_2_);
    }
 
    protected void func_70670_a(EffectInstance p_70670_1_) {
@@ -1092,6 +1399,15 @@
    }
 
    public void func_147100_a(CClientSettingsPacket p_147100_1_) {
+      if (this.func_184591_cq() != p_147100_1_.func_186991_f()) {
+         PlayerChangedMainHandEvent event = new PlayerChangedMainHandEvent(this.getBukkitEntity(), (this.func_184591_cq() == HandSide.LEFT) ? MainHand.LEFT : MainHand.RIGHT);
+         this.field_71133_b.server.getPluginManager().callEvent(event);
+      }
+      if (!this.field_71148_cg.equals(p_147100_1_.func_149524_c())) {
+         PlayerLocaleChangeEvent event2 = new PlayerLocaleChangeEvent(this.getBukkitEntity(), p_147100_1_.func_149524_c());
+         this.field_71133_b.server.getPluginManager().callEvent(event2);
+      }
+      this.clientViewDistance = p_147100_1_.field_149528_b;
       this.field_71148_cg = p_147100_1_.func_149524_c();
       this.field_71143_cn = p_147100_1_.func_149523_e();
       this.field_71140_co = p_147100_1_.func_149520_f();
@@ -1155,7 +1471,7 @@
       this.field_175401_bS = (Entity)(p_175399_1_ == null ? this : p_175399_1_);
       if (entity != this.field_175401_bS) {
          this.field_71135_a.func_147359_a(new SCameraPacket(this.field_175401_bS));
-         this.func_70634_a(this.field_175401_bS.func_226277_ct_(), this.field_175401_bS.func_226278_cu_(), this.field_175401_bS.func_226281_cx_());
+         this.field_71135_a.setPlayerLocation(this.field_175401_bS.func_226277_ct_(), this.field_175401_bS.func_226278_cu_(), this.field_175401_bS.func_226281_cx_(), this.field_70177_z, this.field_70125_A, PlayerTeleportEvent.TeleportCause.SPECTATE);
       }
 
    }
@@ -1182,7 +1498,7 @@
 
    @Nullable
    public ITextComponent func_175396_E() {
-      return null;
+      return listName; // CraftBukkit
    }
 
    public void func_184609_a(Hand p_184609_1_) {
@@ -1203,29 +1519,38 @@
    }
 
    public void func_200619_a(ServerWorld p_200619_1_, double p_200619_2_, double p_200619_4_, double p_200619_6_, float p_200619_8_, float p_200619_9_) {
+      this.teleport(p_200619_1_, p_200619_2_, p_200619_4_, p_200619_2_, p_200619_8_, p_200619_9_, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.UNKNOWN);
+   }
+
+   public void teleport(ServerWorld newWorld, double x, double y, double z, float yaw, float pitch, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause) {
       this.func_175399_e(this);
       this.func_184210_p();
-      if (p_200619_1_ == this.field_70170_p) {
-         this.field_71135_a.func_147364_a(p_200619_2_, p_200619_4_, p_200619_6_, p_200619_8_, p_200619_9_);
-      } else {
-         ServerWorld serverworld = this.func_71121_q();
-         this.field_71093_bK = p_200619_1_.field_73011_w.func_186058_p();
-         WorldInfo worldinfo = p_200619_1_.func_72912_H();
-         this.field_71135_a.func_147359_a(new SRespawnPacket(this.field_71093_bK, WorldInfo.func_227498_c_(worldinfo.func_76063_b()), worldinfo.func_76067_t(), this.field_71134_c.func_73081_b()));
-         this.field_71135_a.func_147359_a(new SServerDifficultyPacket(worldinfo.func_176130_y(), worldinfo.func_176123_z()));
-         this.field_71133_b.func_184103_al().func_187243_f(this);
-         serverworld.func_217434_e(this);
-         this.field_70128_L = false;
-         this.func_70012_b(p_200619_2_, p_200619_4_, p_200619_6_, p_200619_8_, p_200619_9_);
-         this.func_70029_a(p_200619_1_);
-         p_200619_1_.func_217446_a(this);
+      /* CraftBukkit start - replace with bukkit handling for multi-world
+      if (newWorld == this.world) {
+         this.connection.setPlayerLocation(x, y, z, yaw, pitch);
+      } else if (net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, newWorld.dimension.getType())) {
+         DimensionType oldDimension = this.dimension;
+         ServerWorld serverworld = this.getServerWorld();
+         this.dimension = newWorld.dimension.getType();
+         WorldInfo worldinfo = newWorld.getWorldInfo();
+         net.minecraftforge.fml.network.NetworkHooks.sendDimensionDataPacket(this.connection.netManager, this);
+         this.connection.sendPacket(new SRespawnPacket(this.dimension, WorldInfo.byHashing(worldinfo.getSeed()), worldinfo.getGenerator(), this.interactionManager.getGameType()));
+         this.connection.sendPacket(new SServerDifficultyPacket(worldinfo.getDifficulty(), worldinfo.isDifficultyLocked()));
+         this.server.getPlayerList().updatePermissionLevel(this);
+         serverworld.removePlayer(this, true); //Forge: The player entity itself is moved, and not cloned. So we need to keep the data alive with no matching invalidate call later.
+         this.revive();
+         this.setLocationAndAngles(x, y, z, yaw, pitch);
+         this.setWorld(newWorld);
+         newWorld.addDuringCommandTeleport(this);
          this.func_213846_b(serverworld);
-         this.field_71135_a.func_147364_a(p_200619_2_, p_200619_4_, p_200619_6_, p_200619_8_, p_200619_9_);
-         this.field_71134_c.func_73080_a(p_200619_1_);
-         this.field_71133_b.func_184103_al().func_72354_b(this, p_200619_1_);
-         this.field_71133_b.func_184103_al().func_72385_f(this);
+         this.connection.setPlayerLocation(x, y, z, yaw, pitch);
+         this.interactionManager.setWorld(newWorld);
+         this.server.getPlayerList().sendWorldInfo(this, newWorld);
+         this.server.getPlayerList().sendInventory(this);
+         net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerChangedDimensionEvent(this, oldDimension, this.dimension);
       }
-
+      */
+      this.getBukkitEntity().teleport(new Location(newWorld.getWorldCB(), x, y, x, yaw, pitch), cause);
    }
 
    public void func_213844_a(ChunkPos p_213844_1_, IPacket<?> p_213844_2_, IPacket<?> p_213844_3_) {
@@ -1261,6 +1586,8 @@
       if (itementity == null) {
          return null;
       } else {
+         if (captureDrops() != null) captureDrops().add(itementity);
+         else
          this.field_70170_p.func_217376_c(itementity);
          ItemStack itemstack = itementity.func_92059_d();
          if (p_146097_3_) {
@@ -1274,4 +1601,126 @@
          return itementity;
       }
    }
+
+   @Override
+   public CraftPlayer getBukkitEntity() {
+      return (CraftPlayer) super.getBukkitEntity();
+   }
+
+    public long getPlayerTime() {
+       if (this.relativeTime) {
+          return this.field_70170_p.func_72820_D() + this.timeOffset;
+       }
+       return this.field_70170_p.func_72820_D() - this.field_70170_p.func_72820_D() % 24000L + this.timeOffset;
+    }
+
+   public WeatherType getPlayerWeather() {
+      return this.weather;
+   }
+
+   public void setPlayerWeather(final WeatherType type, final boolean plugin) {
+      if (!plugin && this.weather != null) {
+         return;
+      }
+      if (plugin) {
+         this.weather = type;
+      }
+      if (type == WeatherType.DOWNFALL) {
+         this.field_71135_a.func_147359_a(new SChangeGameStatePacket(2, 0.0f));
+      }
+      else {
+         this.field_71135_a.func_147359_a(new SChangeGameStatePacket(1, 0.0f));
+      }
+   }
+
+   public void updateWeather(final float oldRain, final float newRain, final float oldThunder, final float newThunder) {
+      if (this.weather == null) {
+         if (oldRain != newRain) {
+            this.field_71135_a.func_147359_a(new SChangeGameStatePacket(7, newRain));
+         }
+      }
+      else if (this.pluginRainPositionPrevious != this.pluginRainPosition) {
+         this.field_71135_a.func_147359_a(new SChangeGameStatePacket(7, this.pluginRainPosition));
+      }
+      if (oldThunder != newThunder) {
+         if (this.weather == WeatherType.DOWNFALL || this.weather == null) {
+            this.field_71135_a.func_147359_a(new SChangeGameStatePacket(8, newThunder));
+         }
+         else {
+            this.field_71135_a.func_147359_a(new SChangeGameStatePacket(8, 0.0f));
+         }
+      }
+   }
+
+   public void tickWeather() {
+      if (this.weather == null) {
+         return;
+      }
+      this.pluginRainPositionPrevious = this.pluginRainPosition;
+      if (this.weather == WeatherType.DOWNFALL) {
+         this.pluginRainPosition += (float)0.01;
+      }
+      else {
+         this.pluginRainPosition -= (float)0.01;
+      }
+      this.pluginRainPosition = MathHelper.func_76131_a(this.pluginRainPosition, 0.0f, 1.0f);
+   }
+
+   public void resetPlayerWeather() {
+      this.weather = null;
+      this.setPlayerWeather(this.field_70170_p.func_72912_H().func_76059_o() ? WeatherType.DOWNFALL : WeatherType.CLEAR, false);
+   }
+
+   @Override
+   public String toString() {
+      return String.valueOf(super.toString()) + "(" + this.func_195047_I_() + " at " + this.func_226277_ct_() + "," + this.func_226278_cu_() + "," + this.func_226281_cx_() + ")";
+   }
+
+   public void forceSetPositionRotation(double x, double y, double z, float yaw, float pitch) {
+      this.func_70012_b(x, y, z, yaw, pitch);
+      this.field_71135_a.func_184342_d();
+   }
+
+   @Override
+   public boolean func_70610_aX() {
+      return super.func_70610_aX() || !this.getBukkitEntity().isOnline();
+   }
+
+   @Override
+   public Scoreboard func_96123_co() {
+      return this.getBukkitEntity().getScoreboard().getHandle();
+   }
+
+   public void reset() {
+      float exp = 0.0f;
+      final boolean keepInventory = this.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223600_c);
+      if (this.keepLevel || keepInventory) {
+         exp = this.field_71106_cc;
+         this.newTotalExp = this.field_71067_cb;
+         this.newLevel = this.field_71068_ca;
+      }
+      this.func_70606_j(this.func_110138_aP());
+      this.field_190534_ay = 0;
+      this.field_70143_R = 0.0f;
+      this.field_71100_bB = new FoodStats(this);
+      this.field_71068_ca = this.newLevel;
+      this.field_71067_cb = this.newTotalExp;
+      this.field_71106_cc = 0.0f;
+      this.func_85034_r(this.field_70725_aQ = 0);
+      this.clearActivePotions(EntityPotionEffectEvent.Cause.DEATH);
+      this.field_70752_e = true;
+      this.field_71070_bA = this.field_71069_bz;
+      this.field_70717_bb = null;
+      this.field_70755_b = null;
+      this.field_94063_bt = new CombatTracker(this);
+      this.field_71144_ck = -1;
+      if (this.keepLevel || keepInventory) {
+         this.field_71106_cc = exp;
+      }
+      else {
+         this.func_195068_e(this.newExp);
+      }
+      this.keepLevel = false;
+   }
+   // CraftBukkit end
 }
