--- ../src-base/minecraft/org/bukkit/configuration/file/YamlConstructor.java
+++ ../src-work/minecraft/org/bukkit/configuration/file/YamlConstructor.java
@@ -2,14 +2,12 @@
 
 import java.util.LinkedHashMap;
 import java.util.Map;
-
-import org.yaml.snakeyaml.nodes.Node;
+import org.bukkit.configuration.serialization.ConfigurationSerialization;
 import org.yaml.snakeyaml.constructor.SafeConstructor;
 import org.yaml.snakeyaml.error.YAMLException;
+import org.yaml.snakeyaml.nodes.Node;
 import org.yaml.snakeyaml.nodes.Tag;
 
-import org.bukkit.configuration.serialization.ConfigurationSerialization;
-
 public class YamlConstructor extends SafeConstructor {
 
     public YamlConstructor() {
