--- ../src-base/minecraft/net/minecraft/entity/item/EntityTNTPrimed.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityTNTPrimed.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.item;
 
+import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
@@ -9,7 +10,10 @@
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.util.EnumParticleTypes;
+import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+import red.mohist.util.ThermiteRandom;
 
 public class EntityTNTPrimed extends Entity
 {
@@ -18,6 +22,10 @@
     private EntityLivingBase tntPlacedBy;
     private int fuse;
 
+    public float yield = 4;
+    public boolean isIncendiary = false;
+    private static final Random random = new ThermiteRandom();
+
     public EntityTNTPrimed(World worldIn)
     {
         super(worldIn);
@@ -31,10 +39,10 @@
     {
         this(worldIn);
         this.setPosition(x, y, z);
-        float f = (float)(Math.random() * (Math.PI * 2D));
-        this.motionX = (double)(-((float)Math.sin((double)f)) * 0.02F);
+        float f = (float)(random.nextDouble() * (Math.PI * 2D));
+        this.motionX = (double)(-(MathHelper.sin(f)) * 0.02F);
         this.motionY = 0.20000000298023224D;
-        this.motionZ = (double)(-((float)Math.cos((double)f)) * 0.02F);
+        this.motionZ = (double)(-(MathHelper.cos(f)) * 0.02F);
         this.setFuse(80);
         this.prevPosX = x;
         this.prevPosY = y;
@@ -59,6 +67,7 @@
 
     public void onUpdate()
     {
+        if (world.spigotConfig.currentPrimedTnt++ > world.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
         this.prevPosX = this.posX;
         this.prevPosY = this.posY;
         this.prevPosZ = this.posZ;
@@ -69,39 +78,48 @@
         }
 
         this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
-        this.motionX *= 0.9800000190734863D;
-        this.motionY *= 0.9800000190734863D;
-        this.motionZ *= 0.9800000190734863D;
+            this.motionX *= 0.9800000190734863D;
+            this.motionY *= 0.9800000190734863D;
+            this.motionZ *= 0.9800000190734863D;
 
-        if (this.onGround)
-        {
-            this.motionX *= 0.699999988079071D;
-            this.motionZ *= 0.699999988079071D;
-            this.motionY *= -0.5D;
-        }
+            if (this.onGround)
+			{
+                this.motionX *= 0.699999988079071D;
+                this.motionZ *= 0.699999988079071D;
+                this.motionY *= -0.5D;
+            }
 
-        --this.fuse;
+            --this.fuse;
 
-        if (this.fuse <= 0)
-        {
-            this.setDead();
+            if (this.fuse <= 0)
+			{
+                // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
+                // this.setDead();
 
-            if (!this.world.isRemote)
-            {
-                this.explode();
+                if (!this.world.isRemote)
+				{
+                    this.explode();
+                }
+                this.setDead();
             }
+			else
+			{
+                this.handleWaterMovement();
+                this.world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, this.posX, this.posY + 0.5D, this.posZ, 0.0D, 0.0D, 0.0D);
         }
-        else
-        {
-            this.handleWaterMovement();
-            this.world.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, this.posX, this.posY + 0.5D, this.posZ, 0.0D, 0.0D, 0.0D);
-        }
     }
 
     private void explode()
     {
-        float f = 4.0F;
-        this.world.createExplosion(this, this.posX, this.posY + (double)(this.height / 16.0F), this.posZ, 4.0F, true);
+        // float f = 4.0F;
+        // this.world.createExplosion(this, this.posX, this.posY + (double)(this.height / 16.0F), this.posZ, 4.0F, true);
+        org.bukkit.craftbukkit.v1_12_R1.CraftServer server = this.world.getServer();
+        ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) org.bukkit.craftbukkit.v1_12_R1.entity.CraftEntity.getEntity(server, this));
+        server.getPluginManager().callEvent(event);
+
+        if (!event.isCancelled()) {
+            this.world.newExplosion(this, this.posX, this.posY + (double) (this.height / 16.0F), this.posZ, event.getRadius(), event.getFire(), true);
+        }
     }
 
     protected void writeEntityToNBT(NBTTagCompound compound)
