--- a/net/minecraft/item/crafting/RecipeManager.java
+++ b/net/minecraft/item/crafting/RecipeManager.java
@@ -2,21 +_,24 @@
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Maps;
-import com.google.common.collect.ImmutableMap.Builder;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 import com.google.gson.JsonSyntaxException;
+import com.mohistmc.util.i18n.i18n;
+
+import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Objects;
 import java.util.Optional;
-import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import net.minecraft.client.resources.JsonReloadListener;
@@ -39,6 +_,7 @@
    private static final Gson field_223401_a = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
    private static final Logger field_199521_c = LogManager.getLogger();
    public Map<IRecipeType<?>, Map<ResourceLocation, IRecipe<?>>> field_199522_d = ImmutableMap.of();
+   public Map<IRecipeType<?>, Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>>> recipesCB = ImmutableMap.of(); // CraftBukkit
    private boolean field_199523_e;
 
    public RecipeManager() {
@@ -47,31 +_,90 @@
 
    protected void func_212853_a_(Map<ResourceLocation, JsonElement> p_212853_1_, IResourceManager p_212853_2_, IProfiler p_212853_3_) {
       this.field_199523_e = false;
-      Map<IRecipeType<?>, Builder<ResourceLocation, IRecipe<?>>> map = Maps.newHashMap();
+
+      // CraftBukkit start - SPIGOT-5667 make sure all types are populated and mutable
+      Map<IRecipeType<?>, Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>>> map1 = Maps.newHashMap();
+      for (IRecipeType<?> recipeType : Registry.field_218367_H) {
+         map1.put(recipeType, new Object2ObjectLinkedOpenHashMap<>());
+      }
+      // CraftBukkit end
+
+      Map<IRecipeType<?>, ImmutableMap.Builder<ResourceLocation, IRecipe<?>>> map = Maps.newHashMap();
+      ImmutableMap.Builder<ResourceLocation, IRecipe<?>> builder = ImmutableMap.builder();
 
       for(Entry<ResourceLocation, JsonElement> entry : p_212853_1_.entrySet()) {
          ResourceLocation resourcelocation = entry.getKey();
+         if (resourcelocation.func_110623_a().startsWith("_")) continue; //Forge: filter anything beginning with "_" as it's used for metadata.
 
          try {
+            if (entry.getValue().isJsonObject() && !net.minecraftforge.common.crafting.CraftingHelper.processConditions(entry.getValue().getAsJsonObject(), "conditions")) {
+               field_199521_c.debug(i18n.get("recipemanager.1", resourcelocation));
+               continue;
+            }
             IRecipe<?> irecipe = func_215377_a(resourcelocation, JSONUtils.func_151210_l(entry.getValue(), "top element"));
+            if (irecipe == null) {
+               continue;
+            }
             map.computeIfAbsent(irecipe.func_222127_g(), (p_223391_0_) -> {
                return ImmutableMap.builder();
             }).put(resourcelocation, irecipe);
+
+            // CraftBukkit start
+            (map1.computeIfAbsent(irecipe.func_222127_g(), (recipes) -> {
+               return new Object2ObjectLinkedOpenHashMap<>();
+               // CraftBukkit end
+            })).put(resourcelocation, irecipe);
+            builder.put(resourcelocation, irecipe);
          } catch (IllegalArgumentException | JsonParseException jsonparseexception) {
-            field_199521_c.error("Parsing error loading recipe {}", resourcelocation, jsonparseexception);
+            field_199521_c.debug(i18n.get("recipemanager.3", resourcelocation), jsonparseexception);
          }
       }
 
       this.field_199522_d = map.entrySet().stream().collect(ImmutableMap.toImmutableMap(Entry::getKey, (p_223400_0_) -> {
          return p_223400_0_.getValue().build();
       }));
-      field_199521_c.info("Loaded {} recipes", (int)map.size());
-   }
+
+      this.recipesCB = map1.entrySet().stream().collect(ImmutableMap.toImmutableMap(Entry::getKey, (entry1) -> {
+         return entry1.getValue(); // CraftBukkit
+      }));
+   }
+
+   // CraftBukkit start
+   public void addRecipe(final IRecipe<?> irecipe) {
+      if (this.field_199522_d instanceof ImmutableMap) {
+         this.field_199522_d = new HashMap<>(field_199522_d);
+      }
+
+      Map<ResourceLocation, IRecipe<?>> map = this.field_199522_d.get(irecipe.func_222127_g());
+      Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>> map0;
+      if (!(map instanceof Object2ObjectLinkedOpenHashMap)) {
+         Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>> hashMap = new Object2ObjectLinkedOpenHashMap<>();
+         hashMap.putAll(map);
+         this.field_199522_d.put(irecipe.func_222127_g(), hashMap);
+         map0 = hashMap;
+      } else {
+         map0 = ((Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>>) map);
+      }
+
+      Object2ObjectLinkedOpenHashMap<ResourceLocation, IRecipe<?>> craftbukkit = this.recipesCB.get(irecipe.func_222127_g()); // CraftBukkit
+
+      if (map0.containsKey(irecipe.func_199560_c()) || craftbukkit.containsKey(irecipe.func_199560_c())) {
+         throw new IllegalStateException("Duplicate recipe ignored with ID " + irecipe.func_199560_c());
+      } else {
+         map0.putAndMoveToFirst(irecipe.func_199560_c(), irecipe);
+         craftbukkit.putAndMoveToFirst(irecipe.func_199560_c(), irecipe); // CraftBukkit - SPIGOT-4638: last recipe gets priority
+      }
+   }
+   // CraftBukkit end
 
    public <C extends IInventory, T extends IRecipe<C>> Optional<T> func_215371_a(IRecipeType<T> p_215371_1_, C p_215371_2_, World p_215371_3_) {
-      return this.func_215366_a(p_215371_1_).values().stream().flatMap((p_215372_3_) -> {
+      // CraftBukkit start
+      Optional<T> recipe = this.func_215366_a(p_215371_1_).values().stream().flatMap((p_215372_3_) -> {
          return Util.func_215081_a(p_215371_1_.func_222148_a(p_215372_3_, p_215371_3_, p_215371_2_));
       }).findFirst();
+      p_215371_2_.setCurrentRecipe(recipe.orElse(null)); // CraftBukkit - Clear recipe when no recipe is found
+      // CraftBukkit end
+      return recipe;
    }
 
    public <C extends IInventory, T extends IRecipe<C>> List<T> func_241447_a_(IRecipeType<T> p_241447_1_) {
@@ -147,4 +_,16 @@
       });
       this.field_199522_d = ImmutableMap.copyOf(map);
    }
+
+   // CraftBukkit start
+   public void clearRecipes() {
+      this.field_199522_d = Maps.newHashMap();
+      this.recipesCB = Maps.newHashMap();
+      for (IRecipeType<?> recipeType : Registry.field_218367_H) {
+         this.field_199522_d.put(recipeType, ImmutableMap.of());
+         this.recipesCB.put(recipeType, new Object2ObjectLinkedOpenHashMap<>());
+      }
+   }
+   // CraftBukkit end
+
 }
