--- a/net/minecraft/item/crafting/ServerRecipePlacer.java
+++ b/net/minecraft/item/crafting/ServerRecipePlacer.java
@@ -69,7 +_,7 @@
             ItemStack itemstack1 = itemstack.func_77946_l();
             itemstack1.func_190920_e(1);
             if (!this.field_201514_c.func_191971_c(i, itemstack1)) {
-               field_194330_a.error("Can't find any space for item in the inventory");
+               field_194330_a.error(com.mohistmc.util.i18n.i18n.get("serverrecipeplacer.1"));
             }
          }
 
