--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -8,10 +_,10 @@
 import it.unimi.dsi.fastutil.shorts.Short2ObjectMap;
 import it.unimi.dsi.fastutil.shorts.Short2ObjectOpenHashMap;
 import java.util.Map;
-import java.util.Map.Entry;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.tags.BlockTags;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
@@ -29,6 +_,13 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.Bukkit;
+import org.bukkit.block.BlockFace;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.data.CraftBlockData;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockFromToEvent;
+import org.bukkit.event.block.FluidLevelChangeEvent;
 
 public abstract class FlowingFluid extends Fluid {
     public static final BooleanProperty FALLING = BlockStateProperties.FALLING;
@@ -121,6 +_,15 @@
             BlockState blockstate1 = p_255851_.getBlockState(blockpos);
             FluidState fluidstate = this.getNewLiquid(p_255851_, blockpos, blockstate1);
             if (this.canSpreadTo(p_255851_, p_76012_, blockstate, Direction.DOWN, blockpos, blockstate1, p_255851_.getFluidState(blockpos), fluidstate.getType())) {
+            // CraftBukkit start
+            org.bukkit.block.Block source = CraftBlock.at(p_255851_, p_76012_);
+            BlockFromToEvent event = new BlockFromToEvent(source, BlockFace.DOWN);
+            Bukkit.getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
                 this.spreadTo(p_255851_, blockpos, blockstate1, Direction.DOWN, fluidstate);
                 if (this.sourceNeighborCount(p_255851_, p_76012_) >= 3) {
                     this.spreadToSides(p_255851_, p_76012_, p_76013_, blockstate);
@@ -146,6 +_,15 @@
                 BlockPos blockpos = p_76016_.relative(direction);
                 BlockState blockstate = p_256644_.getBlockState(blockpos);
                 if (this.canSpreadTo(p_256644_, p_76016_, p_76018_, direction, blockpos, blockstate, p_256644_.getFluidState(blockpos), fluidstate.getType())) {
+               // CraftBukkit start
+               org.bukkit.block.Block source = CraftBlock.at(p_256644_, p_76016_);
+               BlockFromToEvent event = new BlockFromToEvent(source, CraftBlock.notchToBlockFace(direction));
+               Bukkit.getPluginManager().callEvent(event);
+
+               if (event.isCancelled()) {
+                  continue;
+               }
+               // CraftBukkit end
                     this.spreadTo(p_256644_, blockpos, blockstate, direction, fluidstate);
                 }
             }
@@ -161,7 +_,7 @@
             BlockState blockstate = p_256464_.getBlockState(blockpos);
             FluidState fluidstate = blockstate.getFluidState();
             if (fluidstate.getType().isSame(this) && this.canPassThroughWall(direction, p_256464_, p_76037_, p_76038_, blockpos, blockstate)) {
-                if (fluidstate.isSource()) {
+                if (fluidstate.isSource() && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(p_256464_, blockpos, blockstate, fluidstate.canConvertToSource(p_256464_, blockpos))) {
                     j++;
                 }
 
@@ -169,7 +_,7 @@
             }
         }
 
-        if (this.canConvertToSource(p_256464_) && j >= 2) {
+        if (j >= 2) {
             BlockState blockstate1 = p_256464_.getBlockState(p_76037_.below());
             FluidState fluidstate1 = blockstate1.getFluidState();
             if (blockstate1.isSolid() || this.isSourceBlockOfThisType(fluidstate1)) {
@@ -235,6 +_,15 @@
         return this.getSource().defaultFluidState().setValue(FALLING, Boolean.valueOf(p_76069_));
     }
 
+    @Override
+    public boolean canConvertToSource(FluidState state, Level level, BlockPos pos) {
+       return this.canConvertToSource(level);
+    }
+
+    /**
+     * @deprecated Forge: Use {@link #canConvertToSource(FluidState, Level, BlockPos)} instead.
+     */
+    @Deprecated
     protected abstract boolean canConvertToSource(Level p_256009_);
 
     protected void spreadTo(LevelAccessor p_76005_, BlockPos p_76006_, BlockState p_76007_, Direction p_76008_, FluidState p_76009_) {
@@ -436,11 +_,23 @@
             int i = this.getSpreadDelay(p_75995_, p_75996_, p_75997_, fluidstate);
             if (fluidstate.isEmpty()) {
                 p_75997_ = fluidstate;
-                p_75995_.setBlock(p_75996_, Blocks.AIR.defaultBlockState(), 3);
+            // CraftBukkit start
+            FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(p_75995_, p_75996_, Blocks.AIR.defaultBlockState());
+            if (event.isCancelled()) {
+               return;
+            }
+            p_75995_.setBlock(p_75996_, ((CraftBlockData) event.getNewData()).getState(), 3);
+            // CraftBukkit end
             } else if (!fluidstate.equals(p_75997_)) {
                 p_75997_ = fluidstate;
                 BlockState blockstate = fluidstate.createLegacyBlock();
-                p_75995_.setBlock(p_75996_, blockstate, 2);
+            // CraftBukkit start
+            FluidLevelChangeEvent event = CraftEventFactory.callFluidLevelChangeEvent(p_75995_, p_75996_, blockstate);
+            if (event.isCancelled()) {
+               return;
+            }
+            p_75995_.setBlock(p_75996_, ((CraftBlockData) event.getNewData()).getState(), 2);
+            // CraftBukkit end
                 p_75995_.scheduleTick(p_75996_, fluidstate.getType(), i);
                 p_75995_.updateNeighborsAt(p_75996_, blockstate.getBlock());
             }
