--- ../src-base/minecraft/org/bukkit/command/defaults/AchievementCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/AchievementCommand.java
@@ -1,15 +1,15 @@
 package org.bukkit.command.defaults;
 
+import com.google.common.collect.ImmutableList;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Achievement;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
-import org.bukkit.Statistic;
 import org.bukkit.Material;
+import org.bukkit.Statistic;
 import org.bukkit.Statistic.Type;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandSender;
@@ -18,8 +18,6 @@
 import org.bukkit.event.player.PlayerAchievementAwardedEvent;
 import org.bukkit.event.player.PlayerStatisticIncrementEvent;
 
-import com.google.common.collect.ImmutableList;
-
 public class AchievementCommand extends VanillaCommand {
     public AchievementCommand() {
         super("achievement");
