--- ../src-base/minecraft/org/bukkit/command/defaults/GameModeCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/GameModeCommand.java
@@ -1,8 +1,8 @@
 package org.bukkit.command.defaults;
 
+import com.google.common.collect.ImmutableList;
 import java.util.ArrayList;
 import java.util.List;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
@@ -12,8 +12,6 @@
 import org.bukkit.entity.Player;
 import org.bukkit.util.StringUtil;
 
-import com.google.common.collect.ImmutableList;
-
 public class GameModeCommand extends VanillaCommand {
     private static final List<String> GAMEMODE_NAMES = ImmutableList.of("adventure", "creative", "survival");
 
@@ -50,7 +48,7 @@
 
             GameMode mode = GameMode.getByValue(value);
 
-            if (mode == null) {
+            if (mode == null || mode == GameMode.NOT_SET) { // Cauldron
                 if (modeArg.equalsIgnoreCase("creative") || modeArg.equalsIgnoreCase("c")) {
                     mode = GameMode.CREATIVE;
                 } else if (modeArg.equalsIgnoreCase("adventure") || modeArg.equalsIgnoreCase("a")) {
