--- ../src-base/minecraft/org/bukkit/command/defaults/GameRuleCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/GameRuleCommand.java
@@ -1,20 +1,19 @@
 package org.bukkit.command.defaults;
 
 import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 import org.apache.commons.lang.Validate;
+import org.bukkit.Bukkit;
 import org.bukkit.ChatColor;
+import org.bukkit.World;
 import org.bukkit.command.BlockCommandSender;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandSender;
+import org.bukkit.entity.HumanEntity;
 import org.bukkit.util.StringUtil;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-import org.bukkit.Bukkit;
-import org.bukkit.World;
-import org.bukkit.entity.HumanEntity;
-
 public class GameRuleCommand extends VanillaCommand {
     private static final List<String> GAMERULE_STATES = ImmutableList.of("true", "false");
 
