--- ../src-base/minecraft/org/bukkit/command/defaults/OpCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/OpCommand.java
@@ -12,6 +12,8 @@
 import org.bukkit.command.CommandSender;
 import org.bukkit.entity.Player;
 import org.bukkit.util.StringUtil;
+import thermos.Thermos;
+import net.minecraft.server.MinecraftServer;
 
 import com.google.common.collect.ImmutableList;
 
@@ -32,10 +34,26 @@
         }
 
         OfflinePlayer player = Bukkit.getOfflinePlayer(args[0]);
-        player.setOp(true);
 
-        Command.broadcastCommandMessage(sender, "Opped " + args[0]);
-        return true;
+        if(MinecraftServer.thermosConfig.opConsoleOnly.getValue()) {
+
+            if(!(sender instanceof Player)) {
+                player.setOp(true);
+
+                Command.broadcastCommandMessage(sender, "Opped " + args[0]);
+                return true;
+            } else {
+                sender.sendMessage(ChatColor.RED + "This command can only be run in the console.");
+                return false;
+            }
+
+        } else {
+            player.setOp(true);
+
+            Command.broadcastCommandMessage(sender, "Opped " + args[0]);
+            return true;
+        }
+
     }
 
     @Override
