--- ../src-base/minecraft/org/bukkit/command/defaults/SeedCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/SeedCommand.java
@@ -1,14 +1,12 @@
 package org.bukkit.command.defaults;
 
+import com.google.common.collect.ImmutableList;
 import java.util.List;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Bukkit;
 import org.bukkit.command.CommandSender;
 import org.bukkit.entity.Player;
 
-import com.google.common.collect.ImmutableList;
-
 public class SeedCommand extends VanillaCommand {
     public SeedCommand() {
         super("seed");
