--- ../src-base/minecraft/org/bukkit/command/defaults/StopCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/StopCommand.java
@@ -1,17 +1,14 @@
 package org.bukkit.command.defaults;
 
+import com.google.common.collect.ImmutableList;
 import java.util.List;
-
 import org.apache.commons.lang.StringUtils;
 import org.apache.commons.lang.Validate;
-
 import org.bukkit.Bukkit;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandSender;
 import org.bukkit.entity.Player;
 
-import com.google.common.collect.ImmutableList;
-
 public class StopCommand extends VanillaCommand {
     public StopCommand() {
         super("stop");
