--- a/net/minecraft/SystemReport.java
+++ b/net/minecraft/SystemReport.java
@@ -1,6 +_,7 @@
 package net.minecraft;
 
 import com.google.common.collect.Maps;
+import com.mohistmc.MohistMC;
 import com.mojang.logging.LogUtils;
 import java.util.List;
 import java.util.Locale;
@@ -26,6 +_,7 @@
    private final Map<String, String> entries = Maps.newLinkedHashMap();
 
    public SystemReport() {
+      this.setDetail("Mohist Version", MohistMC.versionInfo.mohist());
       this.setDetail("Minecraft Version", SharedConstants.getCurrentVersion().getName());
       this.setDetail("Minecraft Version ID", SharedConstants.getCurrentVersion().getId());
       this.setDetail("Operating System", OPERATING_SYSTEM);
