--- a/net/minecraft/server/network/LegacyQueryHandler.java
+++ b/net/minecraft/server/network/LegacyQueryHandler.java
@@ -28,9 +_,10 @@
          if (bytebuf.readUnsignedByte() == 254) {
             SocketAddress socketaddress = p_9686_.channel().remoteAddress();
             int i = bytebuf.readableBytes();
+            org.bukkit.event.server.ServerListPingEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callServerListPingEvent(socketaddress, server.getMotd(), server.getPlayerCount(), server.getMaxPlayers()); // CraftBukkit
             if (i == 0) {
                LOGGER.debug("Ping: (<1.3.x) from {}", (Object)socketaddress);
-               String s = createVersion0Response(this.server);
+               String s = createVersion0Response(this.server, event);
                sendFlushAndClose(p_9686_, createLegacyDisconnectPacket(p_9686_.alloc(), s));
             } else {
                if (bytebuf.readUnsignedByte() != 1) {
@@ -47,7 +_,7 @@
                   LOGGER.debug("Ping: (1.4-1.5.x) from {}", (Object)socketaddress);
                }
 
-               String s1 = createVersion1Response(this.server);
+               String s1 = createVersion1Response(this.server, event);
                sendFlushAndClose(p_9686_, createLegacyDisconnectPacket(p_9686_.alloc(), s1));
             }
 
@@ -94,12 +_,12 @@
       }
    }
 
-   private static String createVersion0Response(ServerInfo p_300881_) {
-      return String.format(Locale.ROOT, "%s\u00a7%d\u00a7%d", p_300881_.getMotd(), p_300881_.getPlayerCount(), p_300881_.getMaxPlayers());
+   private static String createVersion0Response(ServerInfo p_300881_, org.bukkit.event.server.ServerListPingEvent event) {
+      return String.format(Locale.ROOT, "%s\u00a7%d\u00a7%d", event.getMotd(), event.getNumPlayers(), event.getMaxPlayers());
    }
 
-   private static String createVersion1Response(ServerInfo p_297753_) {
-      return String.format(Locale.ROOT, "\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, p_297753_.getServerVersion(), p_297753_.getMotd(), p_297753_.getPlayerCount(), p_297753_.getMaxPlayers());
+   private static String createVersion1Response(ServerInfo p_297753_, org.bukkit.event.server.ServerListPingEvent event) {
+      return String.format(Locale.ROOT, "\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", 127, p_297753_.getServerVersion(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers());
    }
 
    private static void sendFlushAndClose(ChannelHandlerContext p_9681_, ByteBuf p_9682_) {
