--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -21,7 +_,9 @@
 import net.minecraft.network.protocol.common.ServerboundResourcePackPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ClientInformation;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.VisibleForDebug;
+import org.bukkit.craftbukkit.v1_20_R2.entity.CraftPlayer;
 import org.slf4j.Logger;
 
 public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener {
@@ -29,18 +_,30 @@
    public static final int LATENCY_CHECK_INTERVAL = 15000;
    private static final Component TIMEOUT_DISCONNECTION_MESSAGE = Component.translatable("disconnect.timeout");
    protected final MinecraftServer server;
-   protected final Connection connection;
+   public final Connection connection;
    private long keepAliveTime;
    private boolean keepAlivePending;
    private long keepAliveChallenge;
    private int latency;
    private volatile boolean suspendFlushingOnServerThread = false;
 
-   public ServerCommonPacketListenerImpl(MinecraftServer p_299469_, Connection p_300872_, CommonListenerCookie p_300277_) {
+   public ServerCommonPacketListenerImpl(MinecraftServer p_299469_, Connection p_300872_, CommonListenerCookie p_300277_, ServerPlayer player) {
       this.server = p_299469_;
       this.connection = p_300872_;
       this.keepAliveTime = Util.getMillis();
       this.latency = p_300277_.latency();
+      // CraftBukkit start - add fields and methods
+      this.player = player;
+      this.cserver = p_299469_.server;
+   }
+
+   protected final ServerPlayer player;
+   protected final org.bukkit.craftbukkit.v1_20_R2.CraftServer cserver;
+   public boolean processedDisconnect;
+
+   public CraftPlayer getCraftPlayer() {
+      return (this.player == null) ? null : (CraftPlayer) this.player.getBukkitEntity();
+      // CraftBukkit end
    }
 
    public void onDisconnect(Component p_300550_) {
@@ -66,6 +_,7 @@
    }
 
    public void handleCustomPayload(ServerboundCustomPayloadPacket p_300164_) {
+      net.minecraftforge.common.ForgeHooks.onCustomPayload(p_300164_, this.connection);
    }
 
    public void handleResourcePackResponse(ServerboundResourcePackPacket p_300656_) {
@@ -147,5 +_,9 @@
 
    protected CommonListenerCookie createCookie(ClientInformation p_297318_) {
       return new CommonListenerCookie(this.playerProfile(), this.latency, p_297318_);
+   }
+
+   public Connection getConnection() {
+      return this.connection;
    }
 }
