--- a/net/minecraft/util/DefaultUncaughtExceptionHandler.java
+++ b/net/minecraft/util/DefaultUncaughtExceptionHandler.java
@@ -1,5 +_,6 @@
 package net.minecraft.util;
 
+import com.mohistmc.util.i18n.i18n;
 import java.lang.Thread.UncaughtExceptionHandler;
 import org.apache.logging.log4j.Logger;
 
@@ -11,6 +_,6 @@
    }
 
    public void uncaughtException(Thread p_uncaughtException_1_, Throwable p_uncaughtException_2_) {
-      this.field_201709_a.error("Caught previously unhandled exception :", p_uncaughtException_2_);
+      this.field_201709_a.error(i18n.get("defaultuncaughtexceptionhandler.1", p_uncaughtException_2_));
    }
 }
