--- a/net/minecraft/util/LoggingPrintStream.java
+++ b/net/minecraft/util/LoggingPrintStream.java
@@ -1,5 +_,6 @@
 package net.minecraft.util;
 
+import com.mohistmc.util.i18n.i18n;
 import java.io.OutputStream;
 import java.io.PrintStream;
 import javax.annotation.Nullable;
@@ -24,6 +_,6 @@
    }
 
    protected void func_179882_a(@Nullable String p_179882_1_) {
-      field_179884_a.info("[{}]: {}", this.field_179883_b, p_179882_1_);
+      field_179884_a.info(i18n.get("loggingprintstream.1", this.field_179883_b, p_179882_1_));
    }
 }
