--- a/net/minecraft/util/MinecraftVersion.java
+++ b/net/minecraft/util/MinecraftVersion.java
@@ -2,6 +_,7 @@
 
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import com.mohistmc.util.i18n.i18n;
 import com.mojang.bridge.game.GameVersion;
 import java.io.IOException;
 import java.io.InputStream;
@@ -49,7 +_,7 @@
    public static GameVersion func_214957_a() {
       try (InputStream inputstream = MinecraftVersion.class.getResourceAsStream("/version.json")) {
          if (inputstream == null) {
-            field_214958_a.warn("Missing version information!");
+            field_214958_a.warn(i18n.get("minecraftversion.1"));
             return field_240808_a_;
          } else {
             MinecraftVersion minecraftversion;
