--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -1,6 +_,7 @@
 package net.minecraft.util;
 
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -15,8 +_,23 @@
 import net.minecraft.world.level.block.StainedGlassBlock;
 import net.minecraft.world.level.block.StainedGlassPaneBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class SpawnUtil {
+
+   // Mohist start
+   private static final AtomicReference<SpawnReason> trySpawnMob$reason = new AtomicReference<>(SpawnReason.DEFAULT);
+
+   public static void trySpawnMob$reason(SpawnReason reason) {
+      trySpawnMob$reason.set(reason);
+   }
+
+   public static <T extends Mob> Optional<T> trySpawnMob(EntityType<T> pEntityType, MobSpawnType pSpawnType, ServerLevel pLevel, BlockPos pPos, int pAttempts, int p_216409_, int pYOffset, SpawnUtil.Strategy pStrategy, SpawnReason reason) {
+      trySpawnMob$reason(reason);
+      return m_216403_(pEntityType, pSpawnType, pLevel, pPos, pAttempts, p_216409_, pYOffset, pStrategy);
+   }
+   // Mohist end
+
    public static <T extends Mob> Optional<T> m_216403_(EntityType<T> p_216404_, MobSpawnType p_216405_, ServerLevel p_216406_, BlockPos p_216407_, int p_216408_, int p_216409_, int p_216410_, SpawnUtil.Strategy p_216411_) {
       BlockPos.MutableBlockPos blockpos$mutableblockpos = p_216407_.m_122032_();
 
@@ -27,8 +_,8 @@
          if (p_216406_.m_6857_().m_61937_(blockpos$mutableblockpos) && m_216398_(p_216406_, p_216410_, blockpos$mutableblockpos, p_216411_)) {
             T t = p_216404_.m_262451_(p_216406_, (CompoundTag)null, (Consumer<T>)null, blockpos$mutableblockpos, p_216405_, false, false);
             if (t != null) {
-               if (t.m_5545_(p_216406_, p_216405_) && t.m_6914_(p_216406_)) {
-                  p_216406_.m_47205_(t);
+               if (net.minecraftforge.event.ForgeEventFactory.checkSpawnPosition(t, p_216406_, p_216405_)) {
+                  p_216406_.addFreshEntityWithPassengers(t, trySpawnMob$reason.getAndSet(SpawnReason.DEFAULT)); // CraftBukkit // Mohist
                   return Optional.of(t);
                }
 
