--- ../src-base/minecraft/net/minecraftforge/fml/common/registry/EntityEntryBuilder.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/registry/EntityEntryBuilder.java
@@ -306,7 +306,7 @@
             }
             catch (final IllegalAccessException | InstantiationException | InvocationTargetException e)
             {
-                FMLLog.log.error("Encountered an exception while constructing entity '{}'", this.describeEntity(), e);
+                //FMLLog.log.error("Encountered an exception while constructing entity '{}'", this.describeEntity(), e);
                 return null;
             }
         }
