--- ../src-base/minecraft/net/minecraft/network/NettyCompressionDecoder.java
+++ ../src-work/minecraft/net/minecraft/network/NettyCompressionDecoder.java
@@ -38,9 +38,9 @@
                     throw new DecoderException("Badly compressed packet - size of " + i + " is below server threshold of " + this.field_179304_b);
                 }
 
-                if (i > 2097152)
+                if (i > 8388608)
                 {
-                    throw new DecoderException("Badly compressed packet - size of " + i + " is larger than protocol maximum of " + 2097152);
+                    throw new DecoderException("Badly compressed packet - size of " + i + " is larger than protocol maximum of " + 8388608);
                 }
 
                 byte[] abyte = new byte[packetbuffer.readableBytes()];
