--- ../src-base/minecraft/net/minecraft/network/NettyPacketEncoder.java
+++ ../src-work/minecraft/net/minecraft/network/NettyPacketEncoder.java
@@ -36,7 +36,10 @@
             {
                 LOGGER.debug(RECEIVED_PACKET_MARKER, "OUT: [{}:{}] {}", p_encode_1_.channel().attr(NetworkManager.PROTOCOL_ATTRIBUTE_KEY).get(), integer, p_encode_2_.getClass().getName());
             }
-
+            if (integer == null && (integer = EnumConnectionState.LOGIN.getPacketId(this.direction, p_encode_2_)) != null)
+            {
+                LOGGER.debug("current state is {}, but send packet {}", enumconnectionstate, p_encode_2_.getClass().getSimpleName());
+            }
             if (integer == null)
             {
                 throw new IOException("Can't serialize unregistered packet");
