--- a/net/minecraft/world/chunk/storage/RegionSectionCache.java
+++ b/net/minecraft/world/chunk/storage/RegionSectionCache.java
@@ -107,7 +_,7 @@
       try {
          return this.field_227173_b_.func_227090_a_(p_223138_1_);
       } catch (IOException ioexception) {
-         field_219120_a.error("Error reading chunk {} data from disk", p_223138_1_, ioexception);
+         field_219120_a.error(com.mohistmc.util.i18n.i18n.get("regionsectioncache.1", p_223138_1_, ioexception));
          return null;
       }
    }
@@ -151,7 +_,7 @@
       if (inbt instanceof CompoundNBT) {
          this.field_227173_b_.func_227093_a_(p_219117_1_, (CompoundNBT)inbt);
       } else {
-         field_219120_a.error("Expected compound tag, got {}", (Object)inbt);
+         field_219120_a.error(com.mohistmc.util.i18n.i18n.get("regionsectioncache.2", (Object)inbt));
       }
 
    }
@@ -185,7 +_,7 @@
       if (optional != null && optional.isPresent()) {
          this.field_219122_d.add(p_219116_1_);
       } else {
-         field_219120_a.warn("No data for position: {}", (Object)SectionPos.func_218170_a(p_219116_1_));
+         field_219120_a.warn(com.mohistmc.util.i18n.i18n.get("regionsectioncache.3", (Object)SectionPos.func_218170_a(p_219116_1_)));
       }
    }
 
