--- ../src-base/minecraft/org/bukkit/event/server/RemoteServerCommandEvent.java
+++ ../src-work/minecraft/org/bukkit/event/server/RemoteServerCommandEvent.java
@@ -14,12 +14,12 @@
         super(sender, command);
     }
 
-    @Override
-    public HandlerList getHandlers() {
+    public static HandlerList getHandlerList() {
         return handlers;
     }
 
-    public static HandlerList getHandlerList() {
+    @Override
+    public HandlerList getHandlers() {
         return handlers;
     }
 }
