--- a/net/minecraft/core/dispenser/AbstractProjectileDispenseBehavior.java
+++ b/net/minecraft/core/dispenser/AbstractProjectileDispenseBehavior.java
@@ -6,6 +_,9 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.DispenserBlock;
+import org.bukkit.craftbukkit.v1_20_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public abstract class AbstractProjectileDispenseBehavior extends DefaultDispenseItemBehavior {
    public ItemStack execute(BlockSource p_123366_, ItemStack p_123367_) {
@@ -13,9 +_,37 @@
       Position position = DispenserBlock.getDispensePosition(p_123366_);
       Direction direction = p_123366_.state().getValue(DispenserBlock.FACING);
       Projectile projectile = this.getProjectile(level, position, p_123367_);
-      projectile.shoot((double)direction.getStepX(), (double)((float)direction.getStepY() + 0.1F), (double)direction.getStepZ(), this.getPower(), this.getUncertainty());
+
+      // CraftBukkit start
+      ItemStack itemstack1 = p_123367_.split(1);
+      org.bukkit.block.Block block = CraftBlock.at(level, p_123366_.pos());
+      CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+      
+      BlockDispenseEvent event = new BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) direction.getStepX(), (double) ((float) direction.getStepY() + 0.1F), (double) direction.getStepZ()));
+      if (!DispenserBlock.eventFired) {
+         level.getCraftServer().getPluginManager().callEvent(event);
+      }
+      
+      if (event.isCancelled()) {
+         p_123367_.grow(1);
+         return p_123367_;
+      }
+      
+      if (!event.getItem().equals(craftItem)) {
+         p_123367_.grow(1);
+         // Chain to handler for new item
+         ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+         DispenseItemBehavior idispensebehavior = DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+         if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != this) {
+            idispensebehavior.dispense(p_123366_, eventStack);
+            return p_123367_;
+         }
+      }
+
+      projectile.shoot(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), this.getPower(), this.getUncertainty());
+      projectile.projectileSource = new org.bukkit.craftbukkit.v1_20_R2.projectiles.CraftBlockProjectileSource(p_123366_.blockEntity());
+      // CraftBukkit end
       level.addFreshEntity(projectile);
-      p_123367_.shrink(1);
       return p_123367_;
    }
 
