--- a/net/minecraft/entity/item/HangingEntity.java
+++ b/net/minecraft/entity/item/HangingEntity.java
@@ -27,7 +27,7 @@
       return p_210144_0_ instanceof HangingEntity;
    };
    private int field_70520_f;
-   protected BlockPos field_174861_a;
+   public BlockPos field_174861_a;
    protected Direction field_174860_b = Direction.SOUTH;
 
    protected HangingEntity(EntityType<? extends HangingEntity> p_i48561_1_, World p_i48561_2_) {
@@ -39,10 +39,14 @@
       this.field_174861_a = p_i48562_3_;
    }
 
-   protected void func_70088_a() {
+    public static AxisAlignedBB calculateBoundingBox(Object o, BlockPos pos, Direction opposite, int width, int height) {
+      return null;
+    }
+
+    protected void func_70088_a() {
    }
 
-   protected void func_174859_a(Direction p_174859_1_) {
+   public void func_174859_a(Direction p_174859_1_) {
       Validate.notNull(p_174859_1_);
       Validate.isTrue(p_174859_1_.func_176740_k().func_176722_c());
       this.field_174860_b = p_174859_1_;
@@ -113,6 +117,8 @@
                int j1 = (j - 1) / -2;
                blockpos$mutable.func_189533_g(blockpos).func_189534_c(direction, k + i1).func_189534_c(Direction.UP, l + j1);
                BlockState blockstate = this.field_70170_p.func_180495_p(blockpos$mutable);
+               if (net.minecraft.block.Block.func_220055_a(this.field_70170_p, blockpos$mutable, this.field_174860_b))
+                  continue;
                if (!blockstate.func_185904_a().func_76220_a() && !RedstoneDiodeBlock.func_185546_B(blockstate)) {
                   return false;
                }
