--- a/net/minecraft/entity/item/ItemFrameEntity.java
+++ b/net/minecraft/entity/item/ItemFrameEntity.java
@@ -57,7 +57,7 @@
       this.func_184212_Q().func_187214_a(field_184526_d, 0);
    }
 
-   protected void func_174859_a(Direction p_174859_1_) {
+   public void func_174859_a(Direction p_174859_1_) { // protected->public CraftBukkit
       Validate.notNull(p_174859_1_);
       this.field_174860_b = p_174859_1_;
       if (p_174859_1_.func_176740_k().func_176722_c()) {
@@ -75,15 +75,24 @@
 
    protected void func_174856_o() {
       if (this.field_174860_b != null) {
-         double d0 = 0.46875D;
-         double d1 = (double)this.field_174861_a.func_177958_n() + 0.5D - (double)this.field_174860_b.func_82601_c() * 0.46875D;
-         double d2 = (double)this.field_174861_a.func_177956_o() + 0.5D - (double)this.field_174860_b.func_96559_d() * 0.46875D;
-         double d3 = (double)this.field_174861_a.func_177952_p() + 0.5D - (double)this.field_174860_b.func_82599_e() * 0.46875D;
-         this.func_226288_n_(d1, d2, d3);
-         double d4 = (double)this.func_82329_d();
-         double d5 = (double)this.func_82330_g();
-         double d6 = (double)this.func_82329_d();
-         Direction.Axis direction$axis = this.field_174860_b.func_176740_k();
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.func_174826_a(calculateBoundingBox(this, this.field_174861_a, this.field_174860_b, this.func_82329_d(), this.func_82330_g()));
+      }
+   }
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AxisAlignedBB calculateBoundingBox(@Nullable Entity entity, BlockPos blockPosition, Direction direction, int width, int height) {
+      final double d0 = 0.46875D;
+      final double d1 = blockPosition.func_177958_n() + 0.5D - direction.func_82601_c() * d0;
+      final double d2 = blockPosition.func_177956_o() + 0.5D - direction.func_96559_d() * d0;
+      final double d3 = blockPosition.func_177952_p() + 0.5D - direction.func_82599_e() * d0;
+      if (entity != null) {
+         entity.func_226288_n_(d1, d2, d3);
+      }
+      double d4 = width;
+      double d5 = height;
+      double d6 = width;
+      final Direction.Axis direction$axis = direction.func_176740_k();
          switch(direction$axis) {
          case X:
             d4 = 1.0D;
@@ -98,9 +107,9 @@
          d4 = d4 / 32.0D;
          d5 = d5 / 32.0D;
          d6 = d6 / 32.0D;
-         this.func_174826_a(new AxisAlignedBB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6));
+         return new AxisAlignedBB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6);
       }
-   }
+   // CraftBukkit end
 
    public boolean func_70518_d() {
       if (!this.field_70170_p.func_226669_j_(this)) {
@@ -125,6 +134,11 @@
          return false;
       } else if (!p_70097_1_.func_94541_c() && !this.func_82335_i().func_190926_b()) {
          if (!this.field_70170_p.field_72995_K) {
+            // CraftBukkit start - fire EntityDamageEvent
+            if (org.bukkit.craftbukkit.v1_15_R1.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false) || this.field_70128_L) {
+               return true;
+            }
+            // CraftBukkit end
             this.func_146065_b(p_70097_1_.func_76346_g(), false);
             this.func_184185_a(SoundEvents.field_187629_cO, 1.0F, 1.0F);
          }
@@ -192,7 +206,7 @@
    }
 
    private void func_110131_b(ItemStack p_110131_1_) {
-      if (p_110131_1_.func_77973_b() == Items.field_151098_aY) {
+      if (p_110131_1_.func_77973_b() instanceof net.minecraft.item.FilledMapItem) {
          MapData mapdata = FilledMapItem.func_195950_a(p_110131_1_, this.field_70170_p);
          mapdata.func_212441_a(this.field_174861_a, this.func_145782_y());
          mapdata.func_76186_a(true);
@@ -210,14 +224,24 @@
    }
 
    public void func_174864_a(ItemStack p_174864_1_, boolean p_174864_2_) {
-      if (!p_174864_1_.func_190926_b()) {
-         p_174864_1_ = p_174864_1_.func_77946_l();
-         p_174864_1_.func_190920_e(1);
-         p_174864_1_.func_82842_a(this);
+      this.setItem(p_174864_1_, p_174864_2_, true);
+   }
+
+   public void setItem(ItemStack stack, final boolean p_174864_2_) {
+      // CraftBukkit start
+      this.setItem(stack, p_174864_2_, true);
+   }
+
+   public void setItem(ItemStack stack, boolean p_174864_2_, boolean playSound) {
+      // CraftBukkit end
+      if (!stack.func_190926_b()) {
+         stack = stack.func_77946_l();
+         stack.func_190920_e(1);
+         stack.func_82842_a(this);
       }
 
-      this.func_184212_Q().func_187227_b(field_184525_c, p_174864_1_);
-      if (!p_174864_1_.func_190926_b()) {
+      this.func_184212_Q().func_187227_b(field_184525_c, stack);
+      if (!stack.func_190926_b() && playSound) { // CraftBukkit
          this.func_184185_a(SoundEvents.field_187620_cL, 1.0F, 1.0F);
       }
 
