--- ../src-base/minecraft/org/bukkit/command/FormattedCommandAlias.java
+++ ../src-work/minecraft/org/bukkit/command/FormattedCommandAlias.java
@@ -1,13 +1,7 @@
 package org.bukkit.command;
 
 import java.util.ArrayList;
-import java.util.logging.Level;
-
 import org.bukkit.Bukkit;
-import org.bukkit.entity.Player;
-import org.bukkit.event.player.PlayerCommandPreprocessEvent;
-import org.bukkit.event.server.RemoteServerCommandEvent;
-import org.bukkit.event.server.ServerCommandEvent;
 
 public class FormattedCommandAlias extends Command {
     private final String[] formatStrings;
