--- ../src-base/minecraft/org/bukkit/command/SimpleCommandMap.java
+++ ../src-work/minecraft/org/bukkit/command/SimpleCommandMap.java
@@ -1,7 +1,5 @@
 package org.bukkit.command;
 
-import static org.bukkit.util.Java15Compat.Arrays_copyOfRange;
-
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -10,11 +8,56 @@
 import java.util.List;
 import java.util.Map;
 import java.util.regex.Pattern;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Server;
-import org.bukkit.command.defaults.*;
+import org.bukkit.command.defaults.AchievementCommand;
+import org.bukkit.command.defaults.BanCommand;
+import org.bukkit.command.defaults.BanIpCommand;
+import org.bukkit.command.defaults.BanListCommand;
+import org.bukkit.command.defaults.ClearCommand;
+import org.bukkit.command.defaults.DefaultGameModeCommand;
+import org.bukkit.command.defaults.DeopCommand;
+import org.bukkit.command.defaults.DifficultyCommand;
+import org.bukkit.command.defaults.EffectCommand;
+import org.bukkit.command.defaults.EnchantCommand;
+import org.bukkit.command.defaults.ExpCommand;
+import org.bukkit.command.defaults.GameModeCommand;
+import org.bukkit.command.defaults.GameRuleCommand;
+import org.bukkit.command.defaults.GiveCommand;
+import org.bukkit.command.defaults.HelpCommand;
+import org.bukkit.command.defaults.KickCommand;
+import org.bukkit.command.defaults.KillCommand;
+import org.bukkit.command.defaults.ListCommand;
+import org.bukkit.command.defaults.MeCommand;
+import org.bukkit.command.defaults.OpCommand;
+import org.bukkit.command.defaults.PardonCommand;
+import org.bukkit.command.defaults.PardonIpCommand;
+import org.bukkit.command.defaults.PlaySoundCommand;
+import org.bukkit.command.defaults.PluginsCommand;
+import org.bukkit.command.defaults.ReloadCommand;
+import org.bukkit.command.defaults.SaveCommand;
+import org.bukkit.command.defaults.SaveOffCommand;
+import org.bukkit.command.defaults.SaveOnCommand;
+import org.bukkit.command.defaults.SayCommand;
+import org.bukkit.command.defaults.ScoreboardCommand;
+import org.bukkit.command.defaults.SeedCommand;
+import org.bukkit.command.defaults.SetIdleTimeoutCommand;
+import org.bukkit.command.defaults.SetWorldSpawnCommand;
+import org.bukkit.command.defaults.SpawnpointCommand;
+import org.bukkit.command.defaults.SpreadPlayersCommand;
+import org.bukkit.command.defaults.StopCommand;
+import org.bukkit.command.defaults.TeleportCommand;
+import org.bukkit.command.defaults.TellCommand;
+import org.bukkit.command.defaults.TestForCommand;
+import org.bukkit.command.defaults.TimeCommand;
+import org.bukkit.command.defaults.TimingsCommand;
+import org.bukkit.command.defaults.ToggleDownfallCommand;
+import org.bukkit.command.defaults.VanillaCommand;
+import org.bukkit.command.defaults.VersionCommand;
+import org.bukkit.command.defaults.WeatherCommand;
+import org.bukkit.command.defaults.WhitelistCommand;
 import org.bukkit.entity.Player;
+import static org.bukkit.util.Java15Compat.Arrays_copyOfRange;
 import org.bukkit.util.StringUtil;
 
 public class SimpleCommandMap implements CommandMap {
@@ -176,11 +219,15 @@
         }
 
         try {
+            target.timings.startTiming(); // Spigot
             // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
             target.execute(sender, sentCommandLabel, Arrays_copyOfRange(args, 1, args.length));
+            target.timings.stopTiming(); // Spigot
         } catch (CommandException ex) {
+            target.timings.stopTiming(); // Spigot
             throw ex;
         } catch (Throwable ex) {
+            target.timings.stopTiming(); // Spigot
             throw new CommandException("Unhandled exception executing '" + commandLine + "' in " + target, ex);
         }
 
