--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.level.block.entity;
 
+import java.util.ArrayList;
+import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.Vec3i;
@@ -18,10 +_,12 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BarrelBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.v1_20_R2.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
    private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
-   public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
+   public ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
       protected void onOpen(Level p_155062_, BlockPos p_155063_, BlockState p_155064_) {
          BarrelBlockEntity.this.playSound(p_155064_, SoundEvents.BARREL_OPEN);
          BarrelBlockEntity.this.updateBlockState(p_155064_, true);
@@ -107,15 +_,51 @@
 
    }
 
-   void updateBlockState(BlockState p_58607_, boolean p_58608_) {
+   public void updateBlockState(BlockState p_58607_, boolean p_58608_) {
       this.level.setBlock(this.getBlockPos(), p_58607_.setValue(BarrelBlock.OPEN, Boolean.valueOf(p_58608_)), 3);
    }
 
-   void playSound(BlockState p_58601_, SoundEvent p_58602_) {
+   public void playSound(BlockState p_58601_, SoundEvent p_58602_) {
       Vec3i vec3i = p_58601_.getValue(BarrelBlock.FACING).getNormal();
       double d0 = (double)this.worldPosition.getX() + 0.5D + (double)vec3i.getX() / 2.0D;
       double d1 = (double)this.worldPosition.getY() + 0.5D + (double)vec3i.getY() / 2.0D;
       double d2 = (double)this.worldPosition.getZ() + 0.5D + (double)vec3i.getZ() / 2.0D;
       this.level.playSound((Player)null, d0, d1, d2, p_58602_, SoundSource.BLOCKS, 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
    }
+
+   // CraftBukkit start - add fields and methods
+   public List<HumanEntity> transaction = new ArrayList<>();
+   private int maxStack = MAX_STACK;
+
+   @Override
+   public List<ItemStack> getContents() {
+      return this.items;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   @Override
+   public int getMaxStackSize() {
+      return maxStack;
+   }
+
+   @Override
+   public void setMaxStackSize(int i) {
+      maxStack = i;
+   }
+
+   // CraftBukkit end
 }
