--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.block.entity;
 
+import com.mohistmc.bukkit.inventory.InventoryOwner;
 import com.mojang.logging.LogUtils;
 import javax.annotation.Nullable;
 import net.minecraft.CrashReportCategory;
@@ -13,9 +_,12 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.v1_20_R3.persistence.CraftPersistentDataContainer;
+import org.bukkit.craftbukkit.v1_20_R3.persistence.CraftPersistentDataTypeRegistry;
+import org.bukkit.inventory.InventoryHolder;
 import org.slf4j.Logger;
 
-public abstract class BlockEntity {
+public abstract class BlockEntity extends net.minecraftforge.common.capabilities.CapabilityProvider<BlockEntity> implements net.minecraftforge.common.extensions.IForgeBlockEntity {
    private static final Logger LOGGER = LogUtils.getLogger();
    private final BlockEntityType<?> type;
    @Nullable
@@ -23,11 +_,19 @@
    protected final BlockPos worldPosition;
    protected boolean remove;
    private BlockState blockState;
+   private CompoundTag customPersistentData;
+
+   // CraftBukkit start - data containers
+   private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
+   public CraftPersistentDataContainer persistentDataContainer;
+   // CraftBukkit end
 
    public BlockEntity(BlockEntityType<?> p_155228_, BlockPos p_155229_, BlockState p_155230_) {
+      super(BlockEntity.class);
       this.type = p_155228_;
       this.worldPosition = p_155229_.immutable();
       this.blockState = p_155230_;
+      this.gatherCapabilities();
    }
 
    public static BlockPos getPosFromTag(CompoundTag p_187473_) {
@@ -48,9 +_,20 @@
    }
 
    public void load(CompoundTag p_155245_) {
+      if (p_155245_.contains("ForgeData")) this.customPersistentData = p_155245_.getCompound("ForgeData");
+      if (getCapabilities() != null && p_155245_.contains("ForgeCaps")) deserializeCaps(p_155245_.getCompound("ForgeCaps"));
+
+      this.persistentDataContainer = new CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
+
+      net.minecraft.nbt.Tag persistentDataTag = p_155245_.get("PublicBukkitValues");
+      if (persistentDataTag instanceof CompoundTag) {
+         this.persistentDataContainer.putAll((CompoundTag) persistentDataTag);
+      }
    }
 
    protected void saveAdditional(CompoundTag p_187471_) {
+      if (this.customPersistentData != null) p_187471_.put("ForgeData", this.customPersistentData.copy());
+      if (getCapabilities() != null) p_187471_.put("ForgeCaps", serializeCaps());
    }
 
    public final CompoundTag saveWithFullMetadata() {
@@ -68,6 +_,11 @@
    public final CompoundTag saveWithoutMetadata() {
       CompoundTag compoundtag = new CompoundTag();
       this.saveAdditional(compoundtag);
+      // CraftBukkit start - store container
+      if (this.persistentDataContainer != null && !this.persistentDataContainer.isEmpty()) {
+         compoundtag.put("PublicBukkitValues", this.persistentDataContainer.toTagCompound());
+      }
+      // CraftBukkit end
       return compoundtag;
    }
 
@@ -163,6 +_,13 @@
 
    public void setRemoved() {
       this.remove = true;
+      this.invalidateCaps();
+      requestModelDataUpdate();
+   }
+
+   @Override
+   public void onChunkUnloaded() {
+      this.invalidateCaps();
    }
 
    public void clearRemoved() {
@@ -191,9 +_,22 @@
       return this.type;
    }
 
+   @Override
+   public CompoundTag getPersistentData() {
+      if (this.customPersistentData == null)
+         this.customPersistentData = new CompoundTag();
+      return this.customPersistentData;
+   }
+
    /** @deprecated */
    @Deprecated
    public void setBlockState(BlockState p_155251_) {
       this.blockState = p_155251_;
    }
+
+   // CraftBukkit start - add method
+   public InventoryHolder getOwner() {
+      return InventoryOwner.get(this);
+   }
+   // CraftBukkit end
 }
