--- a/net/minecraft/world/level/block/entity/DecoratedPotPatterns.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotPatterns.java
@@ -10,32 +_,58 @@
 import net.minecraft.world.item.Items;
 
 public class DecoratedPotPatterns {
-    public static final ResourceKey<DecoratedPotPattern> BLANK = create("blank");
-    public static final ResourceKey<DecoratedPotPattern> ANGLER = create("angler");
-    public static final ResourceKey<DecoratedPotPattern> ARCHER = create("archer");
-    public static final ResourceKey<DecoratedPotPattern> ARMS_UP = create("arms_up");
-    public static final ResourceKey<DecoratedPotPattern> BLADE = create("blade");
-    public static final ResourceKey<DecoratedPotPattern> BREWER = create("brewer");
-    public static final ResourceKey<DecoratedPotPattern> BURN = create("burn");
-    public static final ResourceKey<DecoratedPotPattern> DANGER = create("danger");
-    public static final ResourceKey<DecoratedPotPattern> EXPLORER = create("explorer");
-    public static final ResourceKey<DecoratedPotPattern> FLOW = create("flow");
-    public static final ResourceKey<DecoratedPotPattern> FRIEND = create("friend");
-    public static final ResourceKey<DecoratedPotPattern> GUSTER = create("guster");
-    public static final ResourceKey<DecoratedPotPattern> HEART = create("heart");
-    public static final ResourceKey<DecoratedPotPattern> HEARTBREAK = create("heartbreak");
-    public static final ResourceKey<DecoratedPotPattern> HOWL = create("howl");
-    public static final ResourceKey<DecoratedPotPattern> MINER = create("miner");
-    public static final ResourceKey<DecoratedPotPattern> MOURNER = create("mourner");
-    public static final ResourceKey<DecoratedPotPattern> PLENTY = create("plenty");
-    public static final ResourceKey<DecoratedPotPattern> PRIZE = create("prize");
-    public static final ResourceKey<DecoratedPotPattern> SCRAPE = create("scrape");
-    public static final ResourceKey<DecoratedPotPattern> SHEAF = create("sheaf");
-    public static final ResourceKey<DecoratedPotPattern> SHELTER = create("shelter");
-    public static final ResourceKey<DecoratedPotPattern> SKULL = create("skull");
-    public static final ResourceKey<DecoratedPotPattern> SNORT = create("snort");
-    private static final Map<Item, ResourceKey<DecoratedPotPattern>> ITEM_TO_POT_TEXTURE = Map.ofEntries(
-        Map.entry(Items.BRICK, BLANK),
+    private static final String BASE_NAME = "decorated_pot_base";
+    public static final ResourceKey<String> BASE = create("decorated_pot_base");
+    private static final String BRICK_NAME = "decorated_pot_side";
+    private static final String ANGLER_NAME = "angler_pottery_pattern";
+    private static final String ARCHER_NAME = "archer_pottery_pattern";
+    private static final String ARMS_UP_NAME = "arms_up_pottery_pattern";
+    private static final String BLADE_NAME = "blade_pottery_pattern";
+    private static final String BREWER_NAME = "brewer_pottery_pattern";
+    private static final String BURN_NAME = "burn_pottery_pattern";
+    private static final String DANGER_NAME = "danger_pottery_pattern";
+    private static final String EXPLORER_NAME = "explorer_pottery_pattern";
+    private static final String FLOW_NAME = "flow_pottery_pattern";
+    private static final String FRIEND_NAME = "friend_pottery_pattern";
+    private static final String GUSTER_NAME = "guster_pottery_pattern";
+    private static final String HEART_NAME = "heart_pottery_pattern";
+    private static final String HEARTBREAK_NAME = "heartbreak_pottery_pattern";
+    private static final String HOWL_NAME = "howl_pottery_pattern";
+    private static final String MINER_NAME = "miner_pottery_pattern";
+    private static final String MOURNER_NAME = "mourner_pottery_pattern";
+    private static final String PLENTY_NAME = "plenty_pottery_pattern";
+    private static final String PRIZE_NAME = "prize_pottery_pattern";
+    private static final String SCRAPE_NAME = "scrape_pottery_pattern";
+    private static final String SHEAF_NAME = "sheaf_pottery_pattern";
+    private static final String SHELTER_NAME = "shelter_pottery_pattern";
+    private static final String SKULL_NAME = "skull_pottery_pattern";
+    private static final String SNORT_NAME = "snort_pottery_pattern";
+    private static final ResourceKey<String> BRICK = create("decorated_pot_side");
+    private static final ResourceKey<String> ANGLER = create("angler_pottery_pattern");
+    private static final ResourceKey<String> ARCHER = create("archer_pottery_pattern");
+    private static final ResourceKey<String> ARMS_UP = create("arms_up_pottery_pattern");
+    private static final ResourceKey<String> BLADE = create("blade_pottery_pattern");
+    private static final ResourceKey<String> BREWER = create("brewer_pottery_pattern");
+    private static final ResourceKey<String> BURN = create("burn_pottery_pattern");
+    private static final ResourceKey<String> DANGER = create("danger_pottery_pattern");
+    private static final ResourceKey<String> EXPLORER = create("explorer_pottery_pattern");
+    private static final ResourceKey<String> FLOW = create("flow_pottery_pattern");
+    private static final ResourceKey<String> FRIEND = create("friend_pottery_pattern");
+    private static final ResourceKey<String> GUSTER = create("guster_pottery_pattern");
+    private static final ResourceKey<String> HEART = create("heart_pottery_pattern");
+    private static final ResourceKey<String> HEARTBREAK = create("heartbreak_pottery_pattern");
+    private static final ResourceKey<String> HOWL = create("howl_pottery_pattern");
+    private static final ResourceKey<String> MINER = create("miner_pottery_pattern");
+    private static final ResourceKey<String> MOURNER = create("mourner_pottery_pattern");
+    private static final ResourceKey<String> PLENTY = create("plenty_pottery_pattern");
+    private static final ResourceKey<String> PRIZE = create("prize_pottery_pattern");
+    private static final ResourceKey<String> SCRAPE = create("scrape_pottery_pattern");
+    private static final ResourceKey<String> SHEAF = create("sheaf_pottery_pattern");
+    private static final ResourceKey<String> SHELTER = create("shelter_pottery_pattern");
+    private static final ResourceKey<String> SKULL = create("skull_pottery_pattern");
+    private static final ResourceKey<String> SNORT = create("snort_pottery_pattern");
+    private static final Map<Item, ResourceKey<String>> ITEM_TO_POT_TEXTURE = Map.ofEntries(
+        Map.entry(Items.BRICK, BRICK),
         Map.entry(Items.ANGLER_POTTERY_SHERD, ANGLER),
         Map.entry(Items.ARCHER_POTTERY_SHERD, ARCHER),
         Map.entry(Items.ARMS_UP_POTTERY_SHERD, ARMS_UP),
@@ -61,43 +_,44 @@
         Map.entry(Items.SNORT_POTTERY_SHERD, SNORT)
     );
 
+    private static ResourceKey<String> create(String p_272919_) {
+        return ResourceKey.create(Registries.DECORATED_POT_PATTERNS, new ResourceLocation(p_272919_));
+    }
+
+    public static ResourceLocation location(ResourceKey<String> p_273198_) {
+        return p_273198_.location().withPrefix("entity/decorated_pot/");
+    }
+
     @Nullable
-    public static ResourceKey<DecoratedPotPattern> getPatternFromItem(Item p_273094_) {
+    public static ResourceKey<String> getResourceKey(Item p_273094_) {
         return ITEM_TO_POT_TEXTURE.get(p_273094_);
     }
 
-    private static ResourceKey<DecoratedPotPattern> create(String p_272919_) {
-        return ResourceKey.create(Registries.DECORATED_POT_PATTERN, ResourceLocation.withDefaultNamespace(p_272919_));
-    }
-
-    public static DecoratedPotPattern bootstrap(Registry<DecoratedPotPattern> p_273479_) {
-        register(p_273479_, ANGLER, "angler_pottery_pattern");
-        register(p_273479_, ARCHER, "archer_pottery_pattern");
-        register(p_273479_, ARMS_UP, "arms_up_pottery_pattern");
-        register(p_273479_, BLADE, "blade_pottery_pattern");
-        register(p_273479_, BREWER, "brewer_pottery_pattern");
-        register(p_273479_, BURN, "burn_pottery_pattern");
-        register(p_273479_, DANGER, "danger_pottery_pattern");
-        register(p_273479_, EXPLORER, "explorer_pottery_pattern");
-        register(p_273479_, FLOW, "flow_pottery_pattern");
-        register(p_273479_, FRIEND, "friend_pottery_pattern");
-        register(p_273479_, GUSTER, "guster_pottery_pattern");
-        register(p_273479_, HEART, "heart_pottery_pattern");
-        register(p_273479_, HEARTBREAK, "heartbreak_pottery_pattern");
-        register(p_273479_, HOWL, "howl_pottery_pattern");
-        register(p_273479_, MINER, "miner_pottery_pattern");
-        register(p_273479_, MOURNER, "mourner_pottery_pattern");
-        register(p_273479_, PLENTY, "plenty_pottery_pattern");
-        register(p_273479_, PRIZE, "prize_pottery_pattern");
-        register(p_273479_, SCRAPE, "scrape_pottery_pattern");
-        register(p_273479_, SHEAF, "sheaf_pottery_pattern");
-        register(p_273479_, SHELTER, "shelter_pottery_pattern");
-        register(p_273479_, SKULL, "skull_pottery_pattern");
-        register(p_273479_, SNORT, "snort_pottery_pattern");
-        return register(p_273479_, BLANK, "decorated_pot_side");
-    }
-
-    private static DecoratedPotPattern register(Registry<DecoratedPotPattern> p_345377_, ResourceKey<DecoratedPotPattern> p_345043_, String p_344357_) {
-        return Registry.register(p_345377_, p_345043_, new DecoratedPotPattern(ResourceLocation.withDefaultNamespace(p_344357_)));
+    public static String bootstrap(Registry<String> p_273479_) {
+        Registry.register(p_273479_, BRICK, "decorated_pot_side");
+        Registry.register(p_273479_, ANGLER, "angler_pottery_pattern");
+        Registry.register(p_273479_, ARCHER, "archer_pottery_pattern");
+        Registry.register(p_273479_, ARMS_UP, "arms_up_pottery_pattern");
+        Registry.register(p_273479_, BLADE, "blade_pottery_pattern");
+        Registry.register(p_273479_, BREWER, "brewer_pottery_pattern");
+        Registry.register(p_273479_, BURN, "burn_pottery_pattern");
+        Registry.register(p_273479_, DANGER, "danger_pottery_pattern");
+        Registry.register(p_273479_, EXPLORER, "explorer_pottery_pattern");
+        Registry.register(p_273479_, FLOW, "flow_pottery_pattern");
+        Registry.register(p_273479_, FRIEND, "friend_pottery_pattern");
+        Registry.register(p_273479_, GUSTER, "guster_pottery_pattern");
+        Registry.register(p_273479_, HEART, "heart_pottery_pattern");
+        Registry.register(p_273479_, HEARTBREAK, "heartbreak_pottery_pattern");
+        Registry.register(p_273479_, HOWL, "howl_pottery_pattern");
+        Registry.register(p_273479_, MINER, "miner_pottery_pattern");
+        Registry.register(p_273479_, MOURNER, "mourner_pottery_pattern");
+        Registry.register(p_273479_, PLENTY, "plenty_pottery_pattern");
+        Registry.register(p_273479_, PRIZE, "prize_pottery_pattern");
+        Registry.register(p_273479_, SCRAPE, "scrape_pottery_pattern");
+        Registry.register(p_273479_, SHEAF, "sheaf_pottery_pattern");
+        Registry.register(p_273479_, SHELTER, "shelter_pottery_pattern");
+        Registry.register(p_273479_, SKULL, "skull_pottery_pattern");
+        Registry.register(p_273479_, SNORT, "snort_pottery_pattern");
+        return Registry.register(p_273479_, BASE, "decorated_pot_base");
     }
 }
