--- a/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
@@ -20,7 +_,7 @@
 
 public class DecoratedPotBlockEntity extends BlockEntity {
    public static final String TAG_SHERDS = "sherds";
-   private DecoratedPotBlockEntity.Decorations decorations = DecoratedPotBlockEntity.Decorations.EMPTY;
+   public DecoratedPotBlockEntity.Decorations decorations = DecoratedPotBlockEntity.Decorations.EMPTY;
 
    public DecoratedPotBlockEntity(BlockPos p_273660_, BlockState p_272831_) {
       super(BlockEntityType.DECORATED_POT, p_273660_, p_272831_);
