--- a/net/minecraft/world/level/block/entity/EnchantmentTableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/EnchantmentTableBlockEntity.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.level.block.entity;
 
 import javax.annotation.Nullable;
+
+import com.mohistmc.MohistConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
@@ -46,6 +_,7 @@
    }
 
    public static void m_155503_(Level p_155504_, BlockPos p_155505_, BlockState p_155506_, EnchantmentTableBlockEntity p_155507_) {
+      if (!MohistConfig.bookAnimationTick) return; // Mirai - skip enchantment table book tick
       p_155507_.f_59257_ = p_155507_.f_59256_;
       p_155507_.f_59259_ = p_155507_.f_59258_;
       Player player = p_155504_.m_45924_((double)p_155505_.m_123341_() + 0.5D, (double)p_155505_.m_123342_() + 0.5D, (double)p_155505_.m_123343_() + 0.5D, 3.0D, false);
