--- a/net/minecraft/world/level/block/entity/EnchantmentTableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/EnchantmentTableBlockEntity.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.level.block.entity;
 
 import javax.annotation.Nullable;
+
+import com.mohistmc.MohistConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
@@ -46,6 +_,7 @@
    }
 
    public static void bookAnimationTick(Level p_155504_, BlockPos p_155505_, BlockState p_155506_, EnchantmentTableBlockEntity p_155507_) {
+      if (!MohistConfig.bookAnimationTick) return; // Mirai - skip enchantment table book tick
       p_155507_.oOpen = p_155507_.open;
       p_155507_.oRot = p_155507_.rot;
       Player player = p_155504_.getNearestPlayer((double)p_155505_.getX() + 0.5D, (double)p_155505_.getY() + 0.5D, (double)p_155505_.getZ() + 0.5D, 3.0D, false);
