--- a/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HangingSignBlockEntity.java
@@ -4,7 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class HangingSignBlockEntity extends SignBlockEntity {
-   private static final int f_244379_ = 60;
+   private static final int f_244379_ = 50;
    private static final int f_244305_ = 9;
 
    public HangingSignBlockEntity(BlockPos p_250603_, BlockState p_251674_) {
@@ -16,6 +_,6 @@
    }
 
    public int m_245123_() {
-      return 60;
+      return 50;
    }
 }
