--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -8,6 +_,10 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_20_R3.command.CraftRemoteConsoleCommandSender;
+
+import java.net.SocketAddress;
 
 public class RconConsoleSource implements CommandSource {
    private static final String RCON = "Rcon";
@@ -15,8 +_,14 @@
    private final StringBuffer buffer = new StringBuffer();
    private final MinecraftServer server;
 
-   public RconConsoleSource(MinecraftServer p_11505_) {
-      this.server = p_11505_;
+   // CraftBukkit start
+   public final SocketAddress socketAddress;
+   private final CraftRemoteConsoleCommandSender remoteConsole = new CraftRemoteConsoleCommandSender(this);
+
+   public RconConsoleSource(MinecraftServer pServer, SocketAddress socketAddress) {
+      this.socketAddress = socketAddress;
+      // CraftBukkit end
+      this.server = pServer;
    }
 
    public void prepareForCommand() {
@@ -33,7 +_,7 @@
    }
 
    public void sendSystemMessage(Component p_215653_) {
-      this.buffer.append(p_215653_.getString());
+      this.buffer.append(p_215653_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean acceptsSuccess() {
@@ -47,4 +_,16 @@
    public boolean shouldInformAdmins() {
       return this.server.shouldRconBroadcast();
    }
+
+
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.buffer.append(message);
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return this.remoteConsole;
+   }
+   // CraftBukkit end
 }
