--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.rcon;
 
+import java.net.InetSocketAddress;
 import net.minecraft.commands.CommandSource;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.network.chat.Component;
@@ -8,6 +_,10 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_20_R1.command.CraftRemoteConsoleCommandSender;
+
+import java.net.SocketAddress;
 
 public class RconConsoleSource implements CommandSource {
    private static final String f_144022_ = "Rcon";
@@ -15,8 +_,19 @@
    private final StringBuffer f_11501_ = new StringBuffer();
    private final MinecraftServer f_11502_;
 
+   // CraftBukkit start
+   public final SocketAddress socketAddress;
+   private final CraftRemoteConsoleCommandSender remoteConsole = new CraftRemoteConsoleCommandSender(this);
+
    public RconConsoleSource(MinecraftServer p_11505_) {
       this.f_11502_ = p_11505_;
+      this.socketAddress = new InetSocketAddress(p_11505_.rconPort);
+   }
+
+   public RconConsoleSource(MinecraftServer pServer, SocketAddress socketAddress) {
+      this.socketAddress = socketAddress;
+      // CraftBukkit end
+      this.f_11502_ = pServer;
    }
 
    public void m_11512_() {
@@ -33,7 +_,7 @@
    }
 
    public void m_213846_(Component p_215653_) {
-      this.f_11501_.append(p_215653_.getString());
+      this.f_11501_.append(p_215653_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean m_6999_() {
@@ -47,4 +_,16 @@
    public boolean m_6102_() {
       return this.f_11502_.m_6983_();
    }
+
+
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.f_11501_.append(message);
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return this.remoteConsole;
+   }
+   // CraftBukkit end
 }
