--- a/net/minecraft/world/entity/npc/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/WanderingTraderSpawner.java
@@ -103,7 +_,7 @@
             if (p_35916_.getBiome(blockpos2).is(BiomeTags.WITHOUT_WANDERING_TRADER_SPAWNS)) {
                return false;
             }
-
+            EntityType.WANDERING_TRADER.spawnReasonAtomicReference.set(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit // Mohist
             WanderingTrader wanderingtrader = EntityType.WANDERING_TRADER.spawn(p_35916_, blockpos2, MobSpawnType.EVENT);
             if (wanderingtrader != null) {
                for(int j = 0; j < 2; ++j) {
@@ -125,6 +_,7 @@
    private void tryToSpawnLlamaFor(ServerLevel p_35918_, WanderingTrader p_35919_, int p_35920_) {
       BlockPos blockpos = this.findSpawnPositionNear(p_35918_, p_35919_.blockPosition(), p_35920_);
       if (blockpos != null) {
+         EntityType.TRADER_LLAMA.spawnReasonAtomicReference.set(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit // Mohist
          TraderLlama traderllama = EntityType.TRADER_LLAMA.spawn(p_35918_, blockpos, MobSpawnType.EVENT);
          if (traderllama != null) {
             traderllama.setLeashedTo(p_35919_, true);
