--- a/net/minecraft/world/level/gameevent/GameEventDispatcher.java
+++ b/net/minecraft/world/level/gameevent/GameEventDispatcher.java
@@ -9,6 +_,10 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R2.CraftGameEvent;
+import org.bukkit.craftbukkit.v1_20_R2.util.CraftLocation;
+import org.bukkit.event.world.GenericGameEvent;
 
 public class GameEventDispatcher {
    private final ServerLevel level;
@@ -20,6 +_,14 @@
    public void post(GameEvent p_251754_, Vec3 p_250613_, GameEvent.Context p_251777_) {
       int i = p_251754_.getNotificationRadius();
       BlockPos blockpos = BlockPos.containing(p_250613_);
+      // CraftBukkit start
+      GenericGameEvent event = new GenericGameEvent(CraftGameEvent.minecraftToBukkit(p_251754_), CraftLocation.toBukkit(blockpos, level.getWorld()), (p_251777_.sourceEntity() == null) ? null : p_251777_.sourceEntity().getBukkitEntity(), i, !Bukkit.isPrimaryThread());
+      level.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      i = event.getRadius();
+      // CraftBukkit end
       int j = SectionPos.blockToSectionCoord(blockpos.getX() - i);
       int k = SectionPos.blockToSectionCoord(blockpos.getY() - i);
       int l = SectionPos.blockToSectionCoord(blockpos.getZ() - i);
