--- a/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicate.java
+++ b/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicate.java
@@ -130,12 +_,4 @@
     static BlockPredicate alwaysTrue() {
         return TrueBlockPredicate.INSTANCE;
     }
-
-    static BlockPredicate unobstructed(Vec3i p_344704_) {
-        return new UnobstructedPredicate(p_344704_);
-    }
-
-    static BlockPredicate unobstructed() {
-        return unobstructed(Vec3i.ZERO);
-    }
 }
