--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -487,7 +_,8 @@
 
    }
 
-   private void respawnDragon(List<EndCrystal> p_64092_) {
+   public boolean respawnDragon = false;
+   public void respawnDragon(List<EndCrystal> p_64092_) {
       if (this.dragonKilled && this.respawnStage == null) {
          for(BlockPattern.BlockPatternMatch blockpattern$blockpatternmatch = this.findExitPortal(); blockpattern$blockpatternmatch != null; blockpattern$blockpatternmatch = this.findExitPortal()) {
             for(int i = 0; i < this.exitPortalPattern.getWidth(); ++i) {
@@ -506,6 +_,7 @@
          this.respawnTime = 0;
          this.spawnExitPortal(false);
          this.respawnCrystals = p_64092_;
+         respawnDragon = true;
       }
 
    }
@@ -517,7 +_,14 @@
             endcrystal.setBeamTarget((BlockPos)null);
          }
       }
-
+   }
+
+   public void addPlayer(ServerPlayer player) {
+      this.dragonEvent.addPlayer(player);
+   }
+
+   public void removePlayer(ServerPlayer player) {
+      this.dragonEvent.removePlayer(player);
    }
 
    @Nullable
