--- ../src-base/minecraft/net/minecraft/advancements/AdvancementList.java
+++ ../src-work/minecraft/net/minecraft/advancements/AdvancementList.java
@@ -18,7 +18,7 @@
 public class AdvancementList
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private final Map<ResourceLocation, Advancement> advancements = Maps.<ResourceLocation, Advancement>newHashMap();
+    public final Map<ResourceLocation, Advancement> advancements = Maps.<ResourceLocation, Advancement>newHashMap();
     private final Set<Advancement> roots = Sets.<Advancement>newLinkedHashSet();
     private final Set<Advancement> nonRoots = Sets.<Advancement>newLinkedHashSet();
     private AdvancementList.Listener listener;
@@ -128,12 +128,12 @@
                     }
 
                     Entry<ResourceLocation, Advancement.Builder> entry1 = (Entry)iterator.next();
-                    LOGGER.error("Couldn't load advancement " + entry1.getKey() + ": " + entry1.getValue());
+                    LOGGER.debug("Couldn't load advancement " + entry1.getKey() + ": " + entry1.getValue());
                 }
             }
         }
 
-        LOGGER.info("Loaded " + this.advancements.size() + " advancements");
+       // LOGGER.info("Loaded " + this.advancements.size() + " advancements"); // CraftBukkit - moved to AdvancementDataWorld#reload
     }
 
     public void clear()
