--- ../src-base/minecraft/net/minecraft/advancements/FunctionManager.java
+++ ../src-work/minecraft/net/minecraft/advancements/FunctionManager.java
@@ -29,8 +29,12 @@
     private FunctionObject gameLoopFunction;
     private final ArrayDeque<FunctionManager.QueuedCommand> commandQueue = new ArrayDeque<FunctionManager.QueuedCommand>();
     private boolean isExecuting = false;
-    private final ICommandSender gameLoopFunctionSender = new ICommandSender()
+    private final ICommandSender gameLoopFunctionSender = new CustomFunctionListener();
+
+    public class CustomFunctionListener implements ICommandSender
     {
+
+        public org.bukkit.command.CommandSender sender = new org.bukkit.craftbukkit.v1_12_R1.command.CraftFunctionCommandSender(this);
         public String getName()
         {
             return FunctionManager.this.currentGameLoopFunctionId;
@@ -39,9 +43,10 @@
         {
             return permLevel <= 2;
         }
+
         public World getEntityWorld()
         {
-            return FunctionManager.this.server.worlds[0];
+            return FunctionManager.this.server.worldServerList.get(0);
         }
         public MinecraftServer getServer()
         {
@@ -69,7 +74,7 @@
 
     public int getMaxCommandChainLength()
     {
-        return this.server.worlds[0].getGameRules().getInt("maxCommandChainLength");
+        return this.server.worldServerList.get(0).getGameRules().getInt("maxCommandChainLength");
     }
 
     public Map<ResourceLocation, FunctionObject> getFunctions()
@@ -79,7 +84,7 @@
 
     public void update()
     {
-        String s = this.server.worlds[0].getGameRules().getString("gameLoopFunction");
+        String s = this.server.worldServerList.get(0).getGameRules().getString("gameLoopFunction");
 
         if (!s.equals(this.currentGameLoopFunctionId))
         {
