--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
@@ -4,7 +_,6 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.Set;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -36,10 +_,14 @@
         Entity entity = p_81967_.getParamOrNull(LootContextParams.KILLER_ENTITY);
         int i = 0;
         if (entity instanceof LivingEntity) {
-            i = EnchantmentHelper.getMobLooting((LivingEntity)entity);
-        }
-
-        return p_81967_.getRandom().nextFloat() < this.percent + (float)i * this.lootingMultiplier;
+            i = EnchantmentHelper.getMobLooting((LivingEntity) entity);
+        }
+        // CraftBukkit start - only use lootingModifier if set by Bukkit
+        if (p_81967_.hasParam(LootContextParams.LOOTING_MOD)) {
+            i = p_81967_.getParamOrNull(LootContextParams.LOOTING_MOD);
+        }
+        // CraftBukkit end
+        return p_81967_.getRandom().nextFloat() < this.percent + (float) i * this.lootingMultiplier;
     }
 
     public static LootItemCondition.Builder randomChanceAndLootingBoost(float p_81964_, float p_81965_) {
