--- a/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacerType.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacerType.java
@@ -12,7 +_,7 @@
         return Registry.register(BuiltInRegistries.ROOT_PLACER_TYPE, p_225905_, new RootPlacerType<>(p_329098_));
     }
 
-    private RootPlacerType(MapCodec<P> p_333995_) {
+    public RootPlacerType(MapCodec<P> p_333995_) {
         this.codec = p_333995_;
     }
 
