--- a/net/minecraft/commands/arguments/MessageArgument.java
+++ b/net/minecraft/commands/arguments/MessageArgument.java
@@ -29,8 +_,12 @@
    }
 
    public static Component m_96835_(CommandContext<CommandSourceStack> p_96836_, String p_96837_) throws CommandSyntaxException {
-      MessageArgument.Message messageargument$message = p_96836_.getArgument(p_96837_, MessageArgument.Message.class);
-      return messageargument$message.m_232196_(p_96836_.getSource());
+      try {
+         MessageArgument.Message messageargument$message = p_96836_.getArgument(p_96837_, MessageArgument.Message.class);
+         return messageargument$message.m_232196_(p_96836_.getSource());
+      } catch (IllegalArgumentException e) {
+         return Component.m_237119_();
+      }
    }
 
    public static void m_245478_(CommandContext<CommandSourceStack> p_249433_, String p_248718_, Consumer<PlayerChatMessage> p_249460_) throws CommandSyntaxException {
@@ -50,10 +_,10 @@
    private static void m_247736_(Consumer<PlayerChatMessage> p_250000_, CommandSourceStack p_252335_, PlayerChatMessage p_249420_) {
       MinecraftServer minecraftserver = p_252335_.m_81377_();
       CompletableFuture<FilteredText> completablefuture = m_246606_(p_252335_, p_249420_);
-      CompletableFuture<Component> completablefuture1 = minecraftserver.m_236742_().m_236961_(p_252335_.m_230896_(), p_249420_.m_245692_());
+      CompletableFuture<net.minecraft.network.chat.ChatDecorator.Result> completablefuture1 = minecraftserver.m_236742_().decorate(p_252335_.m_230896_(), p_252335_, p_249420_.m_245692_());
       p_252335_.m_241923_().m_241849_((p_247979_) -> {
          return CompletableFuture.allOf(completablefuture, completablefuture1).thenAcceptAsync((p_247970_) -> {
-            PlayerChatMessage playerchatmessage = p_249420_.m_241956_(completablefuture1.join()).m_243072_(completablefuture.join().f_243010_());
+            PlayerChatMessage playerchatmessage = p_249420_.m_241956_(completablefuture1.join().component()).m_243072_(completablefuture.join().f_243010_());
             p_250000_.accept(playerchatmessage);
          }, p_247979_);
       });
@@ -61,10 +_,10 @@
 
    private static void m_247098_(Consumer<PlayerChatMessage> p_249162_, CommandSourceStack p_248759_, PlayerChatMessage p_252332_) {
       MinecraftServer minecraftserver = p_248759_.m_81377_();
-      CompletableFuture<Component> completablefuture = minecraftserver.m_236742_().m_236961_(p_248759_.m_230896_(), p_252332_.m_245692_());
+      CompletableFuture<net.minecraft.network.chat.ChatDecorator.Result> completablefuture = minecraftserver.m_236742_().decorate(p_248759_.m_230896_(), p_248759_, p_252332_.m_245692_());
       p_248759_.m_241923_().m_241849_((p_247974_) -> {
          return completablefuture.thenAcceptAsync((p_247965_) -> {
-            p_249162_.accept(p_252332_.m_241956_(p_247965_));
+            p_249162_.accept(p_252332_.m_241956_(p_247965_.component()));
          }, p_247974_);
       });
    }
@@ -100,7 +_,7 @@
       }
 
       Component m_232196_(CommandSourceStack p_232197_) throws CommandSyntaxException {
-         return this.m_96849_(p_232197_, p_232197_.m_6761_(2));
+         return this.m_96849_(p_232197_, net.minecraftforge.common.ForgeHooks.canUseEntitySelectors(p_232197_));
       }
 
       public Component m_96849_(CommandSourceStack p_96850_, boolean p_96851_) throws CommandSyntaxException {
