--- a/net/minecraft/item/BucketItem.java
+++ b/net/minecraft/item/BucketItem.java
@@ -11,33 +11,49 @@
 import net.minecraft.fluid.FlowingFluid;
 import net.minecraft.fluid.Fluid;
 import net.minecraft.fluid.Fluids;
+import net.minecraft.network.play.server.SChangeBlockPacket;
 import net.minecraft.particles.ParticleTypes;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.FluidTags;
-import net.minecraft.util.ActionResult;
-import net.minecraft.util.Direction;
-import net.minecraft.util.Hand;
-import net.minecraft.util.SoundCategory;
-import net.minecraft.util.SoundEvent;
-import net.minecraft.util.SoundEvents;
+import net.minecraft.util.*;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.BlockRayTraceResult;
 import net.minecraft.util.math.RayTraceContext;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_15_R1.entity.CraftItem;
+import org.bukkit.craftbukkit.v1_15_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_15_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_15_R1.util.DummyGeneratorAccess;
+import org.bukkit.event.player.PlayerBucketEmptyEvent;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class BucketItem extends Item {
    private final Fluid field_77876_a;
 
+   // Forge: Use the other constructor that takes a Supplier
+   @Deprecated
    public BucketItem(Fluid p_i49025_1_, Item.Properties p_i49025_2_) {
       super(p_i49025_2_);
       this.field_77876_a = p_i49025_1_;
+      this.fluidSupplier = p_i49025_1_.delegate;
    }
 
+   /**
+    * @param supplier A fluid supplier such as {@link net.minecraftforge.fml.RegistryObject<Fluid>}
+    */
+   public BucketItem(java.util.function.Supplier<? extends Fluid> supplier, Item.Properties builder) {
+      super(builder);
+      this.field_77876_a = null;
+      this.fluidSupplier = supplier;
+   }
+
    public ActionResult<ItemStack> func_77659_a(World p_77659_1_, PlayerEntity p_77659_2_, Hand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
       RayTraceResult raytraceresult = func_219968_a(p_77659_1_, p_77659_2_, this.field_77876_a == Fluids.field_204541_a ? RayTraceContext.FluidMode.SOURCE_ONLY : RayTraceContext.FluidMode.NONE);
+      ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(p_77659_2_, p_77659_1_, itemstack, raytraceresult);
+      if (ret != null) return ret;
       if (raytraceresult.func_216346_c() == RayTraceResult.Type.MISS) {
          return ActionResult.func_226250_c_(itemstack);
       } else if (raytraceresult.func_216346_c() != RayTraceResult.Type.BLOCK) {
@@ -51,11 +67,23 @@
             if (this.field_77876_a == Fluids.field_204541_a) {
                BlockState blockstate1 = p_77659_1_.func_180495_p(blockpos);
                if (blockstate1.func_177230_c() instanceof IBucketPickupHandler) {
+                  // CraftBukkit start
+                  Fluid dummyFluid = ((IBucketPickupHandler) blockstate1.func_177230_c()).func_204508_a(DummyGeneratorAccess.INSTANCE, blockpos, blockstate1);
+                  PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_77659_1_, p_77659_2_, blockpos, blockpos, ((BlockRayTraceResult) raytraceresult).func_216354_b(), itemstack, dummyFluid.func_204524_b());
+                  if (event.isCancelled()) {
+                     ((ServerPlayerEntity) p_77659_2_).field_71135_a.func_147359_a(new SChangeBlockPacket(p_77659_1_, blockpos)); // SPIGOT-5163 (see PlayerInteractManager)
+                     ((ServerPlayerEntity) p_77659_2_).getBukkitEntity().updateInventory(); // SPIGOT-4541
+                     return new ActionResult<>(ActionResultType.FAIL, itemstack);
+                  }
+                  // CraftBukkit end
                   Fluid fluid = ((IBucketPickupHandler)blockstate1.func_177230_c()).func_204508_a(p_77659_1_, blockpos, blockstate1);
                   if (fluid != Fluids.field_204541_a) {
                      p_77659_2_.func_71029_a(Stats.field_75929_E.func_199076_b(this));
-                     p_77659_2_.func_184185_a(fluid.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187633_N : SoundEvents.field_187630_M, 1.0F, 1.0F);
-                     ItemStack itemstack1 = this.func_150910_a(itemstack, p_77659_2_, fluid.func_204524_b());
+
+                     SoundEvent soundevent = this.field_77876_a.getAttributes().getFillSound();
+                     if (soundevent == null) soundevent = fluid.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187633_N : SoundEvents.field_187630_M;
+                     p_77659_2_.func_184185_a(soundevent, 1.0F, 1.0F);
+                     ItemStack itemstack1 = this.fillBucket(itemstack, p_77659_2_, fluid.func_204524_b(), event.getItemStack()); // CraftBukkit
                      if (!p_77659_1_.field_72995_K) {
                         CriteriaTriggers.field_204813_j.func_204817_a((ServerPlayerEntity)p_77659_2_, new ItemStack(fluid.func_204524_b()));
                      }
@@ -67,8 +95,8 @@
                return ActionResult.func_226251_d_(itemstack);
             } else {
                BlockState blockstate = p_77659_1_.func_180495_p(blockpos);
-               BlockPos blockpos2 = blockstate.func_177230_c() instanceof ILiquidContainer && this.field_77876_a == Fluids.field_204546_a ? blockpos : blockpos1;
-               if (this.func_180616_a(p_77659_2_, p_77659_1_, blockpos2, blockraytraceresult)) {
+               BlockPos blockpos2 = canBlockContainFluid(p_77659_1_, blockpos, blockstate) ? blockpos : blockpos1;
+               if (this.tryPlaceContainedLiquid(p_77659_2_, p_77659_1_, blockpos2, blockraytraceresult, blockraytraceresult.func_216354_b(), blockpos, itemstack)) { // CraftBukkit
                   this.func_203792_a(p_77659_1_, itemstack, blockpos2);
                   if (p_77659_2_ instanceof ServerPlayerEntity) {
                      CriteriaTriggers.field_193137_x.func_193173_a((ServerPlayerEntity)p_77659_2_, blockpos2, itemstack);
@@ -93,62 +121,98 @@
    public void func_203792_a(World p_203792_1_, ItemStack p_203792_2_, BlockPos p_203792_3_) {
    }
 
-   private ItemStack func_150910_a(ItemStack p_150910_1_, PlayerEntity p_150910_2_, Item p_150910_3_) {
-      if (p_150910_2_.field_71075_bZ.field_75098_d) {
-         return p_150910_1_;
+   // CraftBukkit - added ob.ItemStack result - TODO: Is this... the right way to handle this?
+   private ItemStack fillBucket(ItemStack emptyBuckets, PlayerEntity player, Item fullBucket, org.bukkit.inventory.ItemStack result) {
+      if (player.field_71075_bZ.field_75098_d) {
+         return emptyBuckets;
       } else {
-         p_150910_1_.func_190918_g(1);
-         if (p_150910_1_.func_190926_b()) {
-            return new ItemStack(p_150910_3_);
+         emptyBuckets.func_190918_g(1);
+         if (emptyBuckets.func_190926_b()) {
+            // CraftBukkit start
+            return CraftItemStack.asNMSCopy(result);
          } else {
-            if (!p_150910_2_.field_71071_by.func_70441_a(new ItemStack(p_150910_3_))) {
-               p_150910_2_.func_71019_a(new ItemStack(p_150910_3_), false);
+            if (!player.field_71071_by.func_70441_a(CraftItemStack.asNMSCopy(result))) {
+               player.func_71019_a(CraftItemStack.asNMSCopy(result), false);
+               // CraftBukkit end
             }
 
-            return p_150910_1_;
+            return emptyBuckets;
          }
       }
    }
 
+   // CraftBukkit start
    public boolean func_180616_a(@Nullable PlayerEntity p_180616_1_, World p_180616_2_, BlockPos p_180616_3_, @Nullable BlockRayTraceResult p_180616_4_) {
+      return tryPlaceContainedLiquid(p_180616_1_, p_180616_2_, p_180616_3_, p_180616_4_, null, null, null);
+   }
+   public boolean tryPlaceContainedLiquid(@Nullable PlayerEntity player, World worldIn, BlockPos posIn, @Nullable BlockRayTraceResult p_180616_4_, Direction enumdirection, BlockPos clicked, ItemStack itemstack) {
+      // CraftBukkit end
       if (!(this.field_77876_a instanceof FlowingFluid)) {
          return false;
       } else {
-         BlockState blockstate = p_180616_2_.func_180495_p(p_180616_3_);
+         BlockState blockstate = worldIn.func_180495_p(posIn);
          Material material = blockstate.func_185904_a();
          boolean flag = blockstate.func_227032_a_(this.field_77876_a);
-         if (blockstate.func_196958_f() || flag || blockstate.func_177230_c() instanceof ILiquidContainer && ((ILiquidContainer)blockstate.func_177230_c()).func_204510_a(p_180616_2_, p_180616_3_, blockstate, this.field_77876_a)) {
-            if (p_180616_2_.field_73011_w.func_177500_n() && this.field_77876_a.func_207185_a(FluidTags.field_206959_a)) {
-               int i = p_180616_3_.func_177958_n();
-               int j = p_180616_3_.func_177956_o();
-               int k = p_180616_3_.func_177952_p();
-               p_180616_2_.func_184133_a(p_180616_1_, p_180616_3_, SoundEvents.field_187646_bt, SoundCategory.BLOCKS, 0.5F, 2.6F + (p_180616_2_.field_73012_v.nextFloat() - p_180616_2_.field_73012_v.nextFloat()) * 0.8F);
+         boolean canContainFluid = canBlockContainFluid(worldIn, posIn, blockstate);
+         if (blockstate.func_196958_f() || flag || canContainFluid) {
+            // CraftBukkit start
+            if(player != null){
+               PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(worldIn, player, posIn, clicked, enumdirection, itemstack);
+               if(event.isCancelled()){
+                  ((ServerPlayerEntity) player).field_71135_a.func_147359_a(new SChangeBlockPacket(worldIn, posIn)); // SPIGOT-4238: needed when looking through entity
+                  ((ServerPlayerEntity) player).getBukkitEntity().updateInventory(); // SPIGOT-4541
+                  return false;
+               }
+            }
+            // CraftBukkit end
+            if (worldIn.field_73011_w.func_177500_n() && this.field_77876_a.func_207185_a(FluidTags.field_206959_a)) {
+               int i = posIn.func_177958_n();
+               int j = posIn.func_177956_o();
+               int k = posIn.func_177952_p();
+               worldIn.func_184133_a(player, posIn, SoundEvents.field_187646_bt, SoundCategory.BLOCKS, 0.5F, 2.6F + (worldIn.field_73012_v.nextFloat() - worldIn.field_73012_v.nextFloat()) * 0.8F);
 
                for(int l = 0; l < 8; ++l) {
-                  p_180616_2_.func_195594_a(ParticleTypes.field_197594_E, (double)i + Math.random(), (double)j + Math.random(), (double)k + Math.random(), 0.0D, 0.0D, 0.0D);
+                  worldIn.func_195594_a(ParticleTypes.field_197594_E, (double)i + Math.random(), (double)j + Math.random(), (double)k + Math.random(), 0.0D, 0.0D, 0.0D);
                }
-            } else if (blockstate.func_177230_c() instanceof ILiquidContainer && this.field_77876_a == Fluids.field_204546_a) {
-               if (((ILiquidContainer)blockstate.func_177230_c()).func_204509_a(p_180616_2_, p_180616_3_, blockstate, ((FlowingFluid)this.field_77876_a).func_207204_a(false))) {
-                  this.func_203791_b(p_180616_1_, p_180616_2_, p_180616_3_);
+            } else if (canContainFluid) {
+               if (((ILiquidContainer)blockstate.func_177230_c()).func_204509_a(worldIn, posIn, blockstate, ((FlowingFluid)this.field_77876_a).func_207204_a(false))) {
+                  this.func_203791_b(player, worldIn, posIn);
                }
             } else {
-               if (!p_180616_2_.field_72995_K && flag && !material.func_76224_d()) {
-                  p_180616_2_.func_175655_b(p_180616_3_, true);
+               if (!worldIn.field_72995_K && flag && !material.func_76224_d()) {
+                  worldIn.func_175655_b(posIn, true);
                }
 
-               this.func_203791_b(p_180616_1_, p_180616_2_, p_180616_3_);
-               p_180616_2_.func_180501_a(p_180616_3_, this.field_77876_a.func_207188_f().func_206883_i(), 11);
+               this.func_203791_b(player, worldIn, posIn);
+               worldIn.func_180501_a(posIn, this.field_77876_a.func_207188_f().func_206883_i(), 11);
             }
 
             return true;
          } else {
-            return p_180616_4_ == null ? false : this.func_180616_a(p_180616_1_, p_180616_2_, p_180616_4_.func_216350_a().func_177972_a(p_180616_4_.func_216354_b()), (BlockRayTraceResult)null);
+            return p_180616_4_ == null ? false : this.tryPlaceContainedLiquid(player, worldIn, p_180616_4_.func_216350_a().func_177972_a(p_180616_4_.func_216354_b()), (BlockRayTraceResult)null, enumdirection, clicked, itemstack); // CraftBukkit
          }
       }
    }
 
    protected void func_203791_b(@Nullable PlayerEntity p_203791_1_, IWorld p_203791_2_, BlockPos p_203791_3_) {
-      SoundEvent soundevent = this.field_77876_a.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187627_L : SoundEvents.field_187624_K;
+      SoundEvent soundevent = this.field_77876_a.getAttributes().getEmptySound();
+      if(soundevent == null) soundevent = this.field_77876_a.func_207185_a(FluidTags.field_206960_b) ? SoundEvents.field_187627_L : SoundEvents.field_187624_K;
       p_203791_2_.func_184133_a(p_203791_1_, p_203791_3_, soundevent, SoundCategory.BLOCKS, 1.0F, 1.0F);
    }
+
+   @Override
+   public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.CompoundNBT nbt) {
+      if (this.getClass() == BucketItem.class)
+         return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+      else
+         return super.initCapabilities(stack, nbt);
+   }
+
+   private final java.util.function.Supplier<? extends Fluid> fluidSupplier;
+   public Fluid getFluid() { return fluidSupplier.get(); }
+
+   private boolean canBlockContainFluid(World worldIn, BlockPos posIn, BlockState blockstate)
+   {
+      return blockstate.func_177230_c() instanceof ILiquidContainer && ((ILiquidContainer)blockstate.func_177230_c()).func_204510_a(worldIn, posIn, blockstate, this.field_77876_a);
+   }
 }
