--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/CoreModManager.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/CoreModManager.java
@@ -72,7 +72,7 @@
 public class CoreModManager {
     private static final Attributes.Name COREMODCONTAINSFMLMOD = new Attributes.Name("FMLCorePluginContainsFMLMod");
     private static final Attributes.Name MODTYPE = new Attributes.Name("ModType");
-    private static String[] rootPlugins = { "net.minecraftforge.fml.relauncher.FMLCorePlugin", "net.minecraftforge.classloading.FMLForgePlugin" };
+    private static String[] rootPlugins = { "net.minecraftforge.fml.relauncher.FMLCorePlugin", "net.minecraftforge.classloading.FMLForgePlugin", "red.mohist.forge.MohistCorePlugin" };
     private static List<String> ignoredModFiles = Lists.newArrayList();
     private static Map<String, List<String>> transformers = Maps.newHashMap();
     private static List<FMLPluginWrapper> loadPlugins;
@@ -223,7 +223,7 @@
             {
                 FMLLog.log.info("Detected deobfuscated environment, loading log configs for colored console logs.");
                 // use server logging configs in deobfuscated environment so developers get nicely colored console logs
-                System.setProperty("log4j.configurationFile", "log4j2_server.xml");
+                System.setProperty("log4j.configurationFile", "log4j2_mohist.xml");
                 ((LoggerContext) LogManager.getContext(false)).reconfigure();
             }
         }
@@ -579,7 +579,7 @@
                 }
                 else // This is a probably a normal minecraft workspace - log at warn
                 {
-                    FMLLog.log.warn("The coremod {} ({}) is not signed!", coreModName, coreModClass);
+                    FMLLog.log.debug("The coremod {} ({}) is not signed!", coreModName, coreModClass);
                 }
             }
             else
