--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/ReflectionHelper.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/ReflectionHelper.java
@@ -142,7 +142,7 @@
      *
      * @param clazz        The class to find the field on.
      * @param fieldName    The name of the field to find (used in developer environments, i.e. "maxStackSize").
-     * @param fieldObfName The obfuscated name of the field to find (used in obfuscated environments, i.e. "maxStackSize").
+     * @param fieldObfName The obfuscated name of the field to find (used in obfuscated environments, i.e. "field_77777_bU").
      *                     If the name you are looking for is on a class that is never obfuscated, this should be null.
      *
      * @return The field with the specified name in the given class.
@@ -279,7 +279,7 @@
      *
      * @param clazz          The class to find the method on.
      * @param methodName     The name of the method to find (used in developer environments, i.e. "getWorldTime").
-     * @param methodObfName  The obfuscated name of the method to find (used in obfuscated environments, i.e. "getWorldTime").
+     * @param methodObfName  The obfuscated name of the method to find (used in obfuscated environments, i.e. "func_72820_D").
      *                       If the name you are looking for is on a class that is never obfuscated, this should be null.
      * @param parameterTypes The parameter types of the method to find.
      * @return The method with the specified name and parameters in the given class.
