--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
@@ -21,8 +21,11 @@
 
 import java.lang.reflect.Method;
 
+import java.util.Objects;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.core.LoggerContext;
+import red.mohist.Mohist;
+import red.mohist.util.i18n.Message;
 
 public class ServerLaunchWrapper {
 
@@ -44,34 +47,41 @@
         if (System.getProperty("log4j.configurationFile") == null)
         {
             // Set this early so we don't need to reconfigure later
-            System.setProperty("log4j.configurationFile", "log4j2_server.xml");
+            System.setProperty("log4j.configurationFile", "log4j2_mohist.xml");
         }
         Class<?> launchwrapper = null;
-        try
-        {
-            launchwrapper = Class.forName("net.minecraft.launchwrapper.Launch",true,getClass().getClassLoader());
-            Class.forName("org.objectweb.asm.Type",true,getClass().getClassLoader());
-        }
-        catch (Exception e)
-        {
-            System.err.printf("We appear to be missing one or more essential library files.\n" +
-            		"You will need to add them to your server before FML and Forge will run successfully.");
+        try {
+            launchwrapper = Class.forName("net.minecraft.launchwrapper.Launch", true, Mohist.class.getClassLoader());
+            Class.forName("org.objectweb.asm.Type", true, Mohist.class.getClassLoader());
+            System.out.println("");
+            System.out.println("                   __                     __      ");
+            System.out.println(" /'\\_/`\\          /\\ \\       __          /\\ \\__   ");
+            System.out.println("/\\      \\     ___ \\ \\ \\___  /\\_\\     ____\\ \\ ,_\\  ");
+            System.out.println("\\ \\ \\__\\ \\   / __`\\\\ \\  _ `\\\\/\\ \\   /',__\\\\ \\ \\/  ");
+            System.out.println(" \\ \\ \\_/\\ \\ /\\ \\L\\ \\\\ \\ \\ \\ \\\\ \\ \\ /\\__, `\\\\ \\ \\_ ");
+            System.out.println("  \\ \\_\\\\ \\_\\\\ \\____/ \\ \\_\\ \\_\\\\ \\_\\\\/\\____/ \\ \\__\\");
+            System.out.println("   \\/_/ \\/_/ \\/___/   \\/_/\\/_/ \\/_/ \\/___/   \\/__/");
+            System.out.println("");
+            System.out.println("");
+            System.out.println("                        " + Message.getString("forge.serverlanunchwrapper.1"));
+            System.out.println(Message.getString("mohist.start"));
+            System.out.println(Message.getString("load.libraries"));
+            Mohist.LOGGER = LogManager.getLogger("Mohist");
+        } catch (Exception e) {
+            System.out.println(Message.getString("mohist.start.error.nothavelibrary"));
             e.printStackTrace(System.err);
             System.exit(1);
         }
 
-        try
-        {
-            Method main = launchwrapper.getMethod("main", String[].class);
+        try {
+            Method main = launchwrapper != null ? launchwrapper.getMethod("main", String[].class) : null;
             String[] allArgs = new String[args.length + 2];
             allArgs[0] = "--tweakClass";
             allArgs[1] = "net.minecraftforge.fml.common.launcher.FMLServerTweaker";
             System.arraycopy(args, 0, allArgs, 2, args.length);
-            main.invoke(null,(Object)allArgs);
-        }
-        catch (Exception e)
-        {
-            System.err.printf("A problem occurred running the Server launcher.");
+            Objects.requireNonNull(main).invoke(null, (Object) allArgs);
+        } catch (Exception e) {
+            System.out.println(Message.getString("mohist.start.error"));
             e.printStackTrace(System.err);
             System.exit(1);
         }
