--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.monster;
 
 import java.util.Collection;
+import java.util.concurrent.atomic.AtomicBoolean;
 import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.syncher.EntityDataAccessor;
@@ -40,6 +_,10 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.v1_20_R4.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class Creeper extends Monster implements PowerableMob {
    private static final EntityDataAccessor<Integer> DATA_SWELL_DIR = SynchedEntityData.defineId(Creeper.class, EntityDataSerializers.INT);
@@ -50,6 +_,7 @@
    public int maxSwell = 30;
    public int explosionRadius = 3;
    private int droppedSkulls;
+   private Player entityIgniter; // CraftBukkit
 
    public Creeper(EntityType<? extends Creeper> p_32278_, Level p_32279_) {
       super(p_32278_, p_32279_);
@@ -196,17 +_,28 @@
 
    public void thunderHit(ServerLevel p_32286_, LightningBolt p_32287_) {
       super.thunderHit(p_32286_, p_32287_);
+      // CraftBukkit start
+      if (CraftEventFactory.callCreeperPowerEvent(this, p_32287_, org.bukkit.event.entity.CreeperPowerEvent.PowerCause.LIGHTNING).isCancelled()) {
+         return;
+      }
+
       this.entityData.set(DATA_IS_POWERED, true);
    }
 
+   public void setPowered(boolean powered) {
+      this.entityData.set(DATA_IS_POWERED, powered);
+   }
+   // CraftBukkit end
+
    protected InteractionResult mobInteract(Player p_32301_, InteractionHand p_32302_) {
       ItemStack itemstack = p_32301_.getItemInHand(p_32302_);
       if (itemstack.is(ItemTags.CREEPER_IGNITERS)) {
          SoundEvent soundevent = itemstack.is(Items.FIRE_CHARGE) ? SoundEvents.FIRECHARGE_USE : SoundEvents.FLINTANDSTEEL_USE;
          this.level().playSound(p_32301_, this.getX(), this.getY(), this.getZ(), soundevent, this.getSoundSource(), 1.0F, this.random.nextFloat() * 0.4F + 0.8F);
          if (!this.level().isClientSide) {
+            this.entityIgniter = p_32301_; // CraftBukkit
             this.ignite();
-            if (!itemstack.isDamageableItem()) {
+            if (itemstack.getItem().getMaxDamage() == 0) { // CraftBukkit - fix MC-264285: unbreakable flint and steels are completely consumed when igniting a creeper
                itemstack.shrink(1);
             } else {
                itemstack.hurtAndBreak(1, p_32301_, (p_32290_) -> {
@@ -221,13 +_,21 @@
       }
    }
 
-   private void explodeCreeper() {
+   public void explodeCreeper() {
       if (!this.level().isClientSide) {
          float f = this.isPowered() ? 2.0F : 1.0F;
-         this.dead = true;
-         this.level().explode(this, this.getX(), this.getY(), this.getZ(), (float)this.explosionRadius * f, Level.ExplosionInteraction.MOB);
-         this.discard();
-         this.spawnLingeringCloud();
+         // CraftBukkit start
+         ExplosionPrimeEvent event = CraftEventFactory.callExplosionPrimeEvent(this, this.explosionRadius * f, false);
+         if (!event.isCancelled()) {
+            this.dead = true;
+            this.level().explode(this,  net.minecraft.world.level.Explosion.getDefaultDamageSource(this.level(), this).customCausingEntity(this.entityIgniter), null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB);
+            this.putRemoveCause(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
+            this.discard();
+            this.spawnLingeringCloud();
+         } else {
+            swell = 0;
+         }
+         // CraftBukkit end
       }
 
    }
@@ -236,6 +_,7 @@
       Collection<MobEffectInstance> collection = this.getActiveEffects();
       if (!collection.isEmpty()) {
          AreaEffectCloud areaeffectcloud = new AreaEffectCloud(this.level(), this.getX(), this.getY(), this.getZ());
+         areaeffectcloud.setOwner(this); // CraftBukkit
          areaeffectcloud.setRadius(2.5F);
          areaeffectcloud.setRadiusOnUse(-0.5F);
          areaeffectcloud.setWaitTime(10);
@@ -245,7 +_,7 @@
          for(MobEffectInstance mobeffectinstance : collection) {
             areaeffectcloud.addEffect(new MobEffectInstance(mobeffectinstance));
          }
-
+         areaeffectcloud.spawnReason(CreatureSpawnEvent.SpawnReason.EXPLOSION); // CraftBukkit
          this.level().addFreshEntity(areaeffectcloud);
       }
 
@@ -255,9 +_,23 @@
       return this.entityData.get(DATA_IS_IGNITED);
    }
 
+   // Mohist start - paper api
+   public AtomicBoolean ignited = new AtomicBoolean(true);
    public void ignite() {
-      this.entityData.set(DATA_IS_IGNITED, true);
-   }
+      boolean i = ignited.getAndSet(true);
+      if (isIgnited() != i) {
+         com.destroystokyo.paper.event.entity.CreeperIgniteEvent event = new com.destroystokyo.paper.event.entity.CreeperIgniteEvent((org.bukkit.entity.Creeper) getBukkitEntity(), i);
+         if (event.callEvent()) {
+            this.entityData.set(DATA_IS_IGNITED, event.isIgnited());
+         }
+      }
+   }
+
+   public void setIgnited(boolean ignited) {
+      this.ignited.set(ignited);
+      ignite();
+   }
+   // Mohist end
 
    public boolean canDropMobsSkull() {
       return this.isPowered() && this.droppedSkulls < 1;
