--- a/net/minecraft/world/entity/monster/Drowned.java
+++ b/net/minecraft/world/entity/monster/Drowned.java
@@ -55,8 +_,8 @@
 public class Drowned extends Zombie implements RangedAttackMob {
     public static final float NAUTILUS_SHELL_CHANCE = 0.03F;
     boolean searchingForLand;
-    public final WaterBoundPathNavigation waterNavigation;
-    public final GroundPathNavigation groundNavigation;
+    protected final WaterBoundPathNavigation waterNavigation;
+    protected final GroundPathNavigation groundNavigation;
 
     public Drowned(EntityType<? extends Drowned> p_32344_, Level p_32345_) {
         super(p_32344_, p_32345_);
