--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -59,7 +_,7 @@
    private static final AttributeModifier f_33395_ = new AttributeModifier(f_33394_, "Covered armor bonus", 20.0D, AttributeModifier.Operation.ADDITION);
    protected static final EntityDataAccessor<Direction> f_33392_ = SynchedEntityData.m_135353_(Shulker.class, EntityDataSerializers.f_135040_);
    protected static final EntityDataAccessor<Byte> f_33401_ = SynchedEntityData.m_135353_(Shulker.class, EntityDataSerializers.f_135027_);
-   protected static final EntityDataAccessor<Byte> f_33393_ = SynchedEntityData.m_135353_(Shulker.class, EntityDataSerializers.f_135027_);
+   public static final EntityDataAccessor<Byte> f_33393_ = SynchedEntityData.m_135353_(Shulker.class, EntityDataSerializers.f_135027_);
    private static final int f_149756_ = 6;
    private static final byte f_149757_ = 16;
    private static final byte f_149758_ = 16;
@@ -434,7 +_,7 @@
       return this.f_19804_.m_135370_(f_33392_);
    }
 
-   private void m_149788_(Direction p_149789_) {
+   public void m_149788_(Direction p_149789_) {
       this.f_19804_.m_135381_(f_33392_, p_149789_);
    }
 
@@ -446,11 +_,11 @@
       super.m_7350_(p_33434_);
    }
 
-   private int m_33463_() {
+   public int m_33463_() {
       return this.f_19804_.m_135370_(f_33401_);
    }
 
-   void m_33418_(int p_33419_) {
+   public void m_33418_(int p_33419_) {
       if (!this.f_19853_.f_46443_) {
          this.m_21051_(Attributes.f_22284_).m_22130_(f_33395_);
          if (p_33419_ == 0) {
