--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -54,6 +_,9 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.entity.EntityTeleportEvent;
 import org.joml.Vector3f;
 import org.joml.Vector3fc;
 
@@ -354,6 +_,13 @@
             if (blockpos1.getY() > this.level().getMinBuildHeight() && this.level().isEmptyBlock(blockpos1) && this.level().getWorldBorder().isWithinBounds(blockpos1) && this.level().noCollision(this, (new AABB(blockpos1)).deflate(1.0E-6D))) {
                Direction direction = this.findAttachableSurface(blockpos1);
                if (direction != null) {
+                  net.minecraftforge.event.entity.EntityTeleportEvent.EnderEntity event = net.minecraftforge.event.ForgeEventFactory.onEnderTeleport(this, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ());
+                  EntityTeleportEvent teleportEvent = CraftEventFactory.callEntityTeleportEvent(this, event.getTargetX(), event.getTargetY(), event.getTargetZ()); // Mohist hook forge event
+                  if (event.isCanceled() || teleportEvent.isCancelled()) direction = null;
+                  blockpos1 = CraftLocation.toBlockPosition(teleportEvent.getTo()); // Mohist hook forge event
+               }
+
+               if (direction != null) {
                   this.unRide();
                   this.setAttachFace(direction);
                   this.playSound(SoundEvents.SHULKER_TELEPORT, 1.0F, 1.0F);
@@ -417,9 +_,9 @@
             if (shulker != null) {
                shulker.setVariant(this.getVariant());
                shulker.moveTo(vec3);
+               shulker.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);  // CraftBukkit - the mysteries of life
                this.level().addFreshEntity(shulker);
             }
-
          }
       }
    }
@@ -432,7 +_,7 @@
       return this.entityData.get(DATA_ATTACH_FACE_ID);
    }
 
-   private void setAttachFace(Direction p_149789_) {
+   public void setAttachFace(Direction p_149789_) {
       this.entityData.set(DATA_ATTACH_FACE_ID, p_149789_);
    }
 
@@ -444,11 +_,11 @@
       super.onSyncedDataUpdated(p_33434_);
    }
 
-   private int getRawPeekAmount() {
+   public int getRawPeekAmount() {
       return this.entityData.get(DATA_PEEK_ID);
    }
 
-   void setRawPeekAmount(int p_33419_) {
+   public void setRawPeekAmount(int p_33419_) {
       if (!this.level().isClientSide) {
          this.getAttribute(Attributes.ARMOR).removeModifier(COVERED_ARMOR_MODIFIER.getId());
          if (p_33419_ == 0) {
