--- a/net/minecraft/world/entity/monster/Silverfish.java
+++ b/net/minecraft/world/entity/monster/Silverfish.java
@@ -25,13 +_,13 @@
 import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
 import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.InfestedBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.joml.Vector3f;
 
 public class Silverfish extends Monster {
@@ -141,7 +_,7 @@
             return false;
          } else {
             RandomSource randomsource = this.mob.getRandom();
-            if (this.mob.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && randomsource.nextInt(reducedTickDelay(10)) == 0) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.mob.level(), this.mob) && randomsource.nextInt(reducedTickDelay(10)) == 0) {
                this.selectedDirection = Direction.getRandom(randomsource);
                BlockPos blockpos = BlockPos.containing(this.mob.getX(), this.mob.getY() + 0.5D, this.mob.getZ()).relative(this.selectedDirection);
                BlockState blockstate = this.mob.level().getBlockState(blockpos);
@@ -168,6 +_,11 @@
             BlockPos blockpos = BlockPos.containing(this.mob.getX(), this.mob.getY() + 0.5D, this.mob.getZ()).relative(this.selectedDirection);
             BlockState blockstate = levelaccessor.getBlockState(blockpos);
             if (InfestedBlock.isCompatibleHostBlock(blockstate)) {
+               // CraftBukkit start
+               if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos, InfestedBlock.infestedStateByHost(blockstate))) {
+                  return;
+               }
+               // CraftBukkit end
                levelaccessor.setBlock(blockpos, InfestedBlock.infestedStateByHost(blockstate), 3);
                this.mob.spawnAnim();
                this.mob.discard();
@@ -210,7 +_,12 @@
                      BlockState blockstate = level.getBlockState(blockpos1);
                      Block block = blockstate.getBlock();
                      if (block instanceof InfestedBlock) {
-                        if (level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockpos1, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState())) {
+                           continue;
+                        }
+                        // CraftBukkit end
+                        if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(level, this.silverfish)) {
                            level.destroyBlock(blockpos1, true, this.silverfish);
                         } else {
                            level.setBlock(blockpos1, ((InfestedBlock)block).hostStateByInfested(level.getBlockState(blockpos1)), 3);
