--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -76,7 +_,7 @@
    private static final EntityDataAccessor<Integer> f_33854_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135028_);
    private static final EntityDataAccessor<Boolean> f_33855_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135035_);
    private static final EntityDataAccessor<Boolean> f_33856_ = SynchedEntityData.m_135353_(Strider.class, EntityDataSerializers.f_135035_);
-   private final ItemBasedSteering f_33857_ = new ItemBasedSteering(this.f_19804_, f_33854_, f_33856_);
+   public final ItemBasedSteering f_33857_ = new ItemBasedSteering(this.f_19804_, f_33854_, f_33856_);
    @Nullable
    private TemptGoal f_33858_;
    @Nullable
