--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -50,8 +_,8 @@
     Mob owner;
     @Nullable
     private BlockPos boundOrigin;
-    private boolean hasLimitedLife;
-    private int limitedLifeTicks;
+    public boolean hasLimitedLife;
+    public int limitedLifeTicks;
 
     public Vex(EntityType<? extends Vex> p_33984_, Level p_33985_) {
         super(p_33984_, p_33985_);
