--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -265,7 +_,7 @@
       }
 
       public void m_8056_() {
-         Vex.this.m_6710_(Vex.this.f_33980_.m_5448_());
+         Vex.this.setTarget(Vex.this.f_33980_.m_5448_(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true);
          super.m_8056_();
       }
    }
