--- a/net/minecraft/world/entity/monster/Vindicator.java
+++ b/net/minecraft/world/entity/monster/Vindicator.java
@@ -52,7 +_,7 @@
    static final Predicate<Difficulty> f_34070_ = (p_34082_) -> {
       return p_34082_ == Difficulty.NORMAL || p_34082_ == Difficulty.HARD;
    };
-   boolean f_34071_;
+   public boolean f_34071_;
 
    public Vindicator(EntityType<? extends Vindicator> p_34074_, Level p_34075_) {
       super(p_34074_, p_34075_);
@@ -77,7 +_,7 @@
 
    protected void m_8024_() {
       if (!this.m_21525_() && GoalUtils.m_26894_(this)) {
-         boolean flag = ((ServerLevel)this.m_9236_()).m_8843_(this.m_20183_());
+         boolean flag = ((ServerLevel)this.f_19853_).m_8843_(this.m_20183_());
          ((GroundPathNavigation)this.m_21573_()).m_26477_(flag);
       }
 
