--- a/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -46,7 +_,7 @@
       return SoundEvents.f_12561_;
    }
 
-   protected SoundEvent m_5592_() {
+   public SoundEvent m_5592_() { // Mohist - PAIL protected -> public
       return SoundEvents.f_12560_;
    }
 
@@ -90,7 +_,7 @@
          return false;
       } else {
          if (p_34169_ instanceof LivingEntity) {
-            ((LivingEntity)p_34169_).m_147207_(new MobEffectInstance(MobEffects.f_19615_, 200), this);
+            ((LivingEntity)p_34169_).addEffect(new MobEffectInstance(MobEffects.f_19615_, 200), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
          }
 
          return true;
