--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -12,6 +_,7 @@
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
@@ -64,6 +_,11 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 import org.joml.Vector3f;
 
 public class Zombie extends Monster {
@@ -85,6 +_,7 @@
    private boolean canBreakDoors;
    private int inWaterTime;
    public int conversionTime;
+   private int lastTick = MinecraftServer.currentTick; // CraftBukkit - add field
 
    public Zombie(EntityType<? extends Zombie> p_34271_, Level p_34272_) {
       super(p_34271_, p_34272_);
@@ -107,7 +_,7 @@
       this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0D));
       this.targetSelector.addGoal(1, (new HurtByTargetGoal(this)).setAlertOthers(ZombifiedPiglin.class));
       this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, true));
-      this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false));
+      if ( level.spigotConfig.zombieAggressiveTowardsVillager ) this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false));
       this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, IronGolem.class, true));
       this.targetSelector.addGoal(5, new NearestAttackableTargetGoal<>(this, Turtle.class, 10, true, false, Turtle.BABY_ON_LAND_SELECTOR));
    }
@@ -192,8 +_,11 @@
    public void tick() {
       if (!this.level().isClientSide && this.isAlive() && !this.isNoAi()) {
          if (this.isUnderWaterConverting()) {
-            --this.conversionTime;
-            if (this.conversionTime < 0) {
+            // CraftBukkit start - Use wall time instead of ticks for conversion
+            int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+            this.conversionTime -= elapsedTicks;
+            // CraftBukkit end
+            if (this.conversionTime < 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.DROWNED, (timer) -> this.conversionTime = timer)) {
                this.doUnderWaterConversion();
             }
          } else if (this.convertsInWater()) {
@@ -209,6 +_,7 @@
       }
 
       super.tick();
+      this.lastTick = MinecraftServer.currentTick; // CraftBukkit
    }
 
    public void aiStep() {
@@ -237,7 +_,8 @@
       super.aiStep();
    }
 
-   private void startUnderWaterConversion(int p_34279_) {
+   public void startUnderWaterConversion(int p_34279_) {
+      this.lastTick = MinecraftServer.currentTick; // CraftBukkit
       this.conversionTime = p_34279_;
       this.getEntityData().set(DATA_DROWNED_CONVERSION_ID, true);
    }
@@ -251,10 +_,15 @@
    }
 
    protected void convertToZombieType(EntityType<? extends Zombie> p_34311_) {
-      Zombie zombie = this.convertTo(p_34311_, true);
+      Zombie zombie = this.convertTo(p_34311_, true, EntityTransformEvent.TransformReason.DROWNED, CreatureSpawnEvent.SpawnReason.DROWNED);
       if (zombie != null) {
          zombie.handleAttributes(zombie.level().getCurrentDifficultyAt(zombie.blockPosition()).getSpecialMultiplier());
          zombie.setCanBreakDoors(zombie.supportsBreakDoorGoal() && this.canBreakDoors());
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombie);
+         // CraftBukkit start - SPIGOT-5208: End conversion to stop event spam
+      } else {
+         ((org.bukkit.entity.Zombie) getBukkitEntity()).setConversionTime(-1);
+         // CraftBukkit end
       }
 
    }
@@ -275,11 +_,15 @@
             livingentity = (LivingEntity)p_34288_.getEntity();
          }
 
-         if (livingentity != null && this.level().getDifficulty() == Difficulty.HARD && (double)this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE) && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
             int i = Mth.floor(this.getX());
             int j = Mth.floor(this.getY());
             int k = Mth.floor(this.getZ());
-            Zombie zombie = new Zombie(this.level());
+         net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent event = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, level(), i, j, k, livingentity, this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).getValue());
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW  ||
+            livingentity != null && this.level().getDifficulty() == Difficulty.HARD && (double)this.random.nextFloat() < this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).getValue() && this.level()
+                    .getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+            Zombie zombie = event.getCustomSummonedAid() != null && event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW ? event.getCustomSummonedAid() : EntityType.ZOMBIE.create(this.level());
 
             for(int l = 0; l < 50; ++l) {
                int i1 = i + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
@@ -291,8 +_,10 @@
                if (NaturalSpawner.isSpawnPositionOk(spawnplacements$type, this.level(), blockpos, entitytype) && SpawnPlacements.checkSpawnRules(entitytype, serverlevel, MobSpawnType.REINFORCEMENT, blockpos, this.level().random)) {
                   zombie.setPos((double)i1, (double)j1, (double)k1);
                   if (!this.level().hasNearbyAlivePlayer((double)i1, (double)j1, (double)k1, 7.0D) && this.level().isUnobstructed(zombie) && this.level().noCollision(zombie) && !this.level().containsAnyLiquid(zombie.getBoundingBox())) {
-                     zombie.setTarget(livingentity);
-                     zombie.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(zombie.blockPosition()), MobSpawnType.REINFORCEMENT, (SpawnGroupData)null, (CompoundTag)null);
+                     if (livingentity != null)
+                     zombie.setTarget(livingentity, EntityTargetEvent.TargetReason.REINFORCEMENT_TARGET, true); // CraftBukkit
+                     zombie.finalizeSpawn(serverlevel, this.level.getCurrentDifficultyAt(zombie.blockPosition()), MobSpawnType.REINFORCEMENT, (SpawnGroupData)null, (CompoundTag)null);
+                     zombie.spawnReason(CreatureSpawnEvent.SpawnReason.REINFORCEMENTS);
                      serverlevel.addFreshEntityWithPassengers(zombie);
                      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).addPermanentModifier(new AttributeModifier("Zombie reinforcement caller charge", (double)-0.05F, AttributeModifier.Operation.ADDITION));
                      zombie.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).addPermanentModifier(new AttributeModifier("Zombie reinforcement callee charge", (double)-0.05F, AttributeModifier.Operation.ADDITION));
@@ -311,7 +_,14 @@
       if (flag) {
          float f = this.level().getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
          if (this.getMainHandItem().isEmpty() && this.isOnFire() && this.random.nextFloat() < f * 0.3F) {
-            p_34276_.setSecondsOnFire(2 * (int)f);
+            // CraftBukkit start
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent(this.getBukkitEntity(), p_34276_.getBukkitEntity(), 2 * (int) f); // PAIL: fixme
+            this.level.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+               p_34276_.setSecondsOnFire(event.getDuration(), false);
+            }
+            // CraftBukkit end
          }
       }
 
@@ -376,11 +_,12 @@
 
    public boolean killedEntity(ServerLevel p_219160_, LivingEntity p_219161_) {
       boolean flag = super.killedEntity(p_219160_, p_219161_);
-      if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager) {
+      if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(p_219161_, EntityType.ZOMBIE_VILLAGER, (timer) -> {})) {
          if (p_219160_.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
             return flag;
          }
 
+         villager.transformReasonAndspawnReason(EntityTransformEvent.TransformReason.INFECTION, CreatureSpawnEvent.SpawnReason.INFECTION); // Mohist
          ZombieVillager zombievillager = villager.convertTo(EntityType.ZOMBIE_VILLAGER, false);
          if (zombievillager != null) {
             zombievillager.finalizeSpawn(p_219160_, p_219160_.getCurrentDifficultyAt(zombievillager.blockPosition()), MobSpawnType.CONVERSION, new Zombie.ZombieGroupData(false, true), (CompoundTag)null);
@@ -388,6 +_,7 @@
             zombievillager.setGossips(villager.getGossips().store(NbtOps.INSTANCE));
             zombievillager.setTradeOffers(villager.getOffers().createTag());
             zombievillager.setVillagerXp(villager.getVillagerXp());
+            net.minecraftforge.event.ForgeEventFactory.onLivingConvert(p_219161_, zombievillager);
             if (!this.isSilent()) {
                p_219160_.levelEvent((Player)null, 1026, this.blockPosition(), 0);
             }
@@ -399,6 +_,22 @@
       return flag;
    }
 
+   public static ZombieVillager zombifyVillager(ServerLevel worldserver, Villager entityvillager, BlockPos blockPosition, boolean silent, SpawnReason spawnReason) {
+      ZombieVillager zombievillager = entityvillager.convertTo(EntityType.ZOMBIE_VILLAGER, false, EntityTransformEvent.TransformReason.INFECTION, spawnReason);
+      if (zombievillager != null) {
+         zombievillager.finalizeSpawn(worldserver, worldserver.getCurrentDifficultyAt(zombievillager.blockPosition()), MobSpawnType.CONVERSION, new Zombie.ZombieGroupData(false, true), (CompoundTag) null);
+         zombievillager.setVillagerData(entityvillager.getVillagerData());
+         zombievillager.setGossips(entityvillager.getGossips().store(NbtOps.INSTANCE));
+         zombievillager.setTradeOffers(entityvillager.getOffers().createTag());
+         zombievillager.setVillagerXp(entityvillager.getVillagerXp());
+         // CraftBukkit start
+         if (!silent) {
+            worldserver.levelEvent(null, 1026, blockPosition, 0);
+         }
+      }
+      return zombievillager;
+   }
+
    protected float getStandingEyeHeight(Pose p_34313_, EntityDimensions p_34314_) {
       return this.isBaby() ? 0.93F : 1.74F;
    }
@@ -439,7 +_,8 @@
                      chicken1.finalizeSpawn(p_34297_, p_34298_, MobSpawnType.JOCKEY, (SpawnGroupData)null, (CompoundTag)null);
                      chicken1.setChickenJockey(true);
                      this.startRiding(chicken1);
-                     p_34297_.addFreshEntity(chicken1);
+                     chicken1.spawnReason(CreatureSpawnEvent.SpawnReason.MOUNT);
+                     p_34297_.addFreshEntity(chicken1); // CraftBukkit
                   }
                }
             }
@@ -465,7 +_,7 @@
    }
 
    public static boolean getSpawnAsBabyOdds(RandomSource p_219163_) {
-      return p_219163_.nextFloat() < 0.05F;
+      return p_219163_.nextFloat() < net.minecraftforge.common.ForgeConfig.SERVER.zombieBabyChance.get();
    }
 
    protected void handleAttributes(float p_34340_) {
@@ -485,7 +_,7 @@
    }
 
    protected void randomizeReinforcementsChance() {
-      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * (double)0.1F);
+      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * net.minecraftforge.common.ForgeConfig.SERVER.zombieBaseSummonChance.get());
    }
 
    protected Vector3f getPassengerAttachmentPoint(Entity p_300549_, EntityDimensions p_298020_, float p_298000_) {
