--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
@@ -37,213 +_,221 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.pathfinder.PathType;
+import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.phys.AABB;
+import org.joml.Vector3f;
 
 public class ZombifiedPiglin extends Zombie implements NeutralMob {
-    private static final EntityDimensions BABY_DIMENSIONS = EntityType.ZOMBIFIED_PIGLIN.getDimensions().scale(0.5F).withEyeHeight(0.97F);
-    private static final UUID SPEED_MODIFIER_ATTACKING_UUID = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
-    private static final AttributeModifier SPEED_MODIFIER_ATTACKING = new AttributeModifier(SPEED_MODIFIER_ATTACKING_UUID, "Attacking speed boost", 0.05, AttributeModifier.Operation.ADD_VALUE);
-    private static final UniformInt FIRST_ANGER_SOUND_DELAY = TimeUtil.rangeOfSeconds(0, 1);
-    private int playFirstAngerSoundIn;
-    private static final UniformInt PERSISTENT_ANGER_TIME = TimeUtil.rangeOfSeconds(20, 39);
-    private int remainingPersistentAngerTime;
-    @Nullable
-    private UUID persistentAngerTarget;
-    private static final int ALERT_RANGE_Y = 10;
-    private static final UniformInt ALERT_INTERVAL = TimeUtil.rangeOfSeconds(4, 6);
-    private int ticksUntilNextAlert;
-
-    public ZombifiedPiglin(EntityType<? extends ZombifiedPiglin> p_34427_, Level p_34428_) {
-        super(p_34427_, p_34428_);
-        this.setPathfindingMalus(PathType.LAVA, 8.0F);
-    }
-
-    @Override
-    public void setPersistentAngerTarget(@Nullable UUID p_34444_) {
-        this.persistentAngerTarget = p_34444_;
-    }
-
-    @Override
-    protected void addBehaviourGoals() {
-        this.goalSelector.addGoal(2, new ZombieAttackGoal(this, 1.0, false));
-        this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0));
-        this.targetSelector.addGoal(1, new HurtByTargetGoal(this).setAlertOthers());
-        this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
-        this.targetSelector.addGoal(3, new ResetUniversalAngerTargetGoal<>(this, true));
-    }
-
-    public static AttributeSupplier.Builder createAttributes() {
-        return Zombie.createAttributes().add(Attributes.SPAWN_REINFORCEMENTS_CHANCE, 0.0).add(Attributes.MOVEMENT_SPEED, 0.23F).add(Attributes.ATTACK_DAMAGE, 5.0);
-    }
-
-    @Override
-    public EntityDimensions getDefaultDimensions(Pose p_334497_) {
-        return this.isBaby() ? BABY_DIMENSIONS : super.getDefaultDimensions(p_334497_);
-    }
-
-    @Override
-    protected boolean convertsInWater() {
-        return false;
-    }
-
-    @Override
-    protected void customServerAiStep() {
-        AttributeInstance attributeinstance = this.getAttribute(Attributes.MOVEMENT_SPEED);
-        if (this.isAngry()) {
-            if (!this.isBaby() && !attributeinstance.hasModifier(SPEED_MODIFIER_ATTACKING)) {
-                attributeinstance.addTransientModifier(SPEED_MODIFIER_ATTACKING);
-            }
-
-            this.maybePlayFirstAngerSound();
-        } else if (attributeinstance.hasModifier(SPEED_MODIFIER_ATTACKING)) {
-            attributeinstance.removeModifier(SPEED_MODIFIER_ATTACKING.id());
-        }
-
-        this.updatePersistentAnger((ServerLevel)this.level(), true);
-        if (this.getTarget() != null) {
-            this.maybeAlertOthers();
-        }
-
-        if (this.isAngry()) {
-            this.lastHurtByPlayerTime = this.tickCount;
-        }
-
-        super.customServerAiStep();
-    }
-
-    private void maybePlayFirstAngerSound() {
-        if (this.playFirstAngerSoundIn > 0) {
-            this.playFirstAngerSoundIn--;
-            if (this.playFirstAngerSoundIn == 0) {
-                this.playAngerSound();
-            }
-        }
-    }
-
-    private void maybeAlertOthers() {
-        if (this.ticksUntilNextAlert > 0) {
-            this.ticksUntilNextAlert--;
-        } else {
-            if (this.getSensing().hasLineOfSight(this.getTarget())) {
-                this.alertOthers();
-            }
-
-            this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
-        }
-    }
-
-    private void alertOthers() {
-        double d0 = this.getAttributeValue(Attributes.FOLLOW_RANGE);
-        AABB aabb = AABB.unitCubeFromLowerCorner(this.position()).inflate(d0, 10.0, d0);
-        this.level()
-            .getEntitiesOfClass(ZombifiedPiglin.class, aabb, EntitySelector.NO_SPECTATORS)
-            .stream()
-            .filter(p_34463_ -> p_34463_ != this)
-            .filter(p_327015_ -> p_327015_.getTarget() == null)
-            .filter(p_327016_ -> !p_327016_.isAlliedTo(this.getTarget()))
-            .forEach(p_327014_ -> p_327014_.setTarget(this.getTarget()));
-    }
-
-    private void playAngerSound() {
-        this.playSound(SoundEvents.ZOMBIFIED_PIGLIN_ANGRY, this.getSoundVolume() * 2.0F, this.getVoicePitch() * 1.8F);
-    }
-
-    @Override
-    public void setTarget(@Nullable LivingEntity p_34478_) {
-        if (this.getTarget() == null && p_34478_ != null) {
-            this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
-            this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
-        }
-
-        if (p_34478_ instanceof Player) {
-            this.setLastHurtByPlayer((Player)p_34478_);
-        }
-
-        super.setTarget(p_34478_);
-    }
-
-    @Override
-    public void startPersistentAngerTimer() {
-        this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.sample(this.random));
-    }
-
-    public static boolean checkZombifiedPiglinSpawnRules(
-        EntityType<ZombifiedPiglin> p_219174_, LevelAccessor p_219175_, MobSpawnType p_219176_, BlockPos p_219177_, RandomSource p_219178_
-    ) {
-        return p_219175_.getDifficulty() != Difficulty.PEACEFUL && !p_219175_.getBlockState(p_219177_.below()).is(Blocks.NETHER_WART_BLOCK);
-    }
-
-    @Override
-    public boolean checkSpawnObstruction(LevelReader p_34442_) {
-        return p_34442_.isUnobstructed(this) && !p_34442_.containsAnyLiquid(this.getBoundingBox());
-    }
-
-    @Override
-    public void addAdditionalSaveData(CompoundTag p_34458_) {
-        super.addAdditionalSaveData(p_34458_);
-        this.addPersistentAngerSaveData(p_34458_);
-    }
-
-    @Override
-    public void readAdditionalSaveData(CompoundTag p_34446_) {
-        super.readAdditionalSaveData(p_34446_);
-        this.readPersistentAngerSaveData(this.level(), p_34446_);
-    }
-
-    @Override
-    public void setRemainingPersistentAngerTime(int p_34448_) {
-        this.remainingPersistentAngerTime = p_34448_;
-    }
-
-    @Override
-    public int getRemainingPersistentAngerTime() {
-        return this.remainingPersistentAngerTime;
-    }
-
-    @Override
-    protected SoundEvent getAmbientSound() {
-        return this.isAngry() ? SoundEvents.ZOMBIFIED_PIGLIN_ANGRY : SoundEvents.ZOMBIFIED_PIGLIN_AMBIENT;
-    }
-
-    @Override
-    protected SoundEvent getHurtSound(DamageSource p_34466_) {
-        return SoundEvents.ZOMBIFIED_PIGLIN_HURT;
-    }
-
-    @Override
-    protected SoundEvent getDeathSound() {
-        return SoundEvents.ZOMBIFIED_PIGLIN_DEATH;
-    }
-
-    @Override
-    protected void populateDefaultEquipmentSlots(RandomSource p_219171_, DifficultyInstance p_219172_) {
-        this.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
-    }
-
-    @Override
-    protected ItemStack getSkull() {
-        return ItemStack.EMPTY;
-    }
-
-    @Override
-    protected void randomizeReinforcementsChance() {
-        this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(0.0);
-    }
-
-    @Nullable
-    @Override
-    public UUID getPersistentAngerTarget() {
-        return this.persistentAngerTarget;
-    }
-
-    @Override
-    public boolean isPreventingPlayerRest(Player p_34475_) {
-        return this.isAngryAt(p_34475_);
-    }
-
-    @Override
-    public boolean wantsToPickUp(ItemStack p_182402_) {
-        return this.canHoldItem(p_182402_);
-    }
+   private static final UUID SPEED_MODIFIER_ATTACKING_UUID = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
+   private static final AttributeModifier SPEED_MODIFIER_ATTACKING = new AttributeModifier(SPEED_MODIFIER_ATTACKING_UUID, "Attacking speed boost", 0.05D, AttributeModifier.Operation.ADDITION);
+   private static final UniformInt FIRST_ANGER_SOUND_DELAY = TimeUtil.rangeOfSeconds(0, 1);
+   private int playFirstAngerSoundIn;
+   private static final UniformInt PERSISTENT_ANGER_TIME = TimeUtil.rangeOfSeconds(20, 39);
+   private int remainingPersistentAngerTime;
+   @Nullable
+   private UUID persistentAngerTarget;
+   private static final int ALERT_RANGE_Y = 10;
+   private static final UniformInt ALERT_INTERVAL = TimeUtil.rangeOfSeconds(4, 6);
+   private int ticksUntilNextAlert;
+   private static final float ZOMBIFIED_PIGLIN_EYE_HEIGHT = 1.79F;
+   private static final float ZOMBIFIED_PIGLIN_BABY_EYE_HEIGHT_ADJUSTMENT = 0.82F;
+
+   public ZombifiedPiglin(EntityType<? extends ZombifiedPiglin> p_34427_, Level p_34428_) {
+      super(p_34427_, p_34428_);
+      this.setPathfindingMalus(BlockPathTypes.LAVA, 8.0F);
+   }
+
+   public void setPersistentAngerTarget(@Nullable UUID p_34444_) {
+      this.persistentAngerTarget = p_34444_;
+   }
+
+   protected void addBehaviourGoals() {
+      this.goalSelector.addGoal(2, new ZombieAttackGoal(this, 1.0D, false));
+      this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0D));
+      this.targetSelector.addGoal(1, (new HurtByTargetGoal(this)).setAlertOthers());
+      this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, this::isAngryAt));
+      this.targetSelector.addGoal(3, new ResetUniversalAngerTargetGoal<>(this, true));
+   }
+
+   public static AttributeSupplier.Builder createAttributes() {
+      return Zombie.createAttributes().add(Attributes.SPAWN_REINFORCEMENTS_CHANCE, 0.0D).add(Attributes.MOVEMENT_SPEED, (double)0.23F).add(Attributes.ATTACK_DAMAGE, 5.0D);
+   }
+
+   protected float getStandingEyeHeight(Pose p_259553_, EntityDimensions p_259614_) {
+      return this.isBaby() ? 0.96999997F : 1.79F;
+   }
+
+   protected boolean convertsInWater() {
+      return false;
+   }
+
+   protected void customServerAiStep() {
+      AttributeInstance attributeinstance = this.getAttribute(Attributes.MOVEMENT_SPEED);
+      if (this.isAngry()) {
+         if (!this.isBaby() && !attributeinstance.hasModifier(SPEED_MODIFIER_ATTACKING)) {
+            attributeinstance.addTransientModifier(SPEED_MODIFIER_ATTACKING);
+         }
+
+         this.maybePlayFirstAngerSound();
+      } else if (attributeinstance.hasModifier(SPEED_MODIFIER_ATTACKING)) {
+         attributeinstance.removeModifier(SPEED_MODIFIER_ATTACKING.getId());
+      }
+
+      this.updatePersistentAnger((ServerLevel)this.level(), true);
+      if (this.getTarget() != null) {
+         this.maybeAlertOthers();
+      }
+
+      if (this.isAngry()) {
+         this.lastHurtByPlayerTime = this.tickCount;
+      }
+
+      super.customServerAiStep();
+   }
+
+   private void maybePlayFirstAngerSound() {
+      if (this.playFirstAngerSoundIn > 0) {
+         --this.playFirstAngerSoundIn;
+         if (this.playFirstAngerSoundIn == 0) {
+            this.playAngerSound();
+         }
+      }
+
+   }
+
+   private void maybeAlertOthers() {
+      if (this.ticksUntilNextAlert > 0) {
+         --this.ticksUntilNextAlert;
+      } else {
+         if (this.getSensing().hasLineOfSight(this.getTarget())) {
+            this.alertOthers();
+         }
+
+         this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
+      }
+   }
+
+   private void alertOthers() {
+      double d0 = this.getAttributeValue(Attributes.FOLLOW_RANGE);
+      AABB aabb = AABB.unitCubeFromLowerCorner(this.position()).inflate(d0, 10.0D, d0);
+      this.level().getEntitiesOfClass(ZombifiedPiglin.class, aabb, EntitySelector.NO_SPECTATORS).stream().filter((p_34463_) -> {
+         return p_34463_ != this;
+      }).filter((p_296831_) -> {
+         return p_296831_.getTarget() == null;
+      }).filter((p_296830_) -> {
+         return !p_296830_.isAlliedTo(this.getTarget());
+      }).forEach((p_296829_) -> {
+         p_296829_.setTarget(this.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
+      });
+   }
+
+   private void playAngerSound() {
+      this.playSound(SoundEvents.ZOMBIFIED_PIGLIN_ANGRY, this.getSoundVolume() * 2.0F, this.getVoicePitch() * 1.8F);
+   }
+
+   public void setTarget(@Nullable LivingEntity p_34478_) {
+      if (this.getTarget() == null && p_34478_ != null) {
+         this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
+         this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
+      }
+
+      if (p_34478_ instanceof Player) {
+         this.setLastHurtByPlayer((Player)p_34478_);
+      }
+
+      super.setTarget(p_34478_);
+   }
+
+   @Override
+   public boolean setTarget(@Nullable LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason,boolean fireEvent){ // CraftBukkit - signature
+      if (this.getTarget() == null && entityliving != null) {
+         this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
+         this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
+      }
+
+      if (entityliving instanceof Player) {
+         this.setLastHurtByPlayer((Player)entityliving);
+      }
+      return super.setTarget(entityliving, reason, fireEvent); // CraftBukkit
+   }
+
+   public void startPersistentAngerTimer() {
+      // CraftBukkit start
+      Entity entity = ((ServerLevel) this.level).getEntity(getPersistentAngerTarget());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), PERSISTENT_ANGER_TIME.sample(this.random));
+      this.level.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.setPersistentAngerTarget(null);
+         return;
+      }
+      this.setRemainingPersistentAngerTime(event.getNewAnger());
+      // CraftBukkit end
+   }
+
+   public static boolean checkZombifiedPiglinSpawnRules(EntityType<ZombifiedPiglin> p_219174_, LevelAccessor p_219175_, MobSpawnType p_219176_, BlockPos p_219177_, RandomSource p_219178_) {
+      return p_219175_.getDifficulty() != Difficulty.PEACEFUL && !p_219175_.getBlockState(p_219177_.below()).is(Blocks.NETHER_WART_BLOCK);
+   }
+
+   public boolean checkSpawnObstruction(LevelReader p_34442_) {
+      return p_34442_.isUnobstructed(this) && !p_34442_.containsAnyLiquid(this.getBoundingBox());
+   }
+
+   public void addAdditionalSaveData(CompoundTag p_34458_) {
+      super.addAdditionalSaveData(p_34458_);
+      this.addPersistentAngerSaveData(p_34458_);
+   }
+
+   public void readAdditionalSaveData(CompoundTag p_34446_) {
+      super.readAdditionalSaveData(p_34446_);
+      this.readPersistentAngerSaveData(this.level(), p_34446_);
+   }
+
+   public void setRemainingPersistentAngerTime(int p_34448_) {
+      this.remainingPersistentAngerTime = p_34448_;
+   }
+
+   public int getRemainingPersistentAngerTime() {
+      return this.remainingPersistentAngerTime;
+   }
+
+   protected SoundEvent getAmbientSound() {
+      return this.isAngry() ? SoundEvents.ZOMBIFIED_PIGLIN_ANGRY : SoundEvents.ZOMBIFIED_PIGLIN_AMBIENT;
+   }
+
+   protected SoundEvent getHurtSound(DamageSource p_34466_) {
+      return SoundEvents.ZOMBIFIED_PIGLIN_HURT;
+   }
+
+   protected SoundEvent getDeathSound() {
+      return SoundEvents.ZOMBIFIED_PIGLIN_DEATH;
+   }
+
+   protected void populateDefaultEquipmentSlots(RandomSource p_219171_, DifficultyInstance p_219172_) {
+      this.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
+   }
+
+   protected ItemStack getSkull() {
+      return ItemStack.EMPTY;
+   }
+
+   protected void randomizeReinforcementsChance() {
+      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(0.0D);
+   }
+
+   @Nullable
+   public UUID getPersistentAngerTarget() {
+      return this.persistentAngerTarget;
+   }
+
+   public boolean isPreventingPlayerRest(Player p_34475_) {
+      return this.isAngryAt(p_34475_);
+   }
+
+   public boolean wantsToPickUp(ItemStack p_182402_) {
+      return this.canHoldItem(p_182402_);
+   }
+
+   protected Vector3f getPassengerAttachmentPoint(Entity p_299119_, EntityDimensions p_297425_, float p_300666_) {
+      return new Vector3f(0.0F, p_297425_.height + 0.05F * p_300666_, 0.0F);
+   }
 }
