--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -142,7 +_,7 @@
       }).filter((p_296830_) -> {
          return !p_296830_.isAlliedTo(this.getTarget());
       }).forEach((p_296829_) -> {
-         p_296829_.setTarget(this.getTarget());
+         p_296829_.setTarget(this.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -163,8 +_,30 @@
       super.setTarget(p_34478_);
    }
 
+   @Override
+   public boolean setTarget(@Nullable LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason,boolean fireEvent){ // CraftBukkit - signature
+      if (this.getTarget() == null && entityliving != null) {
+         this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
+         this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
+      }
+
+      if (entityliving instanceof Player) {
+         this.setLastHurtByPlayer((Player)entityliving);
+      }
+      return super.setTarget(entityliving, reason, fireEvent); // CraftBukkit
+   }
+
    public void startPersistentAngerTimer() {
-      this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.sample(this.random));
+      // CraftBukkit start
+      Entity entity = ((ServerLevel) this.level).getEntity(getPersistentAngerTarget());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), PERSISTENT_ANGER_TIME.sample(this.random));
+      this.level.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.setPersistentAngerTarget(null);
+         return;
+      }
+      this.setRemainingPersistentAngerTime(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean checkZombifiedPiglinSpawnRules(EntityType<ZombifiedPiglin> p_219174_, LevelAccessor p_219175_, MobSpawnType p_219176_, BlockPos p_219177_, RandomSource p_219178_) {
