--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -14,6 +_,7 @@
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
@@ -46,6 +_,8 @@
 import net.minecraft.world.level.block.BedBlock;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 import org.joml.Vector3f;
 import org.slf4j.Logger;
 
@@ -65,6 +_,7 @@
    @Nullable
    private CompoundTag tradeOffers;
    private int villagerXp;
+   private int lastTick = MinecraftServer.currentTick; // CraftBukkit - add field
 
    public ZombieVillager(EntityType<? extends ZombieVillager> p_34368_, Level p_34369_) {
       super(p_34368_, p_34369_);
@@ -128,13 +_,18 @@
    public void tick() {
       if (!this.level().isClientSide && this.isAlive() && this.isConverting()) {
          int i = this.getConversionProgress();
+         // CraftBukkit start - Use wall time instead of ticks for villager conversion
+         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+         i *= elapsedTicks;
+         // CraftBukkit end
          this.villagerConversionTime -= i;
-         if (this.villagerConversionTime <= 0) {
+         if (this.villagerConversionTime <= 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.VILLAGER, (timer) -> this.villagerConversionTime = timer)) {
             this.finishConversion((ServerLevel)this.level());
          }
       }
 
       super.tick();
+      this.lastTick = MinecraftServer.currentTick; // CraftBukkit
    }
 
    public InteractionResult mobInteract(Player p_34394_, InteractionHand p_34395_) {
@@ -170,19 +_,21 @@
       return this.getEntityData().get(DATA_CONVERTING_ID);
    }
 
-   private void startConverting(@Nullable UUID p_34384_, int p_34385_) {
+   public void startConverting(@Nullable UUID p_34384_, int p_34385_) {
       this.conversionStarter = p_34384_;
       this.villagerConversionTime = p_34385_;
       this.getEntityData().set(DATA_CONVERTING_ID, true);
-      this.removeEffect(MobEffects.WEAKNESS);
-      this.addEffect(new MobEffectInstance(MobEffects.DAMAGE_BOOST, p_34385_, Math.min(this.level().getDifficulty().getId() - 1, 0)));
-      this.level().broadcastEntityEvent(this, (byte)16);
+      // CraftBukkit start
+      this.removeEffect(MobEffects.WEAKNESS, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+      this.addEffect(new MobEffectInstance(MobEffects.DAMAGE_BOOST, p_34385_, Math.min(this.level.getDifficulty().getId() - 1, 0)), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+      // CraftBukkit end
+      this.level.broadcastEntityEvent(this, (byte)16);
    }
 
    public void handleEntityEvent(byte p_34372_) {
       if (p_34372_ == 16) {
          if (!this.isSilent()) {
-            this.level().playLocalSound(this.getX(), this.getEyeY(), this.getZ(), SoundEvents.ZOMBIE_VILLAGER_CURE, this.getSoundSource(), 1.0F + this.random.nextFloat(), this.random.nextFloat() * 0.7F + 0.3F, false);
+            this.level.playLocalSound(this.getX(), this.getEyeY(), this.getZ(), SoundEvents.ZOMBIE_VILLAGER_CURE, this.getSoundSource(), 1.0F + this.random.nextFloat(), this.random.nextFloat() * 0.7F + 0.3F, false);
          }
 
       } else {
@@ -191,7 +_,14 @@
    }
 
    private void finishConversion(ServerLevel p_34399_) {
+      // CraftBukkit start
+      this.transformReasonAndspawnReason(EntityTransformEvent.TransformReason.CURED, CreatureSpawnEvent.SpawnReason.CURED); // Mohist
       Villager villager = this.convertTo(EntityType.VILLAGER, false);
+      if (villager == null) {
+         ((org.bukkit.entity.ZombieVillager) getBukkitEntity()).setConversionTime(-1); // SPIGOT-5208: End conversion to stop event spam
+         return;
+      }
+      // CraftBukkit end
 
       for(EquipmentSlot equipmentslot : EquipmentSlot.values()) {
          ItemStack itemstack = this.getItemBySlot(equipmentslot);
@@ -201,7 +_,9 @@
             } else {
                double d0 = (double)this.getEquipmentDropChance(equipmentslot);
                if (d0 > 1.0D) {
+                  this.forceDrops = true; // CraftBukkit
                   this.spawnAtLocation(itemstack);
+                  this.forceDrops = false; // CraftBukkit
                }
             }
          }
@@ -227,11 +_,11 @@
          }
       }
 
-      villager.addEffect(new MobEffectInstance(MobEffects.CONFUSION, 200, 0));
+      villager.addEffect(new MobEffectInstance(MobEffects.CONFUSION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION); // CraftBukkit);
       if (!this.isSilent()) {
          p_34399_.levelEvent((Player)null, 1027, this.blockPosition(), 0);
       }
-
+      net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, villager);
    }
 
    private int getConversionProgress() {
