--- a/net/minecraft/command/impl/DebugCommand.java
+++ b/net/minecraft/command/impl/DebugCommand.java
@@ -92,7 +92,7 @@
          p_225389_0_.func_197030_a(new TranslationTextComponent("commands.debug.reportSaved", s), false);
          return 1;
       } catch (IOException ioexception) {
-         field_225390_a.error("Failed to save debug dump", (Throwable)ioexception);
+         field_225390_a.error(com.mohistmc.util.i18n.i18n.get("debugcommand.1", (Throwable)ioexception));
          p_225389_0_.func_197021_a(new TranslationTextComponent("commands.debug.reportFailed"));
          return 0;
       }
