--- a/net/minecraft/world/level/levelgen/LegacyRandomSource.java
+++ b/net/minecraft/world/level/levelgen/LegacyRandomSource.java
@@ -73,11 +_,6 @@
             return new LegacyRandomSource((long)i ^ this.seed);
         }
 
-        @Override
-        public RandomSource fromSeed(long p_344211_) {
-            return new LegacyRandomSource(p_344211_);
-        }
-
         @VisibleForTesting
         @Override
         public void parityConfigString(StringBuilder p_188596_) {
