--- a/net/minecraft/world/level/levelgen/NoiseRouterData.java
+++ b/net/minecraft/world/level/levelgen/NoiseRouterData.java
@@ -60,7 +_,7 @@
     private static final ResourceKey<DensityFunction> SPAGHETTI_2D = createKey("overworld/caves/spaghetti_2d");
 
     private static ResourceKey<DensityFunction> createKey(String p_209537_) {
-        return ResourceKey.create(Registries.DENSITY_FUNCTION, ResourceLocation.withDefaultNamespace(p_209537_));
+        return ResourceKey.create(Registries.DENSITY_FUNCTION, new ResourceLocation(p_209537_));
     }
 
     public static Holder<? extends DensityFunction> bootstrap(BootstrapContext<DensityFunction> p_335193_) {
