--- a/net/minecraft/world/level/levelgen/NoiseGeneratorSettings.java
+++ b/net/minecraft/world/level/levelgen/NoiseGeneratorSettings.java
@@ -50,23 +_,23 @@
       p_256365_.m_255272_(f_64437_, m_255230_(p_256365_));
    }
 
-   private static NoiseGeneratorSettings m_255186_(BootstapContext<?> p_256478_) {
+   public static NoiseGeneratorSettings m_255186_(BootstapContext<?> p_256478_) {
       return new NoiseGeneratorSettings(NoiseSettings.f_209631_, Blocks.f_50259_.m_49966_(), Blocks.f_50016_.m_49966_(), NoiseRouterData.m_255104_(p_256478_.m_255420_(Registries.f_257040_)), SurfaceRuleData.m_194813_(), List.of(), 0, true, false, false, true);
    }
 
-   private static NoiseGeneratorSettings m_255410_(BootstapContext<?> p_256180_) {
+   public static NoiseGeneratorSettings m_255410_(BootstapContext<?> p_256180_) {
       return new NoiseGeneratorSettings(NoiseSettings.f_209630_, Blocks.f_50134_.m_49966_(), Blocks.f_49991_.m_49966_(), NoiseRouterData.m_255404_(p_256180_.m_255420_(Registries.f_257040_), p_256180_.m_255420_(Registries.f_256865_)), SurfaceRuleData.m_194812_(), List.of(), 32, false, false, false, true);
    }
 
-   private static NoiseGeneratorSettings m_255226_(BootstapContext<?> p_256460_, boolean p_256427_, boolean p_256318_) {
+   public static NoiseGeneratorSettings m_255226_(BootstapContext<?> p_256460_, boolean p_256427_, boolean p_256318_) {
       return new NoiseGeneratorSettings(NoiseSettings.f_224519_, Blocks.f_50069_.m_49966_(), Blocks.f_49990_.m_49966_(), NoiseRouterData.m_255262_(p_256460_.m_255420_(Registries.f_257040_), p_256460_.m_255420_(Registries.f_256865_), p_256318_, p_256427_), SurfaceRuleData.m_194807_(), (new OverworldBiomeBuilder()).m_187154_(), 63, false, true, true, false);
    }
 
-   private static NoiseGeneratorSettings m_255038_(BootstapContext<?> p_255690_) {
+   public static NoiseGeneratorSettings m_255038_(BootstapContext<?> p_255690_) {
       return new NoiseGeneratorSettings(NoiseSettings.f_209632_, Blocks.f_50069_.m_49966_(), Blocks.f_49990_.m_49966_(), NoiseRouterData.m_255020_(p_255690_.m_255420_(Registries.f_257040_), p_255690_.m_255420_(Registries.f_256865_)), SurfaceRuleData.m_198380_(false, true, true), List.of(), 32, false, false, false, true);
    }
 
-   private static NoiseGeneratorSettings m_255230_(BootstapContext<?> p_255885_) {
+   public static NoiseGeneratorSettings m_255230_(BootstapContext<?> p_255885_) {
       return new NoiseGeneratorSettings(NoiseSettings.f_209633_, Blocks.f_50069_.m_49966_(), Blocks.f_49990_.m_49966_(), NoiseRouterData.m_254860_(p_255885_.m_255420_(Registries.f_257040_), p_255885_.m_255420_(Registries.f_256865_)), SurfaceRuleData.m_198380_(false, false, false), List.of(), -64, false, false, false, true);
    }
 
