--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -15,6 +_,13 @@
    private int f_38699_;
    private int f_38700_ = 20;
 
+   // CraftBukkit start
+   private Player entityhuman;
+   public int saturatedRegenRate = 10;
+   public int unsaturatedRegenRate = 80;
+   public int starvationRate = 80;
+   // CraftBukkit end
+
    public FoodData() {
       this.f_38697_ = 5.0F;
    }
