--- a/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractChestedHorse.java
@@ -69,6 +_,14 @@
 
    }
 
+   // Paper start
+   protected void postDeathDropItems(org.bukkit.event.entity.EntityDeathEvent event) {
+      if (this.m_30502_() && (event == null || !event.isCancelled())) {
+         this.m_30504_(false);
+      }
+   }
+   // Paper end
+
    public void m_7380_(CompoundTag p_30496_) {
       super.m_7380_(p_30496_);
       p_30496_.m_128379_("ChestedHorse", this.m_30502_());
