--- ../src-base/minecraft/org/bukkit/block/Block.java
+++ ../src-work/minecraft/org/bukkit/block/Block.java
@@ -1,11 +1,10 @@
 package org.bukkit.block;
 
 import java.util.Collection;
-
 import org.bukkit.Chunk;
+import org.bukkit.Location;
 import org.bukkit.Material;
 import org.bukkit.World;
-import org.bukkit.Location;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.metadata.Metadatable;
 
