--- ../src-base/minecraft/org/bukkit/block/Structure.java
+++ ../src-work/minecraft/org/bukkit/block/Structure.java
@@ -37,13 +37,6 @@
     String getAuthor();
 
     /**
-     * Set the name of whoever created this structure.
-     *
-     * @param author whoever created this structure
-     */
-    void setAuthor(String author);
-
-    /**
      * Set the name of whoever created this structure using a
      * {@link LivingEntity}.
      *
@@ -52,6 +45,13 @@
     void setAuthor(LivingEntity livingEntity);
 
     /**
+     * Set the name of whoever created this structure.
+     *
+     * @param author whoever created this structure
+     */
+    void setAuthor(String author);
+
+    /**
      * The relative position of the structure outline based on the position of
      * the structure block. Maximum allowed distance is 32 blocks in any
      * direction.
@@ -91,6 +91,13 @@
     void setStructureSize(BlockVector vector);
 
     /**
+     * How this structure is mirrored.
+     *
+     * @return the current mirroring method
+     */
+    Mirror getMirror();
+
+    /**
      * Sets the mirroring of the structure.
      *
      * @param mirror the new mirroring method
@@ -98,11 +105,11 @@
     void setMirror(Mirror mirror);
 
     /**
-     * How this structure is mirrored.
+     * Get how this structure is rotated.
      *
-     * @return the current mirroring method
+     * @return the new rotation
      */
-    Mirror getMirror();
+    StructureRotation getRotation();
 
     /**
      * Set how this structure is rotated.
@@ -112,11 +119,11 @@
     void setRotation(StructureRotation rotation);
 
     /**
-     * Get how this structure is rotated.
+     * Get the {@link UsageMode} of this structure block.
      *
-     * @return the new rotation
+     * @return the mode this block is currently in.
      */
-    StructureRotation getRotation();
+    UsageMode getUsageMode();
 
     /**
      * Set the {@link UsageMode} of this structure block.
@@ -126,11 +133,11 @@
     void setUsageMode(UsageMode mode);
 
     /**
-     * Get the {@link UsageMode} of this structure block.
+     * Get if this structure block should ignore entities.
      *
-     * @return the mode this block is currently in.
+     * @return true if the appropriate {@link UsageMode} should ignore entities.
      */
-    UsageMode getUsageMode();
+    boolean isIgnoreEntities();
 
     /**
      * While in {@link UsageMode#SAVE} mode, this will ignore any entities when
@@ -144,11 +151,11 @@
     void setIgnoreEntities(boolean ignoreEntities);
 
     /**
-     * Get if this structure block should ignore entities.
+     * Check if this structure block is currently showing all air blocks
      *
-     * @return true if the appropriate {@link UsageMode} should ignore entities.
+     * @return true if the structure block is showing all air blocks
      */
-    boolean isIgnoreEntities();
+    boolean isShowAir();
 
     /**
      * Set if the structure outline should show air blocks.
@@ -158,11 +165,11 @@
     void setShowAir(boolean showAir);
 
     /**
-     * Check if this structure block is currently showing all air blocks
+     * Get if this structure block is currently showing the bounding box.
      *
-     * @return true if the structure block is showing all air blocks
+     * @return true if the bounding box is shown
      */
-    boolean isShowAir();
+    boolean isBoundingBoxVisible();
 
     /**
      * Set if this structure box should show the bounding box.
@@ -172,11 +179,11 @@
     void setBoundingBoxVisible(boolean showBoundingBox);
 
     /**
-     * Get if this structure block is currently showing the bounding box.
+     * Get the integrity of this structure.
      *
-     * @return true if the bounding box is shown
+     * @return the integrity of this structure
      */
-    boolean isBoundingBoxVisible();
+    float getIntegrity();
 
     /**
      * Set the integrity of the structure. Integrity must be between 0.0 and 1.0
@@ -189,11 +196,12 @@
     void setIntegrity(float integrity);
 
     /**
-     * Get the integrity of this structure.
+     * The seed used to determine how many blocks are removed upon loading of
+     * this structure.
      *
-     * @return the integrity of this structure
+     * @return the seed used
      */
-    float getIntegrity();
+    long getSeed();
 
     /**
      * The seed used to determine which blocks will be removed upon loading.
@@ -205,12 +213,14 @@
     void setSeed(long seed);
 
     /**
-     * The seed used to determine how many blocks are removed upon loading of
-     * this structure.
+     * Get the metadata function this structure block will perform when
+     * activated. Consult the
+     * <a href="https://minecraft.gamepedia.com/Structure_Block#Data">Minecraft
+     * Wiki</a> for more information.
      *
-     * @return the seed used
+     * @return the function that will be performed when this block is activated
      */
-    long getSeed();
+    String getMetadata();
 
     /**
      * Only applicable while in {@link UsageMode#DATA}. Metadata are specific
@@ -221,14 +231,4 @@
      * @param metadata the function to perform on the selected location
      */
     void setMetadata(String metadata);
-
-    /**
-     * Get the metadata function this structure block will perform when
-     * activated. Consult the
-     * <a href="https://minecraft.gamepedia.com/Structure_Block#Data">Minecraft
-     * Wiki</a> for more information.
-     *
-     * @return the function that will be performed when this block is activated
-     */
-    String getMetadata();
 }
