--- a/net/minecraft/world/gen/feature/structure/OceanRuinStructure.java
+++ b/net/minecraft/world/gen/feature/structure/OceanRuinStructure.java
@@ -8,6 +8,7 @@
 import net.minecraft.util.Rotation;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MutableBoundingBox;
+import net.minecraft.world.World;
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.gen.ChunkGenerator;
 import net.minecraft.world.gen.feature.Feature;
@@ -34,12 +35,19 @@
       return p_211745_1_.func_201496_a_().func_211727_m();
    }
 
+   @Override
+   protected int func_202382_c() {
+      return 0; // Default
+   }
+
    public Structure.IStartFactory func_214557_a() {
       return OceanRuinStructure.Start::new;
    }
 
-   protected int func_202382_c() {
-      return 14357621;
+   // Spigot start
+   protected int getSeedModifier(World world) {
+      return  world.spigotConfig.oceanSeed;
+      // Spigot end
    }
 
    public static class Start extends StructureStart {
