--- a/net/minecraft/world/gen/feature/structure/TemplateStructurePiece.java
+++ b/net/minecraft/world/gen/feature/structure/TemplateStructurePiece.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.gen.feature.structure;
 
+import com.mohistmc.util.i18n.i18n;
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import java.util.Random;
@@ -75,10 +_,10 @@
                   if (blockstate1 != null) {
                      blockstate = blockstate1;
                   } else {
-                     field_214825_d.error("Error while parsing blockstate {} in jigsaw block @ {}", s, template$blockinfo1.field_186242_a);
+                     field_214825_d.error(i18n.get("templatestructurepiece.1", s, template$blockinfo1.field_186242_a));
                   }
                } catch (CommandSyntaxException commandsyntaxexception) {
-                  field_214825_d.error("Error while parsing blockstate {} in jigsaw block @ {}", s, template$blockinfo1.field_186242_a);
+                  field_214825_d.error(i18n.get("templatestructurepiece.2", s, template$blockinfo1.field_186242_a));
                }
 
                p_230383_1_.func_180501_a(template$blockinfo1.field_186242_a, blockstate, 3);
