From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Tue, 11 Feb 2020 21:56:48 -0600
Subject: [PATCH] Implement EntityMoveEvent


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 60077cbcca5d2d8f3238cffab3bb89511a0e9f52..63acc1294e1273ac9fa3db442511195a57424447 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -17,6 +17,7 @@ import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
 import javax.annotation.Nullable;
+import net.pl3x.purpur.event.entity.EntityMoveEvent; // Purpur
 import org.apache.commons.lang3.tuple.Pair;
 
 // CraftBukkit start
@@ -2660,6 +2661,20 @@ public abstract class EntityLiving extends Entity {
 
         this.collideNearby();
         this.world.getMethodProfiler().exit();
+        // Purpur start
+        if (EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0) {
+            if (lastX != locX() || lastY != locY() || lastZ != locZ() || lastYaw != yaw || lastPitch != pitch) {
+                Location from = new Location(world.getWorld(), lastX, lastY, lastZ, lastYaw, lastPitch);
+                Location to = new Location (world.getWorld(), locX(), locY(), locZ(), yaw, pitch);
+                EntityMoveEvent event = new EntityMoveEvent(getBukkitLivingEntity(), from, to.clone());
+                if (!event.callEvent()) {
+                    setLocation(from.getX(), from.getY(), from.getZ(), from.getYaw(), from.getPitch());
+                } else if (!to.equals(event.getTo())) {
+                    setLocation(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+                }
+            }
+        }
+        // Purpur end
     }
 
     private void n() {
