--- a/net/minecraft/server/players/GameProfileCache.java
+++ b/net/minecraft/server/players/GameProfileCache.java
@@ -10,6 +_,7 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import com.mohistmc.MohistConfig;
 import com.mojang.authlib.Agent;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.GameProfileRepository;
@@ -90,6 +_,7 @@
             atomicreference.set((GameProfile)null);
          }
       };
+      if (MohistConfig.isProxyOnlineMode())
       p_10994_.findProfilesByNames(new String[]{p_10995_}, Agent.MINECRAFT, profilelookupcallback);
       GameProfile gameprofile = atomicreference.get();
       if (!m_11007_() && gameprofile == null) {
@@ -115,7 +_,7 @@
       Date date = calendar.getTime();
       GameProfileCache.GameProfileInfo gameprofilecache$gameprofileinfo = new GameProfileCache.GameProfileInfo(p_10992_, date);
       this.m_10979_(gameprofilecache$gameprofileinfo);
-      this.m_11006_();
+      if( !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly ) this.m_11006_(); // Spigot - skip saving if disabled
    }
 
    private long m_11008_() {
@@ -138,14 +_,14 @@
          gameprofilecache$gameprofileinfo.m_11029_(this.m_11008_());
          optional = Optional.of(gameprofilecache$gameprofileinfo.m_11028_());
       } else {
-         optional = m_10993_(this.f_10968_, s);
+         optional = m_10993_(this.f_10968_, p_10997_);  // Spigot - use correct case for offline players
          if (optional.isPresent()) {
             this.m_10991_(optional.get());
             flag = false;
          }
       }
 
-      if (flag) {
+      if (flag && !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) { // Spigot - skip saving if disabled
          this.m_11006_();
       }
 
@@ -220,7 +_,7 @@
    public void m_11006_() {
       JsonArray jsonarray = new JsonArray();
       DateFormat dateformat = m_11009_();
-      this.m_10977_(1000).forEach((p_143962_) -> {
+      this.m_10977_(org.spigotmc.SpigotConfig.userCacheCap).forEach((p_143962_) -> { // Spigot
          jsonarray.add(m_10981_(p_143962_, dateformat));
       });
       String s = this.f_10969_.toJson((JsonElement)jsonarray);
