--- ../src-base/minecraft/org/bukkit/event/enchantment/EnchantItemEvent.java
+++ ../src-work/minecraft/org/bukkit/event/enchantment/EnchantItemEvent.java
@@ -2,7 +2,6 @@
 
 import java.util.HashMap;
 import java.util.Map;
-
 import org.bukkit.block.Block;
 import org.bukkit.enchantments.Enchantment;
 import org.bukkit.entity.Player;
@@ -20,10 +19,10 @@
     private static final HandlerList handlers = new HandlerList();
     private final Block table;
     private final ItemStack item;
+    private final Map<Enchantment, Integer> enchants;
+    private final Player enchanter;
     private int level;
     private boolean cancelled;
-    private final Map<Enchantment,Integer> enchants;
-    private final Player enchanter;
     private int button;
 
     public EnchantItemEvent(final Player enchanter, final InventoryView view, final Block table, final ItemStack item, final int level, final Map<Enchantment, Integer> enchants, final int i) {
@@ -37,6 +36,10 @@
         this.button = i;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the player enchanting the item
      *
@@ -114,8 +117,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
