--- ../src-base/minecraft/net/minecraftforge/event/world/BlockEvent.java
+++ ../src-work/minecraft/net/minecraftforge/event/world/BlockEvent.java
@@ -24,21 +24,20 @@
 
 import net.minecraft.block.BlockPortal;
 import net.minecraft.block.state.IBlockState;
-import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
-import net.minecraft.init.Enchantments;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.world.World;
-import net.minecraftforge.common.ForgeHooks;
 import net.minecraftforge.common.util.BlockSnapshot;
 import net.minecraftforge.fml.common.eventhandler.Cancelable;
 import net.minecraftforge.fml.common.eventhandler.Event;
 
 import com.google.common.collect.ImmutableList;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
@@ -125,16 +124,17 @@
             super(world, pos, state);
             this.player = player;
 
-            if (state == null || !ForgeHooks.canHarvestBlock(state.getBlock(), player, world, pos) || // Handle empty block or player unable to break block scenario
-                (state.getBlock().canSilkHarvest(world, pos, world.getBlockState(pos), player) && EnchantmentHelper.getEnchantmentLevel(Enchantments.SILK_TOUCH, player.getHeldItemMainhand()) > 0)) // If the block is being silk harvested, the exp dropped is 0
-            {
-                this.exp = 0;
+            // Cauldron start - handle event on bukkit side
+            org.bukkit.event.block.BlockBreakEvent bukkitEvent = CraftEventFactory.callBlockBreakEvent(world, pos, state, (EntityPlayerMP) player);
+
+            if(bukkitEvent.isCancelled()){
+                this.setCanceled(true);
             }
             else
             {
-                int bonusLevel = EnchantmentHelper.getEnchantmentLevel(Enchantments.FORTUNE, player.getHeldItemMainhand());
-                this.exp = state.getBlock().getExpDrop(state, world, pos, bonusLevel);
+                this.exp = bukkitEvent.getExpToDrop();
             }
+            // Cauldron end
         }
 
         public EntityPlayer getPlayer()
@@ -213,6 +213,13 @@
             super(blockSnapshot, placedAgainst, player);
             this.player = player;
             this.hand = hand;
+            // Cauldron start - handle event on bukkit side
+            org.bukkit.craftbukkit.v1_12_R1.block.CraftBlockState blockstate = org.bukkit.craftbukkit.v1_12_R1.block.CraftBlockState.getBlockState(getWorld(), getPos().getX(), getPos().getY(), getPos().getZ());
+            org.bukkit.event.block.BlockPlaceEvent bukkitEvent = CraftEventFactory.callBlockPlaceEvent(getWorld(), player, hand , blockstate, getPos().getX(), getPos().getY(), getPos().getZ());
+            if (bukkitEvent.isCancelled() || !bukkitEvent.canBuild()) {
+                this.setCanceled(true);
+            }
+            // Cauldron end
             if (DEBUG)
             {
                 System.out.printf("Created PlaceEvent - [PlacedBlock: %s ][PlacedAgainst: %s ][ItemStack: %s ][Player: %s ][Hand: %s]\n", getPlacedBlock(), placedAgainst, player.getHeldItem(hand), player, hand);
@@ -399,6 +406,7 @@
         }
     }
 
+
     /**
      * Fired when a crop block grows.  See subevents.
      *
@@ -462,25 +470,20 @@
     @Cancelable
     public static class FarmlandTrampleEvent extends BlockEvent
     {
-
         private final Entity entity;
         private final float fallDistance;
-
         public FarmlandTrampleEvent(World world, BlockPos pos, IBlockState state, float fallDistance, Entity entity)
         {
             super(world, pos, state);
             this.entity = entity;
             this.fallDistance = fallDistance;
         }
-
         public Entity getEntity() {
             return entity;
         }
-
         public float getFallDistance() {
             return fallDistance;
         }
-
     }
 
     /* Fired when an attempt is made to spawn a nether portal from
@@ -492,13 +495,11 @@
     public static class PortalSpawnEvent extends BlockEvent
     {
         private final BlockPortal.Size size;
-
         public PortalSpawnEvent(World world, BlockPos pos, IBlockState state, BlockPortal.Size size)
         {
             super(world, pos, state);
             this.size = size;
         }
-
         public BlockPortal.Size getPortalSize()
         {
             return size;
