--- a/net/minecraft/world/level/storage/loot/LootDataType.java
+++ b/net/minecraft/world/level/storage/loot/LootDataType.java
@@ -42,7 +_,10 @@
       dataresult.error().ifPresent((p_297003_) -> {
          f_278425_.error("Couldn't parse element {}:{} - {}", this.f_278462_, p_279253_, p_297003_.message());
       });
-      return dataresult.result();
+      var ret = dataresult.result();
+      if (this == f_278413_ && ret.isPresent())
+         ret = Optional.ofNullable((T)net.minecraftforge.event.ForgeEventFactory.onLoadLootTable(p_279253_, (LootTable)ret.get()));
+      return ret;
    }
 
    public static Stream<LootDataType<?>> m_278779_() {
