--- a/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
@@ -44,7 +44,7 @@
    }
 
    public void func_75249_e() {
-      this.field_75305_a.func_70624_b(this.field_75304_b);
+      this.field_75305_a.setGoalTarget(this.field_75304_b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
       super.func_75249_e();
    }
 }
