--- a/net/minecraft/world/entity/animal/Animal.java
+++ b/net/minecraft/world/entity/animal/Animal.java
@@ -29,12 +_,15 @@
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityEnterLoveModeEvent;
 
 public abstract class Animal extends AgeableMob {
    protected static final int PARENT_AGE_AFTER_BREEDING = 6000;
    public int inLove;
    @Nullable
    public UUID loveCause;
+   public ItemStack breedItem; // CraftBukkit - Add breedItem variable
 
    protected Animal(EntityType<? extends Animal> p_27557_, Level p_27558_) {
       super(p_27557_, p_27558_);
@@ -72,8 +_,13 @@
       if (this.isInvulnerableTo(p_27567_)) {
          return false;
       } else {
-         this.inLove = 0;
-         return super.hurt(p_27567_, p_27568_);
+         // CraftBukkit start
+         boolean result = super.hurt(p_27567_, p_27568_);
+         if (result) {
+            this.inLove = 0;
+         }
+         return result;
+         // CraftBukkit end
       }
    }
 
@@ -159,7 +_,15 @@
    public void setInLove(@Nullable Player p_27596_) {
       this.inLove = 600;
       if (p_27596_ != null) {
+         // CraftBukkit start
+         EntityEnterLoveModeEvent entityEnterLoveModeEvent = CraftEventFactory.callEntityEnterLoveModeEvent(p_27596_, this, 600);
+         if (entityEnterLoveModeEvent.isCancelled()) {
+            return;
+         }
+         this.setInLoveTime(entityEnterLoveModeEvent.getTicksInLove());
+         // CraftBukkit end
          this.loveCause = p_27596_.getUUID();
+         this.breedItem = p_27596_.getInventory().getSelected(); // CraftBukkit
       }
 
       this.level().broadcastEntityEvent(this, (byte)18);
@@ -203,28 +_,61 @@
 
    public void spawnChildFromBreeding(ServerLevel p_27564_, Animal p_27565_) {
       AgeableMob ageablemob = this.getBreedOffspring(p_27564_, p_27565_);
+      final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(this, p_27565_, ageablemob);
+      final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+      ageablemob = event.getChild();
+      if (cancelled) {
+         //Reset the "inLove" state for the animals
+         this.setAge(6000);
+         p_27565_.setAge(6000);
+         this.resetLove();
+         p_27565_.resetLove();
+         return;
+      }
       if (ageablemob != null) {
          ageablemob.setBaby(true);
          ageablemob.moveTo(this.getX(), this.getY(), this.getZ(), 0.0F, 0.0F);
          this.finalizeSpawnChildFromBreeding(p_27564_, p_27565_, ageablemob);
          p_27564_.addFreshEntityWithPassengers(ageablemob);
+         // CraftBukkit start - call EntityBreedEvent
+         ServerPlayer breeder = Optional.ofNullable(this.getLoveCause()).or(() -> {
+            return Optional.ofNullable(p_27565_.getLoveCause());
+         }).orElse(null);
+         int experience = this.getRandom().nextInt(7) + 1;
+         org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreedEvent(ageablemob, this, p_27565_, breeder, this.breedItem, experience);
+         if (entityBreedEvent.isCancelled()) {
+            return;
+         }
+         experience = entityBreedEvent.getExperience();
+         this.finalizeSpawnChildFromBreeding(p_27564_, p_27565_, ageablemob, experience);
+         ageablemob.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);
+         p_27564_.addFreshEntityWithPassengers(ageablemob);
+         // CraftBukkit end
       }
    }
-
+   
    public void finalizeSpawnChildFromBreeding(ServerLevel p_277963_, Animal p_277357_, @Nullable AgeableMob p_277516_) {
+      // CraftBukkit start
+      this.finalizeSpawnChildFromBreeding(p_277963_, p_277357_, p_277516_, this.getRandom().nextInt(7) + 1);
+   }
+   public void finalizeSpawnChildFromBreeding(ServerLevel pLevel, Animal pAnimal, @Nullable AgeableMob pBaby, int experience) {
       Optional.ofNullable(this.getLoveCause()).or(() -> {
-         return Optional.ofNullable(p_277357_.getLoveCause());
+         return Optional.ofNullable(pAnimal.getLoveCause());
       }).ifPresent((p_277486_) -> {
          p_277486_.awardStat(Stats.ANIMALS_BRED);
-         CriteriaTriggers.BRED_ANIMALS.trigger(p_277486_, this, p_277357_, p_277516_);
+         CriteriaTriggers.BRED_ANIMALS.trigger(p_277486_, this, pAnimal, pBaby);
       });
       this.setAge(6000);
-      p_277357_.setAge(6000);
+      pAnimal.setAge(6000);
       this.resetLove();
-      p_277357_.resetLove();
-      p_277963_.broadcastEntityEvent(this, (byte)18);
-      if (p_277963_.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
-         p_277963_.addFreshEntity(new ExperienceOrb(p_277963_, this.getX(), this.getY(), this.getZ(), this.getRandom().nextInt(7) + 1));
+      pAnimal.resetLove();
+      pLevel.broadcastEntityEvent(this, (byte)18);
+      if (pLevel.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+         // CraftBukkit start - use event experience
+         if (experience > 0) {
+            pLevel.addFreshEntity(new ExperienceOrb(pLevel, this.getX(), this.getY(), this.getZ(), experience));
+         }
+         // CraftBukkit end
       }
 
    }
