--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -79,7 +_,6 @@
 import net.minecraft.world.level.block.SweetBerryBushBlock;
 import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntity;
-import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
@@ -130,7 +_,8 @@
    @Nullable
    BlockPos savedFlowerPos;
    @Nullable
-   public BlockPos hivePos;
+   public
+   BlockPos hivePos;
    Bee.BeePollinateGoal beePollinateGoal;
    Bee.BeeGoToHiveGoal goToHiveGoal;
    private Bee.BeeGoToKnownFlowerGoal goToKnownFlowerGoal;
@@ -179,12 +_,19 @@
    }
 
    public void addAdditionalSaveData(CompoundTag p_27823_) {
+      // CraftBukkit start - selectively save data
+      addAdditionalSaveData(p_27823_, true);
+   }
+
+   @Override
+   public void addAdditionalSaveData(CompoundTag p_27823_, boolean includeAll) {
+      // CraftBukkit end
       super.addAdditionalSaveData(p_27823_);
-      if (this.hasHive()) {
+      if (includeAll && this.hasHive()) { // CraftBukkit - selectively save hive
          p_27823_.put("HivePos", NbtUtils.writeBlockPos(this.getHivePos()));
       }
 
-      if (this.hasSavedFlowerPos()) {
+      if (includeAll && this.hasSavedFlowerPos()) { // CraftBukkit - selectively save flower
          p_27823_.put("FlowerPos", NbtUtils.writeBlockPos(this.getSavedFlowerPos()));
       }
 
@@ -230,7 +_,7 @@
             }
 
             if (i > 0) {
-               ((LivingEntity)p_27722_).addEffect(new MobEffectInstance(MobEffects.POISON, i * 20, 0), this);
+               ((LivingEntity)p_27722_).addEffect(new MobEffectInstance(MobEffects.POISON, i * 20, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
          }
 
@@ -473,7 +_,7 @@
          return false;
       } else {
          BlockEntity blockentity = this.level().getBlockEntity(this.hivePos);
-         return blockentity != null && blockentity.getType() == BlockEntityType.BEEHIVE;
+         return blockentity instanceof BeehiveBlockEntity;
       }
    }
 
@@ -481,7 +_,7 @@
       return this.getFlag(8);
    }
 
-   void setHasNectar(boolean p_27920_) {
+   public void setHasNectar(boolean p_27920_) {
       if (p_27920_) {
          this.resetTicksWithoutNectarSinceExitingHive();
       }
@@ -493,7 +_,7 @@
       return this.getFlag(4);
    }
 
-   private void setHasStung(boolean p_27926_) {
+   public void setHasStung(boolean p_27926_) {
       this.setFlag(4, p_27926_);
    }
 
@@ -600,11 +_,14 @@
       if (this.isInvulnerableTo(p_27762_)) {
          return false;
       } else {
-         if (!this.level().isClientSide) {
+         // CraftBukkit start - Only stop pollinating if entity was damaged
+         boolean result = super.hurt(p_27762_, p_27763_);
+         if (result && !this.level().isClientSide) {
+            // CraftBukkit end
             this.beePollinateGoal.stopPollinating();
          }
 
-         return super.hurt(p_27762_, p_27763_);
+         return result; // CraftBukkit
       }
    }
 
@@ -612,10 +_,20 @@
       return MobType.ARTHROPOD;
    }
 
+   @Deprecated // FORGE: use jumpInFluid instead
    protected void jumpInLiquid(TagKey<Fluid> p_204061_) {
+      this.jumpInLiquidInternal();
+   }
+
+   private void jumpInLiquidInternal() {
       this.setDeltaMovement(this.getDeltaMovement().add(0.0D, 0.01D, 0.0D));
    }
 
+   @Override
+   public void jumpInFluid(net.minecraftforge.fluids.FluidType type) {
+      this.jumpInLiquidInternal();
+   }
+
    public Vec3 getLeashOffset() {
       return new Vec3(0.0D, (double)(0.5F * this.getEyeHeight()), (double)(this.getBbWidth() * 0.2F));
    }
@@ -710,7 +_,7 @@
    @VisibleForDebug
    public class BeeGoToHiveGoal extends Bee.BaseBeeGoal {
       public static final int MAX_TRAVELLING_TICKS = 600;
-      int travellingTicks = Bee.this.level().random.nextInt(10);
+      int travellingTicks = Bee.this.random.nextInt(10); // CraftBukkit - SPIGOT-7495: Give Bees another chance and let them use their own random, avoid concurrency issues
       private static final int MAX_BLACKLISTED_TARGETS = 3;
       final List<BlockPos> blacklistedTargets = Lists.newArrayList();
       @Nullable
@@ -822,7 +_,7 @@
 
    public class BeeGoToKnownFlowerGoal extends Bee.BaseBeeGoal {
       private static final int MAX_TRAVELLING_TICKS = 600;
-      int travellingTicks = Bee.this.level().random.nextInt(10);
+      int travellingTicks = Bee.this.random.nextInt(10); // CraftBukkit - SPIGOT-7495: Give Bees another chance and let them use their own random, avoid concurrency issues
 
       BeeGoToKnownFlowerGoal() {
          this.setFlags(EnumSet.of(Goal.Flag.MOVE));
@@ -911,7 +_,7 @@
                      ((BonemealableBlock)blockstate.getBlock()).performBonemeal((ServerLevel)Bee.this.level(), Bee.this.random, blockpos, blockstate);
                   }
 
-                  if (blockstate1 != null) {
+                  if (blockstate1 != null && org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityChangeBlockEvent(Bee.this, blockpos, blockstate1)) { // CraftBukkit
                      Bee.this.level().levelEvent(2005, blockpos, 0);
                      Bee.this.level().setBlockAndUpdate(blockpos, blockstate1);
                      Bee.this.incrementNumCropsGrownSincePollination();
@@ -934,7 +_,7 @@
 
       protected void alertOther(Mob p_28035_, LivingEntity p_28036_) {
          if (p_28035_ instanceof Bee && this.mob.hasLineOfSight(p_28036_)) {
-            p_28035_.setTarget(p_28036_);
+            p_28035_.setTarget(p_28036_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
          }
 
       }
