--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -31,6 +_,7 @@
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
@@ -64,6 +_,8 @@
 import net.minecraft.world.item.DyeItem;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.BedBlock;
@@ -74,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
+import org.joml.Vector3f;
 
 public class Cat extends TamableAnimal implements VariantHolder<Holder<CatVariant>> {
     public static final double TEMPT_SPEED_MOD = 0.6;
@@ -152,7 +_,7 @@
         return DyeColor.byId(this.entityData.get(DATA_COLLAR_COLOR));
     }
 
-    private void setCollarColor(DyeColor p_28132_) {
+    public void setCollarColor(DyeColor p_28132_) {
         this.entityData.set(DATA_COLLAR_COLOR, p_28132_.getId());
     }
 
@@ -379,9 +_,9 @@
                     }
                 } else if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
                     if (!this.level().isClientSide()) {
-                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                         FoodProperties foodproperties = itemstack.get(DataComponents.FOOD);
                         this.heal(foodproperties != null ? (float)foodproperties.nutrition() : 1.0F);
+                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
@@ -441,7 +_,7 @@
     }
 
     private void tryToTame(Player p_333297_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_333297_)) {
             this.tame(p_333297_);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
@@ -579,20 +_,19 @@
                 .withParameter(LootContextParams.THIS_ENTITY, this.cat)
                 .create(LootContextParamSets.GIFT);
 
-            for (ItemStack itemstack : loottable.getRandomItems(lootparams)) {
-                this.cat
-                    .level()
-                    .addFreshEntity(
-                        new ItemEntity(
-                            this.cat.level(),
-                            (double)blockpos$mutableblockpos.getX() - (double)Mth.sin(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
-                            (double)blockpos$mutableblockpos.getY(),
-                            (double)blockpos$mutableblockpos.getZ() + (double)Mth.cos(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
-                            itemstack
-                        )
-                    );
+         for(ItemStack itemstack : loottable.getRandomItems(lootparams)) {
+            // CraftBukkit start
+            ItemEntity entityitem = new ItemEntity(this.cat.level, (double) blockpos$mutableblockpos.getX() - (double) Mth.sin(this.cat.yBodyRot * 0.017453292F), (double) blockpos$mutableblockpos.getY(), (double) blockpos$mutableblockpos.getZ() + (double) Mth.cos(this.cat.yBodyRot * 0.017453292F), itemstack);
+            org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.cat.getBukkitEntity(), (org.bukkit.entity.Item) entityitem.getBukkitEntity());
+            entityitem.level.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               continue;
             }
-        }
+            this.cat.level.addFreshEntity(entityitem);
+            // CraftBukkit end
+         }
+
+      }
 
         @Override
         public void tick() {
