--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -366,8 +_,8 @@
             if (this.isOwnedBy(p_28153_)) {
                if (!(item instanceof DyeItem)) {
                   if (item.isEdible() && this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
+                     this.heal((float)itemstack.getFoodProperties(this).getNutrition());
                      this.usePlayerItem(p_28153_, p_28154_, itemstack);
-                     this.heal((float)item.getFoodProperties().getNutrition());
                      return InteractionResult.CONSUME;
                   }
 
@@ -392,7 +_,7 @@
             }
          } else if (this.isFood(itemstack)) {
             this.usePlayerItem(p_28153_, p_28154_, itemstack);
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && (!net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_28153_) || !org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTameEvent(this, p_28153_).isCancelled())) {
                this.tame(p_28153_);
                this.setOrderedToSit(true);
                this.level().broadcastEntityEvent(this, (byte)7);
@@ -551,7 +_,15 @@
          LootParams lootparams = (new LootParams.Builder((ServerLevel)this.cat.level())).withParameter(LootContextParams.ORIGIN, this.cat.position()).withParameter(LootContextParams.THIS_ENTITY, this.cat).create(LootContextParamSets.GIFT);
 
          for(ItemStack itemstack : loottable.getRandomItems(lootparams)) {
-            this.cat.level().addFreshEntity(new ItemEntity(this.cat.level(), (double)blockpos$mutableblockpos.getX() - (double)Mth.sin(this.cat.yBodyRot * ((float)Math.PI / 180F)), (double)blockpos$mutableblockpos.getY(), (double)blockpos$mutableblockpos.getZ() + (double)Mth.cos(this.cat.yBodyRot * ((float)Math.PI / 180F)), itemstack));
+            // CraftBukkit start
+            ItemEntity entityitem = new ItemEntity(this.cat.level, (double) blockpos$mutableblockpos.getX() - (double) Mth.sin(this.cat.yBodyRot * 0.017453292F), (double) blockpos$mutableblockpos.getY(), (double) blockpos$mutableblockpos.getZ() + (double) Mth.cos(this.cat.yBodyRot * 0.017453292F), itemstack);
+            org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.cat.getBukkitEntity(), (org.bukkit.entity.Item) entityitem.getBukkitEntity());
+            entityitem.level.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               continue;
+            }
+            this.cat.level.addFreshEntity(entityitem);
+            // CraftBukkit end
          }
 
       }
