--- a/net/minecraft/world/entity/animal/Ocelot.java
+++ b/net/minecraft/world/entity/animal/Ocelot.java
@@ -67,11 +_,11 @@
       this.reassessTrustingGoals();
    }
 
-   boolean isTrusting() {
+   public boolean isTrusting() {
       return this.entityData.get(DATA_TRUSTING);
    }
 
-   private void setTrusting(boolean p_29046_) {
+   public void setTrusting(boolean p_29046_) {
       this.entityData.set(DATA_TRUSTING, p_29046_);
       this.reassessTrustingGoals();
    }
@@ -162,7 +_,7 @@
       if ((this.temptGoal == null || this.temptGoal.isRunning()) && !this.isTrusting() && this.isFood(itemstack) && p_29021_.distanceToSqr(this) < 9.0D) {
          this.usePlayerItem(p_29021_, p_29022_, itemstack);
          if (!this.level().isClientSide) {
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && (!net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_29021_) || !org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTameEvent(this, p_29021_).isCancelled())) {
                this.setTrusting(true);
                this.spawnTrustingParticles(true);
                this.level().broadcastEntityEvent(this, (byte)41);
