--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -58,6 +_,8 @@
 import net.minecraft.world.entity.monster.Monster;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
@@ -70,6 +_,8 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
+import org.joml.Vector3f;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class Panda extends Animal {
     private static final EntityDataAccessor<Integer> UNHAPPY_COUNTER = SynchedEntityData.defineId(Panda.class, EntityDataSerializers.INT);
@@ -552,7 +_,8 @@
 
     @Override
     protected void pickUpItem(ItemEntity p_29121_) {
-        if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_29121_)) {
+      	//if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && PANDA_ITEMS.test(pItemEntity)) {
+      	if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, p_29121_, 0, !(this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_29121_))).isCancelled()) { // CraftBukkit
             this.onItemPickup(p_29121_);
             ItemStack itemstack = p_29121_.getItem();
             this.setItemSlot(EquipmentSlot.MAINHAND, itemstack);
@@ -887,7 +_,7 @@
         @Override
         protected void alertOther(Mob p_29295_, LivingEntity p_29296_) {
             if (p_29295_ instanceof Panda && p_29295_.isAggressive()) {
-                p_29295_.setTarget(p_29296_);
+            p_29295_.setTarget(p_29296_, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit
             }
         }
     }
