--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -48,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.joml.Vector3f;
+import org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory;
 
 public class Pig extends Animal implements ItemSteerable, Saddleable {
    private static final EntityDataAccessor<Boolean> DATA_SADDLE_ID = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.BOOLEAN);
@@ -203,7 +_,7 @@
    }
 
    public void thunderHit(ServerLevel p_29473_, LightningBolt p_29474_) {
-      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL) {
+      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> {})) {
          ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(p_29473_);
          if (zombifiedpiglin != null) {
             zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
@@ -216,7 +_,15 @@
             }
 
             zombifiedpiglin.setPersistenceRequired();
+            net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglin);
+            // CraftBukkit start
+            if (CraftEventFactory.callPigZapEvent(this, p_29474_, zombifiedpiglin).isCancelled()) {
+               return;
+            }
+            // CraftBukkit - added a reason for spawning this creature
+            zombifiedpiglin.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
             p_29473_.addFreshEntity(zombifiedpiglin);
+            // CraftBukkit end
             this.discard();
          } else {
             super.thunderHit(p_29473_, p_29474_);
