--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -11,12 +_,13 @@
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
-import net.minecraft.tags.ItemTags;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.AgeableMob;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EntityDimensions;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.ItemBasedSteering;
@@ -41,227 +_,233 @@
 import net.minecraft.world.entity.vehicle.DismountHelper;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
+import net.minecraft.world.item.crafting.Ingredient;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.joml.Vector3f;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class Pig extends Animal implements ItemSteerable, Saddleable {
-    private static final EntityDataAccessor<Boolean> DATA_SADDLE_ID = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.BOOLEAN);
-    private static final EntityDataAccessor<Integer> DATA_BOOST_TIME = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.INT);
-    public final ItemBasedSteering steering = new ItemBasedSteering(this.entityData, DATA_BOOST_TIME, DATA_SADDLE_ID);
-
-    public Pig(EntityType<? extends Pig> p_29462_, Level p_29463_) {
-        super(p_29462_, p_29463_);
-    }
-
-    @Override
-    protected void registerGoals() {
-        this.goalSelector.addGoal(0, new FloatGoal(this));
-        this.goalSelector.addGoal(1, new PanicGoal(this, 1.25));
-        this.goalSelector.addGoal(3, new BreedGoal(this, 1.0));
-        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, p_332330_ -> p_332330_.is(Items.CARROT_ON_A_STICK), false));
-        this.goalSelector.addGoal(4, new TemptGoal(this, 1.2, p_332514_ -> p_332514_.is(ItemTags.PIG_FOOD), false));
-        this.goalSelector.addGoal(5, new FollowParentGoal(this, 1.1));
-        this.goalSelector.addGoal(6, new WaterAvoidingRandomStrollGoal(this, 1.0));
-        this.goalSelector.addGoal(7, new LookAtPlayerGoal(this, Player.class, 6.0F));
-        this.goalSelector.addGoal(8, new RandomLookAroundGoal(this));
-    }
-
-    public static AttributeSupplier.Builder createAttributes() {
-        return Mob.createMobAttributes().add(Attributes.MAX_HEALTH, 10.0).add(Attributes.MOVEMENT_SPEED, 0.25);
-    }
-
-    @Nullable
-    @Override
-    public LivingEntity getControllingPassenger() {
-        return (LivingEntity)(this.isSaddled() && this.getFirstPassenger() instanceof Player player && player.isHolding(Items.CARROT_ON_A_STICK) ? player : super.getControllingPassenger());
-    }
-
-    @Override
-    public void onSyncedDataUpdated(EntityDataAccessor<?> p_29480_) {
-        if (DATA_BOOST_TIME.equals(p_29480_) && this.level().isClientSide) {
-            this.steering.onSynced();
-        }
-
-        super.onSyncedDataUpdated(p_29480_);
-    }
-
-    @Override
-    protected void defineSynchedData(SynchedEntityData.Builder p_335143_) {
-        super.defineSynchedData(p_335143_);
-        p_335143_.define(DATA_SADDLE_ID, false);
-        p_335143_.define(DATA_BOOST_TIME, 0);
-    }
-
-    @Override
-    public void addAdditionalSaveData(CompoundTag p_29495_) {
-        super.addAdditionalSaveData(p_29495_);
-        this.steering.addAdditionalSaveData(p_29495_);
-    }
-
-    @Override
-    public void readAdditionalSaveData(CompoundTag p_29478_) {
-        super.readAdditionalSaveData(p_29478_);
-        this.steering.readAdditionalSaveData(p_29478_);
-    }
-
-    @Override
-    protected SoundEvent getAmbientSound() {
-        return SoundEvents.PIG_AMBIENT;
-    }
-
-    @Override
-    protected SoundEvent getHurtSound(DamageSource p_29502_) {
-        return SoundEvents.PIG_HURT;
-    }
-
-    @Override
-    protected SoundEvent getDeathSound() {
-        return SoundEvents.PIG_DEATH;
-    }
-
-    @Override
-    protected void playStepSound(BlockPos p_29492_, BlockState p_29493_) {
-        this.playSound(SoundEvents.PIG_STEP, 0.15F, 1.0F);
-    }
-
-    @Override
-    public InteractionResult mobInteract(Player p_29489_, InteractionHand p_29490_) {
-        boolean flag = this.isFood(p_29489_.getItemInHand(p_29490_));
-        if (!flag && this.isSaddled() && !this.isVehicle() && !p_29489_.isSecondaryUseActive()) {
-            if (!this.level().isClientSide) {
-                p_29489_.startRiding(this);
-            }
-
-            return InteractionResult.sidedSuccess(this.level().isClientSide);
-        } else {
-            InteractionResult interactionresult = super.mobInteract(p_29489_, p_29490_);
-            if (!interactionresult.consumesAction()) {
-                ItemStack itemstack = p_29489_.getItemInHand(p_29490_);
-                return itemstack.is(Items.SADDLE) ? itemstack.interactLivingEntity(p_29489_, this, p_29490_) : InteractionResult.PASS;
-            } else {
-                return interactionresult;
-            }
-        }
-    }
-
-    @Override
-    public boolean isSaddleable() {
-        return this.isAlive() && !this.isBaby();
-    }
-
-    @Override
-    protected void dropEquipment() {
-        super.dropEquipment();
-        if (this.isSaddled()) {
-            this.spawnAtLocation(Items.SADDLE);
-        }
-    }
-
-    @Override
-    public boolean isSaddled() {
-        return this.steering.hasSaddle();
-    }
-
-    @Override
-    public void equipSaddle(@Nullable SoundSource p_29476_) {
-        this.steering.setSaddle(true);
-        if (p_29476_ != null) {
-            this.level().playSound(null, this, SoundEvents.PIG_SADDLE, p_29476_, 0.5F, 1.0F);
-        }
-    }
-
-    @Override
-    public Vec3 getDismountLocationForPassenger(LivingEntity p_29487_) {
-        Direction direction = this.getMotionDirection();
-        if (direction.getAxis() == Direction.Axis.Y) {
-            return super.getDismountLocationForPassenger(p_29487_);
-        } else {
-            int[][] aint = DismountHelper.offsetsForDirection(direction);
-            BlockPos blockpos = this.blockPosition();
-            BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
-
-            for (Pose pose : p_29487_.getDismountPoses()) {
-                AABB aabb = p_29487_.getLocalBoundsForPose(pose);
-
-                for (int[] aint1 : aint) {
-                    blockpos$mutableblockpos.set(blockpos.getX() + aint1[0], blockpos.getY(), blockpos.getZ() + aint1[1]);
-                    double d0 = this.level().getBlockFloorHeight(blockpos$mutableblockpos);
-                    if (DismountHelper.isBlockFloorValid(d0)) {
-                        Vec3 vec3 = Vec3.upFromBottomCenterOf(blockpos$mutableblockpos, d0);
-                        if (DismountHelper.canDismountTo(this.level(), p_29487_, aabb.move(vec3))) {
-                            p_29487_.setPose(pose);
-                            return vec3;
-                        }
-                    }
-                }
-            }
-
-            return super.getDismountLocationForPassenger(p_29487_);
-        }
-    }
-
-    @Override
-    public void thunderHit(ServerLevel p_29473_, LightningBolt p_29474_) {
-        if (p_29473_.getDifficulty() != Difficulty.PEACEFUL) {
-            ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(p_29473_);
-            if (zombifiedpiglin != null) {
-                zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
-                zombifiedpiglin.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
-                zombifiedpiglin.setNoAi(this.isNoAi());
-                zombifiedpiglin.setBaby(this.isBaby());
-                if (this.hasCustomName()) {
-                    zombifiedpiglin.setCustomName(this.getCustomName());
-                    zombifiedpiglin.setCustomNameVisible(this.isCustomNameVisible());
-                }
-
-                zombifiedpiglin.setPersistenceRequired();
-                p_29473_.addFreshEntity(zombifiedpiglin);
-                this.discard();
-            } else {
-                super.thunderHit(p_29473_, p_29474_);
-            }
-        } else {
+   private static final EntityDataAccessor<Boolean> DATA_SADDLE_ID = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.BOOLEAN);
+   private static final EntityDataAccessor<Integer> DATA_BOOST_TIME = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.INT);
+   private static final Ingredient FOOD_ITEMS = Ingredient.of(Items.CARROT, Items.POTATO, Items.BEETROOT);
+   public final ItemBasedSteering steering = new ItemBasedSteering(this.entityData, DATA_BOOST_TIME, DATA_SADDLE_ID);
+
+   public Pig(EntityType<? extends Pig> p_29462_, Level p_29463_) {
+      super(p_29462_, p_29463_);
+   }
+
+   protected void registerGoals() {
+      this.goalSelector.addGoal(0, new FloatGoal(this));
+      this.goalSelector.addGoal(1, new PanicGoal(this, 1.25D));
+      this.goalSelector.addGoal(3, new BreedGoal(this, 1.0D));
+      this.goalSelector.addGoal(4, new TemptGoal(this, 1.2D, Ingredient.of(Items.CARROT_ON_A_STICK), false));
+      this.goalSelector.addGoal(4, new TemptGoal(this, 1.2D, FOOD_ITEMS, false));
+      this.goalSelector.addGoal(5, new FollowParentGoal(this, 1.1D));
+      this.goalSelector.addGoal(6, new WaterAvoidingRandomStrollGoal(this, 1.0D));
+      this.goalSelector.addGoal(7, new LookAtPlayerGoal(this, Player.class, 6.0F));
+      this.goalSelector.addGoal(8, new RandomLookAroundGoal(this));
+   }
+
+   public static AttributeSupplier.Builder createAttributes() {
+      return Mob.createMobAttributes().add(Attributes.MAX_HEALTH, 10.0D).add(Attributes.MOVEMENT_SPEED, 0.25D);
+   }
+
+   @Nullable
+   public LivingEntity getControllingPassenger() {
+      if (this.isSaddled()) {
+         Entity entity = this.getFirstPassenger();
+         if (entity instanceof Player) {
+            Player player = (Player)entity;
+            if (player.isHolding(Items.CARROT_ON_A_STICK)) {
+               return player;
+            }
+         }
+      }
+
+      return super.getControllingPassenger();
+   }
+
+   public void onSyncedDataUpdated(EntityDataAccessor<?> p_29480_) {
+      if (DATA_BOOST_TIME.equals(p_29480_) && this.level().isClientSide) {
+         this.steering.onSynced();
+      }
+
+      super.onSyncedDataUpdated(p_29480_);
+   }
+
+   protected void defineSynchedData() {
+      super.defineSynchedData();
+      this.entityData.define(DATA_SADDLE_ID, false);
+      this.entityData.define(DATA_BOOST_TIME, 0);
+   }
+
+   public void addAdditionalSaveData(CompoundTag p_29495_) {
+      super.addAdditionalSaveData(p_29495_);
+      this.steering.addAdditionalSaveData(p_29495_);
+   }
+
+   public void readAdditionalSaveData(CompoundTag p_29478_) {
+      super.readAdditionalSaveData(p_29478_);
+      this.steering.readAdditionalSaveData(p_29478_);
+   }
+
+   protected SoundEvent getAmbientSound() {
+      return SoundEvents.PIG_AMBIENT;
+   }
+
+   protected SoundEvent getHurtSound(DamageSource p_29502_) {
+      return SoundEvents.PIG_HURT;
+   }
+
+   protected SoundEvent getDeathSound() {
+      return SoundEvents.PIG_DEATH;
+   }
+
+   protected void playStepSound(BlockPos p_29492_, BlockState p_29493_) {
+      this.playSound(SoundEvents.PIG_STEP, 0.15F, 1.0F);
+   }
+
+   public InteractionResult mobInteract(Player p_29489_, InteractionHand p_29490_) {
+      boolean flag = this.isFood(p_29489_.getItemInHand(p_29490_));
+      if (!flag && this.isSaddled() && !this.isVehicle() && !p_29489_.isSecondaryUseActive()) {
+         if (!this.level().isClientSide) {
+            p_29489_.startRiding(this);
+         }
+
+         return InteractionResult.sidedSuccess(this.level().isClientSide);
+      } else {
+         InteractionResult interactionresult = super.mobInteract(p_29489_, p_29490_);
+         if (!interactionresult.consumesAction()) {
+            ItemStack itemstack = p_29489_.getItemInHand(p_29490_);
+            return itemstack.is(Items.SADDLE) ? itemstack.interactLivingEntity(p_29489_, this, p_29490_) : InteractionResult.PASS;
+         } else {
+            return interactionresult;
+         }
+      }
+   }
+
+   public boolean isSaddleable() {
+      return this.isAlive() && !this.isBaby();
+   }
+
+   protected void dropEquipment() {
+      super.dropEquipment();
+      if (this.isSaddled()) {
+         this.spawnAtLocation(Items.SADDLE);
+      }
+
+   }
+
+   public boolean isSaddled() {
+      return this.steering.hasSaddle();
+   }
+
+   public void equipSaddle(@Nullable SoundSource p_29476_) {
+      this.steering.setSaddle(true);
+      if (p_29476_ != null) {
+         this.level().playSound((Player)null, this, SoundEvents.PIG_SADDLE, p_29476_, 0.5F, 1.0F);
+      }
+
+   }
+
+   public Vec3 getDismountLocationForPassenger(LivingEntity p_29487_) {
+      Direction direction = this.getMotionDirection();
+      if (direction.getAxis() == Direction.Axis.Y) {
+         return super.getDismountLocationForPassenger(p_29487_);
+      } else {
+         int[][] aint = DismountHelper.offsetsForDirection(direction);
+         BlockPos blockpos = this.blockPosition();
+         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
+
+         for(Pose pose : p_29487_.getDismountPoses()) {
+            AABB aabb = p_29487_.getLocalBoundsForPose(pose);
+
+            for(int[] aint1 : aint) {
+               blockpos$mutableblockpos.set(blockpos.getX() + aint1[0], blockpos.getY(), blockpos.getZ() + aint1[1]);
+               double d0 = this.level().getBlockFloorHeight(blockpos$mutableblockpos);
+               if (DismountHelper.isBlockFloorValid(d0)) {
+                  Vec3 vec3 = Vec3.upFromBottomCenterOf(blockpos$mutableblockpos, d0);
+                  if (DismountHelper.canDismountTo(this.level(), p_29487_, aabb.move(vec3))) {
+                     p_29487_.setPose(pose);
+                     return vec3;
+                  }
+               }
+            }
+         }
+
+         return super.getDismountLocationForPassenger(p_29487_);
+      }
+   }
+
+   public void thunderHit(ServerLevel p_29473_, LightningBolt p_29474_) {
+      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> {})) {
+         ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(p_29473_);
+         if (zombifiedpiglin != null) {
+            zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
+            zombifiedpiglin.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
+            zombifiedpiglin.setNoAi(this.isNoAi());
+            zombifiedpiglin.setBaby(this.isBaby());
+            if (this.hasCustomName()) {
+               zombifiedpiglin.setCustomName(this.getCustomName());
+               zombifiedpiglin.setCustomNameVisible(this.isCustomNameVisible());
+            }
+
+            zombifiedpiglin.setPersistenceRequired();
+            net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglin);
+            // CraftBukkit start
+            if (CraftEventFactory.callPigZapEvent(this, p_29474_, zombifiedpiglin).isCancelled()) {
+               return;
+            }
+            // CraftBukkit - added a reason for spawning this creature
+            zombifiedpiglin.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+            p_29473_.addFreshEntity(zombifiedpiglin);
+            // CraftBukkit end
+            this.discard();
+         } else {
             super.thunderHit(p_29473_, p_29474_);
-        }
-    }
-
-    @Override
-    protected void tickRidden(Player p_278330_, Vec3 p_278267_) {
-        super.tickRidden(p_278330_, p_278267_);
-        this.setRot(p_278330_.getYRot(), p_278330_.getXRot() * 0.5F);
-        this.yRotO = this.yBodyRot = this.yHeadRot = this.getYRot();
-        this.steering.tickBoost();
-    }
-
-    @Override
-    protected Vec3 getRiddenInput(Player p_278309_, Vec3 p_275479_) {
-        return new Vec3(0.0, 0.0, 1.0);
-    }
-
-    @Override
-    protected float getRiddenSpeed(Player p_278258_) {
-        return (float)(this.getAttributeValue(Attributes.MOVEMENT_SPEED) * 0.225 * (double)this.steering.boostFactor());
-    }
-
-    @Override
-    public boolean boost() {
-        return this.steering.boost(this.getRandom());
-    }
-
-    @Nullable
-    public Pig getBreedOffspring(ServerLevel p_149001_, AgeableMob p_149002_) {
-        return EntityType.PIG.create(p_149001_);
-    }
-
-    @Override
-    public boolean isFood(ItemStack p_29508_) {
-        return p_29508_.is(ItemTags.PIG_FOOD);
-    }
-
-    @Override
-    public Vec3 getLeashOffset() {
-        return new Vec3(0.0, (double)(0.6F * this.getEyeHeight()), (double)(this.getBbWidth() * 0.4F));
-    }
+         }
+      } else {
+         super.thunderHit(p_29473_, p_29474_);
+      }
+
+   }
+
+   protected void tickRidden(Player p_278330_, Vec3 p_278267_) {
+      super.tickRidden(p_278330_, p_278267_);
+      this.setRot(p_278330_.getYRot(), p_278330_.getXRot() * 0.5F);
+      this.yRotO = this.yBodyRot = this.yHeadRot = this.getYRot();
+      this.steering.tickBoost();
+   }
+
+   protected Vec3 getRiddenInput(Player p_278309_, Vec3 p_275479_) {
+      return new Vec3(0.0D, 0.0D, 1.0D);
+   }
+
+   protected float getRiddenSpeed(Player p_278258_) {
+      return (float)(this.getAttributeValue(Attributes.MOVEMENT_SPEED) * 0.225D * (double)this.steering.boostFactor());
+   }
+
+   public boolean boost() {
+      return this.steering.boost(this.getRandom());
+   }
+
+   @Nullable
+   public Pig getBreedOffspring(ServerLevel p_149001_, AgeableMob p_149002_) {
+      return EntityType.PIG.create(p_149001_);
+   }
+
+   public boolean isFood(ItemStack p_29508_) {
+      return FOOD_ITEMS.test(p_29508_);
+   }
+
+   public Vec3 getLeashOffset() {
+      return new Vec3(0.0D, (double)(0.6F * this.getEyeHeight()), (double)(this.getBbWidth() * 0.4F));
+   }
+
+   protected Vector3f getPassengerAttachmentPoint(Entity p_300896_, EntityDimensions p_301237_, float p_298918_) {
+      return new Vector3f(0.0F, p_301237_.height - 0.03125F * p_298918_, 0.0F);
+   }
 }
