--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -92,7 +_,7 @@
       this.m_29725_(0.0D);
    }
 
-   protected void m_8099_() {
+   public void m_8099_() {
       this.f_21345_.m_25352_(1, new FloatGoal(this));
       this.f_21345_.m_25352_(1, new ClimbOnTopOfPowderSnowGoal(this, this.f_19853_));
       this.f_21345_.m_25352_(1, new Rabbit.RabbitPanicGoal(this, 2.2D));
@@ -511,7 +_,7 @@
 
       public boolean m_8036_() {
          if (this.f_25600_ <= 0) {
-            if (!this.f_29778_.f_19853_.m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_29778_.f_19853_, this.f_29778_)) {
                return false;
             }
 
