--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -284,15 +_,19 @@
          return false;
       } else {
          Entity entity = p_30386_.getEntity();
-         if (!this.level().isClientSide) {
-            this.setOrderedToSit(false);
-         }
+         // CraftBukkit - move diff down
 
          if (entity != null && !(entity instanceof Player) && !(entity instanceof AbstractArrow)) {
             p_30387_ = (p_30387_ + 1.0F) / 2.0F;
          }
 
-         return super.hurt(p_30386_, p_30387_);
+         // CraftBukkit start
+         boolean result = super.hurt(p_30386_, p_30387_);
+         if (!this.level().isClientSide && result) {
+            this.setOrderedToSit(false);
+         }
+         return result;
+         // CraftBukkit end
       }
    }
 
@@ -309,7 +_,7 @@
       super.setTame(p_30443_);
       if (p_30443_) {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(20.0D);
-         this.setHealth(20.0F);
+         this.setHealth(this.getMaxHealth()); // CraftBukkit - 20.0 -> getMaxHealth()
       } else {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(8.0D);
       }
@@ -325,11 +_,12 @@
          return flag ? InteractionResult.CONSUME : InteractionResult.PASS;
       } else if (this.isTame()) {
          if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
+            this.heal((float)itemstack.getFoodProperties(this).getNutrition());
             if (!p_30412_.getAbilities().instabuild) {
                itemstack.shrink(1);
             }
 
-            this.heal((float)item.getFoodProperties().getNutrition());
+            this.gameEvent(GameEvent.EAT, this);
             return InteractionResult.SUCCESS;
          } else {
             if (item instanceof DyeItem) {
@@ -365,7 +_,7 @@
             itemstack.shrink(1);
          }
 
-         if (this.random.nextInt(3) == 0) {
+         if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_30412_)) {
             this.tame(p_30412_);
             this.navigation.stop();
             this.setTarget((LivingEntity)null);
@@ -404,7 +_,7 @@
 
    public boolean isFood(ItemStack p_30440_) {
       Item item = p_30440_.getItem();
-      return item.isEdible() && item.getFoodProperties().isMeat();
+      return item.isEdible() && p_30440_.getFoodProperties(this).isMeat();
    }
 
    public int getMaxSpawnClusterSize() {
