--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -498,7 +_,7 @@
     }
 
     private void tryToTame(Player p_336244_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_336244_)) {
             this.tame(p_336244_);
             this.navigation.stop();
             this.setTarget(null);
@@ -578,7 +_,7 @@
         return !this.getBodyArmorItem().isEmpty();
     }
 
-    private void setCollarColor(DyeColor p_30398_) {
+    public void setCollarColor(DyeColor p_30398_) {
         this.entityData.set(DATA_COLLAR_COLOR, p_30398_.getId());
     }
 
