--- ../src-base/minecraft/org/bukkit/event/player/PlayerBedLeaveEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerBedLeaveEvent.java
@@ -16,6 +16,10 @@
         this.bed = bed;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Returns the bed block involved in this event.
      *
@@ -29,8 +33,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
