--- ../src-base/minecraft/org/bukkit/event/player/PlayerDropItemEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerDropItemEvent.java
@@ -18,6 +18,10 @@
         this.drop = drop;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the ItemDrop created by the player
      *
@@ -39,8 +43,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
