--- ../src-base/minecraft/org/bukkit/event/player/PlayerChatTabCompleteEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerChatTabCompleteEvent.java
@@ -1,7 +1,6 @@
 package org.bukkit.event.player;
 
 import java.util.Collection;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.entity.Player;
 import org.bukkit.event.HandlerList;
@@ -29,6 +28,10 @@
         this.completions = completions;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the chat message being tab-completed.
      *
@@ -63,8 +66,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
