--- ../src-base/minecraft/org/bukkit/event/player/PlayerVelocityEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerVelocityEvent.java
@@ -18,6 +18,10 @@
         this.velocity = velocity;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public boolean isCancelled() {
         return cancel;
     }
@@ -48,8 +52,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
