--- ../src-base/minecraft/org/bukkit/event/player/PlayerChangedMainHandEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerChangedMainHandEvent.java
@@ -18,6 +18,10 @@
         this.mainHand = mainHand;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the new main hand of the player. The old hand is still momentarily
      * available via {@link Player#getMainHand()}.
@@ -32,8 +36,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
