--- ../src-base/minecraft/org/bukkit/event/player/PlayerLevelChangeEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerLevelChangeEvent.java
@@ -17,6 +17,10 @@
         this.newLevel = newLevel;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the old level of the player
      *
@@ -39,8 +43,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
