--- ../src-base/minecraft/org/bukkit/event/player/PlayerChatEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerChatEvent.java
@@ -2,7 +2,6 @@
 
 import java.util.HashSet;
 import java.util.Set;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Warning;
 import org.bukkit.entity.Player;
@@ -19,14 +18,13 @@
  *     causes delays for chat. {@link AsyncPlayerChatEvent} is the encouraged
  *     alternative for thread safe implementations.
  */
-@Deprecated
-@Warning(reason="Listening to this event forces chat to wait for the main thread, delaying chat messages.")
+@Warning(reason = "Listening to this event forces chat to wait for the main thread, delaying chat messages.")
 public class PlayerChatEvent extends PlayerEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
+    private final Set<Player> recipients;
     private boolean cancel = false;
     private String message;
     private String format;
-    private final Set<Player> recipients;
 
     public PlayerChatEvent(final Player player, final String message) {
         super(player);
@@ -42,6 +40,10 @@
         this.recipients = recipients;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public boolean isCancelled() {
         return cancel;
     }
@@ -118,8 +120,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
