--- ../src-base/minecraft/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
@@ -26,6 +26,10 @@
         this.position = position;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public Vector getClickedPosition() {
         return position.clone();
     }
@@ -34,8 +38,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
