--- ../src-base/minecraft/org/bukkit/event/player/PlayerMoveEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerMoveEvent.java
@@ -21,6 +21,10 @@
         this.to = to;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the cancellation state of this event. A cancelled event will not
      * be executed in the server, but will still pass to other plugins
@@ -96,8 +100,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
