--- ../src-base/minecraft/org/bukkit/event/player/PlayerRespawnEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerRespawnEvent.java
@@ -10,8 +10,8 @@
  */
 public class PlayerRespawnEvent extends PlayerEvent {
     private static final HandlerList handlers = new HandlerList();
-    private Location respawnLocation;
     private final boolean isBedSpawn;
+    private Location respawnLocation;
 
     public PlayerRespawnEvent(final Player respawnPlayer, final Location respawnLocation, final boolean isBedSpawn) {
         super(respawnPlayer);
@@ -19,6 +19,10 @@
         this.isBedSpawn = isBedSpawn;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the current respawn location
      *
@@ -53,8 +57,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
