--- ../src-base/minecraft/org/bukkit/event/player/PlayerPortalEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerPortalEvent.java
@@ -26,6 +26,10 @@
         this.travelAgent = pta;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Sets whether or not the Travel Agent will be used.
      * <p>
@@ -80,8 +84,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
