--- ../src-base/minecraft/org/bukkit/event/player/PlayerTeleportEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerTeleportEvent.java
@@ -21,6 +21,10 @@
         this.cause = cause;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the cause of this teleportation event
      *
@@ -30,6 +34,11 @@
         return cause;
     }
 
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
     public enum TeleportCause {
         /**
          * Indicates the teleporation was caused by a player throwing an Ender
@@ -71,18 +80,24 @@
          */
         CHORUS_FRUIT,
         /**
+         * Indicates the teleportation was caused by a player entering a
+         * Mod portal
+         */
+        MOD,
+        DEATH,
+        /**
          * Indicates the teleportation was caused by an event not covered by
          * this enum
          */
         UNKNOWN;
-    }
 
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
+        public boolean isPortal() {
+            switch(this) {
+                case NETHER_PORTAL:
+                case END_PORTAL:
+                    return true;
+            }
+            return false;
+        }
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
