--- ../src-base/minecraft/org/bukkit/event/player/PlayerGameModeChangeEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerGameModeChangeEvent.java
@@ -10,14 +10,18 @@
  */
 public class PlayerGameModeChangeEvent extends PlayerEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
-    private boolean cancelled;
     private final GameMode newGameMode;
+    private boolean cancelled;
 
     public PlayerGameModeChangeEvent(final Player player, final GameMode newGameMode) {
         super(player);
         this.newGameMode = newGameMode;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public boolean isCancelled() {
         return cancelled;
     }
@@ -29,7 +33,7 @@
     /**
      * Gets the GameMode the player is switched to.
      *
-     * @return  player's new GameMode
+     * @return player's new GameMode
      */
     public GameMode getNewGameMode() {
         return newGameMode;
@@ -39,8 +43,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
