--- ../src-base/minecraft/org/bukkit/event/player/PlayerEditBookEvent.java
+++ ../src-work/minecraft/org/bukkit/event/player/PlayerEditBookEvent.java
@@ -36,6 +36,10 @@
         this.cancel = false;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the book meta currently on the book.
      * <p>
@@ -62,18 +66,6 @@
     }
 
     /**
-     * Gets the inventory slot number for the book item that triggered this
-     * event.
-     * <p>
-     * This is a slot number on the player's hotbar in the range 0-8.
-     *
-     * @return the inventory slot number that the book item occupies
-     */
-    public int getSlot() {
-        return slot;
-    }
-
-    /**
      * Sets the book meta that will actually be added to the book.
      *
      * @param newBookMeta new book meta
@@ -86,6 +78,18 @@
     }
 
     /**
+     * Gets the inventory slot number for the book item that triggered this
+     * event.
+     * <p>
+     * This is a slot number on the player's hotbar in the range 0-8.
+     *
+     * @return the inventory slot number that the book item occupies
+     */
+    public int getSlot() {
+        return slot;
+    }
+
+    /**
      * Gets whether or not the book is being signed. If a book is signed the
      * Material changes from BOOK_AND_QUILL to WRITTEN_BOOK.
      *
@@ -110,10 +114,6 @@
         return handlers;
     }
 
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
-
     public boolean isCancelled() {
         return cancel;
     }
