--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -3,7 +_,7 @@
 import net.minecraft.network.chat.Component;
 
 public interface CommandSource {
-    CommandSource NULL = new CommandSource() {
+   CommandSource NULL = new CommandSource() {
         @Override
         public void sendSystemMessage(Component p_230799_) {
         }
@@ -22,7 +_,14 @@
         public boolean shouldInformAdmins() {
             return false;
         }
-    };
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return wrapper.getServer().console;
+      }
+      // CraftBukkit end
+   };
 
     void sendSystemMessage(Component p_230797_);
 
@@ -35,4 +_,8 @@
     default boolean alwaysAccepts() {
         return false;
     }
+
+   default org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return NULL.getBukkitSender(wrapper);
+   }
 }
