--- a/net/minecraft/world/level/levelgen/structure/StructurePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/StructurePiece.java
@@ -48,7 +_,7 @@
     private Rotation rotation;
     protected int genDepth;
     private final StructurePieceType type;
-    private static final Set<Block> SHAPE_CHECK_BLOCKS = ImmutableSet.<Block>builder()
+    public static final Set<Block> SHAPE_CHECK_BLOCKS = ImmutableSet.<Block>builder()
         .add(Blocks.NETHER_BRICK_FENCE)
         .add(Blocks.TORCH)
         .add(Blocks.WALL_TORCH)
@@ -91,6 +_,9 @@
     }
 
     public final CompoundTag createTag(StructurePieceSerializationContext p_192645_) {
+        if (BuiltInRegistries.STRUCTURE_PIECE.getKey(this.getType()) == null) { // FORGE: Friendlier error then the Null String error below.
+            throw new RuntimeException("StructurePiece \"" + this.getClass().getName() + "\": \"" + this.getType() + "\" unregistered, serializing impossible.");
+        }
         CompoundTag compoundtag = new CompoundTag();
         compoundtag.putString("id", BuiltInRegistries.STRUCTURE_PIECE.getKey(this.getType()).toString());
         BoundingBox.CODEC
