--- ../src-base/minecraft/org/bukkit/util/ChatPaginator.java
+++ ../src-work/minecraft/org/bukkit/util/ChatPaginator.java
@@ -1,10 +1,9 @@
 package org.bukkit.util;
 
-import org.bukkit.ChatColor;
-
 import java.util.Arrays;
 import java.util.LinkedList;
 import java.util.List;
+import org.bukkit.ChatColor;
 
 /**
  * The ChatPaginator takes a raw string of arbitrary length and breaks it down
@@ -63,12 +62,12 @@
     public static String[] wordWrap(String rawString, int lineLength) {
         // A null string is a single line
         if (rawString == null) {
-            return new String[] {""};
+            return new String[]{""};
         }
 
         // A string shorter than the lineWidth is a single line
         if (rawString.length() <= lineLength && !rawString.contains("\n")) {
-            return new String[] {rawString};
+            return new String[]{rawString};
         }
 
         char[] rawChars = (rawString + ' ').toCharArray(); // add a trailing space to trigger pagination
