--- ../src-base/minecraft/org/bukkit/util/NumberConversions.java
+++ ../src-work/minecraft/org/bukkit/util/NumberConversions.java
@@ -4,7 +4,8 @@
  * Utils for casting number types to other number types
  */
 public final class NumberConversions {
-    private NumberConversions() {}
+    private NumberConversions() {
+    }
 
     public static int floor(double num) {
         final int floor = (int) num;
@@ -31,8 +32,7 @@
 
         try {
             return Integer.parseInt(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
@@ -44,8 +44,7 @@
 
         try {
             return Float.parseFloat(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
@@ -57,8 +56,7 @@
 
         try {
             return Double.parseDouble(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
@@ -70,8 +68,7 @@
 
         try {
             return Long.parseLong(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
@@ -83,8 +80,7 @@
 
         try {
             return Short.parseShort(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
@@ -96,8 +92,7 @@
 
         try {
             return Byte.parseByte(object.toString());
-        } catch (NumberFormatException e) {
-        } catch (NullPointerException e) {
+        } catch (NumberFormatException | NullPointerException e) {
         }
         return 0;
     }
