--- a/net/minecraft/stats/RecipeBookSettings.java
+++ b/net/minecraft/stats/RecipeBookSettings.java
@@ -148,4 +_,8 @@
          return "[open=" + this.open + ", filtering=" + this.filtering + "]";
       }
    }
+   //FORGE -- called automatically on Enum creation - used for serialization
+   public static void addTagsForType(RecipeBookType type, String openTag, String filteringTag) {
+      TAG_FIELDS.put(type, Pair.of(openTag, filteringTag));
+   }
 }
