--- a/net/minecraft/stats/StatsCounter.java
+++ b/net/minecraft/stats/StatsCounter.java
@@ -14,6 +_,12 @@
 
    public void increment(Player p_13024_, Stat<?> p_13025_, int p_13026_) {
       int i = (int)Math.min((long)this.getValue(p_13025_) + (long)p_13026_, 2147483647L);
+      // CraftBukkit start - fire Statistic events
+      org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.handleStatisticsIncrease(p_13024_, p_13025_, this.getValue(p_13025_), i);
+      if (cancellable != null && cancellable.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       this.setValue(p_13024_, p_13025_, i);
    }
 
