--- ../src-base/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
@@ -155,12 +155,10 @@
 
                         if (l > 4096 * k)
                         {
-                            net.minecraftforge.fml.common.FMLLog.log.warn("Invalid chunk: ({}, {}) Offset: {} Invalid Size: {}>{} {}", x, z, j, l, k * 4096, fileName);
                             return null;
                         }
                         else if (l <= 0)
                         {
-                            net.minecraftforge.fml.common.FMLLog.log.warn("Invalid chunk: ({}, {}) Offset: {} Invalid Size: {} {}", x, z, j, l, fileName);
                             return null;
                         }
                         else
@@ -216,8 +214,7 @@
 
             if (l >= 256)
             {
-                if (!FORGE_ENABLE_EXTENDED_SAVE) return;
-                net.minecraftforge.fml.common.FMLLog.log.warn("Large Chunk Detected: ({}, {}) Size: {} {}", x, z, l, fileName);
+               if (!FORGE_ENABLE_EXTENDED_SAVE) return;
             }
 
             if (j != 0 && k == l)
