--- a/net/minecraft/world/entity/monster/piglin/RememberIfHoglinWasKilled.java
+++ b/net/minecraft/world/entity/monster/piglin/RememberIfHoglinWasKilled.java
@@ -10,10 +_,10 @@
    public static BehaviorControl<LivingEntity> m_257745_() {
       return BehaviorBuilder.m_258034_((p_260168_) -> {
          return p_260168_.group(p_260168_.m_257495_(MemoryModuleType.f_26372_), p_260168_.m_257492_(MemoryModuleType.f_26340_)).apply(p_260168_, (p_259214_, p_260031_) -> {
-            return (p_289478_, p_289479_, p_289480_) -> {
+            return (p_275129_, p_275130_, p_275131_) -> {
                LivingEntity livingentity = p_260168_.m_258051_(p_259214_);
                if (livingentity.m_6095_() == EntityType.f_20456_ && livingentity.m_21224_()) {
-                  p_260031_.m_257465_(true, (long)PiglinAi.f_34795_.m_214085_(p_289479_.m_9236_().f_46441_));
+                  p_260031_.m_257465_(true, (long)PiglinAi.f_34795_.m_214085_(p_275130_.f_19853_.f_46441_));
                }
 
                return true;
