--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -12,6 +_,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.PathType;
 import net.minecraft.world.level.pathfinder.WalkNodeEvaluator;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTeleportEvent;
 
 public class FollowOwnerGoal extends Goal {
     public static final int TELEPORT_WHEN_DISTANCE_IS = 12;
@@ -120,7 +_,15 @@
         } else if (!this.canTeleportTo(new BlockPos(p_25304_, p_25305_, p_25306_))) {
             return false;
         } else {
-            this.tamable.moveTo((double)p_25304_ + 0.5, (double)p_25305_, (double)p_25306_ + 0.5, this.tamable.getYRot(), this.tamable.getXRot());
+         	// this.tamable.moveTo((double)pX + 0.5D, (double)pY, (double)pZ + 0.5D, this.tamable.getYRot(), this.tamable.getXRot());
+         	// CraftBukkit start
+         	EntityTeleportEvent event = CraftEventFactory.callEntityTeleportEvent(this.tamable, (double) p_25304_ + 0.5D, (double) p_25305_, (double) p_25306_ + 0.5D);
+         	if (event.isCancelled()) {
+            	return false;
+         	}
+         	Location to = event.getTo();
+         	this.tamable.moveTo(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         	// CraftBukkit end
             this.navigation.stop();
             return true;
         }
