--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -20,6 +_,8 @@
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_20_R2.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory;
 
 public class RemoveBlockGoal extends MoveToBlockGoal {
    private final Block blockToRemove;
@@ -34,7 +_,7 @@
    }
 
    public boolean canUse() {
-      if (!this.removerMob.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.removerMob.level(), this.removerMob)) {
          return false;
       } else if (this.nextStartTick > 0) {
          --this.nextStartTick;
@@ -89,6 +_,11 @@
          }
 
          if (this.ticksSinceReachedGoal > 60) {
+            // CraftBukkit start - Step on eggs
+            if (!CraftEventFactory.callEntityInteractEvent(this.removerMob, CraftBlock.at(level, blockpos1))) {
+               return;
+            }
+            // CraftBukkit end
             level.removeBlock(blockpos1, false);
             if (!level.isClientSide) {
                for(int i = 0; i < 20; ++i) {
@@ -129,6 +_,7 @@
       if (chunkaccess == null) {
          return false;
       } else {
+         if (!chunkaccess.getBlockState(p_25851_).canEntityDestroy(p_25850_, p_25851_, this.removerMob)) return false;
          return chunkaccess.getBlockState(p_25851_).is(this.blockToRemove) && chunkaccess.getBlockState(p_25851_.above()).isAir() && chunkaccess.getBlockState(p_25851_.above(2)).isAir();
       }
    }
