--- a/net/minecraft/world/BossEvent.java
+++ b/net/minecraft/world/BossEvent.java
@@ -6,10 +_,10 @@
 
 public abstract class BossEvent {
    private final UUID f_18847_;
-   protected Component f_18840_;
+   public Component f_18840_;
    protected float f_146638_;
-   protected BossEvent.BossBarColor f_18842_;
-   protected BossEvent.BossBarOverlay f_18843_;
+   public BossEvent.BossBarColor f_18842_;
+   public BossEvent.BossBarOverlay f_18843_;
    protected boolean f_18844_;
    protected boolean f_18845_;
    protected boolean f_18846_;
