--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -1,7 +_,10 @@
 package net.minecraft.world;
 
+import java.util.Collections;
 import java.util.Set;
 import java.util.function.Predicate;
+
+import com.mohistmc.bukkit.inventory.InventoryOwner;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
@@ -9,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 
+import net.minecraft.world.item.crafting.Recipe;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftHumanEntity;
+
 public interface Container extends Clearable {
    int f_146642_ = 64;
    int f_271421_ = 8;
@@ -92,4 +_,47 @@
          return p_272670_.m_20275_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.5D, (double)blockpos.m_123343_() + 0.5D) <= (double)(p_273411_ * p_273411_);
       }
    }
+
+   // CraftBukkit start
+   default java.util.List<ItemStack> getContents() {
+      return Collections.emptyList();
+   }
+
+   default void onOpen(CraftHumanEntity who) {
+   }
+
+   default void onClose(CraftHumanEntity who) {
+   }
+
+   // Mohist start - fix exoticbirds mod
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+      return Collections.emptyList();
+   }
+   // Mohist end
+
+   default org.bukkit.inventory.InventoryHolder getOwner() {
+      return this instanceof BlockEntity blockEntity ? InventoryOwner.get(blockEntity) : null;
+   }
+
+   default void setMaxStackSize(int size) {
+   }
+
+   default org.bukkit.Location getLocation(){
+      if (this instanceof BlockEntity entity) {
+         BlockPos blockPos = entity.m_58899_();
+         return new Location(entity.m_58904_().getWorld(), blockPos.m_123341_(), blockPos.m_123342_(), blockPos.m_123343_());
+      } else {
+         return null;
+      }
+   }
+
+   default Recipe getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(Recipe recipe) {
+   }
+
+   int MAX_STACK = 64;
+   // CraftBukkit end
 }
