--- a/net/minecraft/world/Region.java
+++ b/net/minecraft/world/Region.java
@@ -53,6 +_,13 @@
 
    }
 
+   // Mohist start - Add isPosLoaded method to IBlockReader
+   @Override
+   public boolean isPosLoaded(BlockPos pos) {
+      return this.field_72815_e.isPosLoaded(pos);
+   }
+   // Mohist end
+
    private IChunk func_226703_d_(BlockPos p_226703_1_) {
       return this.func_226702_a_(p_226703_1_.func_177958_n() >> 4, p_226703_1_.func_177952_p() >> 4);
    }
