--- ../src-base/minecraft/cpw/mods/fml/common/network/handshake/FMLHandshakeServerState.java
+++ ../src-work/minecraft/cpw/mods/fml/common/network/handshake/FMLHandshakeServerState.java
@@ -9,6 +9,7 @@
 import cpw.mods.fml.common.network.internal.FMLNetworkHandler;
 import cpw.mods.fml.common.registry.GameData;
 import cpw.mods.fml.relauncher.Side;
+import net.minecraft.server.MinecraftServer;
 
 enum FMLHandshakeServerState implements IHandshakeState<FMLHandshakeServerState>
 {
@@ -37,8 +38,13 @@
             }
 
             FMLHandshakeMessage.ModList client = (FMLHandshakeMessage.ModList)msg;
-            FMLLog.info("Client attempting to join with %d mods : %s", client.modListSize(), client.modListAsString());
+            if (net.minecraft.server.MinecraftServer.uraniumConfig.loggingClientModList.getValue()) {
+                FMLLog.info("Client attempting to join with %d mods : %s", client.modListSize(), client.modListAsString());
+            } else {
+                FMLLog.info("Client attempting to join with %d mods", client.modListSize());
+            }
             String result = FMLNetworkHandler.checkModList(client, Side.CLIENT);
+
             if (result != null)
             {
                 NetworkDispatcher dispatcher = ctx.channel().attr(NetworkDispatcher.FML_DISPATCHER).get();
