--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -53,6 +_,8 @@
 import net.minecraft.world.level.pathfinder.BlockPathTypes;
 import net.minecraft.world.phys.Vec3;
 import org.joml.Vector3f;
+import org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack;
 
 public class Goat extends Animal {
    public static final EntityDimensions LONG_JUMPING_DIMENSIONS = EntityDimensions.scalable(0.9F, 1.3F).scale(0.7F);
@@ -64,8 +_,8 @@
    public static final double GOAT_SCREAMING_CHANCE = 0.02D;
    public static final double UNIHORN_CHANCE = (double)0.1F;
    private static final EntityDataAccessor<Boolean> DATA_IS_SCREAMING_GOAT = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
-   private static final EntityDataAccessor<Boolean> DATA_HAS_LEFT_HORN = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
-   private static final EntityDataAccessor<Boolean> DATA_HAS_RIGHT_HORN = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
+   public static final EntityDataAccessor<Boolean> DATA_HAS_LEFT_HORN = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
+   public static final EntityDataAccessor<Boolean> DATA_HAS_RIGHT_HORN = SynchedEntityData.defineId(Goat.class, EntityDataSerializers.BOOLEAN);
    private boolean isLoweringHead;
    private int lowerHeadTick;
 
@@ -194,8 +_,15 @@
    public InteractionResult mobInteract(Player p_149379_, InteractionHand p_149380_) {
       ItemStack itemstack = p_149379_.getItemInHand(p_149380_);
       if (itemstack.is(Items.BUCKET) && !this.isBaby()) {
+         // CraftBukkit start - Got milk?
+         org.bukkit.event.player.PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent((ServerLevel) p_149379_.level, p_149379_, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET, p_149380_);
+
+         if (event.isCancelled()) {
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
          p_149379_.playSound(this.getMilkingSound(), 1.0F, 1.0F);
-         ItemStack itemstack1 = ItemUtils.createFilledResult(itemstack, p_149379_, Items.MILK_BUCKET.getDefaultInstance());
+         ItemStack itemstack1 = ItemUtils.createFilledResult(itemstack, p_149379_, CraftItemStack.asNMSCopy(event.getItemStack())); // CraftBukkit
          p_149379_.setItemInHand(p_149380_, itemstack1);
          return InteractionResult.sidedSuccess(this.level().isClientSide);
       } else {
