From dcacc8aab3402f8c997e2be3c79742641214c305 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sat, 4 Aug 2018 02:52:29 -0700
Subject: [PATCH] Call BlockFormEvent for obsidian and cobblestone formation
 due to lava / water interacting.


diff --git a/src/main/java/net/minecraft/server/BlockFluids.java b/src/main/java/net/minecraft/server/BlockFluids.java
index db73f5d9..4d4eba14 100644
--- a/src/main/java/net/minecraft/server/BlockFluids.java
+++ b/src/main/java/net/minecraft/server/BlockFluids.java
@@ -148,17 +148,36 @@ public abstract class BlockFluids extends Block {
             if (flag) {
                 Integer integer = (Integer) iblockdata.get(BlockFluids.LEVEL);
 
-                if (integer.intValue() == 0) {
-                    world.setTypeUpdate(blockposition, Blocks.OBSIDIAN.getBlockData());
-                    this.fizz(world, blockposition);
-                    return true;
+                // CraftBukkit start
+                Block formed = null;
+                if (integer == 0) {
+                    formed = Blocks.OBSIDIAN;
+                    } else if (integer <= 4) {
+                    formed = Blocks.COBBLESTONE;
                 }
 
-                if (integer.intValue() > 0) { // PaperSpigot
-                    world.setTypeUpdate(blockposition, Blocks.COBBLESTONE.getBlockData());
-                    this.fizz(world, blockposition);
-                    return true;
+                if (formed != null) {
+                    org.bukkit.block.Block block = world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+                    org.bukkit.block.BlockState newState = block.getState();
+                    newState.setType(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(formed));
+                    org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(block, newState);
+                    world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        if (integer.intValue() == 0) {
+                            world.setTypeUpdate(blockposition, Blocks.OBSIDIAN.getBlockData());
+                            this.fizz(world, blockposition);
+                            return true;
+                        }
+
+                        if (integer.intValue() > 0) {
+                            world.setTypeUpdate(blockposition, Blocks.COBBLESTONE.getBlockData());
+                            this.fizz(world, blockposition);
+                            return true;
+                        }
+                    }
                 }
+                // CraftBukkit end
             }
         }
 
-- 
2.18.0

