From 560a26cfce9a6075887f8bfdcddaca6cdaad40bc Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sat, 4 Aug 2018 17:18:52 -0700
Subject: [PATCH] Enable multiworld maps


diff --git a/src/main/java/net/minecraft/server/ItemMapEmpty.java b/src/main/java/net/minecraft/server/ItemMapEmpty.java
index 6bf90c47..85a04c91 100644
--- a/src/main/java/net/minecraft/server/ItemMapEmpty.java
+++ b/src/main/java/net/minecraft/server/ItemMapEmpty.java
@@ -7,12 +7,11 @@ public class ItemMapEmpty extends ItemWorldMapBase {
     }
 
     public ItemStack a(ItemStack itemstack, World world, EntityHuman entityhuman) {
-        World worldMain = world.getServer().getServer().worlds.get(0); // CraftBukkit - store reference to primary world
-        ItemStack itemstack1 = new ItemStack(Items.FILLED_MAP, 1, worldMain.b("map")); // CraftBukkit - use primary world for maps
+        ItemStack itemstack1 = new ItemStack(Items.FILLED_MAP, 1, world.b("map"));
         String s = "map_" + itemstack1.getData();
         WorldMap worldmap = new WorldMap(s);
 
-        worldMain.a(s, (PersistentBase) worldmap); // CraftBukkit
+        world.a(s, (PersistentBase) worldmap); // CraftBukkit
         worldmap.scale = 0;
         worldmap.a(entityhuman.locX, entityhuman.locZ, worldmap.scale);
         worldmap.map = (byte) ((WorldServer) world).dimension; // CraftBukkit - use bukkit dimension
diff --git a/src/main/java/net/minecraft/server/ItemWorldMap.java b/src/main/java/net/minecraft/server/ItemWorldMap.java
index 6473ee5b..95987cd4 100644
--- a/src/main/java/net/minecraft/server/ItemWorldMap.java
+++ b/src/main/java/net/minecraft/server/ItemWorldMap.java
@@ -16,19 +16,18 @@ public class ItemWorldMap extends ItemWorldMapBase {
     }
 
     public WorldMap getSavedMap(ItemStack itemstack, World world) {
-        World worldMain = world.getServer().getServer().worlds.get(0); // CraftBukkit - store reference to primary world
         String s = "map_" + itemstack.getData();
-        WorldMap worldmap = (WorldMap) worldMain.a(WorldMap.class, s); // CraftBukkit - use primary world for maps
+        WorldMap worldmap = (WorldMap) world.a(WorldMap.class, s);
 
         if (worldmap == null && !world.isClientSide) {
-            itemstack.setData(worldMain.b("map")); // CraftBukkit - use primary world for maps
+            itemstack.setData(world.b("map"));
             s = "map_" + itemstack.getData();
             worldmap = new WorldMap(s);
             worldmap.scale = 3;
             worldmap.a((double) world.getWorldData().c(), (double) world.getWorldData().e(), worldmap.scale);
             worldmap.map = (byte) ((WorldServer) world).dimension; // CraftBukkit - fixes Bukkit multiworld maps
             worldmap.c();
-            worldMain.a(s, (PersistentBase) worldmap); // CraftBukkit - use primary world for maps
+            world.a(s, (PersistentBase) worldmap);
 
             // CraftBukkit start
             MapInitializeEvent event = new MapInitializeEvent(worldmap.mapView);
@@ -193,8 +192,6 @@ public class ItemWorldMap extends ItemWorldMapBase {
         if (itemstack.hasTag() && itemstack.getTag().getBoolean("map_is_scaling")) {
             WorldMap worldmap = Items.FILLED_MAP.getSavedMap(itemstack, world);
 
-            world = world.getServer().getServer().worlds.get(0); // CraftBukkit - use primary world for maps
-
             itemstack.setData(world.b("map"));
             WorldMap worldmap1 = new WorldMap("map_" + itemstack.getData());
 
-- 
2.18.0

