From 5e8a3c36e3796ae4029003f8a64540e33596a495 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sat, 4 Aug 2018 19:35:51 -0700
Subject: [PATCH] Add PlayerPickupExperienceEvent


diff --git a/src/main/java/net/minecraft/server/EntityExperienceOrb.java b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
index aefa9fa87..dae728188 100644
--- a/src/main/java/net/minecraft/server/EntityExperienceOrb.java
+++ b/src/main/java/net/minecraft/server/EntityExperienceOrb.java
@@ -158,10 +158,15 @@ public class EntityExperienceOrb extends Entity {
         if (!this.world.isClientSide) {
             if (this.c == 0 && entityhuman.bp == 0) {
                 entityhuman.bp = 2;
-                this.world.makeSound(entityhuman, "random.orb", 0.1F, 0.5F * ((this.random.nextFloat() - this.random.nextFloat()) * 0.7F + 1.8F));
-                entityhuman.receive(this, 1);
-                entityhuman.giveExp(CraftEventFactory.callPlayerExpChangeEvent(entityhuman, this.value).getAmount()); // CraftBukkit - this.value -> event.getAmount()
-                this.die();
+                // SportBukkit start
+                org.bukkit.event.player.PlayerPickupExperienceEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerPickupExperienceEvent((EntityPlayer) entityhuman, this);
+                if(!event.isCancelled()) {
+                    // SportBukkit end
+                    this.world.makeSound(entityhuman, "random.orb", 0.1F, 0.5F * ((this.random.nextFloat() - this.random.nextFloat()) * 0.7F + 1.8F));
+                    entityhuman.receive(this, 1);
+                    entityhuman.giveExp(CraftEventFactory.callPlayerExpChangeEvent(entityhuman, this.value).getAmount()); // CraftBukkit - this.value -> event.getAmount()
+                    this.die();
+                } // SportBukkit
             }
 
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index db915033b..532a80520 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -918,6 +918,13 @@ public class CraftEventFactory {
         return event;
     }
 
+    public static PlayerPickupExperienceEvent callPlayerPickupExperienceEvent(EntityPlayer player, net.minecraft.server.EntityExperienceOrb orb) {
+        World world = player.world;
+        PlayerPickupExperienceEvent event = new PlayerPickupExperienceEvent(player.getBukkitEntity(), new org.bukkit.craftbukkit.entity.CraftExperienceOrb(world.getServer(), orb));
+        world.getServer().getPluginManager().callEvent(event);
+        return event;
+    }
+
     public static Cancellable handleStatisticsIncrease(EntityHuman entityHuman, net.minecraft.server.Statistic statistic, int current, int incrementation) {
         Player player = ((EntityPlayer) entityHuman).getBukkitEntity();
         Event event;
-- 
2.18.0

