From e69c1b27aa8b1c29092a70cfdb5938446e989644 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sun, 5 Aug 2018 02:19:32 -0700
Subject: [PATCH] Add API to disable potion particles


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index c8b4e2928..1c3677a46 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -14,6 +14,7 @@ import java.util.UUID;
 import java.util.ArrayList;
 import com.google.common.base.Function;
 import com.google.common.collect.Lists;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Vehicle;
@@ -537,10 +538,16 @@ public abstract class EntityLiving extends Entity {
             this.bj();
             this.setInvisible(false);
         } else {
-            int i = PotionBrewer.a(this.effects.values());
+            // CraftBukkit start - optionally disable potion particles
+            if(((CraftLivingEntity) getBukkitEntity()).getPotionParticles()) {
+                int i = PotionBrewer.a(this.effects.values());
 
-            this.datawatcher.watch(8, Byte.valueOf((byte) (PotionBrewer.b(this.effects.values()) ? 1 : 0)));
-            this.datawatcher.watch(7, Integer.valueOf(i));
+                this.datawatcher.watch(8, Byte.valueOf((byte) (PotionBrewer.b(this.effects.values()) ? 1 : 0)));
+                this.datawatcher.watch(7, Integer.valueOf(i));
+            } else {
+                this.bj();
+            }
+            // CraftBukkit end
             this.setInvisible(this.hasEffect(MobEffectList.INVISIBILITY.id));
         }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 0b033fd12..5f544d82f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -65,6 +65,7 @@ import org.bukkit.util.Vector;
 
 public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     private CraftEntityEquipment equipment;
+    private boolean potionParticles = true;
 
     public CraftLivingEntity(final CraftServer server, final EntityLiving entity) {
         super(server, entity);
@@ -333,6 +334,21 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return effects;
     }
 
+    @Override
+    public void setPotionParticles(boolean enabled) {
+        this.potionParticles = enabled;
+        if(enabled) {
+            this.getHandle().updateEffects = true;
+        } else {
+            this.getHandle().getDataWatcher().watch(7, (int) 0); // particle color
+            this.getHandle().getDataWatcher().watch(8, (byte) 0); // ambient flag
+        }
+    }
+
+    public boolean getPotionParticles() {
+        return this.potionParticles;
+    }
+
     public <T extends Projectile> T launchProjectile(Class<? extends T> projectile) {
         return launchProjectile(projectile, null);
     }
-- 
2.18.0

