From 3c13b91ed96366df6907e8ffbfd571015ba2de06 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sun, 5 Aug 2018 18:17:01 -0700
Subject: [PATCH] Keep players in the player list while respawning


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 3c2d3197f..f38ab7156 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -559,6 +559,7 @@ public abstract class PlayerList {
         BlockPosition blockposition1;
 
         // CraftBukkit start - fire PlayerRespawnEvent
+        this.players.add(entityplayer1); // Add player back to this list earlier than vanilla does
         if (location == null) {
             boolean isBedSpawn = false;
             CraftWorld cworld = (CraftWorld) this.server.server.getWorld(entityplayer.spawnWorld);
@@ -619,10 +620,12 @@ public abstract class PlayerList {
         entityplayer1.playerConnection.sendPacket(new PacketPlayOutExperience(entityplayer1.exp, entityplayer1.expTotal, entityplayer1.expLevel));
         this.b(entityplayer1, worldserver);
 
-        if (!entityplayer.playerConnection.isDisconnected()) {
+        // Don't re-add player to player list if disconnected
+        if (entityplayer.playerConnection.isDisconnected()) {
+            this.players.remove(entityplayer1);
+        } else {
             worldserver.getPlayerChunkMap().addPlayer(entityplayer1);
             worldserver.addEntity(entityplayer1);
-            this.players.add(entityplayer1);
             this.playersByName.put(entityplayer1.getName(), entityplayer1); // Spigot
             this.j.put(entityplayer1.getUniqueID(), entityplayer1);
         }
-- 
2.18.0

