From f0e0a8fd87c788cd0776224d68b2e05193542ab0 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sun, 5 Aug 2018 22:07:28 -0700
Subject: [PATCH] Fix combustion events


diff --git a/src/main/java/net/minecraft/server/DispenserRegistry.java b/src/main/java/net/minecraft/server/DispenserRegistry.java
index 329071051..9577483db 100644
--- a/src/main/java/net/minecraft/server/DispenserRegistry.java
+++ b/src/main/java/net/minecraft/server/DispenserRegistry.java
@@ -188,6 +188,7 @@ public class DispenserRegistry {
                 CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
 
                 EntitySmallFireball entitysmallfireball = new EntitySmallFireball(world, d0, d1, d2, d3, d4, d5);
+                entitysmallfireball.projectileSource = new org.bukkit.craftbukkit.projectiles.CraftBlockProjectileSource((TileEntityDispenser) isourceblock.getTileEntity());
                 BlockDispenseEntityEvent event = new BlockDispenseEntityEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5), entitysmallfireball.getBukkitEntity());
                 if (!BlockDispenser.eventFired) {
                     world.getServer().getPluginManager().callEvent(event);
@@ -210,7 +211,6 @@ public class DispenserRegistry {
                 }
 
                 entitysmallfireball.setDirection(event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
-                entitysmallfireball.projectileSource = new org.bukkit.craftbukkit.projectiles.CraftBlockProjectileSource((TileEntityDispenser) isourceblock.getTileEntity());
 
                 world.addEntity(entitysmallfireball);
                 // itemstack.a(1); // Handled during event processing
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 9746cafda..1f8dd0e2d 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -18,6 +18,8 @@ import org.bukkit.entity.Painting;
 import org.bukkit.entity.Vehicle;
 import co.aikar.timings.SpigotTimings; // Spigot
 import co.aikar.timings.Timing; // Spigot
+import org.bukkit.event.entity.EntityExtinguishEvent;
+import org.bukkit.event.entity.EntityCombustByBlockEvent;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.entity.EntityDespawnInVoidEvent;
 import org.bukkit.event.hanging.HangingBreakByEntityEvent;
@@ -95,6 +97,7 @@ public abstract class Entity implements ICommandListener {
     public int ticksLived;
     public int maxFireTicks;
     public int fireTicks;
+    public boolean wasOnFire; // CraftBukkit - to detect when the fire goes out
     public boolean inWater; // Spigot - protected -> public // PAIL
     public int noDamageTicks;
     protected boolean justCreated;
@@ -368,21 +371,23 @@ public abstract class Entity implements ICommandListener {
             this.damageEntity(DamageSource.LAVA, 4.0F);
 
             // CraftBukkit start - Fallen in lava TODO: this event spams!
+            Vec3D lavaPos = this.world.getLargestBlockIntersection(this.boundingBox.shrink(0.001D, 0.001D, 0.001D), Material.LAVA);
+            org.bukkit.block.Block lavaBlock = lavaPos == null ? null : this.world.getWorld().getBlockAt((int) lavaPos.a, (int) lavaPos.b, (int) lavaPos.c);
+            try {
+                CraftEventFactory.blockDamage = lavaBlock;
+                this.damageEntity(DamageSource.LAVA, 4);
+            } finally {
+                CraftEventFactory.blockDamage = null;
+            }
             if (this instanceof EntityLiving) {
-                if (fireTicks <= 0) {
-                    // not on fire yet
-                    // TODO: shouldn't be sending null for the block
-                    org.bukkit.block.Block damager = null; // ((WorldServer) this.l).getWorld().getBlockAt(i, j, k);
-                    org.bukkit.entity.Entity damagee = this.getBukkitEntity();
-                    EntityCombustEvent combustEvent = new org.bukkit.event.entity.EntityCombustByBlockEvent(damager, damagee, 15);
-                    this.world.getServer().getPluginManager().callEvent(combustEvent);
-
-                    if (!combustEvent.isCancelled()) {
-                        this.setOnFire(combustEvent.getDuration());
-                    }
-                } else {
-                    // This will be called every single tick the entity is in lava, so don't throw an event
-                    this.setOnFire(15);
+                // Note that in order for cancelling or custom duration to work properly,
+                // this event must be fired every tick, thus we cannot avoid "spamming" it.
+                org.bukkit.entity.Entity damagee = this.getBukkitEntity();
+                EntityCombustEvent combustEvent = new org.bukkit.event.entity.EntityCombustByBlockEvent(lavaBlock, damagee, 15);
+                this.world.getServer().getPluginManager().callEvent(combustEvent);
+
+                if (!combustEvent.isCancelled()) {
+                    this.setOnFire(combustEvent.getDuration());
                 }
                 return;
             }
@@ -744,21 +749,32 @@ public abstract class Entity implements ICommandListener {
 
             boolean flag2 = this.U();
 
-            if (this.world.e(this.getBoundingBox().shrink(0.001D, 0.001D, 0.001D))) {
-                this.burn(1);
+            // CraftBukkit start - get the location of the fire block
+            Vec3D firePos = this.world.getLargestBlockIntersection(this.boundingBox.shrink(0.001D, 0.001D, 0.001D), Material.FIRE);
+            if (firePos != null) {
+                org.bukkit.block.Block fireBlock = this.bukkitEntity.getWorld().getBlockAt((int) firePos.a, (int) firePos.b, (int) firePos.c);
+                try {
+                    CraftEventFactory.blockDamage = fireBlock;
+                    this.burn(1);
+                } finally {
+                    CraftEventFactory.blockDamage = null;
+                }
                 if (!flag2) {
-                    ++this.fireTicks;
-                    // CraftBukkit start - Not on fire yet
-                    if (this.fireTicks <= 0) { // Only throw events on the first combust, otherwise it spams
-                        EntityCombustEvent event = new EntityCombustEvent(getBukkitEntity(), 8);
-                        world.getServer().getPluginManager().callEvent(event);
-
-                        if (!event.isCancelled()) {
-                            setOnFire(event.getDuration());
+                    EntityCombustByBlockEvent event = new EntityCombustByBlockEvent(fireBlock, this.getBukkitEntity(), 8);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        // Note carefully how this works: when fireTicks is negative, the entity is
+                        // "heating up" but not on fire yet. When fireTicks reaches 0, the entity
+                        // "ignites" and fireTicks jumps to 160. It will then stay at that value as
+                        // long as the player remains in fire (because the ++ below will cancel out
+                        // the -- in the entity tick). For the event cancelling to work, it has to
+                        // be fired every tick, thus we cannot avoid "spamming" it.
+                        ++this.fireTicks;
+                        if (this.fireTicks == 0) {
+                            this.setOnFire(event.getDuration());
                         }
-                    } else {
-                        // CraftBukkit end
-                        this.setOnFire(8);
+
                     }
                 }
             } else if (this.fireTicks <= 0) {
@@ -770,6 +786,15 @@ public abstract class Entity implements ICommandListener {
                 this.fireTicks = -this.maxFireTicks;
             }
 
+            // CraftBukkit start
+            if(this.fireTicks > 0) {
+                this.wasOnFire = true;
+            } else if(this.wasOnFire && this.fireTicks <= 0) {
+                this.wasOnFire = false;
+                this.world.getServer().getPluginManager().callEvent(new EntityExtinguishEvent(this.getBukkitEntity()));
+            }
+            // CraftBukkit end
+
             this.world.methodProfiler.b();
         }
     }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 897ef18b0..1c61fd7dd 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1872,6 +1872,35 @@ public abstract class World implements IBlockAccess {
         return false;
     }
 
+    // CraftBukkit start
+    public Vec3D getLargestBlockIntersection(AxisAlignedBB aabb, Material material) {
+        int xMin = MathHelper.floor(aabb.a);
+        int xMax = MathHelper.floor(aabb.d + 1.0D);
+        int yMin = MathHelper.floor(aabb.b);
+        int yMax = MathHelper.floor(aabb.e + 1.0D);
+        int zMin = MathHelper.floor(aabb.c);
+        int zMax = MathHelper.floor(aabb.f + 1.0D);
+        double maxVolume = 0;
+        Vec3D block = null;
+        for (int x = xMin; x < xMax; ++x) {
+            for (int y = yMin; y < yMax; ++y) {
+                for (int z = zMin; z < zMax; ++z) {
+                    IBlockData type = this.getType(new BlockPosition(x, y, z));
+                    if (material == type.getBlock().getMaterial()) {
+                        AxisAlignedBB i = new AxisAlignedBB(x, x+1, y, y+1, z, z+1).a(aabb);
+                        double volume = (i.d - i.a) * (i.e - i.b) * (i.f - i.c);
+                        if(volume > maxVolume) {
+                            maxVolume = volume;
+                            block = new Vec3D(x, y, z);
+                        }
+                    }
+                }
+            }
+        }
+        return block;
+    }
+    // CraftBukkit end
+
     public boolean b(AxisAlignedBB axisalignedbb, Material material) {
         int i = MathHelper.floor(axisalignedbb.a);
         int j = MathHelper.floor(axisalignedbb.d + 1.0D);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 532a80520..52adc874e 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -194,7 +194,7 @@ public class CraftEventFactory {
     public static PlayerInteractEvent callPlayerInteractEvent(EntityHuman who, Action action, BlockPosition position, EnumDirection direction, ItemStack itemstack) {
         return callPlayerInteractEvent(who, action, position, direction, itemstack, false);
     }
-    
+
     public static PlayerInteractEvent callPlayerInteractEvent(EntityHuman who, Action action, BlockPosition position, EnumDirection direction, ItemStack itemstack, boolean cancelledBlock) {
         Player player = (who == null) ? null : (Player) who.getBukkitEntity();
         CraftItemStack itemInHand = CraftItemStack.asCraftMirror(itemstack);
@@ -470,18 +470,16 @@ public class CraftEventFactory {
                 event.getEntity().setLastDamageCause(event);
             }
             return event;
-        } else if (source == DamageSource.LAVA) {
-            EntityDamageEvent event = callEvent(new EntityDamageByBlockEvent(null, entity.getBukkitEntity(), DamageCause.LAVA, modifiers, modifierFunctions));
-            if (!event.isCancelled()) {
-                event.getEntity().setLastDamageCause(event);
-            }
-            return event;
         } else if (blockDamage != null) {
             DamageCause cause = null;
             Block damager = blockDamage;
             blockDamage = null;
             if (source == DamageSource.CACTUS) {
                 cause = DamageCause.CONTACT;
+            } else if (source == DamageSource.FIRE) {
+                cause = DamageCause.FIRE;
+            } else if (source == DamageSource.LAVA) {
+                cause = DamageCause.LAVA;
             } else {
                 throw new RuntimeException(String.format("Unhandled damage of %s by %s from %s", entity, damager, source.translationIndex)); // Spigot
             }
@@ -513,6 +511,8 @@ public class CraftEventFactory {
         DamageCause cause = null;
         if (source == DamageSource.FIRE) {
             cause = DamageCause.FIRE;
+        } else if (source == DamageSource.LAVA) {
+            cause = DamageCause.LAVA;
         } else if (source == DamageSource.STARVE) {
             cause = DamageCause.STARVATION;
         } else if (source == DamageSource.WITHER) {
-- 
2.18.0

