From 1ac85d05984fe239c72c8f0ac50943aa6376f74d Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Mon, 6 Aug 2018 13:29:52 -0700
Subject: [PATCH] Add setting to disable skull fetching


diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 58014c5f1..885050fe7 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -17,6 +17,7 @@ import java.util.concurrent.TimeUnit;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.authlib.Agent;
 import com.mojang.authlib.ProfileLookupCallback;
+import org.spigotmc.SpigotConfig;
 // Spigot end
 
 public class TileEntitySkull extends TileEntity {
@@ -67,7 +68,11 @@ public class TileEntitySkull extends TileEntity {
 
                         if ( property == null )
                         {
-                            profile = MinecraftServer.getServer().aD().fillProfileProperties( profile, true );
+                            if(SpigotConfig.fetchSkulls) {
+                                profile = MinecraftServer.getServer().aD().fillProfileProperties(profile, true);
+                            } else {
+                                org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for \"" + profile.getName() + "\"");
+                            }
                         }
                     }
 
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index b9d185d99..7ab410398 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -226,6 +226,12 @@ public class SpigotConfig
         WatchdogThread.doStart( timeoutTime, restartOnCrash );
     }
 
+    public static boolean fetchSkulls = true;
+    private static void sportPaper()
+    {
+        fetchSkulls = getBoolean("settings.fetch-skulls", fetchSkulls);
+    }
+
     public static boolean bungee;
     private static void bungee() {
         if ( version < 4 )
-- 
2.18.0

