From 24587907bae96daeee7fe07774d052899456c02a Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Tue, 7 Aug 2018 02:31:47 -0700
Subject: [PATCH] Correct falling block vertical offset


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 764007093..ae93d76e7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -890,7 +890,7 @@ public class CraftWorld implements World {
         Validate.isTrue(material.isBlock(), "Material must be a block");
 
         double x = location.getBlockX() + 0.5;
-        double y = location.getBlockY() + 0.5;
+        double y = location.getBlockY();
         double z = location.getBlockZ() + 0.5;
 
         // PaperSpigot start - Add FallingBlock source location API
-- 
2.18.0

