From aa0fa541fbdd9c3d346b26257a9472068ce4128a Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Fri, 24 Aug 2018 02:39:05 -0700
Subject: [PATCH] Empty chunk API


diff --git a/src/main/java/net/minecraft/server/ChunkSection.java b/src/main/java/net/minecraft/server/ChunkSection.java
index 907c57b8..7a251439 100644
--- a/src/main/java/net/minecraft/server/ChunkSection.java
+++ b/src/main/java/net/minecraft/server/ChunkSection.java
@@ -71,6 +71,7 @@ public class ChunkSection {
         return iblockdata.getBlock().toLegacyData(iblockdata);
     }
 
+    public boolean isEmpty() { return this.a(); } // SportPaper - alias for following method
     public boolean a() {
         return this.nonEmptyBlockCount == 0;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 80de4eb1..14b2f040 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -103,6 +103,14 @@ public class CraftChunk implements Chunk {
         return blocks;
     }
 
+    @Override
+    public boolean isEmpty() {
+        for(ChunkSection chunkSections : getHandle().getSections()) {
+            if(chunkSections != null && !chunkSections.isEmpty()) return false;
+        }
+        return true;
+    }
+
     public Entity[] getEntities() {
         int count = 0, index = 0;
         net.minecraft.server.Chunk chunk = getHandle();
-- 
2.19.0

