From 9088d14c325a4eb233b2293efc27dc3476381347 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Mon, 27 Aug 2018 00:03:32 -0700
Subject: [PATCH] Add world broadcast API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 5bdeb949..8a7c3345 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -12,6 +12,7 @@ import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
 
+import net.md_5.bungee.api.chat.BaseComponent;
 import net.minecraft.server.*;
 
 import org.apache.commons.lang.Validate;
@@ -850,6 +851,21 @@ public class CraftWorld implements World {
         return list;
     }
 
+    @Override
+    public void broadcastMessage(String message) {
+        getPlayers().forEach(player -> player.sendMessage(message));
+    }
+
+    @Override
+    public void broadcast(BaseComponent component) {
+        getPlayers().forEach(player -> player.sendMessage(component));
+    }
+
+    @Override
+    public void broadcast(BaseComponent... components) {
+        getPlayers().forEach(player -> player.sendMessage(components));
+    }
+
     public void save() {
     // Spigot start
         save(true);
-- 
2.19.0

