From d7bd96183495c83e5f1177aadb01f16a466803cd Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Wed, 26 Sep 2018 23:04:03 -0500
Subject: [PATCH] Improve locale API and add methods for translating components


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 49a2227..24d1538 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1031,6 +1031,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.locale = packetplayinsettings.a();
         if (!this.locale.equals(oldLocale)) {
             CraftEventFactory.callPlayerLocaleChangeEvent(this, oldLocale, this.locale);
+            this.getBukkitEntity().updateLocale(this.locale);
         }
         // PaperSpigot end
         this.bR = packetplayinsettings.c();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index b85669b..67be13f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -88,6 +88,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private double healthScale = 20;
     private Skin realSkin;
     private boolean showInvisibles;
+    private java.util.Locale cachedLocale = java.util.Locale.US; // SportPaper
 
     private final Map<CommandSender, String> fakeNames = new WeakHashMap<CommandSender, String>();
     private final Map<CommandSender, String> fakeDisplayNames = new WeakHashMap<CommandSender, String>();
@@ -1984,4 +1985,34 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return spigot;
     }
     // Spigot end
+
+    // SportPaper start
+    @Override
+    public java.util.Locale getLocale() {
+        return this.cachedLocale;
+    }
+
+    public void updateLocale(String locale) {
+        this.cachedLocale = parseLocale(locale);
+    }
+
+    private static java.util.Locale parseLocale(String locale) {
+        try {
+            final String[] split = locale.split("[-_]");
+            switch (split.length) {
+                case 1: // language
+                    return new java.util.Locale(split[0]);
+                case 2: // language and country
+                    return new java.util.Locale(split[0], split[1]);
+                case 3: // language, country, and variant
+                    return new java.util.Locale(split[0], split[1], split[2]);
+            }
+        } catch (IllegalArgumentException e) {
+            // ignore
+        }
+
+        // bad locale sent?
+        return java.util.Locale.US;
+    }
+    // SportPaper end
 }
-- 
2.6.4

