From b6576302e9cce23cb1ce678242a4ba544d8f8c92 Mon Sep 17 00:00:00 2001
From: Austin Mayes <almayes@trevecca.edu>
Date: Fri, 8 Mar 2019 16:01:32 -0600
Subject: [PATCH] Expose PacketPlayOutSpawnEntity constructor


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
index 69dd590a..8a6accf5 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 import java.io.IOException;
+import java.util.UUID;
 
 public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
 
@@ -22,6 +23,22 @@ public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
         this(entity, i, 0);
     }
 
+    // SportPaper start - add constructor
+    public PacketPlayOutSpawnEntity(int id, double xPos, double yPos, double zPos, int xVel, int yVel, int zVel, int pitch, int yaw, int type, int data) {
+        this.a = id;
+        this.b = MathHelper.floor(xPos * 32.0D);
+        this.c = MathHelper.floor(yPos * 32.0D);
+        this.d = MathHelper.floor(zPos * 32.0D);
+        this.e = xVel;
+        this.f = yVel;
+        this.g = zVel;
+        this.h = MathHelper.d(yaw * 256.0F / 360.0F);;
+        this.i = MathHelper.d(pitch * 256.0F / 360.0F);;
+        this.j = type;
+        this.k = data;
+    }
+    // SportPaper end
+
     public PacketPlayOutSpawnEntity(Entity entity, int i, int j) {
         this.a = entity.getId();
         this.b = MathHelper.floor(entity.locX * 32.0D);
@@ -129,8 +146,4 @@ public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
     public void f(int i) {
         this.g = i;
     }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayOut) packetlistener);
-    }
 }
-- 
2.20.1

