From 0c65cd292b9c0b2e63338e31348c6ec232d99d75 Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Thu, 28 Mar 2019 18:09:05 -0500
Subject: [PATCH] Fix NPE in CraftChunk#getBlocks


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 7907c983..abab23d6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -94,7 +94,7 @@ public class CraftChunk implements Chunk {
             for(int i = 0; i < blockIds.length; i++) {
                 // This does a lookup in the block registry, but does not create any objects, so should be pretty efficient
                 IBlockData blockData = (IBlockData) net.minecraft.server.Block.d.a(blockIds[i]);
-                if(blockData.getBlock() == nmsBlock) {
+                if(blockData != null && blockData.getBlock() == nmsBlock) {
                     blocks.add(getBlock(i & 0xf, section.getYPosition() | (i >> 8), (i >> 4) & 0xf));
                 }
             }
-- 
2.20.1

