From 8d2332e50ed6b4aaa49a49de8d19a3fb8e91bfa7 Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Sun, 14 Apr 2019 16:11:30 -0500
Subject: [PATCH] Add reason to potion effect add event


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index d4b8a0ef..1bcb177f 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -24,6 +24,7 @@ import org.bukkit.event.entity.EntityDamageEvent;
 import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
 import org.bukkit.event.entity.EntityRegainHealthEvent;
 import org.bukkit.event.entity.PotionEffectAddEvent;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 import org.bukkit.event.entity.PotionEffectExpireEvent;
 import org.bukkit.event.entity.PotionEffectExtendEvent;
 import org.bukkit.event.entity.PotionEffectRemoveEvent;
@@ -601,13 +602,17 @@ public abstract class EntityLiving extends Entity {
         return (MobEffect) this.effects.get(Integer.valueOf(mobeffectlist.id));
     }
 
-    public void addEffect(MobEffect mobeffect) {
+    public void addEffect(MobEffect effect) {
+        addEffect(effect, EffectAddReason.UNKNOWN);
+    }
+
+    public void addEffect(MobEffect mobeffect, EffectAddReason reason) {
         org.spigotmc.AsyncCatcher.catchOp( "effect add"); // Spigot
         if (this.d(mobeffect)) {
             if (this.effects.containsKey(Integer.valueOf(mobeffect.getEffectId()))) {
                 // SportBukkit start - fire event
                 PotionEffectExtendEvent event = new PotionEffectExtendEvent((LivingEntity) this.getBukkitEntity(),
-                        CraftPotionUtils.toBukkit(mobeffect),
+                        CraftPotionUtils.toBukkit(mobeffect), reason,
                         CraftPotionUtils.toBukkit(this.effects.get(mobeffect.getEffectId())));
                 this.world.getServer().getPluginManager().callEvent(event);
                 if(event.isCancelled()) return;
@@ -617,7 +622,7 @@ public abstract class EntityLiving extends Entity {
             } else {
                 // SportBukkit start - fire event
                 PotionEffectAddEvent event = new PotionEffectAddEvent((LivingEntity) this.getBukkitEntity(),
-                        CraftPotionUtils.toBukkit(mobeffect));
+                        CraftPotionUtils.toBukkit(mobeffect), reason);
                 this.world.getServer().getPluginManager().callEvent(event);
                 if(event.isCancelled()) return;
                 // SportBukkit end
diff --git a/src/main/java/net/minecraft/server/EntityPotion.java b/src/main/java/net/minecraft/server/EntityPotion.java
index fd174c34..f4837eac 100644
--- a/src/main/java/net/minecraft/server/EntityPotion.java
+++ b/src/main/java/net/minecraft/server/EntityPotion.java
@@ -8,6 +8,7 @@ import java.util.HashMap;
 
 import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 // CraftBukkit end
 
 public class EntityPotion extends EntityProjectile {
@@ -120,7 +121,7 @@ public class EntityPotion extends EntityProjectile {
                                     int j = (int) (d1 * (double) mobeffect.getDuration() + 0.5D);
 
                                     if (j > 20) {
-                                        entityliving.addEffect(new MobEffect(i, j, mobeffect.getAmplifier()));
+                                        entityliving.addEffect(new MobEffect(i, j, mobeffect.getAmplifier()), EffectAddReason.POTION_SPLASH);
                                     }
                                 }
                             }
diff --git a/src/main/java/net/minecraft/server/EntitySkeleton.java b/src/main/java/net/minecraft/server/EntitySkeleton.java
index 46bff95d..b284903c 100644
--- a/src/main/java/net/minecraft/server/EntitySkeleton.java
+++ b/src/main/java/net/minecraft/server/EntitySkeleton.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.Calendar;
 
 import org.bukkit.event.entity.EntityCombustEvent; // CraftBukkit
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 
 public class EntitySkeleton extends EntityMonster implements IRangedEntity {
 
@@ -56,7 +57,7 @@ public class EntitySkeleton extends EntityMonster implements IRangedEntity {
     public boolean r(Entity entity) {
         if (super.r(entity)) {
             if (this.getSkeletonType() == 1 && entity instanceof EntityLiving) {
-                ((EntityLiving) entity).addEffect(new MobEffect(MobEffectList.WITHER.id, 200));
+                ((EntityLiving) entity).addEffect(new MobEffect(MobEffectList.WITHER.id, 200), EffectAddReason.WITHER_SKELETON);
             }
 
             return true;
diff --git a/src/main/java/net/minecraft/server/EntitySpider.java b/src/main/java/net/minecraft/server/EntitySpider.java
index 68a253c9..e3ca4ae8 100644
--- a/src/main/java/net/minecraft/server/EntitySpider.java
+++ b/src/main/java/net/minecraft/server/EntitySpider.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 import java.util.Random;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 
 public class EntitySpider extends EntityMonster {
 
@@ -127,7 +128,7 @@ public class EntitySpider extends EntityMonster {
             int i = ((EntitySpider.GroupDataSpider) object).a;
 
             if (i > 0 && MobEffectList.byId[i] != null) {
-                this.addEffect(new MobEffect(i, Integer.MAX_VALUE));
+                this.addEffect(new MobEffect(i, Integer.MAX_VALUE), EffectAddReason.SPIDER_POWERUP);
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityVillager.java b/src/main/java/net/minecraft/server/EntityVillager.java
index 0d31f10b..cbaf4e2f 100644
--- a/src/main/java/net/minecraft/server/EntityVillager.java
+++ b/src/main/java/net/minecraft/server/EntityVillager.java
@@ -3,6 +3,7 @@ package net.minecraft.server;
 import java.util.Iterator;
 import java.util.Random;
 import org.bukkit.craftbukkit.entity.CraftVillager; // CraftBukkit
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 
 public class EntityVillager extends EntityAgeable implements IMerchant, NPC {
 
@@ -119,7 +120,7 @@ public class EntityVillager extends EntityAgeable implements IMerchant, NPC {
                     }
                 }
 
-                this.addEffect(new MobEffect(MobEffectList.REGENERATION.id, 200, 0));
+                this.addEffect(new MobEffect(MobEffectList.REGENERATION.id, 200, 0), EffectAddReason.VILLAGER_LEVELUP);
             }
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityWitherSkull.java b/src/main/java/net/minecraft/server/EntityWitherSkull.java
index f0e627ba..c3785a18 100644
--- a/src/main/java/net/minecraft/server/EntityWitherSkull.java
+++ b/src/main/java/net/minecraft/server/EntityWitherSkull.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 import org.bukkit.event.entity.ExplosionPrimeEvent; // CraftBukkit
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 
 public class EntityWitherSkull extends EntityFireball {
 
@@ -62,7 +63,7 @@ public class EntityWitherSkull extends EntityFireball {
                     }
 
                     if (b0 > 0) {
-                        ((EntityLiving) movingobjectposition.entity).addEffect(new MobEffect(MobEffectList.WITHER.id, 20 * b0, 1));
+                        ((EntityLiving) movingobjectposition.entity).addEffect(new MobEffect(MobEffectList.WITHER.id, 20 * b0, 1), EffectAddReason.WITHER_SKULL);
                     }
                 }
             }
diff --git a/src/main/java/net/minecraft/server/EntityZombie.java b/src/main/java/net/minecraft/server/EntityZombie.java
index 6d3b002b..a284d366 100644
--- a/src/main/java/net/minecraft/server/EntityZombie.java
+++ b/src/main/java/net/minecraft/server/EntityZombie.java
@@ -5,11 +5,11 @@ import java.util.List;
 import java.util.UUID;
 
 //CraftBukkit start
-import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.entity.EntityCombustEvent;
 import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 //CraftBukkit end
 
 public class EntityZombie extends EntityMonster {
@@ -464,7 +464,7 @@ public class EntityZombie extends EntityMonster {
         this.bn = i;
         this.getDataWatcher().watch(14, Byte.valueOf((byte) 1));
         this.removeEffect(MobEffectList.WEAKNESS.id);
-        this.addEffect(new MobEffect(MobEffectList.INCREASE_DAMAGE.id, i, Math.min(this.world.getDifficulty().a() - 1, 0)));
+        this.addEffect(new MobEffect(MobEffectList.INCREASE_DAMAGE.id, i, Math.min(this.world.getDifficulty().a() - 1, 0)), EffectAddReason.VILLAGER_CURE);
         this.world.broadcastEntityEffect(this, (byte) 16);
     }
 
@@ -494,7 +494,7 @@ public class EntityZombie extends EntityMonster {
         }
 
         this.world.addEntity(entityvillager, CreatureSpawnEvent.SpawnReason.CURED); // CraftBukkit - add SpawnReason
-        entityvillager.addEffect(new MobEffect(MobEffectList.CONFUSION.id, 200, 0));
+        entityvillager.addEffect(new MobEffect(MobEffectList.CONFUSION.id, 200, 0), EffectAddReason.VILLAGER_CURE);
         this.world.a((EntityHuman) null, 1017, new BlockPosition((int) this.locX, (int) this.locY, (int) this.locZ), 0);
     }
 
diff --git a/src/main/java/net/minecraft/server/ItemGoldenApple.java b/src/main/java/net/minecraft/server/ItemGoldenApple.java
index 1537d213..b558cde5 100644
--- a/src/main/java/net/minecraft/server/ItemGoldenApple.java
+++ b/src/main/java/net/minecraft/server/ItemGoldenApple.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
+
 public class ItemGoldenApple extends ItemFood {
 
     public ItemGoldenApple(int i, float f, boolean flag) {
@@ -13,14 +15,14 @@ public class ItemGoldenApple extends ItemFood {
 
     protected void c(ItemStack itemstack, World world, EntityHuman entityhuman) {
         if (!world.isClientSide) {
-            entityhuman.addEffect(new MobEffect(MobEffectList.ABSORBTION.id, 2400, 0));
+            entityhuman.addEffect(new MobEffect(MobEffectList.ABSORBTION.id, 2400, 0), EffectAddReason.GOLDEN_APPLE);
         }
 
         if (itemstack.getData() > 0) {
             if (!world.isClientSide) {
-                entityhuman.addEffect(new MobEffect(MobEffectList.REGENERATION.id, 600, 4));
-                entityhuman.addEffect(new MobEffect(MobEffectList.RESISTANCE.id, 6000, 0));
-                entityhuman.addEffect(new MobEffect(MobEffectList.FIRE_RESISTANCE.id, 6000, 0));
+                entityhuman.addEffect(new MobEffect(MobEffectList.REGENERATION.id, 600, 4), EffectAddReason.GOLDEN_APPLE);
+                entityhuman.addEffect(new MobEffect(MobEffectList.RESISTANCE.id, 6000, 0), EffectAddReason.GOLDEN_APPLE);
+                entityhuman.addEffect(new MobEffect(MobEffectList.FIRE_RESISTANCE.id, 6000, 0), EffectAddReason.GOLDEN_APPLE);
             }
         } else {
             super.c(itemstack, world, entityhuman);
diff --git a/src/main/java/net/minecraft/server/ItemPotion.java b/src/main/java/net/minecraft/server/ItemPotion.java
index 45dd15bd..134c1376 100644
--- a/src/main/java/net/minecraft/server/ItemPotion.java
+++ b/src/main/java/net/minecraft/server/ItemPotion.java
@@ -6,6 +6,7 @@ import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 
 public class ItemPotion extends Item {
 
@@ -71,7 +72,7 @@ public class ItemPotion extends Item {
                 while (iterator.hasNext()) {
                     MobEffect mobeffect = (MobEffect) iterator.next();
 
-                    entityhuman.addEffect(new MobEffect(mobeffect));
+                    entityhuman.addEffect(new MobEffect(mobeffect), EffectAddReason.POTION_DRINK);
                 }
             }
         }
diff --git a/src/main/java/net/minecraft/server/TileEntityBeacon.java b/src/main/java/net/minecraft/server/TileEntityBeacon.java
index f4717af7..164c52e3 100644
--- a/src/main/java/net/minecraft/server/TileEntityBeacon.java
+++ b/src/main/java/net/minecraft/server/TileEntityBeacon.java
@@ -13,6 +13,7 @@ import org.bukkit.entity.HumanEntity;
 // PaperSpigot start
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.entity.Player;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 import org.bukkit.potion.PotionEffect;
 import org.bukkit.potion.PotionEffectType;
 import org.github.paperspigot.event.block.BeaconEffectEvent;
@@ -96,7 +97,7 @@ public class TileEntityBeacon extends TileEntityContainer implements IUpdatePlay
                 if (CraftEventFactory.callEvent(event).isCancelled()) continue;
 
                 PotionEffect effect = event.getEffect();
-                entityhuman.addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()));
+                entityhuman.addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()), EffectAddReason.BEACON);
                 // PaperSpigot end
             }
 
@@ -111,7 +112,7 @@ public class TileEntityBeacon extends TileEntityContainer implements IUpdatePlay
                     if (CraftEventFactory.callEvent(event).isCancelled()) continue;
 
                     PotionEffect effect = event.getEffect();
-                    entityhuman.addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()));
+                    entityhuman.addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()), EffectAddReason.BEACON);
                     // PaperSpigot end
                 }
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 4351e304..f74d6654 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -56,6 +56,7 @@ import org.bukkit.entity.Snowball;
 import org.bukkit.entity.ThrownExpBottle;
 import org.bukkit.entity.ThrownPotion;
 import org.bukkit.entity.WitherSkull;
+import org.bukkit.event.entity.PotionEffectAddEvent.EffectAddReason;
 import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.inventory.EntityEquipment;
 import org.bukkit.inventory.ItemStack;
@@ -295,24 +296,36 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     }
 
     public boolean addPotionEffect(PotionEffect effect) {
-        return addPotionEffect(effect, false);
+        return addPotionEffect(effect, EffectAddReason.CUSTOM);
+    }
+
+    public boolean addPotionEffect(PotionEffect effect, EffectAddReason reason) {
+        return addPotionEffect(effect, false, reason);
     }
 
     public boolean addPotionEffect(PotionEffect effect, boolean force) {
+        return addPotionEffect(effect, force, EffectAddReason.CUSTOM);
+    }
+
+    public boolean addPotionEffect(PotionEffect effect, boolean force, EffectAddReason reason) {
         if (hasPotionEffect(effect.getType())) {
             if (!force) {
                 return false;
             }
             removePotionEffect(effect.getType());
         }
-        getHandle().addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()));
+        getHandle().addEffect(new MobEffect(effect.getType().getId(), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles()), reason);
         return true;
     }
 
     public boolean addPotionEffects(Collection<PotionEffect> effects) {
+        return addPotionEffects(effects, EffectAddReason.CUSTOM);
+    }
+
+    public boolean addPotionEffects(Collection<PotionEffect> effects, EffectAddReason reason) {
         boolean success = true;
         for (PotionEffect effect : effects) {
-            success &= addPotionEffect(effect);
+            success &= addPotionEffect(effect, reason);
         }
         return success;
     }
-- 
2.20.1

