From ed5dbef3f88d2aa18d187d6b20d1b1255f58b82b Mon Sep 17 00:00:00 2001
From: Ashcon Partovi <ashcon@partovi.net>
Date: Thu, 2 May 2019 15:46:23 -0700
Subject: [PATCH] Add willBeOnline API


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 9f1039012..d09f7453a 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -366,6 +366,8 @@ public abstract class PlayerList {
     public String disconnect(EntityPlayer entityplayer) { // CraftBukkit - return string
         entityplayer.b(StatisticList.f);
 
+        ((CraftPlayer) cserver.getPlayer(entityplayer)).startQuitting(); // SportPaper
+
         // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
         org.bukkit.craftbukkit.event.CraftEventFactory.handleInventoryCloseEvent(entityplayer);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 45217866b..b4ca64010 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -43,6 +43,12 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         return getPlayer() != null;
     }
 
+    @Override
+    public boolean willBeOnline() {
+        final Player player = getPlayer();
+        return player != null && player != this && player.willBeOnline();
+    }
+
     public String getName() {
         Player player = getPlayer();
         if (player != null) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 442c8119f..60c81cdc7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -89,6 +89,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private Skin realSkin;
     private boolean showInvisibles;
     private java.util.Locale cachedLocale = java.util.Locale.US; // SportPaper
+    private boolean quitting; // SportPaper
 
     private final Map<CommandSender, String> fakeNames = new WeakHashMap<CommandSender, String>();
     private final Map<CommandSender, String> fakeDisplayNames = new WeakHashMap<CommandSender, String>();
@@ -127,6 +128,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return server.getPlayer(getUniqueId()) != null;
     }
 
+    @Override
+    public boolean willBeOnline() {
+        return !quitting;
+    }
+
+    public void startQuitting() {
+        quitting = true;
+    }
+
     public InetSocketAddress getAddress() {
         if (getHandle().playerConnection == null) return null;
 
-- 
2.20.1

