From ff2f9cfedfc710f221187daffa7872d2c25d07f7 Mon Sep 17 00:00:00 2001
From: "Maxwell C. West" <maxwell@blueoxygen.net>
Date: Wed, 22 Apr 2020 17:24:09 -0400
Subject: [PATCH] Remove double chunk lookup.


diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index d3767d2a..82416a53 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -128,9 +128,11 @@ public class ActivationRange
             {
                 for ( int j1 = k; j1 <= l; ++j1 )
                 {
-                    if ( world.getWorld().isChunkLoaded( i1, j1 ) )
+                    Chunk chunk = world.getChunkIfLoaded( i1, j1 ); // SportPaper - remove double chunk lookup
+
+                    if ( chunk != null )
                     {
-                        activateChunkEntities( world.getChunkAt( i1, j1 ) );
+                        activateChunkEntities( chunk );
                     }
                 }
             }
-- 
2.25.0.windows.1

