--- ../src-base/minecraft/net/minecraft/entity/passive/EntityLlama.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityLlama.java
@@ -66,7 +66,7 @@
         this.setSize(0.9F, 1.87F);
     }
 
-    private void setStrength(int strengthIn)
+    public void setStrength(int strengthIn)
     {
         this.dataManager.set(DATA_STRENGTH_ID, Integer.valueOf(Math.max(1, Math.min(5, strengthIn))));
     }
