--- ../src-base/minecraft/net/minecraft/entity/passive/EntityTameable.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityTameable.java
@@ -27,6 +27,7 @@
     protected static final DataParameter<Byte> TAMED = EntityDataManager.<Byte>createKey(EntityTameable.class, DataSerializers.BYTE);
     protected static final DataParameter<Optional<UUID>> OWNER_UNIQUE_ID = EntityDataManager.<Optional<UUID>>createKey(EntityTameable.class, DataSerializers.OPTIONAL_UNIQUE_ID);
     protected EntityAISit aiSit;
+    @Nullable private Optional<UUID> cachedOwnerId;
 
     public EntityTameable(World worldIn)
     {
@@ -180,12 +181,14 @@
     @Nullable
     public UUID getOwnerId()
     {
-        return (UUID)((Optional)this.dataManager.get(OWNER_UNIQUE_ID)).orNull();
+        if (cachedOwnerId == null) cachedOwnerId = (UUID)((Optional)dataManager.get(OWNER_UNIQUE_ID);
+        return cachedOwnerId.orNull();
     }
 
     public void setOwnerId(@Nullable UUID p_184754_1_)
     {
-        this.dataManager.set(OWNER_UNIQUE_ID, Optional.fromNullable(p_184754_1_));
+        cachedOwnerId = Optional.fromNullable(p_184754_1_);
+        dataManager.set(OWNER_UNIQUE_ID, cachedOwnerId);
     }
 
     public void setTamedBy(EntityPlayer player)
@@ -223,6 +226,11 @@
         return this.aiSit;
     }
 
+    public void setAISit(EntityAISit aiSit)
+    {
+        this.aiSit = aiSit;
+    }
+
     public boolean shouldAttackEntity(EntityLivingBase target, EntityLivingBase owner)
     {
         return true;
