--- a/net/minecraft/block/BeehiveBlock.java
+++ b/net/minecraft/block/BeehiveBlock.java
@@ -85,11 +85,12 @@
       List<BeeEntity> list = p_226881_1_.func_217357_a(BeeEntity.class, (new AxisAlignedBB(p_226881_2_)).func_72314_b(8.0D, 6.0D, 8.0D));
       if (!list.isEmpty()) {
          List<PlayerEntity> list1 = p_226881_1_.func_217357_a(PlayerEntity.class, (new AxisAlignedBB(p_226881_2_)).func_72314_b(8.0D, 6.0D, 8.0D));
+         if (list1.isEmpty()) return; //Forge: Prevent Error when no players are around.
          int i = list1.size();
 
          for(BeeEntity beeentity : list) {
             if (beeentity.func_70638_az() == null) {
-               beeentity.func_70624_b(list1.get(p_226881_1_.field_73012_v.nextInt(i)));
+               beeentity.setGoalTarget(list1.get(p_226881_1_.field_73012_v.nextInt(i)), org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit
             }
          }
       }
