--- a/net/minecraft/block/PortalInfo.java
+++ b/net/minecraft/block/PortalInfo.java
@@ -1,14 +1,22 @@
 package net.minecraft.block;
 
 import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.event.CraftPortalEvent;
 
 public class PortalInfo {
    public final Vector3d field_222505_a;
    public final Vector3d field_222506_b;
    public final float field_242960_c;
    public final float field_242961_d;
+   // CraftBukkit start
+   public final ServerWorld world;
+   public final CraftPortalEvent portalEventInfo;
 
-   public PortalInfo(Vector3d p_i242042_1_, Vector3d p_i242042_2_, float p_i242042_3_, float p_i242042_4_) {
+   public PortalInfo(Vector3d p_i242042_1_, Vector3d p_i242042_2_, float p_i242042_3_, float p_i242042_4_, ServerWorld world, CraftPortalEvent portalEventInfo) {
+      this.world = world;
+      this.portalEventInfo = portalEventInfo;
+      // CraftBukkit end
       this.field_222505_a = p_i242042_1_;
       this.field_222506_b = p_i242042_2_;
       this.field_242960_c = p_i242042_3_;
