--- a/net/minecraft/block/PortalSize.java
+++ b/net/minecraft/block/PortalSize.java
@@ -1,6 +_,7 @@
 package net.minecraft.block;
 
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.entity.EntitySize;
@@ -13,10 +_,15 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlockState;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftPortalEvent;
+import org.bukkit.event.world.PortalCreateEvent;
 
 public class PortalSize {
    private static final AbstractBlock.IPositionPredicate field_242962_a = (p_242966_0_, p_242966_1_, p_242966_2_) -> {
-      return p_242966_0_.func_203425_a(Blocks.field_150343_Z);
+      return p_242966_0_.isPortalFrame(p_242966_1_, p_242966_2_);
    };
    private final IWorld field_150867_a;
    private final Direction.Axis field_150865_b;
@@ -26,6 +_,7 @@
    private BlockPos field_150861_f;
    private int field_150862_g;
    private int field_150868_h;
+   org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator blocks; // CraftBukkit - add field
 
    public static Optional<PortalSize> func_242964_a(IWorld p_242964_0_, BlockPos p_242964_1_, Direction.Axis p_242964_2_) {
       return func_242965_a(p_242964_0_, p_242964_1_, (p_242968_0_) -> {
@@ -44,6 +_,7 @@
    }
 
    public PortalSize(IWorld p_i48740_1_, BlockPos p_i48740_2_, Direction.Axis p_i48740_3_) {
+      blocks = new org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator(p_i48740_1_.getMinecraftWorld()); // CraftBukkit
       this.field_150867_a = p_i48740_1_;
       this.field_150865_b = p_i48740_3_;
       this.field_150866_c = p_i48740_3_ == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
@@ -84,6 +_,7 @@
          BlockState blockstate = this.field_150867_a.func_180495_p(blockpos$mutable);
          if (!func_196900_a(blockstate)) {
             if (field_242962_a.test(blockstate, this.field_150867_a, blockpos$mutable)) {
+               blocks.func_180501_a(blockpos$mutable, blockstate, 18); // CraftBukkit - lower left / right
                return i;
             }
             break;
@@ -93,6 +_,7 @@
          if (!field_242962_a.test(blockstate1, this.field_150867_a, blockpos$mutable)) {
             break;
          }
+         blocks.func_180501_a(blockpos$mutable, blockstate1, 18); // CraftBukkit - bottom row
       }
 
       return 0;
@@ -110,6 +_,7 @@
          if (!field_242962_a.test(this.field_150867_a.func_180495_p(blockpos$mutable), this.field_150867_a, blockpos$mutable)) {
             return false;
          }
+         blocks.func_180501_a(blockpos$mutable, this.field_150867_a.func_180495_p(blockpos$mutable), 18); // CraftBukkit - upper row
       }
 
       return true;
@@ -138,6 +_,11 @@
                ++this.field_150864_e;
             }
          }
+
+         // CraftBukkit start - left and right
+         blocks.func_180501_a(p_242969_1_.func_189533_g(this.field_150861_f).func_189534_c(Direction.UP, i).func_189534_c(this.field_150866_c, -1), this.field_150867_a.func_180495_p(p_242969_1_), 18);
+         blocks.func_180501_a(p_242969_1_.func_189533_g(this.field_150861_f).func_189534_c(Direction.UP, i).func_189534_c(this.field_150866_c, this.field_150868_h), this.field_150867_a.func_180495_p(p_242969_1_), 18);
+         // CraftBukkit end
       }
 
       return 21;
@@ -151,6 +_,7 @@
       return this.field_150861_f != null && this.field_150868_h >= 2 && this.field_150868_h <= 21 && this.field_150862_g >= 3 && this.field_150862_g <= 21;
    }
 
+
    public void func_150859_c() {
       BlockState blockstate = Blocks.field_150427_aO.func_176223_P().func_206870_a(NetherPortalBlock.field_176550_a, this.field_150865_b);
       BlockPos.func_218278_a(this.field_150861_f, this.field_150861_f.func_177967_a(Direction.UP, this.field_150862_g - 1).func_177967_a(this.field_150866_c, this.field_150868_h - 1)).forEach((p_242967_2_) -> {
@@ -158,6 +_,31 @@
       });
    }
 
+   // Mohist start
+   // CraftBukkit start - return boolean
+   public boolean createPortal() {
+      org.bukkit.World bworld = this.field_150867_a.getMinecraftWorld().getWorld();
+
+      // Copy below for loop
+      BlockState blockstate = Blocks.field_150427_aO.func_176223_P().func_206870_a(NetherPortalBlock.field_176550_a, this.field_150865_b);
+      BlockPos.func_218278_a(this.field_150861_f, this.field_150861_f.func_177967_a(Direction.UP, this.field_150862_g - 1).func_177967_a(this.field_150866_c, this.field_150868_h - 1)).forEach((p_242967_2_) -> {
+         blocks.func_180501_a(p_242967_2_, blockstate, 18);
+      });
+
+      PortalCreateEvent event = new PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks.getList(), bworld, null, PortalCreateEvent.CreateReason.FIRE);
+      Bukkit.getPluginManager().callEvent(event);
+
+      if (event.isCancelled()) {
+         return false;
+      }
+      // CraftBukkit end
+      BlockPos.func_218278_a(this.field_150861_f, this.field_150861_f.func_177967_a(Direction.UP, this.field_150862_g - 1).func_177967_a(this.field_150866_c, this.field_150868_h - 1)).forEach((p_242967_2_) -> {
+         this.field_150867_a.func_180501_a(p_242967_2_, blockstate, 18);
+      });
+      return true; // CraftBukkit
+      // Mohist end
+   }
+
    public boolean func_208508_f() {
       return this.func_150860_b() && this.field_150864_e == this.field_150868_h * this.field_150862_g;
    }
@@ -187,6 +_,9 @@
       return new Vector3d(d2, d4, d3);
    }
 
+   // Mohist start
+   public static AtomicReference<CraftPortalEvent> mohist$portalEventInfo = new AtomicReference<>();
+
    public static PortalInfo func_242963_a(ServerWorld p_242963_0_, TeleportationRepositioner.Result p_242963_1_, Direction.Axis p_242963_2_, Vector3d p_242963_3_, EntitySize p_242963_4_, Vector3d p_242963_5_, float p_242963_6_, float p_242963_7_) {
       BlockPos blockpos = p_242963_1_.field_243679_a;
       BlockState blockstate = p_242963_0_.func_180495_p(blockpos);
@@ -200,6 +_,15 @@
       double d4 = 0.5D + p_242963_3_.func_82616_c();
       boolean flag = direction$axis == Direction.Axis.X;
       Vector3d vector3d1 = new Vector3d((double)blockpos.func_177958_n() + (flag ? d2 : d4), (double)blockpos.func_177956_o() + d3, (double)blockpos.func_177952_p() + (flag ? d4 : d2));
-      return new PortalInfo(vector3d1, vector3d, p_242963_6_ + (float)i, p_242963_7_);
+      PortalInfo portalInfo = new PortalInfo(vector3d1, vector3d, p_242963_6_ + (float)i, p_242963_7_);
+      if (mohist$portalEventInfo.get() != null){
+         portalInfo.putCB(p_242963_0_, mohist$portalEventInfo.getAndSet(null));
+      }
+      return portalInfo; // CraftBukkit
+   }
+
+   public static PortalInfo createPortalInfo(ServerWorld world, TeleportationRepositioner.Result result, Direction.Axis axis, Vector3d offsetVector, EntitySize size, Vector3d motion, float rotationYaw, float rotationPitch, CraftPortalEvent portalEventInfo) {
+      mohist$portalEventInfo.set(portalEventInfo);
+      return func_242963_a(world, result, axis, offsetVector, size, motion, rotationYaw, rotationPitch); // CraftBukkit
    }
 }
