--- ../src-base/minecraft/org/bukkit/help/HelpMap.java
+++ ../src-work/minecraft/org/bukkit/help/HelpMap.java
@@ -7,7 +7,7 @@
  * The HelpMap tracks all help topics registered in a Bukkit server. When the
  * server starts up or is reloaded, help is processed and topics are added in
  * the following order:
- * 
+ *
  * <ol>
  * <li>General topics are loaded from the help.yml
  * <li>Plugins load and optionally call {@code addTopic()}
@@ -32,7 +32,7 @@
      * @return All the registered help topics.
      */
     public Collection<HelpTopic> getHelpTopics();
-    
+
     /**
      * Adds a topic to the server's help index.
      *
