--- ../src-base/minecraft/org/bukkit/help/IndexHelpTopic.java
+++ ../src-work/minecraft/org/bukkit/help/IndexHelpTopic.java
@@ -1,13 +1,12 @@
 package org.bukkit.help;
 
+import java.util.Collection;
 import org.bukkit.ChatColor;
 import org.bukkit.command.CommandSender;
 import org.bukkit.command.ConsoleCommandSender;
 import org.bukkit.entity.Player;
 import org.bukkit.util.ChatPaginator;
 
-import java.util.Collection;
-
 /**
  * This help topic generates a list of other help topics. This class is useful
  * for adding your own index help topics. To enforce a particular order, use a
