--- a/net/minecraft/nbt/NBTUtil.java
+++ b/net/minecraft/nbt/NBTUtil.java
@@ -2,6 +_,7 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableMap;
+import com.mohistmc.util.i18n.i18n;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.serialization.Dynamic;
@@ -207,7 +_,7 @@
       if (optional.isPresent()) {
          return p_193590_0_.func_206870_a(p_193590_1_, optional.get());
       } else {
-         field_193591_a.warn("Unable to read property: {} with value: {} for blockstate: {}", p_193590_2_, p_193590_3_.func_74779_i(p_193590_2_), p_193590_4_.toString());
+         field_193591_a.warn(i18n.get("nbtutil.1", p_193590_2_, p_193590_3_.func_74779_i(p_193590_2_), p_193590_4_.toString()));
          return p_193590_0_;
       }
    }
