--- a/net/minecraft/world/chunk/ChunkPrimer.java
+++ b/net/minecraft/world/chunk/ChunkPrimer.java
@@ -3,6 +_,7 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+import com.mohistmc.util.i18n.i18n;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
@@ -64,6 +_,16 @@
    private final Map<GenerationStage.Carving, BitSet> field_205769_s = new Object2ObjectArrayMap<>();
    private volatile boolean field_217335_u;
 
+   // Paper start - track last save time
+   public long lastSaveTime;
+
+   @Override
+   public void setLastSaved(long ticks) {
+      this.lastSaveTime = ticks;
+   }
+
+   // Paper end
+
    public ChunkPrimer(ChunkPos p_i48700_1_, UpgradeData p_i48700_2_) {
       this(p_i48700_1_, p_i48700_2_, (ChunkSection[])null, new ChunkPrimerTickList<>((p_205332_0_) -> {
          return p_205332_0_ == null || p_205332_0_.func_176223_P().func_196958_f();
@@ -81,7 +_,7 @@
          if (this.field_201661_i.length == p_i49941_3_.length) {
             System.arraycopy(p_i49941_3_, 0, this.field_201661_i, 0, this.field_201661_i.length);
          } else {
-            field_201653_a.warn("Could not set level chunk sections, array length is {} instead of {}", p_i49941_3_.length, this.field_201661_i.length);
+            field_201653_a.warn(i18n.get("chunkprimer.1", p_i49941_3_.length, this.field_201661_i.length));
          }
       }
 
@@ -138,13 +_,13 @@
          if (this.field_201661_i[j >> 4] == Chunk.field_186036_a && p_177436_2_.func_203425_a(Blocks.field_150350_a)) {
             return p_177436_2_;
          } else {
-            if (p_177436_2_.func_185906_d() > 0) {
+            if (p_177436_2_.getLightValue(this, p_177436_1_) > 0) {
                this.field_201663_k.add(new BlockPos((i & 15) + this.func_76632_l().func_180334_c(), j, (k & 15) + this.func_76632_l().func_180333_d()));
             }
 
             ChunkSection chunksection = this.func_217332_a(j >> 4);
             BlockState blockstate = chunksection.func_222629_a(i & 15, j & 15, k & 15, p_177436_2_);
-            if (this.field_201658_f.func_209003_a(ChunkStatus.field_222613_i) && p_177436_2_ != blockstate && (p_177436_2_.func_200016_a(this, p_177436_1_) != blockstate.func_200016_a(this, p_177436_1_) || p_177436_2_.func_185906_d() != blockstate.func_185906_d() || p_177436_2_.func_215691_g() || blockstate.func_215691_g())) {
+            if (this.field_201658_f.func_209003_a(ChunkStatus.field_222613_i) && p_177436_2_ != blockstate && (p_177436_2_.func_200016_a(this, p_177436_1_) != blockstate.func_200016_a(this, p_177436_1_) || p_177436_2_.getLightValue(this, p_177436_1_) != blockstate.getLightValue(this, p_177436_1_) || p_177436_2_.func_215691_g() || blockstate.func_215691_g())) {
                WorldLightManager worldlightmanager = this.func_217307_e();
                worldlightmanager.func_215568_a(p_177436_1_);
             }
