--- ../src-base/minecraft/org/bukkit/conversations/ConversationContext.java
+++ ../src-work/minecraft/org/bukkit/conversations/ConversationContext.java
@@ -1,9 +1,8 @@
 package org.bukkit.conversations;
 
+import java.util.Map;
 import org.bukkit.plugin.Plugin;
 
-import java.util.Map;
-
 /**
  * A ConversationContext provides continuity between nodes in the prompt graph
  * by giving the developer access to the subject of the conversation and a
