--- ../src-base/minecraft/org/bukkit/conversations/Conversation.java
+++ ../src-work/minecraft/org/bukkit/conversations/Conversation.java
@@ -1,11 +1,10 @@
 package org.bukkit.conversations;
 
-import org.bukkit.plugin.Plugin;
-
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import org.bukkit.plugin.Plugin;
 
 /**
  * The Conversation class is responsible for tracking the current state of a
@@ -34,8 +33,6 @@
  */
 public class Conversation {
 
-    private Prompt firstPrompt;
-    private boolean abandoned;
     protected Prompt currentPrompt;
     protected ConversationContext context;
     protected boolean modal;
@@ -43,6 +40,8 @@
     protected ConversationPrefix prefix;
     protected List<ConversationCanceller> cancellers;
     protected List<ConversationAbandonedListener> abandonedListeners;
+    private Prompt firstPrompt;
+    private boolean abandoned;
 
     /**
      * Initializes a new Conversation.
