--- a/net/minecraft/server/packs/repository/FolderRepositorySource.java
+++ b/net/minecraft/server/packs/repository/FolderRepositorySource.java
@@ -88,8 +_,8 @@
         }
     }
 
-    static class FolderPackDetector extends PackDetector<Pack.ResourcesSupplier> {
-        protected FolderPackDetector(DirectoryValidator p_301187_) {
+    public static class FolderPackDetector extends PackDetector<Pack.ResourcesSupplier> {
+        public FolderPackDetector(DirectoryValidator p_301187_) {
             super(p_301187_);
         }
 
