--- a/net/minecraft/inventory/container/AbstractFurnaceContainer.java
+++ b/net/minecraft/inventory/container/AbstractFurnaceContainer.java
@@ -18,13 +18,30 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_15_R1.inventory.CraftInventoryFurnace;
+import org.bukkit.craftbukkit.v1_15_R1.inventory.CraftInventoryView;
 
 public abstract class AbstractFurnaceContainer extends RecipeBookContainer<IInventory> {
    private final IInventory field_217063_d;
    private final IIntArray field_217064_e;
    protected final World field_217062_c;
    private final IRecipeType<? extends AbstractCookingRecipe> field_217065_f;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
 
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryFurnace inventory = new CraftInventoryFurnace((AbstractFurnaceTileEntity) this.field_217063_d);
+      bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    protected AbstractFurnaceContainer(ContainerType<?> p_i50103_1_, IRecipeType<? extends AbstractCookingRecipe> p_i50103_2_, int p_i50103_3_, PlayerInventory p_i50103_4_) {
       this(p_i50103_1_, p_i50103_2_, p_i50103_3_, p_i50103_4_, new Inventory(3), new IntArray(4));
    }
@@ -40,7 +57,7 @@
       this.func_75146_a(new Slot(p_i50104_5_, 0, 56, 17));
       this.func_75146_a(new FurnaceFuelSlot(this, p_i50104_5_, 1, 56, 53));
       this.func_75146_a(new FurnaceResultSlot(p_i50104_4_.field_70458_d, p_i50104_5_, 2, 116, 35));
-
+      this.player = p_i50104_4_; // CraftBukkit - save player
       for(int i = 0; i < 3; ++i) {
          for(int j = 0; j < 9; ++j) {
             this.func_75146_a(new Slot(p_i50104_4_, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
@@ -91,6 +108,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_217063_d.func_70300_a(p_75145_1_);
    }
 
