--- a/net/minecraft/inventory/container/BlastFurnaceContainer.java
+++ b/net/minecraft/inventory/container/BlastFurnaceContainer.java
@@ -4,6 +4,7 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.item.crafting.IRecipeType;
 import net.minecraft.util.IIntArray;
+import org.bukkit.inventory.InventoryView;
 
 public class BlastFurnaceContainer extends AbstractFurnaceContainer {
    public BlastFurnaceContainer(int p_i50097_1_, PlayerInventory p_i50097_2_) {
