--- a/net/minecraft/inventory/container/LecternContainer.java
+++ b/net/minecraft/inventory/container/LecternContainer.java
@@ -1,6 +1,7 @@
 package net.minecraft.inventory.container;
 
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
@@ -8,16 +9,38 @@
 import net.minecraft.util.IntArray;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLectern;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerTakeLecternBookEvent;
 
 public class LecternContainer extends Container {
+   private CraftInventoryView bukkitEntity; // CraftBukkit
+   private Player player; // CraftBukkit
    private final IInventory field_217018_c;
    private final IIntArray field_217019_d;
 
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (this.bukkitEntity != null) {
+         return this.bukkitEntity;
+      }
+      final CraftInventoryLectern inventory = new CraftInventoryLectern(this.field_217018_c);
+      return this.bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+   }
+
+   public LecternContainer(final int i, PlayerInventory playerinventory) {
+      this(i, new net.minecraft.inventory.Inventory(1), new IntArray(1), playerinventory);
+   }
+   // CraftBukkit start
+
    public LecternContainer(int p_i50075_1_) {
-      this(p_i50075_1_, new Inventory(1), new IntArray(1));
+      this(p_i50075_1_, new Inventory(1), new IntArray(1), null);
    }
 
-   public LecternContainer(int p_i50076_1_, IInventory p_i50076_2_, IIntArray p_i50076_3_) {
+   public LecternContainer(int p_i50076_1_, IInventory p_i50076_2_, IIntArray p_i50076_3_, PlayerInventory playerinventory) {
       super(ContainerType.field_221523_q, p_i50076_1_);
       func_216962_a(p_i50076_2_, 1);
       func_216959_a(p_i50076_3_, 1);
@@ -30,6 +53,7 @@
          }
       });
       this.func_216961_a(p_i50076_3_);
+      this.player = (Player)playerinventory.field_70458_d.getBukkitEntity(); // CraftBukkit
    }
 
    public boolean func_75140_a(PlayerEntity p_75140_1_, int p_75140_2_) {
@@ -52,6 +76,14 @@
                return false;
             }
 
+            // CraftBukkit start - Event for taking the book
+            PlayerTakeLecternBookEvent event = new PlayerTakeLecternBookEvent(player, ((CraftInventoryLectern) getBukkitView().getTopInventory()).getHolder());
+            Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return false;
+            }
+            // CraftBukkit end
+
             ItemStack itemstack = this.field_217018_c.func_70304_b(0);
             this.field_217018_c.func_70296_d();
             if (!p_75140_1_.field_71071_by.func_70441_a(itemstack)) {
@@ -71,6 +103,7 @@
    }
 
    public boolean func_75145_c(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.field_217018_c.func_70300_a(p_75145_1_);
    }
 
