--- a/net/minecraft/inventory/container/SmokerContainer.java
+++ b/net/minecraft/inventory/container/SmokerContainer.java
@@ -4,6 +4,7 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.item.crafting.IRecipeType;
 import net.minecraft.util.IIntArray;
+import org.bukkit.inventory.InventoryView;
 
 public class SmokerContainer extends AbstractFurnaceContainer {
    public SmokerContainer(int p_i50061_1_, PlayerInventory p_i50061_2_) {
