--- a/net/minecraft/inventory/container/WorkbenchContainer.java
+++ b/net/minecraft/inventory/container/WorkbenchContainer.java
@@ -18,11 +18,12 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.inventory.InventoryView;
 
 public class WorkbenchContainer extends RecipeBookContainer<CraftingInventory> {
    private final CraftingInventory field_75162_e = new CraftingInventory(this, 3, 3);
    private final CraftResultInventory field_75160_f = new CraftResultInventory();
-   private final IWorldPosCallable field_217070_e;
+   public final IWorldPosCallable field_217070_e;
    private final PlayerEntity field_192390_i;
 
    public WorkbenchContainer(int p_i50089_1_, PlayerInventory p_i50089_2_) {
@@ -100,6 +101,11 @@
       return func_216963_a(this.field_217070_e, p_75145_1_, Blocks.field_150462_ai);
    }
 
+   @Override
+   public InventoryView getBukkitView() {
+      return null;
+   }
+
    public ItemStack func_82846_b(PlayerEntity p_82846_1_, int p_82846_2_) {
       ItemStack itemstack = ItemStack.field_190927_a;
       Slot slot = this.field_75151_b.get(p_82846_2_);
