--- ../src-base/minecraft/org/bukkit/BlockChangeDelegate.java
+++ ../src-work/minecraft/org/bukkit/BlockChangeDelegate.java
@@ -4,9 +4,8 @@
  * A delegate for handling block changes. This serves as a direct interface
  * between generation algorithms in the server implementation and utilizing
  * code.
- * @deprecated rarely used API that was largely for implementation purposes
  */
-@Deprecated
+
 public interface BlockChangeDelegate {
 
     /**
@@ -23,7 +22,7 @@
      * @return true if the block was set successfully
      * @deprecated Magic value
      */
-    @Deprecated
+
     public boolean setRawTypeId(int x, int y, int z, int typeId);
 
     /**
@@ -41,7 +40,7 @@
      * @return true if the block was set successfully
      * @deprecated Magic value
      */
-    @Deprecated
+
     public boolean setRawTypeIdAndData(int x, int y, int z, int typeId, int data);
 
     /**
@@ -56,7 +55,7 @@
      * @return true if the block was set successfully
      * @deprecated Magic value
      */
-    @Deprecated
+
     public boolean setTypeId(int x, int y, int z, int typeId);
 
     /**
@@ -72,7 +71,7 @@
      * @return true if the block was set successfully
      * @deprecated Magic value
      */
-    @Deprecated
+
     public boolean setTypeIdAndData(int x, int y, int z, int typeId, int data);
 
     /**
@@ -84,7 +83,7 @@
      * @return The block ID
      * @deprecated Magic value
      */
-    @Deprecated
+
     public int getTypeId(int x, int y, int z);
 
     /**
