--- ../src-base/minecraft/org/bukkit/Bukkit.java
+++ ../src-work/minecraft/org/bukkit/Bukkit.java
@@ -1,5 +1,6 @@
 package org.bukkit;
 
+import com.avaje.ebean.config.ServerConfig;
 import java.awt.image.BufferedImage;
 import java.io.File;
 import java.util.Collection;
@@ -9,7 +10,6 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.logging.Logger;
-
 import org.bukkit.Warning.WarningState;
 import org.bukkit.command.CommandException;
 import org.bukkit.command.CommandSender;
@@ -18,10 +18,10 @@
 import org.bukkit.entity.Player;
 import org.bukkit.event.inventory.InventoryType;
 import org.bukkit.help.HelpMap;
-import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.Inventory;
 import org.bukkit.inventory.InventoryHolder;
 import org.bukkit.inventory.ItemFactory;
+import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.Recipe;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.PluginManager;
@@ -31,8 +31,6 @@
 import org.bukkit.scoreboard.ScoreboardManager;
 import org.bukkit.util.CachedServerIcon;
 
-import com.avaje.ebean.config.ServerConfig;
-
 /**
  * Represents the Bukkit core, for version and Server singleton handling
  */
@@ -301,6 +299,7 @@
      */
     public static void reload() {
         server.reload();
+        org.spigotmc.CustomTimingsHandler.reload(); // Spigot
     }
 
     /**
