--- ../src-base/minecraft/org/bukkit/FireworkEffect.java
+++ ../src-work/minecraft/org/bukkit/FireworkEffect.java
@@ -1,15 +1,13 @@
 package org.bukkit;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import java.util.List;
 import java.util.Map;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.configuration.serialization.SerializableAs;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-
 /**
  * Represents a single firework effect.
  */
