--- ../src-base/minecraft/org/bukkit/GameMode.java
+++ ../src-work/minecraft/org/bukkit/GameMode.java
@@ -11,7 +11,13 @@
  * have
  */
 public enum GameMode {
+    // Cauldron start - FakePlayers do not set their gametype and use the default
     /**
+     * Default mode
+     */
+    NOT_SET(-1),
+    // Cauldron end
+    /**
      * Creative mode may fly, build instantly, become invulnerable and create
      * free items.
      */
