--- ../src-base/minecraft/org/bukkit/UnsafeValues.java
+++ ../src-work/minecraft/org/bukkit/UnsafeValues.java
@@ -1,7 +1,6 @@
 package org.bukkit;
 
 import java.util.List;
-
 import org.bukkit.advancement.Advancement;
 import org.bukkit.inventory.ItemStack;
 
@@ -13,7 +12,7 @@
  * may be poorly named, throw exceptions, have misleading parameters, or any
  * other bad programming practice.
  */
-@Deprecated
+
 public interface UnsafeValues {
 
     Material getMaterialFromInternalName(String name);
