--- ../src-base/minecraft/org/bukkit/material/Cauldron.java
+++ ../src-work/minecraft/org/bukkit/material/Cauldron.java
@@ -19,7 +19,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Cauldron(int type, byte data) {
         super(type, data);
     }
@@ -29,7 +29,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Cauldron(byte data) {
         super(Material.CAULDRON, data);
     }
