--- ../src-base/minecraft/org/bukkit/material/CocoaPlant.java
+++ ../src-work/minecraft/org/bukkit/material/CocoaPlant.java
@@ -8,21 +8,15 @@
  */
 public class CocoaPlant extends MaterialData implements Directional, Attachable {
 
-    public enum CocoaPlantSize {
-        SMALL,
-        MEDIUM,
-        LARGE
-    }
-
     public CocoaPlant() {
         super(Material.COCOA);
     }
 
     /**
-     * @param type the raw type id     
+     * @param type the raw type id
      * @deprecated Magic value
      */
-    @Deprecated
+
     public CocoaPlant(final int type) {
         super(type);
     }
@@ -32,7 +26,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public CocoaPlant(final int type, final byte data) {
         super(type, data);
     }
@@ -130,4 +124,10 @@
     public String toString() {
         return super.toString() + " facing " + getFacing() + " " + getSize();
     }
+
+    public enum CocoaPlantSize {
+        SMALL,
+        MEDIUM,
+        LARGE
+    }
 }
