--- ../src-base/minecraft/org/bukkit/material/Door.java
+++ ../src-work/minecraft/org/bukkit/material/Door.java
@@ -27,7 +27,7 @@
     /**
      * @deprecated Artifact of old API, equivalent to new <code>Door(Material.WOODEN_DOOR);</code>
      */
-    @Deprecated
+
     public Door() {
         super(Material.WOODEN_DOOR);
     }
@@ -36,7 +36,7 @@
      * @param type the raw type id
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Door(final int type) {
         super(type);
     }
@@ -170,7 +170,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Door(final int type, final byte data) {
         super(type, data);
     }
@@ -180,7 +180,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Door(final Material type, final byte data) {
         super(type, data);
     }
@@ -250,7 +250,7 @@
      * @return BlockFace.SELF
      * @deprecated This method should not be used; use hinge and facing accessors instead.
      */
-    @Deprecated
+
     public BlockFace getHingeCorner() {
         return BlockFace.SELF;
     }
