--- ../src-base/minecraft/org/bukkit/material/PoweredRail.java
+++ ../src-work/minecraft/org/bukkit/material/PoweredRail.java
@@ -14,7 +14,7 @@
      * @param type the raw type id
      * @deprecated Magic value
      */
-    @Deprecated
+
     public PoweredRail(final int type) {
         super(type);
     }
@@ -28,7 +28,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public PoweredRail(final int type, final byte data) {
         super(type, data);
     }
@@ -38,7 +38,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public PoweredRail(final Material type, final byte data) {
         super(type, data);
     }
