--- ../src-base/minecraft/org/bukkit/material/Step.java
+++ ../src-work/minecraft/org/bukkit/material/Step.java
@@ -2,7 +2,6 @@
 
 import java.util.ArrayList;
 import java.util.List;
-
 import org.bukkit.Material;
 
 /**
@@ -10,6 +9,7 @@
  */
 public class Step extends TexturedMaterial {
     private static final List<Material> textures = new ArrayList<Material>();
+
     static {
         textures.add(Material.STONE);
         textures.add(Material.SANDSTONE);
@@ -29,7 +29,7 @@
      * @param type the raw type id
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Step(final int type) {
         super(type);
     }
@@ -46,7 +46,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Step(final int type, final byte data) {
         super(type, data);
     }
@@ -56,7 +56,7 @@
      * @param data the raw data value
      * @deprecated Magic value
      */
-    @Deprecated
+
     public Step(final Material type, final byte data) {
         super(type, data);
     }
@@ -93,7 +93,7 @@
      *
      * @deprecated Magic value
      */
-    @Deprecated
+
     @Override
     protected int getTextureIndex() {
         return getData() & 0x7;
@@ -103,7 +103,7 @@
      *
      * @deprecated Magic value
      */
-    @Deprecated
+
     @Override
     protected void setTextureIndex(int idx) {
         setData((byte) ((getData() & 0x8) | idx));
