--- ../src-base/minecraft/net/minecraftforge/server/command/CommandSetDimension.java
+++ ../src-work/minecraft/net/minecraftforge/server/command/CommandSetDimension.java
@@ -33,6 +33,7 @@
 import javax.annotation.Nullable;
 import java.util.Collections;
 import java.util.List;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 
 public class CommandSetDimension extends CommandBase
 {
