--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -14,8 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
 
 public class BoatItem extends Item {
    private static final Predicate<Entity> f_40615_ = EntitySelector.f_20408_.and(Entity::m_6087_);
@@ -30,7 +_,7 @@
 
    public InteractionResultHolder<ItemStack> m_7203_(Level p_40622_, Player p_40623_, InteractionHand p_40624_) {
       ItemStack itemstack = p_40623_.m_21120_(p_40624_);
-      HitResult hitresult = m_41435_(p_40622_, p_40623_, ClipContext.Fluid.ANY);
+      BlockHitResult hitresult = m_41435_(p_40622_, p_40623_, ClipContext.Fluid.ANY);
       if (hitresult.m_6662_() == HitResult.Type.MISS) {
          return InteractionResultHolder.m_19098_(itemstack);
       } else {
@@ -49,6 +_,13 @@
          }
 
          if (hitresult.m_6662_() == HitResult.Type.BLOCK) {
+            // CraftBukkit start - Boat placement
+            org.bukkit.event.player.PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent(p_40623_, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, hitresult.m_82425_(), hitresult.m_82434_(), itemstack, false, p_40624_, hitresult.m_82450_());
+
+            if (event.isCancelled()) {
+               return InteractionResultHolder.m_19098_(itemstack);
+            }
+            // CraftBukkit end
             Boat boat = this.m_220016_(p_40622_, hitresult);
             boat.m_28464_(this.f_40616_);
             boat.m_146922_(p_40623_.m_146908_());
@@ -56,7 +_,15 @@
                return InteractionResultHolder.m_19100_(itemstack);
             } else {
                if (!p_40622_.f_46443_) {
-                  p_40622_.m_7967_(boat);
+                  // CraftBukkit start
+                  if (CraftEventFactory.callEntityPlaceEvent(p_40622_, hitresult.m_82425_(), hitresult.m_82434_(), p_40623_, boat, p_40624_).isCancelled()) {
+                     return InteractionResultHolder.m_19100_(itemstack);
+                  }
+
+                  if (!p_40622_.m_7967_(boat)) {
+                     return InteractionResultHolder.m_19098_(itemstack);
+                  }
+                  // CraftBukkit end
                   p_40622_.m_220400_(p_40623_, GameEvent.f_157810_, hitresult.m_82450_());
                   if (!p_40623_.m_150110_().f_35937_) {
                      itemstack.m_41774_(1);
