--- a/net/minecraft/world/item/EggItem.java
+++ b/net/minecraft/world/item/EggItem.java
@@ -16,13 +_,20 @@
 
    public InteractionResultHolder<ItemStack> use(Level p_41128_, Player p_41129_, InteractionHand p_41130_) {
       ItemStack itemstack = p_41129_.getItemInHand(p_41130_);
-      p_41128_.playSound((Player)null, p_41129_.getX(), p_41129_.getY(), p_41129_.getZ(), SoundEvents.EGG_THROW, SoundSource.PLAYERS, 0.5F, 0.4F / (p_41128_.getRandom().nextFloat() * 0.4F + 0.8F));
       if (!p_41128_.isClientSide) {
          ThrownEgg thrownegg = new ThrownEgg(p_41128_, p_41129_);
          thrownegg.setItem(itemstack);
          thrownegg.shootFromRotation(p_41129_, p_41129_.getXRot(), p_41129_.getYRot(), 0.0F, 1.5F, 1.0F);
-         p_41128_.addFreshEntity(thrownegg);
+         // CraftBukkit start
+         if (!p_41128_.addFreshEntity(thrownegg)) {
+            if (p_41129_ instanceof net.minecraft.server.level.ServerPlayer) {
+               ((net.minecraft.server.level.ServerPlayer) p_41129_).getBukkitEntity().updateInventory();
+            }
+            return InteractionResultHolder.fail(itemstack);
+         }
+         // CraftBukkit end
       }
+      p_41128_.playSound((Player)null, p_41129_.getX(), p_41129_.getY(), p_41129_.getZ(), SoundEvents.EGG_THROW, SoundSource.PLAYERS, 0.5F, 0.4F / (p_41128_.getRandom().nextFloat() * 0.4F + 0.8F));
 
       p_41129_.awardStat(Stats.ITEM_USED.get(this));
       if (!p_41129_.getAbilities().instabuild) {
