--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -40,6 +_,11 @@
                if (level instanceof ServerLevel) {
                   EndCrystal endcrystal = new EndCrystal(level, d0 + 0.5D, d1, d2 + 0.5D);
                   endcrystal.setShowBottom(false);
+                  // CraftBukkit start
+                  if (org.bukkit.craftbukkit.v1_20_R3.event.CraftEventFactory.callEntityPlaceEvent(p_41176_, endcrystal).isCancelled()) {
+                     return InteractionResult.FAIL;
+                  }
+                  // CraftBukkit end
                   level.addFreshEntity(endcrystal);
                   level.gameEvent(p_41176_.getPlayer(), GameEvent.ENTITY_PLACE, blockpos1);
                   EndDragonFight enddragonfight = ((ServerLevel)level).getDragonFight();
