--- a/net/minecraft/world/item/Equipable.java
+++ b/net/minecraft/world/item/Equipable.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.item;
 
+import com.mohistmc.mohist.plugins.ban.bans.BanItem;
 import javax.annotation.Nullable;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
@@ -22,6 +_,7 @@
 
    default InteractionResultHolder<ItemStack> swapWithEquipmentSlot(Item p_270453_, Level p_270395_, Player p_270300_, InteractionHand p_270262_) {
       ItemStack itemstack = p_270300_.getItemInHand(p_270262_);
+      if (BanItem.check(p_270300_, itemstack)) return InteractionResultHolder.fail(itemstack);
       EquipmentSlot equipmentslot = Mob.getEquipmentSlotForItem(itemstack);
       ItemStack itemstack1 = p_270300_.getItemBySlot(equipmentslot);
       if ((!EnchantmentHelper.hasBindingCurse(itemstack1) || p_270300_.isCreative()) && !ItemStack.matches(itemstack, itemstack1)) {
