--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -3,9 +_,11 @@
 import com.google.common.collect.HashMultimap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Multimap;
+import com.mohistmc.plugins.ban.bans.BanItem;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Codec;
+import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.text.DecimalFormat;
 import java.text.DecimalFormatSymbols;
@@ -31,6 +_,7 @@
 import net.minecraft.core.registries.Registries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
@@ -39,11 +_,13 @@
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.Style;
 import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.RandomSource;
+import net.minecraft.util.datafix.fixes.References;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.InteractionResultHolder;
@@ -70,17 +_,25 @@
 import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.ItemLike;
 import net.minecraft.world.level.Level;
-import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.*;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.pattern.BlockInWorld;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R1.util.CraftMagicNumbers;
+import org.bukkit.event.player.PlayerItemDamageEvent;
 import org.slf4j.Logger;
 
-public final class ItemStack {
+public final class ItemStack extends net.minecraftforge.common.capabilities.CapabilityProvider<ItemStack> implements net.minecraftforge.common.extensions.IForgeItemStack {
    public static final Codec<ItemStack> f_41582_ = RecordCodecBuilder.create((p_258963_) -> {
       return p_258963_.group(BuiltInRegistries.f_257033_.m_194605_().fieldOf("id").forGetter(ItemStack::m_41720_), Codec.INT.fieldOf("Count").forGetter(ItemStack::m_41613_), CompoundTag.f_128325_.optionalFieldOf("tag").forGetter((p_281115_) -> {
          return Optional.ofNullable(p_281115_.m_41783_());
       })).apply(p_258963_, ItemStack::new);
    });
+   @org.jetbrains.annotations.Nullable
+   private net.minecraft.core.Holder.Reference<Item> delegate;
+   private CompoundTag capNBT;
+
    private static final Logger f_41585_ = LogUtils.getLogger();
    public static final ItemStack f_41583_ = new ItemStack((Void)null);
    public static final DecimalFormat f_41584_ = Util.m_137469_(new DecimalFormat("#.##"), (p_41704_) -> {
@@ -104,7 +_,6 @@
    private int f_41588_;
    /** @deprecated */
    @Deprecated
-   @Nullable
    private Item f_41589_;
    @Nullable
    private CompoundTag f_41590_;
@@ -136,32 +_,88 @@
       this(p_220155_.m_203334_(), p_220156_);
    }
 
-   public ItemStack(ItemLike p_41601_, int p_41602_) {
-      this.f_41589_ = p_41601_.m_5456_();
-      this.f_41587_ = p_41602_;
-      if (this.f_41589_.m_41465_()) {
+   public ItemStack(ItemLike p_41601_, int p_41602_) { this(p_41601_, p_41602_, (CompoundTag) null); }
+   public ItemStack(ItemLike p_41604_, int p_41605_, @Nullable CompoundTag p_41606_) {
+      super(ItemStack.class, true);
+      this.capNBT = p_41606_;
+      this.f_41589_ = p_41604_.m_5456_();
+      this.delegate = net.minecraftforge.registries.ForgeRegistries.ITEMS.getDelegateOrThrow(p_41604_.m_5456_());
+      this.f_41587_ = p_41605_;
+      this.forgeInit();
+      if (this.f_41589_.isDamageable(this)) {
          this.m_41721_(this.m_41773_());
       }
 
    }
 
    private ItemStack(@Nullable Void p_282703_) {
+      super(ItemStack.class, true);
       this.f_41589_ = null;
-   }
-
-   private ItemStack(CompoundTag p_41608_) {
+      this.delegate = null;
+   }
+
+   // Called to run this stack through the data converter to handle older storage methods and serialized items
+   public void convertStack(int version) {
+      if (0 < version && version < CraftMagicNumbers.INSTANCE.getDataVersion()) {
+         CompoundTag savedStack = new CompoundTag();
+         this.m_41739_(savedStack);
+         savedStack = (CompoundTag) MinecraftServer.getServer().m_129933_().update(References.f_16782_, new Dynamic(NbtOps.f_128958_, savedStack), version, CraftMagicNumbers.INSTANCE.getDataVersion()).getValue();
+         this.load(savedStack);
+      }
+   }
+
+   // Mohist start fix 'Bigger Stacks' mod mixin
+   // CraftBukkit start - break into own method
+   private void load(CompoundTag compound) {
+      this.capNBT = compound.m_128441_("ForgeCaps") ? compound.m_128469_("ForgeCaps") : null;
+      Item rawItem =
+              this.f_41589_ = BuiltInRegistries.f_257033_.m_7745_(new ResourceLocation(compound.m_128461_("id")));
+      this.delegate = net.minecraftforge.registries.ForgeRegistries.ITEMS.getDelegateOrThrow(rawItem);
+      this.f_41587_ = compound.m_128445_("Count");
+      if (compound.m_128425_("tag", 10)) {
+         // CraftBukkit start - make defensive copy as this data may be coming from the save thread
+         this.f_41590_ = compound.m_128469_("tag").m_6426_();
+         this.m_41720_().m_142312_(this.f_41590_);
+         // CraftBukkit end
+      }
+
+      this.forgeInit();
+      if (this.m_41720_().isDamageable(this)) {
+         this.m_41721_(this.m_41773_());
+      }
+   }
+
+   public ItemStack(CompoundTag p_41608_) {
+      super(ItemStack.class, true);
+      this.capNBT = p_41608_.m_128441_("ForgeCaps") ? p_41608_.m_128469_("ForgeCaps") : null;
+      Item rawItem =
       this.f_41589_ = BuiltInRegistries.f_257033_.m_7745_(new ResourceLocation(p_41608_.m_128461_("id")));
+      this.delegate = net.minecraftforge.registries.ForgeRegistries.ITEMS.getDelegateOrThrow(rawItem);
       this.f_41587_ = p_41608_.m_128445_("Count");
       if (p_41608_.m_128425_("tag", 10)) {
+         // CraftBukkit start - make defensive copy as this data may be coming from the save thread
          this.f_41590_ = p_41608_.m_128469_("tag");
          this.m_41720_().m_142312_(this.f_41590_);
+         // CraftBukkit end
       }
 
-      if (this.m_41720_().m_41465_()) {
+      this.forgeInit();
+      if (this.m_41720_().isDamageable(this)) {
          this.m_41721_(this.m_41773_());
       }
-
-   }
+   }
+
+   public CompoundTag getForgeCaps() {
+      return this.serializeCaps();
+   }
+
+   public void setForgeCaps(CompoundTag nbt) {
+      this.capNBT = nbt;
+      if (nbt != null)
+         this.deserializeCaps(nbt);
+   }
+
+   // Mohist end
 
    public static ItemStack m_41712_(CompoundTag p_41713_) {
       try {
@@ -173,7 +_,7 @@
    }
 
    public boolean m_41619_() {
-      return this == f_41583_ || this.f_41589_ == Items.f_41852_ || this.f_41587_ <= 0;
+      return this == f_41583_ || this.f_41587_ <= 0 || this.delegate.get() == Items.f_41852_;
    }
 
    public boolean m_246617_(FeatureFlagSet p_250869_) {
@@ -198,7 +_,7 @@
    }
 
    public Item m_41720_() {
-      return this.m_41619_() ? Items.f_41852_ : this.f_41589_;
+      return this.m_41619_() ? Items.f_41852_ : this.delegate.get();
    }
 
    public Holder<Item> m_220173_() {
@@ -226,14 +_,24 @@
    }
 
    public InteractionResult m_41661_(UseOnContext p_41662_) {
+      if (!p_41662_.m_43725_().f_46443_) return net.minecraftforge.common.ForgeHooks.onPlaceItemIntoWorld(p_41662_);
+      return onItemUse(p_41662_, (c) -> m_41720_().m_6225_(p_41662_));
+   }
+
+   public InteractionResult onItemUseFirst(UseOnContext p_41662_) {
+      return onItemUse(p_41662_, (c) -> m_41720_().onItemUseFirst(this, p_41662_));
+   }
+
+   private InteractionResult onItemUse(UseOnContext p_41662_, java.util.function.Function<UseOnContext, InteractionResult> callback) {
       Player player = p_41662_.m_43723_();
+      if (BanItem.check(player, p_41662_.m_43722_())) return InteractionResult.FAIL;
       BlockPos blockpos = p_41662_.m_8083_();
       BlockInWorld blockinworld = new BlockInWorld(p_41662_.m_43725_(), blockpos, false);
       if (player != null && !player.m_150110_().f_35938_ && !this.m_204121_(p_41662_.m_43725_().m_9598_().m_175515_(Registries.f_256747_), blockinworld)) {
          return InteractionResult.PASS;
       } else {
          Item item = this.m_41720_();
-         InteractionResult interactionresult = item.m_6225_(p_41662_);
+         InteractionResult interactionresult = callback.apply(p_41662_);
          if (player != null && interactionresult.m_146666_()) {
             player.m_36246_(Stats.f_12982_.m_12902_(item));
          }
@@ -262,11 +_,15 @@
          p_41740_.m_128365_("tag", this.f_41590_.m_6426_());
       }
 
+      CompoundTag cnbt = this.serializeCaps();
+      if (cnbt != null && !cnbt.m_128456_()) {
+         p_41740_.m_128365_("ForgeCaps", cnbt);
+      }
       return p_41740_;
    }
 
    public int m_41741_() {
-      return this.m_41720_().m_41459_();
+      return this.m_41720_().getMaxStackSize(this);
    }
 
    public boolean m_41753_() {
@@ -274,7 +_,7 @@
    }
 
    public boolean m_41763_() {
-      if (!this.m_41619_() && this.m_41720_().m_41462_() > 0) {
+      if (!this.m_41619_() && this.m_41720_().isDamageable(this)) {
          CompoundTag compoundtag = this.m_41783_();
          return compoundtag == null || !compoundtag.m_128471_("Unbreakable");
       } else {
@@ -283,19 +_,19 @@
    }
 
    public boolean m_41768_() {
-      return this.m_41763_() && this.m_41773_() > 0;
+      return this.m_41763_() && m_41720_().isDamaged(this);
    }
 
    public int m_41773_() {
-      return this.f_41590_ == null ? 0 : this.f_41590_.m_128451_("Damage");
+      return this.m_41720_().getDamage(this);
    }
 
    public void m_41721_(int p_41722_) {
-      this.m_41784_().m_128405_("Damage", Math.max(0, p_41722_));
+      this.m_41720_().setDamage(this, p_41722_);
    }
 
    public int m_41776_() {
-      return this.m_41720_().m_41462_();
+      return this.m_41720_().getMaxDamage(this);
    }
 
    public boolean m_220157_(int p_220158_, RandomSource p_220159_, @Nullable ServerPlayer p_220160_) {
@@ -313,6 +_,21 @@
             }
 
             p_220158_ -= j;
+            // CraftBukkit start
+            if (p_220160_ != null) {
+               PlayerItemDamageEvent event = new PlayerItemDamageEvent(p_220160_.getBukkitEntity(), CraftItemStack.asCraftMirror(this), p_220158_);
+               event.getPlayer().getServer().getPluginManager().callEvent(event);
+
+               if (p_220158_ != event.getDamage() || event.isCancelled()) {
+                  event.getPlayer().updateInventory();
+               }
+               if (event.isCancelled()) {
+                  return false;
+               }
+
+               p_220158_ = event.getDamage();
+            }
+            // CraftBukkit end
             if (p_220158_ <= 0) {
                return false;
             }
@@ -331,9 +_,15 @@
    public <T extends LivingEntity> void m_41622_(int p_41623_, T p_41624_, Consumer<T> p_41625_) {
       if (!p_41624_.m_9236_().f_46443_ && (!(p_41624_ instanceof Player) || !((Player)p_41624_).m_150110_().f_35937_)) {
          if (this.m_41763_()) {
+            p_41623_ = this.m_41720_().damageItem(this, p_41623_, p_41624_, p_41625_);
             if (this.m_220157_(p_41623_, p_41624_.m_217043_(), p_41624_ instanceof ServerPlayer ? (ServerPlayer)p_41624_ : null)) {
                p_41625_.accept(p_41624_);
                Item item = this.m_41720_();
+               // CraftBukkit start - Check for item breaking
+               if (this.f_41587_ == 1 && p_41624_ instanceof Player) {
+                  CraftEventFactory.callPlayerItemBreakEvent((Player) p_41624_, this);
+               }
+               // CraftBukkit end
                this.m_41774_(1);
                if (p_41624_ instanceof Player) {
                   ((Player)p_41624_).m_36246_(Stats.f_12983_.m_12902_(item));
@@ -383,7 +_,7 @@
    }
 
    public boolean m_41735_(BlockState p_41736_) {
-      return this.m_41720_().m_8096_(p_41736_);
+      return this.m_41720_().isCorrectToolForDrops(this, p_41736_);
    }
 
    public InteractionResult m_41647_(Player p_41648_, LivingEntity p_41649_, InteractionHand p_41650_) {
@@ -394,7 +_,7 @@
       if (this.m_41619_()) {
          return f_41583_;
       } else {
-         ItemStack itemstack = new ItemStack(this.m_41720_(), this.f_41587_);
+         ItemStack itemstack = new ItemStack(this.m_41720_(), this.f_41587_, this.serializeCaps());
          itemstack.m_41754_(this.m_41612_());
          if (this.f_41590_ != null) {
             itemstack.f_41590_ = this.f_41590_.m_6426_();
@@ -430,7 +_,7 @@
       if (!p_150943_.m_150930_(p_150944_.m_41720_())) {
          return false;
       } else {
-         return p_150943_.m_41619_() && p_150944_.m_41619_() ? true : Objects.equals(p_150943_.f_41590_, p_150944_.f_41590_);
+         return p_150943_.m_41619_() && p_150944_.m_41619_() ? true : Objects.equals(p_150943_.f_41590_, p_150944_.f_41590_) && p_150943_.areCapsCompatible(p_150944_);
       }
    }
 
@@ -483,6 +_,17 @@
       return this.f_41590_;
    }
 
+   // CraftBukkit start
+   @Nullable
+   private CompoundTag getTagClone() {
+      return this.f_41590_ == null ? null : this.f_41590_.m_6426_();
+   }
+
+   private void setTagClone(@Nullable CompoundTag nbtttagcompound) {
+      this.m_41751_(nbtttagcompound == null ? null : nbtttagcompound.m_6426_());
+   }
+   // CraftBukkit end
+
    public CompoundTag m_41784_() {
       if (this.f_41590_ == null) {
          this.m_41751_(new CompoundTag());
@@ -520,9 +_,28 @@
       return this.f_41590_ != null ? this.f_41590_.m_128437_("Enchantments", 10) : new ListTag();
    }
 
+   // Paper start - (this is just a good no conflict location)
+   public org.bukkit.inventory.ItemStack asBukkitMirror() {
+      return CraftItemStack.asCraftMirror(this);
+   }
+   public org.bukkit.inventory.ItemStack asBukkitCopy() {
+      return CraftItemStack.asCraftMirror(this.m_41777_());
+   }
+   public static ItemStack fromBukkitCopy(org.bukkit.inventory.ItemStack itemstack) {
+      return CraftItemStack.asNMSCopy(itemstack);
+   }
+   private org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack bukkitStack;
+   public org.bukkit.inventory.ItemStack getBukkitStack() {
+      if (bukkitStack == null || bukkitStack.handle != this) {
+         bukkitStack = org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack.asCraftMirror(this);
+      }
+      return bukkitStack;
+   }
+   // Paper end
+
    public void m_41751_(@Nullable CompoundTag p_41752_) {
       this.f_41590_ = p_41752_;
-      if (this.m_41720_().m_41465_()) {
+      if (this.m_41720_().isDamageable(this)) {
          this.m_41721_(this.m_41773_());
       }
 
@@ -583,7 +_,7 @@
 
    public List<Component> m_41651_(@Nullable Player p_41652_, TooltipFlag p_41653_) {
       List<Component> list = Lists.newArrayList();
-      MutableComponent mutablecomponent = Component.m_237119_().m_7220_(this.m_41786_()).m_130940_(this.m_41791_().f_43022_);
+      MutableComponent mutablecomponent = Component.m_237119_().m_7220_(this.m_41786_()).m_130938_(this.m_41791_().getStyleModifier());
       if (this.m_41788_()) {
          mutablecomponent.m_130940_(ChatFormatting.ITALIC);
       }
@@ -730,6 +_,7 @@
          list.add(f_243862_);
       }
 
+      net.minecraftforge.event.ForgeEventFactory.onItemTooltip(this, p_41652_, list, p_41653_);
       return list;
    }
 
@@ -738,7 +_,7 @@
    }
 
    private int m_41618_() {
-      return this.m_41782_() && this.f_41590_.m_128425_("HideFlags", 99) ? this.f_41590_.m_128451_("HideFlags") : 0;
+      return this.m_41782_() && this.f_41590_.m_128425_("HideFlags", 99) ? this.f_41590_.m_128451_("HideFlags") : this.m_41720_().getDefaultTooltipHideFlags(this);
    }
 
    public void m_41654_(ItemStack.TooltipPart p_41655_) {
@@ -831,6 +_,12 @@
    }
 
    public void m_41742_(int p_41743_) {
+      // CraftBukkit start - remove RepairCost tag when 0 (SPIGOT-3945)
+      if (p_41743_ == 0) {
+         this.m_41749_("RepairCost");
+         return;
+      }
+      // CraftBukkit end
       this.m_41784_().m_128405_("RepairCost", p_41743_);
    }
 
@@ -853,9 +_,10 @@
             }
          }
       } else {
-         multimap = this.m_41720_().m_7167_(p_41639_);
+         multimap = this.m_41720_().getAttributeModifiers(p_41639_, this);
       }
 
+      multimap = net.minecraftforge.common.ForgeHooks.getAttributeModifiers(this, p_41639_, multimap);
       return multimap;
    }
 
@@ -883,7 +_,7 @@
 
       MutableComponent mutablecomponent1 = ComponentUtils.m_130748_(mutablecomponent);
       if (!this.m_41619_()) {
-         mutablecomponent1.m_130940_(this.m_41791_().f_43022_).m_130938_((p_220170_) -> {
+         mutablecomponent1.m_130938_(this.m_41791_().getStyleModifier()).m_130938_((p_220170_) -> {
             return p_220170_.m_131144_(new HoverEvent(HoverEvent.Action.f_130832_, new HoverEvent.ItemStackInfo(this)));
          });
       }
@@ -935,6 +_,8 @@
       this.m_41720_().m_5929_(p_41732_, p_41733_, this, p_41734_);
    }
 
+   /** @deprecated Forge: Use {@linkplain net.minecraftforge.common.extensions.IForgeItemStack#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
+   @Deprecated
    public void m_150924_(ItemEntity p_150925_) {
       this.m_41720_().m_142023_(p_150925_);
    }
@@ -943,6 +_,23 @@
       return this.m_41720_().m_41472_();
    }
 
+   // FORGE START
+   public void deserializeNBT(CompoundTag nbt) {
+      final ItemStack itemStack = ItemStack.m_41712_(nbt);
+      this.m_41751_(itemStack.m_41783_());
+      if (itemStack.capNBT != null) deserializeCaps(itemStack.capNBT);
+   }
+
+   /**
+    * Set up forge's ItemStack additions.
+    */
+   private void forgeInit() {
+      if (this.delegate != null) {
+         this.gatherCapabilities(() -> f_41589_.initCapabilities(this, this.capNBT));
+         if (this.capNBT != null) deserializeCaps(this.capNBT);
+      }
+   }
+
    public SoundEvent m_41615_() {
       return this.m_41720_().m_6023_();
    }
@@ -950,6 +_,14 @@
    public SoundEvent m_41616_() {
       return this.m_41720_().m_6061_();
    }
+
+   // CraftBukkit start
+   @Deprecated
+   public void setItem(Item item) {
+      this.f_41589_ = item;
+      this.delegate = net.minecraftforge.registries.ForgeRegistries.ITEMS.getDelegateOrThrow(item.m_5456_());
+   }
+   // CraftBukkit end
 
    public static enum TooltipPart {
       ENCHANTMENTS,
