--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -1,6 +_,5 @@
 package net.minecraft.world.item;
 
-import java.util.List;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerPlayer;
@@ -11,59 +_,90 @@
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.component.Tool;
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.GrowingPlantHeadBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class ShearsItem extends Item {
-    public ShearsItem(Item.Properties p_43074_) {
-        super(p_43074_);
-    }
-
-    public static Tool createToolProperties() {
-        return new Tool(
-            List.of(
-                Tool.Rule.minesAndDrops(List.of(Blocks.COBWEB), 15.0F),
-                Tool.Rule.overrideSpeed(BlockTags.LEAVES, 15.0F),
-                Tool.Rule.overrideSpeed(BlockTags.WOOL, 5.0F),
-                Tool.Rule.overrideSpeed(List.of(Blocks.VINE, Blocks.GLOW_LICHEN), 2.0F)
-            ),
-            1.0F,
-            1
-        );
-    }
-
-    @Override
-    public boolean mineBlock(ItemStack p_43078_, Level p_43079_, BlockState p_43080_, BlockPos p_43081_, LivingEntity p_43082_) {
-        if (!p_43079_.isClientSide && !p_43080_.is(BlockTags.FIRE)) {
-            p_43078_.hurtAndBreak(1, p_43082_, EquipmentSlot.MAINHAND);
-        }
-
-        return p_43080_.is(BlockTags.LEAVES)
-            || p_43080_.is(Blocks.COBWEB)
-            || p_43080_.is(Blocks.SHORT_GRASS)
-            || p_43080_.is(Blocks.FERN)
-            || p_43080_.is(Blocks.DEAD_BUSH)
-            || p_43080_.is(Blocks.HANGING_ROOTS)
-            || p_43080_.is(Blocks.VINE)
-            || p_43080_.is(Blocks.TRIPWIRE)
-            || p_43080_.is(BlockTags.WOOL);
-    }
-
-    @Override
-    public InteractionResult useOn(UseOnContext p_186371_) {
-        Level level = p_186371_.getLevel();
-        BlockPos blockpos = p_186371_.getClickedPos();
-        BlockState blockstate = level.getBlockState(blockpos);
-        if (blockstate.getBlock() instanceof GrowingPlantHeadBlock growingplantheadblock && !growingplantheadblock.isMaxAge(blockstate)) {
+   public ShearsItem(Item.Properties p_43074_) {
+      super(p_43074_);
+   }
+
+   public boolean mineBlock(ItemStack p_43078_, Level p_43079_, BlockState p_43080_, BlockPos p_43081_, LivingEntity p_43082_) {
+      if (!p_43079_.isClientSide && !p_43080_.is(BlockTags.FIRE)) {
+         p_43078_.hurtAndBreak(1, p_43082_, (p_43076_) -> {
+            p_43076_.broadcastBreakEvent(EquipmentSlot.MAINHAND);
+         });
+      }
+
+      return !p_43080_.is(BlockTags.LEAVES) && !p_43080_.is(Blocks.COBWEB) && !p_43080_.is(Blocks.SHORT_GRASS) && !p_43080_.is(Blocks.FERN) && !p_43080_.is(Blocks.DEAD_BUSH) && !p_43080_.is(Blocks.HANGING_ROOTS) && !p_43080_.is(Blocks.VINE) && !p_43080_.is(Blocks.TRIPWIRE) && !p_43080_.is(BlockTags.WOOL) ? super.mineBlock(p_43078_, p_43079_, p_43080_, p_43081_, p_43082_) : true;
+   }
+
+   public boolean isCorrectToolForDrops(BlockState p_43087_) {
+      return p_43087_.is(Blocks.COBWEB) || p_43087_.is(Blocks.REDSTONE_WIRE) || p_43087_.is(Blocks.TRIPWIRE);
+   }
+
+   public float getDestroySpeed(ItemStack p_43084_, BlockState p_43085_) {
+      if (!p_43085_.is(Blocks.COBWEB) && !p_43085_.is(BlockTags.LEAVES)) {
+         if (p_43085_.is(BlockTags.WOOL)) {
+            return 5.0F;
+         } else {
+            return !p_43085_.is(Blocks.VINE) && !p_43085_.is(Blocks.GLOW_LICHEN) ? super.getDestroySpeed(p_43084_, p_43085_) : 2.0F;
+         }
+      } else {
+         return 15.0F;
+      }
+   }
+
+   @Override
+   public net.minecraft.world.InteractionResult interactLivingEntity(ItemStack stack, net.minecraft.world.entity.player.Player playerIn, LivingEntity entity, net.minecraft.world.InteractionHand hand) {
+      if (entity instanceof net.minecraftforge.common.IForgeShearable target) {
+         if (entity.level().isClientSide) return net.minecraft.world.InteractionResult.SUCCESS;
+         BlockPos pos = BlockPos.containing(entity.position());
+         if (target.isShearable(stack, entity.level(), pos)) {
+            // CraftBukkit start
+            if (!CraftEventFactory.handlePlayerShearEntityEvent(playerIn, entity, stack, hand)) {
+               return InteractionResult.PASS;
+            }
+            // CraftBukkit end
+            java.util.List<ItemStack> drops = target.onSheared(playerIn, stack, entity.level(), pos,
+                    net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.BLOCK_FORTUNE, stack));
+
+            java.util.Random rand = new java.util.Random();
+            drops.forEach(d -> {
+               entity.forceDrops = true; // CraftBukkit
+               net.minecraft.world.entity.item.ItemEntity ent = entity.spawnAtLocation(d, 1.0F);
+               entity.forceDrops = false; // CraftBukkit
+               ent.setDeltaMovement(ent.getDeltaMovement().add((double)((rand.nextFloat() - rand.nextFloat()) * 0.1F), (double)(rand.nextFloat() * 0.05F), (double)((rand.nextFloat() - rand.nextFloat()) * 0.1F)));
+            });
+            stack.hurtAndBreak(1, playerIn, e -> e.broadcastBreakEvent(hand));
+         }
+         return net.minecraft.world.InteractionResult.SUCCESS;
+      }
+      return net.minecraft.world.InteractionResult.PASS;
+   }
+
+   @Override
+   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
+      return net.minecraftforge.common.ToolActions.DEFAULT_SHEARS_ACTIONS.contains(toolAction);
+   }
+
+   public InteractionResult useOn(UseOnContext p_186371_) {
+      Level level = p_186371_.getLevel();
+      BlockPos blockpos = p_186371_.getClickedPos();
+      BlockState blockstate = level.getBlockState(blockpos);
+      Block block = blockstate.getBlock();
+      if (block instanceof GrowingPlantHeadBlock growingplantheadblock) {
+         if (!growingplantheadblock.isMaxAge(blockstate)) {
             Player player = p_186371_.getPlayer();
             ItemStack itemstack = p_186371_.getItemInHand();
             if (player instanceof ServerPlayer) {
-                CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger((ServerPlayer)player, blockpos, itemstack);
+               CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger((ServerPlayer)player, blockpos, itemstack);
             }
 
             level.playSound(player, blockpos, SoundEvents.GROWING_PLANT_CROP, SoundSource.BLOCKS, 1.0F, 1.0F);
@@ -71,12 +_,15 @@
             level.setBlockAndUpdate(blockpos, blockstate1);
             level.gameEvent(GameEvent.BLOCK_CHANGE, blockpos, GameEvent.Context.of(p_186371_.getPlayer(), blockstate1));
             if (player != null) {
-                itemstack.hurtAndBreak(1, player, LivingEntity.getSlotForHand(p_186371_.getHand()));
+               itemstack.hurtAndBreak(1, player, (p_186374_) -> {
+                  p_186374_.broadcastBreakEvent(p_186371_.getHand());
+               });
             }
 
             return InteractionResult.sidedSuccess(level.isClientSide);
-        }
+         }
+      }
 
-        return super.useOn(p_186371_);
-    }
+      return super.useOn(p_186371_);
+   }
 }
