--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -7,11 +_,13 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SignItem extends StandingAndWallBlockItem {
-    public SignItem(Item.Properties p_43126_, Block p_43127_, Block p_43128_) {
+   public static BlockPos openSign; // CraftBukkit
+   public SignItem(Item.Properties p_43126_, Block p_43127_, Block p_43128_) {
         super(p_43127_, p_43128_, p_43126_, Direction.DOWN);
     }
 
