--- a/net/minecraft/world/item/Tier.java
+++ b/net/minecraft/world/item/Tier.java
@@ -1,25 +_,20 @@
 package net.minecraft.world.item;
 
-import java.util.List;
-import net.minecraft.tags.TagKey;
-import net.minecraft.world.item.component.Tool;
 import net.minecraft.world.item.crafting.Ingredient;
-import net.minecraft.world.level.block.Block;
 
 public interface Tier {
-    int getUses();
-
-    float getSpeed();
-
-    float getAttackDamageBonus();
-
-    TagKey<Block> getIncorrectBlocksForDrops();
-
-    int getEnchantmentValue();
-
-    Ingredient getRepairIngredient();
-
-    default Tool createToolProperties(TagKey<Block> p_331434_) {
-        return new Tool(List.of(Tool.Rule.deniesDrops(this.getIncorrectBlocksForDrops()), Tool.Rule.minesAndDrops(p_331434_, this.getSpeed())), 1.0F, 1);
-    }
+   int getUses();
+
+   float getSpeed();
+
+   float getAttackDamageBonus();
+
+   @Deprecated // FORGE: Use TierSortingRegistry to define which tiers are better than others
+   int getLevel();
+
+   int getEnchantmentValue();
+
+   Ingredient getRepairIngredient();
+
+   @org.jetbrains.annotations.Nullable default net.minecraft.tags.TagKey<net.minecraft.world.level.block.Block> getTag() { return null; }
 }
