--- ../src-base/minecraft/net/minecraft/command/CommandSenderWrapper.java
+++ ../src-work/minecraft/net/minecraft/command/CommandSenderWrapper.java
@@ -11,7 +11,7 @@
 
 public class CommandSenderWrapper implements ICommandSender
 {
-    private final ICommandSender field_193043_a;
+    public final ICommandSender field_193043_a;
     @Nullable
     private final Vec3d field_194002_b;
     @Nullable
