--- a/net/minecraft/command/FunctionObject.java
+++ b/net/minecraft/command/FunctionObject.java
@@ -15,12 +_,22 @@
 public class FunctionObject {
    private final FunctionObject.IEntry[] field_193530_b;
    private final ResourceLocation field_197002_b;
+   // Paper start
+    public co.aikar.timings.Timing timing;
+    public co.aikar.timings.Timing getTiming() {
+       if (timing == null) {
+          timing = co.aikar.timings.MinecraftTimings.getCommandFunctionTiming(this);
+       }
+       return timing;
+    }
+    // Paper end
 
    public FunctionObject(ResourceLocation p_i47973_1_, FunctionObject.IEntry[] p_i47973_2_) {
       this.field_197002_b = p_i47973_1_;
       this.field_193530_b = p_i47973_2_;
    }
 
+   public final ResourceLocation getMinecraftKey() { return this.func_197001_a(); } // Paper - OBFHELPER
    public ResourceLocation func_197001_a() {
       return this.field_197002_b;
    }
