--- ../src-base/minecraft/net/minecraftforge/fml/common/launcher/FMLServerTweaker.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/launcher/FMLServerTweaker.java
@@ -37,7 +37,7 @@
 
         if (System.getProperty("log4j.configurationFile") == null)
         {
-            System.setProperty("log4j.configurationFile", "log4j2_server.xml");
+            System.setProperty("log4j.configurationFile", "log4j2_mohist.xml");
             ((LoggerContext) LogManager.getContext(false)).reconfigure();
         }
     }
@@ -57,7 +57,7 @@
 
         classLoader.addClassLoaderExclusion("org.jline.");
         classLoader.addClassLoaderExclusion("com.sun.jna.");
-        classLoader.addClassLoaderExclusion("net.minecraftforge.server.terminalconsole.");
+        classLoader.addClassLoaderExclusion("red.mohist.console.log4j.TerminalConsoleAppender");
 
         FMLLaunchHandler.configureForServerLaunch(classLoader, this);
         FMLLaunchHandler.appendCoreMods();
