--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityStructure.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityStructure.java
@@ -36,20 +36,20 @@
 
 public class TileEntityStructure extends TileEntity
 {
-    private String name = "";
-    private String author = "";
-    private String metadata = "";
-    private BlockPos position = new BlockPos(0, 1, 0);
-    private BlockPos size = BlockPos.ORIGIN;
-    private Mirror mirror = Mirror.NONE;
-    private Rotation rotation = Rotation.NONE;
-    private TileEntityStructure.Mode mode = TileEntityStructure.Mode.DATA;
-    private boolean ignoreEntities = true;
-    private boolean powered;
-    private boolean showAir;
-    private boolean showBoundingBox = true;
-    private float integrity = 1.0F;
-    private long seed;
+    public String name = "";
+    public String author = "";
+    public String metadata = "";
+    public BlockPos position = new BlockPos(0, 1, 0);
+    public BlockPos size = BlockPos.ORIGIN;
+    public Mirror mirror = Mirror.NONE;
+    public Rotation rotation = Rotation.NONE;
+    public Mode mode = Mode.DATA;
+    public boolean ignoreEntities = true;
+    public boolean powered;
+    public boolean showAir;
+    public boolean showBoundingBox = true;
+    public float integrity = 1.0F;
+    public long seed;
 
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
