--- ../src-base/minecraft/net/minecraftforge/fml/common/network/internal/FMLNetworkHandler.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/network/internal/FMLNetworkHandler.java
@@ -33,9 +33,13 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.inventory.Container;
+import net.minecraft.inventory.IInventory;
 import net.minecraft.network.NetworkManager;
 import net.minecraft.network.Packet;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.management.PlayerList;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.FakePlayer;
 import net.minecraftforge.fml.common.FMLCommonHandler;
@@ -58,6 +62,12 @@
 import com.google.gson.JsonArray;
 import com.google.gson.JsonObject;
 import org.apache.commons.lang3.tuple.Pair;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView;
+import org.bukkit.event.inventory.InventoryType;
+import red.mohist.MohistConfig;
 
 import javax.annotation.Nullable;
 
@@ -88,6 +98,25 @@
             Container remoteGuiContainer = NetworkRegistry.INSTANCE.getRemoteGuiContainer(mc, entityPlayerMP, modGuiId, world, x, y, z);
             if (remoteGuiContainer != null)
             {
+                // Cauldron start - create bukkitView for passed container then fire open event.
+                if (remoteGuiContainer.getBukkitView() == null)
+                {
+                    TileEntity tileEntity = entityPlayer.world.getTileEntity(new BlockPos(x, y, z));
+                    if (tileEntity instanceof IInventory)
+                    {
+                        CraftInventory inventory = new CraftInventory((IInventory)tileEntity);
+                        remoteGuiContainer.bukkitView = new CraftInventoryView(entityPlayer.getBukkitEntity(), inventory, remoteGuiContainer);
+                    }
+                    else
+                    {
+                        remoteGuiContainer.bukkitView = new CraftInventoryView(entityPlayer.getBukkitEntity(), Bukkit.createInventory(entityPlayer.getBukkitEntity(), InventoryType.CHEST), remoteGuiContainer);
+                    }
+                }
+                remoteGuiContainer = CraftEventFactory.callInventoryOpenEvent((EntityPlayerMP) entityPlayer, remoteGuiContainer, false);
+                if (remoteGuiContainer == null) {
+                    return;
+                }
+                // Cauldron end
                 entityPlayerMP.getNextWindowId();
                 entityPlayerMP.closeContainer();
                 int windowId = entityPlayerMP.currentWindowId;
@@ -210,7 +239,7 @@
     public static void enhanceStatusQuery(JsonObject jsonobject)
     {
         JsonObject fmlData = new JsonObject();
-        fmlData.addProperty("type", "FML");
+        fmlData.addProperty("type", MohistConfig.server_type);
         JsonArray modList = new JsonArray();
         for (ModContainer mc : Loader.instance().getActiveModList())
         {
