--- ../src-base/minecraft/net/minecraftforge/fml/common/network/internal/NetworkModHolder.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/network/internal/NetworkModHolder.java
@@ -251,7 +251,9 @@
             }
             catch (Exception e)
             {
-                FMLLog.log.warn("The declared version check handler method {} on network mod id {} is not accessible", networkCheckHandlerMethod, container.getModId(), e);
+                if(!"checkModLists".equals(networkCheckHandlerMethod)) {
+                    FMLLog.log.warn("The declared version check handler method {} on network mod id {} is not accessible", networkCheckHandlerMethod, container.getModId(), e);
+                }
             }
         }
         if (this.checkHandler != null)
