--- ../src-base/minecraft/org/bukkit/enchantments/EnchantmentTarget.java
+++ ../src-work/minecraft/org/bukkit/enchantments/EnchantmentTarget.java
@@ -24,9 +24,9 @@
         @Override
         public boolean includes(Material item) {
             return ARMOR_FEET.includes(item)
-                || ARMOR_LEGS.includes(item)
-                || ARMOR_HEAD.includes(item)
-                || ARMOR_TORSO.includes(item);
+                    || ARMOR_LEGS.includes(item)
+                    || ARMOR_HEAD.includes(item)
+                    || ARMOR_TORSO.includes(item);
         }
     },
 
@@ -37,10 +37,10 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.LEATHER_BOOTS)
-                || item.equals(Material.CHAINMAIL_BOOTS)
-                || item.equals(Material.IRON_BOOTS)
-                || item.equals(Material.DIAMOND_BOOTS)
-                || item.equals(Material.GOLD_BOOTS);
+                    || item.equals(Material.CHAINMAIL_BOOTS)
+                    || item.equals(Material.IRON_BOOTS)
+                    || item.equals(Material.DIAMOND_BOOTS)
+                    || item.equals(Material.GOLD_BOOTS);
         }
     },
 
@@ -51,10 +51,10 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.LEATHER_LEGGINGS)
-                || item.equals(Material.CHAINMAIL_LEGGINGS)
-                || item.equals(Material.IRON_LEGGINGS)
-                || item.equals(Material.DIAMOND_LEGGINGS)
-                || item.equals(Material.GOLD_LEGGINGS);
+                    || item.equals(Material.CHAINMAIL_LEGGINGS)
+                    || item.equals(Material.IRON_LEGGINGS)
+                    || item.equals(Material.DIAMOND_LEGGINGS)
+                    || item.equals(Material.GOLD_LEGGINGS);
         }
     },
 
@@ -65,10 +65,10 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.LEATHER_CHESTPLATE)
-                || item.equals(Material.CHAINMAIL_CHESTPLATE)
-                || item.equals(Material.IRON_CHESTPLATE)
-                || item.equals(Material.DIAMOND_CHESTPLATE)
-                || item.equals(Material.GOLD_CHESTPLATE);
+                    || item.equals(Material.CHAINMAIL_CHESTPLATE)
+                    || item.equals(Material.IRON_CHESTPLATE)
+                    || item.equals(Material.DIAMOND_CHESTPLATE)
+                    || item.equals(Material.GOLD_CHESTPLATE);
         }
     },
 
@@ -79,10 +79,10 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.LEATHER_HELMET)
-                || item.equals(Material.CHAINMAIL_HELMET)
-                || item.equals(Material.DIAMOND_HELMET)
-                || item.equals(Material.IRON_HELMET)
-                || item.equals(Material.GOLD_HELMET);
+                    || item.equals(Material.CHAINMAIL_HELMET)
+                    || item.equals(Material.DIAMOND_HELMET)
+                    || item.equals(Material.IRON_HELMET)
+                    || item.equals(Material.GOLD_HELMET);
         }
     },
 
@@ -93,10 +93,10 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.WOOD_SWORD)
-                || item.equals(Material.STONE_SWORD)
-                || item.equals(Material.IRON_SWORD)
-                || item.equals(Material.DIAMOND_SWORD)
-                || item.equals(Material.GOLD_SWORD);
+                    || item.equals(Material.STONE_SWORD)
+                    || item.equals(Material.IRON_SWORD)
+                    || item.equals(Material.DIAMOND_SWORD)
+                    || item.equals(Material.GOLD_SWORD);
         }
     },
 
@@ -108,27 +108,27 @@
         @Override
         public boolean includes(Material item) {
             return item.equals(Material.WOOD_SPADE)
-                || item.equals(Material.STONE_SPADE)
-                || item.equals(Material.IRON_SPADE)
-                || item.equals(Material.DIAMOND_SPADE)
-                || item.equals(Material.GOLD_SPADE)
-                || item.equals(Material.WOOD_PICKAXE)
-                || item.equals(Material.STONE_PICKAXE)
-                || item.equals(Material.IRON_PICKAXE)
-                || item.equals(Material.DIAMOND_PICKAXE)
-                || item.equals(Material.GOLD_PICKAXE)
-                || item.equals(Material.WOOD_HOE)
-                || item.equals(Material.STONE_HOE)
-                || item.equals(Material.IRON_HOE)
-                || item.equals(Material.DIAMOND_HOE)
-                || item.equals(Material.GOLD_HOE)
-                || item.equals(Material.WOOD_AXE)
-                || item.equals(Material.STONE_AXE)
-                || item.equals(Material.IRON_AXE)
-                || item.equals(Material.DIAMOND_AXE)
-                || item.equals(Material.GOLD_AXE)
-                || item.equals(Material.SHEARS)
-                || item.equals(Material.FLINT_AND_STEEL);
+                    || item.equals(Material.STONE_SPADE)
+                    || item.equals(Material.IRON_SPADE)
+                    || item.equals(Material.DIAMOND_SPADE)
+                    || item.equals(Material.GOLD_SPADE)
+                    || item.equals(Material.WOOD_PICKAXE)
+                    || item.equals(Material.STONE_PICKAXE)
+                    || item.equals(Material.IRON_PICKAXE)
+                    || item.equals(Material.DIAMOND_PICKAXE)
+                    || item.equals(Material.GOLD_PICKAXE)
+                    || item.equals(Material.WOOD_HOE)
+                    || item.equals(Material.STONE_HOE)
+                    || item.equals(Material.IRON_HOE)
+                    || item.equals(Material.DIAMOND_HOE)
+                    || item.equals(Material.GOLD_HOE)
+                    || item.equals(Material.WOOD_AXE)
+                    || item.equals(Material.STONE_AXE)
+                    || item.equals(Material.IRON_AXE)
+                    || item.equals(Material.DIAMOND_AXE)
+                    || item.equals(Material.GOLD_AXE)
+                    || item.equals(Material.SHEARS)
+                    || item.equals(Material.FLINT_AND_STEEL);
         }
     },
 
