--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -128,9 +_,16 @@
         if (!p_49263_.fireImmune()) {
             p_49263_.setRemainingFireTicks(p_49263_.getRemainingFireTicks() + 1);
             if (p_49263_.getRemainingFireTicks() == 0) {
-                p_49263_.igniteForSeconds(8.0F);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(org.bukkit.craftbukkit.block.CraftBlock.at(p_49261_, p_49262_), p_49263_.getBukkitEntity(), 8);
+            p_49261_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+               p_49263_.igniteForSeconds(event.getDuration(), false);
             }
-        }
+            // CraftBukkit end
+         }
+      }
 
         p_49263_.hurt(p_49261_.damageSources().inFire(), this.fireDamage);
         super.entityInside(p_49260_, p_49261_, p_49262_, p_49263_);
@@ -141,6 +_,7 @@
         if (!p_49282_.is(p_49279_.getBlock())) {
             if (inPortalDimension(p_49280_)) {
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(p_49280_, p_49281_, Direction.Axis.X);
+                optional = net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_49280_, p_49281_, optional);
                 if (optional.isPresent()) {
                     optional.get().createPortalBlocks();
                     return;
@@ -183,7 +_,7 @@
             boolean flag = false;
 
             for (Direction direction : Direction.values()) {
-                if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).is(Blocks.OBSIDIAN)) {
+                if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).isPortalFrame(p_49270_, blockpos$mutableblockpos)) {
                     flag = true;
                     break;
                 }
@@ -199,4 +_,12 @@
             }
         }
     }
+
+   	// CraftBukkit start
+   	protected void fireExtinguished(net.minecraft.world.level.LevelAccessor world, BlockPos position) {
+      	if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, position, Blocks.AIR.defaultBlockState()).isCancelled()) {
+         	world.removeBlock(position, false);
+      	}
+   	}
+   	// CraftBukkit end
 }
