--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.portal.PortalShape;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
@@ -112,7 +_,14 @@
       if (!p_49263_.fireImmune()) {
          p_49263_.setRemainingFireTicks(p_49263_.getRemainingFireTicks() + 1);
          if (p_49263_.getRemainingFireTicks() == 0) {
-            p_49263_.setSecondsOnFire(8);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(org.bukkit.craftbukkit.v1_20_R2.block.CraftBlock.at(p_49261_, p_49262_), p_49263_.getBukkitEntity(), 8);
+            p_49261_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+               p_49263_.setSecondsOnFire(event.getDuration(), false);
+            }
+            // CraftBukkit end
          }
       }
 
@@ -124,6 +_,7 @@
       if (!p_49282_.is(p_49279_.getBlock())) {
          if (inPortalDimension(p_49280_)) {
             Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(p_49280_, p_49281_, Direction.Axis.X);
+            optional = net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_49280_, p_49281_, optional);
             if (optional.isPresent()) {
                optional.get().createPortalBlocks();
                return;
@@ -131,14 +_,14 @@
          }
 
          if (!p_49279_.canSurvive(p_49280_, p_49281_)) {
-            p_49280_.removeBlock(p_49281_, false);
+            fireExtinguished(p_49280_, p_49281_);
          }
 
       }
    }
 
    private static boolean inPortalDimension(Level p_49249_) {
-      return p_49249_.dimension() == Level.OVERWORLD || p_49249_.dimension() == Level.NETHER;
+      return p_49249_.getTypeKey() == LevelStem.OVERWORLD || p_49249_.getTypeKey() == LevelStem.NETHER;
    }
 
    protected void spawnDestroyParticles(Level p_152139_, Player p_152140_, BlockPos p_152141_, BlockState p_152142_) {
@@ -169,7 +_,7 @@
          boolean flag = false;
 
          for(Direction direction : Direction.values()) {
-            if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).is(Blocks.OBSIDIAN)) {
+            if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).isPortalFrame(p_49270_, blockpos$mutableblockpos)) {
                flag = true;
                break;
             }
@@ -183,4 +_,12 @@
          }
       }
    }
+
+   // CraftBukkit start
+   protected void fireExtinguished(net.minecraft.world.level.LevelAccessor world, BlockPos position) {
+      if (!org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callBlockFadeEvent(world, position, Blocks.AIR.defaultBlockState()).isCancelled()) {
+         world.removeBlock(position, false);
+      }
+   }
+   // CraftBukkit end
 }
