--- a/net/minecraft/world/level/block/BeetrootBlock.java
+++ b/net/minecraft/world/level/block/BeetrootBlock.java
@@ -25,12 +_,11 @@
    public MapCodec<BeetrootBlock> codec() {
       return CODEC;
    }
-
    public BeetrootBlock(BlockBehaviour.Properties p_49661_) {
       super(p_49661_);
    }
 
-   protected IntegerProperty getAgeProperty() {
+   public IntegerProperty getAgeProperty() {
       return AGE;
    }
 
