--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -247,4 +_,16 @@
    public boolean isPathfindable(BlockState p_49717_, BlockGetter p_49718_, BlockPos p_49719_, PathComputationType p_49720_) {
       return false;
    }
+
+   // CraftBukkit start - fix MC-253819
+   @Override
+   public BlockState rotate(BlockState iblockdata, Rotation enumblockrotation) {
+      return iblockdata.setValue(BellBlock.FACING, enumblockrotation.rotate(iblockdata.getValue(BellBlock.FACING)));
+   }
+
+   @Override
+   public BlockState mirror(BlockState iblockdata, Mirror enumblockmirror) {
+      return iblockdata.rotate(enumblockmirror.getRotation(iblockdata.getValue(BellBlock.FACING)));
+   }
+   // CraftBukkit end
 }
