--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -39,6 +_,9 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class BigDripleafBlock extends HorizontalDirectionalBlock implements BonemealableBlock, SimpleWaterloggedBlock {
    private static final BooleanProperty f_152200_ = BlockStateProperties.f_61362_;
@@ -104,7 +_,7 @@
    }
 
    public void m_5581_(Level p_152228_, BlockState p_152229_, BlockHitResult p_152230_, Projectile p_152231_) {
-      this.m_152282_(p_152229_, p_152228_, p_152230_.m_82425_(), Tilt.FULL, SoundEvents.f_144131_);
+      this.setTiltAndScheduleTick(p_152229_, p_152228_, p_152230_.m_82425_(), Tilt.FULL, SoundEvents.f_144131_, p_152231_); // CraftBukkit
    }
 
    public FluidState m_5888_(BlockState p_152312_) {
@@ -152,7 +_,20 @@
    public void m_7892_(BlockState p_152266_, Level p_152267_, BlockPos p_152268_, Entity p_152269_) {
       if (!p_152267_.f_46443_) {
          if (p_152266_.m_61143_(f_152201_) == Tilt.NONE && m_152301_(p_152268_, p_152269_) && !p_152267_.m_276867_(p_152268_)) {
-            this.m_152282_(p_152266_, p_152267_, p_152268_, Tilt.UNSTABLE, (SoundEvent)null);
+            // CraftBukkit start - tilt dripleaf
+            org.bukkit.event.Cancellable cancellable;
+            if (p_152269_ instanceof Player) {
+               cancellable = CraftEventFactory.callPlayerInteractEvent((Player) p_152269_, org.bukkit.event.block.Action.PHYSICAL, p_152268_, null, null, null);
+            } else {
+               cancellable = new EntityInteractEvent(p_152269_.getBukkitEntity(), CraftBlock.at(p_152267_, p_152268_));
+               p_152267_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled()) {
+               return;
+            }
+            this.setTiltAndScheduleTick(p_152266_, p_152267_, p_152268_, Tilt.UNSTABLE, (SoundEvent) null, p_152269_);
+            // CraftBukkit end
          }
 
       }
@@ -164,9 +_,9 @@
       } else {
          Tilt tilt = p_220798_.m_61143_(f_152201_);
          if (tilt == Tilt.UNSTABLE) {
-            this.m_152282_(p_220798_, p_220799_, p_220800_, Tilt.PARTIAL, SoundEvents.f_144131_);
+            this.setTiltAndScheduleTick(p_220798_, p_220799_, p_220800_, Tilt.PARTIAL, SoundEvents.f_144131_, null); // CraftBukkit
          } else if (tilt == Tilt.PARTIAL) {
-            this.m_152282_(p_220798_, p_220799_, p_220800_, Tilt.FULL, SoundEvents.f_144131_);
+            this.setTiltAndScheduleTick(p_220798_, p_220799_, p_220800_, Tilt.FULL, SoundEvents.f_144131_, null); // CraftBukkit
          } else if (tilt == Tilt.FULL) {
             m_152313_(p_220798_, p_220799_, p_220800_);
          }
@@ -190,34 +_,41 @@
       return p_152303_.m_20096_() && p_152303_.m_20182_().f_82480_ > (double)((float)p_152302_.m_123342_() + 0.6875F);
    }
 
-   private void m_152282_(BlockState p_152283_, Level p_152284_, BlockPos p_152285_, Tilt p_152286_, @Nullable SoundEvent p_152287_) {
-      m_152277_(p_152283_, p_152284_, p_152285_, p_152286_);
-      if (p_152287_ != null) {
-         m_152232_(p_152284_, p_152285_, p_152287_);
+   private void setTiltAndScheduleTick(BlockState pState, Level pLevel, BlockPos pPos, Tilt pTilt, @Nullable SoundEvent pSound, @Nullable Entity entity) {
+      if (!setTilt(pState, pLevel, pPos, pTilt, entity)) return;
+      if (pSound != null) {
+         m_152232_(pLevel, pPos, pSound);
       }
 
-      int i = f_152203_.getInt(p_152286_);
+      int i = f_152203_.getInt(pTilt);
       if (i != -1) {
-         p_152284_.m_186460_(p_152285_, this, i);
+         pLevel.m_186460_(pPos, this, i);
       }
 
    }
 
    private static void m_152313_(BlockState p_152314_, Level p_152315_, BlockPos p_152316_) {
-      m_152277_(p_152314_, p_152315_, p_152316_, Tilt.NONE);
+      setTilt(p_152314_, p_152315_, p_152316_, Tilt.NONE, null);
       if (p_152314_.m_61143_(f_152201_) != Tilt.NONE) {
          m_152232_(p_152315_, p_152316_, SoundEvents.f_144132_);
       }
 
    }
 
-   private static void m_152277_(BlockState p_152278_, Level p_152279_, BlockPos p_152280_, Tilt p_152281_) {
-      Tilt tilt = p_152278_.m_61143_(f_152201_);
-      p_152279_.m_7731_(p_152280_, p_152278_.m_61124_(f_152201_, p_152281_), 2);
-      if (p_152281_.m_156084_() && p_152281_ != tilt) {
-         p_152279_.m_142346_((Entity)null, GameEvent.f_157792_, p_152280_);
+   private static boolean setTilt(BlockState pState, Level pLevel, BlockPos pPos, Tilt pTilt, @Nullable Entity entity) {
+      if (entity != null) {
+         if (!CraftEventFactory.callEntityChangeBlockEvent(entity, pPos, pState.m_61124_(BigDripleafBlock.f_152201_, pTilt))) {
+            return false;
+         }
+      }
+      // CraftBukkit end
+      Tilt tilt = pState.m_61143_(f_152201_);
+      pLevel.m_7731_(pPos, pState.m_61124_(f_152201_, pTilt), 2);
+      if (pTilt.m_156084_() && pTilt != tilt) {
+         pLevel.m_142346_((Entity)null, GameEvent.f_157792_, pPos);
       }
 
+      return true; // CraftBukkit
    }
 
    public VoxelShape m_5939_(BlockState p_152307_, BlockGetter p_152308_, BlockPos p_152309_, CollisionContext p_152310_) {
