--- a/net/minecraft/world/level/block/ButtonBlock.java
+++ b/net/minecraft/world/level/block/ButtonBlock.java
@@ -3,7 +_,6 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.function.BiConsumer;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -12,6 +_,7 @@
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.util.RandomSource;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
@@ -32,182 +_,224 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class ButtonBlock extends FaceAttachedHorizontalDirectionalBlock {
-    public static final MapCodec<ButtonBlock> CODEC = RecordCodecBuilder.mapCodec(
-        p_310359_ -> p_310359_.group(
-                    BlockSetType.CODEC.fieldOf("block_set_type").forGetter(p_312681_ -> p_312681_.type),
-                    Codec.intRange(1, 1024).fieldOf("ticks_to_stay_pressed").forGetter(p_312686_ -> p_312686_.ticksToStayPressed),
-                    propertiesCodec()
-                )
-                .apply(p_310359_, ButtonBlock::new)
-    );
-    public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
-    private static final int PRESSED_DEPTH = 1;
-    private static final int UNPRESSED_DEPTH = 2;
-    protected static final int HALF_AABB_HEIGHT = 2;
-    protected static final int HALF_AABB_WIDTH = 3;
-    protected static final VoxelShape CEILING_AABB_X = Block.box(6.0, 14.0, 5.0, 10.0, 16.0, 11.0);
-    protected static final VoxelShape CEILING_AABB_Z = Block.box(5.0, 14.0, 6.0, 11.0, 16.0, 10.0);
-    protected static final VoxelShape FLOOR_AABB_X = Block.box(6.0, 0.0, 5.0, 10.0, 2.0, 11.0);
-    protected static final VoxelShape FLOOR_AABB_Z = Block.box(5.0, 0.0, 6.0, 11.0, 2.0, 10.0);
-    protected static final VoxelShape NORTH_AABB = Block.box(5.0, 6.0, 14.0, 11.0, 10.0, 16.0);
-    protected static final VoxelShape SOUTH_AABB = Block.box(5.0, 6.0, 0.0, 11.0, 10.0, 2.0);
-    protected static final VoxelShape WEST_AABB = Block.box(14.0, 6.0, 5.0, 16.0, 10.0, 11.0);
-    protected static final VoxelShape EAST_AABB = Block.box(0.0, 6.0, 5.0, 2.0, 10.0, 11.0);
-    protected static final VoxelShape PRESSED_CEILING_AABB_X = Block.box(6.0, 15.0, 5.0, 10.0, 16.0, 11.0);
-    protected static final VoxelShape PRESSED_CEILING_AABB_Z = Block.box(5.0, 15.0, 6.0, 11.0, 16.0, 10.0);
-    protected static final VoxelShape PRESSED_FLOOR_AABB_X = Block.box(6.0, 0.0, 5.0, 10.0, 1.0, 11.0);
-    protected static final VoxelShape PRESSED_FLOOR_AABB_Z = Block.box(5.0, 0.0, 6.0, 11.0, 1.0, 10.0);
-    protected static final VoxelShape PRESSED_NORTH_AABB = Block.box(5.0, 6.0, 15.0, 11.0, 10.0, 16.0);
-    protected static final VoxelShape PRESSED_SOUTH_AABB = Block.box(5.0, 6.0, 0.0, 11.0, 10.0, 1.0);
-    protected static final VoxelShape PRESSED_WEST_AABB = Block.box(15.0, 6.0, 5.0, 16.0, 10.0, 11.0);
-    protected static final VoxelShape PRESSED_EAST_AABB = Block.box(0.0, 6.0, 5.0, 1.0, 10.0, 11.0);
-    private final BlockSetType type;
-    private final int ticksToStayPressed;
-
-    @Override
-    public MapCodec<ButtonBlock> codec() {
-        return CODEC;
-    }
-
-    public ButtonBlock(BlockSetType p_273462_, int p_273212_, BlockBehaviour.Properties p_273290_) {
-        super(p_273290_.sound(p_273462_.soundType()));
-        this.type = p_273462_;
-        this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, Boolean.valueOf(false)).setValue(FACE, AttachFace.WALL)
-        );
-        this.ticksToStayPressed = p_273212_;
-    }
-
-    @Override
-    protected VoxelShape getShape(BlockState p_51104_, BlockGetter p_51105_, BlockPos p_51106_, CollisionContext p_51107_) {
-        Direction direction = p_51104_.getValue(FACING);
-        boolean flag = p_51104_.getValue(POWERED);
-        switch ((AttachFace)p_51104_.getValue(FACE)) {
-            case FLOOR:
-                if (direction.getAxis() == Direction.Axis.X) {
-                    return flag ? PRESSED_FLOOR_AABB_X : FLOOR_AABB_X;
-                }
-
-                return flag ? PRESSED_FLOOR_AABB_Z : FLOOR_AABB_Z;
-            case WALL:
-                return switch (direction) {
-                    case EAST -> flag ? PRESSED_EAST_AABB : EAST_AABB;
-                    case WEST -> flag ? PRESSED_WEST_AABB : WEST_AABB;
-                    case SOUTH -> flag ? PRESSED_SOUTH_AABB : SOUTH_AABB;
-                    case NORTH, UP, DOWN -> flag ? PRESSED_NORTH_AABB : NORTH_AABB;
-                };
-            case CEILING:
-            default:
-                if (direction.getAxis() == Direction.Axis.X) {
-                    return flag ? PRESSED_CEILING_AABB_X : CEILING_AABB_X;
-                } else {
-                    return flag ? PRESSED_CEILING_AABB_Z : CEILING_AABB_Z;
-                }
-        }
-    }
-
-    @Override
-    protected InteractionResult useWithoutItem(BlockState p_329418_, Level p_334611_, BlockPos p_332004_, Player p_330636_, BlockHitResult p_327724_) {
-        if (p_329418_.getValue(POWERED)) {
-            return InteractionResult.CONSUME;
-        } else {
-            this.press(p_329418_, p_334611_, p_332004_);
-            this.playSound(p_330636_, p_334611_, p_332004_, true);
-            p_334611_.gameEvent(p_330636_, GameEvent.BLOCK_ACTIVATE, p_332004_);
-            return InteractionResult.sidedSuccess(p_334611_.isClientSide);
-        }
-    }
-
-    @Override
-    protected void onExplosionHit(BlockState p_310762_, Level p_312485_, BlockPos p_312982_, Explosion p_311820_, BiConsumer<ItemStack, BlockPos> p_312672_) {
-        if (p_311820_.getBlockInteraction() == Explosion.BlockInteraction.TRIGGER_BLOCK && !p_312485_.isClientSide() && !p_310762_.getValue(POWERED)) {
-            this.press(p_310762_, p_312485_, p_312982_);
-        }
-
-        super.onExplosionHit(p_310762_, p_312485_, p_312982_, p_311820_, p_312672_);
-    }
-
-    public void press(BlockState p_51117_, Level p_51118_, BlockPos p_51119_) {
-        p_51118_.setBlock(p_51119_, p_51117_.setValue(POWERED, Boolean.valueOf(true)), 3);
-        this.updateNeighbours(p_51117_, p_51118_, p_51119_);
-        p_51118_.scheduleTick(p_51119_, this, this.ticksToStayPressed);
-    }
-
-    protected void playSound(@Nullable Player p_51068_, LevelAccessor p_51069_, BlockPos p_51070_, boolean p_51071_) {
-        p_51069_.playSound(p_51071_ ? p_51068_ : null, p_51070_, this.getSound(p_51071_), SoundSource.BLOCKS);
-    }
-
-    protected SoundEvent getSound(boolean p_51102_) {
-        return p_51102_ ? this.type.buttonClickOn() : this.type.buttonClickOff();
-    }
-
-    @Override
-    protected void onRemove(BlockState p_51095_, Level p_51096_, BlockPos p_51097_, BlockState p_51098_, boolean p_51099_) {
-        if (!p_51099_ && !p_51095_.is(p_51098_.getBlock())) {
-            if (p_51095_.getValue(POWERED)) {
-                this.updateNeighbours(p_51095_, p_51096_, p_51097_);
-            }
-
-            super.onRemove(p_51095_, p_51096_, p_51097_, p_51098_, p_51099_);
-        }
-    }
-
-    @Override
-    protected int getSignal(BlockState p_51078_, BlockGetter p_51079_, BlockPos p_51080_, Direction p_51081_) {
-        return p_51078_.getValue(POWERED) ? 15 : 0;
-    }
-
-    @Override
-    protected int getDirectSignal(BlockState p_51109_, BlockGetter p_51110_, BlockPos p_51111_, Direction p_51112_) {
-        return p_51109_.getValue(POWERED) && getConnectedDirection(p_51109_) == p_51112_ ? 15 : 0;
-    }
-
-    @Override
-    protected boolean isSignalSource(BlockState p_51114_) {
-        return true;
-    }
-
-    @Override
-    protected void tick(BlockState p_220903_, ServerLevel p_220904_, BlockPos p_220905_, RandomSource p_220906_) {
-        if (p_220903_.getValue(POWERED)) {
-            this.checkPressed(p_220903_, p_220904_, p_220905_);
-        }
-    }
-
-    @Override
-    protected void entityInside(BlockState p_51083_, Level p_51084_, BlockPos p_51085_, Entity p_51086_) {
-        if (!p_51084_.isClientSide && this.type.canButtonBeActivatedByArrows() && !p_51083_.getValue(POWERED)) {
-            this.checkPressed(p_51083_, p_51084_, p_51085_);
-        }
-    }
-
-    protected void checkPressed(BlockState p_51121_, Level p_51122_, BlockPos p_51123_) {
-        AbstractArrow abstractarrow = this.type.canButtonBeActivatedByArrows()
-            ? p_51122_.getEntitiesOfClass(AbstractArrow.class, p_51121_.getShape(p_51122_, p_51123_).bounds().move(p_51123_)).stream().findFirst().orElse(null)
-            : null;
-        boolean flag = abstractarrow != null;
-        boolean flag1 = p_51121_.getValue(POWERED);
-        if (flag != flag1) {
-            p_51122_.setBlock(p_51123_, p_51121_.setValue(POWERED, Boolean.valueOf(flag)), 3);
-            this.updateNeighbours(p_51121_, p_51122_, p_51123_);
-            this.playSound(null, p_51122_, p_51123_, flag);
-            p_51122_.gameEvent(abstractarrow, flag ? GameEvent.BLOCK_ACTIVATE : GameEvent.BLOCK_DEACTIVATE, p_51123_);
-        }
-
-        if (flag) {
-            p_51122_.scheduleTick(new BlockPos(p_51123_), this, this.ticksToStayPressed);
-        }
-    }
-
-    private void updateNeighbours(BlockState p_51125_, Level p_51126_, BlockPos p_51127_) {
-        p_51126_.updateNeighborsAt(p_51127_, this);
-        p_51126_.updateNeighborsAt(p_51127_.relative(getConnectedDirection(p_51125_).getOpposite()), this);
-    }
-
-    @Override
-    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> p_51101_) {
-        p_51101_.add(FACING, POWERED, FACE);
-    }
+   public static final MapCodec<ButtonBlock> CODEC = RecordCodecBuilder.mapCodec((p_310359_) -> {
+      return p_310359_.group(BlockSetType.CODEC.fieldOf("block_set_type").forGetter((p_312681_) -> {
+         return p_312681_.type;
+      }), Codec.intRange(1, 1024).fieldOf("ticks_to_stay_pressed").forGetter((p_312686_) -> {
+         return p_312686_.ticksToStayPressed;
+      }), propertiesCodec()).apply(p_310359_, ButtonBlock::new);
+   });
+   public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
+   private static final int PRESSED_DEPTH = 1;
+   private static final int UNPRESSED_DEPTH = 2;
+   protected static final int HALF_AABB_HEIGHT = 2;
+   protected static final int HALF_AABB_WIDTH = 3;
+   protected static final VoxelShape CEILING_AABB_X = Block.box(6.0D, 14.0D, 5.0D, 10.0D, 16.0D, 11.0D);
+   protected static final VoxelShape CEILING_AABB_Z = Block.box(5.0D, 14.0D, 6.0D, 11.0D, 16.0D, 10.0D);
+   protected static final VoxelShape FLOOR_AABB_X = Block.box(6.0D, 0.0D, 5.0D, 10.0D, 2.0D, 11.0D);
+   protected static final VoxelShape FLOOR_AABB_Z = Block.box(5.0D, 0.0D, 6.0D, 11.0D, 2.0D, 10.0D);
+   protected static final VoxelShape NORTH_AABB = Block.box(5.0D, 6.0D, 14.0D, 11.0D, 10.0D, 16.0D);
+   protected static final VoxelShape SOUTH_AABB = Block.box(5.0D, 6.0D, 0.0D, 11.0D, 10.0D, 2.0D);
+   protected static final VoxelShape WEST_AABB = Block.box(14.0D, 6.0D, 5.0D, 16.0D, 10.0D, 11.0D);
+   protected static final VoxelShape EAST_AABB = Block.box(0.0D, 6.0D, 5.0D, 2.0D, 10.0D, 11.0D);
+   protected static final VoxelShape PRESSED_CEILING_AABB_X = Block.box(6.0D, 15.0D, 5.0D, 10.0D, 16.0D, 11.0D);
+   protected static final VoxelShape PRESSED_CEILING_AABB_Z = Block.box(5.0D, 15.0D, 6.0D, 11.0D, 16.0D, 10.0D);
+   protected static final VoxelShape PRESSED_FLOOR_AABB_X = Block.box(6.0D, 0.0D, 5.0D, 10.0D, 1.0D, 11.0D);
+   protected static final VoxelShape PRESSED_FLOOR_AABB_Z = Block.box(5.0D, 0.0D, 6.0D, 11.0D, 1.0D, 10.0D);
+   protected static final VoxelShape PRESSED_NORTH_AABB = Block.box(5.0D, 6.0D, 15.0D, 11.0D, 10.0D, 16.0D);
+   protected static final VoxelShape PRESSED_SOUTH_AABB = Block.box(5.0D, 6.0D, 0.0D, 11.0D, 10.0D, 1.0D);
+   protected static final VoxelShape PRESSED_WEST_AABB = Block.box(15.0D, 6.0D, 5.0D, 16.0D, 10.0D, 11.0D);
+   protected static final VoxelShape PRESSED_EAST_AABB = Block.box(0.0D, 6.0D, 5.0D, 1.0D, 10.0D, 11.0D);
+   private final BlockSetType type;
+   private final int ticksToStayPressed;
+
+   public MapCodec<ButtonBlock> codec() {
+      return CODEC;
+   }
+
+   public ButtonBlock(BlockSetType p_273462_, int p_273212_, BlockBehaviour.Properties p_273290_) {
+      super(p_273290_.sound(p_273462_.soundType()));
+      this.type = p_273462_;
+      this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, Boolean.valueOf(false)).setValue(FACE, AttachFace.WALL));
+      this.ticksToStayPressed = p_273212_;
+   }
+
+   public VoxelShape getShape(BlockState p_51104_, BlockGetter p_51105_, BlockPos p_51106_, CollisionContext p_51107_) {
+      Direction direction = p_51104_.getValue(FACING);
+      boolean flag = p_51104_.getValue(POWERED);
+      switch ((AttachFace)p_51104_.getValue(FACE)) {
+         case FLOOR:
+            if (direction.getAxis() == Direction.Axis.X) {
+               return flag ? PRESSED_FLOOR_AABB_X : FLOOR_AABB_X;
+            }
+
+            return flag ? PRESSED_FLOOR_AABB_Z : FLOOR_AABB_Z;
+         case WALL:
+            VoxelShape voxelshape;
+            switch (direction) {
+               case EAST:
+                  voxelshape = flag ? PRESSED_EAST_AABB : EAST_AABB;
+                  break;
+               case WEST:
+                  voxelshape = flag ? PRESSED_WEST_AABB : WEST_AABB;
+                  break;
+               case SOUTH:
+                  voxelshape = flag ? PRESSED_SOUTH_AABB : SOUTH_AABB;
+                  break;
+               case NORTH:
+               case UP:
+               case DOWN:
+                  voxelshape = flag ? PRESSED_NORTH_AABB : NORTH_AABB;
+                  break;
+               default:
+                  throw new IncompatibleClassChangeError();
+            }
+
+            return voxelshape;
+         case CEILING:
+         default:
+            if (direction.getAxis() == Direction.Axis.X) {
+               return flag ? PRESSED_CEILING_AABB_X : CEILING_AABB_X;
+            } else {
+               return flag ? PRESSED_CEILING_AABB_Z : CEILING_AABB_Z;
+            }
+      }
+   }
+
+   public InteractionResult use(BlockState p_51088_, Level p_51089_, BlockPos p_51090_, Player p_51091_, InteractionHand p_51092_, BlockHitResult p_51093_) {
+      if (p_51088_.getValue(POWERED)) {
+         return InteractionResult.CONSUME;
+      } else {
+         // CraftBukkit start
+         boolean powered = p_51088_.getValue(POWERED);
+         org.bukkit.block.Block block = CraftBlock.at(p_51089_, p_51090_);
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_51089_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
+         this.press(p_51088_, p_51089_, p_51090_);
+         this.playSound(p_51091_, p_51089_, p_51090_, true);
+         p_51089_.gameEvent(p_51091_, GameEvent.BLOCK_ACTIVATE, p_51090_);
+         return InteractionResult.sidedSuccess(p_51089_.isClientSide);
+      }
+   }
+
+   public void onExplosionHit(BlockState p_310762_, Level p_312485_, BlockPos p_312982_, Explosion p_311820_, BiConsumer<ItemStack, BlockPos> p_312672_) {
+      if (p_311820_.getBlockInteraction() == Explosion.BlockInteraction.TRIGGER_BLOCK && !p_312485_.isClientSide() && !p_310762_.getValue(POWERED)) {
+         this.press(p_310762_, p_312485_, p_312982_);
+      }
+
+      super.onExplosionHit(p_310762_, p_312485_, p_312982_, p_311820_, p_312672_);
+   }
+
+   public void press(BlockState p_51117_, Level p_51118_, BlockPos p_51119_) {
+      p_51118_.setBlock(p_51119_, p_51117_.setValue(POWERED, Boolean.valueOf(true)), 3);
+      this.updateNeighbours(p_51117_, p_51118_, p_51119_);
+      p_51118_.scheduleTick(p_51119_, this, this.ticksToStayPressed);
+   }
+
+   protected void playSound(@Nullable Player p_51068_, LevelAccessor p_51069_, BlockPos p_51070_, boolean p_51071_) {
+      p_51069_.playSound(p_51071_ ? p_51068_ : null, p_51070_, this.getSound(p_51071_), SoundSource.BLOCKS);
+   }
+
+   protected SoundEvent getSound(boolean p_51102_) {
+      return p_51102_ ? this.type.buttonClickOn() : this.type.buttonClickOff();
+   }
+
+   public void onRemove(BlockState p_51095_, Level p_51096_, BlockPos p_51097_, BlockState p_51098_, boolean p_51099_) {
+      if (!p_51099_ && !p_51095_.is(p_51098_.getBlock())) {
+         if (p_51095_.getValue(POWERED)) {
+            this.updateNeighbours(p_51095_, p_51096_, p_51097_);
+         }
+
+         super.onRemove(p_51095_, p_51096_, p_51097_, p_51098_, p_51099_);
+      }
+   }
+
+   public int getSignal(BlockState p_51078_, BlockGetter p_51079_, BlockPos p_51080_, Direction p_51081_) {
+      return p_51078_.getValue(POWERED) ? 15 : 0;
+   }
+
+   public int getDirectSignal(BlockState p_51109_, BlockGetter p_51110_, BlockPos p_51111_, Direction p_51112_) {
+      return p_51109_.getValue(POWERED) && getConnectedDirection(p_51109_) == p_51112_ ? 15 : 0;
+   }
+
+   public boolean isSignalSource(BlockState p_51114_) {
+      return true;
+   }
+
+   public void tick(BlockState p_220903_, ServerLevel p_220904_, BlockPos p_220905_, RandomSource p_220906_) {
+      if (p_220903_.getValue(POWERED)) {
+         this.checkPressed(p_220903_, p_220904_, p_220905_);
+      }
+   }
+
+   public void entityInside(BlockState p_51083_, Level p_51084_, BlockPos p_51085_, Entity p_51086_) {
+      if (!p_51084_.isClientSide && this.type.canButtonBeActivatedByArrows() && !p_51083_.getValue(POWERED)) {
+         this.checkPressed(p_51083_, p_51084_, p_51085_);
+      }
+   }
+
+   protected void checkPressed(BlockState p_51121_, Level p_51122_, BlockPos p_51123_) {
+      AbstractArrow abstractarrow = this.type.canButtonBeActivatedByArrows() ? p_51122_.getEntitiesOfClass(AbstractArrow.class, p_51121_.getShape(p_51122_, p_51123_).bounds().move(p_51123_)).stream().findFirst().orElse((AbstractArrow)null) : null;
+      boolean flag = abstractarrow != null;
+      boolean flag1 = p_51121_.getValue(POWERED);
+
+      // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+      if (flag1 != flag && flag) {
+         org.bukkit.block.Block block = CraftBlock.at(p_51122_, p_51123_);
+         EntityInteractEvent event = new EntityInteractEvent(abstractarrow.getBukkitEntity(), block);
+         p_51122_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (event.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
+
+      if (flag != flag1) {
+         // CraftBukkit start
+         boolean powered = flag1;
+         org.bukkit.block.Block block = CraftBlock.at(p_51122_, p_51123_);
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_51122_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((flag && eventRedstone.getNewCurrent() <= 0) || (!flag && eventRedstone.getNewCurrent() > 0)) {
+            return;
+         }
+         // CraftBukkit end
+         p_51122_.setBlock(p_51123_, p_51121_.setValue(POWERED, Boolean.valueOf(flag)), 3);
+         this.updateNeighbours(p_51121_, p_51122_, p_51123_);
+         this.playSound((Player)null, p_51122_, p_51123_, flag);
+         p_51122_.gameEvent(abstractarrow, flag ? GameEvent.BLOCK_ACTIVATE : GameEvent.BLOCK_DEACTIVATE, p_51123_);
+      }
+
+      if (flag) {
+         p_51122_.scheduleTick(new BlockPos(p_51123_), this, this.ticksToStayPressed);
+      }
+
+   }
+
+   private void updateNeighbours(BlockState p_51125_, Level p_51126_, BlockPos p_51127_) {
+      p_51126_.updateNeighborsAt(p_51127_, this);
+      p_51126_.updateNeighborsAt(p_51127_.relative(getConnectedDirection(p_51125_).getOpposite()), this);
+   }
+
+   protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> p_51101_) {
+      p_51101_.add(FACING, POWERED, FACE);
+   }
 }
