--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -3,6 +_,7 @@
 import com.mojang.serialization.MapCodec;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.cauldron.CauldronInteraction;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class CauldronBlock extends AbstractCauldronBlock {
     public static final MapCodec<CauldronBlock> CODEC = simpleCodec(CauldronBlock::new);
@@ -60,14 +_,13 @@
     protected void receiveStalactiteDrip(BlockState p_152940_, Level p_152941_, BlockPos p_152942_, Fluid p_152943_) {
         if (p_152943_ == Fluids.WATER) {
             BlockState blockstate = Blocks.WATER_CAULDRON.defaultBlockState();
-            p_152941_.setBlockAndUpdate(p_152942_, blockstate);
-            p_152941_.gameEvent(GameEvent.BLOCK_CHANGE, p_152942_, GameEvent.Context.of(blockstate));
-            p_152941_.levelEvent(1047, p_152942_, 0);
-        } else if (p_152943_ == Fluids.LAVA) {
-            BlockState blockstate1 = Blocks.LAVA_CAULDRON.defaultBlockState();
-            p_152941_.setBlockAndUpdate(p_152942_, blockstate1);
-            p_152941_.gameEvent(GameEvent.BLOCK_CHANGE, p_152942_, GameEvent.Context.of(blockstate1));
-            p_152941_.levelEvent(1046, p_152942_, 0);
-        }
-    }
+         LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, blockstate, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
+         p_152941_.levelEvent(1047, p_152942_, 0);
+      } else if (p_152943_ == Fluids.LAVA) {
+         BlockState blockstate1 = Blocks.LAVA_CAULDRON.defaultBlockState();
+         LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, blockstate1, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
+         p_152941_.levelEvent(1046, p_152942_, 0);
+      }
+
+   }
 }
