--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -17,7 +_,7 @@
       float f = 0.05688889F;
       if (p_310572_.nextFloat() < 0.05688889F) {
          this.getNextState(p_311790_, p_309416_, p_310092_, p_310572_).ifPresent((p_153039_) -> {
-            p_309416_.setBlockAndUpdate(p_310092_, p_153039_);
+            org.bukkit.craftbukkit.v1_20_R4.event.CraftEventFactory.handleBlockFormEvent(p_309416_, p_310092_, p_153039_); // CraftBukkit
          });
       }
 
