--- a/net/minecraft/world/level/block/ChestBlock.java
+++ b/net/minecraft/world/level/block/ChestBlock.java
@@ -18,14 +_,17 @@
 import net.minecraft.world.CompoundContainer;
 import net.minecraft.world.Container;
 import net.minecraft.world.Containers;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.MenuProvider;
+import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.animal.Cat;
 import net.minecraft.world.entity.monster.piglin.PiglinAi;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ChestMenu;
+import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
@@ -79,29 +_,8 @@
     };
     private static final DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>> MENU_PROVIDER_COMBINER = new DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>>() {
         public Optional<MenuProvider> acceptDouble(final ChestBlockEntity p_51604_, final ChestBlockEntity p_51605_) {
-            final Container container = new CompoundContainer(p_51604_, p_51605_);
-            return Optional.of(new MenuProvider() {
-                @Nullable
-                @Override
-                public AbstractContainerMenu createMenu(int p_51622_, Inventory p_51623_, Player p_51624_) {
-                    if (p_51604_.canOpen(p_51624_) && p_51605_.canOpen(p_51624_)) {
-                        p_51604_.unpackLootTable(p_51623_.player);
-                        p_51605_.unpackLootTable(p_51623_.player);
-                        return ChestMenu.sixRows(p_51622_, p_51623_, container);
-                    } else {
-                        return null;
-                    }
-                }
-
-                @Override
-                public Component getDisplayName() {
-                    if (p_51604_.hasCustomName()) {
-                        return p_51604_.getDisplayName();
-                    } else {
-                        return (Component)(p_51605_.hasCustomName() ? p_51605_.getDisplayName() : Component.translatable("container.chestDouble"));
-                    }
-                }
-            });
+            final CompoundContainer container = new CompoundContainer(p_51604_, p_51605_); // Mohist TODO Container -> CompoundContainer
+            return Optional.of(new DoubleInventory(p_51604_, p_51605_, container)); // CraftBukkit
         }
 
         public Optional<MenuProvider> acceptSingle(ChestBlockEntity p_51602_) {
@@ -121,7 +_,7 @@
     public ChestBlock(BlockBehaviour.Properties p_51490_, Supplier<BlockEntityType<? extends ChestBlockEntity>> p_51491_) {
         super(p_51490_, p_51491_);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TYPE, ChestType.SINGLE).setValue(WATERLOGGED, Boolean.valueOf(false))
+                this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TYPE, ChestType.SINGLE).setValue(WATERLOGGED, Boolean.valueOf(false))
         );
     }
 
@@ -148,9 +_,9 @@
         if (p_51557_.is(this) && p_51556_.getAxis().isHorizontal()) {
             ChestType chesttype = p_51557_.getValue(TYPE);
             if (p_51555_.getValue(TYPE) == ChestType.SINGLE
-                && chesttype != ChestType.SINGLE
-                && p_51555_.getValue(FACING) == p_51557_.getValue(FACING)
-                && getConnectedDirection(p_51557_) == p_51556_.getOpposite()) {
+                    && chesttype != ChestType.SINGLE
+                    && p_51555_.getValue(FACING) == p_51557_.getValue(FACING)
+                    && getConnectedDirection(p_51557_) == p_51556_.getOpposite()) {
                 return p_51555_.setValue(TYPE, chesttype.getOpposite());
             }
         } else if (getConnectedDirection(p_51555_) == p_51556_) {
@@ -208,9 +_,9 @@
         }
 
         return this.defaultBlockState()
-            .setValue(FACING, direction)
-            .setValue(TYPE, chesttype)
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidstate.getType() == Fluids.WATER));
+                .setValue(FACING, direction)
+                .setValue(TYPE, chesttype)
+                .setValue(WATERLOGGED, Boolean.valueOf(fluidstate.getType() == Fluids.WATER));
     }
 
     @Override
@@ -261,7 +_,7 @@
 
     @Override
     public DoubleBlockCombiner.NeighborCombineResult<? extends ChestBlockEntity> combine(
-        BlockState p_51544_, Level p_51545_, BlockPos p_51546_, boolean p_51547_
+            BlockState p_51544_, Level p_51545_, BlockPos p_51546_, boolean p_51547_
     ) {
         BiPredicate<LevelAccessor, BlockPos> bipredicate;
         if (p_51547_) {
@@ -271,7 +_,7 @@
         }
 
         return DoubleBlockCombiner.combineWithNeigbour(
-            this.blockEntityType.get(), ChestBlock::getBlockType, ChestBlock::getConnectedDirection, FACING, p_51544_, p_51545_, p_51546_, bipredicate
+                this.blockEntityType.get(), ChestBlock::getBlockType, ChestBlock::getConnectedDirection, FACING, p_51544_, p_51545_, p_51546_, bipredicate
         );
     }
 
@@ -281,6 +_,13 @@
         return this.combine(p_51574_, p_51575_, p_51576_, false).apply(MENU_PROVIDER_COMBINER).orElse(null);
     }
 
+    // CraftBukkit start
+    @Nullable
+    public MenuProvider getMenuProvider(BlockState pState, Level pLevel, BlockPos pPos, boolean ignoreObstructions) {
+        return this.combine(pState, pLevel, pPos, ignoreObstructions).apply(MENU_PROVIDER_COMBINER).orElse((MenuProvider) null);
+    }
+    // CraftBukkit end
+
     public static DoubleBlockCombiner.Combiner<ChestBlockEntity, Float2FloatFunction> opennessCombiner(final LidBlockEntity p_51518_) {
         return new DoubleBlockCombiner.Combiner<ChestBlockEntity, Float2FloatFunction>() {
             public Float2FloatFunction acceptDouble(ChestBlockEntity p_51633_, ChestBlockEntity p_51634_) {
@@ -319,15 +_,15 @@
 
     private static boolean isCatSittingOnChest(LevelAccessor p_51564_, BlockPos p_51565_) {
         List<Cat> list = p_51564_.getEntitiesOfClass(
-            Cat.class,
-            new AABB(
-                (double)p_51565_.getX(),
-                (double)(p_51565_.getY() + 1),
-                (double)p_51565_.getZ(),
-                (double)(p_51565_.getX() + 1),
-                (double)(p_51565_.getY() + 2),
-                (double)(p_51565_.getZ() + 1)
-            )
+                Cat.class,
+                new AABB(
+                        (double) p_51565_.getX(),
+                        (double) (p_51565_.getY() + 1),
+                        (double) p_51565_.getZ(),
+                        (double) (p_51565_.getX() + 1),
+                        (double) (p_51565_.getY() + 2),
+                        (double) (p_51565_.getZ() + 1)
+                )
         );
         if (!list.isEmpty()) {
             for (Cat cat : list) {
@@ -357,7 +_,8 @@
 
     @Override
     protected BlockState mirror(BlockState p_51549_, Mirror p_51550_) {
-        return p_51549_.rotate(p_51550_.getRotation(p_51549_.getValue(FACING)));
+        BlockState rotated = p_51549_.rotate(p_51550_.getRotation(p_51549_.getValue(FACING)));
+        return p_51550_ == Mirror.NONE ? rotated : rotated.setValue(TYPE, rotated.getValue(TYPE).getOpposite());  // Forge: Fixed MC-134110 Structure mirroring breaking apart double chests
     }
 
     @Override
@@ -374,7 +_,38 @@
     protected void tick(BlockState p_220958_, ServerLevel p_220959_, BlockPos p_220960_, RandomSource p_220961_) {
         BlockEntity blockentity = p_220959_.getBlockEntity(p_220960_);
         if (blockentity instanceof ChestBlockEntity) {
-            ((ChestBlockEntity)blockentity).recheckOpen();
-        }
-    }
+            ((ChestBlockEntity) blockentity).recheckOpen();
+        }
+    }
+
+    // CraftBukkit start
+    public static class DoubleInventory implements MenuProvider {
+
+        private final ChestBlockEntity tileentitychest;
+        private final ChestBlockEntity tileentitychest1;
+        public final CompoundContainer inventorylargechest;
+
+        public DoubleInventory(ChestBlockEntity tileentitychest, ChestBlockEntity tileentitychest1, CompoundContainer inventorylargechest) {
+            this.tileentitychest = tileentitychest;
+            this.tileentitychest1 = tileentitychest1;
+            this.inventorylargechest = inventorylargechest;
+        }
+
+        @Nullable
+        @Override
+        public AbstractContainerMenu createMenu(int i, net.minecraft.world.entity.player.Inventory playerinventory, net.minecraft.world.entity.player.Player entityhuman) {
+            if (this.tileentitychest.canOpen(entityhuman) && this.tileentitychest1.canOpen(entityhuman)) {
+                this.tileentitychest.unpackLootTable(playerinventory.player);
+                this.tileentitychest1.unpackLootTable(playerinventory.player);
+                return ChestMenu.sixRows(i, playerinventory, this.inventorylargechest);
+            }
+            return null;
+        }
+
+        @Override
+        public net.minecraft.network.chat.Component getDisplayName() {
+            return this.tileentitychest.hasCustomName() ? this.tileentitychest.getDisplayName() : (this.tileentitychest1.hasCustomName() ? this.tileentitychest1.getDisplayName() : (tileentitychest1.hasCustomName() ? tileentitychest1.getDisplayName() : Component.translatable("container.chestDouble")));
+        }
+    }
+    // CraftBukkit end
 }
