--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -7,7 +_,6 @@
 import net.minecraft.core.Direction;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
@@ -36,6 +_,11 @@
 
    public void tick(BlockState p_221020_, ServerLevel p_221021_, BlockPos p_221022_, RandomSource p_221023_) {
       if (!this.scanForWater(p_221021_, p_221022_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(p_221021_, p_221022_, this.deadBlock.defaultBlockState()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_221021_.setBlock(p_221022_, this.deadBlock.defaultBlockState(), 2);
       }
 
@@ -50,9 +_,10 @@
    }
 
    protected boolean scanForWater(BlockGetter p_52135_, BlockPos p_52136_) {
+      BlockState state = p_52135_.getBlockState(p_52136_);
       for(Direction direction : Direction.values()) {
          FluidState fluidstate = p_52135_.getFluidState(p_52136_.relative(direction));
-         if (fluidstate.is(FluidTags.WATER)) {
+         if (state.canBeHydrated(p_52135_, p_52136_, fluidstate, p_52136_.relative(direction))) {
             return true;
          }
       }
