--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -24,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.Property;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DetectorRailBlock extends BaseRailBlock {
    public static final MapCodec<DetectorRailBlock> CODEC = simpleCodec(DetectorRailBlock::new);
@@ -37,6 +_,10 @@
 
    public DetectorRailBlock(BlockBehaviour.Properties p_52431_) {
       super(true, p_52431_);
+      this.registerDefaultState();
+   }
+
+   protected void registerDefaultState() {
       this.registerDefaultState(this.stateDefinition.any().setValue(POWERED, Boolean.valueOf(false)).setValue(SHAPE, RailShape.NORTH_SOUTH).setValue(WATERLOGGED, Boolean.valueOf(false)));
    }
 
@@ -80,6 +_,17 @@
          if (!list.isEmpty()) {
             flag1 = true;
          }
+
+         // CraftBukkit start
+         if (flag != flag1) {
+            org.bukkit.block.Block block = CraftBlock.at(p_52433_, p_52434_);
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+            p_52433_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+         }
+         // CraftBukkit end
 
          if (flag1 && !flag) {
             BlockState blockstate = p_52435_.setValue(POWERED, Boolean.valueOf(true));
