--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -36,6 +_,8 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DoorBlock extends Block {
    public static final MapCodec<DoorBlock> CODEC = RecordCodecBuilder.mapCodec((p_310521_) -> {
@@ -206,8 +_,22 @@
    }
 
    public void neighborChanged(BlockState p_52776_, Level p_52777_, BlockPos p_52778_, Block p_52779_, BlockPos p_52780_, boolean p_52781_) {
-      boolean flag = p_52777_.hasNeighborSignal(p_52778_) || p_52777_.hasNeighborSignal(p_52778_.relative(p_52776_.getValue(HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
-      if (!this.defaultBlockState().is(p_52779_) && flag != p_52776_.getValue(POWERED)) {
+      // CraftBukkit start
+      BlockPos otherHalf = p_52778_.relative(p_52776_.getValue(HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN);
+
+      org.bukkit.block.Block bukkitBlock = CraftBlock.at(p_52777_, p_52778_);
+      org.bukkit.block.Block blockTop = CraftBlock.at(p_52777_, otherHalf);
+
+      int power = bukkitBlock.getBlockPower();
+      int powerTop = blockTop.getBlockPower();
+      if (powerTop > power) power = powerTop;
+      int oldPower = (Boolean) p_52776_.getValue(POWERED) ? 15 : 0;
+
+      if (oldPower == 0 ^ power == 0) {
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bukkitBlock, oldPower, power);
+         p_52777_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+         boolean flag = eventRedstone.getNewCurrent() > 0;
+         // CraftBukkit end
          if (flag != p_52776_.getValue(OPEN)) {
             this.playSound((Entity)null, p_52777_, p_52778_, flag);
             p_52777_.gameEvent((Entity)null, flag ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, p_52778_);
