--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -49,7 +_,7 @@
    private final SoundEvent f_244293_;
    private final SoundEvent f_244089_;
 
-   protected DoorBlock(BlockBehaviour.Properties p_250467_, SoundEvent p_249791_, SoundEvent p_250722_) {
+   public DoorBlock(BlockBehaviour.Properties p_250467_, SoundEvent p_249791_, SoundEvent p_250722_) {
       super(p_250467_);
       this.m_49959_(this.f_49792_.m_61090_().m_61124_(f_52726_, Direction.NORTH).m_61124_(f_52727_, Boolean.valueOf(false)).m_61124_(f_52728_, DoorHingeSide.LEFT).m_61124_(f_52729_, Boolean.valueOf(false)).m_61124_(f_52730_, DoubleBlockHalf.LOWER));
       this.f_244293_ = p_249791_;
