--- a/net/minecraft/world/level/block/DragonEggBlock.java
+++ b/net/minecraft/world/level/block/DragonEggBlock.java
@@ -4,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.util.Mth;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.BlockGetter;
@@ -15,6 +_,8 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockFromToEvent;
 
 public class DragonEggBlock extends FallingBlock {
     public static final MapCodec<DragonEggBlock> CODEC = simpleCodec(DragonEggBlock::new);
@@ -55,6 +_,18 @@
                 p_52937_.random.nextInt(16) - p_52937_.random.nextInt(16)
             );
             if (p_52937_.getBlockState(blockpos).isAir() && worldborder.isWithinBounds(blockpos)) {
+            	// CraftBukkit start
+            	org.bukkit.block.Block from = CraftBlock.at(p_52937_, p_52938_);
+            	org.bukkit.block.Block to = CraftBlock.at(p_52937_, blockpos);
+            	BlockFromToEvent event = new BlockFromToEvent(from, to);
+            	org.bukkit.Bukkit.getPluginManager().callEvent(event);
+
+            	if (event.isCancelled()) {
+               	return;
+            	}
+
+            	blockpos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+            	// CraftBukkit end
                 if (p_52937_.isClientSide) {
                     for (int j = 0; j < 128; j++) {
                         double d0 = p_52937_.random.nextDouble();
