--- a/net/minecraft/world/level/block/IceBlock.java
+++ b/net/minecraft/world/level/block/IceBlock.java
@@ -54,6 +_,11 @@
    }
 
    protected void melt(BlockState p_54169_, Level p_54170_, BlockPos p_54171_) {
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(p_54170_, p_54171_, p_54170_.dimensionType().ultraWarm() ? Blocks.AIR.defaultBlockState() : Blocks.WATER.defaultBlockState()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       if (p_54170_.dimensionType().ultraWarm()) {
          p_54170_.removeBlock(p_54171_, false);
       } else {
