--- a/net/minecraft/world/level/block/IceBlock.java
+++ b/net/minecraft/world/level/block/IceBlock.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LightLayer;
 import net.minecraft.world.level.block.entity.BlockEntity;
@@ -55,6 +_,11 @@
     }
 
     protected void melt(BlockState p_54169_, Level p_54170_, BlockPos p_54171_) {
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(p_54170_, p_54171_, p_54170_.dimensionType().ultraWarm() ? Blocks.AIR.defaultBlockState() : Blocks.WATER.defaultBlockState()).isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
         if (p_54170_.dimensionType().ultraWarm()) {
             p_54170_.removeBlock(p_54171_, false);
         } else {
