--- a/net/minecraft/world/level/block/InfestedBlock.java
+++ b/net/minecraft/world/level/block/InfestedBlock.java
@@ -9,15 +_,16 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.EnchantmentTags;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.monster.Silverfish;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.Property;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class InfestedBlock extends Block {
     public static final MapCodec<InfestedBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -52,6 +_,7 @@
         Silverfish silverfish = EntityType.SILVERFISH.create(p_54181_);
         if (silverfish != null) {
             silverfish.moveTo((double)p_54182_.getX() + 0.5, (double)p_54182_.getY(), (double)p_54182_.getZ() + 0.5, 0.0F, 0.0F);
+         	silverfish.spawnReason(CreatureSpawnEvent.SpawnReason.SILVERFISH_BLOCK);
             p_54181_.addFreshEntity(silverfish);
             silverfish.spawnAnim();
         }
