--- a/net/minecraft/world/level/block/InfestedRotatedPillarBlock.java
+++ b/net/minecraft/world/level/block/InfestedRotatedPillarBlock.java
@@ -12,8 +_,8 @@
 
 public class InfestedRotatedPillarBlock extends InfestedBlock {
     public static final MapCodec<InfestedRotatedPillarBlock> CODEC = RecordCodecBuilder.mapCodec(
-        p_341833_ -> p_341833_.group(BuiltInRegistries.BLOCK.byNameCodec().fieldOf("host").forGetter(InfestedBlock::getHostBlock), propertiesCodec())
-                .apply(p_341833_, InfestedRotatedPillarBlock::new)
+        p_312491_ -> p_312491_.group(BuiltInRegistries.BLOCK.byNameCodec().fieldOf("host").forGetter(InfestedBlock::getHostBlock), propertiesCodec())
+                .apply(p_312491_, InfestedRotatedPillarBlock::new)
     );
 
     @Override
