--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -5,13 +_,11 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.component.DataComponents;
-import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
-import net.minecraft.world.ItemInteractionResult;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.JukeboxPlayable;
+import net.minecraft.world.item.RecordItem;
 import net.minecraft.world.item.component.CustomData;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
@@ -52,7 +_,7 @@
     @Override
     protected InteractionResult useWithoutItem(BlockState p_54281_, Level p_54282_, BlockPos p_54283_, Player p_54284_, BlockHitResult p_54286_) {
         if (p_54281_.getValue(HAS_RECORD) && p_54282_.getBlockEntity(p_54283_) instanceof JukeboxBlockEntity jukeboxblockentity) {
-            jukeboxblockentity.popOutTheItem();
+            jukeboxblockentity.popOutRecord();
             return InteractionResult.sidedSuccess(p_54282_.isClientSide);
         } else {
             return InteractionResult.PASS;
@@ -60,23 +_,10 @@
     }
 
     @Override
-    protected ItemInteractionResult useItemOn(
-        ItemStack p_345342_, BlockState p_343906_, Level p_342356_, BlockPos p_342905_, Player p_343973_, InteractionHand p_345093_, BlockHitResult p_345506_
-    ) {
-        if (p_343906_.getValue(HAS_RECORD)) {
-            return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
-        } else {
-            ItemStack itemstack = p_343973_.getItemInHand(p_345093_);
-            ItemInteractionResult iteminteractionresult = JukeboxPlayable.tryInsertIntoJukebox(p_342356_, p_342905_, itemstack, p_343973_);
-            return !iteminteractionresult.consumesAction() ? ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION : iteminteractionresult;
-        }
-    }
-
-    @Override
     protected void onRemove(BlockState p_54288_, Level p_54289_, BlockPos p_54290_, BlockState p_54291_, boolean p_54292_) {
         if (!p_54288_.is(p_54291_.getBlock())) {
             if (p_54289_.getBlockEntity(p_54290_) instanceof JukeboxBlockEntity jukeboxblockentity) {
-                jukeboxblockentity.popOutTheItem();
+                jukeboxblockentity.popOutRecord();
             }
 
             super.onRemove(p_54288_, p_54289_, p_54290_, p_54291_, p_54292_);
@@ -95,7 +_,7 @@
 
     @Override
     public int getSignal(BlockState p_272942_, BlockGetter p_273232_, BlockPos p_273524_, Direction p_272902_) {
-        if (p_273232_.getBlockEntity(p_273524_) instanceof JukeboxBlockEntity jukeboxblockentity && jukeboxblockentity.getSongPlayer().isPlaying()) {
+        if (p_273232_.getBlockEntity(p_273524_) instanceof JukeboxBlockEntity jukeboxblockentity && jukeboxblockentity.isRecordPlaying()) {
             return 15;
         }
 
@@ -109,7 +_,12 @@
 
     @Override
     protected int getAnalogOutputSignal(BlockState p_54277_, Level p_54278_, BlockPos p_54279_) {
-        return p_54278_.getBlockEntity(p_54279_) instanceof JukeboxBlockEntity jukeboxblockentity ? jukeboxblockentity.getComparatorOutput() : 0;
+        if (p_54278_.getBlockEntity(p_54279_) instanceof JukeboxBlockEntity jukeboxblockentity
+            && jukeboxblockentity.getTheItem().getItem() instanceof RecordItem recorditem) {
+            return recorditem.getAnalogOutput();
+        }
+
+        return 0;
     }
 
     @Override
@@ -125,6 +_,6 @@
     @Nullable
     @Override
     public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level p_239682_, BlockState p_239683_, BlockEntityType<T> p_239684_) {
-        return p_239683_.getValue(HAS_RECORD) ? createTickerHelper(p_239684_, BlockEntityType.JUKEBOX, JukeboxBlockEntity::tick) : null;
+        return p_239683_.getValue(HAS_RECORD) ? createTickerHelper(p_239684_, BlockEntityType.JUKEBOX, JukeboxBlockEntity::playRecordTick) : null;
     }
 }
