--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -23,8 +_,10 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.LeavesDecayEvent;
 
-public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
+public class LeavesBlock extends Block implements SimpleWaterloggedBlock, net.minecraftforge.common.IForgeShearable {
     public static final MapCodec<LeavesBlock> CODEC = simpleCodec(LeavesBlock::new);
     public static final int DECAY_DISTANCE = 7;
     public static final IntegerProperty DISTANCE = BlockStateProperties.DISTANCE;
@@ -61,6 +_,14 @@
     @Override
     protected void randomTick(BlockState p_221379_, ServerLevel p_221380_, BlockPos p_221381_, RandomSource p_221382_) {
         if (this.decaying(p_221379_)) {
+         	// CraftBukkit start
+        	LeavesDecayEvent event = new LeavesDecayEvent(CraftBlock.at(p_221380_, p_221381_));
+         	p_221380_.getCraftServer().getPluginManager().callEvent(event);
+
+         	if (event.isCancelled() || p_221380_.getBlockState(p_221381_).getBlock() != this) {
+            	return;
+         	}
+         	// CraftBukkit end
             dropResources(p_221379_, p_221380_, p_221381_);
             p_221380_.removeBlock(p_221381_, false);
         }
