--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -1,6 +_,5 @@
 package net.minecraft.world.level.block;
 
-import java.util.OptionalInt;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleTypes;
@@ -22,8 +_,12 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
-
-public class LeavesBlock extends Block implements SimpleWaterloggedBlock {
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.event.block.LeavesDecayEvent;
+
+import java.util.OptionalInt;
+
+public class LeavesBlock extends Block implements SimpleWaterloggedBlock, net.minecraftforge.common.IForgeShearable {
    public static final int f_153563_ = 7;
    public static final IntegerProperty f_54418_ = BlockStateProperties.f_61414_;
    public static final BooleanProperty f_54419_ = BlockStateProperties.f_61447_;
@@ -44,11 +_,7 @@
    }
 
    public void m_213898_(BlockState p_221379_, ServerLevel p_221380_, BlockPos p_221381_, RandomSource p_221382_) {
-      if (this.m_221385_(p_221379_)) {
-         m_49950_(p_221379_, p_221380_, p_221381_);
-         p_221380_.m_7471_(p_221381_, false);
-      }
-
+      // Mohist remove
    }
 
    protected boolean m_221385_(BlockState p_221386_) {
@@ -56,7 +_,7 @@
    }
 
    public void m_213897_(BlockState p_221369_, ServerLevel p_221370_, BlockPos p_221371_, RandomSource p_221372_) {
-      p_221370_.m_7731_(p_221371_, m_54435_(p_221369_, p_221370_, p_221371_), 3);
+      // Mohist not update distance
    }
 
    public int m_7753_(BlockState p_54460_, BlockGetter p_54461_, BlockPos p_54462_) {
