--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -17,6 +_,8 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class RedstoneTorchBlock extends BaseTorchBlock {
     public static final MapCodec<RedstoneTorchBlock> CODEC = simpleCodec(RedstoneTorchBlock::new);
@@ -71,18 +_,44 @@
             list.remove(0);
         }
 
-        if (p_221949_.getValue(LIT)) {
-            if (flag) {
-                p_221950_.setBlock(p_221951_, p_221949_.setValue(LIT, Boolean.valueOf(false)), 3);
-                if (isToggledTooFrequently(p_221950_, p_221951_, true)) {
-                    p_221950_.levelEvent(1502, p_221951_, 0);
-                    p_221950_.scheduleTick(p_221951_, p_221950_.getBlockState(p_221951_).getBlock(), 160);
-                }
-            }
-        } else if (!flag && !isToggledTooFrequently(p_221950_, p_221951_, false)) {
-            p_221950_.setBlock(p_221951_, p_221949_.setValue(LIT, Boolean.valueOf(true)), 3);
-        }
-    }
+      // CraftBukkit start
+      org.bukkit.plugin.PluginManager manager = p_221950_.getCraftServer().getPluginManager();
+      org.bukkit.block.Block block = CraftBlock.at(p_221950_, p_221951_);
+      int oldCurrent = ((Boolean) p_221949_.getValue(LIT)).booleanValue() ? 15 : 0;
+
+      BlockRedstoneEvent event = new BlockRedstoneEvent(block, oldCurrent, oldCurrent);
+      // CraftBukkit end
+      if (p_221949_.getValue(LIT)) {
+         if (flag) {
+            // CraftBukkit start
+            if (oldCurrent != 0) {
+               event.setNewCurrent(0);
+               manager.callEvent(event);
+               if (event.getNewCurrent() != 0) {
+                  return;
+               }
+            }
+            // CraftBukkit end
+            p_221950_.setBlock(p_221951_, p_221949_.setValue(LIT, Boolean.valueOf(false)), 3);
+            if (isToggledTooFrequently(p_221950_, p_221951_, true)) {
+               p_221950_.levelEvent(1502, p_221951_, 0);
+               p_221950_.scheduleTick(p_221951_, p_221950_.getBlockState(p_221951_).getBlock(), 160);
+            }
+         }
+      } else if (!flag && !isToggledTooFrequently(p_221950_, p_221951_, false)) {
+         // CraftBukkit start
+         if (oldCurrent != 15) {
+            event.setNewCurrent(15);
+            manager.callEvent(event);
+            if (event.getNewCurrent() != 15) {
+               return;
+            }
+         }
+         // CraftBukkit end
+         p_221950_.setBlock(p_221951_, p_221949_.setValue(LIT, Boolean.valueOf(true)), 3);
+      }
+
+   }
 
     @Override
     protected void neighborChanged(BlockState p_55699_, Level p_55700_, BlockPos p_55701_, Block p_55702_, BlockPos p_55703_, boolean p_55704_) {
