--- a/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -78,6 +_,7 @@
          if (!p_55875_.isClientSide) {
             ServerPlayer serverplayer = (ServerPlayer)p_55877_;
             if (serverplayer.getRespawnDimension() != p_55875_.dimension() || !p_55876_.equals(serverplayer.getRespawnPosition())) {
+               serverplayer.cause = org.bukkit.event.player.PlayerSpawnChangeEvent.Cause.RESPAWN_ANCHOR; // CraftBukkit // Mohist
                serverplayer.setRespawnPosition(p_55875_.dimension(), p_55876_, 0.0F, false, true);
                p_55875_.playSound((Player)null, (double)p_55876_.getX() + 0.5D, (double)p_55876_.getY() + 0.5D, (double)p_55876_.getZ() + 0.5D, SoundEvents.RESPAWN_ANCHOR_SET_SPAWN, SoundSource.BLOCKS, 1.0F, 1.0F);
                return InteractionResult.SUCCESS;
