--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.block;
 
 import com.mojang.serialization.MapCodec;
+import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.sounds.SoundEvents;
@@ -12,6 +_,11 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlockState;
+import org.bukkit.craftbukkit.v1_20_R3.util.BlockStateListPopulator;
+import org.bukkit.event.block.SpongeAbsorbEvent;
 
 public class SpongeBlock extends Block {
    public static final MapCodec<SpongeBlock> CODEC = simpleCodec(SpongeBlock::new);
@@ -47,7 +_,9 @@
    }
 
    private boolean removeWaterBreadthFirstSearch(Level p_56808_, BlockPos p_56809_) {
-      return BlockPos.breadthFirstTraversal(p_56809_, 6, 65, (p_277519_, p_277492_) -> {
+      BlockState spongeState = p_56808_.getBlockState(p_56809_);
+      BlockStateListPopulator blockList = new BlockStateListPopulator(p_56808_); // CraftBukkit - Use BlockStateListPopulator
+      BlockPos.breadthFirstTraversal(p_56809_, 6, 65, (p_277519_, p_277492_) -> {
          for(Direction direction : ALL_DIRECTIONS) {
             p_277492_.accept(p_277519_.relative(direction));
          }
@@ -56,15 +_,17 @@
          if (p_296944_.equals(p_56809_)) {
             return true;
          } else {
-            BlockState blockstate = p_56808_.getBlockState(p_296944_);
-            FluidState fluidstate = p_56808_.getFluidState(p_296944_);
-            if (!fluidstate.is(FluidTags.WATER)) {
+            // CraftBukkit start
+            BlockState blockstate = blockList.getBlockState(p_296944_);
+            FluidState fluidstate = blockList.getFluidState(p_296944_);
+            // CraftBukkit end
+            if (!spongeState.canBeHydrated(p_56808_, p_56809_, fluidstate, p_296944_)) {
                return false;
             } else {
                Block block = blockstate.getBlock();
                if (block instanceof BucketPickup) {
                   BucketPickup bucketpickup = (BucketPickup)block;
-                  if (!bucketpickup.pickupBlock((Player)null, p_56808_, p_296944_, blockstate).isEmpty()) {
+                  if (!bucketpickup.pickupBlock(null, blockList, p_296944_, blockstate).isEmpty()) { // CraftBukkit
                      return true;
                   }
                }
@@ -76,14 +_,50 @@
                      return false;
                   }
 
-                  BlockEntity blockentity = blockstate.hasBlockEntity() ? p_56808_.getBlockEntity(p_296944_) : null;
-                  dropResources(blockstate, p_56808_, p_296944_, blockentity);
-                  p_56808_.setBlock(p_296944_, Blocks.AIR.defaultBlockState(), 3);
+                  // CraftBukkit start
+                  // BlockEntity blockentity = blockstate.hasBlockEntity() ? pLevel.getBlockEntity(p_279054_) : null;
+                  // dropResources(blockstate, pLevel, p_279054_, blockentity);
+                  blockList.setBlock(p_296944_, Blocks.AIR.defaultBlockState(), 3);
                }
 
                return true;
             }
          }
-      }) > 1;
+      });
+      // CraftBukkit start
+      List<CraftBlockState> blocks = blockList.getList(); // Is a clone
+      if (!blocks.isEmpty()) {
+         final org.bukkit.block.Block bblock = CraftBlock.at(p_56808_, p_56809_);
+
+         SpongeAbsorbEvent event = new SpongeAbsorbEvent(bblock, (List<org.bukkit.block.BlockState>) (List) blocks);
+         Bukkit.getPluginManager().callEvent(event);
+
+         if (event.isCancelled()) {
+            return false;
+         }
+
+         for (CraftBlockState block : blocks) {
+            BlockPos blockposition1 = block.getPosition();
+            BlockState iblockdata = p_56808_.getBlockState(blockposition1);
+            FluidState fluid = p_56808_.getFluidState(blockposition1);
+
+            if (fluid.is(FluidTags.WATER)) {
+               if (iblockdata.getBlock() instanceof BucketPickup && !((BucketPickup) iblockdata.getBlock()).pickupBlock(null, blockList, blockposition1, iblockdata).isEmpty()) {
+                  // NOP
+               } else if (iblockdata.getBlock() instanceof LiquidBlock) {
+                  // NOP
+               } else if (iblockdata.is(Blocks.KELP) || iblockdata.is(Blocks.KELP_PLANT) || iblockdata.is(Blocks.SEAGRASS) || iblockdata.is(Blocks.TALL_SEAGRASS)) {
+                  BlockEntity tileentity = iblockdata.hasBlockEntity() ? p_56808_.getBlockEntity(blockposition1) : null;
+
+                  dropResources(iblockdata, p_56808_, blockposition1, tileentity);
+               }
+            }
+            p_56808_.setBlock(blockposition1, block.getHandle(), block.getFlag());
+         }
+
+         return true;
+      }
+      return false;
+      // CraftBukkit end
    }
 }
