--- a/net/minecraft/world/level/block/SweetBerryBushBlock.java
+++ b/net/minecraft/world/level/block/SweetBerryBushBlock.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.block;
 
 import com.mojang.serialization.MapCodec;
+import java.util.Collections;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvents;
@@ -27,6 +_,9 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerHarvestBlockEvent;
 
 public class SweetBerryBushBlock extends BushBlock implements BonemealableBlock {
    public static final MapCodec<SweetBerryBushBlock> CODEC = simpleCodec(SweetBerryBushBlock::new);
@@ -63,10 +_,11 @@
 
    public void randomTick(BlockState p_222563_, ServerLevel p_222564_, BlockPos p_222565_, RandomSource p_222566_) {
       int i = p_222563_.getValue(AGE);
-      if (i < 3 && p_222566_.nextInt(5) == 0 && p_222564_.getRawBrightness(p_222565_.above(), 0) >= 9) {
+      if (i < 3 && p_222564_.getRawBrightness(p_222565_.above(), 0) >= 9 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_222564_, p_222565_, p_222563_, p_222566_.nextFloat() < (p_222564_.spigotConfig.sweetBerryModifier / (100.0f * 5)) && p_222564_.getRawBrightness(p_222565_.above(), 0) >= 9)) {  // Spigot - SPIGOT-7159: Better modifier resolution
          BlockState blockstate = p_222563_.setValue(AGE, Integer.valueOf(i + 1));
-         p_222564_.setBlock(p_222565_, blockstate, 2);
+         if (!CraftEventFactory.handleBlockGrowEvent(p_222564_, p_222565_, blockstate, 2)) return; // CraftBukkit
          p_222564_.gameEvent(GameEvent.BLOCK_CHANGE, p_222565_, GameEvent.Context.of(blockstate));
+         net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_222564_, p_222565_, p_222563_);
       }
 
    }
@@ -78,7 +_,7 @@
             double d0 = Math.abs(p_57273_.getX() - p_57273_.xOld);
             double d1 = Math.abs(p_57273_.getZ() - p_57273_.zOld);
             if (d0 >= (double)0.003F || d1 >= (double)0.003F) {
-               p_57273_.hurt(p_57271_.damageSources().sweetBerryBush(), 1.0F);
+               p_57273_.hurt(p_57271_.damageSources().sweetBerryBush().directBlock(p_57271_, p_57272_), 1.0F); // CraftBukkit
             }
          }
 
@@ -92,7 +_,15 @@
          return InteractionResult.PASS;
       } else if (i > 1) {
          int j = 1 + p_57276_.random.nextInt(2);
-         popResource(p_57276_, p_57277_, new ItemStack(Items.SWEET_BERRIES, j + (flag ? 1 : 0)));
+         // CraftBukkit start
+         PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(p_57276_, p_57277_, p_57278_, p_57279_, Collections.singletonList(new ItemStack(Items.SWEET_BERRIES, j + (flag ? 1 : 0))));
+         if (event.isCancelled()) {
+            return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+         }
+         for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+            popResource(p_57276_, p_57277_, CraftItemStack.asNMSCopy(itemStack));
+         }
+         // CraftBukkit end
          p_57276_.playSound((Player)null, p_57277_, SoundEvents.SWEET_BERRY_BUSH_PICK_BERRIES, SoundSource.BLOCKS, 1.0F, 0.8F + p_57276_.random.nextFloat() * 0.4F);
          BlockState blockstate = p_57275_.setValue(AGE, Integer.valueOf(1));
          p_57276_.setBlock(p_57277_, blockstate, 2);
