--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -31,6 +_,8 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_20_R4.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleWaterloggedBlock {
    public static final MapCodec<TrapDoorBlock> CODEC = RecordCodecBuilder.mapCodec((p_309899_) -> {
@@ -128,6 +_,18 @@
       if (!p_57548_.isClientSide) {
          boolean flag = p_57548_.hasNeighborSignal(p_57549_);
          if (flag != p_57547_.getValue(POWERED)) {
+            // CraftBukkit start
+            org.bukkit.block.Block bblock = CraftBlock.at(p_57548_, p_57549_);
+
+            int power = bblock.getBlockPower();
+            int oldPower = (Boolean) p_57547_.getValue(OPEN) ? 15 : 0;
+
+            if (oldPower == 0 ^ power == 0 || p_57550_.defaultBlockState().isSignalSource()) {
+               BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bblock, oldPower, power);
+               p_57548_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+               flag = eventRedstone.getNewCurrent() > 0;
+            }
+            // CraftBukkit end
             if (p_57547_.getValue(OPEN) != flag) {
                p_57547_ = p_57547_.setValue(OPEN, Boolean.valueOf(flag));
                this.playSound((Player)null, p_57548_, p_57549_, flag);
@@ -174,6 +_,19 @@
 
       return super.updateShape(p_57554_, p_57555_, p_57556_, p_57557_, p_57558_, p_57559_);
    }
+
+   //Forge Start
+   @Override
+   public boolean isLadder(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, net.minecraft.world.entity.LivingEntity entity) {
+      if (state.getValue(OPEN)) {
+         BlockPos downPos = pos.below();
+         BlockState down = world.getBlockState(downPos);
+         return down.getBlock().makesOpenTrapdoorAboveClimbable(down, world, downPos, state);
+      }
+      return false;
+   }
+   //Forge End
+
 
    protected BlockSetType getType() {
       return this.type;
