--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -24,6 +_,8 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlock;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class TripWireBlock extends Block {
    public static final MapCodec<TripWireBlock> CODEC = RecordCodecBuilder.mapCodec((p_312765_) -> {
@@ -81,7 +_,7 @@
    }
 
    public BlockState playerWillDestroy(Level p_57615_, BlockPos p_57616_, BlockState p_57617_, Player p_57618_) {
-      if (!p_57615_.isClientSide && !p_57618_.getMainHandItem().isEmpty() && p_57618_.getMainHandItem().is(Items.SHEARS)) {
+      if (!p_57615_.isClientSide && !p_57618_.getMainHandItem().isEmpty() && p_57618_.getMainHandItem().canPerformAction(net.minecraftforge.common.ToolActions.SHEARS_DISARM)) {
          p_57615_.setBlock(p_57616_, p_57617_.setValue(DISARMED, Boolean.valueOf(true)), 4);
          p_57615_.gameEvent(p_57618_, GameEvent.SHEAR, p_57616_);
       }
@@ -136,6 +_,39 @@
             }
          }
       }
+
+      // CraftBukkit start - Call interact even when triggering connected tripwire
+      if (flag != flag1 && flag1 && (Boolean)blockstate.getValue(ATTACHED)) {
+         org.bukkit.plugin.PluginManager manager = p_57608_.getCraftServer().getPluginManager();
+         org.bukkit.block.Block block = CraftBlock.at(p_57608_, p_57609_);
+         boolean allowed = false;
+
+         // If all of the events are cancelled block the tripwire trigger, else allow
+         for (Object object : list) {
+            if (object != null) {
+               org.bukkit.event.Cancellable cancellable;
+
+               if (object instanceof Player player) {
+                  cancellable = org.bukkit.craftbukkit.v1_20_R3.event.CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.PHYSICAL, p_57609_, null, null, null);
+               } else if (object instanceof Entity) {
+                  cancellable = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+                  manager.callEvent((EntityInteractEvent) cancellable);
+               } else {
+                  continue;
+               }
+
+               if (!cancellable.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+
+         if (!allowed) {
+            return;
+         }
+      }
+      // CraftBukkit end
 
       if (flag1 != flag) {
          blockstate = blockstate.setValue(POWERED, Boolean.valueOf(flag1));
