--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -18,7 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -29,6 +_,9 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_20_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class TurtleEggBlock extends Block {
    public static final MapCodec<TurtleEggBlock> CODEC = simpleCodec(TurtleEggBlock::new);
@@ -68,6 +_,19 @@
    private void destroyEgg(Level p_154851_, BlockState p_154852_, BlockPos p_154853_, Entity p_154854_, int p_154855_) {
       if (this.canDestroyEgg(p_154851_, p_154854_)) {
          if (!p_154851_.isClientSide && p_154851_.random.nextInt(p_154855_) == 0 && p_154852_.is(Blocks.TURTLE_EGG)) {
+            // CraftBukkit start - Step on eggs
+            org.bukkit.event.Cancellable cancellable;
+            if (p_154854_ instanceof Player player) {
+               cancellable = CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.PHYSICAL, p_154853_, null, null, null);
+            } else {
+               cancellable = new EntityInteractEvent(p_154854_.getBukkitEntity(), CraftBlock.at(p_154851_, p_154853_));
+               p_154851_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             this.decreaseEggs(p_154851_, p_154853_, p_154852_);
          }
 
@@ -91,13 +_,21 @@
       if (this.shouldUpdateHatchLevel(p_222645_) && onSand(p_222645_, p_222646_)) {
          int i = p_222644_.getValue(HATCH);
          if (i < 2) {
+            // CraftBukkit start - Call BlockGrowEvent
+            if (!CraftEventFactory.handleBlockGrowEvent(p_222645_, p_222646_, p_222644_.setValue(HATCH, i + 1), 2)) {
+               return;
+            }
+            // CraftBukkit end
             p_222645_.playSound((Player)null, p_222646_, SoundEvents.TURTLE_EGG_CRACK, SoundSource.BLOCKS, 0.7F, 0.9F + p_222647_.nextFloat() * 0.2F);
-            p_222645_.setBlock(p_222646_, p_222644_.setValue(HATCH, Integer.valueOf(i + 1)), 2);
-            p_222645_.gameEvent(GameEvent.BLOCK_CHANGE, p_222646_, GameEvent.Context.of(p_222644_));
+           // pLevel.setBlock(pPos, pState.setValue(HATCH, Integer.valueOf(i + 1)), 2);
          } else {
+            // CraftBukkit start - Call BlockFadeEvent
+            if (CraftEventFactory.callBlockFadeEvent(p_222645_, p_222646_, Blocks.AIR.defaultBlockState()).isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             p_222645_.playSound((Player)null, p_222646_, SoundEvents.TURTLE_EGG_HATCH, SoundSource.BLOCKS, 0.7F, 0.9F + p_222647_.nextFloat() * 0.2F);
             p_222645_.removeBlock(p_222646_, false);
-            p_222645_.gameEvent(GameEvent.BLOCK_DESTROY, p_222646_, GameEvent.Context.of(p_222644_));
 
             for(int j = 0; j < p_222644_.getValue(EGGS); ++j) {
                p_222645_.levelEvent(2001, p_222646_, Block.getId(p_222644_));
@@ -106,6 +_,7 @@
                   turtle.setAge(-24000);
                   turtle.setHomePos(p_222646_);
                   turtle.moveTo((double)p_222646_.getX() + 0.3D + (double)j * 0.2D, (double)p_222646_.getY(), (double)p_222646_.getZ() + 0.3D, 0.0F, 0.0F);
+                  turtle.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG);
                   p_222645_.addFreshEntity(turtle);
                }
             }
@@ -166,7 +_,7 @@
          if (!(p_57769_ instanceof LivingEntity)) {
             return false;
          } else {
-            return p_57769_ instanceof Player || p_57768_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            return p_57769_ instanceof Player || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_57768_, p_57769_);
          }
       } else {
          return false;
