--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -22,8 +_,9 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty UP = PipeBlock.UP;
    public static final BooleanProperty NORTH = PipeBlock.NORTH;
    public static final BooleanProperty EAST = PipeBlock.EAST;
@@ -158,7 +_,7 @@
 
    public void randomTick(BlockState p_222655_, ServerLevel p_222656_, BlockPos p_222657_, RandomSource p_222658_) {
       if (p_222656_.getGameRules().getBoolean(GameRules.RULE_DO_VINES_SPREAD)) {
-         if (p_222658_.nextInt(4) == 0) {
+      if (p_222656_.random.nextFloat() < (p_222656_.spigotConfig.vineModifier / (100.0f * 4)) && p_222656_.isAreaLoaded(p_222657_, 4)) { // Forge: check area to prevent loading unloaded chunks
             Direction direction = Direction.getRandom(p_222658_);
             BlockPos blockpos = p_222657_.above();
             if (direction.getAxis().isHorizontal() && !p_222655_.getValue(getPropertyForFace(direction))) {
@@ -172,19 +_,24 @@
                      boolean flag1 = p_222655_.getValue(getPropertyForFace(direction4));
                      BlockPos blockpos2 = blockpos4.relative(direction3);
                      BlockPos blockpos3 = blockpos4.relative(direction4);
+
+                     // CraftBukkit start - Call BlockSpreadEvent
+                     BlockPos source = p_222657_;
+
                      if (flag && isAcceptableNeighbour(p_222656_, blockpos2, direction3)) {
-                        p_222656_.setBlock(blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction3), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
                      } else if (flag1 && isAcceptableNeighbour(p_222656_, blockpos3, direction4)) {
-                        p_222656_.setBlock(blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction4), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction4), true), 2);
                      } else {
                         Direction direction1 = direction.getOpposite();
                         if (flag && p_222656_.isEmptyBlock(blockpos2) && isAcceptableNeighbour(p_222656_, p_222657_.relative(direction3), direction1)) {
-                           p_222656_.setBlock(blockpos2, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos2, this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
                         } else if (flag1 && p_222656_.isEmptyBlock(blockpos3) && isAcceptableNeighbour(p_222656_, p_222657_.relative(direction4), direction1)) {
-                           p_222656_.setBlock(blockpos3, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos3, this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
                         } else if ((double)p_222658_.nextFloat() < 0.05D && isAcceptableNeighbour(p_222656_, blockpos4.above(), Direction.UP)) {
-                           p_222656_.setBlock(blockpos4, this.defaultBlockState().setValue(UP, Boolean.valueOf(true)), 2);
+                           CraftEventFactory.handleBlockSpreadEvent(p_222656_, source, blockpos4, this.defaultBlockState().setValue(UP, true), 2);
                         }
+                        // CraftBukkit end
                      }
                   } else if (isAcceptableNeighbour(p_222656_, blockpos4, direction)) {
                      p_222656_.setBlock(p_222657_, p_222655_.setValue(getPropertyForFace(direction), Boolean.valueOf(true)), 2);
@@ -212,7 +_,7 @@
                      }
 
                      if (this.hasHorizontalConnection(blockstate3)) {
-                        p_222656_.setBlock(blockpos, blockstate3, 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, p_222657_, blockpos, blockstate3, 2); // CraftBukkit
                      }
 
                      return;
@@ -226,7 +_,7 @@
                      BlockState blockstate1 = blockstate.isAir() ? this.defaultBlockState() : blockstate;
                      BlockState blockstate2 = this.copyRandomFaces(p_222655_, blockstate1, p_222658_);
                      if (blockstate1 != blockstate2 && this.hasHorizontalConnection(blockstate2)) {
-                        p_222656_.setBlock(blockpos1, blockstate2, 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_222656_, p_222657_, blockpos1, blockstate2, 2); // CraftBukkit
                      }
                   }
                }
