--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -3,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
@@ -10,6 +_,8 @@
 import net.minecraft.world.level.block.state.properties.BlockSetType;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
+import org.bukkit.craftbukkit.v1_20_R1.block.CraftBlock;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public class WeightedPressurePlateBlock extends BasePressurePlateBlock {
    public static final IntegerProperty f_58198_ = BlockStateProperties.f_61426_;
@@ -22,7 +_,31 @@
    }
 
    protected int m_6693_(Level p_58213_, BlockPos p_58214_) {
-      int i = Math.min(m_289607_(p_58213_, f_49287_.m_82338_(p_58214_), Entity.class), this.f_58199_);
+      // CraftBukkit start
+      // int i = Math.min(pLevel.getEntitiesOfClass(Entity.class, TOUCH_AABB.move(pPos)).size(), this.maxWeight);
+      int i = 0;
+      java.util.Iterator iterator = p_58213_.m_45976_(Entity.class, f_49287_.m_82338_(p_58214_)).iterator();
+
+      while (iterator.hasNext()) {
+         Entity entity = (Entity) iterator.next();
+
+         org.bukkit.event.Cancellable cancellable;
+
+         if (entity instanceof Player player) {
+            cancellable = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.PHYSICAL, p_58214_, null, null, null);
+         } else {
+            cancellable = new EntityInteractEvent(entity.getBukkitEntity(), CraftBlock.at(p_58213_, p_58214_));
+            p_58213_.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+         }
+
+         // We only want to block turning the plate on if all events are cancelled
+         if (!cancellable.isCancelled()) {
+            i++;
+         }
+      }
+
+      i = Math.min(i, this.f_58199_);
+      // CraftBukkit end
       if (i > 0) {
          float f = (float)Math.min(this.f_58199_, i) / (float)this.f_58199_;
          return Mth.m_14167_(f * 15.0F);
