--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -20,7 +_,7 @@
    private final SoundEvent f_243985_;
    private final SoundEvent f_243907_;
 
-   protected WeightedPressurePlateBlock(int p_250012_, BlockBehaviour.Properties p_249299_, SoundEvent p_251068_, SoundEvent p_249071_) {
+   public WeightedPressurePlateBlock(int p_250012_, BlockBehaviour.Properties p_249299_, SoundEvent p_251068_, SoundEvent p_249071_) {
       super(p_249299_);
       this.m_49959_(this.f_49792_.m_61090_().m_61124_(f_58198_, Integer.valueOf(0)));
       this.f_58199_ = p_250012_;
