--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.level.block.state.pattern.BlockPattern;
 import net.minecraft.world.level.block.state.pattern.BlockPatternBuilder;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class WitherSkullBlock extends SkullBlock {
    public static final MapCodec<WitherSkullBlock> CODEC = simpleCodec(WitherSkullBlock::new);
@@ -48,6 +_,7 @@
    }
 
    public static void checkSpawn(Level p_58256_, BlockPos p_58257_, SkullBlockEntity p_58258_) {
+      if (p_58256_.captureBlockStates) return; // CraftBukkit
       if (!p_58256_.isClientSide) {
          BlockState blockstate = p_58258_.getBlockState();
          boolean flag = blockstate.is(Blocks.WITHER_SKELETON_SKULL) || blockstate.is(Blocks.WITHER_SKELETON_WALL_SKULL);
@@ -56,17 +_,22 @@
             if (blockpattern$blockpatternmatch != null) {
                WitherBoss witherboss = EntityType.WITHER.create(p_58256_);
                if (witherboss != null) {
-                  CarvedPumpkinBlock.clearPatternBlocks(p_58256_, blockpattern$blockpatternmatch);
+                  // CarvedPumpkinBlock.clearPatternBlocks(pLevel, blockpattern$blockpatternmatch);
                   BlockPos blockpos = blockpattern$blockpatternmatch.getBlock(1, 2, 0).getPos();
                   witherboss.moveTo((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.55D, (double)blockpos.getZ() + 0.5D, blockpattern$blockpatternmatch.getForwards().getAxis() == Direction.Axis.X ? 0.0F : 90.0F, 0.0F);
                   witherboss.yBodyRot = blockpattern$blockpatternmatch.getForwards().getAxis() == Direction.Axis.X ? 0.0F : 90.0F;
                   witherboss.makeInvulnerable();
-
+                  // CraftBukkit start
+                  if (!p_58256_.addFreshEntity(witherboss, CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
+                     return;
+                  }
+                  CarvedPumpkinBlock.clearPatternBlocks(p_58256_, blockpattern$blockpatternmatch); // CraftBukkit - from above
+                  // CraftBukkit end
                   for(ServerPlayer serverplayer : p_58256_.getEntitiesOfClass(ServerPlayer.class, witherboss.getBoundingBox().inflate(50.0D))) {
                      CriteriaTriggers.SUMMONED_ENTITY.trigger(serverplayer, witherboss);
                   }
 
-                  p_58256_.addFreshEntity(witherboss);
+                  // pLevel.addFreshEntity(witherboss);
                   CarvedPumpkinBlock.updatePatternBlocks(p_58256_, blockpattern$blockpatternmatch);
                }
 
