--- a/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
@@ -250,6 +_,9 @@
             BlockPos blockpos = p_227506_.below();
             if (p_227509_.isInside(blockpos)) {
                RandomizableContainerBlockEntity.setLootTable(p_227507_, p_227508_, blockpos, BuiltInLootTables.END_CITY_TREASURE);
+               // CraftBukkit start - ensure block transformation
+               setCraftLootTable(p_227507_, blockpos, p_227508_, BuiltInLootTables.END_CITY_TREASURE);
+               // CraftBukkit end
             }
          } else if (p_227509_.isInside(p_227506_) && Level.isInSpawnableBounds(p_227506_)) {
             if (p_227505_.startsWith("Sentry")) {
