--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -24,7 +_,6 @@
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
 import net.minecraft.world.level.levelgen.structure.templatesystem.BlockIgnoreProcessor;
-import net.minecraft.world.level.levelgen.structure.templatesystem.LiquidSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
@@ -32,9 +_,9 @@
 
 public class IglooPieces {
     public static final int GENERATION_HEIGHT = 90;
-    static final ResourceLocation STRUCTURE_LOCATION_IGLOO = ResourceLocation.withDefaultNamespace("igloo/top");
-    private static final ResourceLocation STRUCTURE_LOCATION_LADDER = ResourceLocation.withDefaultNamespace("igloo/middle");
-    private static final ResourceLocation STRUCTURE_LOCATION_LABORATORY = ResourceLocation.withDefaultNamespace("igloo/bottom");
+    static final ResourceLocation STRUCTURE_LOCATION_IGLOO = new ResourceLocation("igloo/top");
+    private static final ResourceLocation STRUCTURE_LOCATION_LADDER = new ResourceLocation("igloo/middle");
+    private static final ResourceLocation STRUCTURE_LOCATION_LABORATORY = new ResourceLocation("igloo/bottom");
     static final Map<ResourceLocation, BlockPos> PIVOTS = ImmutableMap.of(
         STRUCTURE_LOCATION_IGLOO, new BlockPos(3, 5, 5), STRUCTURE_LOCATION_LADDER, new BlockPos(1, 3, 1), STRUCTURE_LOCATION_LABORATORY, new BlockPos(3, 6, 7)
     );
@@ -79,8 +_,7 @@
                 .setRotation(p_227576_)
                 .setMirror(Mirror.NONE)
                 .setRotationPivot(IglooPieces.PIVOTS.get(p_227577_))
-                .addProcessor(BlockIgnoreProcessor.STRUCTURE_BLOCK)
-                .setLiquidSettings(LiquidSettings.IGNORE_WATERLOGGING);
+                .addProcessor(BlockIgnoreProcessor.STRUCTURE_BLOCK);
         }
 
         private static BlockPos makePosition(ResourceLocation p_227564_, BlockPos p_227565_, int p_227566_) {
@@ -114,7 +_,7 @@
             ChunkPos p_227573_,
             BlockPos p_227574_
         ) {
-            ResourceLocation resourcelocation = ResourceLocation.parse(this.templateName);
+            ResourceLocation resourcelocation = new ResourceLocation(this.templateName);
             StructurePlaceSettings structureplacesettings = makeSettings(this.placeSettings.getRotation(), resourcelocation);
             BlockPos blockpos = IglooPieces.OFFSETS.get(resourcelocation);
             BlockPos blockpos1 = this.templatePosition
