--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -79,7 +_,9 @@
             p_227584_.setBlock(p_227583_, Blocks.AIR.defaultBlockState(), 3);
             BlockEntity blockentity = p_227584_.getBlockEntity(p_227583_.below());
             if (blockentity instanceof ChestBlockEntity) {
-               ((ChestBlockEntity)blockentity).setLootTable(BuiltInLootTables.IGLOO_CHEST, p_227585_.nextLong());
+               // CraftBukkit start - ensure block transformation
+               setCraftLootTable(p_227584_, p_227583_.below(), p_227585_, BuiltInLootTables.IGLOO_CHEST);
+               // CraftBukkit end
             }
 
          }
