--- a/net/minecraft/world/level/levelgen/structure/structures/OceanRuinPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanRuinPieces.java
@@ -25,8 +_,6 @@
 import net.minecraft.world.level.block.ChestBlock;
 import net.minecraft.world.level.block.Mirror;
 import net.minecraft.world.level.block.Rotation;
-import net.minecraft.world.level.block.entity.BlockEntity;
-import net.minecraft.world.level.block.entity.ChestBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -184,11 +_,12 @@
 
       protected void handleDataMarker(String p_229046_, BlockPos p_229047_, ServerLevelAccessor p_229048_, RandomSource p_229049_, BoundingBox p_229050_) {
          if ("chest".equals(p_229046_)) {
-            p_229048_.setBlock(p_229047_, Blocks.CHEST.defaultBlockState().setValue(ChestBlock.WATERLOGGED, Boolean.valueOf(p_229048_.getFluidState(p_229047_).is(FluidTags.WATER))), 2);
-            BlockEntity blockentity = p_229048_.getBlockEntity(p_229047_);
-            if (blockentity instanceof ChestBlockEntity) {
-               ((ChestBlockEntity)blockentity).setLootTable(this.isLarge ? BuiltInLootTables.UNDERWATER_RUIN_BIG : BuiltInLootTables.UNDERWATER_RUIN_SMALL, p_229049_.nextLong());
-            }
+            // CraftBukkit start - transform block to ensure loot table is accessible
+            org.bukkit.craftbukkit.v1_20_R2.block.CraftChest craftChest = (org.bukkit.craftbukkit.v1_20_R2.block.CraftChest) org.bukkit.craftbukkit.v1_20_R2.block.CraftBlockStates.getBlockState(p_229047_, Blocks.CHEST.defaultBlockState().setValue(ChestBlock.WATERLOGGED, p_229048_.getFluidState(p_229047_).is(FluidTags.WATER)), null);
+            craftChest.setSeed(p_229049_.nextLong());
+            craftChest.setLootTable(org.bukkit.Bukkit.getLootTable(org.bukkit.craftbukkit.v1_20_R2.util.CraftNamespacedKey.fromMinecraft(this.isLarge ? BuiltInLootTables.UNDERWATER_RUIN_BIG : BuiltInLootTables.UNDERWATER_RUIN_SMALL)));
+            placeCraftBlockEntity(p_229048_, p_229047_, craftChest, 2);
+            // CraftBukkit end
          } else if ("drowned".equals(p_229046_)) {
             Drowned drowned = EntityType.DROWNED.create(p_229048_.getLevel());
             if (drowned != null) {
