--- a/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalStructure.java
@@ -102,9 +_,9 @@
             ruinedportalpiece$properties.replaceWithBlackstone = ruinedportalstructure$setup3.replaceWithBlackstone();
             ResourceLocation resourcelocation;
             if (worldgenrandom.nextFloat() < 0.05F) {
-                resourcelocation = ResourceLocation.withDefaultNamespace(STRUCTURE_LOCATION_GIANT_PORTALS[worldgenrandom.nextInt(STRUCTURE_LOCATION_GIANT_PORTALS.length)]);
+                resourcelocation = new ResourceLocation(STRUCTURE_LOCATION_GIANT_PORTALS[worldgenrandom.nextInt(STRUCTURE_LOCATION_GIANT_PORTALS.length)]);
             } else {
-                resourcelocation = ResourceLocation.withDefaultNamespace(STRUCTURE_LOCATION_PORTALS[worldgenrandom.nextInt(STRUCTURE_LOCATION_PORTALS.length)]);
+                resourcelocation = new ResourceLocation(STRUCTURE_LOCATION_PORTALS[worldgenrandom.nextInt(STRUCTURE_LOCATION_PORTALS.length)]);
             }
 
             StructureTemplate structuretemplate = p_229285_.structureTemplateManager().getOrCreate(resourcelocation);
