--- a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
@@ -1222,7 +_,7 @@
         }
 
         private static ResourceLocation makeLocation(String p_230211_) {
-            return ResourceLocation.withDefaultNamespace("woodland_mansion/" + p_230211_);
+            return new ResourceLocation("woodland_mansion/" + p_230211_);
         }
 
         private static StructurePlaceSettings makeSettings(Mirror p_230205_, Rotation p_230206_) {
