--- a/net/minecraft/world/flag/FeatureFlagRegistry.java
+++ b/net/minecraft/world/flag/FeatureFlagRegistry.java
@@ -17,7 +_,7 @@
 public class FeatureFlagRegistry {
    private static final Logger LOGGER = LogUtils.getLogger();
    private final FeatureFlagUniverse universe;
-   private final Map<ResourceLocation, FeatureFlag> names;
+   public final Map<ResourceLocation, FeatureFlag> names;
    private final FeatureFlagSet allFlags;
 
    FeatureFlagRegistry(FeatureFlagUniverse p_249715_, FeatureFlagSet p_249277_, Map<ResourceLocation, FeatureFlag> p_249557_) {
