--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.alchemy.PotionContents;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class Arrow extends AbstractArrow {
     private static final int EXPOSED_POTION_DECAY_TIME = 600;
@@ -103,6 +_,7 @@
         PotionContents potioncontents = this.getPotionContents();
         if (potioncontents.potion().isPresent()) {
             for (MobEffectInstance mobeffectinstance : potioncontents.potion().get().value().getEffects()) {
+                p_36873_.addEffectCause(EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
                 p_36873_.addEffect(
                     new MobEffectInstance(
                         mobeffectinstance.getEffect(),
@@ -117,6 +_,7 @@
         }
 
         for (MobEffectInstance mobeffectinstance1 : potioncontents.customEffects()) {
+            p_36873_.addEffectCause(EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
             p_36873_.addEffect(mobeffectinstance1, entity);
         }
     }
