--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.alchemy.PotionContents;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class Arrow extends AbstractArrow {
     private static final int EXPOSED_POTION_DECAY_TIME = 600;
@@ -36,11 +_,11 @@
         this.updateColor();
     }
 
-    private PotionContents getPotionContents() {
+    public PotionContents getPotionContents() {
         return this.getPickupItemStackOrigin().getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY);
     }
 
-    private void setPotionContents(PotionContents p_328713_) {
+    public void setPotionContents(PotionContents p_328713_) {
         this.getPickupItemStackOrigin().set(DataComponents.POTION_CONTENTS, p_328713_);
         this.updateColor();
     }
@@ -51,7 +_,7 @@
         this.updateColor();
     }
 
-    private void updateColor() {
+    public void updateColor() {
         PotionContents potioncontents = this.getPotionContents();
         this.entityData.set(ID_EFFECT_COLOR, potioncontents.equals(PotionContents.EMPTY) ? -1 : potioncontents.getColor());
     }
@@ -104,6 +_,7 @@
         PotionContents potioncontents = this.getPotionContents();
         if (potioncontents.potion().isPresent()) {
             for (MobEffectInstance mobeffectinstance : potioncontents.potion().get().value().getEffects()) {
+                p_36873_.addEffectCause(EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
                 p_36873_.addEffect(
                     new MobEffectInstance(
                         mobeffectinstance.getEffect(),
@@ -118,6 +_,7 @@
         }
 
         for (MobEffectInstance mobeffectinstance1 : potioncontents.customEffects()) {
+            p_36873_.addEffectCause(EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
             p_36873_.addEffect(mobeffectinstance1, entity);
         }
     }
