--- a/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.TraceableEntity;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class EvokerFangs extends Entity implements TraceableEntity {
     public static final int ATTACK_DURATION = 20;
@@ -106,6 +_,7 @@
             }
 
             if (--this.lifeTicks < 0) {
+                this.putRemoveCause(EntityRemoveEvent.Cause.DESPAWN);
                 this.discard();
             }
         }
@@ -115,7 +_,7 @@
         LivingEntity livingentity = this.getOwner();
         if (p_36945_.isAlive() && !p_36945_.isInvulnerable() && p_36945_ != livingentity) {
             if (livingentity == null) {
-                p_36945_.hurt(this.damageSources().magic(), 6.0F);
+                p_36945_.hurt(this.damageSources().magic().customCausingEntity(this), 6.0F); // CraftBukkit
             } else {
                 if (livingentity.isAlliedTo(p_36945_)) {
                     return;
