--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -75,6 +_,22 @@
       this.m_5602_(p_37052_);
    }
 
+   // Spigot Start - copied from tick
+   @Override
+   public void inactiveTick() {
+      this.f_37022_ += 1;
+
+      if (!this.f_19853_.f_46443_ && this.f_37022_ > this.f_37023_) {
+         // CraftBukkit start
+         if (!org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+            this.m_37080_();
+         }
+         // CraftBukkit end
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected void m_8097_() {
       this.f_19804_.m_135372_(f_37019_, ItemStack.f_41583_);
       this.f_19804_.m_135372_(f_37020_, OptionalInt.empty());
@@ -149,6 +_,13 @@
          this.m_37080_();
       }
 
+   }
+
+   @Override
+   protected void m_6532_(HitResult result) {
+      if (result.m_6662_() == HitResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.m_6532_(result);
+      }
    }
 
    private void m_37080_() {
