--- a/net/minecraft/world/entity/projectile/Fireball.java
+++ b/net/minecraft/world/entity/projectile/Fireball.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.projectile;
 
+import net.minecraft.Util;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
@@ -32,7 +_,7 @@
 
    }
 
-   protected ItemStack getItemRaw() {
+   public ItemStack getItemRaw() {
       return this.getEntityData().get(DATA_ITEM_STACK);
    }
 
@@ -57,6 +_,6 @@
    public void readAdditionalSaveData(CompoundTag p_37009_) {
       super.readAdditionalSaveData(p_37009_);
       ItemStack itemstack = ItemStack.of(p_37009_.getCompound("Item"));
-      this.setItem(itemstack);
+      if (!itemstack.isEmpty()) this.setItem(itemstack); // CraftBukkit - SPIGOT-5474 probably came from bugged earlier versions
    }
 }
