--- a/net/minecraft/world/entity/projectile/LlamaSpit.java
+++ b/net/minecraft/world/entity/projectile/LlamaSpit.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class LlamaSpit extends Projectile {
     public LlamaSpit(EntityType<? extends LlamaSpit> p_37224_, Level p_37225_) {
@@ -43,15 +_,18 @@
         super.tick();
         Vec3 vec3 = this.getDeltaMovement();
         HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
-        this.hitTargetOrDeflectSelf(hitresult);
+        if (hitresult.getType() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult))
+            this.preHitTargetOrDeflectSelf(hitresult);
         double d0 = this.getX() + vec3.x;
         double d1 = this.getY() + vec3.y;
         double d2 = this.getZ() + vec3.z;
         this.updateRotation();
         float f = 0.99F;
         if (this.level().getBlockStates(this.getBoundingBox()).noneMatch(BlockBehaviour.BlockStateBase::isAir)) {
+            this.putRemoveCause(EntityRemoveEvent.Cause.DESPAWN);
             this.discard();
         } else if (this.isInWaterOrBubble()) {
+            this.putRemoveCause(EntityRemoveEvent.Cause.DESPAWN);
             this.discard();
         } else {
             this.setDeltaMovement(vec3.scale(0.99F));
@@ -76,6 +_,7 @@
     protected void onHitBlock(BlockHitResult p_37239_) {
         super.onHitBlock(p_37239_);
         if (!this.level().isClientSide) {
+            this.putRemoveCause(EntityRemoveEvent.Cause.HIT);
             this.discard();
         }
     }
