--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.projectiles.ProjectileSource;
 
 public abstract class Projectile extends Entity implements TraceableEntity {
    @Nullable
@@ -32,6 +_,10 @@
    private boolean leftOwner;
    private boolean hasBeenShot;
 
+   // CraftBukkit start
+   private boolean hitCancelled = false;
+   // CraftBukkit end
+
    protected Projectile(EntityType<? extends Projectile> p_37248_, Level p_37249_) {
       super(p_37248_, p_37249_);
    }
@@ -41,6 +_,7 @@
          this.ownerUUID = p_37263_.getUUID();
          this.cachedOwner = p_37263_;
       }
+      this.projectileSource = (p_37263_ != null && p_37263_.getBukkitEntity() instanceof ProjectileSource) ? (ProjectileSource) p_37263_.getBukkitEntity() : null; // CraftBukkit
 
    }
 
@@ -147,6 +_,16 @@
       this.setDeltaMovement(this.getDeltaMovement().add(vec3.x, p_37252_.onGround() ? 0.0D : vec3.y, vec3.z));
    }
 
+   // CraftBukkit start - call projectile hit event
+   protected void preOnHit(HitResult movingobjectposition) {
+      org.bukkit.event.entity.ProjectileHitEvent event = org.bukkit.craftbukkit.v1_20_R4.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition);
+      this.hitCancelled = event != null && event.isCancelled();
+      if (movingobjectposition.getType() == HitResult.Type.BLOCK || !this.hitCancelled) {
+         this.onHit(movingobjectposition);
+      }
+   }
+   // CraftBukkit end
+
    protected void onHit(HitResult p_37260_) {
       HitResult.Type hitresult$type = p_37260_.getType();
       if (hitresult$type == HitResult.Type.ENTITY) {
@@ -165,6 +_,11 @@
    }
 
    protected void onHitBlock(BlockHitResult p_37258_) {
+      // CraftBukkit start - cancellable hit event
+      if (hitCancelled) {
+         return;
+      }
+      // CraftBukkit end
       BlockState blockstate = this.level().getBlockState(p_37258_.getBlockPos());
       blockstate.onProjectileHit(this.level(), blockstate, p_37258_, this);
    }
@@ -229,7 +_,7 @@
       if (entity instanceof Player) {
          return entity.mayInteract(p_150167_, p_150168_);
       } else {
-         return entity == null || p_150167_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+         return entity == null || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_150167_, entity);
       }
    }
 
