--- a/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -6,12 +_,12 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
 
 public class SmallFireball extends Fireball {
    public SmallFireball(EntityType<? extends SmallFireball> p_37364_, Level p_37365_) {
@@ -24,6 +_,9 @@
 
    public SmallFireball(Level p_37367_, double p_37368_, double p_37369_, double p_37370_, double p_37371_, double p_37372_, double p_37373_) {
       super(EntityType.SMALL_FIREBALL, p_37368_, p_37369_, p_37370_, p_37371_, p_37372_, p_37373_, p_37367_);
+      // CraftBukkit start
+      isIncendiary = !(this.getOwner() instanceof Mob) || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level(), this.getOwner()); // Mohist
+      // CraftBukkit end
    }
 
    protected void onHitEntity(EntityHitResult p_37386_) {
@@ -32,7 +_,14 @@
          Entity entity = p_37386_.getEntity();
          Entity entity1 = this.getOwner();
          int i = entity.getRemainingFireTicks();
-         entity.setSecondsOnFire(5);
+         // CraftBukkit start - Entity damage by entity event + combust event
+         EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+         entity.level().getCraftServer().getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            entity.setSecondsOnFire(event.getDuration(), false);
+         }
+         // CraftBukkit end
          if (!entity.hurt(this.damageSources().fireball(this, entity1), 5.0F)) {
             entity.setRemainingFireTicks(i);
          } else if (entity1 instanceof LivingEntity) {
@@ -46,10 +_,10 @@
       super.onHitBlock(p_37384_);
       if (!this.level().isClientSide) {
          Entity entity = this.getOwner();
-         if (!(entity instanceof Mob) || this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+         if (isIncendiary) { // CraftBukkit
             BlockPos blockpos = p_37384_.getBlockPos().relative(p_37384_.getDirection());
-            if (this.level().isEmptyBlock(blockpos)) {
-               this.level().setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level(), blockpos));
+            if (this.level().isEmptyBlock(blockpos) && !org.bukkit.craftbukkit.v1_20_R3.event.CraftEventFactory.callBlockIgniteEvent(level(), blockpos, this).isCancelled()) {
+               this.level().setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level, blockpos));
             }
          }
 
