--- a/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -16,6 +_,9 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+
 public class SmallFireball extends Fireball {
     public SmallFireball(EntityType<? extends SmallFireball> p_37364_, Level p_37365_) {
         super(p_37364_, p_37365_);
@@ -27,7 +_,10 @@
 
     public SmallFireball(Level p_37375_, double p_37377_, double p_37378_, double p_37379_, Vec3 p_343728_) {
         super(EntityType.SMALL_FIREBALL, p_37377_, p_37378_, p_37379_, p_343728_, p_37375_);
-    }
+        // CraftBukkit start
+        isIncendiary = !(this.getOwner() instanceof Mob) || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level(), this.getOwner()); // Mohist
+        // CraftBukkit end
+   }
 
     @Override
     protected void onHitEntity(EntityHitResult p_37386_) {
@@ -36,7 +_,15 @@
             Entity entity1 = p_37386_.getEntity();
             Entity $$4 = this.getOwner();
             int $$5 = entity1.getRemainingFireTicks();
-            entity1.igniteForSeconds(5.0F);
+
+            // CraftBukkit start - Entity damage by entity event + combust event
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity1.getBukkitEntity(), 5);
+            entity1.level().getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                entity1.igniteForSeconds(event.getDuration(), false);
+            }
+            // CraftBukkit end
             DamageSource $$6 = this.damageSources().fireball(this, $$4);
             if (!entity1.hurt($$6, 5.0F)) {
                 entity1.setRemainingFireTicks($$5);
@@ -51,9 +_,9 @@
         super.onHitBlock(p_37384_);
         if (!this.level().isClientSide) {
             Entity entity = this.getOwner();
-            if (!(entity instanceof Mob) || this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (isIncendiary) { // CraftBukkit
                 BlockPos blockpos = p_37384_.getBlockPos().relative(p_37384_.getDirection());
-                if (this.level().isEmptyBlock(blockpos)) {
+                if (this.level().isEmptyBlock(blockpos) && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level(), blockpos, this).isCancelled()) {
                     this.level().setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level(), blockpos));
                 }
             }
@@ -64,6 +_,7 @@
     protected void onHit(HitResult p_37388_) {
         super.onHit(p_37388_);
         if (!this.level().isClientSide) {
+            this.putRemoveCause(EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
