--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class SpectralArrow extends AbstractArrow {
    public int f_37409_ = 200;
@@ -40,6 +_,7 @@
    protected void m_7761_(LivingEntity p_37422_) {
       super.m_7761_(p_37422_);
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.f_19619_, this.f_37409_, 0);
+      p_37422_.addEffectCause(EntityPotionEffectEvent.Cause.ARROW);  // CraftBukkit // Mohist
       p_37422_.m_147207_(mobeffectinstance, this.m_150173_());
    }
 
