--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -11,7 +_,7 @@
 import net.minecraft.world.level.Level;
 
 public class SpectralArrow extends AbstractArrow {
-   private int f_37409_ = 200;
+   public int f_37409_ = 200;
 
    public SpectralArrow(EntityType<? extends SpectralArrow> p_37411_, Level p_37412_) {
       super(p_37411_, p_37412_);
