--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -40,7 +_,7 @@
    protected void doPostHurtEffects(LivingEntity p_37422_) {
       super.doPostHurtEffects(p_37422_);
       MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
-      p_37422_.addEffect(mobeffectinstance, this.getEffectSource());
+      p_37422_.addEffect(mobeffectinstance, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
    }
 
    public void readAdditionalSaveData(CompoundTag p_37424_) {
