--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -26,7 +_,7 @@
 public class ThrownTrident extends AbstractArrow {
    private static final EntityDataAccessor<Byte> f_37558_ = SynchedEntityData.m_135353_(ThrownTrident.class, EntityDataSerializers.f_135027_);
    private static final EntityDataAccessor<Boolean> f_37554_ = SynchedEntityData.m_135353_(ThrownTrident.class, EntityDataSerializers.f_135035_);
-   private ItemStack f_37555_ = new ItemStack(Items.f_42713_);
+   public ItemStack f_37555_ = new ItemStack(Items.f_42713_);
    private boolean f_37556_;
    public int f_37557_;
 
