--- a/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
+++ b/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
@@ -35,8 +_,15 @@
    protected void onHit(HitResult p_37521_) {
       super.onHit(p_37521_);
       if (this.level() instanceof ServerLevel) {
-         this.level().levelEvent(2002, this.blockPosition(), PotionUtils.getColor(Potions.WATER));
-         int i = 3 + this.level().random.nextInt(5) + this.level().random.nextInt(5);
+         int i = 3 + this.level.random.nextInt(5) + this.level.random.nextInt(5);
+         // CraftBukkit start
+         org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callExpBottleEvent(this, i);
+         i = event.getExperience();
+         if (event.getShowEffect()) {
+            this.level().levelEvent(2002, this.blockPosition(), PotionUtils.getColor(Potions.WATER));
+         }
+         // CraftBukkit end
+
          ExperienceOrb.award((ServerLevel)this.level(), this.position(), i);
          this.discard();
       }
