--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class ThrownEgg extends ThrowableItemProjectile {
     private static final EntityDimensions ZERO_SIZED_DIMENSIONS = EntityDimensions.fixed(0.0F, 0.0F);
@@ -78,6 +_,7 @@
             }
 
             this.level().broadcastEntityEvent(this, (byte)3);
+            this.putRemoveCause(EntityRemoveEvent.Cause.HIT);
             this.discard();
         }
     }
