--- a/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
+++ b/net/minecraft/world/entity/projectile/ThrownExperienceBottle.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class ThrownExperienceBottle extends ThrowableItemProjectile {
     public ThrownExperienceBottle(EntityType<? extends ThrownExperienceBottle> p_37510_, Level p_37511_) {
@@ -38,9 +_,17 @@
     protected void onHit(HitResult p_37521_) {
         super.onHit(p_37521_);
         if (this.level() instanceof ServerLevel) {
-            this.level().levelEvent(2002, this.blockPosition(), PotionContents.getColor(Potions.WATER));
-            int i = 3 + this.level().random.nextInt(5) + this.level().random.nextInt(5);
-            ExperienceOrb.award((ServerLevel)this.level(), this.position(), i);
+            int i = 3 + this.level.random.nextInt(5) + this.level.random.nextInt(5);
+            // CraftBukkit start
+            org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExpBottleEvent(this, p_37521_, i);
+            i = event.getExperience();
+            if (event.getShowEffect()) {
+                this.level().levelEvent(2002, this.blockPosition(), PotionContents.getColor(Potions.WATER));
+            }
+            // CraftBukkit end
+
+            ExperienceOrb.award((ServerLevel) this.level(), this.position(), i);
+            this.putRemoveCause(EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
