--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -17,6 +_,8 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
+import org.bukkit.craftbukkit.v1_20_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class ThrownEnderpearl extends ThrowableItemProjectile {
    public ThrownEnderpearl(EntityType<? extends ThrownEnderpearl> p_37491_, Level p_37492_) {
@@ -48,10 +_,13 @@
          if (entity instanceof ServerPlayer) {
             ServerPlayer serverplayer = (ServerPlayer)entity;
             if (serverplayer.connection.isAcceptingMessages() && serverplayer.level() == this.level() && !serverplayer.isSleeping()) {
+               net.minecraftforge.event.entity.EntityTeleportEvent.EnderPearl event = net.minecraftforge.event.ForgeEventFactory.onEnderPearlLand(serverplayer, this.getX(), this.getY(), this.getZ(), this, 5.0F, p_37504_);
+               if (!event.isCanceled()) { // Don't indent to lower patch size
                if (this.random.nextFloat() < 0.05F && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
                   Endermite endermite = EntityType.ENDERMITE.create(this.level());
                   if (endermite != null) {
                      endermite.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
+                     endermite.spawnReason(CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
                      this.level().addFreshEntity(endermite);
                   }
                }
@@ -62,9 +_,13 @@
                   entity.teleportTo(this.getX(), this.getY(), this.getZ());
                }
 
+               entity.teleportTo(event.getTargetX(), event.getTargetY(), event.getTargetZ());
                entity.resetFallDistance();
-               entity.hurt(this.damageSources().fall(), 5.0F);
+               CraftEventFactory.entityDamage = this;
+               entity.hurt(this.damageSources().fall(), event.getAttackDamage());
                this.level().playSound((Player)null, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_TELEPORT, SoundSource.PLAYERS);
+               CraftEventFactory.entityDamage = null;
+               } //Forge: End
             }
          } else if (entity != null) {
             entity.teleportTo(this.getX(), this.getY(), this.getZ());
@@ -87,12 +_,12 @@
    }
 
    @Nullable
-   public Entity changeDimension(ServerLevel p_37506_) {
+   public Entity changeDimension(ServerLevel p_37506_, net.minecraftforge.common.util.ITeleporter teleporter) {
       Entity entity = this.getOwner();
       if (entity != null && entity.level().dimension() != p_37506_.dimension()) {
          this.setOwner((Entity)null);
       }
 
-      return super.changeDimension(p_37506_);
+      return super.changeDimension(p_37506_, teleporter);
    }
 }
