--- a/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class ThrownTrident extends AbstractArrow {
     private static final EntityDataAccessor<Byte> ID_LOYALTY = SynchedEntityData.defineId(ThrownTrident.class, EntityDataSerializers.BYTE);
@@ -65,7 +_,7 @@
                 if (!this.level().isClientSide && this.pickup == AbstractArrow.Pickup.ALLOWED) {
                     this.spawnAtLocation(this.getPickupItem(), 0.1F);
                 }
-
+                this.putRemoveCause(EntityRemoveEvent.Cause.DROP); // CraftBukkit
                 this.discard();
             } else {
                 this.setNoPhysics(true);
@@ -139,7 +_,7 @@
                 if (lightningbolt != null) {
                     lightningbolt.moveTo(Vec3.atBottomCenterOf(blockpos));
                     lightningbolt.setCause(entity1 instanceof ServerPlayer ? (ServerPlayer)entity1 : null);
-                    this.level().addFreshEntity(lightningbolt);
+                    this.level().addFreshEntity(lightningbolt); // TODO Mohist
                     soundevent = SoundEvents.TRIDENT_THUNDER;
                     f1 = 5.0F;
                 }
