--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -60,8 +_,8 @@
             }
         }
 
-        if (hitresult.getType() != HitResult.Type.MISS && !flag) {
-            this.hitTargetOrDeflectSelf(hitresult);
+        if (hitresult.getType() != HitResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
+            this.preHitTargetOrDeflectSelf(hitresult); // CraftBukkit - projectile hit event
         }
 
         this.checkInsideBlocks();
