--- ../src-base/minecraft/org/bukkit/event/inventory/FurnaceSmeltEvent.java
+++ ../src-work/minecraft/org/bukkit/event/inventory/FurnaceSmeltEvent.java
@@ -22,6 +22,10 @@
         this.cancelled = false;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the smelted ItemStack for this event
      *
@@ -61,8 +65,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
