--- ../src-base/minecraft/org/bukkit/event/inventory/InventoryPickupItemEvent.java
+++ ../src-work/minecraft/org/bukkit/event/inventory/InventoryPickupItemEvent.java
@@ -11,9 +11,9 @@
  */
 public class InventoryPickupItemEvent extends Event implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
-    private boolean cancelled;
     private final Inventory inventory;
     private final Item item;
+    private boolean cancelled;
 
     public InventoryPickupItemEvent(final Inventory inventory, final Item item) {
         super();
@@ -21,6 +21,10 @@
         this.item = item;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the Inventory that picked up the item
      *
@@ -51,8 +55,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
