--- ../src-base/minecraft/org/bukkit/event/inventory/InventoryType.java
+++ ../src-work/minecraft/org/bukkit/event/inventory/InventoryType.java
@@ -6,11 +6,11 @@
      * A chest inventory, with 0, 9, 18, 27, 36, 45, or 54 slots of type
      * CONTAINER.
      */
-    CHEST(27,"Chest"),
+    CHEST(27, "Chest"),
     /**
      * A dispenser inventory, with 9 slots of type CONTAINER.
      */
-    DISPENSER(9,"Dispenser"),
+    DISPENSER(9, "Dispenser"),
     /**
      * A dropper inventory, with 9 slots of type CONTAINER.
      */
@@ -19,45 +19,45 @@
      * A furnace inventory, with a RESULT slot, a CRAFTING slot, and a FUEL
      * slot.
      */
-    FURNACE(3,"Furnace"),
+    FURNACE(3, "Furnace"),
     /**
      * A workbench inventory, with 9 CRAFTING slots and a RESULT slot.
      */
-    WORKBENCH(10,"Crafting"),
+    WORKBENCH(10, "Crafting"),
     /**
      * A player's crafting inventory, with 4 CRAFTING slots and a RESULT slot.
      * Also implies that the 4 ARMOR slots are accessible.
      */
-    CRAFTING(5,"Crafting"),
+    CRAFTING(5, "Crafting"),
     /**
      * An enchantment table inventory, with two CRAFTING slots and three
      * enchanting buttons.
      */
-    ENCHANTING(2,"Enchanting"),
+    ENCHANTING(2, "Enchanting"),
     /**
      * A brewing stand inventory, with one FUEL slot and three CRAFTING slots.
      */
-    BREWING(5,"Brewing"),
+    BREWING(5, "Brewing"),
     /**
      * A player's inventory, with 9 QUICKBAR slots, 27 CONTAINER slots, 4 ARMOR
      * slots and 1 offhand slot. The ARMOR and offhand slots may not be visible
      * to the player, though.
      */
-    PLAYER(41,"Player"),
+    PLAYER(41, "Player"),
     /**
      * The creative mode inventory, with only 9 QUICKBAR slots and nothing
      * else. (The actual creative interface with the items is client-side and
      * cannot be altered by the server.)
      */
-    CREATIVE(9,"Creative"),
+    CREATIVE(9, "Creative"),
     /**
      * The merchant inventory, with 2 TRADE-IN slots, and 1 RESULT slot.
      */
-    MERCHANT(3,"Villager"),
+    MERCHANT(3, "Villager"),
     /**
      * The ender chest inventory, with 27 slots.
      */
-    ENDER_CHEST(27,"Ender Chest"),
+    ENDER_CHEST(27, "Ender Chest"),
     /**
      * An anvil inventory, with 2 CRAFTING slots and 1 RESULT slot
      */
@@ -73,8 +73,7 @@
     /**
      * A shulker box inventory, with 27 slots of type CONTAINER.
      */
-    SHULKER_BOX(27, "Shulker Box"),
-    ;
+    SHULKER_BOX(27, "Shulker Box"),;
 
     private final int size;
     private final String title;
