--- ../src-base/minecraft/org/bukkit/event/entity/FoodLevelChangeEvent.java
+++ ../src-work/minecraft/org/bukkit/event/entity/FoodLevelChangeEvent.java
@@ -17,6 +17,10 @@
         this.level = level;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     @Override
     public HumanEntity getEntity() {
         return (HumanEntity) entity;
@@ -42,7 +46,9 @@
      *     event should be set to
      */
     public void setFoodLevel(int level) {
-        if (level < 0) level = 0;
+        if (level < 0) {
+            level = 0;
+        }
 
         this.level = level;
     }
@@ -59,8 +65,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
