--- ../src-base/minecraft/net/minecraft/entity/EntityList.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityList.java
@@ -61,6 +61,7 @@
 import net.minecraft.entity.projectile.EntitySnowball;
 import net.minecraft.entity.projectile.EntityWitherSkull;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.stats.StatBase;
 import net.minecraft.stats.StatList;
 import net.minecraft.world.World;
@@ -166,6 +167,13 @@
             exception.printStackTrace();
         }
 
+        //Uranium start
+        String entityId=p_75615_0_.getString("id");
+        if(MinecraftServer.entityConfig.entitiesBlacklist.contains(entityId)){
+            logger.warn("Skipping Entity with id " + entityId);
+            return null;
+        }
+        //Uranium end
         if (entity != null)
         {
             try
