--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -26,8 +_,10 @@
    }
 
    public ClientboundInitializeBorderPacket(WorldBorder p_178877_) {
-      this.newCenterX = p_178877_.getCenterX();
-      this.newCenterZ = p_178877_.getCenterZ();
+      // CraftBukkit start - multiply out nether border
+      this.newCenterX = p_178877_.getCenterX() * p_178877_.world.dimensionType().coordinateScale();
+      this.newCenterZ = p_178877_.getCenterZ() * p_178877_.world.dimensionType().coordinateScale();
+      // CraftBukkit end
       this.oldSize = p_178877_.getSize();
       this.newSize = p_178877_.getLerpTarget();
       this.lerpTime = p_178877_.getLerpRemainingTime();
