--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -5,6 +_,18 @@
 import net.minecraft.network.protocol.Packet;
 
 public record ClientboundSystemChatPacket(Component f_237849_, boolean f_240374_) implements Packet<ClientGamePacketListener> {
+
+   // Mohist start
+   public ClientboundSystemChatPacket(String f_237849_, boolean f_240374_) {
+      this(Component.Serializer.m_130701_(f_237849_), f_240374_);
+   }
+   // Mohist end
+
+   // Spigot start
+   public ClientboundSystemChatPacket(net.md_5.bungee.api.chat.BaseComponent[] content, boolean overlay) {
+      this(net.md_5.bungee.chat.ComponentSerializer.toString(content), overlay);
+   }
+   // Spigot end
    public ClientboundSystemChatPacket(FriendlyByteBuf p_237852_) {
       this(p_237852_.m_130238_(), p_237852_.readBoolean());
    }
