--- a/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
@@ -9,6 +_,7 @@
    private final BlockHitResult blockHit;
    private final InteractionHand hand;
    private final int sequence;
+   public long timestamp; // Spigot
 
    public ServerboundUseItemOnPacket(InteractionHand p_238005_, BlockHitResult p_238006_, int p_238007_) {
       this.hand = p_238005_;
@@ -17,6 +_,7 @@
    }
 
    public ServerboundUseItemOnPacket(FriendlyByteBuf p_179796_) {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.hand = p_179796_.readEnum(InteractionHand.class);
       this.blockHit = p_179796_.readBlockHitResult();
       this.sequence = p_179796_.readVarInt();
