--- a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -7,6 +_,7 @@
 public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener> {
    private final InteractionHand hand;
    private final int sequence;
+   public long timestamp; // Spigot
 
    public ServerboundUseItemPacket(InteractionHand p_238011_, int p_238012_) {
       this.hand = p_238011_;
@@ -14,6 +_,7 @@
    }
 
    public ServerboundUseItemPacket(FriendlyByteBuf p_179798_) {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.hand = p_179798_.readEnum(InteractionHand.class);
       this.sequence = p_179798_.readVarInt();
    }
