--- a/net/minecraft/tileentity/EndGatewayTileEntity.java
+++ b/net/minecraft/tileentity/EndGatewayTileEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.tileentity;
 
+import com.mohistmc.util.i18n.i18n;
 import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
@@ -20,6 +_,7 @@
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
@@ -31,14 +_,18 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndGatewayTileEntity extends EndPortalTileEntity implements ITickableTileEntity {
    private static final Logger field_195503_a = LogManager.getLogger();
-   private long field_195504_f;
+   public long field_195504_f;
    private int field_195505_g;
    @Nullable
-   private BlockPos field_195506_h;
-   private boolean field_195507_i;
+   public BlockPos field_195506_h;
+   public boolean field_195507_i;
 
    public EndGatewayTileEntity() {
       super(TileEntityType.field_200991_v);
@@ -148,7 +_,7 @@
    public void func_195496_a(Entity p_195496_1_) {
       if (this.field_145850_b instanceof ServerWorld && !this.func_195500_d()) {
          this.field_195505_g = 100;
-         if (this.field_195506_h == null && this.field_145850_b.func_234923_W_() == World.field_234920_i_) {
+         if (this.field_195506_h == null && this.field_145850_b.getTypeKey() == DimensionType.field_236001_e_) { // CraftBukkit - work in alternate worlds
             this.func_227015_a_((ServerWorld)this.field_145850_b);
          }
 
@@ -171,6 +_,24 @@
                entity = p_195496_1_.func_184208_bv();
             }
 
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (p_195496_1_ instanceof ServerPlayerEntity) {
+               org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer player = (CraftPlayer) p_195496_1_.getBukkitEntity();
+               org.bukkit.Location location = new Location(field_145850_b.getWorld(), (double) blockpos.func_177958_n() + 0.5D, (double) blockpos.func_177956_o() + 0.5D, (double) blockpos.func_177952_p() + 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+               PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+               ((ServerPlayerEntity) p_195496_1_).field_71135_a.teleport(teleEvent.getTo());
+               this.func_195500_d(); // CraftBukkit - call at end of method
+               return;
+            }
+            // CraftBukkit end
+
             entity.func_242279_ag();
             entity.func_223102_j((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p() + 0.5D);
          }
@@ -181,7 +_,7 @@
 
    private BlockPos func_195502_i() {
       BlockPos blockpos = func_195494_a(this.field_145850_b, this.field_195506_h.func_177982_a(0, 2, 0), 5, false);
-      field_195503_a.debug("Best exit position for portal at {} is {}", this.field_195506_h, blockpos);
+      field_195503_a.debug(i18n.get("endgatewaytileentity.1", this.field_195506_h, blockpos));
       return blockpos.func_177984_a();
    }
 
@@ -190,26 +_,26 @@
       Vector3d vector3d1 = vector3d.func_186678_a(1024.0D);
 
       for(int i = 16; func_195495_a(p_227015_1_, vector3d1).func_76625_h() > 0 && i-- > 0; vector3d1 = vector3d1.func_178787_e(vector3d.func_186678_a(-16.0D))) {
-         field_195503_a.debug("Skipping backwards past nonempty chunk at {}", (Object)vector3d1);
+         field_195503_a.debug(i18n.get("endgatewaytileentity.2", (Object)vector3d1));
       }
 
       for(int j = 16; func_195495_a(p_227015_1_, vector3d1).func_76625_h() == 0 && j-- > 0; vector3d1 = vector3d1.func_178787_e(vector3d.func_186678_a(16.0D))) {
-         field_195503_a.debug("Skipping forward past empty chunk at {}", (Object)vector3d1);
+         field_195503_a.debug(i18n.get("endgatewaytileentity.3", (Object)vector3d1));
       }
 
-      field_195503_a.debug("Found chunk at {}", (Object)vector3d1);
+      field_195503_a.debug(i18n.get("endgatewaytileentity.4", (Object)vector3d1));
       Chunk chunk = func_195495_a(p_227015_1_, vector3d1);
       this.field_195506_h = func_195498_a(chunk);
       if (this.field_195506_h == null) {
          this.field_195506_h = new BlockPos(vector3d1.field_72450_a + 0.5D, 75.0D, vector3d1.field_72449_c + 0.5D);
-         field_195503_a.debug("Failed to find suitable block, settling on {}", (Object)this.field_195506_h);
+         field_195503_a.debug(i18n.get("endgatewaytileentity.5", (Object)this.field_195506_h));
          Features.field_243945_e.func_242765_a(p_227015_1_, p_227015_1_.func_72863_F().func_201711_g(), new Random(this.field_195506_h.func_218275_a()), this.field_195506_h);
       } else {
-         field_195503_a.debug("Found block at {}", (Object)this.field_195506_h);
+         field_195503_a.debug(i18n.get("endgatewaytileentity.6", (Object)this.field_195506_h));
       }
 
       this.field_195506_h = func_195494_a(p_227015_1_, this.field_195506_h, 16, true);
-      field_195503_a.debug("Creating portal at {}", (Object)this.field_195506_h);
+      field_195503_a.debug(i18n.get("endgatewaytileentity.7", (Object)this.field_195506_h));
       this.field_195506_h = this.field_195506_h.func_177981_b(10);
       this.func_227016_a_(p_227015_1_, this.field_195506_h);
       this.func_70296_d();
