--- a/net/minecraft/tileentity/ShulkerBoxTileEntity.java
+++ b/net/minecraft/tileentity/ShulkerBoxTileEntity.java
@@ -30,6 +30,8 @@
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_15_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public class ShulkerBoxTileEntity extends LockableLootTileEntity implements ISidedInventory, ITickableTileEntity {
    private static final int[] field_190595_a = IntStream.range(0, 27).toArray();
@@ -42,6 +44,41 @@
    private DyeColor field_190602_l;
    private boolean field_204400_o;
 
+   // CraftBukkit start - add fields and methods
+   private int maxStack = MAX_STACK;
+   public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+
+   @Override
+   public int func_70297_j_() {
+      return maxStack;
+   }
+
+   @Override
+   public List<ItemStack> getContents() {
+      return field_190596_f;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+      this.transaction.add(who);
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+      this.transaction.remove(who);
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return this.transaction;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+      this.maxStack = size;
+   }
+   // CraftBukkit end
+
    public ShulkerBoxTileEntity(@Nullable DyeColor p_i47242_1_) {
       super(TileEntityType.field_200993_x);
       this.field_190602_l = p_i47242_1_;
@@ -280,6 +317,11 @@
       return new ShulkerBoxContainer(p_213906_1_, p_213906_2_, this);
    }
 
+   @Override
+   protected net.minecraftforge.items.IItemHandler createUnSidedHandler() {
+      return new net.minecraftforge.items.wrapper.SidedInvWrapper(this, Direction.UP);
+   }
+
    public static enum AnimationStatus {
       CLOSED,
       OPENING,
