From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Mon, 22 Jun 2020 04:42:51 -0700
Subject: [PATCH] Fix error in light engine update

Fixes the light engine using high amounts of CPU

diff --git a/src/main/java/net/minecraft/server/LightEngineThreaded.java b/src/main/java/net/minecraft/server/LightEngineThreaded.java
index 604fe85313969bf1ad5b5ef75b46f9a23f79764a..bf22ee3d7a56c32d8c0666001b2af842f575be44 100644
--- a/src/main/java/net/minecraft/server/LightEngineThreaded.java
+++ b/src/main/java/net/minecraft/server/LightEngineThreaded.java
@@ -282,17 +282,24 @@ public class LightEngineThreaded extends LightEngine implements AutoCloseable {
     }
 
     // Paper start - replace impl
+    private final java.util.List<Runnable> pre = new java.util.ArrayList<>(); // Tuinity - stop allocating
+    private final java.util.List<Runnable> post = new java.util.ArrayList<>(); // Tuinity - stop allocating
     private void b() {
-        java.util.List<Runnable> pre = new java.util.ArrayList<>();
-        java.util.List<Runnable> post = new java.util.ArrayList<>();
+        // Tuinity - stop allocating
+        boolean ran = false; // Tuinity - fix error in updating
         int i = Math.min(queue.size(), 4);
         while (i-- > 0 && queue.poll(pre, post)) {
             pre.forEach(Runnable::run);
             pre.clear();
-            super.a(Integer.MAX_VALUE, true, true);
+            super.a(Integer.MAX_VALUE, true, true); ran = true; // Tuinity - fix error in updating
             post.forEach(Runnable::run);
             post.clear();
         }
+        // Tuinity start - fix error in updating
+        if (!ran) {
+            super.a(Integer.MAX_VALUE, true, true);
+        }
+        // Tuinity end - fix error in updating
         // Paper end
     }
 
