--- ../src-base/minecraft/net/minecraft/server/management/UserListOps.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserListOps.java
@@ -35,7 +35,8 @@
 
     protected String func_152699_b(GameProfile p_152699_1_)
     {
-        return p_152699_1_.getId().toString();
+        if (p_152699_1_ == null) return "invalid"; // Cauldron - handle GameProfiles with no ID
+        return p_152699_1_.getName().toLowerCase();
     }
 
     public GameProfile func_152700_a(String p_152700_1_)
