--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -5,11 +_,12 @@
 import net.minecraft.core.Vec3i;
 import net.minecraft.util.Unit;
 import net.minecraft.world.level.ChunkPos;
+import org.bukkit.plugin.Plugin;
 
 public class TicketType<T> {
     private final String name;
     private final Comparator<T> comparator;
-    public long timeout;
+   	public long timeout;
     public static final TicketType<Unit> START = create("start", (p_9471_, p_9472_) -> 0);
     public static final TicketType<Unit> DRAGON = create("dragon", (p_9460_, p_9461_) -> 0);
     public static final TicketType<ChunkPos> PLAYER = create("player", Comparator.comparingLong(ChunkPos::toLong));
@@ -18,6 +_,9 @@
     public static final TicketType<BlockPos> PORTAL = create("portal", Vec3i::compareTo, 300);
     public static final TicketType<Integer> POST_TELEPORT = create("post_teleport", Integer::compareTo, 5);
     public static final TicketType<ChunkPos> UNKNOWN = create("unknown", Comparator.comparingLong(ChunkPos::toLong), 1);
+
+   	public static final TicketType<Unit> PLUGIN = create("plugin", (a, b) -> 0); // CraftBukkit
+   	public static final TicketType<Plugin> PLUGIN_TICKET = create("plugin_ticket", Comparator.comparing(plugin -> plugin.getClass().getName())); // CraftBukkit
 
     public static <T> TicketType<T> create(String p_9463_, Comparator<T> p_9464_) {
         return new TicketType<>(p_9463_, p_9464_, 0L);
