--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -190,7 +_,7 @@
       if (blockstate.m_60795_()) {
          return false;
       } else {
-         if (p_9551_) {
+         if (false) { // CraftBukkit - SPIGOT-6833: Do not drop during world generation
             BlockEntity blockentity = blockstate.m_155947_() ? this.m_7702_(p_9550_) : null;
             Block.m_49881_(blockstate, this.f_9479_, p_9550_, blockentity, p_9552_, ItemStack.f_41583_);
          }
@@ -226,7 +_,7 @@
          }
 
          if (blockstate.m_155947_()) {
-            f_9474_.warn("Tried to access a block entity before it was created. {}", (Object)p_9582_);
+            f_9474_.debug("Tried to access a block entity before it was created. {}", (Object)p_9582_);
          }
 
          return null;
@@ -297,10 +_,16 @@
    }
 
    public boolean m_7967_(Entity p_9580_) {
+      if (p_9580_ instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
       int i = SectionPos.m_123171_(p_9580_.m_146903_());
       int j = SectionPos.m_123171_(p_9580_.m_146907_());
       this.m_6325_(i, j).m_6286_(p_9580_);
       return true;
+   }
+
+   @Override
+   public boolean addFreshEntity(Entity pEntity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      return m_7967_(pEntity);
    }
 
    public boolean m_7471_(BlockPos p_9547_, boolean p_9548_) {
