--- a/net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ b/net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -58,6 +_,7 @@
       if (p_44639_ instanceof LivingEntity livingentity) {
          if (this.type == 2 && p_44640_ > 0 && livingentity.getMobType() == MobType.ARTHROPOD) {
             int i = 20 + p_44638_.getRandom().nextInt(10 * p_44640_);
+            livingentity.addEffectCause.set(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // Mohist
             livingentity.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, i, 3));
          }
       }
