--- a/net/minecraft/world/item/crafting/BlastingRecipe.java
+++ b/net/minecraft/world/item/crafting/BlastingRecipe.java
@@ -2,6 +_,11 @@
 
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftBlastingRecipe;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe;
+import org.bukkit.inventory.Recipe;
 
 public class BlastingRecipe extends AbstractCookingRecipe {
    public BlastingRecipe(String p_251053_, CookingBookCategory p_249936_, Ingredient p_251550_, ItemStack p_251027_, float p_250843_, int p_249841_) {
@@ -15,4 +_,17 @@
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44092_;
    }
+
+   // CraftBukkit start
+   @Override
+   public Recipe toBukkitRecipe(NamespacedKey id) {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+
+      CraftBlastingRecipe recipe = new CraftBlastingRecipe(id, result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+      recipe.setCategory(CraftRecipe.getCategory(this.m_245534_()));
+
+      return recipe;
+   }
+   // CraftBukkit end
 }
