--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -2,6 +_,11 @@
 
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftCampfireRecipe;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R2.inventory.CraftRecipe;
+import org.bukkit.inventory.Recipe;
 
 public class CampfireCookingRecipe extends AbstractCookingRecipe {
    public CampfireCookingRecipe(String p_250140_, CookingBookCategory p_251808_, Ingredient p_249826_, ItemStack p_251839_, float p_251432_, int p_251471_) {
@@ -15,4 +_,17 @@
    public RecipeSerializer<?> m_7707_() {
       return RecipeSerializer.f_44094_;
    }
+
+   // CraftBukkit start
+   @Override
+   public Recipe toBukkitRecipe(NamespacedKey id) {
+      CraftItemStack result = CraftItemStack.asCraftMirror(this.f_43730_);
+
+      CraftCampfireRecipe recipe = new CraftCampfireRecipe(id, result, CraftRecipe.toBukkit(this.f_43729_), this.f_43731_, this.f_43732_);
+      recipe.setGroup(this.f_43728_);
+      recipe.setCategory(CraftRecipe.getCategory(this.m_245534_()));
+      
+      return recipe;
+   }
+   // CraftBukkit end
 }
