--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -1,7 +_,9 @@
 package net.minecraft.world.item.crafting;
 
 import net.minecraft.core.HolderLookup;
+import net.minecraft.core.RegistryAccess;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.NamespacedKey;
 
 public abstract class CustomRecipe implements CraftingRecipe {
     private final CraftingBookCategory category;
@@ -24,4 +_,11 @@
     public CraftingBookCategory category() {
         return this.category;
     }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(NamespacedKey id) {
+      return new org.bukkit.craftbukkit.inventory.CraftComplexRecipe(id,this);
+   }
+   // CraftBukkit end
 }
