--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -1,16 +_,13 @@
 package net.minecraft.world.item.crafting;
 
-import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.Optional;
 import java.util.stream.Stream;
 import net.minecraft.core.Holder;
-import net.minecraft.core.HolderLookup;
-import net.minecraft.core.component.DataComponents;
+import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.network.RegistryFriendlyByteBuf;
-import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.world.Container;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
@@ -20,115 +_,115 @@
 import net.minecraft.world.item.armortrim.TrimPattern;
 import net.minecraft.world.item.armortrim.TrimPatterns;
 import net.minecraft.world.level.Level;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.inventory.CraftSmithingTrimRecipe;
 
 public class SmithingTrimRecipe implements SmithingRecipe {
-    final Ingredient template;
-    final Ingredient base;
-    final Ingredient addition;
-
-    public SmithingTrimRecipe(Ingredient p_267298_, Ingredient p_266862_, Ingredient p_267050_) {
-        this.template = p_267298_;
-        this.base = p_266862_;
-        this.addition = p_267050_;
-    }
-
-    @Override
-    public boolean matches(Container p_267224_, Level p_266798_) {
-        return this.template.test(p_267224_.getItem(0)) && this.base.test(p_267224_.getItem(1)) && this.addition.test(p_267224_.getItem(2));
-    }
-
-    @Override
-    public ItemStack assemble(Container p_334062_, HolderLookup.Provider p_330268_) {
-        ItemStack itemstack = p_334062_.getItem(1);
-        if (this.base.test(itemstack)) {
-            Optional<Holder.Reference<TrimMaterial>> optional = TrimMaterials.getFromIngredient(p_330268_, p_334062_.getItem(2));
-            Optional<Holder.Reference<TrimPattern>> optional1 = TrimPatterns.getFromTemplate(p_330268_, p_334062_.getItem(0));
-            if (optional.isPresent() && optional1.isPresent()) {
-                ArmorTrim armortrim = itemstack.get(DataComponents.TRIM);
-                if (armortrim != null && armortrim.hasPatternAndMaterial(optional1.get(), optional.get())) {
-                    return ItemStack.EMPTY;
-                }
-
-                ItemStack itemstack1 = itemstack.copyWithCount(1);
-                itemstack1.set(DataComponents.TRIM, new ArmorTrim(optional.get(), optional1.get()));
-                return itemstack1;
-            }
-        }
-
-        return ItemStack.EMPTY;
-    }
-
-    @Override
-    public ItemStack getResultItem(HolderLookup.Provider p_335603_) {
-        ItemStack itemstack = new ItemStack(Items.IRON_CHESTPLATE);
-        Optional<Holder.Reference<TrimPattern>> optional = p_335603_.lookupOrThrow(Registries.TRIM_PATTERN).listElements().findFirst();
-        Optional<Holder.Reference<TrimMaterial>> optional1 = p_335603_.lookupOrThrow(Registries.TRIM_MATERIAL).get(TrimMaterials.REDSTONE);
-        if (optional.isPresent() && optional1.isPresent()) {
-            itemstack.set(DataComponents.TRIM, new ArmorTrim(optional1.get(), optional.get()));
-        }
-
-        return itemstack;
-    }
-
-    @Override
-    public boolean isTemplateIngredient(ItemStack p_266762_) {
-        return this.template.test(p_266762_);
-    }
-
-    @Override
-    public boolean isBaseIngredient(ItemStack p_266795_) {
-        return this.base.test(p_266795_);
-    }
-
-    @Override
-    public boolean isAdditionIngredient(ItemStack p_266922_) {
-        return this.addition.test(p_266922_);
-    }
-
-    @Override
-    public RecipeSerializer<?> getSerializer() {
-        return RecipeSerializer.SMITHING_TRIM;
-    }
-
-    @Override
-    public boolean isIncomplete() {
-        return Stream.of(this.template, this.base, this.addition).anyMatch(Ingredient::isEmpty);
-    }
-
-    public static class Serializer implements RecipeSerializer<SmithingTrimRecipe> {
-        private static final MapCodec<SmithingTrimRecipe> CODEC = RecordCodecBuilder.mapCodec(
-            p_297394_ -> p_297394_.group(
-                        Ingredient.CODEC.fieldOf("template").forGetter(p_298441_ -> p_298441_.template),
-                        Ingredient.CODEC.fieldOf("base").forGetter(p_297838_ -> p_297838_.base),
-                        Ingredient.CODEC.fieldOf("addition").forGetter(p_299309_ -> p_299309_.addition)
-                    )
-                    .apply(p_297394_, SmithingTrimRecipe::new)
-        );
-        public static final StreamCodec<RegistryFriendlyByteBuf, SmithingTrimRecipe> STREAM_CODEC = StreamCodec.of(
-            SmithingTrimRecipe.Serializer::toNetwork, SmithingTrimRecipe.Serializer::fromNetwork
-        );
-
-        @Override
-        public MapCodec<SmithingTrimRecipe> codec() {
-            return CODEC;
-        }
-
-        @Override
-        public StreamCodec<RegistryFriendlyByteBuf, SmithingTrimRecipe> streamCodec() {
-            return STREAM_CODEC;
-        }
-
-        private static SmithingTrimRecipe fromNetwork(RegistryFriendlyByteBuf p_333367_) {
-            Ingredient ingredient = Ingredient.CONTENTS_STREAM_CODEC.decode(p_333367_);
-            Ingredient ingredient1 = Ingredient.CONTENTS_STREAM_CODEC.decode(p_333367_);
-            Ingredient ingredient2 = Ingredient.CONTENTS_STREAM_CODEC.decode(p_333367_);
-            return new SmithingTrimRecipe(ingredient, ingredient1, ingredient2);
-        }
-
-        private static void toNetwork(RegistryFriendlyByteBuf p_335485_, SmithingTrimRecipe p_335201_) {
-            Ingredient.CONTENTS_STREAM_CODEC.encode(p_335485_, p_335201_.template);
-            Ingredient.CONTENTS_STREAM_CODEC.encode(p_335485_, p_335201_.base);
-            Ingredient.CONTENTS_STREAM_CODEC.encode(p_335485_, p_335201_.addition);
-        }
-    }
+   final Ingredient template;
+   final Ingredient base;
+   final Ingredient addition;
+
+   public SmithingTrimRecipe(Ingredient p_267298_, Ingredient p_266862_, Ingredient p_267050_) {
+      this.template = p_267298_;
+      this.base = p_266862_;
+      this.addition = p_267050_;
+   }
+
+   public boolean matches(Container p_267224_, Level p_266798_) {
+      return this.template.test(p_267224_.getItem(0)) && this.base.test(p_267224_.getItem(1)) && this.addition.test(p_267224_.getItem(2));
+   }
+
+   public ItemStack assemble(Container p_267320_, RegistryAccess p_267280_) {
+      ItemStack itemstack = p_267320_.getItem(1);
+      if (this.base.test(itemstack)) {
+         Optional<Holder.Reference<TrimMaterial>> optional = TrimMaterials.getFromIngredient(p_267280_, p_267320_.getItem(2));
+         Optional<Holder.Reference<TrimPattern>> optional1 = TrimPatterns.getFromTemplate(p_267280_, p_267320_.getItem(0));
+         if (optional.isPresent() && optional1.isPresent()) {
+            Optional<ArmorTrim> optional2 = ArmorTrim.getTrim(p_267280_, itemstack, false);
+            if (optional2.isPresent() && optional2.get().hasPatternAndMaterial(optional1.get(), optional.get())) {
+               return ItemStack.EMPTY;
+            }
+
+            ItemStack itemstack1 = itemstack.copy();
+            itemstack1.setCount(1);
+            ArmorTrim armortrim = new ArmorTrim(optional.get(), optional1.get());
+            if (ArmorTrim.setTrim(p_267280_, itemstack1, armortrim)) {
+               return itemstack1;
+            }
+         }
+      }
+
+      return ItemStack.EMPTY;
+   }
+
+   public ItemStack getResultItem(RegistryAccess p_266948_) {
+      ItemStack itemstack = new ItemStack(Items.IRON_CHESTPLATE);
+      Optional<Holder.Reference<TrimPattern>> optional = p_266948_.registryOrThrow(Registries.TRIM_PATTERN).holders().findFirst();
+      if (optional.isPresent()) {
+         Optional<Holder.Reference<TrimMaterial>> optional1 = p_266948_.registryOrThrow(Registries.TRIM_MATERIAL).getHolder(TrimMaterials.REDSTONE);
+         if (optional1.isPresent()) {
+            ArmorTrim armortrim = new ArmorTrim(optional1.get(), optional.get());
+            ArmorTrim.setTrim(p_266948_, itemstack, armortrim);
+         }
+      }
+
+      return itemstack;
+   }
+
+   public boolean isTemplateIngredient(ItemStack p_266762_) {
+      return this.template.test(p_266762_);
+   }
+
+   public boolean isBaseIngredient(ItemStack p_266795_) {
+      return this.base.test(p_266795_);
+   }
+
+   public boolean isAdditionIngredient(ItemStack p_266922_) {
+      return this.addition.test(p_266922_);
+   }
+
+   public RecipeSerializer<?> getSerializer() {
+      return RecipeSerializer.SMITHING_TRIM;
+   }
+
+   public boolean isIncomplete() {
+      return Stream.of(this.template, this.base, this.addition).anyMatch(net.minecraftforge.common.ForgeHooks::hasNoElements);
+   }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe(NamespacedKey id) {
+      return new CraftSmithingTrimRecipe(id, CraftRecipe.toBukkit(this.template), CraftRecipe.toBukkit(this.base), CraftRecipe.toBukkit(this.addition));
+   }
+   // CraftBukkit end
+
+   public static class Serializer implements RecipeSerializer<SmithingTrimRecipe> {
+      private static final Codec<SmithingTrimRecipe> CODEC = RecordCodecBuilder.create((p_297394_) -> {
+         return p_297394_.group(Ingredient.CODEC.fieldOf("template").forGetter((p_298441_) -> {
+            return p_298441_.template;
+         }), Ingredient.CODEC.fieldOf("base").forGetter((p_297838_) -> {
+            return p_297838_.base;
+         }), Ingredient.CODEC.fieldOf("addition").forGetter((p_299309_) -> {
+            return p_299309_.addition;
+         })).apply(p_297394_, SmithingTrimRecipe::new);
+      });
+
+      public Codec<SmithingTrimRecipe> codec() {
+         return CODEC;
+      }
+
+      public SmithingTrimRecipe fromNetwork(FriendlyByteBuf p_266888_) {
+         Ingredient ingredient = Ingredient.fromNetwork(p_266888_);
+         Ingredient ingredient1 = Ingredient.fromNetwork(p_266888_);
+         Ingredient ingredient2 = Ingredient.fromNetwork(p_266888_);
+         return new SmithingTrimRecipe(ingredient, ingredient1, ingredient2);
+      }
+
+      public void toNetwork(FriendlyByteBuf p_266901_, SmithingTrimRecipe p_266893_) {
+         p_266893_.template.toNetwork(p_266901_);
+         p_266893_.base.toNetwork(p_266901_);
+         p_266893_.addition.toNetwork(p_266901_);
+      }
+   }
 }
