--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -1,9 +_,14 @@
 package net.minecraft.world.item.crafting;
 
+import com.mohistmc.mohist.bukkit.inventory.MohistSpecialRecipe;
 import net.minecraft.world.Container;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.inventory.CraftStonecuttingRecipe;
 
 public class StonecutterRecipe extends SingleItemRecipe {
     public StonecutterRecipe(String p_44479_, Ingredient p_44480_, ItemStack p_301701_) {
@@ -19,4 +_,19 @@
     public ItemStack getToastSymbol() {
         return new ItemStack(Blocks.STONECUTTER);
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe(NamespacedKey id) {
+        if (this.result.isEmpty()) {
+            return new MohistSpecialRecipe(id, this);
+        }
+        CraftItemStack result = CraftItemStack.asCraftMirror(this.result);
+
+        CraftStonecuttingRecipe recipe = new CraftStonecuttingRecipe(id, result, CraftRecipe.toBukkit(this.ingredient));
+        recipe.setGroup(this.group);
+
+        return recipe;
+    }
+    // CraftBukkit end
 }
