--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -30,7 +_,7 @@
    }
 
    public Component canPlayerLogin(SocketAddress p_120007_, GameProfile p_120008_) {
-      return (Component)(this.getServer().isSingleplayerOwner(p_120008_) && this.getPlayerByName(p_120008_.getName()) != null ? Component.translatable("multiplayer.disconnect.name_taken") : super.canPlayerLogin(p_120007_, p_120008_));
+      return (Component)(this.getServer().isSingleplayerOwner(p_120008_) && this.getPlayerByName(p_120008_.getName()) != null ? Component.translatable("multiplayer.disconnect.name_taken") : super.canPlayerLogin(null, p_120008_));
    }
 
    public IntegratedServer getServer() {
