--- a/net/minecraft/village/PointOfInterestData.java
+++ b/net/minecraft/village/PointOfInterestData.java
@@ -3,6 +_,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+import com.mohistmc.util.i18n.i18n;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import it.unimi.dsi.fastutil.shorts.Short2ObjectMap;
@@ -102,7 +_,7 @@
          field_218255_a.error("POI data mismatch: never registered at " + p_218248_1_);
       } else {
          this.field_218257_c.get(pointofinterest.func_218260_g()).remove(pointofinterest);
-         field_218255_a.debug("Removed POI of type {} @ {}", pointofinterest::func_218260_g, pointofinterest::func_218261_f);
+         field_218255_a.debug(i18n.get("pointofinterestdata.1", pointofinterest.func_218260_g(), pointofinterest.func_218261_f()));
          this.field_218258_d.run();
       }
    }
