--- ../src-base/minecraft/net/minecraftforge/common/crafting/CraftingHelper.java
+++ ../src-work/minecraft/net/minecraftforge/common/crafting/CraftingHelper.java
@@ -725,12 +725,12 @@
                 }
                 catch (JsonParseException e)
                 {
-                    FMLLog.log.error("Parsing error loading recipe {}", key, e);
+                    FMLLog.log.debug("Parsing error loading recipe {}", key, e);
                     return false;
                 }
                 catch (IOException e)
                 {
-                    FMLLog.log.error("Couldn't read recipe {} from {}", key, file, e);
+                    FMLLog.log.debug("Couldn't read recipe {} from {}", key, file, e);
                     return false;
                 }
                 return true;
