--- ../src-base/minecraft/net/minecraftforge/fml/common/network/handshake/FMLHandshakeServerState.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/network/handshake/FMLHandshakeServerState.java
@@ -19,6 +19,8 @@
 
 package net.minecraftforge.fml.common.network.handshake;
 
+import red.mohist.api.PlayerAPI;
+import red.mohist.i18n.Message;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelHandlerContext;
 
@@ -66,7 +68,12 @@
             FMLHandshakeMessage.ModList client = (FMLHandshakeMessage.ModList)msg;
             NetworkDispatcher dispatcher = ctx.channel().attr(NetworkDispatcher.FML_DISPATCHER).get();
             dispatcher.setModList(client.modList());
-            FMLLog.log.info("Client attempting to join with {} mods : {}", client.modListSize(), client.modListAsString());
+            if (client.modListSize() > 0) {
+                PlayerAPI.mods.put(dispatcher.player, client.modListSize());
+                PlayerAPI.modlist.put(dispatcher.player, client.modListAsString());
+            }
+            FMLLog.log.info(Message.getFormatString(Message.client_join_mods, new Object[]{ client.modListSize() }));
+            FMLLog.log.debug("Client attempting to join with {} mods : {}", client.modListSize(), client.modListAsString());
             String modRejections = FMLNetworkHandler.checkModList(client, Side.CLIENT);
             if (modRejections != null)
             {
