--- ../src-base/minecraft/net/minecraftforge/fml/common/network/handshake/NetworkDispatcher.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/network/handshake/NetworkDispatcher.java
@@ -103,7 +103,7 @@
     public static final AttributeKey<Map<ResourceLocation, ForgeRegistry.Snapshot>> FML_GAMEDATA_SNAPSHOT = AttributeKey.valueOf("fml:gameDataSnapshot");
     public final NetworkManager manager;
     private final PlayerList scm;
-    private EntityPlayerMP player;
+    public EntityPlayerMP player;
     private ConnectionState state;
     private ConnectionType connectionType;
     private final Side side;
@@ -208,6 +208,10 @@
                 if (this.player.connection != this) return;
                 super.update();
             }
+
+            public void sendPacket(Packet<?> packetIn) {
+                super.sendPacket(packetIn);
+            }
         };
         this.netHandler = serverHandler;
         // NULL the play server here - we restore it further on. If not, there are packets sent before the login
@@ -597,7 +601,7 @@
             }
             else
             {
-                FMLLog.log.error("NetworkDispatcher exception", cause);
+                FMLLog.log.debug("NetworkDispatcher exception", cause);
             }
         }
         super.exceptionCaught(ctx, cause);
