--- a/net/minecraft/world/level/chunk/storage/SectionStorage.java
+++ b/net/minecraft/world/level/chunk/storage/SectionStorage.java
@@ -40,7 +_,6 @@
     private final Function<Runnable, Codec<R>> codec;
     private final Function<Runnable, R> factory;
     private final RegistryAccess registryAccess;
-    private final ChunkIOErrorReporter errorReporter;
     protected final LevelHeightAccessor levelHeightAccessor;
 
     public SectionStorage(
@@ -48,14 +_,12 @@
         Function<Runnable, Codec<R>> p_223510_,
         Function<Runnable, R> p_223511_,
         RegistryAccess p_223515_,
-        ChunkIOErrorReporter p_345160_,
         LevelHeightAccessor p_223516_
     ) {
         this.simpleRegionStorage = p_335141_;
         this.codec = p_223510_;
         this.factory = p_223511_;
         this.registryAccess = p_223515_;
-        this.errorReporter = p_345160_;
         this.levelHeightAccessor = p_223516_;
     }
 
@@ -121,13 +_,12 @@
     }
 
     private CompletableFuture<Optional<CompoundTag>> tryRead(ChunkPos p_223533_) {
-        return this.simpleRegionStorage.read(p_223533_).exceptionally(p_341893_ -> {
-            if (p_341893_ instanceof IOException ioexception) {
+        return this.simpleRegionStorage.read(p_223533_).exceptionally(p_223526_ -> {
+            if (p_223526_ instanceof IOException ioexception) {
                 LOGGER.error("Error reading chunk {} data from disk", p_223533_, ioexception);
-                this.errorReporter.reportChunkLoadFailure(ioexception, this.simpleRegionStorage.storageInfo(), p_223533_);
                 return Optional.empty();
             } else {
-                throw new CompletionException(p_341893_);
+                throw new CompletionException(p_223526_);
             }
         });
     }
@@ -166,10 +_,7 @@
         Dynamic<Tag> dynamic = this.writeColumn(p_63826_, registryops);
         Tag tag = dynamic.getValue();
         if (tag instanceof CompoundTag) {
-            this.simpleRegionStorage.write(p_63826_, (CompoundTag)tag).exceptionally(p_341891_ -> {
-                this.errorReporter.reportChunkSaveFailure(p_341891_, this.simpleRegionStorage.storageInfo(), p_63826_);
-                return null;
-            });
+            this.simpleRegionStorage.write(p_63826_, (CompoundTag)tag);
         } else {
             LOGGER.error("Expected compound tag, got {}", tag);
         }
