--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -99,6 +_,7 @@
    }, WorldPresets.f_226437_.m_135782_().toString()));
    @Nullable
    private WorldGenSettings f_139727_;
+   public final boolean debug = this.m_139836_("debug", false); // CraftBukkit
 
    public DedicatedServerProperties(Properties p_180926_) {
       super(p_180926_);
