--- ../src-base/minecraft/net/minecraft/network/rcon/RConThreadClient.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConThreadClient.java
@@ -40,25 +40,20 @@
 
     public void run()
     {
-        while (true)
-        {
-            try
-            {
-                if (!this.field_72619_a)
-                {
+        try {
+            while (true) {
+                if (!this.field_72619_a || field_72659_h == null) {
                     return;
                 }
 
                 BufferedInputStream bufferedinputstream = new BufferedInputStream(this.field_72659_h.getInputStream());
                 int i = bufferedinputstream.read(this.field_72660_i, 0, 1460);
 
-                if (10 <= i)
-                {
+                if (10 <= i) {
                     int j = 0;
                     int k = RConUtils.func_72665_b(this.field_72660_i, 0, i);
 
-                    if (k != i - 4)
-                    {
+                    if (k != i - 4) {
                         return;
                     }
 
@@ -68,68 +63,58 @@
                     int i1 = RConUtils.func_72662_b(this.field_72660_i, j);
                     j = j + 4;
 
-                    switch (i1)
-                    {
+                    switch (i1) {
                         case 2:
 
-                            if (this.field_72657_g)
-                            {
+                            if (this.field_72657_g) {
                                 String s1 = RConUtils.func_72661_a(this.field_72660_i, j, i);
 
-                                try
-                                {
+                                try {
                                     this.func_72655_a(l, this.field_72617_b.func_71252_i(s1));
-                                }
-                                catch (Exception exception)
-                                {
+                                } catch (Exception exception) {
                                     this.func_72655_a(l, "Error executing: " + s1 + " (" + exception.getMessage() + ")");
                                 }
 
-                                continue;
+                                break;
                             }
 
                             this.func_72656_f();
-                            continue;
+                            break;
                         case 3:
                             String s = RConUtils.func_72661_a(this.field_72660_i, j, i);
                             int j1 = j + s.length();
 
-                            if (!s.isEmpty() && s.equals(this.field_72658_j))
-                            {
+                            if (!s.isEmpty() && s.equals(this.field_72658_j)) {
                                 this.field_72657_g = true;
                                 this.func_72654_a(l, 2, "");
-                                continue;
+                                break;
                             }
 
                             this.field_72657_g = false;
                             this.func_72656_f();
-                            continue;
+                            break;
                         default:
                             this.func_72655_a(l, String.format("Unknown request %s", Integer.toHexString(i1)));
-                            continue;
+                            break;
                     }
                 }
             }
-            catch (SocketTimeoutException var17)
-            {
-                return;
-            }
-            catch (IOException var18)
-            {
-                return;
-            }
-            catch (Exception exception1)
-            {
-                field_164005_h.error("Exception whilst parsing RCON input", (Throwable)exception1);
-                return;
-            }
-            finally
-            {
-                this.func_72653_g();
-            }
-
+        }
+        catch (SocketTimeoutException var17)
+        {
             return;
         }
+        catch (IOException var18)
+        {
+            return;
+        }
+        catch (Exception exception1)
+        {
+            field_164005_h.error("Exception whilst parsing RCON input", (Throwable)exception1);
+        }
+        finally {
+            this.func_72653_g();
+        }
     }
 
     private void func_72654_a(int p_72654_1_, int p_72654_2_, String p_72654_3_) throws IOException
