--- a/net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -9,20 +_,24 @@
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.DamageTypeTags;
 import net.minecraft.world.damagesource.DamageSource;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.AbstractArrow;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Explosion;
+import net.minecraft.world.level.ExplosionDamageCalculator;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class MinecartTNT extends AbstractMinecart {
-    private static final byte EVENT_PRIME = 10;
-    private int fuse = -1;
+   	private static final byte EVENT_PRIME = 10;
+   	public int fuse = -1;
 
     public MinecartTNT(EntityType<? extends MinecartTNT> p_38649_, Level p_38650_) {
         super(p_38649_, p_38650_);
@@ -92,28 +_,26 @@
         this.explode(null, p_38689_);
     }
 
-    protected void explode(@Nullable DamageSource p_259539_, double p_260287_) {
-        if (!this.level().isClientSide) {
-            double d0 = Math.sqrt(p_260287_);
-            if (d0 > 5.0) {
-                d0 = 5.0;
-            }
-
-            this.level()
-                .explode(
-                    this,
-                    p_259539_,
-                    null,
-                    this.getX(),
-                    this.getY(),
-                    this.getZ(),
-                    (float)(4.0 + this.random.nextDouble() * 1.5 * d0),
-                    false,
-                    Level.ExplosionInteraction.TNT
-                );
-            this.discard();
-        }
-    }
+   protected void explode(@Nullable DamageSource p_259539_, double p_260287_) {
+      if (!this.level().isClientSide) {
+         double d0 = Math.sqrt(p_260287_);
+         if (d0 > 5.0D) {
+            d0 = 5.0D;
+         }
+
+         // CraftBukkit start
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), (float) (4.0D + this.random.nextDouble() * 1.5D * d0), false);
+         this.level.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            fuse = -1;
+            return;
+         }
+         this.level().explode(this, p_259539_, (ExplosionDamageCalculator) null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.TNT);
+         // CraftBukkit end
+         this.discard();
+      }
+
+   }
 
     @Override
     public boolean causeFallDamage(float p_150347_, float p_150348_, DamageSource p_150349_) {
