--- a/net/minecraft/world/gen/feature/template/TemplateManager.java
+++ b/net/minecraft/world/gen/feature/template/TemplateManager.java
@@ -29,7 +29,7 @@
 
 public class TemplateManager {
    private static final Logger field_195431_a = LogManager.getLogger();
-   private final Map<ResourceLocation, Template> field_186240_a = Maps.newHashMap();
+   private final Map<ResourceLocation, Template> field_186240_a = Maps.newConcurrentMap(); // SPIGOT-5287
    private final DataFixer field_191154_c;
    private IResourceManager field_237130_d_;
    private final Path field_195433_e;
