--- a/net/minecraft/entity/monster/SlimeEntity.java
+++ b/net/minecraft/entity/monster/SlimeEntity.java
@@ -1,6 +1,8 @@
 package net.minecraft.entity.monster;
 
+import java.util.ArrayList;
 import java.util.EnumSet;
+import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.entity.Entity;
@@ -37,10 +39,13 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
-import net.minecraft.world.WorldType;
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.biome.Biomes;
 import net.minecraft.world.storage.loot.LootTables;
+import org.bukkit.craftbukkit.v1_15_R1.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.SlimeSplitEvent;
 
 public class SlimeEntity extends MobEntity implements IMob {
    private static final DataParameter<Integer> field_184711_bt = EntityDataManager.func_187226_a(SlimeEntity.class, DataSerializers.field_187192_b);
@@ -75,7 +80,7 @@
       this.func_110140_aT().func_111150_b(SharedMonsterAttributes.field_111264_e);
    }
 
-   protected void func_70799_a(int p_70799_1_, boolean p_70799_2_) {
+   public void func_70799_a(int p_70799_1_, boolean p_70799_2_) { // protected->public CraftBukkit
       this.field_70180_af.func_187227_b(field_184711_bt, p_70799_1_);
       this.func_226264_Z_();
       this.func_213323_x_();
@@ -129,6 +134,7 @@
       if (this.field_70122_E && !this.field_175452_bi) {
          int i = this.func_70809_q();
 
+         if (spawnCustomParticles()) i = 0; // don't spawn particles if it's handled by the implementation itself
          for(int j = 0; j < i * 8; ++j) {
             float f = this.field_70146_Z.nextFloat() * ((float)Math.PI * 2F);
             float f1 = this.field_70146_Z.nextFloat() * 0.5F + 0.5F;
@@ -180,11 +186,24 @@
       return (EntityType<? extends SlimeEntity>) super.func_200600_R();
    }
 
-   public void func_70106_y() {
+   @Override
+   public void remove(boolean keepData) {
       int i = this.func_70809_q();
-      if (!this.field_70170_p.field_72995_K && i > 1 && this.func_110143_aJ() <= 0.0F) {
+      if (!this.field_70170_p.field_72995_K && i > 1 && this.func_110143_aJ() <= 0.0F && !this.field_70128_L) {
          int j = 2 + this.field_70146_Z.nextInt(3);
 
+         // CraftBukkit start
+         SlimeSplitEvent event = new SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), j);
+         this.field_70170_p.getServerCB().getPluginManager().callEvent(event);
+         if (!event.isCancelled() && event.getCount() > 0) {
+            j = event.getCount();
+         } else {
+            super.func_70106_y();
+            return;
+         }
+         List<LivingEntity> slimes = new ArrayList<>(j);
+         // CraftBukkit end
+
          for(int k = 0; k < j; ++k) {
             float f = ((float)(k % 2) - 0.5F) * (float)i / 4.0F;
             float f1 = ((float)(k / 2) - 0.5F) * (float)i / 4.0F;
@@ -200,11 +219,18 @@
             slimeentity.func_184224_h(this.func_190530_aW());
             slimeentity.func_70799_a(i / 2, true);
             slimeentity.func_70012_b(this.func_226277_ct_() + (double)f, this.func_226278_cu_() + 0.5D, this.func_226281_cx_() + (double)f1, this.field_70146_Z.nextFloat() * 360.0F, 0.0F);
-            this.field_70170_p.func_217376_c(slimeentity);
+            slimes.add(slimeentity);
          }
+
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityTransformEvent(this, slimes, EntityTransformEvent.TransformReason.SPLIT).isCancelled()) {
+            return;
+         }
+         slimes.forEach(living -> this.field_70170_p.addEntity(living, CreatureSpawnEvent.SpawnReason.SLIME_SPLIT));   // CraftBukkit - SpawnReason
+         // CraftBukkit end
       }
 
-      super.func_70106_y();
+      super.remove(keepData);
    }
 
    public void func_70108_f(Entity p_70108_1_) {
@@ -257,12 +283,12 @@
       return this.func_189101_db() ? SoundEvents.field_187900_fz : SoundEvents.field_187886_fs;
    }
 
-   protected ResourceLocation func_184647_J() {
+   public ResourceLocation func_184647_J() { // protected->public CraftBukkit
       return this.func_70809_q() == 1 ? this.func_200600_R().func_220348_g() : LootTables.field_186419_a;
    }
 
    public static boolean func_223366_c(EntityType<SlimeEntity> p_223366_0_, IWorld p_223366_1_, SpawnReason p_223366_2_, BlockPos p_223366_3_, Random p_223366_4_) {
-      if (p_223366_1_.func_72912_H().func_76067_t() == WorldType.field_77138_c && p_223366_4_.nextInt(4) != 1) {
+      if (p_223366_1_.func_72912_H().func_76067_t().handleSlimeSpawnReduction(p_223366_4_, p_223366_1_) && p_223366_4_.nextInt(4) != 1) {
          return false;
       } else {
          if (p_223366_1_.func_175659_aa() != Difficulty.PEACEFUL) {
@@ -272,7 +298,7 @@
             }
 
             ChunkPos chunkpos = new ChunkPos(p_223366_3_);
-            boolean flag = SharedSeedRandom.func_205190_a(chunkpos.field_77276_a, chunkpos.field_77275_b, p_223366_1_.func_72905_C(), 987234911L).nextInt(10) == 0;
+            boolean flag = SharedSeedRandom.func_205190_a(chunkpos.field_77276_a, chunkpos.field_77275_b, p_223366_1_.func_72905_C(), p_223366_1_.func_201672_e().spigotConfig.slimeSeed).nextInt(10) == 0; // Spigot
             if (p_223366_4_.nextInt(10) == 0 && flag && p_223366_3_.func_177956_o() < 40) {
                return func_223315_a(p_223366_0_, p_223366_1_, p_223366_2_, p_223366_3_, p_223366_4_);
             }
@@ -320,6 +346,12 @@
       return super.func_213305_a(p_213305_1_).func_220313_a(0.255F * (float)this.func_70809_q());
    }
 
+   /**
+    * Called when the slime spawns particles on landing, see onUpdate.
+    * Return true to prevent the spawning of the default particles.
+    */
+   protected boolean spawnCustomParticles() { return false; }
+
    static class AttackGoal extends Goal {
       private final SlimeEntity field_179466_a;
       private int field_179465_b;
