--- ../src-base/minecraft/net/minecraft/network/rcon/RConThreadMain.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConThreadMain.java
@@ -28,7 +28,7 @@
         super(p_i1538_1_, "RCON Listener");
         this.rconPort = p_i1538_1_.getIntProperty("rcon.port", 0);
         this.rconPassword = p_i1538_1_.getStringProperty("rcon.password", "");
-        this.hostname = p_i1538_1_.getHostname();
+        this.hostname = p_i1538_1_.getStringProperty("rcon.ip", p_i1538_1_.getHostname()); // Paper
         this.serverPort = p_i1538_1_.getPort();
 
         if (0 == this.rconPort)
