--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -19,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.ProjectileWeaponItem;
 import net.minecraft.world.level.pathfinder.PathComputationType;
@@ -79,14 +_,22 @@
         throwItem(p_22614_, p_22615_, p_22616_, vec3, 0.3F);
     }
 
-    public static void throwItem(LivingEntity p_217134_, ItemStack p_217135_, Vec3 p_217136_, Vec3 p_217137_, float p_217138_) {
-        double d0 = p_217134_.getEyeY() - (double)p_217138_;
+   public static void throwItem(LivingEntity p_217134_, ItemStack p_217135_, Vec3 p_217136_, Vec3 p_217137_, float p_217138_) {
+      	if (p_217135_.isEmpty()) return; // CraftBukkit - SPIGOT-4940: no empty loot
+      	double d0 = p_217134_.getEyeY() - (double)p_217138_;
         ItemEntity itementity = new ItemEntity(p_217134_.level(), p_217134_.getX(), d0, p_217134_.getZ(), p_217135_);
         itementity.setThrower(p_217134_);
         Vec3 vec3 = p_217136_.subtract(p_217134_.position());
         vec3 = vec3.normalize().multiply(p_217137_.x, p_217137_.y, p_217137_.z);
         itementity.setDeltaMovement(vec3);
         itementity.setDefaultPickUpDelay();
+	    // CraftBukkit start
+	    org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(p_217134_.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+	    itementity.level.getCraftServer().getPluginManager().callEvent(event);
+	    if (event.isCancelled()) {
+	        return;
+	    }
+	    // CraftBukkit end
         p_217134_.level().addFreshEntity(itementity);
     }
 
