--- a/net/minecraft/client/renderer/chunk/ChunkRenderCache.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderCache.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.IBlockDisplayReader;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.chunk.IChunk;
 import net.minecraft.world.level.ColorResolver;
 import net.minecraft.world.lighting.WorldLightManager;
 import net.minecraftforge.api.distmarker.Dist;
