--- ../src-base/minecraft/org/bukkit/configuration/serialization/ConfigurationSerialization.java
+++ ../src-work/minecraft/org/bukkit/configuration/serialization/ConfigurationSerialization.java
@@ -8,10 +8,10 @@
 import java.util.Map;
 import java.util.logging.Level;
 import java.util.logging.Logger;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Color;
 import org.bukkit.FireworkEffect;
+import org.bukkit.Location;
 import org.bukkit.configuration.Configuration;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.potion.PotionEffect;
@@ -33,6 +33,7 @@
         registerClass(Color.class);
         registerClass(PotionEffect.class);
         registerClass(FireworkEffect.class);
+        registerClass(Location.class);
     }
 
     protected ConfigurationSerialization(Class<? extends ConfigurationSerializable> clazz) {
