From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] EMC Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 5ad0aacf01d714c5036d5be602a17cde36e7745f..bc3c9136eeb71a9f9126e543802dca77a1a90ad5 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -803,6 +803,16 @@ public abstract class EntityLiving extends Entity {
         }
         // CraftBukkit end
 
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), new org.bukkit.potion.PotionEffect(
+            org.bukkit.potion.PotionEffectType.getById(MobEffectList.getId(mobeffect.getMobEffect())),
+            mobeffect.getDuration(), mobeffect.getAmplifier()));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        final org.bukkit.potion.PotionEffect effect = addedEvent.getEffect();
+        mobeffect = new MobEffect(MobEffectList.fromId(effect.getType().getId()), effect.getDuration(), effect.getAmplifier(), effect.isAmbient(), effect.hasParticles());
+        // EMC end
         if (!this.d(mobeffect)) {
             return false;
         } else {
