From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 29 Nov 2019 02:48:17 -0600
Subject: [PATCH] EMC Add the PlayerThrowTridentEvent


diff --git a/src/main/java/net/minecraft/server/ItemTrident.java b/src/main/java/net/minecraft/server/ItemTrident.java
index f571a1586d131be7f6caff96f455604f4d576e59..2997c0e981be1e73218547796412bebc053d2729 100644
--- a/src/main/java/net/minecraft/server/ItemTrident.java
+++ b/src/main/java/net/minecraft/server/ItemTrident.java
@@ -1,6 +1,8 @@
 package net.minecraft.server;
 
 import com.google.common.collect.Multimap;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.Player;
 
 public class ItemTrident extends Item {
 
@@ -62,6 +64,7 @@ public class ItemTrident extends Item {
                                 }
                                 return;
                             }
+                            (new com.destroystokyo.paper.event.player.PlayerThrowTridentEvent((Player) entityhuman.getBukkitEntity(), CraftItemStack.asCraftMirror(itemstack), entitythrowntrident.getBukkitEntity())).callEvent(); // Paper
 
                             itemstack.damage(1, entityhuman, (entityhuman1) -> {
                                 entityhuman1.broadcastItemBreak(entityliving.getRaisedHand());
