--- ../src-base/minecraft/org/bukkit/event/Event.java
+++ ../src-work/minecraft/org/bukkit/event/Event.java
@@ -1,5 +1,6 @@
 package org.bukkit.event;
 
+import org.bukkit.Bukkit;
 import org.bukkit.plugin.PluginManager;
 
 /**
@@ -8,11 +9,11 @@
  * All events require a static method named getHandlerList() which returns the same {@link HandlerList} as {@link #getHandlers()}.
  *
  * @see PluginManager#callEvent(Event)
- * @see PluginManager#registerEvents(Listener,Plugin)
+ * @see PluginManager#registerEvents(Listener, Plugin)
  */
 public abstract class Event {
-    private String name;
     private final boolean async;
+    private String name;
 
     /**
      * The default constructor is defined for cleaner code. This constructor
@@ -33,7 +34,24 @@
         this.async = isAsync;
     }
 
+    // Paper start
+
     /**
+     * Calls the event and tests if cancelled.
+     *
+     * @return false if event was cancelled, if cancellable. otherwise true.
+     */
+    public boolean callEvent() {
+        Bukkit.getPluginManager().callEvent(this);
+        if (this instanceof Cancellable) {
+            return !((Cancellable) this).isCancelled();
+        } else {
+            return true;
+        }
+    }
+    // Paper end
+
+    /**
      * Convenience method for providing a user-friendly identifier. By
      * default, it is the event's class's {@linkplain Class#getSimpleName()
      * simple name}.
