--- ../src-base/minecraft/net/minecraftforge/fml/client/config/GuiSelectStringEntries.java
+++ ../src-work/minecraft/net/minecraftforge/fml/client/config/GuiSelectStringEntries.java
@@ -195,7 +195,7 @@
         }
 
         @Override
-        public void updatePosition(int slotIndex, int x, int y, float partialTicks){}
+        public void updatePosition(int p_192633_1_, int p_192633_2_, int p_192633_3_, float p_192633_4_){}
     }
 
     public static interface IGuiSelectStringListEntry extends GuiListExtended.IGuiListEntry
