--- ../src-base/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
+++ ../src-work/minecraft/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -68,7 +68,8 @@
 
     public String getWorldName()
     {
-        return this.delegate.getWorldName();
+        String name = super.getWorldName();
+        return name != null ? name : this.delegate.getWorldName();
     }
 
     public int getSaveVersion()
@@ -218,13 +219,33 @@
     {
     }
 
+    @Deprecated
     public void setDimensionData(DimensionType dimensionIn, NBTTagCompound compound)
     {
         this.delegate.setDimensionData(dimensionIn, compound);
     }
 
+    @Deprecated
     public NBTTagCompound getDimensionData(DimensionType dimensionIn)
     {
         return this.delegate.getDimensionData(dimensionIn);
     }
+
+    public void setDimensionData(int dimensionID, NBTTagCompound compound)
+    {
+        this.delegate.setDimensionData(dimensionID, compound);
+    }
+
+    public NBTTagCompound getDimensionData(int dimensionID)
+    {
+        return this.delegate.getDimensionData(dimensionID);
+    }
+
+    public int getDimension() {
+        return this.delegate.getDimension();
+    }
+
+    public void setDimension(int dim) {
+        this.delegate.setDimension(dim);
+    }
 }
