--- ../src-base/minecraft/net/minecraft/world/storage/MapStorage.java
+++ ../src-work/minecraft/net/minecraft/world/storage/MapStorage.java
@@ -14,12 +14,14 @@
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagShort;
+import red.mohist.asyn.ThreadBoxFactory;
+import red.mohist.configuration.MohistConfig;
 
 public class MapStorage
 {
     private final ISaveHandler saveHandler;
     protected Map<String, WorldSavedData> loadedDataMap = Maps.<String, WorldSavedData>newHashMap();
-    private final List<WorldSavedData> loadedDataList = Lists.<WorldSavedData>newArrayList();
+    public final List<WorldSavedData> loadedDataList = Lists.<WorldSavedData>newArrayList();
     private final Map<String, Short> idCounts = Maps.<String, Short>newHashMap();
 
     public MapStorage(ISaveHandler saveHandlerIn)
@@ -105,25 +107,53 @@
 
     private void saveData(WorldSavedData data)
     {
+        if (!MohistConfig.instance.asyn_worldsave.getValue()) {
+            if (this.saveHandler != null) {
+                try {
+                    File file1 = this.saveHandler.getMapFileFromName(data.mapName);
+
+                    if (file1 != null) {
+                        NBTTagCompound nbttagcompound = new NBTTagCompound();
+                        nbttagcompound.setTag("data", data.writeToNBT(new NBTTagCompound()));
+                        FileOutputStream fileoutputstream = new FileOutputStream(file1);
+                        CompressedStreamTools.writeCompressed(nbttagcompound, fileoutputstream);
+                        fileoutputstream.close();
+                    }
+                } catch (Exception exception) {
+                    exception.printStackTrace();
+                }
+            }
+        } else {
+            saveData1(data);
+        }
+    }
+
+    private void saveData1(WorldSavedData data)
+    {
         if (this.saveHandler != null)
         {
-            try
+            File file1 = this.saveHandler.getMapFileFromName(data.mapName);
+            if (file1 != null)
             {
-                File file1 = this.saveHandler.getMapFileFromName(data.mapName);
-
-                if (file1 != null)
+                try {
+                    ThreadBoxFactory.asyncWorldSaver.submit(() -> {
+                        try {
+                            NBTTagCompound nbttagcompound = new NBTTagCompound();
+                            nbttagcompound.setTag("data", data.writeToNBT(new NBTTagCompound()));
+                            FileOutputStream fileoutputstream = new FileOutputStream(file1);
+                            CompressedStreamTools.writeCompressed(nbttagcompound, fileoutputstream);
+                            fileoutputstream.close();
+                        }
+                        catch (Exception exception) {
+                            exception.printStackTrace();
+                        }
+                    }).get();
+                }
+                catch (Exception e)
                 {
-                    NBTTagCompound nbttagcompound = new NBTTagCompound();
-                    nbttagcompound.setTag("data", data.writeToNBT(new NBTTagCompound()));
-                    FileOutputStream fileoutputstream = new FileOutputStream(file1);
-                    CompressedStreamTools.writeCompressed(nbttagcompound, fileoutputstream);
-                    fileoutputstream.close();
+                    e.printStackTrace();
                 }
             }
-            catch (Exception exception)
-            {
-                exception.printStackTrace();
-            }
         }
     }
 
