--- ../src-base/minecraft/net/minecraft/world/storage/ThreadedFileIOBase.java
+++ ../src-work/minecraft/net/minecraft/world/storage/ThreadedFileIOBase.java
@@ -37,14 +37,20 @@
         for (int i = 0; i < this.threadedIOQueue.size(); ++i)
         {
             IThreadedFileIO ithreadedfileio = this.threadedIOQueue.get(i);
-            boolean flag = ithreadedfileio.writeNextIO();
+            boolean retain = false;
+            try {
+                retain = ithreadedfileio.writeNextIO();
+            } catch (Throwable e) {
+                e.printStackTrace();
+            }
 
-            if (!flag)
+            if (!retain)
             {
                 this.threadedIOQueue.remove(i--);
                 ++this.savedIOCounter;
             }
 
+            /*// Spigot start - don't sleep in between chunks so we unload faster.
             try
             {
                 Thread.sleep(this.isThreadWaiting ? 0L : 10L);
@@ -52,7 +58,7 @@
             catch (InterruptedException interruptedexception1)
             {
                 interruptedexception1.printStackTrace();
-            }
+            } */// Spigot end
         }
 
         if (this.threadedIOQueue.isEmpty())
