--- a/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/FeaturePoolElement.java
@@ -21,14 +_,13 @@
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.placement.PlacedFeature;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
-import net.minecraft.world.level.levelgen.structure.templatesystem.LiquidSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 
 public class FeaturePoolElement extends StructurePoolElement {
     public static final MapCodec<FeaturePoolElement> CODEC = RecordCodecBuilder.mapCodec(
-        p_341912_ -> p_341912_.group(PlacedFeature.CODEC.fieldOf("feature").forGetter(p_210215_ -> p_210215_.feature), projectionCodec())
-                .apply(p_341912_, FeaturePoolElement::new)
+        p_210213_ -> p_210213_.group(PlacedFeature.CODEC.fieldOf("feature").forGetter(p_210215_ -> p_210215_.feature), projectionCodec())
+                .apply(p_210213_, FeaturePoolElement::new)
     );
     private final Holder<PlacedFeature> feature;
     private final CompoundTag defaultJigsawNBT;
@@ -91,7 +_,6 @@
         Rotation p_227187_,
         BoundingBox p_227188_,
         RandomSource p_227189_,
-        LiquidSettings p_344422_,
         boolean p_227190_
     ) {
         return this.feature.value().place(p_227182_, p_227184_, p_227189_, p_227185_);
