--- a/net/minecraft/world/level/levelgen/structure/pools/ListPoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/ListPoolElement.java
@@ -14,14 +_,13 @@
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
-import net.minecraft.world.level.levelgen.structure.templatesystem.LiquidSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 
 public class ListPoolElement extends StructurePoolElement {
     public static final MapCodec<ListPoolElement> CODEC = RecordCodecBuilder.mapCodec(
-        p_341932_ -> p_341932_.group(StructurePoolElement.CODEC.listOf().fieldOf("elements").forGetter(p_210369_ -> p_210369_.elements), projectionCodec())
-                .apply(p_341932_, ListPoolElement::new)
+        p_210367_ -> p_210367_.group(StructurePoolElement.CODEC.listOf().fieldOf("elements").forGetter(p_210369_ -> p_210369_.elements), projectionCodec())
+                .apply(p_210367_, ListPoolElement::new)
     );
     private final List<StructurePoolElement> elements;
 
@@ -78,13 +_,10 @@
         Rotation p_227278_,
         BoundingBox p_227279_,
         RandomSource p_227280_,
-        LiquidSettings p_344918_,
         boolean p_227281_
     ) {
         for (StructurePoolElement structurepoolelement : this.elements) {
-            if (!structurepoolelement.place(
-                p_227272_, p_227273_, p_227274_, p_227275_, p_227276_, p_227277_, p_227278_, p_227279_, p_227280_, p_344918_, p_227281_
-            )) {
+            if (!structurepoolelement.place(p_227272_, p_227273_, p_227274_, p_227275_, p_227276_, p_227277_, p_227278_, p_227279_, p_227280_, p_227281_)) {
                 return false;
             }
         }
