--- a/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
@@ -98,7 +_,7 @@
     }
 
     public StructurePoolElement getRandomTemplate(RandomSource p_227356_) {
-        return (StructurePoolElement)(this.templates.isEmpty() ? EmptyPoolElement.INSTANCE : this.templates.get(p_227356_.nextInt(this.templates.size())));
+        return this.templates.get(p_227356_.nextInt(this.templates.size()));
     }
 
     public List<StructurePoolElement> getShuffledTemplates(RandomSource p_227363_) {
