--- a/net/minecraft/world/damagesource/DamageSources.java
+++ b/net/minecraft/world/damagesource/DamageSources.java
@@ -40,9 +_,15 @@
     private final DamageSource stalagmite;
     private final DamageSource outsideBorder;
     private final DamageSource genericKill;
+   	// CraftBukkit start
+   	private final DamageSource melting;
+   	private final DamageSource poison;
 
-    public DamageSources(RegistryAccess p_270740_) {
-        this.damageTypes = p_270740_.registryOrThrow(Registries.DAMAGE_TYPE);
+   	public DamageSources(RegistryAccess p_270740_) {
+      	this.damageTypes = p_270740_.registryOrThrow(Registries.DAMAGE_TYPE);
+      	this.melting = this.source(DamageTypes.ON_FIRE).melting();
+      	this.poison = this.source(DamageTypes.MAGIC).poison();
+      	// CraftBukkit end
         this.inFire = this.source(DamageTypes.IN_FIRE);
         this.lightningBolt = this.source(DamageTypes.LIGHTNING_BOLT);
         this.onFire = this.source(DamageTypes.ON_FIRE);
@@ -80,6 +_,16 @@
         return new DamageSource(this.damageTypes.getHolderOrThrow(p_270076_), p_270656_, p_270242_);
     }
 
+   	// CraftBukkit start
+   	public DamageSource melting() {
+      	return this.melting;
+   	}
+
+   	public DamageSource poison() {
+      	return this.poison;
+   	}
+   	// CraftBukkit end
+
     public DamageSource inFire() {
         return this.inFire;
     }
@@ -241,7 +_,13 @@
     }
 
     public DamageSource explosion(@Nullable Entity p_271016_, @Nullable Entity p_270814_) {
-        return this.source(p_270814_ != null && p_271016_ != null ? DamageTypes.PLAYER_EXPLOSION : DamageTypes.EXPLOSION, p_271016_, p_270814_);
+        // CraftBukkit start
+        return this.explosion(p_271016_, p_270814_, p_270814_ != null && p_271016_ != null ? DamageTypes.PLAYER_EXPLOSION : DamageTypes.EXPLOSION);
+    }
+
+    public DamageSource explosion(@Nullable Entity entity, @Nullable Entity entity1, ResourceKey<DamageType> resourceKey) {
+        return this.source(resourceKey, entity, entity1);
+        // CraftBukkit end
     }
 
     public DamageSource sonicBoom(Entity p_270401_) {
@@ -250,6 +_,10 @@
 
     public DamageSource badRespawnPointExplosion(Vec3 p_270175_) {
         return new DamageSource(this.damageTypes.getHolderOrThrow(DamageTypes.BAD_RESPAWN_POINT), p_270175_);
+    }
+
+    public DamageSource badRespawnPointExplosion(Vec3 vec3d, org.bukkit.block.BlockState blockState) {
+        return new DamageSource(this.damageTypes.getHolderOrThrow(DamageTypes.BAD_RESPAWN_POINT), vec3d).directBlockState(blockState);
     }
 
     public DamageSource outOfBorder() {
