From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 26 May 2015 04:00:58 -0400
Subject: [PATCH] Accurate block hit checking


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -17,6 +17,7 @@ import org.bukkit.material.MaterialData;
 import org.bukkit.metadata.Metadatable;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
 import org.bukkit.util.Consumer;
+import org.bukkit.util.RayBlockIntersection;
 import org.bukkit.util.Vector;
 
 /**
@@ -112,6 +113,26 @@ public interface World extends PluginMessageRecipient, Metadatable {
     public Block getHighestBlockAt(Location location);
 
     /**
+     * Find the first block intersecting the given ray
+     * @param start             Origin AND direction of the ray
+     * @param distance          Length of the ray (so it's not strictly a ray, but whatever)
+     * @param nonSolids         Include non-colliding blocks (e.g. torch, button, flower, etc)
+     * @param liquids           Include liquids (source blocks only)
+     * @return A data structure describing the intersection, or null if no blocks were hit
+     */
+    RayBlockIntersection rayTraceBlock(Location start, double distance, boolean nonSolids, boolean liquids);
+
+    /**
+     * Find the first block intersecting the given ray
+     * @param start             Origin of the ray
+     * @param end               End point of the ray (so it's not strictly a ray, but whatever)
+     * @param nonSolids         Include non-colliding blocks (e.g. torch, button, flower, etc)
+     * @param liquids           Include liquids (source blocks only)
+     * @return A data structure describing the intersection, or null if no blocks were hit
+     */
+    RayBlockIntersection rayTraceBlock(Vector start, Vector end, boolean nonSolids, boolean liquids);
+
+    /**
      * Gets the {@link Chunk} at the given coordinates
      *
      * @param x X-coordinate of the chunk
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -8,6 +8,8 @@ import org.bukkit.World;
 import org.bukkit.Location;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.metadata.Metadatable;
+import org.bukkit.util.RayBlockIntersection;
+import org.bukkit.util.Vector;
 
 /**
  * Represents a block. This is a live object, and only one Block may exist for
@@ -390,4 +392,19 @@ public interface Block extends Metadatable {
      */
     Collection<ItemStack> getDrops(ItemStack tool);
 
+    /**
+     * Find the first intersection of the given ray with this block's bounding box
+     * @param start             Origin of the ray
+     * @param end               End point of the ray (so it's not strictly a ray, but whatever)
+     * @return A data structure describing the intersection, or null if there was no intersection
+     */
+    RayBlockIntersection rayTrace(Vector start, Vector end);
+
+    /**
+     * Find the first intersection of the given ray with this block's bounding box
+     * @param start             Origin AND direction of the ray
+     * @param distance          Length of the ray (so it's not strictly a ray, but whatever)
+     * @return A data structure describing the intersection, or null if there was no intersection
+     */
+    RayBlockIntersection rayTrace(Location start, double distance);
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -24,6 +24,7 @@ import org.bukkit.event.player.PlayerResourcePackStatusEvent;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
 import org.bukkit.scoreboard.Scoreboard;
+import org.bukkit.util.RayBlockIntersection;
 
 /**
  * Represents a player, connected or not
@@ -1619,4 +1620,29 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * or null if they are not digging any block.
      */
     public org.bukkit.block.Block getDiggingBlock();
+
+    /**
+     * @return The distance from which the player can target blocks in their current gamemode
+     */
+    double getBlockReach();
+
+    /**
+     * Find the first block in the player's line-of-sight, up to the given distance away
+     * @param distance          Maximum distance from the player to search
+     * @param nonSolids         Include non-colliding blocks (e.g. torch, button, flower, etc)
+     * @param liquids           Include liquids (source blocks only)
+     * @return Information about the targeted block, or null if no blocks were found
+     */
+    RayBlockIntersection getTargetedBlock(double distance, boolean nonSolids, boolean liquids);
+
+    /**
+     * Find the block targeted by the player, i.e. the block that (probably) has a black outline on their screen.
+     * This accounts for the player's current gamemode, so players in creative mode will have a longer range.
+     * Players in adventure mode or spectator mode are assumed to be able to target any block, with the same range
+     * as survival mode.
+     * @param nonSolids         Include non-colliding blocks (e.g. torch, button, flower, etc)
+     * @param liquids           Include liquids (source blocks only)
+     * @return Information about the targeted block, or null if no blocks were found
+     */
+    RayBlockIntersection getTargetedBlock(boolean nonSolids, boolean liquids);
 }
diff --git a/src/main/java/org/bukkit/util/RayBlockIntersection.java b/src/main/java/org/bukkit/util/RayBlockIntersection.java
new file mode 100644
index 0000000..b1ba1d8
--- /dev/null
+++ b/src/main/java/org/bukkit/util/RayBlockIntersection.java
@@ -0,0 +1,40 @@
+package org.bukkit.util;
+
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+
+/**
+ * Result of a ray-block intersection test
+ */
+public class RayBlockIntersection {
+    private final Block block;
+    private final BlockFace face;
+    private final Vector position;
+
+    public RayBlockIntersection(Block block, BlockFace face, Vector position) {
+        this.block = block;
+        this.face = face;
+        this.position = position;
+    }
+
+    /**
+     * @return The intersected block
+     */
+    public Block getBlock() {
+        return block;
+    }
+
+    /**
+     * @return The first intersected face of the block
+     */
+    public BlockFace getFace() {
+        return face;
+    }
+
+    /**
+     * @return The first intersected point on the surface of the block
+     */
+    public Vector getPosition() {
+        return position;
+    }
+}
