From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 9 May 2014 23:44:41 -0400
Subject: [PATCH] Add API to disable potion particles


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -302,6 +302,13 @@ public interface LivingEntity extends Attributable, Entity, Damageable, Projecti
     public Collection<PotionEffect> getActivePotionEffects();
 
     /**
+     * Enables or disables potion particles visible around this entity when under
+     * the effect of a potion. If this entity is a player, this also affects the
+     * particles they see in front of them.
+     */
+    public void setPotionParticles(boolean enabled);
+
+    /**
      * Checks whether the living entity has block line of sight to another.
      * <p>
      * This uses the same algorithm that hostile mobs use to find the closest
