From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 25 Dec 2012 01:50:07 -0500
Subject: [PATCH] Add onGround API


diff --git a/src/main/java/org/bukkit/event/player/PlayerOnGroundEvent.java b/src/main/java/org/bukkit/event/player/PlayerOnGroundEvent.java
new file mode 100644
index 0000000..4a6c3d4
--- /dev/null
+++ b/src/main/java/org/bukkit/event/player/PlayerOnGroundEvent.java
@@ -0,0 +1,32 @@
+package org.bukkit.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+
+public class PlayerOnGroundEvent extends PlayerEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private boolean onGround;
+
+    public PlayerOnGroundEvent(final Player player, boolean onGround) {
+        super(player);
+        this.onGround = onGround;
+    }
+    /**
+     * Returns true of the player is on the ground after the event
+     *
+     * @return if the player is on the ground
+     */
+    public boolean getOnGround() {
+        return this.onGround;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
