From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 2 Apr 2016 05:08:36 -0400
Subject: [PATCH] Add PlayerUseUnknownEntityEvent


diff --git a/src/main/java/org/bukkit/event/player/PlayerUseUnknownEntityEvent.java b/src/main/java/org/bukkit/event/player/PlayerUseUnknownEntityEvent.java
new file mode 100644
index 0000000..26cda06
--- /dev/null
+++ b/src/main/java/org/bukkit/event/player/PlayerUseUnknownEntityEvent.java
@@ -0,0 +1,35 @@
+package org.bukkit.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.inventory.EquipmentSlot;
+
+public class PlayerUseUnknownEntityEvent extends PlayerActionBase {
+
+    private final int entityId;
+    private final boolean attack;
+    private final EquipmentSlot hand;
+
+    public PlayerUseUnknownEntityEvent(Player who, int entityId, boolean attack, EquipmentSlot hand) {
+        super(who);
+        this.entityId = entityId;
+        this.attack = attack;
+        this.hand = hand;
+    }
+
+    public int getEntityId() {
+        return entityId;
+    }
+
+    public boolean isAttack() {
+        return attack;
+    }
+
+    public EquipmentSlot getHand() {
+        return hand;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+    public static HandlerList getHandlerList() { return handlers; }
+    @Override public HandlerList getHandlers() { return handlers; }
+}
