From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 03:13:18 -0500
Subject: [PATCH] Arrow pickup rule API


diff --git a/src/main/java/org/bukkit/entity/Arrow.java b/src/main/java/org/bukkit/entity/Arrow.java
--- a/src/main/java/org/bukkit/entity/Arrow.java
+++ b/src/main/java/org/bukkit/entity/Arrow.java
@@ -39,4 +39,20 @@ public interface Arrow extends Projectile {
      * @param critical whether or not it should be critical
      */
     public void setCritical(boolean critical);
+
+    /**
+     * Which players can pickup this arrow as an item?
+     *
+     * This is generally true only if the arrow was NOT fired from a bow with the infinity enchantment.
+     */
+    PickupRule getPickupRule();
+
+    /**
+     * Set the rule for which players can pickup this arrow as an item.
+     */
+    void setPickupRule(PickupRule rule);
+
+    enum PickupRule {
+        DISALLOWED, ALLOWED, CREATIVE_ONLY
+    }
 }
