From: md_5 <git@md-5.net>
Date: Sat, 13 Dec 2014 12:59:14 +1100
Subject: [PATCH] BungeeCord Chat API


diff --git a/pom.xml b/pom.xml
--- a/pom.xml
+++ b/pom.xml
@@ -93,6 +93,14 @@
             <version>1.17</version>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>tc.oc</groupId>
+            <artifactId>bungeecord-chat</artifactId>
+            <version>1.11-SNAPSHOT</version>
+            <type>jar</type>
+            <scope>compile</scope>
+        </dependency>
+
         <!-- testing -->
         <dependency>
             <groupId>junit</groupId>
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1192,4 +1192,22 @@ public final class Bukkit {
     public static UnsafeValues getUnsafe() {
         return server.getUnsafe();
     }
+
+    /**
+     * Sends the component to the player
+     *
+     * @param component the components to send
+     */
+    public static void broadcast(net.md_5.bungee.api.chat.BaseComponent component) {
+        server.broadcast(component);
+    }
+
+    /**
+     * Sends an array of components as a single message to the player
+     *
+     * @param components the components to send
+     */
+    public static void broadcast(net.md_5.bungee.api.chat.BaseComponent... components) {
+        server.broadcast(components);
+    }
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -974,4 +974,18 @@ public interface Server extends PluginMessageRecipient {
      */
     @Deprecated
     UnsafeValues getUnsafe();
+
+    /**
+     * Sends the component to the player
+     *
+     * @param component the components to send
+     */
+    public void broadcast(net.md_5.bungee.api.chat.BaseComponent component);
+
+    /**
+     * Sends an array of components as a single message to the player
+     *
+     * @param components the components to send
+     */
+    public void broadcast(net.md_5.bungee.api.chat.BaseComponent... components);
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1495,4 +1495,18 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * effectively setting it to the default of "en_US".
      */
     public void setLocale(String locale);
+
+    /**
+     * Sends the component to the player
+     *
+     * @param component the components to send
+     */
+    public void sendMessage(net.md_5.bungee.api.chat.BaseComponent component);
+
+    /**
+     * Sends an array of components as a single message to the player
+     *
+     * @param components the components to send
+     */
+    public void sendMessage(net.md_5.bungee.api.chat.BaseComponent... components);
  }
