From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 20 Jun 2015 12:27:57 -0400
Subject: [PATCH] Call onDisable before disabling the plugin


diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -394,6 +394,18 @@ public final class SimplePluginManager implements PluginManager {
 
     public void enablePlugin(final Plugin plugin) {
         if (!plugin.isEnabled()) {
+            boolean dependenciesEnabled = true;
+            for(String dependency : plugin.getDescription().getDepend()) {
+                if(!this.isPluginEnabled(dependency)) {
+                    dependenciesEnabled = false;
+                    plugin.getLogger().warning(
+                        "Cannot enable " + plugin.getDescription().getFullName() +
+                        " because dependency " + dependency + " is not enabled"
+                    );
+                }
+            }
+            if(!dependenciesEnabled) return;
+
             List<Command> pluginCommands = PluginCommandYamlParser.parse(plugin);
 
             if (!pluginCommands.isEmpty()) {
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -264,13 +264,19 @@ public abstract class JavaPlugin extends PluginBase {
      * @param enabled true if enabled, otherwise false
      */
     protected final void setEnabled(final boolean enabled) {
-        if (isEnabled != enabled) {
-            isEnabled = enabled;
-
-            if (isEnabled) {
+        if(!isEnabled && enabled) {
+            isEnabled = true;
+            try {
                 onEnable();
-            } else {
+            } catch(RuntimeException e) {
+                isEnabled = false;
+                throw e;
+            }
+        } else if(isEnabled && !enabled) {
+            try {
                 onDisable();
+            } finally {
+                isEnabled = false;
             }
         }
     }
