From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 16 Feb 2015 04:58:20 -0500
Subject: [PATCH] Efficiently search for blocks by material


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -5,6 +5,8 @@ import org.bukkit.block.BlockState;
 import org.bukkit.entity.Entity;
 import org.bukkit.material.MaterialData;
 
+import java.util.Set;
+
 /**
  * Represents a chunk of blocks
  */
@@ -42,6 +44,14 @@ public interface Chunk {
     Block getBlock(int x, int y, int z);
 
     /**
+     * Get all blocks in this chunk that are made of the given {@link Material}
+     *
+     * @param material type of block to search for
+     * @return all blocks found
+     */
+    Set<Block> getBlocks(Material material);
+
+    /**
      * Capture thread-safe read-only snapshot of chunk data
      *
      * @return ChunkSnapshot
