From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 28 Apr 2012 01:39:04 -0400
Subject: [PATCH] Entity collision API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1073,6 +1073,18 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     public void setFlying(boolean value);
 
     /**
+     * Change whether this player can push other entities
+     * @param yes true to push other entities, which is the default
+     */
+    public void setCollidesWithEntities(boolean yes);
+
+    /**
+     * Whether this player can push other entities
+     * @return true if the player can push other entities
+     */
+    public boolean getCollidesWithEntities();
+
+    /**
      * Sets the speed at which a client will fly. Negative values indicate
      * reverse directions.
      *
