From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 29 Mar 2016 02:00:35 -0400
Subject: [PATCH] Entity bounding box API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -7,6 +7,7 @@ import org.bukkit.Nameable;
 import org.bukkit.Server;
 import org.bukkit.World;
 import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.geometry.Cuboid;
 import org.bukkit.metadata.Metadatable;
 import org.bukkit.util.Vector;
 
@@ -493,4 +494,6 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
      * @return true if the tag was successfully removed
      */
     boolean removeScoreboardTag(String tag);
+
+    Cuboid getBoundingBox();
 }
diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -13,6 +13,7 @@ import org.bukkit.inventory.EntityEquipment;
 import org.bukkit.potion.PotionEffect;
 import org.bukkit.potion.PotionEffectType;
 import org.bukkit.projectiles.ProjectileSource;
+import org.bukkit.geometry.Ray;
 
 /**
  * Represents a living entity, such as a monster or player
@@ -43,6 +44,12 @@ public interface LivingEntity extends Attributable, Entity, Damageable, Projecti
     public Location getEyeLocation();
 
     /**
+     * Get a {@link Ray} originating at this entity's eye position,
+     * and pointing in the direction this entity is looking.
+     */
+    Ray getEyeRay();
+
+    /**
      * Gets all blocks along the living entity's line of sight.
      * <p>
      * This list contains all blocks from the living entity's eye position to
