From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 9 Dec 2014 05:23:05 -0500
Subject: [PATCH] EventException knows the source Event


diff --git a/src/main/java/org/bukkit/event/EventException.java b/src/main/java/org/bukkit/event/EventException.java
--- a/src/main/java/org/bukkit/event/EventException.java
+++ b/src/main/java/org/bukkit/event/EventException.java
@@ -2,7 +2,26 @@ package org.bukkit.event;
 
 public class EventException extends Exception {
     private static final long serialVersionUID = 3532808232324183999L;
-    private final Throwable cause;
+    private final Event event;
+
+    /**
+     * Constructs a new EventException based on the given Exception
+     *
+     * @param cause Exception that triggered this Exception
+     */
+    public EventException(Throwable cause, Event event, String message) {
+        super(message, cause);
+        this.event = event;
+    }
+
+    /**
+     * Constructs a new EventException based on the given Exception
+     *
+     * @param throwable Exception that triggered this Exception
+     */
+    public EventException(Throwable throwable, Event event) {
+        this(throwable, event, null);
+    }
 
     /**
      * Constructs a new EventException based on the given Exception
@@ -10,14 +29,14 @@ public class EventException extends Exception {
      * @param throwable Exception that triggered this Exception
      */
     public EventException(Throwable throwable) {
-        cause = throwable;
+        this(throwable, (Event) null);
     }
 
     /**
      * Constructs a new EventException
      */
     public EventException() {
-        cause = null;
+        this((Throwable) null);
     }
 
     /**
@@ -27,8 +46,7 @@ public class EventException extends Exception {
      * @param message The message
      */
     public EventException(Throwable cause, String message) {
-        super(message);
-        this.cause = cause;
+        this(cause, null, message);
     }
 
     /**
@@ -37,17 +55,13 @@ public class EventException extends Exception {
      * @param message The message
      */
     public EventException(String message) {
-        super(message);
-        cause = null;
+        this(null, message);
     }
 
     /**
-     * If applicable, returns the Exception that triggered this Exception
-     *
-     * @return Inner exception, or null if one does not exist
+     * @return the {@link Event} that generated this exception
      */
-    @Override
-    public Throwable getCause() {
-        return cause;
+    public Event getEvent() {
+        return event;
     }
 }
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
--- a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
@@ -295,9 +295,9 @@ public final class JavaPluginLoader implements PluginLoader {
                         }
                         method.invoke(listener, event);
                     } catch (InvocationTargetException ex) {
-                        throw new EventException(ex.getCause());
+                        throw new EventException(ex.getCause(), event);
                     } catch (Throwable t) {
-                        throw new EventException(t);
+                        throw new EventException(t, event);
                     }
                 }
             };
