From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 7 Dec 2016 04:29:15 -0500
Subject: [PATCH] FallingBlock form block flag


diff --git a/src/main/java/org/bukkit/entity/FallingBlock.java b/src/main/java/org/bukkit/entity/FallingBlock.java
--- a/src/main/java/org/bukkit/entity/FallingBlock.java
+++ b/src/main/java/org/bukkit/entity/FallingBlock.java
@@ -59,4 +59,19 @@ public interface FallingBlock extends Entity {
      * @param hurtEntities whether entities will be damaged by this block.
      */
     void setHurtEntities(boolean hurtEntities);
+
+    /**
+     * Can this entity form a block when it lands?
+     */
+    boolean canFormBlock();
+
+    /**
+     * Allow or prevent this entity from forming a block when it lands.
+     *
+     * If this is false, the entity will just disappear when it lands.
+     * If true, it may form a block if other circumstances permit it.
+     *
+     * This is initially true by default.
+     */
+    void setFormBlock(boolean formBlock);
 }
