From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 30 Oct 2014 00:40:30 -0400
Subject: [PATCH] Fix combustion events


diff --git a/src/main/java/org/bukkit/event/entity/EntityExtinguishEvent.java b/src/main/java/org/bukkit/event/entity/EntityExtinguishEvent.java
new file mode 100644
index 0000000..5833ab7
--- /dev/null
+++ b/src/main/java/org/bukkit/event/entity/EntityExtinguishEvent.java
@@ -0,0 +1,25 @@
+package org.bukkit.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+
+/**
+ * Called when a burning entity is extinguished.
+ */
+public class EntityExtinguishEvent extends EntityEvent {
+
+    public EntityExtinguishEvent(Entity combustee) {
+        super(combustee);
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
