From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 21 Mar 2016 01:41:04 -0400
Subject: [PATCH] Multi-passenger API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -248,6 +248,28 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
     public abstract boolean eject();
 
     /**
+     * @return true if the given entity is a passenger of this vehicle
+     */
+    boolean hasPassenger(Entity passenger);
+
+    /**
+     * @return all of the current passengers in this vehicle
+     */
+    List<Entity> getPassengers();
+
+    /**
+     * Make the given entities passengers of this vehicle. Any existing
+     * passengers who are not in the list are ejected.
+     * @return given entities that could NOT become passengers for whatever reason
+     */
+    List<Entity> setPassengers(List<Entity> passengers);
+
+    /**
+     * Eject all passengers from this vehicle
+     */
+    void ejectAll();
+
+    /**
      * Returns the distance this entity has fallen
      *
      * @return The distance.
@@ -327,6 +349,12 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
     public boolean isInsideVehicle();
 
     /**
+     * Enter the given vehicle
+     * @return true if successful
+     */
+    boolean enterVehicle(Entity vehicle);
+
+    /**
      * Leave the current vehicle. If the entity is currently in a vehicle (and
      * is removed from it), true will be returned, otherwise false will be
      * returned.
