From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 21 Mar 2016 04:31:06 -0400
Subject: [PATCH] Natural regeneration API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -980,6 +980,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public void setFoodLevel(int value);
 
+    boolean getFastNaturalRegeneration();
+
+    void setFastNaturalRegeneration(boolean enabled);
+
+    boolean getSlowNaturalRegeneration();
+
+    void setSlowNaturalRegeneration(boolean enabled);
+
     /**
      * Gets the Location where the player will spawn at their bed, null if
      * they have not slept in one or their current bed spawn is invalid.
