From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 28 Feb 2015 04:32:26 -0500
Subject: [PATCH] Title API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1529,4 +1529,57 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      * @param footer content for the bottom of the player list
      */
     public void setPlayerListHeaderFooter(net.md_5.bungee.api.chat.BaseComponent header, net.md_5.bungee.api.chat.BaseComponent footer);
+
+    /**
+     * Set the times for titles displayed to the player
+     * @param fadeInTicks ticks to fade-in
+     * @param stayTicks ticks to stay visible
+     * @param fadeOutTicks ticks to fade-out
+     */
+    public void setTitleTimes(int fadeInTicks, int stayTicks, int fadeOutTicks);
+
+    /**
+     * Set the subtitle of titles displayed to the player
+     */
+    public void setSubtitle(net.md_5.bungee.api.chat.BaseComponent[] subtitle);
+
+    /**
+     * Set the subtitle of titles displayed to the player
+     */
+    public void setSubtitle(net.md_5.bungee.api.chat.BaseComponent subtitle);
+
+    /**
+     * Show the given title to the player, along with the last subtitle set, using the last set times
+     */
+    public void showTitle(net.md_5.bungee.api.chat.BaseComponent[] title);
+
+    /**
+     * Show the given title to the player, along with the last subtitle set, using the last set times
+     */
+    public void showTitle(net.md_5.bungee.api.chat.BaseComponent title);
+
+    /**
+     * Show the given title and subtitle to the player using the given times
+     * @param title big text
+     * @param subtitle little text under it
+     * @param fadeInTicks ticks to fade-in
+     * @param stayTicks ticks to stay visible
+     * @param fadeOutTicks ticks to fade-out
+     */
+    public void showTitle(net.md_5.bungee.api.chat.BaseComponent[] title, net.md_5.bungee.api.chat.BaseComponent[] subtitle, int fadeInTicks, int stayTicks, int fadeOutTicks);
+
+    /**
+     * Show the given title and subtitle to the player using the given times
+     * @param title big text
+     * @param subtitle little text under it
+     * @param fadeInTicks ticks to fade-in
+     * @param stayTicks ticks to stay visible
+     * @param fadeOutTicks ticks to fade-out
+     */
+    public void showTitle(net.md_5.bungee.api.chat.BaseComponent title, net.md_5.bungee.api.chat.BaseComponent subtitle, int fadeInTicks, int stayTicks, int fadeOutTicks);
+
+    /**
+     * Hide any title that is currently visible to the player
+     */
+    public void hideTitle();
 }
