From e5998c42c34d2320b89035198c855437af24d1c6 Mon Sep 17 00:00:00 2001
From: DerFlash <bte@freenet.de>
Date: Tue, 9 Jul 2013 00:11:12 +0200
Subject: [PATCH] Save ticks lived to nbttag


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index aa7727fb2..88871dc25 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1477,6 +1477,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener {
             if (this.persistentInvisibility) {
                 nbttagcompound.setBoolean("Bukkit.invisible", this.persistentInvisibility);
             }
+            nbttagcompound.setInt("Spigot.ticksLived", this.ticksLived);
             // CraftBukkit end
             IChatBaseComponent ichatbasecomponent = this.getCustomName();
 
@@ -1622,6 +1623,8 @@ public abstract class Entity implements INamableTileEntity, ICommandListener {
             if (this instanceof EntityLiving) {
                 EntityLiving entity = (EntityLiving) this;
 
+                this.ticksLived = nbttagcompound.getInt("Spigot.ticksLived");
+
                 // Reset the persistence for tamed animals
                 if (entity instanceof EntityTameableAnimal && !isLevelAtLeast(nbttagcompound, 2) && !nbttagcompound.getBoolean("PersistenceRequired")) {
                     EntityInsentient entityinsentient = (EntityInsentient) entity;
-- 
2.25.1

