From a7cb2cc8a8ca54debbfb5ffb80a507919633e747 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Mon, 14 Apr 2014 09:46:20 +1000
Subject: [PATCH] Use Provided Case for Non Existent Offline Players


diff --git a/src/main/java/net/minecraft/server/players/UserCache.java b/src/main/java/net/minecraft/server/players/UserCache.java
index ca9022723..10a29ccde 100644
--- a/src/main/java/net/minecraft/server/players/UserCache.java
+++ b/src/main/java/net/minecraft/server/players/UserCache.java
@@ -143,7 +143,7 @@ public class UserCache {
             usercache_usercacheentry.a(this.d());
             gameprofile = usercache_usercacheentry.a();
         } else {
-            gameprofile = a(this.e, s1);
+            gameprofile = a(this.e, s); // Spigot - use correct case for offline players
             if (gameprofile != null) {
                 this.a(gameprofile);
                 flag = false;
-- 
2.25.1

