From 181d8727927d33d28e48f64e6561137bf5bb5a9f Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Wed, 22 Apr 2020 12:04:34 +1000
Subject: [PATCH] Fix portals generating outside world border

See SPIGOT-5677, MC-11479

diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 4b90ee6ef..9fddbae24 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2477,8 +2477,10 @@ public abstract class Entity implements INamableTileEntity, ICommandListener {
                     d1 *= 8.0D;
                 }
 
-                double d3 = Math.min(-2.9999872E7D, worldserver1.getWorldBorder().c() + 16.0D);
-                double d4 = Math.min(-2.9999872E7D, worldserver1.getWorldBorder().d() + 16.0D);
+                // Spigot start - SPIGOT-5677, MC-114796: Fix portals generating outside world border
+                double d3 = Math.max(-2.9999872E7D, worldserver1.getWorldBorder().c() + 16.0D);
+                double d4 = Math.max(-2.9999872E7D, worldserver1.getWorldBorder().d() + 16.0D);
+                // Spigot end
                 double d5 = Math.min(2.9999872E7D, worldserver1.getWorldBorder().e() - 16.0D);
                 double d6 = Math.min(2.9999872E7D, worldserver1.getWorldBorder().f() - 16.0D);
 
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index ce4821092..541ddc928 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -763,8 +763,10 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             // this.setPositionRotation(d0, d1, d2, f1, f); // CraftBukkit - PlayerTeleportEvent handles position changes
             worldserver.getMethodProfiler().exit();
             worldserver.getMethodProfiler().enter("placing");
-            double d4 = Math.min(-2.9999872E7D, worldserver1.getWorldBorder().c() + 16.0D);
-            double d5 = Math.min(-2.9999872E7D, worldserver1.getWorldBorder().d() + 16.0D);
+            // Spigot start - SPIGOT-5677, MC-114796: Fix portals generating outside world border
+            double d4 = Math.max(-2.9999872E7D, worldserver1.getWorldBorder().c() + 16.0D);
+            double d5 = Math.max(-2.9999872E7D, worldserver1.getWorldBorder().d() + 16.0D);
+            // Spigot end
             double d6 = Math.min(2.9999872E7D, worldserver1.getWorldBorder().e() - 16.0D);
             double d7 = Math.min(2.9999872E7D, worldserver1.getWorldBorder().f() - 16.0D);
 
-- 
2.25.1

