--- a/net/minecraft/world/entity/monster/hoglin/HoglinAi.java
+++ b/net/minecraft/world/entity/monster/hoglin/HoglinAi.java
@@ -124,7 +_,7 @@
    private static void m_34619_(Hoglin p_34620_, LivingEntity p_34621_) {
       p_34620_.m_6274_().m_21936_(MemoryModuleType.f_26372_);
       p_34620_.m_6274_().m_21936_(MemoryModuleType.f_26370_);
-      p_34620_.m_6274_().m_21882_(MemoryModuleType.f_26383_, p_34621_, (long)f_34568_.m_214085_(p_34620_.m_9236_().f_46441_));
+      p_34620_.m_6274_().m_21882_(MemoryModuleType.f_26383_, p_34621_, (long)f_34568_.m_214085_(p_34620_.f_19853_.f_46441_));
    }
 
    private static Optional<? extends LivingEntity> m_34610_(Hoglin p_34611_) {
