--- ../src-base/minecraft/net/minecraftforge/event/entity/player/PlayerEvent.java
+++ ../src-work/minecraft/net/minecraftforge/event/entity/player/PlayerEvent.java
@@ -31,6 +31,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraftforge.event.entity.living.LivingEvent;
 import net.minecraftforge.fml.common.eventhandler.Event;
+import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer;
 
 /**
  * PlayerEvent is fired whenever an event involving Living entities occurs. <br>
@@ -154,6 +155,11 @@
             super(player);
             this.username = username;
             this.setDisplayname(username);
+            CraftPlayer cplayer = (CraftPlayer) player.getBukkitEntity();
+            String displayName = cplayer.getDisplayName();
+            if(displayName != null) {
+                displayname = displayName;
+            }
         }
 
         public String getUsername()
