--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -15,7 +_,8 @@
     @Override
     public boolean applyEffectTick(LivingEntity p_300845_, int p_301393_) {
         if (this.isHarm == p_300845_.isInvertedHealAndHarm()) {
-            p_300845_.heal((float)Math.max(4 << p_301393_, 0));
+         p_300845_.heal$regainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit // Mohist
+         p_300845_.heal((float)Math.max(4 << p_301393_, 0));
         } else {
             p_300845_.hurt(p_300845_.damageSources().magic(), (float)(6 << p_301393_));
         }
@@ -27,6 +_,7 @@
     public void applyInstantenousEffect(@Nullable Entity p_298495_, @Nullable Entity p_298887_, LivingEntity p_298479_, int p_298172_, double p_298163_) {
         if (this.isHarm == p_298479_.isInvertedHealAndHarm()) {
             int i = (int)(p_298163_ * (double)(4 << p_298172_) + 0.5);
+         	p_298479_.heal$regainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit // Mohist
             p_298479_.heal((float)i);
         } else {
             int j = (int)(p_298163_ * (double)(6 << p_298172_) + 0.5);
