--- a/net/minecraft/world/effect/MobEffect.java
+++ b/net/minecraft/world/effect/MobEffect.java
@@ -17,7 +_,7 @@
 import net.minecraft.world.entity.ai.attributes.AttributeMap;
 import net.minecraft.world.entity.ai.attributes.AttributeModifier;
 
-public class MobEffect {
+public class MobEffect implements net.minecraftforge.common.extensions.IForgeMobEffect {
    private final Map<Attribute, AttributeModifierTemplate> attributeModifiers = Maps.newHashMap();
    private final MobEffectCategory category;
    private final int color;
