--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -48,4 +_,13 @@
       });
       return list;
    }
+
+   // CraftBukkit start
+   public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel p_216947_, @Nullable Entity p_216948_, Vec3 p_216949_, double p_216950_, MobEffectInstance p_216951_, int p_216952_, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause) {
+      // Mohist start
+      LivingEntity.addEffectCause.set(cause);
+      return m_216946_(p_216947_, p_216948_, p_216949_, p_216950_, p_216951_, p_216952_);
+      // Mohist end
+   }
+   // CraftBukkit end
 }
