--- ../src-base/minecraft/net/minecraft/util/math/Vec3i.java
+++ ../src-work/minecraft/net/minecraft/util/math/Vec3i.java
@@ -7,10 +7,19 @@
 public class Vec3i implements Comparable<Vec3i>
 {
     public static final Vec3i NULL_VECTOR = new Vec3i(0, 0, 0);
+    // Paper start - Make mutable and protected for MutableBlockPos and PooledBlockPos
     private final int x;
     private final int y;
     private final int z;
 
+    public final boolean isValidLocation() {
+        return x >= -30000000 && z >= -30000000 && x < 30000000 && z < 30000000 && y >= 0 && y < 256;
+    }
+    public boolean isInvalidYLocation() {
+        return y < 0 || y >= 256;
+    }
+    // Paper end
+
     public Vec3i(int xIn, int yIn, int zIn)
     {
         this.x = xIn;
