--- ../src-base/minecraft/org/bukkit/scoreboard/Criterias.java
+++ ../src-work/minecraft/org/bukkit/scoreboard/Criterias.java
@@ -10,11 +10,12 @@
     public static final String DEATHS;
 
     static {
-        HEALTH="health";
-        PLAYER_KILLS="playerKillCount";
-        TOTAL_KILLS="totalKillCount";
-        DEATHS="deathCount";
+        HEALTH = "health";
+        PLAYER_KILLS = "playerKillCount";
+        TOTAL_KILLS = "totalKillCount";
+        DEATHS = "deathCount";
     }
 
-    private Criterias() {}
+    private Criterias() {
+    }
 }
