--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -27,7 +_,7 @@
 
    static <T extends GameRules.Value<T>> int setRule(CommandContext<CommandSourceStack> p_137755_, GameRules.Key<T> p_137756_) {
       CommandSourceStack commandsourcestack = p_137755_.getSource();
-      T t = commandsourcestack.getServer().getGameRules().getRule(p_137756_);
+      T t = commandsourcestack.getLevel().getGameRules().getRule(p_137756_); // CraftBukkit
       t.setFromArgument(p_137755_, "value");
       commandsourcestack.sendSuccess(() -> {
          return Component.translatable("commands.gamerule.set", p_137756_.getId(), t.toString());
@@ -36,7 +_,7 @@
    }
 
    static <T extends GameRules.Value<T>> int queryRule(CommandSourceStack p_137758_, GameRules.Key<T> p_137759_) {
-      T t = p_137758_.getServer().getGameRules().getRule(p_137759_);
+      T t = p_137758_.getLevel().getGameRules().getRule(p_137759_); // CraftBukkit
       p_137758_.sendSuccess(() -> {
          return Component.translatable("commands.gamerule.query", p_137759_.getId(), t.toString());
       }, false);
