--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -13,6 +_,7 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.ItemTags;
 import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
 import net.minecraft.world.ItemInteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
@@ -29,6 +_,7 @@
 import net.minecraft.world.level.block.entity.BannerPatternLayers;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public interface CauldronInteraction {
     Map<String, CauldronInteraction.InteractionMap> INTERACTIONS = new Object2ObjectArrayMap<>();
@@ -58,35 +_,53 @@
             Blocks.POWDER_SNOW_CAULDRON.defaultBlockState().setValue(LayeredCauldronBlock.LEVEL, Integer.valueOf(3)),
             SoundEvents.BUCKET_EMPTY_POWDER_SNOW
         );
-    CauldronInteraction SHULKER_BOX = (p_340986_, p_340987_, p_340988_, p_340989_, p_340990_, p_340991_) -> {
-        Block block = Block.byItem(p_340991_.getItem());
+    CauldronInteraction SHULKER_BOX = (p_325746_, p_325747_, p_325748_, p_325749_, p_325750_, p_325751_) -> {
+        Block block = Block.byItem(p_325751_.getItem());
         if (!(block instanceof ShulkerBoxBlock)) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
-            if (!p_340987_.isClientSide) {
-                ItemStack itemstack = p_340991_.transmuteCopy(Blocks.SHULKER_BOX, 1);
-                p_340989_.setItemInHand(p_340990_, ItemUtils.createFilledResult(p_340991_, p_340989_, itemstack, false));
-                p_340989_.awardStat(Stats.CLEAN_SHULKER_BOX);
-                LayeredCauldronBlock.lowerFillLevel(p_340986_, p_340987_, p_340988_);
-            }
-
-            return ItemInteractionResult.sidedSuccess(p_340987_.isClientSide);
+            if (!p_325747_.isClientSide) {
+            	// CraftBukkit start
+            	if (!LayeredCauldronBlock.lowerFillLevel(p_325746_, p_325747_, p_325748_, p_325749_, CauldronLevelChangeEvent.ChangeReason.SHULKER_WASH)) {
+                    return ItemInteractionResult.SUCCESS;
+                }
+            	// CraftBukkit end
+
+                p_325749_.setItemInHand(p_325750_, p_325751_.transmuteCopy(Blocks.SHULKER_BOX, 1));
+                p_325749_.awardStat(Stats.CLEAN_SHULKER_BOX);
+            	// LayeredCauldronBlock.lowerFillLevel(p_175662_, p_175663_, p_175664_); // CraftBukkit
+         }
+
+            return ItemInteractionResult.sidedSuccess(p_325747_.isClientSide);
         }
     };
-    CauldronInteraction BANNER = (p_340992_, p_340993_, p_340994_, p_340995_, p_340996_, p_340997_) -> {
-        BannerPatternLayers bannerpatternlayers = p_340997_.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY);
+    CauldronInteraction BANNER = (p_325764_, p_325765_, p_325766_, p_325767_, p_325768_, p_325769_) -> {
+        BannerPatternLayers bannerpatternlayers = p_325769_.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY);
         if (bannerpatternlayers.layers().isEmpty()) {
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
-            if (!p_340993_.isClientSide) {
-                ItemStack itemstack = p_340997_.copyWithCount(1);
+            if (!p_325765_.isClientSide) {
+            	// CraftBukkit start
+            	if (!LayeredCauldronBlock.lowerFillLevel(p_325764_, p_325765_, p_325766_, p_325767_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                    return ItemInteractionResult.SUCCESS;
+            	}
+            	// CraftBukkit end
+                ItemStack itemstack = p_325769_.copyWithCount(1);
                 itemstack.set(DataComponents.BANNER_PATTERNS, bannerpatternlayers.removeLast());
-                p_340995_.setItemInHand(p_340996_, ItemUtils.createFilledResult(p_340997_, p_340995_, itemstack, false));
-                p_340995_.awardStat(Stats.CLEAN_BANNER);
-                LayeredCauldronBlock.lowerFillLevel(p_340992_, p_340993_, p_340994_);
-            }
-
-            return ItemInteractionResult.sidedSuccess(p_340993_.isClientSide);
+                p_325769_.consume(1, p_325767_);
+                if (p_325769_.isEmpty()) {
+                    p_325767_.setItemInHand(p_325768_, itemstack);
+                } else if (p_325767_.getInventory().add(itemstack)) {
+                    p_325767_.inventoryMenu.sendAllDataToRemote();
+                } else {
+                    p_325767_.drop(itemstack, false);
+                }
+
+                p_325767_.awardStat(Stats.CLEAN_BANNER);
+            	// LayeredCauldronBlock.lowerFillLevel(p_278890_, p_278891_, p_278892_); // CraftBukkit
+         }
+
+            return ItemInteractionResult.sidedSuccess(p_325765_.isClientSide);
         }
     };
     CauldronInteraction DYED_ITEM = (p_325770_, p_325771_, p_325772_, p_325773_, p_325774_, p_325775_) -> {
@@ -96,9 +_,14 @@
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_325771_.isClientSide) {
-                p_325775_.remove(DataComponents.DYED_COLOR);
+            	// CraftBukkit start
+            	if (!LayeredCauldronBlock.lowerFillLevel(p_325770_, p_325771_, p_325772_, p_325773_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+               		return ItemInteractionResult.SUCCESS;
+            	}
+            	// CraftBukkit end
+				p_325775_.remove(DataComponents.DYED_COLOR);
                 p_325773_.awardStat(Stats.CLEAN_ARMOR);
-                LayeredCauldronBlock.lowerFillLevel(p_325770_, p_325771_, p_325772_);
+            	// LayeredCauldronBlock.lowerFillLevel(p_175629_, p_175630_, p_175631_); // CraftBukkit
             }
 
             return ItemInteractionResult.sidedSuccess(p_325771_.isClientSide);
@@ -124,11 +_,16 @@
             PotionContents potioncontents = p_175737_.get(DataComponents.POTION_CONTENTS);
             if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                 if (!p_175733_.isClientSide) {
+               		// CraftBukkit start
+               		if (!LayeredCauldronBlock.changeLevel(p_175732_, p_175733_, p_175734_, Blocks.WATER_CAULDRON.defaultBlockState(), p_175735_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                  		return ItemInteractionResult.SUCCESS;
+               		}
+               		// CraftBukkit end
                     Item item = p_175737_.getItem();
                     p_175735_.setItemInHand(p_175736_, ItemUtils.createFilledResult(p_175737_, p_175735_, new ItemStack(Items.GLASS_BOTTLE)));
                     p_175735_.awardStat(Stats.USE_CAULDRON);
                     p_175735_.awardStat(Stats.ITEM_USED.get(item));
-                    p_175733_.setBlockAndUpdate(p_175734_, Blocks.WATER_CAULDRON.defaultBlockState());
+               		// p_175733_.setBlockAndUpdate(p_175734_, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
                     p_175733_.playSound(null, p_175734_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                     p_175733_.gameEvent(null, GameEvent.FLUID_PLACE, p_175734_);
                 }
@@ -156,11 +_,16 @@
         );
         map1.put(Items.GLASS_BOTTLE, (p_325758_, p_325759_, p_325760_, p_325761_, p_325762_, p_325763_) -> {
             if (!p_325759_.isClientSide) {
+            	// CraftBukkit start
+            	if (!LayeredCauldronBlock.lowerFillLevel(p_325758_, p_325759_, p_325760_, p_325761_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+               	    return ItemInteractionResult.SUCCESS;
+            	}
+            	// CraftBukkit end
                 Item item = p_325763_.getItem();
                 p_325761_.setItemInHand(p_325762_, ItemUtils.createFilledResult(p_325763_, p_325761_, PotionContents.createItemStack(Items.POTION, Potions.WATER)));
                 p_325761_.awardStat(Stats.USE_CAULDRON);
                 p_325761_.awardStat(Stats.ITEM_USED.get(item));
-                LayeredCauldronBlock.lowerFillLevel(p_325758_, p_325759_, p_325760_);
+            	// LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_);
                 p_325759_.playSound(null, p_325760_, SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 p_325759_.gameEvent(null, GameEvent.FLUID_PICKUP, p_325760_);
             }
@@ -174,10 +_,15 @@
                 PotionContents potioncontents = p_175709_.get(DataComponents.POTION_CONTENTS);
                 if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                     if (!p_175705_.isClientSide) {
+               			// CraftBukkit start
+               			if (!LayeredCauldronBlock.changeLevel(p_175704_, p_175705_, p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL), p_175707_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                  			return ItemInteractionResult.SUCCESS;
+               			}
+               			// CraftBukkit end
                         p_175707_.setItemInHand(p_175708_, ItemUtils.createFilledResult(p_175709_, p_175707_, new ItemStack(Items.GLASS_BOTTLE)));
                         p_175707_.awardStat(Stats.USE_CAULDRON);
                         p_175707_.awardStat(Stats.ITEM_USED.get(p_175709_.getItem()));
-                        p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL));
+               			// p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL));
                         p_175705_.playSound(null, p_175706_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                         p_175705_.gameEvent(null, GameEvent.FLUID_PLACE, p_175706_);
                     }
@@ -273,14 +_,19 @@
             return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         } else {
             if (!p_175637_.isClientSide) {
+            	// CraftBukkit start
+            	if (!LayeredCauldronBlock.changeLevel(p_175636_, p_175637_, p_175638_, Blocks.CAULDRON.defaultBlockState(), p_175639_, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+               		return ItemInteractionResult.SUCCESS;
+            	}
+            	// CraftBukkit end
                 Item item = p_175641_.getItem();
                 p_175639_.setItemInHand(p_175640_, ItemUtils.createFilledResult(p_175641_, p_175639_, p_175642_));
                 p_175639_.awardStat(Stats.USE_CAULDRON);
                 p_175639_.awardStat(Stats.ITEM_USED.get(item));
-                p_175637_.setBlockAndUpdate(p_175638_, Blocks.CAULDRON.defaultBlockState());
+            	// pLevel.setBlockAndUpdate(pPos, Blocks.CAULDRON.defaultBlockState());
                 p_175637_.playSound(null, p_175638_, p_175644_, SoundSource.BLOCKS, 1.0F, 1.0F);
                 p_175637_.gameEvent(null, GameEvent.FLUID_PICKUP, p_175638_);
-            }
+         }
 
             return ItemInteractionResult.sidedSuccess(p_175637_.isClientSide);
         }
@@ -289,12 +_,17 @@
     static ItemInteractionResult emptyBucket(
         Level p_175619_, BlockPos p_175620_, Player p_175621_, InteractionHand p_175622_, ItemStack p_175623_, BlockState p_175624_, SoundEvent p_175625_
     ) {
-        if (!p_175619_.isClientSide) {
+      if (!p_175619_.isClientSide) {
+         	// CraftBukkit start
+         	if (!LayeredCauldronBlock.changeLevel(p_175624_, p_175619_, p_175620_, p_175624_, p_175621_, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+            	return ItemInteractionResult.SUCCESS;
+         	}
+         	// CraftBukkit end
             Item item = p_175623_.getItem();
             p_175621_.setItemInHand(p_175622_, ItemUtils.createFilledResult(p_175623_, p_175621_, new ItemStack(Items.BUCKET)));
             p_175621_.awardStat(Stats.FILL_CAULDRON);
             p_175621_.awardStat(Stats.ITEM_USED.get(item));
-            p_175619_.setBlockAndUpdate(p_175620_, p_175624_);
+         	// pLevel.setBlockAndUpdate(pPos, pState);
             p_175619_.playSound(null, p_175620_, p_175625_, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175619_.gameEvent(null, GameEvent.FLUID_PLACE, p_175620_);
         }
