--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -28,6 +_,7 @@
 import net.minecraft.world.level.block.entity.BannerBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public interface CauldronInteraction {
    Map<Item, CauldronInteraction> EMPTY = newInteractionMap();
@@ -49,6 +_,11 @@
          return InteractionResult.PASS;
       } else {
          if (!p_175663_.isClientSide) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.lowerFillLevel(p_175662_, p_175663_, p_175664_, p_175665_, CauldronLevelChangeEvent.ChangeReason.SHULKER_WASH)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             ItemStack itemstack = new ItemStack(Blocks.SHULKER_BOX);
             if (p_175667_.hasTag()) {
                itemstack.setTag(p_175667_.getTag().copy());
@@ -126,7 +_,7 @@
                p_175735_.setItemInHand(p_175736_, ItemUtils.createFilledResult(p_175737_, p_175735_, new ItemStack(Items.GLASS_BOTTLE)));
                p_175735_.awardStat(Stats.USE_CAULDRON);
                p_175735_.awardStat(Stats.ITEM_USED.get(item));
-               p_175733_.setBlockAndUpdate(p_175734_, Blocks.WATER_CAULDRON.defaultBlockState());
+               // p_175733_.setBlockAndUpdate(p_175734_, Blocks.WATER_CAULDRON.defaultBlockState());
                p_175733_.playSound((Player)null, p_175734_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                p_175733_.gameEvent((Entity)null, GameEvent.FLUID_PLACE, p_175734_);
             }
@@ -142,11 +_,16 @@
       });
       WATER.put(Items.GLASS_BOTTLE, (p_175718_, p_175719_, p_175720_, p_175721_, p_175722_, p_175723_) -> {
          if (!p_175719_.isClientSide) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_, p_175721_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = p_175723_.getItem();
             p_175721_.setItemInHand(p_175722_, ItemUtils.createFilledResult(p_175723_, p_175721_, PotionUtils.setPotion(new ItemStack(Items.POTION), Potions.WATER)));
             p_175721_.awardStat(Stats.USE_CAULDRON);
             p_175721_.awardStat(Stats.ITEM_USED.get(item));
-            LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_);
+            // LayeredCauldronBlock.lowerFillLevel(p_175718_, p_175719_, p_175720_);
             p_175719_.playSound((Player)null, p_175720_, SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175719_.gameEvent((Entity)null, GameEvent.FLUID_PICKUP, p_175720_);
          }
@@ -156,10 +_,15 @@
       WATER.put(Items.POTION, (p_175704_, p_175705_, p_175706_, p_175707_, p_175708_, p_175709_) -> {
          if (p_175704_.getValue(LayeredCauldronBlock.LEVEL) != 3 && PotionUtils.getPotion(p_175709_) == Potions.WATER) {
             if (!p_175705_.isClientSide) {
+               // CraftBukkit start
+               if (!LayeredCauldronBlock.changeLevel(p_175704_, p_175705_, p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL), p_175707_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                  return InteractionResult.SUCCESS;
+               }
+               // CraftBukkit end
                p_175707_.setItemInHand(p_175708_, ItemUtils.createFilledResult(p_175709_, p_175707_, new ItemStack(Items.GLASS_BOTTLE)));
                p_175707_.awardStat(Stats.USE_CAULDRON);
                p_175707_.awardStat(Stats.ITEM_USED.get(p_175709_.getItem()));
-               p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL));
+               // p_175705_.setBlockAndUpdate(p_175706_, p_175704_.cycle(LayeredCauldronBlock.LEVEL));
                p_175705_.playSound((Player)null, p_175706_, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                p_175705_.gameEvent((Entity)null, GameEvent.FLUID_PLACE, p_175706_);
             }
@@ -231,11 +_,16 @@
          return InteractionResult.PASS;
       } else {
          if (!p_175637_.isClientSide) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.changeLevel(p_175636_, p_175637_, p_175638_, Blocks.CAULDRON.defaultBlockState(), p_175639_, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+               return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = p_175641_.getItem();
             p_175639_.setItemInHand(p_175640_, ItemUtils.createFilledResult(p_175641_, p_175639_, p_175642_));
             p_175639_.awardStat(Stats.USE_CAULDRON);
             p_175639_.awardStat(Stats.ITEM_USED.get(item));
-            p_175637_.setBlockAndUpdate(p_175638_, Blocks.CAULDRON.defaultBlockState());
+            // pLevel.setBlockAndUpdate(pPos, Blocks.CAULDRON.defaultBlockState());
             p_175637_.playSound((Player)null, p_175638_, p_175644_, SoundSource.BLOCKS, 1.0F, 1.0F);
             p_175637_.gameEvent((Entity)null, GameEvent.FLUID_PICKUP, p_175638_);
          }
@@ -246,11 +_,16 @@
 
    static InteractionResult emptyBucket(Level p_175619_, BlockPos p_175620_, Player p_175621_, InteractionHand p_175622_, ItemStack p_175623_, BlockState p_175624_, SoundEvent p_175625_) {
       if (!p_175619_.isClientSide) {
+         // CraftBukkit start
+         if (!LayeredCauldronBlock.changeLevel(p_175624_, p_175619_, p_175620_, p_175624_, p_175621_, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          Item item = p_175623_.getItem();
          p_175621_.setItemInHand(p_175622_, ItemUtils.createFilledResult(p_175623_, p_175621_, new ItemStack(Items.BUCKET)));
          p_175621_.awardStat(Stats.FILL_CAULDRON);
          p_175621_.awardStat(Stats.ITEM_USED.get(item));
-         p_175619_.setBlockAndUpdate(p_175620_, p_175624_);
+         // pLevel.setBlockAndUpdate(pPos, pState);
          p_175619_.playSound((Player)null, p_175620_, p_175625_, SoundSource.BLOCKS, 1.0F, 1.0F);
          p_175619_.gameEvent((Entity)null, GameEvent.FLUID_PLACE, p_175620_);
       }
