From 769edf31c51be7efc0ce210fe32a88a268338972 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Fri, 27 Mar 2020 18:03:45 +0100
Subject: [PATCH] Entity collision improvements

---
 .../java/org/bukkit/entity/LivingEntity.java  | 22 +++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 7f73e4b92..07e566f6f 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -610,6 +610,28 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     public void swingOffHand();
 
+    // Origami start - canCollide API
+    /**
+     * Set if this entity can collide with another entity.
+     * <p>
+     * This is not bidirectional like collidable!
+     * It only needs to be set for one of the entities to stop collisions.
+     *
+     * @param canCollide collide status
+     */
+    void setCanCollide(boolean canCollide);
+
+    /**
+     * Gets if this entity can collide with another entity.
+     * <p>
+     * This is not bidirectional like collidable!
+     * It only needs to be set for one of the entities to stop collisions.
+     *
+     * @return collide status
+     */
+    boolean canCollide();
+    // Origami end
+
     /**
      * Set if this entity will be subject to collisions other entities.
      * <p>
-- 
2.25.1.windows.1

