--- ../src-base/minecraft/net/minecraft/world/gen/FlatLayerInfo.java
+++ ../src-work/minecraft/net/minecraft/world/gen/FlatLayerInfo.java
@@ -18,7 +18,6 @@
 
     public FlatLayerInfo(int p_i45627_1_, int height, Block layerMaterialIn)
     {
-        this.layerCount = 1;
         this.version = p_i45627_1_;
         this.layerCount = height;
         this.layerMaterial = layerMaterialIn.getDefaultState();
