--- ../src-base/minecraft/org/bukkit/util/permissions/DefaultPermissions.java
+++ ../src-work/minecraft/org/bukkit/util/permissions/DefaultPermissions.java
@@ -9,7 +9,8 @@
     private static final String ROOT = "craftbukkit";
     private static final String LEGACY_PREFIX = "craft";
 
-    private DefaultPermissions() {}
+    private DefaultPermissions() {
+    }
 
     public static Permission registerPermission(Permission perm) {
         return registerPermission(perm, true);
