--- a/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java
@@ -12,7 +_,7 @@
    public static final int PODIUM_PILLAR_HEIGHT = 4;
    public static final int RIM_RADIUS = 1;
    public static final float CORNER_ROUNDING = 0.5F;
-   private static final BlockPos END_PODIUM_LOCATION = BlockPos.ZERO;
+   public static final BlockPos END_PODIUM_LOCATION = BlockPos.ZERO;
    private final boolean active;
 
    public static BlockPos getLocation(BlockPos p_287614_) {
