--- ../src-base/minecraft/net/minecraft/world/ChunkCache.java
+++ ../src-work/minecraft/net/minecraft/world/ChunkCache.java
@@ -34,7 +34,7 @@
         {
             for (int l = this.chunkZ; l <= j; ++l)
             {
-                this.chunkArray[k - this.chunkX][l - this.chunkZ] = worldIn.getChunkFromChunkCoords(k, l);
+                this.chunkArray[k - this.chunkX][l - this.chunkZ] = world.getChunkIfLoaded(k, l); // Paper
             }
         }
 
