--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -71,19 +_,19 @@
    public static final EntityDataAccessor<Rotations> f_31550_ = SynchedEntityData.m_135353_(ArmorStand.class, EntityDataSerializers.f_135037_);
    public static final EntityDataAccessor<Rotations> f_31527_ = SynchedEntityData.m_135353_(ArmorStand.class, EntityDataSerializers.f_135037_);
    private static final Predicate<Entity> f_31537_ = (p_31582_) -> {
-      return p_31582_ instanceof AbstractMinecart && ((AbstractMinecart)p_31582_).m_6064_() == AbstractMinecart.Type.RIDEABLE;
+      return p_31582_ instanceof AbstractMinecart && ((AbstractMinecart)p_31582_).canBeRidden();
    };
    private final NonNullList<ItemStack> f_31538_ = NonNullList.m_122780_(2, ItemStack.f_41583_);
    private final NonNullList<ItemStack> f_31539_ = NonNullList.m_122780_(4, ItemStack.f_41583_);
    private boolean f_31540_;
    public long f_31528_;
-   private int f_31541_;
-   private Rotations f_31542_ = f_31529_;
-   private Rotations f_31543_ = f_31530_;
-   private Rotations f_31544_ = f_31531_;
-   private Rotations f_31545_ = f_31532_;
-   private Rotations f_31525_ = f_31533_;
-   private Rotations f_31526_ = f_31534_;
+   public int f_31541_;
+   public Rotations f_31542_ = f_31529_;
+   public Rotations f_31543_ = f_31530_;
+   public Rotations f_31544_ = f_31531_;
+   public Rotations f_31545_ = f_31532_;
+   public Rotations f_31525_ = f_31533_;
+   public Rotations f_31526_ = f_31534_;
 
    public ArmorStand(EntityType<? extends ArmorStand> p_31553_, Level p_31554_) {
       super(p_31553_, p_31554_);
@@ -597,7 +_,7 @@
       return this.m_31677_() ? PushReaction.IGNORE : super.m_7752_();
    }
 
-   private void m_31603_(boolean p_31604_) {
+   public void m_31603_(boolean p_31604_) {
       this.f_19804_.m_135381_(f_31524_, this.m_31569_(this.f_19804_.m_135370_(f_31524_), 1, p_31604_));
    }
 
@@ -605,7 +_,7 @@
       return (this.f_19804_.m_135370_(f_31524_) & 1) != 0;
    }
 
-   private void m_31675_(boolean p_31676_) {
+   public void m_31675_(boolean p_31676_) {
       this.f_19804_.m_135381_(f_31524_, this.m_31569_(this.f_19804_.m_135370_(f_31524_), 4, p_31676_));
    }
 
@@ -613,7 +_,7 @@
       return (this.f_19804_.m_135370_(f_31524_) & 4) != 0;
    }
 
-   private void m_31678_(boolean p_31679_) {
+   public void m_31678_(boolean p_31679_) {
       this.f_19804_.m_135381_(f_31524_, this.m_31569_(this.f_19804_.m_135370_(f_31524_), 8, p_31679_));
    }
 
@@ -621,7 +_,7 @@
       return (this.f_19804_.m_135370_(f_31524_) & 8) != 0;
    }
 
-   private void m_31681_(boolean p_31682_) {
+   public void m_31681_(boolean p_31682_) {
       this.f_19804_.m_135381_(f_31524_, this.m_31569_(this.f_19804_.m_135370_(f_31524_), 16, p_31682_));
    }
 
