--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -40,8 +_,8 @@
 import org.apache.commons.lang3.Validate;
 
 public class ItemFrame extends HangingEntity {
-    private static final EntityDataAccessor<ItemStack> DATA_ITEM = SynchedEntityData.defineId(ItemFrame.class, EntityDataSerializers.ITEM_STACK);
-    private static final EntityDataAccessor<Integer> DATA_ROTATION = SynchedEntityData.defineId(ItemFrame.class, EntityDataSerializers.INT);
+    public static final EntityDataAccessor<ItemStack> DATA_ITEM = SynchedEntityData.defineId(ItemFrame.class, EntityDataSerializers.ITEM_STACK);
+    public static final EntityDataAccessor<Integer> DATA_ROTATION = SynchedEntityData.defineId(ItemFrame.class, EntityDataSerializers.INT);
     public static final int NUM_ROTATIONS = 8;
     private static final float DEPTH = 0.0625F;
     private static final float WIDTH = 0.75F;
@@ -69,7 +_,7 @@
     }
 
     @Override
-    protected void setDirection(Direction p_31793_) {
+    public void setDirection(Direction p_31793_) {
         Validate.notNull(p_31793_);
         this.direction = p_31793_;
         if (p_31793_.getAxis().isHorizontal()) {
@@ -94,6 +_,10 @@
         double d1 = direction$axis == Direction.Axis.Y ? 0.0625 : 0.75;
         double d2 = direction$axis == Direction.Axis.Z ? 0.0625 : 0.75;
         return AABB.ofSize(vec3, d0, d1, d2);
+    }
+
+    public static AABB calculateBoundingBoxStatic(BlockPos blockposition, Direction enumdirection) {
+        return calculateBoundingBoxStatic(blockposition, enumdirection);
     }
 
     @Override
