--- a/net/minecraft/world/item/context/BlockPlaceContext.java
+++ b/net/minecraft/world/item/context/BlockPlaceContext.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.item.context;
 
+import com.mohistmc.mohist.plugins.ban.bans.BanItem;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -53,6 +_,7 @@
     }
 
     public boolean canPlace() {
+      	if (BanItem.check(this.getPlayer(), this.getItemInHand())) return false;
         return this.replaceClicked || this.getLevel().getBlockState(this.getClickedPos()).canBeReplaced(this);
     }
 
