--- a/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacerType.java
+++ b/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacerType.java
@@ -22,7 +_,7 @@
         return Registry.register(BuiltInRegistries.FOLIAGE_PLACER_TYPE, p_68606_, new FoliagePlacerType<>(p_332653_));
     }
 
-    private FoliagePlacerType(MapCodec<P> p_328706_) {
+    public FoliagePlacerType(MapCodec<P> p_328706_) {
         this.codec = p_328706_;
     }
 
