--- a/net/minecraft/world/biome/BiomeContainer.java
+++ b/net/minecraft/world/biome/BiomeContainer.java
@@ -108,4 +108,13 @@
       int k = p_225526_3_ & field_227050_b_;
       return this.field_227054_f_[j << field_227052_d_ + field_227052_d_ | k << field_227052_d_ | i];
    }
+   
+   // CraftBukkit start
+   public void setBiome(final int i, final int j, final int k, final Biome biome) {
+      final int l = i & BiomeContainer.field_227050_b_;
+      final int i2 = MathHelper.func_76125_a(j, 0, BiomeContainer.field_227051_c_);
+      final int j2 = k & BiomeContainer.field_227050_b_;
+      this.field_227054_f_[i2 << BiomeContainer.field_227052_d_ + BiomeContainer.field_227052_d_ | j2 << BiomeContainer.field_227052_d_ | l] = biome;
+   }
+   // CraftBukkit end
 }
