From 7c7fddf2fd9cfbaca8c8fa1d577edf992a49768f Mon Sep 17 00:00:00 2001
From: ninja- <xninja@openmailbox.org>
Date: Tue, 8 Oct 2013 14:35:58 +0200
Subject: [PATCH] Add respawn API.


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index eb3141b6..fb83c016 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2208,6 +2208,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         public InetSocketAddress getRawAddress() {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+
+        /**
+         * Respawns the player if dead.
+         */
+        public void respawn() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
     }
 
     @NotNull
-- 
2.43.2

