--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -13,6 +_,8 @@
 import net.minecraft.world.item.crafting.RecipeType;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity;
+import org.bukkit.craftbukkit.inventory.CraftInventoryFurnace;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public abstract class AbstractFurnaceMenu extends RecipeBookMenu<Container> {
     public static final int INGREDIENT_SLOT = 0;
@@ -30,6 +_,22 @@
     private final RecipeType<? extends AbstractCookingRecipe> recipeType;
     private final RecipeBookType recipeBookType;
 
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Inventory player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryFurnace inventory = new CraftInventoryFurnace((AbstractFurnaceBlockEntity) this.container);
+        bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+
     protected AbstractFurnaceMenu(
         MenuType<?> p_38960_, RecipeType<? extends AbstractCookingRecipe> p_38961_, RecipeBookType p_38962_, int p_38963_, Inventory p_38964_
     ) {
@@ -56,7 +_,7 @@
         this.addSlot(new Slot(p_38971_, 0, 56, 17));
         this.addSlot(new FurnaceFuelSlot(this, p_38971_, 1, 56, 53));
         this.addSlot(new FurnaceResultSlot(p_38970_.player, p_38971_, 2, 116, 35));
-
+      	this.player = p_38970_; // CraftBukkit - save player
         for (int i = 0; i < 3; i++) {
             for (int j = 0; j < 9; j++) {
                 this.addSlot(new Slot(p_38970_, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
@@ -110,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player p_38974_) {
+      	if (!this.checkReachable) return true; // CraftBukkit
         return this.container.stillValid(p_38974_);
     }
 
@@ -167,7 +_,7 @@
     }
 
     protected boolean isFuel(ItemStack p_38989_) {
-        return AbstractFurnaceBlockEntity.isFuel(p_38989_);
+        return net.minecraftforge.common.ForgeHooks.getBurnTime(p_38989_, this.recipeType) > 0;
     }
 
     public float getBurnProgress() {
