--- a/net/minecraft/world/inventory/ContainerLevelAccess.java
+++ b/net/minecraft/world/inventory/ContainerLevelAccess.java
@@ -5,6 +_,7 @@
 import java.util.function.BiFunction;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.level.Level;
+import org.bukkit.World;
 
 public interface ContainerLevelAccess {
    ContainerLevelAccess f_39287_ = new ContainerLevelAccess() {
@@ -18,6 +_,17 @@
          public <T> Optional<T> m_6721_(BiFunction<Level, BlockPos, T> p_39311_) {
             return Optional.of(p_39311_.apply(p_39290_, p_39291_));
          }
+         // CraftBukkit start
+         @Override
+         public Level getWorld() {
+            return p_39290_;
+         }
+
+         @Override
+         public BlockPos getPosition() {
+            return p_39291_;
+         }
+         // CraftBukkit end
       };
    }
 
@@ -33,4 +_,18 @@
          return Optional.empty();
       });
    }
+
+   // CraftBukkit start
+   default Level getWorld() {
+      return this.m_6721_((level, blockpos) -> level).orElse(null); // Mohist
+   }
+
+   default BlockPos getPosition() {
+      return this.m_6721_((level, blockpos) -> blockpos).orElse(null); // Mohist
+   }
+
+   default org.bukkit.Location getLocation() {
+      return new org.bukkit.Location(getWorld().getWorld(), getPosition().m_123341_(), getPosition().m_123342_(), getPosition().m_123343_());
+   }
+   // CraftBukkit end
 }
