--- a/net/minecraft/world/inventory/CraftingContainer.java
+++ b/net/minecraft/world/inventory/CraftingContainer.java
@@ -4,6 +_,7 @@
 import net.minecraft.world.Container;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.CraftingInput;
+import net.minecraft.world.item.crafting.RecipeHolder;
 
 public interface CraftingContainer extends Container, StackedContentsCompatible {
     int getWidth();
@@ -11,6 +_,15 @@
     int getHeight();
 
     List<ItemStack> getItems();
+
+    // CraftBukkit start
+    default RecipeHolder<?> getCurrentRecipe() {
+        return null;
+    }
+
+    default void setCurrentRecipe(RecipeHolder<?> recipe) {
+    }
+    // CraftBukkit end
 
     default CraftingInput asCraftInput() {
         return this.asPositionedCraftInput().input();
