--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -1,7 +_,10 @@
 package net.minecraft.world.inventory;
 
-import com.mojang.datafixers.util.Pair;
 import java.util.List;
+import java.util.Map;
+
+import com.mohistmc.plugins.ban.bans.BanEnchantment;
+import com.mojang.datafixers.util.Pair;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -20,10 +_,21 @@
 import net.minecraft.world.item.EnchantedBookItem;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
+import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.enchantment.EnchantmentInstance;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.EnchantmentTableBlock;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.v1_20_R3.inventory.CraftInventoryEnchanting;
+import org.bukkit.craftbukkit.v1_20_R3.inventory.CraftInventoryView;
+import org.bukkit.craftbukkit.v1_20_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_20_R3.util.CraftNamespacedKey;
+import org.bukkit.enchantments.EnchantmentOffer;
+import org.bukkit.event.enchantment.EnchantItemEvent;
+import org.bukkit.event.enchantment.PrepareItemEnchantEvent;
 
 public class EnchantmentMenu extends AbstractContainerMenu {
    static final ResourceLocation EMPTY_SLOT_LAPIS_LAZULI = new ResourceLocation("item/empty_slot_lapis_lazuli");
@@ -32,6 +_,13 @@
          super.setChanged();
          EnchantmentMenu.this.slotsChanged(this);
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return access.getLocation();
+      }
+      // CraftBukkit end
    };
    private final ContainerLevelAccess access;
    private final RandomSource random = RandomSource.create();
@@ -40,6 +_,11 @@
    public final int[] enchantClue = new int[]{-1, -1, -1};
    public final int[] levelClue = new int[]{-1, -1, -1};
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+   // CraftBukkit end
+
    public EnchantmentMenu(int p_39454_, Inventory p_39455_) {
       this(p_39454_, p_39455_, ContainerLevelAccess.NULL);
    }
@@ -54,7 +_,7 @@
       });
       this.addSlot(new Slot(this.enchantSlots, 1, 35, 47) {
          public boolean mayPlace(ItemStack p_39517_) {
-            return p_39517_.is(Items.LAPIS_LAZULI);
+            return p_39517_.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS);
          }
 
          public Pair<ResourceLocation, ResourceLocation> getNoItemIcon() {
@@ -82,30 +_,34 @@
       this.addDataSlot(DataSlot.shared(this.levelClue, 0));
       this.addDataSlot(DataSlot.shared(this.levelClue, 1));
       this.addDataSlot(DataSlot.shared(this.levelClue, 2));
+      // CraftBukkit start
+      player = (org.bukkit.entity.Player) p_39458_.player.getBukkitEntity();
+      // CraftBukkit end
    }
 
    public void slotsChanged(Container p_39461_) {
       if (p_39461_ == this.enchantSlots) {
          ItemStack itemstack = p_39461_.getItem(0);
-         if (!itemstack.isEmpty() && itemstack.isEnchantable()) {
+         if (!itemstack.isEmpty()) { // CraftBukkit - relax condition
             this.access.execute((p_39485_, p_39486_) -> {
-               int j = 0;
+               float j = 0;
 
                for(BlockPos blockpos : EnchantmentTableBlock.BOOKSHELF_OFFSETS) {
                   if (EnchantmentTableBlock.isValidBookShelf(p_39485_, p_39486_, blockpos)) {
-                     ++j;
+                     j += p_39485_.getBlockState(p_39486_.offset(blockpos)).getEnchantPowerBonus(p_39485_, p_39486_.offset(blockpos));
                   }
                }
 
                this.random.setSeed((long)this.enchantmentSeed.get());
 
                for(int k = 0; k < 3; ++k) {
-                  this.costs[k] = EnchantmentHelper.getEnchantmentCost(this.random, k, j, itemstack);
+                  this.costs[k] = EnchantmentHelper.getEnchantmentCost(this.random, k, (int)j, itemstack);
                   this.enchantClue[k] = -1;
                   this.levelClue[k] = -1;
                   if (this.costs[k] < k + 1) {
                      this.costs[k] = 0;
                   }
+                  this.costs[k] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(p_39485_, p_39486_, k, (int)j, itemstack, costs[k]);
                }
 
                for(int l = 0; l < 3; ++l) {
@@ -119,6 +_,41 @@
                   }
                }
 
+               // CraftBukkit start
+               CraftItemStack item = CraftItemStack.asCraftMirror(itemstack);
+               org.bukkit.enchantments.EnchantmentOffer[] offers = new EnchantmentOffer[3];
+               for (int l = 0; l < 3; ++l) {
+                  org.bukkit.enchantments.Enchantment enchantment = (this.enchantClue[l] >= 0) ? org.bukkit.enchantments.Enchantment.getByKey(CraftNamespacedKey.fromMinecraft(BuiltInRegistries.ENCHANTMENT.getKey(BuiltInRegistries.ENCHANTMENT.byId(this.enchantClue[l])))) : null;
+                  offers[l] = (enchantment != null) ? new EnchantmentOffer(enchantment, this.levelClue[l], this.costs[l]) : null;
+               }
+
+               PrepareItemEnchantEvent event = new PrepareItemEnchantEvent(player, this.getBukkitView(), access.getLocation().getBlock(), item, offers, (int) j);
+               event.setCancelled(!itemstack.isEnchantable());
+               Bukkit.getPluginManager().callEvent(event);
+
+               if (event.isCancelled()) {
+                  for (int l = 0; l < 3; ++l) {
+                     this.costs[l] = 0;
+                     this.enchantClue[l] = -1;
+                     this.levelClue[l] = -1;
+                  }
+                  return;
+               }
+
+               for (int l = 0; l < 3; l++) {
+                  EnchantmentOffer offer = event.getOffers()[l];
+                  if (offer != null) {
+                     this.costs[l] = offer.getCost();
+                     this.enchantClue[l] = BuiltInRegistries.ENCHANTMENT.getId(BuiltInRegistries.ENCHANTMENT.get(CraftNamespacedKey.toMinecraft(offer.getEnchantment().getKey())));
+                     this.levelClue[l] = offer.getEnchantmentLevel();
+                  } else {
+                     this.costs[l] = 0;
+                     this.enchantClue[l] = -1;
+                     this.levelClue[l] = -1;
+                  }
+               }
+               // CraftBukkit end
+
                this.broadcastChanges();
             });
          } else {
@@ -142,12 +_,29 @@
          } else if (this.costs[p_39466_] <= 0 || itemstack.isEmpty() || (p_39465_.experienceLevel < i || p_39465_.experienceLevel < this.costs[p_39466_]) && !p_39465_.getAbilities().instabuild) {
             return false;
          } else {
-            this.access.execute((p_296875_, p_296876_) -> {
+            this.access.execute((p_39481_, p_39482_) -> {
                ItemStack itemstack2 = itemstack;
                List<EnchantmentInstance> list = this.getEnchantmentList(itemstack, p_39466_, this.costs[p_39466_]);
-               if (!list.isEmpty()) {
-                  p_39465_.onEnchantmentPerformed(itemstack, i);
+               if (true || !list.isEmpty()) {
+                  // pPlayer.onEnchantmentPerformed(itemstack, i); // Moved down
                   boolean flag = itemstack.is(Items.BOOK);
+                  Map<org.bukkit.enchantments.Enchantment, Integer> enchants = new java.util.HashMap<>();
+                  for (Object obj : list) {
+                     EnchantmentInstance instance = (EnchantmentInstance) obj;
+                     enchants.put(org.bukkit.enchantments.Enchantment.getByKey(CraftNamespacedKey.fromMinecraft(BuiltInRegistries.ENCHANTMENT.getKey(instance.enchantment))), instance.level);
+                  }
+                  CraftItemStack item = CraftItemStack.asCraftMirror(itemstack2);
+
+                  org.bukkit.enchantments.Enchantment hintedEnchantment = org.bukkit.enchantments.Enchantment.getByKey(CraftNamespacedKey.fromMinecraft(BuiltInRegistries.ENCHANTMENT.getKey(Enchantment.byId(enchantClue[p_39466_]))));
+                  if (BanEnchantment.check(hintedEnchantment)) return;
+                  int hintedEnchantmentLevel = levelClue[p_39466_];
+                  EnchantItemEvent event = new EnchantItemEvent((org.bukkit.entity.Player) p_39465_.getBukkitEntity(), this.getBukkitView(), access.getLocation().getBlock(), item, this.costs[p_39466_], enchants, hintedEnchantment, hintedEnchantmentLevel, p_39466_);
+                  Bukkit.getPluginManager().callEvent(event);
+
+                  int level = event.getExpLevelCost();
+                  if (event.isCancelled() || (level > p_39465_.experienceLevel && !p_39465_.getAbilities().instabuild) || event.getEnchantsToAdd().isEmpty()) {
+                     return;
+                  }
                   if (flag) {
                      itemstack2 = new ItemStack(Items.ENCHANTED_BOOK);
                      CompoundTag compoundtag = itemstack.getTag();
@@ -158,14 +_,29 @@
                      this.enchantSlots.setItem(0, itemstack2);
                   }
 
-                  for(EnchantmentInstance enchantmentinstance : list) {
-                     if (flag) {
-                        EnchantedBookItem.addEnchantment(itemstack2, enchantmentinstance);
-                     } else {
-                        itemstack2.enchant(enchantmentinstance.enchantment, enchantmentinstance.level);
+                  for (Map.Entry<org.bukkit.enchantments.Enchantment, Integer> entry : event.getEnchantsToAdd().entrySet()) {
+                     try {
+                        if (flag) {
+                           NamespacedKey enchantId = entry.getKey().getKey();
+                           net.minecraft.world.item.enchantment.Enchantment nms = BuiltInRegistries.ENCHANTMENT.get(CraftNamespacedKey.toMinecraft(enchantId));
+                           if (nms == null) {
+                              continue;
+                           }
+
+                           EnchantmentInstance weightedrandomenchant = new EnchantmentInstance(nms, entry.getValue());
+                           EnchantedBookItem.addEnchantment(itemstack2, weightedrandomenchant);
+                        } else {
+                           item.addUnsafeEnchantment(entry.getKey(), entry.getValue());
+                        }
+                     } catch (IllegalArgumentException e) {
+                        /* Just swallow invalid enchantments */
                      }
                   }
 
+                  p_39465_.onEnchantmentPerformed(itemstack, i);
+                  // CraftBukkit end
+
+                  // CraftBukkit - TODO: let plugins change this
                   if (!p_39465_.getAbilities().instabuild) {
                      itemstack1.shrink(i);
                      if (itemstack1.isEmpty()) {
@@ -181,7 +_,7 @@
                   this.enchantSlots.setChanged();
                   this.enchantmentSeed.set(p_39465_.getEnchantmentSeed());
                   this.slotsChanged(this.enchantSlots);
-                  p_296875_.playSound((Player)null, p_296876_, SoundEvents.ENCHANTMENT_TABLE_USE, SoundSource.BLOCKS, 1.0F, p_296875_.random.nextFloat() * 0.1F + 0.9F);
+                  p_39481_.playSound((Player)null, p_39482_, SoundEvents.ENCHANTMENT_TABLE_USE, SoundSource.BLOCKS, 1.0F, p_39481_.random.nextFloat() * 0.1F + 0.9F);
                }
 
             });
@@ -220,6 +_,7 @@
    }
 
    public boolean stillValid(Player p_39463_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return stillValid(this.access, p_39463_, Blocks.ENCHANTING_TABLE);
    }
 
@@ -237,7 +_,7 @@
             if (!this.moveItemStackTo(itemstack1, 2, 38, true)) {
                return ItemStack.EMPTY;
             }
-         } else if (itemstack1.is(Items.LAPIS_LAZULI)) {
+         } else if (itemstack1.is(net.minecraftforge.common.Tags.Items.ENCHANTING_FUELS)) {
             if (!this.moveItemStackTo(itemstack1, 1, 2, true)) {
                return ItemStack.EMPTY;
             }
@@ -266,4 +_,17 @@
 
       return itemstack;
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryEnchanting inventory = new CraftInventoryEnchanting(this.enchantSlots);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
