--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -44,10 +_,19 @@
     @Override
     protected void checkTakeAchievements(ItemStack p_39558_) {
         p_39558_.onCraftedBy(this.player.level(), this.player, this.removeCount);
-        if (this.player instanceof ServerPlayer serverplayer && this.container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+      Player player = this.player;
+      if (player instanceof ServerPlayer serverplayer) {
+         Container container = this.container;
+         if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+            // Mohist start
+            AbstractFurnaceBlockEntity.stackAtomicReference.set(p_39558_);
+            AbstractFurnaceBlockEntity.amount.set(this.removeCount);
             abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
-        }
+            // Mohist end
+         }
+      }
 
-        this.removeCount = 0;
-    }
+      this.removeCount = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.player, p_39558_);
+   }
 }
