--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -43,10 +_,15 @@
       if (player instanceof ServerPlayer serverplayer) {
          Container container = this.container;
          if (container instanceof AbstractFurnaceBlockEntity abstractfurnaceblockentity) {
+            // Mohist start
+            AbstractFurnaceBlockEntity.stackAtomicReference.set(p_39558_);
+            AbstractFurnaceBlockEntity.amount.set(this.removeCount);
             abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
+            // Mohist end
          }
       }
 
       this.removeCount = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.player, p_39558_);
    }
 }
