--- a/net/minecraft/world/inventory/HopperMenu.java
+++ b/net/minecraft/world/inventory/HopperMenu.java
@@ -5,11 +_,29 @@
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class HopperMenu extends AbstractContainerMenu {
     public static final int CONTAINER_SIZE = 5;
     private final Container hopper;
 
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Inventory player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventory inventory = new CraftInventory(this.hopper);
+        bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+
     public HopperMenu(int p_39640_, Inventory p_39641_) {
         this(p_39640_, p_39641_, new SimpleContainer(5));
     }
@@ -17,6 +_,7 @@
     public HopperMenu(int p_39643_, Inventory p_39644_, Container p_39645_) {
         super(MenuType.HOPPER, p_39643_);
         this.hopper = p_39645_;
+        this.player = p_39644_; // CraftBukkit - save player
         checkContainerSize(p_39645_, 5);
         p_39645_.startOpen(p_39644_.player);
         int i = 51;
@@ -38,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player p_39647_) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.hopper.stillValid(p_39647_);
     }
 
