--- a/net/minecraft/world/inventory/InventoryMenu.java
+++ b/net/minecraft/world/inventory/InventoryMenu.java
@@ -2,6 +_,7 @@
 
 import com.mojang.datafixers.util.Pair;
 import java.util.Map;
+import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.Container;
 import net.minecraft.world.entity.EquipmentSlot;
@@ -12,6 +_,9 @@
 import net.minecraft.world.item.crafting.CraftingInput;
 import net.minecraft.world.item.crafting.CraftingRecipe;
 import net.minecraft.world.item.crafting.RecipeHolder;
+import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class InventoryMenu extends RecipeBookMenu<CraftingInput, CraftingRecipe> {
     public static final int CONTAINER_ID = 0;
@@ -41,11 +_,29 @@
     private final ResultContainer resultSlots = new ResultContainer();
     public final boolean active;
     private final Player owner;
+   	// CraftBukkit start
+   	private CraftInventoryView bukkitEntity = null;
+   	private Inventory player;
+  	@Override
+   	public CraftInventoryView getBukkitView() {
+      	if (bukkitEntity != null) {
+         	return bukkitEntity;
+      	}
+
+      	CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.craftSlots, this.resultSlots);
+      	bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+      	return bukkitEntity;
+   	}
+   	// CraftBukkit end
 
     public InventoryMenu(Inventory p_39706_, boolean p_39707_, final Player p_39708_) {
         super(null, 0);
         this.active = p_39707_;
         this.owner = p_39708_;
+      	((TransientCraftingContainer)this.craftSlots).resultInventory = this.resultSlots; // CraftBukkit - let InventoryCrafting know about its result slot
+      	((TransientCraftingContainer)this.craftSlots).setOwner(p_39706_.player);
+      	this.player = p_39706_; // CraftBukkit - save player
+        setTitle(Component.translatable("container.crafting")); // SPIGOT-4722: Allocate title for player inventory
         this.addSlot(new ResultSlot(p_39706_.player, this.craftSlots, this.resultSlots, 0, 154, 28));
 
         for (int i = 0; i < 2; i++) {
