--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -23,6 +_,9 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.entity.BannerPatternLayers;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class LoomMenu extends AbstractContainerMenu {
     private static final int PATTERN_NOT_SET = -1;
@@ -48,14 +_,26 @@
             LoomMenu.this.slotsChanged(this);
             LoomMenu.this.slotUpdateListener.run();
         }
-    };
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return LoomMenu.this.access.getLocation();
+      }
+      // CraftBukkit end
+   };
     private final Container outputContainer = new SimpleContainer(1) {
         @Override
         public void setChanged() {
             super.setChanged();
             LoomMenu.this.slotUpdateListener.run();
         }
-    };
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return LoomMenu.this.access.getLocation();
+      }
+      // CraftBukkit end
+   };
 
     public LoomMenu(int p_39856_, Inventory p_39857_) {
         this(p_39856_, p_39857_, ContainerLevelAccess.NULL);
@@ -306,4 +_,20 @@
     public Slot getResultSlot() {
         return this.resultSlot;
     }
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private org.bukkit.entity.Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      CraftInventoryLoom inventory = new CraftInventoryLoom(this.inputContainer, this.outputContainer);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
