--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -10,6 +_,8 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.v1_18_R1.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class MerchantMenu extends AbstractContainerMenu {
    protected static final int f_150619_ = 0;
@@ -28,6 +_,19 @@
    private int f_40029_;
    private boolean f_40030_;
    private boolean f_40031_;
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Inventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity == null) {
+         bukkitEntity = new CraftInventoryView(this.player.f_35978_.getBukkitEntity(), new org.bukkit.craftbukkit.v1_18_R1.inventory.CraftInventoryMerchant(f_40027_, f_40028_), this);
+      }
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 
    public MerchantMenu(int p_40033_, Inventory p_40034_) {
       this(p_40033_, p_40034_, new ClientSideMerchant(p_40034_.f_35978_));
