--- a/net/minecraft/world/level/levelgen/carver/CarverDebugSettings.java
+++ b/net/minecraft/world/level/levelgen/carver/CarverDebugSettings.java
@@ -10,7 +_,7 @@
    public static final Codec<CarverDebugSettings> f_159115_ = RecordCodecBuilder.create((p_159135_) -> {
       return p_159135_.group(Codec.BOOL.optionalFieldOf("debug_mode", Boolean.valueOf(false)).forGetter(CarverDebugSettings::m_159128_), BlockState.f_61039_.optionalFieldOf("air_state", f_159114_.m_159145_()).forGetter(CarverDebugSettings::m_159145_), BlockState.f_61039_.optionalFieldOf("water_state", f_159114_.m_159145_()).forGetter(CarverDebugSettings::m_159146_), BlockState.f_61039_.optionalFieldOf("lava_state", f_159114_.m_159145_()).forGetter(CarverDebugSettings::m_159147_), BlockState.f_61039_.optionalFieldOf("barrier_state", f_159114_.m_159145_()).forGetter(CarverDebugSettings::m_159148_)).apply(p_159135_, CarverDebugSettings::new);
    });
-   private final boolean f_159116_;
+   private boolean f_159116_;
    private final BlockState f_159117_;
    private final BlockState f_159118_;
    private final BlockState f_159119_;
