--- a/net/minecraft/advancements/AdvancementHolder.java
+++ b/net/minecraft/advancements/AdvancementHolder.java
@@ -3,8 +_,8 @@
 import java.util.List;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.ResourceLocation;
+import org.bukkit.craftbukkit.advancement.CraftAdvancement;
 
 public record AdvancementHolder(ResourceLocation id, Advancement value) {
     public static final StreamCodec<RegistryFriendlyByteBuf, AdvancementHolder> STREAM_CODEC = StreamCodec.composite(
@@ -21,9 +_,9 @@
                 return true;
             }
 
-            return false;
-        }
-    }
+         return false;
+      }
+   }
 
     @Override
     public int hashCode() {
@@ -34,4 +_,10 @@
     public String toString() {
         return this.id.toString();
     }
+
+   // CraftBukkit start
+   public final org.bukkit.advancement.Advancement toBukkit() {
+      return new CraftAdvancement(this);
+   }
+   // CraftBukkit end
 }
