--- a/net/minecraft/advancements/AdvancementList.java
+++ b/net/minecraft/advancements/AdvancementList.java
@@ -96,7 +96,8 @@
          }
       }
 
-      field_192091_a.info("Loaded {} advancements", (int)this.field_192092_b.size());
+      net.minecraftforge.common.AdvancementLoadFix.buildSortedTrees(this.field_192093_c);
+      //LOGGER.info("Loaded {} advancements", (int)this.advancements.size()); // CraftBukkit - moved to AdvancementManager#reload
    }
 
    @OnlyIn(Dist.CLIENT)
