--- a/net/minecraft/advancements/PlayerAdvancements.java
+++ b/net/minecraft/advancements/PlayerAdvancements.java
@@ -141,7 +141,11 @@
             for(Entry<ResourceLocation, AdvancementProgress> entry : stream.collect(Collectors.toList())) {
                Advancement advancement = this.field_192756_d.func_191949_aK().func_192778_a(entry.getKey());
                if (advancement == null) {
-                  field_192753_a.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), this.field_192757_e);
+                  // CraftBukkit start
+                  if(entry.getKey().func_110624_b().equals("minecraft")){
+                     field_192753_a.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", entry.getKey(), this.field_192757_e);
+                  }
+                  // CraftBukkit end
                } else {
                   this.func_192743_a(advancement, entry.getValue());
                }
@@ -154,11 +158,13 @@
       }
 
       this.func_192748_e();
+      if (net.minecraftforge.common.ForgeConfig.SERVER.fixAdvancementLoading.get()) net.minecraftforge.common.AdvancementLoadFix.loadVisibility(this, this.field_192759_g, this.field_192760_h, this.field_192758_f, this.field_192761_i, this::func_192738_c); else
       this.func_192752_d();
       this.func_192751_c();
    }
 
    public void func_192749_b() {
+      if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return;
       Map<ResourceLocation, AdvancementProgress> map = Maps.newHashMap();
 
       for(Entry<Advancement, AdvancementProgress> entry : this.field_192758_f.entrySet()) {
@@ -187,6 +193,8 @@
    }
 
    public boolean func_192750_a(Advancement p_192750_1_, String p_192750_2_) {
+      // Forge: don't grant advancements for fake players
+      if (this.field_192762_j instanceof net.minecraftforge.common.util.FakePlayer) return false;
       boolean flag = false;
       AdvancementProgress advancementprogress = this.func_192747_a(p_192750_1_);
       boolean flag1 = advancementprogress.func_192105_a();
@@ -195,10 +203,12 @@
          this.field_192761_i.add(p_192750_1_);
          flag = true;
          if (!flag1 && advancementprogress.func_192105_a()) {
+            this.field_192762_j.field_70170_p.getServerCB().getPluginManager().callEvent(new org.bukkit.event.player.PlayerAdvancementDoneEvent(this.field_192762_j.getBukkitEntity(), p_192750_1_.bukkit)); // CraftBukkit
             p_192750_1_.func_192072_d().func_192113_a(this.field_192762_j);
             if (p_192750_1_.func_192068_c() != null && p_192750_1_.func_192068_c().func_193220_i() && this.field_192762_j.field_70170_p.func_82736_K().func_223586_b(GameRules.field_223620_w)) {
                this.field_192756_d.func_184103_al().func_148539_a(new TranslationTextComponent("chat.type.advancement." + p_192750_1_.func_192068_c().func_192291_d().func_192307_a(), this.field_192762_j.func_145748_c_(), p_192750_1_.func_193123_j()));
             }
+            net.minecraftforge.common.ForgeHooks.onAdvancement(this.field_192762_j, p_192750_1_);
          }
       }
 
