--- a/net/minecraft/world/level/LevelSettings.java
+++ b/net/minecraft/world/level/LevelSettings.java
@@ -4,9 +_,9 @@
 import net.minecraft.world.Difficulty;
 
 public final class LevelSettings {
-   private final String f_46902_;
+   public String f_46902_;
    private final GameType f_46903_;
-   private final boolean f_46904_;
+   public boolean f_46904_;
    private final Difficulty f_46905_;
    private final boolean f_46906_;
    private final GameRules f_46907_;
@@ -62,6 +_,7 @@
    }
 
    public LevelSettings m_46918_(Difficulty p_46919_) {
+      net.minecraftforge.common.ForgeHooks.onDifficultyChange(p_46919_, this.f_46905_);
       return new LevelSettings(this.f_46902_, this.f_46903_, this.f_46904_, p_46919_, this.f_46906_, this.f_46907_, this.f_46908_);
    }
 
