--- a/net/minecraft/world/level/WorldGenLevel.java
+++ b/net/minecraft/world/level/WorldGenLevel.java
@@ -5,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.levelgen.feature.StructureFeature;
 import net.minecraft.world.level.levelgen.structure.StructureStart;
 
@@ -18,5 +_,10 @@
    }
 
    default void m_183406_(@Nullable Supplier<String> p_186618_) {
+   }
+
+   @Override
+   default ServerLevel getMinecraftWorld(){
+      return m_6018_();
    }
 }
