--- a/net/minecraft/server/management/ProfileBanEntry.java
+++ b/net/minecraft/server/management/ProfileBanEntry.java
@@ -36,19 +36,26 @@
    }
 
    private static GameProfile func_152648_b(JsonObject p_152648_0_) {
-      if (p_152648_0_.has("uuid") && p_152648_0_.has("name")) {
+      // Spigot start
+      // this whole method has to be reworked to account for the fact Bukkit only accepts UUID bans and gives no way for usernames to be stored!
+      UUID uuid = null;
+      String name = null;
+      if (p_152648_0_.has("uuid")) {
          String s = p_152648_0_.get("uuid").getAsString();
 
-         UUID uuid;
          try {
             uuid = UUID.fromString(s);
          } catch (Throwable var4) {
-            return null;
          }
-
-         return new GameProfile(uuid, p_152648_0_.get("name").getAsString());
+      }
+      if (p_152648_0_.has("name")) {
+         name = p_152648_0_.get("name").getAsString();
+      }
+      if (uuid != null || name != null) {
+         return new GameProfile(uuid, name);
       } else {
          return null;
       }
+      // Spigot End
    }
 }
