--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -208,7 +_,14 @@
             }
 
             frog.setPersistenceRequired();
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_20_R2.event.CraftEventFactory.callEntityTransformEvent(this, frog, org.bukkit.event.entity.EntityTransformEvent.TransformReason.METAMORPHOSIS).isCancelled()) {
+               this.setAge(0); // Sets the age to 0 for avoid a loop if the event is canceled
+               return;
+            }
+            // CraftBukkit end
             this.playSound(SoundEvents.TADPOLE_GROW_UP, 0.15F, 1.0F);
+            frog.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.METAMORPHOSIS); // CraftBukkit - add SpawnReason
             serverlevel.addFreshEntityWithPassengers(frog);
             this.discard();
          }
