--- a/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -8,7 +_,7 @@
    private static final int MAX_HOST_LENGTH = 255;
 
    public ClientIntentionPacket(FriendlyByteBuf p_179801_) {
-      this(p_179801_.readVarInt(), p_179801_.readUtf(255), p_179801_.readUnsignedShort(), ClientIntent.byId(p_179801_.readVarInt()));
+      this(p_179801_.readVarInt(), p_179801_.readUtf(Short.MAX_VALUE), p_179801_.readUnsignedShort(), ClientIntent.byId(p_179801_.readVarInt()));
    }
 
    public void write(FriendlyByteBuf p_134737_) {
