--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -139,6 +_,7 @@
       WorldDimensions.Complete worlddimensions$complete = worldgensettings.dimensions().bake(p_311619_);
       Lifecycle lifecycle = worlddimensions$complete.lifecycle().add(p_313214_.allRegistriesLifecycle());
       PrimaryLevelData primaryleveldata = PrimaryLevelData.parse(dynamic, levelsettings, worlddimensions$complete.specialWorldProperty(), worldgensettings.options(), lifecycle);
+      primaryleveldata.pdc = ((Dynamic<Tag>) dynamic1).getElement("BukkitValues", null); // CraftBukkit - Add PDC to world
       return new LevelDataAndDimensions(primaryleveldata, worlddimensions$complete);
    }
 
@@ -407,7 +_,7 @@
 
    public class LevelStorageAccess implements AutoCloseable {
       final DirectoryLock lock;
-      final LevelStorageSource.LevelDirectory levelDirectory;
+      public final LevelStorageSource.LevelDirectory levelDirectory;
       private final String levelId;
       private final Map<LevelResource, Path> resources = Maps.newHashMap();
 
