--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -8,6 +_,7 @@
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.Lifecycle;
+
 import java.util.HashSet;
 import java.util.Set;
 import java.util.UUID;
@@ -17,6 +_,7 @@
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.UUIDUtil;
 import net.minecraft.nbt.CompoundTag;
@@ -25,8 +_,11 @@
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.StringTag;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket;
 import net.minecraft.resources.RegistryOps;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.level.GameRules;
@@ -35,11 +_,15 @@
 import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.WorldDataConfiguration;
 import net.minecraft.world.level.border.WorldBorder;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 import net.minecraft.world.level.levelgen.WorldOptions;
 import net.minecraft.world.level.timers.TimerCallbacks;
 import net.minecraft.world.level.timers.TimerQueue;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
 import org.slf4j.Logger;
 
 public class PrimaryLevelData implements ServerLevelData, WorldData {
@@ -82,6 +_,22 @@
    private boolean wasModded;
    private final Set<String> removedFeatureFlags;
    private final TimerQueue<MinecraftServer> scheduledEvents;
+   private boolean confirmedExperimentalWarning = false;
+
+   // CraftBukkit start - Add world and pdc
+   public Registry<LevelStem> customDimensions;
+   private ServerLevel world;
+   protected Tag pdc;
+
+   public void setWorld(ServerLevel world) {
+      if (this.world != null) {
+         return;
+      }
+      this.world = world;
+      if (world.getWorld() != null) world.getWorld().readBukkitValues(pdc);
+      pdc = null;
+   }
+   // CraftBukkit end
 
    private PrimaryLevelData(@Nullable DataFixer p_277859_, int p_277672_, @Nullable CompoundTag p_277888_, boolean p_278109_, int p_277714_, int p_278088_, int p_278037_, float p_277542_, long p_277414_, long p_277635_, int p_277595_, int p_277794_, int p_278007_, boolean p_277943_, int p_277674_, boolean p_277644_, boolean p_277749_, boolean p_278004_, WorldBorder.Settings p_277729_, int p_277856_, int p_278051_, @Nullable UUID p_277341_, Set<String> p_277989_, Set<String> p_277399_, TimerQueue<MinecraftServer> p_277860_, @Nullable CompoundTag p_277936_, EndDragonFight.Data p_289764_, LevelSettings p_278064_, WorldOptions p_278072_, PrimaryLevelData.SpecialWorldProperty p_277548_, Lifecycle p_277915_) {
       this.fixerUpper = p_277859_;
@@ -127,7 +_,7 @@
          return p_78529_.asString().result().stream();
       }).collect(Collectors.toCollection(Sets::newLinkedHashSet)), p_78531_.get("removed_features").asStream().flatMap((p_277335_) -> {
          return p_277335_.asString().result().stream();
-      }).collect(Collectors.toSet()), new TimerQueue<>(TimerCallbacks.SERVER_CALLBACKS, p_78531_.get("ScheduledEvents").asStream()), (CompoundTag)p_78531_.get("CustomBossEvents").orElseEmptyMap().getValue(), p_78531_.get("DragonFight").read(EndDragonFight.Data.CODEC).resultOrPartial(LOGGER::error).orElse(EndDragonFight.Data.DEFAULT), p_78535_, p_251864_, p_250651_, p_78538_);
+      }).collect(Collectors.toSet()), new TimerQueue<>(TimerCallbacks.SERVER_CALLBACKS, p_78531_.get("ScheduledEvents").asStream()), (CompoundTag)p_78531_.get("CustomBossEvents").orElseEmptyMap().getValue(), p_78531_.get("DragonFight").read(EndDragonFight.Data.CODEC).resultOrPartial(LOGGER::error).orElse(EndDragonFight.Data.DEFAULT), p_78535_, p_251864_, p_250651_, p_78538_).withConfirmedWarning(p_78538_ != Lifecycle.stable() && p_78531_.get("confirmedExperimentalSettings").asBoolean(false));
    }
 
    public CompoundTag createTag(RegistryAccess p_78543_, @Nullable CompoundTag p_78544_) {
@@ -178,7 +_,7 @@
       p_78547_.putBoolean("allowCommands", this.settings.allowCommands());
       p_78547_.putBoolean("initialized", this.initialized);
       this.worldBorder.write(p_78547_);
-      p_78547_.putByte("Difficulty", (byte)this.settings.difficulty().getId());
+      p_78547_.putByte("Difficulty", (byte) this.settings.difficulty().getId());
       p_78547_.putBoolean("DifficultyLocked", this.difficultyLocked);
       p_78547_.put("GameRules", this.settings.gameRules().createTag());
       p_78547_.put("DragonFight", Util.getOrThrow(EndDragonFight.Data.CODEC.encodeStart(NbtOps.INSTANCE, this.endDragonFightData), IllegalStateException::new));
@@ -188,9 +_,9 @@
 
       DataResult<Tag> dataresult = WorldDataConfiguration.CODEC.encodeStart(NbtOps.INSTANCE, this.settings.getDataConfiguration());
       dataresult.get().ifLeft((p_248505_) -> {
-         p_78547_.merge((CompoundTag)p_248505_);
+         p_78547_.merge((CompoundTag) p_248505_);
       }).ifRight((p_248506_) -> {
-         LOGGER.warn("Failed to encode configuration {}", (Object)p_248506_.message());
+         LOGGER.warn("Failed to encode configuration {}", (Object) p_248506_.message());
       });
       if (this.customBossEvents != null) {
          p_78547_.put("CustomBossEvents", this.customBossEvents);
@@ -203,6 +_,14 @@
          p_78547_.putUUID("WanderingTraderId", this.wanderingTraderId);
       }
 
+      p_78547_.putString("forgeLifecycle", net.minecraftforge.common.ForgeHooks.encodeLifecycle(this.settings.getLifecycle()));
+      p_78547_.putBoolean("confirmedExperimentalSettings", this.confirmedExperimentalWarning);
+
+      if (Bukkit.getServer() != null) {
+         p_78547_.putString("Bukkit.Version", Bukkit.getName() + "/" + Bukkit.getVersion() + "/" + Bukkit.getBukkitVersion()); // CraftBukkit
+         world.getWorld().storeBukkitValues(p_78547_); // CraftBukkit - add pdc
+      }
+
    }
 
    private static ListTag stringCollectionToTag(Set<String> p_277880_) {
@@ -306,6 +_,20 @@
    }
 
    public void setThundering(boolean p_78562_) {
+      // CraftBukkit start
+      if (this.thundering == p_78562_) {
+         return;
+      }
+
+      org.bukkit.World world = Bukkit.getWorld(getLevelName());
+      if (world != null) {
+         ThunderChangeEvent thunder = new ThunderChangeEvent(world, p_78562_);
+         Bukkit.getServer().getPluginManager().callEvent(thunder);
+         if (thunder.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.thundering = p_78562_;
    }
 
@@ -322,6 +_,20 @@
    }
 
    public void setRaining(boolean p_78576_) {
+      // CraftBukkit start
+      if (this.raining == p_78576_) {
+         return;
+      }
+
+      org.bukkit.World world = Bukkit.getWorld(getLevelName());
+      if (world != null) {
+         WeatherChangeEvent weather = new WeatherChangeEvent(world, p_78576_);
+         Bukkit.getServer().getPluginManager().callEvent(weather);
+         if (weather.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.raining = p_78576_;
    }
 
@@ -375,6 +_,12 @@
 
    public void setDifficulty(Difficulty p_78521_) {
       this.settings = this.settings.withDifficulty(p_78521_);
+      // CraftBukkit start
+      ClientboundChangeDifficultyPacket packet = new ClientboundChangeDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+      for (ServerPlayer player : (java.util.List<ServerPlayer>) (java.util.List) world.players()) {
+         player.connection.send(packet);
+      }
+      // CraftBukkit end
    }
 
    public boolean isDifficultyLocked() {
@@ -485,6 +_,15 @@
       return this.settings.copy();
    }
 
+   public boolean hasConfirmedExperimentalWarning() {
+      return this.confirmedExperimentalWarning;
+   }
+
+   public PrimaryLevelData withConfirmedWarning(boolean confirmedWarning) { // Builder-like to not patch ctor
+      this.confirmedExperimentalWarning = confirmedWarning;
+      return this;
+   }
+
    /** @deprecated */
    @Deprecated
    public static enum SpecialWorldProperty {
@@ -492,4 +_,12 @@
       FLAT,
       DEBUG;
    }
+
+   // CraftBukkit start - Check if the name stored in NBT is the correct one
+   public void checkName(String name) {
+      if (!this.settings.levelName.equals(name)) {
+         this.settings.levelName = name;
+      }
+   }
+   // CraftBukkit end
 }
