--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -21,8 +_,11 @@
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.StringTag;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket;
 import net.minecraft.resources.RegistryWriteOps;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.level.DataPackConfig;
@@ -36,6 +_,9 @@
 import net.minecraft.world.level.timers.TimerQueue;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
 
 public class PrimaryLevelData implements ServerLevelData, WorldData {
    private static final Logger f_78442_ = LogManager.getLogger();
@@ -74,6 +_,7 @@
    private final Set<String> f_78439_;
    private boolean f_78440_;
    private final TimerQueue<MinecraftServer> f_78441_;
+   public ServerLevel world; // CraftBukkit
 
    private PrimaryLevelData(@Nullable DataFixer p_164942_, int p_164943_, @Nullable CompoundTag p_164944_, boolean p_164945_, int p_164946_, int p_164947_, int p_164948_, float p_164949_, long p_164950_, long p_164951_, int p_164952_, int p_164953_, int p_164954_, boolean p_164955_, int p_164956_, boolean p_164957_, boolean p_164958_, boolean p_164959_, WorldBorder.Settings p_164960_, int p_164961_, int p_164962_, @Nullable UUID p_164963_, Set<String> p_164964_, TimerQueue<MinecraftServer> p_164965_, @Nullable CompoundTag p_164966_, CompoundTag p_164967_, LevelSettings p_164968_, WorldGenSettings p_164969_, Lifecycle p_164970_) {
       this.f_78452_ = p_164942_;
@@ -189,6 +_,7 @@
          p_78547_.m_128362_("WanderingTraderId", this.f_78438_);
       }
 
+      p_78547_.m_128359_("Bukkit.Version", Bukkit.getName() + "/" + Bukkit.getVersion() + "/" + Bukkit.getBukkitVersion()); // CraftBukkit
    }
 
    public int m_6789_() {
@@ -286,6 +_,19 @@
    }
 
    public void m_5557_(boolean p_78562_) {
+      // CraftBukkit start
+      if (this.f_78460_ == p_78562_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(m_5462_());
+      if (world != null) {
+         ThunderChangeEvent thunder = new ThunderChangeEvent(world, p_78562_);
+         Bukkit.getServer().getPluginManager().callEvent(thunder);
+         if (thunder.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.f_78460_ = p_78562_;
    }
 
@@ -302,6 +_,19 @@
    }
 
    public void m_5565_(boolean p_78576_) {
+      // CraftBukkit start
+      if (this.f_78458_ == p_78576_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(m_5462_());
+      if (world != null) {
+         WeatherChangeEvent weather = new WeatherChangeEvent(world, p_78576_);
+         Bukkit.getServer().getPluginManager().callEvent(weather);
+         if (weather.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.f_78458_ = p_78576_;
    }
 
@@ -355,6 +_,12 @@
 
    public void m_6166_(Difficulty p_78521_) {
       this.f_78443_ = this.f_78443_.m_46918_(p_78521_);
+      // CraftBukkit start
+      ClientboundChangeDifficultyPacket packet = new ClientboundChangeDifficultyPacket(this.m_5472_(), this.m_5474_());
+      for (ServerPlayer player : (java.util.List<ServerPlayer>) (java.util.List) world.m_6907_()) {
+         player.f_8906_.m_141995_(packet);
+      }
+      // CraftBukkit end
    }
 
    public boolean m_5474_() {
@@ -452,4 +_,12 @@
    public LevelSettings m_5926_() {
       return this.f_78443_.m_46935_();
    }
+
+   // CraftBukkit start - Check if the name stored in NBT is the correct one
+   public void checkName(String name) {
+      if (!this.f_78443_.f_46902_.equals(name)) {
+         this.f_78443_.f_46902_ = name;
+      }
+   }
+   // CraftBukkit end
 }
