From c4b43e9f560626d937392a6b602f809c3336e422 Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Thu, 22 Aug 2019 20:20:02 -0400
Subject: [PATCH] Configuration file


diff --git a/src/main/java/com/canyonmodded/config/CanyonConfig.java b/src/main/java/com/canyonmodded/config/CanyonConfig.java
new file mode 100644
index 00000000..d7dffe62
--- /dev/null
+++ b/src/main/java/com/canyonmodded/config/CanyonConfig.java
@@ -0,0 +1,18 @@
+package com.canyonmodded.config;
+
+import org.bukkit.util.config.Configuration;
+
+import java.io.File;
+
+public class CanyonConfig {
+    // Add settings here. Make sure they're public visibility!
+
+    public static void loadConfiguration(File file) {
+        Configuration configuration = new Configuration(file);
+        configuration.load();
+
+        // Load settings here.
+
+        configuration.save();
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index b2123565..4f1902fc 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2,6 +2,7 @@ package org.bukkit.craftbukkit;
 
 import java.io.FileNotFoundException;
 
+import com.canyonmodded.config.CanyonConfig;
 import org.bukkit.generator.ChunkGenerator;
 import com.avaje.ebean.config.DataSourceConfig;
 import com.avaje.ebean.config.ServerConfig;
@@ -125,6 +126,8 @@ public final class CraftServer implements Server {
             configuration.setProperty("aliases.icanhasbukkit", icanhasbukkit);
         }
         configuration.save();
+
+        CanyonConfig.loadConfiguration((File) console.options.valueOf("canyon-settings")); // Canyon
     }
 
     public void loadPlugins() {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 68b9abd2..394c6a97 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -92,6 +92,14 @@ public class Main {
                         .defaultsTo(new File("bukkit.yml"))
                         .describedAs("Yml file");
 
+                // Canyon Start
+                acceptsAll(asList("c", "canyon-settings"), "File for Canyon settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("canyon.yml"))
+                        .describedAs("Yml file");
+                // Canyon End
+
                 acceptsAll(asList("nojline"), "Disables jline and emulates the vanilla console");
 
                 acceptsAll(asList("v", "version"), "Show the CraftBukkit Version");
-- 
2.24.1.windows.2

