From 75afb5ebc12256bf94764ab6bd0bc5c5526d565d Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Sun, 29 Sep 2019 18:54:13 -0400
Subject: [PATCH] Optimize hash codes for some classes used as keys


diff --git a/src/main/java/net/minecraft/server/NextTickListEntry.java b/src/main/java/net/minecraft/server/NextTickListEntry.java
index b461a80f..76e8c9ff 100644
--- a/src/main/java/net/minecraft/server/NextTickListEntry.java
+++ b/src/main/java/net/minecraft/server/NextTickListEntry.java
@@ -29,7 +29,14 @@ public class NextTickListEntry implements Comparable<NextTickListEntry> { // Can
     }
 
     public int hashCode() {
-        return (this.a * 128 * 1024 + this.c * 128 + this.b) * 256 + this.d;
+        // Canyon start - improve hash code
+        //return (this.a * 128 * 1024 + this.c * 128 + this.b) * 256 + this.d;
+        int result = Integer.hashCode(a);
+        result = 31 * result + Integer.hashCode(c);
+        result = 31 * result + Integer.hashCode(b);
+        result = 31 * result + Integer.hashCode(d);
+        return result;
+        // Canyon end
     }
 
     public NextTickListEntry a(long i) {
-- 
2.25.2

