From e24867ff3d86fd71937a6bee18fc68f946b61bd6 Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Sun, 6 Oct 2019 17:06:39 -0400
Subject: [PATCH] Disable Nagle's algorithm


diff --git a/src/main/java/com/canyonmodded/network/NettyBootstrap.java b/src/main/java/com/canyonmodded/network/NettyBootstrap.java
index 0e58eeff..c4bf0196 100644
--- a/src/main/java/com/canyonmodded/network/NettyBootstrap.java
+++ b/src/main/java/com/canyonmodded/network/NettyBootstrap.java
@@ -76,6 +76,8 @@ public class NettyBootstrap {
                     @Override
                     protected void initChannel(Channel channel) throws Exception {
                         channel.config().setOption(ChannelOption.IP_TOS, 24);
+                        // Disable Nagle's algorithm
+                        channel.config().setOption(ChannelOption.TCP_NODELAY, true);
 
                         NetLoginHandler handler = new NetLoginHandler(server, channel);
                         channel.pipeline()
diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 85618930..e393ad56 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -59,6 +59,7 @@ public class NetworkManager {
         try {
             // CraftBukkit start - cant compile these outside the try
             socket.setSoTimeout(30000);
+            socket.setTcpNoDelay(true); // Canyon - disable Nagle's algorithm
             this.input = new DataInputStream(socket.getInputStream());
             this.output = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream(), 5120));
         } catch (java.io.IOException socketexception) {
-- 
2.25.2

