From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 01:19:40 -0500
Subject: [PATCH] EMC ServerShutdownEvent


diff --git a/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java b/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..43f761a17ea6db23c9bbec686959d3c1f249edf8
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java
@@ -0,0 +1,29 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+public class ServerShutdownEvent extends Event {
+    private String reason;
+    public ServerShutdownEvent(String reason) {
+        this.reason = reason;
+    }
+
+    public String getReason() {
+        return reason;
+    }
+
+    public void setReason(String reason) {
+        this.reason = reason;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
