--- ../src-base/minecraft/net/minecraftforge/fml/server/FMLServerHandler.java
+++ ../src-work/minecraft/net/minecraftforge/fml/server/FMLServerHandler.java
@@ -19,13 +19,17 @@
 
 package net.minecraftforge.fml.server;
 
-import java.io.*;
+import com.google.common.collect.ImmutableList;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
-
 import net.minecraft.network.INetHandler;
 import net.minecraft.network.NetHandlerPlayServer;
 import net.minecraft.network.NetworkManager;
@@ -46,9 +50,9 @@
 import net.minecraftforge.fml.common.network.FMLNetworkEvent;
 import net.minecraftforge.fml.relauncher.FMLLaunchHandler;
 import net.minecraftforge.fml.relauncher.Side;
-
-import com.google.common.collect.ImmutableList;
 import org.apache.commons.io.IOUtils;
+import red.mohist.api.ServerAPI;
+import red.mohist.util.i18n.Message;
 
 /**
  * Handles primary communication from hooked code into the system
@@ -172,8 +176,9 @@
         else
         {
             String text = query.getText() +
-                    "\n\nRun the command /fml confirm or or /fml cancel to proceed." +
-                    "\nAlternatively start the server with -Dfml.queryResult=confirm or -Dfml.queryResult=cancel to preselect the answer.";
+                    "\n\n" + Message.getString("forge.fmlserverhandler.1") +
+                    "\n" + Message.getString("forge.fmlserverhandler.2");
+
             FMLLog.log.warn(text);
 
             if (!query.isSynchronous()) return; // no-op until mc does commands in another thread (if ever)
@@ -195,14 +200,14 @@
 
                         if (cmd.equals("/fml confirm"))
                         {
-                            FMLLog.log.info("confirmed");
+                            FMLLog.log.info(Message.getString("forge.fmlserverhandler.3"));
                             query.setResult(true);
                             done = true;
                             it.remove();
                         }
                         else if (cmd.equals("/fml cancel"))
                         {
-                            FMLLog.log.info("cancelled");
+                            FMLLog.log.info(Message.getString("forge.fmlserverhandler.4"));
                             query.setResult(false);
                             done = true;
                             it.remove();
@@ -236,8 +241,10 @@
     @Override
     public void addModAsResource(ModContainer container)
     {
-        String langFile = "assets/" + container.getModId().toLowerCase() + "/lang/en_us.lang";
-        String langFile2 = "assets/" + container.getModId().toLowerCase() + "/lang/en_US.lang";
+        String langFile = "assets/" + container.getModId().toLowerCase() + "/lang/"+ ServerAPI.getLanguage().toLowerCase() +".lang";
+        String langFile2 = "assets/" + container.getModId().toLowerCase() + "/lang/"+ ServerAPI.getLanguage() +".lang";
+        String langFile3 = "assets/" + container.getModId().toLowerCase() + "/lang/en_us.lang";
+        String langFile4 = "assets/" + container.getModId().toLowerCase() + "/lang/en_US.lang";
         File source = container.getSource();
         InputStream stream = null;
         ZipFile zip = null;
@@ -249,6 +256,10 @@
                 if (!f.exists())
                     f = new File(source.toURI().resolve(langFile2).getPath());
                 if (!f.exists())
+                    f = new File(source.toURI().resolve(langFile3).getPath());
+                if (!f.exists())
+                    f = new File(source.toURI().resolve(langFile4).getPath());
+                if (!f.exists())
                     throw new FileNotFoundException(source.toURI().resolve(langFile).getPath());
                 stream = new FileInputStream(f);
             }
@@ -257,6 +268,8 @@
                 zip = new ZipFile(source);
                 ZipEntry entry = zip.getEntry(langFile);
                 if (entry == null) entry = zip.getEntry(langFile2);
+                if (entry == null) entry = zip.getEntry(langFile3);
+                if (entry == null) entry = zip.getEntry(langFile4);
                 if (entry == null) throw new FileNotFoundException(langFile);
                 stream = zip.getInputStream(entry);
             }
@@ -285,7 +298,7 @@
     @Override
     public String getCurrentLanguage()
     {
-        return "en_US";
+        return ServerAPI.getLanguage();
     }
 
     @Override
