--- a/net/minecraft/world/level/levelgen/feature/stateproviders/BlockStateProviderType.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/BlockStateProviderType.java
@@ -20,7 +_,7 @@
         return Registry.register(BuiltInRegistries.BLOCKSTATE_PROVIDER_TYPE, p_68763_, new BlockStateProviderType<>(p_334907_));
     }
 
-    private BlockStateProviderType(MapCodec<P> p_331592_) {
+    public BlockStateProviderType(MapCodec<P> p_331592_) {
         this.codec = p_331592_;
     }
 
