--- ../src-base/minecraft/net/minecraft/client/util/RecipeItemHelper.java
+++ ../src-work/minecraft/net/minecraft/client/util/RecipeItemHelper.java
@@ -24,10 +24,15 @@
 
     public void accountStack(ItemStack stack)
     {
+        this.accountStack(stack, -1);
+    }
+
+    public void accountStack(ItemStack stack, int forceCount)
+    {
         if (!stack.isEmpty() && !stack.isItemDamaged() && !stack.isItemEnchanted() && !stack.hasDisplayName())
         {
             int i = pack(stack);
-            int j = stack.getCount();
+            int j = forceCount == -1 ? stack.getCount() : forceCount;
             this.increment(i, j);
         }
     }
