--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -150,4 +_,11 @@
       }
 
    }
+
+   // Paper - start
+   public void registerAttribute(Attribute attributeBase) {
+      AttributeInstance attributeModifiable = new AttributeInstance(attributeBase, AttributeInstance::m_22099_);
+      f_22139_.put(attributeBase, attributeModifiable);
+   }
+   // Paper - end
 }
