--- a/net/minecraft/world/entity/ai/attributes/RangedAttribute.java
+++ b/net/minecraft/world/entity/ai/attributes/RangedAttribute.java
@@ -3,8 +_,8 @@
 import net.minecraft.util.Mth;
 
 public class RangedAttribute extends Attribute {
-    private final double minValue;
-    public final double maxValue;
+   	private final double minValue;
+   	public double maxValue;
 
     public RangedAttribute(String p_22310_, double p_22311_, double p_22312_, double p_22313_) {
         super(p_22310_, p_22311_);
