--- a/net/minecraft/network/play/client/CChatMessagePacket.java
+++ b/net/minecraft/network/play/client/CChatMessagePacket.java
@@ -1,6 +_,8 @@
 package net.minecraft.network.play.client;
 
 import java.io.IOException;
+
+import com.mohistmc.common.async.MohistThreadBox;
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.IServerPlayNetHandler;
@@ -20,14 +_,20 @@
    }
 
    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
-      this.field_149440_a = p_148837_1_.func_150789_c(256);
+      this.field_149440_a = org.apache.commons.lang3.StringUtils.normalizeSpace(p_148837_1_.func_150789_c(256)); // CraftBukkit - see PlayerConnection
    }
 
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
       p_148840_1_.func_180714_a(this.field_149440_a);
    }
 
+   // Spigot Start
    public void func_148833_a(IServerPlayNetHandler p_148833_1_) {
+      if (!field_149440_a.startsWith("/")) {
+         MohistThreadBox.ASYNCCHAT.submit(() -> p_148833_1_.func_147354_a(CChatMessagePacket.this));
+         return;
+      }
+      // Spigot End
       p_148833_1_.func_147354_a(this);
    }
 
