--- a/net/minecraft/network/play/client/CCloseWindowPacket.java
+++ b/net/minecraft/network/play/client/CCloseWindowPacket.java
@@ -13,7 +13,7 @@
    public CCloseWindowPacket() {
    }
 
-   @OnlyIn(Dist.CLIENT)
+   //@OnlyIn(Dist.CLIENT) CraftBukkit
    public CCloseWindowPacket(int p_i46881_1_) {
       this.field_149556_a = p_i46881_1_;
    }
@@ -26,6 +26,14 @@
       this.field_149556_a = p_148837_1_.readByte();
    }
 
+   public void constructor() {
+      throw new RuntimeException();
+   }
+
+   public void constructor(int id) {
+      constructor();
+      this.field_149556_a = id;
+   }
    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
       p_148840_1_.writeByte(this.field_149556_a);
    }
