--- a/net/minecraft/network/play/client/CPlayerPacket.java
+++ b/net/minecraft/network/play/client/CPlayerPacket.java
@@ -8,14 +8,14 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class CPlayerPacket implements IPacket<IServerPlayNetHandler> {
-   protected double field_149479_a;
-   protected double field_149477_b;
-   protected double field_149478_c;
-   protected float field_149476_e;
-   protected float field_149473_f;
+   public double field_149479_a; // CraftBukkit protected->public
+   public double field_149477_b; // CraftBukkit protected->public
+   public double field_149478_c; // CraftBukkit protected->public
+   public float field_149476_e; // CraftBukkit protected->public
+   public float field_149473_f; // CraftBukkit protected->public
    protected boolean field_149474_g;
-   protected boolean field_149480_h;
-   protected boolean field_149481_i;
+   public boolean field_149480_h; // CraftBukkit protected->public
+   public boolean field_149481_i; // CraftBukkit protected->public
 
    public CPlayerPacket() {
    }
