--- ../src-base/minecraft/net/minecraftforge/items/ItemStackHandler.java
+++ ../src-work/minecraft/net/minecraftforge/items/ItemStackHandler.java
@@ -31,6 +31,9 @@
 public class ItemStackHandler implements IItemHandler, IItemHandlerModifiable, INBTSerializable<NBTTagCompound>
 {
     protected NonNullList<ItemStack> stacks;
+	public NonNullList<ItemStack> getStacks() {
+        return this.stacks;
+    }
 
     public ItemStackHandler()
     {
