--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -8,6 +_,7 @@
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
 
 public class MerchantOffer {
     public static final Codec<MerchantOffer> CODEC = RecordCodecBuilder.create(
@@ -33,11 +_,24 @@
     public final ItemStack result;
     public int uses;
     public int maxUses;
-    public final boolean rewardExp;
+    public boolean rewardExp;
     public int specialPriceDiff;
     public int demand;
-    public final float priceMultiplier;
-    public final int xp;
+    public float priceMultiplier;
+    public int xp;
+
+    // CraftBukkit start
+    private CraftMerchantRecipe bukkitHandle; // CraftBukkit
+
+    public CraftMerchantRecipe asBukkit() {
+        return (this.bukkitHandle == null) ? (this.bukkitHandle = new CraftMerchantRecipe(this)) : this.bukkitHandle;
+    }
+
+    public MerchantOffer(ItemCost baseCostA, Optional<ItemCost> costB, ItemStack result, int uses, int maxUses, int experience, float priceMultiplier, int demand, CraftMerchantRecipe bukkit) {
+       this(baseCostA, costB, result, uses, maxUses, experience, priceMultiplier, demand);
+       this.bukkitHandle = bukkit;
+    }
+    // CraftBukkit end
 
     private MerchantOffer(
         ItemCost p_329205_,
