--- ../src-base/minecraft/org/bukkit/event/block/BlockExplodeEvent.java
+++ ../src-work/minecraft/org/bukkit/event/block/BlockExplodeEvent.java
@@ -1,18 +1,17 @@
 package org.bukkit.event.block;
 
+import java.util.List;
 import org.bukkit.block.Block;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
 
-import java.util.List;
-
 /**
  * Called when a block explodes
  */
 public class BlockExplodeEvent extends BlockEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
-    private boolean cancel;
     private final List<Block> blocks;
+    private boolean cancel;
     private float yield;
 
     public BlockExplodeEvent(final Block what, final List<Block> blocks, final float yield) {
@@ -22,6 +21,10 @@
         this.cancel = false;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public boolean isCancelled() {
         return cancel;
     }
@@ -62,8 +65,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
