--- ../src-base/minecraft/org/bukkit/event/block/BlockPhysicsEvent.java
+++ ../src-work/minecraft/org/bukkit/event/block/BlockPhysicsEvent.java
@@ -1,7 +1,7 @@
 package org.bukkit.event.block;
 
-import org.bukkit.block.Block;
 import org.bukkit.Material;
+import org.bukkit.block.Block;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
 
@@ -19,19 +19,21 @@
      * @param block the block involved in this event
      * @param changed the changed block's type id
      */
-    @Deprecated
     public BlockPhysicsEvent(final Block block, final int changed) {
         super(block);
         this.changed = changed;
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     /**
      * Gets the type of block that changed, causing this event
      *
      * @return Changed block's type id
      * @deprecated Magic value
      */
-    @Deprecated
     public int getChangedTypeId() {
         return changed;
     }
@@ -57,8 +59,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
