--- ../src-base/minecraft/org/bukkit/event/block/LeavesDecayEvent.java
+++ ../src-work/minecraft/org/bukkit/event/block/LeavesDecayEvent.java
@@ -17,6 +17,10 @@
         super(block);
     }
 
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
     public boolean isCancelled() {
         return cancel;
     }
@@ -29,8 +33,4 @@
     public HandlerList getHandlers() {
         return handlers;
     }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
