--- a/net/minecraft/world/gen/FlatGenerationSettings.java
+++ b/net/minecraft/world/gen/FlatGenerationSettings.java
@@ -86,7 +_,7 @@
       this.field_82655_a.addAll(p_i242012_3_);
       this.func_82645_d();
       if (!p_i242012_6_.isPresent()) {
-         field_211404_q.error("Unknown biome, defaulting to plains");
+         field_211404_q.error(com.mohistmc.util.i18n.i18n.get("flatgenerationsettings.1"));
          this.field_82654_c = () -> {
             return p_i242012_1_.func_243576_d(Biomes.field_76772_c);
          };
@@ -174,7 +_,7 @@
          }
       }
 
-      return (new Biome.Builder()).func_205415_a(biome.func_201851_b()).func_205419_a(biome.func_201856_r()).func_205421_a(biome.func_185355_j()).func_205420_b(biome.func_185360_m()).func_205414_c(biome.func_242445_k()).func_205417_d(biome.func_76727_i()).func_235097_a_(biome.func_235089_q_()).func_242457_a(biomegenerationsettings$builder.func_242508_a()).func_242458_a(biome.func_242433_b()).func_242455_a();
+      return (new Biome.Builder()).func_205415_a(biome.func_201851_b()).func_205419_a(biome.func_201856_r()).func_205421_a(biome.func_185355_j()).func_205420_b(biome.func_185360_m()).func_205414_c(biome.func_242445_k()).func_205417_d(biome.func_76727_i()).func_235097_a_(biome.func_235089_q_()).func_242457_a(biomegenerationsettings$builder.func_242508_a()).func_242458_a(biome.func_242433_b()).func_242455_a().setRegistryName(biome.getRegistryName());
    }
 
    public DimensionStructuresSettings func_236943_d_() {
