--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.saveddata.SavedData;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_20_R4.event.CraftEventFactory;
 
 public class Raids extends SavedData {
    private static final String RAID_FILE_ID = "raids";
@@ -119,19 +_,33 @@
             Raid raid = this.getOrCreateRaid(p_37964_.serverLevel(), blockpos1);
             boolean flag = false;
             if (!raid.isStarted()) {
+               /* CraftBukkit - moved down
                if (!this.raidMap.containsKey(raid.getId())) {
                   this.raidMap.put(raid.getId(), raid);
                }
+               */
 
                flag = true;
-            } else if (raid.getBadOmenLevel() < raid.getMaxBadOmenLevel()) {
+               // CraftBukkit start - fixed a bug with raid: players could add up Bad Omen level even when the raid had finished
+            } else if (raid.isInProgress() && raid.getBadOmenLevel() < raid.getMaxBadOmenLevel()) {
                flag = true;
+               // CraftBukkit end
             } else {
                p_37964_.removeEffect(MobEffects.BAD_OMEN);
                p_37964_.connection.send(new ClientboundEntityEventPacket(p_37964_, (byte)43));
             }
 
             if (flag) {
+               // CraftBukkit start
+               if (!CraftEventFactory.callRaidTriggerEvent(raid, p_37964_)) {
+                  p_37964_.removeEffect(MobEffects.BAD_OMEN);
+                  return null;
+               }
+
+               if (!this.raidMap.containsKey(raid.getId())) {
+                  this.raidMap.put(raid.getId(), raid);
+               }
+               // CraftBukkit end
                raid.absorbBadOmen(p_37964_);
                p_37964_.connection.send(new ClientboundEntityEventPacket(p_37964_, (byte)43));
                if (!raid.hasFirstWaveSpawned()) {
