--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -44,8 +_,8 @@
 
 public abstract class Raider extends PatrollingMonster {
    protected static final EntityDataAccessor<Boolean> IS_CELEBRATING = SynchedEntityData.defineId(Raider.class, EntityDataSerializers.BOOLEAN);
-   static final Predicate<ItemEntity> ALLOWED_ITEMS = (p_309223_) -> {
-      return !p_309223_.hasPickUpDelay() && p_309223_.isAlive() && ItemStack.matches(p_309223_.getItem(), Raid.getLeaderBannerInstance());
+   static final Predicate<ItemEntity> ALLOWED_ITEMS = (p_296862_) -> {
+      return !p_296862_.hasPickUpDelay() && p_296862_.isAlive() && ItemStack.matches(p_296862_.getItem(), Raid.getLeaderBannerInstance());
    };
    @Nullable
    protected Raid raid;
@@ -149,6 +_,7 @@
                i = Mth.clamp(i, 0, 4);
                MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.BAD_OMEN, 120000, i, false, false, true);
                if (!this.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS)) {
+                  player.addEffectCause.set(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                   player.addEffect(mobeffectinstance);
                }
             }
@@ -296,7 +_,7 @@
          this.mob.getNavigation().stop();
 
          for(Raider raider : this.mob.level().getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0D, 8.0D, 8.0D))) {
-            raider.setTarget(this.mob.getTarget());
+            raider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
          }
 
       }
@@ -306,7 +_,7 @@
          LivingEntity livingentity = this.mob.getTarget();
          if (livingentity != null) {
             for(Raider raider : this.mob.level().getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0D, 8.0D, 8.0D))) {
-               raider.setTarget(livingentity);
+               raider.setTarget(livingentity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                raider.setAggressive(true);
             }
 
