--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -26,7 +_,7 @@
 
 public class Raids extends SavedData {
    private static final String f_150234_ = "raids";
-   private final Map<Integer, Raid> f_37951_ = Maps.newHashMap();
+   public final Map<Integer, Raid> f_37951_ = Maps.newHashMap();
    private final ServerLevel f_37952_;
    private int f_37953_;
    private int f_37954_;
