--- ../src-base/minecraft/net/minecraft/block/Block.java
+++ ../src-work/minecraft/net/minecraft/block/Block.java
@@ -27,6 +27,7 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemBlock;
+import net.minecraft.item.ItemShears;
 import net.minecraft.item.ItemStack;
 import net.minecraft.stats.StatList;
 import net.minecraft.tileentity.TileEntity;
@@ -46,6 +47,7 @@
 import net.minecraftforge.common.EnumPlantType;
 import net.minecraftforge.common.ForgeHooks;
 import net.minecraftforge.common.IPlantable;
+import net.minecraftforge.common.IShearable;
 import net.minecraftforge.common.MinecraftForge;
 import net.minecraftforge.common.util.ForgeDirection;
 import net.minecraftforge.common.util.RotationHelper;
@@ -122,6 +124,7 @@
     private String unlocalizedName;
     @SideOnly(Side.CLIENT)
     protected IIcon blockIcon;
+    public boolean isForgeBlock; // Cauldron
     private static final String __OBFID = "CL_00000199";
 
     public final cpw.mods.fml.common.registry.RegistryDelegate<Block> delegate = 
@@ -954,11 +957,15 @@
         if (this.canSilkHarvest(p_149636_1_, p_149636_2_, p_149636_3_, p_149636_4_, p_149636_5_, p_149636_6_) && EnchantmentHelper.getSilkTouchModifier(p_149636_2_))
         {
             ArrayList<ItemStack> items = new ArrayList<ItemStack>();
-            ItemStack itemstack = this.createStackedBlock(p_149636_6_);
-
-            if (itemstack != null)
+            
+            if (!((p_149636_2_.getHeldItem().getItem() instanceof ItemShears) && (this instanceof BlockLeavesBase) && (this instanceof IShearable)))
             {
-                items.add(itemstack);
+                ItemStack itemstack = this.createStackedBlock(p_149636_6_);
+
+                if (itemstack != null)
+                {
+                    items.add(itemstack);
+                }            	
             }
 
             ForgeEventFactory.fireBlockHarvesting(items, p_149636_1_, this, p_149636_3_, p_149636_4_, p_149636_5_, p_149636_6_, 0, 1.0f, true, p_149636_2_);
@@ -1131,6 +1138,23 @@
         return this;
     }
 
+    // Spigot start
+    public static float range(float min, float value, float max)
+    {
+        if (value < min)
+        {
+            return min;
+        }
+
+        if (value > max)
+        {
+            return max;
+        }
+
+        return value;
+    }
+    // Spigot end
+
     @SideOnly(Side.CLIENT)
     protected String getTextureName()
     {
