--- ../src-base/minecraft/net/minecraft/world/storage/loot/LootContext.java
+++ ../src-work/minecraft/net/minecraft/world/storage/loot/LootContext.java
@@ -89,6 +89,16 @@
         }
     }
 
+    public WorldServer getWorld()
+    {
+        return world;
+    }
+
+    public int getLootingModifier()
+    {
+        return net.minecraftforge.common.ForgeHooks.getLootingLevel(getLootedEntity(), getKiller(), damageSource);
+    }
+
     public static class Builder
         {
             private final WorldServer world;
