--- ../src-base/minecraft/net/minecraft/command/server/CommandTeleport.java
+++ ../src-work/minecraft/net/minecraft/command/server/CommandTeleport.java
@@ -91,7 +91,7 @@
             }
 
             p_189862_0_.func_184210_p();
-            ((EntityPlayerMP)p_189862_0_).field_71135_a.func_175089_a(p_189862_1_.func_179628_a(), p_189862_2_.func_179628_a(), p_189862_3_.func_179628_a(), f, f1, set);
+            ((EntityPlayerMP)p_189862_0_).field_71135_a.setPlayerLocation(p_189862_1_.func_179628_a(), p_189862_2_.func_179628_a(), p_189862_3_.func_179628_a(), f, f1, set, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND);
             p_189862_0_.func_70034_d(f);
         }
         else
