--- a/net/minecraft/server/Eula.java
+++ b/net/minecraft/server/Eula.java
@@ -10,47 +_,41 @@
 import org.slf4j.Logger;
 
 public class Eula {
-    private static final Logger LOGGER = LogUtils.getLogger();
-    private final Path file;
-    private final boolean agreed;
-
-    public Eula(Path p_135943_) {
-        this.file = p_135943_;
-        this.agreed = SharedConstants.IS_RUNNING_IN_IDE || this.readFile();
-    }
-
-    private boolean readFile() {
-        try {
-            boolean flag;
-            try (InputStream inputstream = Files.newInputStream(this.file)) {
-                Properties properties = new Properties();
-                properties.load(inputstream);
-                flag = Boolean.parseBoolean(properties.getProperty("eula", "false"));
-            }
-
-            return flag;
-        } catch (Exception exception) {
-            LOGGER.warn("Failed to load {}", this.file);
-            this.saveDefaults();
-            return false;
-        }
-    }
-
-    public boolean hasAgreedToEULA() {
-        return this.agreed;
-    }
-
-    private void saveDefaults() {
-        if (!SharedConstants.IS_RUNNING_IN_IDE) {
-            try (OutputStream outputstream = Files.newOutputStream(this.file)) {
-                Properties properties = new Properties();
-                properties.setProperty("eula", "false");
-                properties.store(
-                    outputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://aka.ms/MinecraftEULA)."
-                );
-            } catch (Exception exception) {
-                LOGGER.warn("Failed to save {}", this.file, exception);
-            }
-        }
-    }
+   private static final Logger LOGGER = LogUtils.getLogger();
+   private final Path file;
+   private final boolean agreed;
+
+   public Eula(Path p_135943_) {
+      this.file = p_135943_;
+      this.agreed = SharedConstants.IS_RUNNING_IN_IDE || net.minecraftforge.gametest.ForgeGameTestHooks.isGametestServer() || this.readFile(); // Forge: Automatically agree to EULA for gametest servers to aid CI
+   }
+
+   private boolean readFile() {
+      try (InputStream inputstream = Files.newInputStream(this.file)) {
+         Properties properties = new Properties();
+         properties.load(inputstream);
+         return Boolean.parseBoolean(properties.getProperty("eula", "false"));
+      } catch (Exception exception) {
+         LOGGER.warn("Failed to load {}", (Object)this.file);
+         this.saveDefaults();
+         return false;
+      }
+   }
+
+   public boolean hasAgreedToEULA() {
+      return this.agreed;
+   }
+
+   private void saveDefaults() {
+      if (!SharedConstants.IS_RUNNING_IN_IDE) {
+         try (OutputStream outputstream = Files.newOutputStream(this.file)) {
+            Properties properties = new Properties();
+            properties.setProperty("eula", "false");
+            properties.store(outputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://aka.ms/MinecraftEULA).");
+         } catch (Exception exception) {
+            LOGGER.warn("Failed to save {}", this.file, exception);
+         }
+
+      }
+   }
 }
