--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -3,7 +_,6 @@
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonElement;
-import com.google.gson.JsonIOException;
 import com.google.gson.JsonParseException;
 import com.google.gson.internal.Streams;
 import com.google.gson.stream.JsonReader;
@@ -25,16 +_,17 @@
 import java.util.function.BiConsumer;
 import javax.annotation.Nullable;
 import net.minecraft.FileUtil;
+import net.minecraft.Util;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.AdvancementHolder;
 import net.minecraft.advancements.AdvancementNode;
 import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.advancements.AdvancementTree;
+import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.advancements.Criterion;
 import net.minecraft.advancements.CriterionProgress;
 import net.minecraft.advancements.CriterionTrigger;
 import net.minecraft.advancements.CriterionTriggerInstance;
-import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.network.protocol.game.ClientboundSelectAdvancementsTabPacket;
 import net.minecraft.network.protocol.game.ClientboundUpdateAdvancementsPacket;
@@ -129,6 +_,7 @@
     }
 
     public void save() {
+		if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return; // Spigot
         JsonElement jsonelement = this.codec.encodeStart(JsonOps.INSTANCE, this.asData()).getOrThrow();
 
         try {
@@ -146,7 +_,8 @@
         p_300341_.forEach((p_296440_, p_296441_) -> {
             AdvancementHolder advancementholder = p_299201_.get(p_296440_);
             if (advancementholder == null) {
-                LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", p_296440_, this.playerSavePath);
+            if (!p_296440_.getNamespace().equals("minecraft")) return; // CraftBukkit
+            LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", p_296440_, this.playerSavePath);
             } else {
                 this.startProgress(advancementholder, p_296441_);
                 this.progressChanged.add(advancementholder);
@@ -173,6 +_,7 @@
             this.unregisterListeners(p_298135_);
             this.progressChanged.add(p_298135_);
             flag = true;
+            net.minecraftforge.event.ForgeEventFactory.onAdvancementGrant(this.player, p_298135_, advancementprogress, p_135990_);
             if (!flag1 && advancementprogress.isDone()) {
                 p_298135_.value().rewards().grant(this.player);
                 p_298135_.value().display().ifPresent(p_326189_ -> {
@@ -180,6 +_,7 @@
                         this.playerList.broadcastSystemMessage(p_326189_.getType().createAnnouncement(p_298135_, this.player), false);
                     }
                 });
+                net.minecraftforge.event.ForgeEventFactory.onAdvancementEarned(this.player, p_298135_);
             }
         }
 
@@ -198,6 +_,7 @@
             this.registerListeners(p_297905_);
             this.progressChanged.add(p_297905_);
             flag = true;
+            net.minecraftforge.event.ForgeEventFactory.onAdvancementRevoke(this.player, p_297905_, advancementprogress, p_136000_);
         }
 
         if (flag1 && !advancementprogress.isDone()) {
