--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -1,8 +_,8 @@
 package net.minecraft.server;
 
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Multimap;
 import com.google.common.collect.ImmutableMap.Builder;
+import com.google.common.collect.Multimap;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonElement;
@@ -11,7 +_,6 @@
 import com.mojang.serialization.JsonOps;
 import java.util.Collection;
 import java.util.Map;
-import java.util.Map.Entry;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 import net.minecraft.advancements.Advancement;
@@ -29,23 +_,43 @@
 import org.slf4j.Logger;
 
 public class ServerAdvancementManager extends SimpleJsonResourceReloadListener {
-    private static final Logger LOGGER = LogUtils.getLogger();
-    public static final Gson GSON = new GsonBuilder().create();
-    public Map<ResourceLocation, AdvancementHolder> advancements = Map.of();
+   private static final Logger LOGGER = LogUtils.getLogger();
+    private static final Gson GSON = new GsonBuilder().create();
+    private Map<ResourceLocation, AdvancementHolder> advancements = Map.of();
     private AdvancementTree tree = new AdvancementTree();
     private final HolderLookup.Provider registries;
+    private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context; //Forge: add context
 
+    /** @deprecated Forge: use {@linkplain ServerAdvancementManager#ServerAdvancementManager(LootDataManager, net.minecraftforge.common.crafting.conditions.ICondition.IContext) constructor with context}. */
+    @Deprecated
     public ServerAdvancementManager(HolderLookup.Provider p_336198_) {
+        this(p_336198_, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+    }
+
+    public ServerAdvancementManager(HolderLookup.Provider p_336198_, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
         super(GSON, "advancements");
         this.registries = p_336198_;
+        this.context = context;
     }
 
     protected void apply(Map<ResourceLocation, JsonElement> p_136034_, ResourceManager p_136035_, ProfilerFiller p_136036_) {
         RegistryOps<JsonElement> registryops = this.registries.createSerializationContext(JsonOps.INSTANCE);
         Builder<ResourceLocation, AdvancementHolder> builder = ImmutableMap.builder();
         p_136034_.forEach((p_326203_, p_326204_) -> {
+            // Spigot start
+            if (org.spigotmc.SpigotConfig.disabledAdvancements != null && (org.spigotmc.SpigotConfig.disabledAdvancements.contains("*")
+                    || org.spigotmc.SpigotConfig.disabledAdvancements.contains(p_326203_.toString())
+                    || org.spigotmc.SpigotConfig.disabledAdvancements.contains(p_326203_.getNamespace()))) {
+                return;
+            }
+            // Spigot end
             try {
-                Advancement advancement = Advancement.CODEC.parse(registryops, p_326204_).getOrThrow(JsonParseException::new);
+                var json = net.minecraftforge.common.ForgeHooks.readConditionalAdvancement(registryops, (com.google.gson.JsonObject)p_326204_);
+                if (json == null) {
+                    LOGGER.debug("Skipping loading advancement {} as its conditions were not met", p_326204_);
+                    return;
+                }
+                Advancement advancement = Advancement.CODEC.parse(registryops, json).getOrThrow(JsonParseException::new);
                 this.validate(p_326203_, advancement);
                 builder.put(p_326203_, new AdvancementHolder(p_326203_, advancement));
             } catch (Exception exception) {
