--- a/net/minecraft/server/WorldLoader.java
+++ b/net/minecraft/server/WorldLoader.java
@@ -54,7 +_,7 @@
       return RegistryDataLoader.load(p_251529_, registryaccess$frozen, p_249516_);
    }
 
-   private static LayeredRegistryAccess<RegistryLayer> loadAndReplaceLayer(ResourceManager p_249913_, LayeredRegistryAccess<RegistryLayer> p_252077_, RegistryLayer p_250346_, List<RegistryDataLoader.RegistryData<?>> p_250589_) {
+   public static LayeredRegistryAccess<RegistryLayer> loadAndReplaceLayer(ResourceManager p_249913_, LayeredRegistryAccess<RegistryLayer> p_252077_, RegistryLayer p_250346_, List<RegistryDataLoader.RegistryData<?>> p_250589_) {
       RegistryAccess.Frozen registryaccess$frozen = loadLayer(p_249913_, p_252077_, p_250346_, p_250589_);
       return p_252077_.replaceFrom(p_250346_, registryaccess$frozen);
    }
