--- a/net/minecraft/entity/ai/brain/BrainUtil.java
+++ b/net/minecraft/entity/ai/brain/BrainUtil.java
@@ -58,6 +58,7 @@
    }
 
    public static void func_220624_a(LivingEntity p_220624_0_, ItemStack p_220624_1_, LivingEntity p_220624_2_) {
+      if (p_220624_1_.func_190926_b()) return; // CraftBukkit - SPIGOT-4940: no empty loot
       double d0 = p_220624_0_.func_226280_cw_() - (double)0.3F;
       ItemEntity itementity = new ItemEntity(p_220624_0_.field_70170_p, p_220624_0_.func_226277_ct_(), d0, p_220624_0_.func_226281_cx_(), p_220624_1_);
       BlockPos blockpos = new BlockPos(p_220624_2_);
