--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerState.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerState.java
@@ -64,10 +_,7 @@
         return switch (this) {
             case INACTIVE -> trialspawnerdata.getOrCreateDisplayEntity(p_310869_, p_313233_, WAITING_FOR_PLAYERS) == null ? this : WAITING_FOR_PLAYERS;
             case WAITING_FOR_PLAYERS -> {
-                if (!p_310869_.canSpawnInLevel(p_313233_)) {
-                    trialspawnerdata.reset();
-                    yield this;
-                } else if (!trialspawnerdata.hasMobToSpawn(p_310869_, p_313233_.random)) {
+                if (!trialspawnerdata.hasMobToSpawn(p_310869_, p_313233_.random)) {
                     yield INACTIVE;
                 } else {
                     trialspawnerdata.tryDetectPlayers(p_313233_, p_313024_, p_310869_);
@@ -75,10 +_,7 @@
                 }
             }
             case ACTIVE -> {
-                if (!p_310869_.canSpawnInLevel(p_313233_)) {
-                    trialspawnerdata.reset();
-                    yield WAITING_FOR_PLAYERS;
-                } else if (!trialspawnerdata.hasMobToSpawn(p_310869_, p_313233_.random)) {
+                if (!trialspawnerdata.hasMobToSpawn(p_310869_, p_313233_.random)) {
                     yield INACTIVE;
                 } else {
                     int i = trialspawnerdata.countAdditionalPlayers(p_313024_);
@@ -141,8 +_,8 @@
                     trialspawnerdata.nextMobSpawnsAt = 0L;
                     yield ACTIVE;
                 } else if (trialspawnerdata.isCooldownFinished(p_313233_)) {
+                    trialspawnerdata.cooldownEndsAt = 0L;
                     p_310869_.removeOminous(p_313233_, p_313024_);
-                    trialspawnerdata.reset();
                     yield WAITING_FOR_PLAYERS;
                 } else {
                     yield this;
@@ -175,10 +_,10 @@
             .map(p_332378_::getPlayerByUUID)
             .filter(Objects::nonNull)
             .filter(
-                p_341872_ -> !p_341872_.isCreative()
-                        && !p_341872_.isSpectator()
-                        && p_341872_.isAlive()
-                        && p_341872_.distanceToSqr(p_330701_.getCenter()) <= (double)Mth.square(p_331338_.getRequiredPlayerRange())
+                p_327387_ -> !p_327387_.isCreative()
+                        && !p_327387_.isSpectator()
+                        && p_327387_.isAlive()
+                        && p_327387_.distanceToSqr(p_330701_.getCenter()) <= (double)Mth.square(p_331338_.getRequiredPlayerRange())
             )
             .toList();
         if (list.isEmpty()) {
