--- ../src-base/minecraft/org/bukkit/inventory/BeaconInventory.java
+++ ../src-work/minecraft/org/bukkit/inventory/BeaconInventory.java
@@ -6,16 +6,16 @@
 public interface BeaconInventory extends Inventory {
 
     /**
-     * Set the item powering the beacon.
+     * Get the item powering the beacon.
      *
-     * @param item The new item
+     * @return The current item.
      */
-    void setItem(ItemStack item);
+    ItemStack getItem();
 
     /**
-     * Get the item powering the beacon.
+     * Set the item powering the beacon.
      *
-     * @return The current item.
+     * @param item The new item
      */
-    ItemStack getItem();
+    void setItem(ItemStack item);
 }
