--- ../src-base/minecraft/org/bukkit/inventory/CraftingInventory.java
+++ ../src-work/minecraft/org/bukkit/inventory/CraftingInventory.java
@@ -13,18 +13,18 @@
     ItemStack getResult();
 
     /**
-     * Get the contents of the crafting matrix.
+     * Set the item in the result slot of the crafting inventory.
      *
-     * @return The contents.
+     * @param newResult The new result item.
      */
-    ItemStack[] getMatrix();
+    void setResult(ItemStack newResult);
 
     /**
-     * Set the item in the result slot of the crafting inventory.
+     * Get the contents of the crafting matrix.
      *
-     * @param newResult The new result item.
+     * @return The contents.
      */
-    void setResult(ItemStack newResult);
+    ItemStack[] getMatrix();
 
     /**
      * Replace the contents of the crafting matrix
