--- ../src-base/minecraft/org/bukkit/inventory/EnchantingInventory.java
+++ ../src-work/minecraft/org/bukkit/inventory/EnchantingInventory.java
@@ -6,13 +6,6 @@
 public interface EnchantingInventory extends Inventory {
 
     /**
-     * Set the item being enchanted.
-     *
-     * @param item The new item
-     */
-    void setItem(ItemStack item);
-
-    /**
      * Get the item being enchanted.
      *
      * @return The current item.
@@ -20,11 +13,11 @@
     ItemStack getItem();
 
     /**
-     * Set the secondary item being used for the enchant.
+     * Set the item being enchanted.
      *
      * @param item The new item
      */
-    void setSecondary(ItemStack item);
+    void setItem(ItemStack item);
 
     /**
      * Get the secondary item being used for the enchant.
@@ -32,4 +25,11 @@
      * @return The second item
      */
     ItemStack getSecondary();
+
+    /**
+     * Set the secondary item being used for the enchant.
+     *
+     * @param item The new item
+     */
+    void setSecondary(ItemStack item);
 }
