--- ../src-base/minecraft/org/bukkit/inventory/ShapedRecipe.java
+++ ../src-work/minecraft/org/bukkit/inventory/ShapedRecipe.java
@@ -3,9 +3,7 @@
 import com.google.common.base.Preconditions;
 import java.util.HashMap;
 import java.util.Map;
-
 import org.apache.commons.lang.Validate;
-
 import org.bukkit.Keyed;
 import org.bukkit.Material;
 import org.bukkit.NamespacedKey;
@@ -20,7 +18,6 @@
     private String[] rows;
     private Map<Character, ItemStack> ingredients = new HashMap<Character, ItemStack>();
 
-    @Deprecated
     public ShapedRecipe(ItemStack result) {
         this.key = NamespacedKey.randomKey();
         this.output = new ItemStack(result);
@@ -116,7 +113,6 @@
      * @return The changed recipe, so you can chain calls.
      * @deprecated Magic value
      */
-    @Deprecated
     public ShapedRecipe setIngredient(char key, Material ingredient, int raw) {
         Validate.isTrue(ingredients.containsKey(key), "Symbol does not appear in the shape:", key);
 
