--- ../src-base/minecraft/org/bukkit/inventory/ShapelessRecipe.java
+++ ../src-work/minecraft/org/bukkit/inventory/ShapelessRecipe.java
@@ -3,10 +3,8 @@
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
-
 import org.apache.commons.lang.Validate;
 import org.bukkit.Keyed;
-
 import org.bukkit.Material;
 import org.bukkit.NamespacedKey;
 import org.bukkit.material.MaterialData;
@@ -20,7 +18,6 @@
     private final ItemStack output;
     private final List<ItemStack> ingredients = new ArrayList<ItemStack>();
 
-    @Deprecated
     public ShapelessRecipe(ItemStack result) {
         this.key = NamespacedKey.randomKey();
         this.output = new ItemStack(result);
@@ -35,10 +32,10 @@
      * @param result The item you want the recipe to create.
      * @see ShapelessRecipe#addIngredient(Material)
      * @see ShapelessRecipe#addIngredient(MaterialData)
-     * @see ShapelessRecipe#addIngredient(Material,int)
-     * @see ShapelessRecipe#addIngredient(int,Material)
-     * @see ShapelessRecipe#addIngredient(int,MaterialData)
-     * @see ShapelessRecipe#addIngredient(int,Material,int)
+     * @see ShapelessRecipe#addIngredient(Material, int)
+     * @see ShapelessRecipe#addIngredient(int, Material)
+     * @see ShapelessRecipe#addIngredient(int, MaterialData)
+     * @see ShapelessRecipe#addIngredient(int, Material, int)
      */
     public ShapelessRecipe(NamespacedKey key, ItemStack result) {
         this.key = key;
@@ -73,7 +70,6 @@
      * @return The changed recipe, so you can chain calls.
      * @deprecated Magic value
      */
-    @Deprecated
     public ShapelessRecipe addIngredient(Material ingredient, int rawdata) {
         return addIngredient(1, ingredient, rawdata);
     }
@@ -109,7 +105,6 @@
      * @return The changed recipe, so you can chain calls.
      * @deprecated Magic value
      */
-    @Deprecated
     public ShapelessRecipe addIngredient(int count, Material ingredient, int rawdata) {
         Validate.isTrue(ingredients.size() + count <= 9, "Shapeless recipes cannot have more than 9 ingredients");
 
@@ -184,7 +179,6 @@
      * @return The changed recipe.
      * @deprecated Magic value
      */
-    @Deprecated
     public ShapelessRecipe removeIngredient(Material ingredient, int rawdata) {
         return removeIngredient(1, ingredient, rawdata);
     }
@@ -200,7 +194,7 @@
      * @return The changed recipe.
      * @deprecated Magic value
      */
-    @Deprecated
+
     public ShapelessRecipe removeIngredient(int count, Material ingredient, int rawdata) {
         Iterator<ItemStack> iterator = ingredients.iterator();
         while (count > 0 && iterator.hasNext()) {
