--- a/net/minecraft/server/packs/VanillaPackResourcesBuilder.java
+++ b/net/minecraft/server/packs/VanillaPackResourcesBuilder.java
@@ -44,7 +_,7 @@
                   URI uri = url.toURI();
                   String s1 = uri.getScheme();
                   if (!"jar".equals(s1) && !"file".equals(s1)) {
-                     LOGGER.warn("Assets URL '{}' uses unexpected schema", (Object)uri);
+                     LOGGER.debug("Assets URL '{}' uses unexpected schema", (Object)uri);
                   }
 
                   Path path = safeGetPath(uri);
