From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 20 Mar 2016 06:45:01 -0400
Subject: [PATCH] Access items by EquipmentSlot


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
@@ -11,6 +11,7 @@ import org.apache.commons.lang.Validate;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.entity.HumanEntity;
 import org.bukkit.inventory.EntityEquipment;
+import org.bukkit.inventory.EquipmentSlot;
 import org.bukkit.inventory.ItemStack;
 
 public class CraftInventoryPlayer extends CraftInventory implements org.bukkit.inventory.PlayerInventory, EntityEquipment {
@@ -62,6 +63,32 @@ public class CraftInventoryPlayer extends CraftInventory implements org.bukkit.i
     }
 
     @Override
+    public ItemStack getItem(EquipmentSlot slot) {
+        switch(slot) {
+            case HAND: return getItemInMainHand();
+            case OFF_HAND: return getItemInOffHand();
+            case HEAD: return getHelmet();
+            case CHEST: return getChestplate();
+            case LEGS: return getLeggings();
+            case FEET: return getBoots();
+        }
+        throw new UnsupportedOperationException(slot.name());
+    }
+
+    @Override
+    public void setItem(EquipmentSlot slot, ItemStack stack) {
+        switch(slot) {
+            case HAND: setItemInMainHand(stack); return;
+            case OFF_HAND: setItemInOffHand(stack); return;
+            case HEAD: setHelmet(stack); return;
+            case CHEST: setChestplate(stack); return;
+            case LEGS: setLeggings(stack); return;
+            case FEET: setBoots(stack); return;
+        }
+        throw new UnsupportedOperationException(slot.name());
+    }
+
+    @Override
     public void setItem(int index, ItemStack item) {
         super.setItem(index, item);
         if (this.getHolder() == null) return;
