From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 14 Jun 2015 04:13:52 -0400
Subject: [PATCH] Add ExplosionPrimeByEntityEvent


diff --git a/src/main/java/net/minecraft/server/BlockTNT.java b/src/main/java/net/minecraft/server/BlockTNT.java
--- a/src/main/java/net/minecraft/server/BlockTNT.java
+++ b/src/main/java/net/minecraft/server/BlockTNT.java
@@ -2,6 +2,7 @@ package net.minecraft.server;
 
 public class BlockTNT extends Block {
     private boolean isPrimed = false; // SportBukkit
+    Entity primer; // SportBukkit
 
     public static final BlockStateBoolean EXPLODE = BlockStateBoolean.of("explode");
 
@@ -41,7 +42,12 @@ public class BlockTNT extends Block {
             EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) blockposition.getX() + 0.5F), (double) blockposition.getY(), (double) ((float) blockposition.getZ() + 0.5F), explosion.getSource());
 
             // SportBukkit start - allow exploded TNT to cancel priming
-            org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+            final org.bukkit.event.entity.ExplosionPrimeEvent event;
+            if(explosion.source != null) {
+                event = new org.bukkit.event.entity.ExplosionPrimeByEntityEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity(), explosion.source.getBukkitEntity());
+            } else {
+                event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+            }
             world.getServer().getPluginManager().callEvent(event);
 
             if (!event.isCancelled()) {
@@ -65,7 +71,12 @@ public class BlockTNT extends Block {
                 EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double) ((float) blockposition.getX() + 0.5F), (double) blockposition.getY(), (double) ((float) blockposition.getZ() + 0.5F), entityliving);
 
                 // SportBukkit start - fire ExplosionPrimeEvent
-                org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+                org.bukkit.event.entity.ExplosionPrimeEvent event;
+                if(primer != null) {
+                    event = new org.bukkit.event.entity.ExplosionPrimeByEntityEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity(), primer.getBukkitEntity());
+                } else {
+                    event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) entitytntprimed.getBukkitEntity());
+                }
                 world.getServer().getPluginManager().callEvent(event);
                 if(event.isCancelled()) return;
                 isPrimed = true;
@@ -82,7 +93,9 @@ public class BlockTNT extends Block {
         ItemStack itemstack = entityhuman.b(enumhand);
 
         if (!itemstack.isEmpty() && (itemstack.getItem() == Items.FLINT_AND_STEEL || itemstack.getItem() == Items.FIRE_CHARGE)) {
+            try { primer = entityhuman; // SportBukkit
             this.a(world, blockposition, iblockdata.set(BlockTNT.EXPLODE, Boolean.valueOf(true)), (EntityLiving) entityhuman);
+            } finally { primer = null; } // SportBukkit
             // SportBukkit start - don't remove block if TNT priming is cancelled
             if (isPrimed) {
                 world.setTypeAndData(blockposition, Blocks.AIR.getBlockData(), 11);
@@ -110,7 +123,9 @@ public class BlockTNT extends Block {
                     return;
                 }
                 // CraftBukkit end
+                try { primer = entityarrow; // SportBukkit
                 this.a(world, blockposition, world.getType(blockposition).set(BlockTNT.EXPLODE, Boolean.valueOf(true)), entityarrow.shooter instanceof EntityLiving ? (EntityLiving) entityarrow.shooter : null);
+                } finally { primer = null; } // SportBukkit
                 // SportBukkit start - don't remove block if TNT priming is cancelled
                 if (isPrimed) {
                     world.setAir(blockposition);
diff --git a/src/main/java/net/minecraft/server/EntityCreeper.java b/src/main/java/net/minecraft/server/EntityCreeper.java
--- a/src/main/java/net/minecraft/server/EntityCreeper.java
+++ b/src/main/java/net/minecraft/server/EntityCreeper.java
@@ -6,6 +6,7 @@ import javax.annotation.Nullable;
 // CraftBukkit start
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
+import org.bukkit.event.entity.ExplosionPrimeByEntityEvent;
 // CraftBukkit end
 
 public class EntityCreeper extends EntityMonster {
@@ -18,6 +19,7 @@ public class EntityCreeper extends EntityMonster {
     private int maxFuseTicks = 30;
     private int explosionRadius = 3;
     private int bA;
+    private Entity primer; // SportBukkit
 
     public EntityCreeper(World world) {
         super(world);
@@ -104,7 +106,13 @@ public class EntityCreeper extends EntityMonster {
 
             if (i > 0 && this.fuseTicks == 0) {
                 // SportBukkit start - fire ExplosionPrimeEvent
-                ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * (this.isPowered() ? 2.0F : 1.0F), false);
+                final float radius = this.explosionRadius * (this.isPowered() ? 2.0F : 1.0F);
+                final ExplosionPrimeEvent event;
+                if(primer != null) {
+                    event = new ExplosionPrimeByEntityEvent(this.getBukkitEntity(), radius, false, primer.getBukkitEntity());
+                } else {
+                    event = new ExplosionPrimeEvent(this.getBukkitEntity(), radius, false);
+                }
                 this.world.getServer().getPluginManager().callEvent(event);
 
                 if (event.isCancelled()) {
@@ -198,6 +206,7 @@ public class EntityCreeper extends EntityMonster {
             this.world.a(entityhuman, this.locX, this.locY, this.locZ, SoundEffects.bI, this.bC(), 1.0F, this.random.nextFloat() * 0.4F + 0.8F);
             entityhuman.a(enumhand);
             if (!this.world.isClientSide) {
+                this.primer = entityhuman; // SportBukkit
                 this.dk();
                 itemstack.damage(1, entityhuman);
                 return true;
@@ -213,7 +222,12 @@ public class EntityCreeper extends EntityMonster {
             float f = this.isPowered() ? 2.0F : 1.0F;
 
             // CraftBukkit start
-            ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+            final ExplosionPrimeEvent event;
+            if(primer != null) {
+                event = new ExplosionPrimeByEntityEvent(this.getBukkitEntity(), this.explosionRadius * f, false, primer.getBukkitEntity());
+            } else {
+                event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+            }
             this.world.getServer().getPluginManager().callEvent(event);
             if (!event.isCancelled()) {
                 this.aU = true;
diff --git a/src/main/java/net/minecraft/server/EntityEnderCrystal.java b/src/main/java/net/minecraft/server/EntityEnderCrystal.java
--- a/src/main/java/net/minecraft/server/EntityEnderCrystal.java
+++ b/src/main/java/net/minecraft/server/EntityEnderCrystal.java
@@ -6,6 +6,7 @@ import javax.annotation.Nullable;
 // CraftBukkit start
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.ExplosionPrimeEvent;
+import org.bukkit.event.entity.ExplosionPrimeByEntityEvent;
 // CraftBukkit end
 
 public class EntityEnderCrystal extends Entity {
@@ -92,7 +93,12 @@ public class EntityEnderCrystal extends Entity {
                 this.die();
                 if (!this.world.isClientSide) {
                     // CraftBukkit start
-                    ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, true);
+                    final ExplosionPrimeEvent event;
+                    if(damagesource.getEntity() != null) {
+                        event = new ExplosionPrimeByEntityEvent(this.getBukkitEntity(), 6.0F, false, damagesource.getEntity().getBukkitEntity());
+                    } else {
+                        event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+                    }
                     this.world.getServer().getPluginManager().callEvent(event);
                     if (event.isCancelled()) {
                         this.dead = false;
