From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Jan 2013 14:12:03 -0500
Subject: [PATCH] Add PlayerOnGroundEvent


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -98,6 +98,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     private int processedMovePackets;
     private boolean processedDisconnect; // CraftBukkit - Added
 
+    boolean clientOnGround; // SportBukkit - only set by client
+
     public PlayerConnection(MinecraftServer minecraftserver, NetworkManager networkmanager, EntityPlayer entityplayer) {
         this.minecraftServer = minecraftserver;
         this.networkManager = networkmanager;
@@ -631,6 +633,13 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                             this.B &= !this.player.hasEffect(MobEffects.LEVITATION) && !this.player.cH() && !worldserver.c(this.player.getBoundingBox().g(0.0625D).b(0.0D, -0.55D, 0.0D));
                             this.player.onGround = packetplayinflying.a();
                             this.minecraftServer.getPlayerList().d(this.player);
+                            // SportBukkit start
+                            if(clientOnGround != this.player.onGround) {
+                                clientOnGround = this.player.onGround;
+                                org.bukkit.event.player.PlayerOnGroundEvent onGroundEvent = new org.bukkit.event.player.PlayerOnGroundEvent(player, this.player.onGround);
+                                this.server.getPluginManager().callEvent(onGroundEvent);
+                            }
+                            // SportBukkit end
                             this.player.a(this.player.locY - d3, packetplayinflying.a());
                             this.o = this.player.locX;
                             this.p = this.player.locY;
