From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 2 Apr 2016 05:09:16 -0400
Subject: [PATCH] Add PlayerUseUnknownEntityEvent


diff --git a/src/main/java/net/minecraft/server/PacketPlayInUseEntity.java b/src/main/java/net/minecraft/server/PacketPlayInUseEntity.java
--- a/src/main/java/net/minecraft/server/PacketPlayInUseEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayInUseEntity.java
@@ -5,11 +5,20 @@ import javax.annotation.Nullable;
 
 public class PacketPlayInUseEntity implements Packet<PacketListenerPlayIn> {
 
-    private int a;
+    private int a; public int getEntityId() { return a; } // SportBukkit - add accessor
     private PacketPlayInUseEntity.EnumEntityUseAction action;
     private Vec3D c;
     private EnumHand d;
 
+    // SportBukkit start - add constructor
+    public PacketPlayInUseEntity(int entityId, EnumEntityUseAction action, Vec3D touchAt, EnumHand hand) {
+        this.a = entityId;
+        this.action = action;
+        this.c = touchAt;
+        this.d = hand;
+    }
+    // SportBukkit end
+
     public PacketPlayInUseEntity() {}
 
     public PacketPlayInUseEntity(Entity entity) {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -42,6 +42,7 @@ import org.bukkit.event.inventory.InventoryType.SlotType;
 import org.bukkit.event.player.AsyncPlayerChatEvent;
 import org.bukkit.event.player.PlayerAnimationEvent;
 import org.bukkit.event.player.PlayerAttackEntityEvent;
+import org.bukkit.event.player.PlayerUseUnknownEntityEvent;
 import org.bukkit.event.player.PlayerChatEvent;
 import org.bukkit.event.player.PlayerCommandPreprocessEvent;
 import org.bukkit.event.player.PlayerInteractAtEntityEvent;
@@ -1685,7 +1686,16 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 }
             }
         }
-
+        // SportBukkit start - fire event
+        else {
+            this.server.getPluginManager().callEvent(new PlayerUseUnknownEntityEvent(
+                this.getPlayer(),
+                packetplayinuseentity.getEntityId(),
+                packetplayinuseentity.a() == PacketPlayInUseEntity.EnumEntityUseAction.ATTACK,
+                packetplayinuseentity.b() == EnumHand.MAIN_HAND ? EquipmentSlot.HAND : EquipmentSlot.OFF_HAND
+            ));
+        }
+        // SportBukkit end
     }
 
     public void a(PacketPlayInClientCommand packetplayinclientcommand) {
