From: YukonAppleGeek <yukonvinecki@gmail.com>
Date: Sun, 26 May 2013 18:43:56 -0700
Subject: [PATCH] Add PlayerSpawnEntityEvent


diff --git a/src/main/java/net/minecraft/server/ItemMonsterEgg.java b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
--- a/src/main/java/net/minecraft/server/ItemMonsterEgg.java
+++ b/src/main/java/net/minecraft/server/ItemMonsterEgg.java
@@ -5,6 +5,12 @@ import java.util.List;
 import java.util.UUID;
 import javax.annotation.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerSpawnEntityEvent;
+// CraftBukkit end
+
 public class ItemMonsterEgg extends Item {
 
     public ItemMonsterEgg() {
@@ -54,6 +60,42 @@ public class ItemMonsterEgg extends Item {
             double d0 = this.a(world, blockposition1);
             Entity entity = a(world, h(itemstack), (double) blockposition1.getX() + 0.5D, (double) blockposition1.getY() + d0, (double) blockposition1.getZ() + 0.5D);
 
+            // SportBukkit start - fire event
+            if (entity != null) {
+                CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack);
+                PlayerSpawnEntityEvent event = new PlayerSpawnEntityEvent(
+                    (org.bukkit.entity.Player) entityhuman.getBukkitEntity(),
+                    entity.getBukkitEntity(),
+                    new org.bukkit.Location(
+                        world.getWorld(),
+                        (double) blockposition1.getX() + 0.5D,
+                        (double) blockposition1.getY() + d0,
+                        (double) blockposition1.getZ() + 0.5D,
+                        MathHelper.g(world.random.nextFloat() * 360.0F),
+                        0.0F
+                    ),
+                    craftItem.clone()
+                );
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    event.getEntity().remove();
+                    return EnumInteractionResult.PASS;
+                }
+
+                if (entity instanceof EntityLiving) {
+                    EntityInsentient entityinsentient = (EntityInsentient) entity;
+
+                    entity.setPositionRotation(event.getLocation().getX(), event.getLocation().getY(), event.getLocation().getZ(), event.getLocation().getYaw(), event.getLocation().getPitch());
+                    entityinsentient.aK = entityinsentient.yaw;
+                    entityinsentient.aM = entityinsentient.yaw;
+                    entityinsentient.prepare(world.D(new BlockPosition(entityinsentient)), (GroupDataEntity) null);
+                    world.addEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+                    entityinsentient.D();
+                }
+            }
+            // SportBukkit end
+
             if (entity != null) {
                 if (entity instanceof EntityLiving && itemstack.hasName()) {
                     entity.setCustomName(itemstack.getName());
