From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 5 Apr 2016 10:38:52 -0400
Subject: [PATCH] Better error logging on IO threads


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -88,7 +88,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
         try {
             this.setProtocol(EnumProtocol.HANDSHAKING);
         } catch (Throwable throwable) {
-            NetworkManager.g.fatal(throwable);
+            NetworkManager.g.fatal("Exception setting protocol", throwable);
         }
 
     }
@@ -108,11 +108,12 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
 
         if (throwable instanceof TimeoutException) {
             chatmessage = new ChatMessage("disconnect.timeout", new Object[0]);
+            NetworkManager.g.debug("Uncaught Exception", throwable);
         } else {
             chatmessage = new ChatMessage("disconnect.genericReason", new Object[] { "Internal Exception: " + throwable});
+            NetworkManager.g.fatal("Uncaught Exception", throwable);
         }
 
-        NetworkManager.g.debug(throwable);
         this.close(chatmessage);
     }
 
