From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 3 Jun 2016 02:02:17 -0400
Subject: [PATCH] Component support for signs


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -11,6 +11,9 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.Set;
+
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.TextComponent;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -26,6 +29,7 @@ import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.util.BungeeChatUtils;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.craftbukkit.util.LazyPlayerSet;
 import org.bukkit.craftbukkit.util.Waitable;
@@ -2223,16 +2227,18 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
             int x = packetplayinupdatesign.a().getX();
             int y = packetplayinupdatesign.a().getY();
             int z = packetplayinupdatesign.a().getZ();
-            String[] lines = new String[4];
+            BaseComponent[] lines = new BaseComponent[4];
 
             for (int i = 0; i < astring.length; ++i) {
-                lines[i] = EnumChatFormat.a(new ChatComponentText(EnumChatFormat.a(astring[i])).toPlainText());
+                lines[i] = new TextComponent(EnumChatFormat.a(astring[i]));
             }
             SignChangeEvent event = new SignChangeEvent((org.bukkit.craftbukkit.block.CraftBlock) player.getWorld().getBlockAt(x, y, z), this.server.getPlayer(this.player), lines);
             this.server.getPluginManager().callEvent(event);
 
             if (!event.isCancelled()) {
-                System.arraycopy(org.bukkit.craftbukkit.block.CraftSign.sanitizeLines(event.getLines()), 0, tileentitysign.lines, 0, 4);
+                for(int i = 0; i < tileentitysign.lines.length; i++) {
+                    tileentitysign.lines[i] = BungeeChatUtils.toNms(event.line(i));
+                }
                 tileentitysign.isEditable = false;
              }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSign.java
@@ -1,85 +1,80 @@
 package org.bukkit.craftbukkit.block;
 
-import net.minecraft.server.ChatComponentText;
-import net.minecraft.server.IChatBaseComponent;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.TextComponent;
 import net.minecraft.server.TileEntitySign;
 import org.bukkit.Material;
 import org.bukkit.block.Block;
 import org.bukkit.block.Sign;
 import org.bukkit.craftbukkit.CraftWorld;
-import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.craftbukkit.util.BungeeChatUtils;
 
 public class CraftSign extends CraftBlockState implements Sign {
     private final TileEntitySign sign;
-    private final String[] lines;
+    private final BaseComponent[] lines = new BaseComponent[4];
 
     public CraftSign(final Block block) {
         super(block);
 
         CraftWorld world = (CraftWorld) block.getWorld();
         sign = (TileEntitySign) world.getTileEntityAt(getX(), getY(), getZ());
-        lines = new String[sign.lines.length];
-        System.arraycopy(revertComponents(sign.lines), 0, lines, 0, lines.length);
+        importFromNms();
     }
 
     public CraftSign(final Material material, final TileEntitySign te) {
         super(material);
         sign = te;
-        lines = new String[sign.lines.length];
-        System.arraycopy(revertComponents(sign.lines), 0, lines, 0, lines.length);
+        importFromNms();
     }
 
-    public String[] getLines() {
+    private void importFromNms() {
+        for(int i = 0; i < lines.length; i++) {
+            lines[i] = sign.lines.length > i ? BungeeChatUtils.toBungee(sign.lines[i])
+                                             : new TextComponent();
+        }
+    }
+
+    @Override
+    public BaseComponent[] lines() {
         return lines;
     }
 
-    public String getLine(int index) throws IndexOutOfBoundsException {
+    @Override
+    public BaseComponent line(int index) {
         return lines[index];
     }
 
-    public void setLine(int index, String line) throws IndexOutOfBoundsException {
+    @Override
+    public void setLine(int index, BaseComponent line) {
         lines[index] = line;
     }
 
+    public String[] getLines() {
+        return BaseComponent.toLegacyArray(lines);
+    }
+
+    public String getLine(int index) throws IndexOutOfBoundsException {
+        return lines[index].toLegacyText();
+    }
+
+    public void setLine(int index, String line) throws IndexOutOfBoundsException {
+        lines[index] = TextComponent.fromLegacyToComponent(line, false);
+    }
+
     @Override
     public boolean update(boolean force, boolean applyPhysics) {
         boolean result = super.update(force, applyPhysics);
 
         if (result) {
-            IChatBaseComponent[] newLines = sanitizeLines(lines);
-            System.arraycopy(newLines, 0, sign.lines, 0, 4);
+            for(int i = 0; i < sign.lines.length; i++) {
+                sign.lines[i] = BungeeChatUtils.toNms(lines[i]);
+            }
             sign.update();
         }
 
         return result;
     }
 
-    public static IChatBaseComponent[] sanitizeLines(String[] lines) {
-        IChatBaseComponent[] components = new IChatBaseComponent[4];
-
-        for (int i = 0; i < 4; i++) {
-            if (i < lines.length && lines[i] != null) {
-                components[i] = CraftChatMessage.fromString(lines[i])[0];
-            } else {
-                components[i] = new ChatComponentText("");
-            }
-        }
-
-        return components;
-    }
-
-    public static String[] revertComponents(IChatBaseComponent[] components) {
-        String[] lines = new String[components.length];
-        for (int i = 0; i < lines.length; i++) {
-            lines[i] = revertComponent(components[i]);
-        }
-        return lines;
-    }
-
-    private static String revertComponent(IChatBaseComponent component) {
-        return CraftChatMessage.fromComponent(component);
-    }
-
     @Override
     public TileEntitySign getTileEntity() {
         return sign;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -58,6 +58,7 @@ import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.map.CraftMapView;
 import org.bukkit.craftbukkit.map.RenderData;
 import org.bukkit.craftbukkit.scoreboard.CraftScoreboard;
+import org.bukkit.craftbukkit.util.BungeeChatUtils;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.craftbukkit.util.Skins;
@@ -701,23 +702,41 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public void sendSignChange(Location loc, String[] lines) {
-        if (getHandle().playerConnection == null) {
-            return;
-        }
-
         if (lines == null) {
             lines = new String[4];
         }
+        final IChatBaseComponent[] nms = new IChatBaseComponent[4];
+        for(int i = 0; i < 4; i++) {
+            nms[i] = lines != null && lines[i] == null ? new ChatComponentText("")
+                                                       : CraftChatMessage.fromString(lines[i])[0];
+        }
+        sendSignChange(loc, nms);
+    }
 
+    @Override
+    public void sendSignChange(Location loc, BaseComponent[] lines) {
+        final IChatBaseComponent[] nms = new IChatBaseComponent[4];
+        for(int i = 0; i < 4; i++) {
+            nms[i] = lines != null && lines[i] == null ? new ChatComponentText("")
+                                                       : BungeeChatUtils.toNms(lines[i]);
+        }
+        sendSignChange(loc, nms);
+    }
+
+    private void sendSignChange(Location loc, IChatBaseComponent[] lines) {
         Validate.notNull(loc, "Location can not be null");
+
         if (lines.length < 4) {
             throw new IllegalArgumentException("Must have at least 4 lines");
         }
 
-        IChatBaseComponent[] components = CraftSign.sanitizeLines(lines);
+        if (getHandle().playerConnection == null) {
+            return;
+        }
+
         TileEntitySign sign = new TileEntitySign();
         sign.setPosition(new BlockPosition(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
-        System.arraycopy(components, 0, sign.lines, 0, sign.lines.length);
+        System.arraycopy(lines, 0, sign.lines, 0, sign.lines.length);
 
         getHandle().playerConnection.sendPacket(sign.getUpdatePacket());
     }
