From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 5 Mar 2016 05:56:05 -0500
Subject: [PATCH] Expose various entity packets


diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java b/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
--- a/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutEntityTeleport.java
@@ -12,6 +12,18 @@ public class PacketPlayOutEntityTeleport implements Packet<PacketListenerPlayOut
     private byte f;
     private boolean g;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutEntityTeleport(int id, double xPos, double yPos, double zPos, byte yaw, byte pitch, boolean onGround) {
+        this.a = id;
+        this.b = xPos;
+        this.c = yPos;
+        this.d = zPos;
+        this.e = yaw;
+        this.f = pitch;
+        this.g = onGround;
+    }
+    // SportBukkit end
+
     public PacketPlayOutEntityTeleport() {}
 
     public PacketPlayOutEntityTeleport(Entity entity) {
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutMount.java b/src/main/java/net/minecraft/server/PacketPlayOutMount.java
--- a/src/main/java/net/minecraft/server/PacketPlayOutMount.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutMount.java
@@ -8,6 +8,13 @@ public class PacketPlayOutMount implements Packet<PacketListenerPlayOut> {
     private int a;
     private int[] b;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutMount(int vehicleId, int... riderIds) {
+        this.a = vehicleId;
+        this.b = riderIds;
+    }
+    // SportBukkit end
+
     public PacketPlayOutMount() {}
 
     public PacketPlayOutMount(Entity entity) {
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
--- a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
@@ -16,6 +16,19 @@ public class PacketPlayOutNamedEntitySpawn implements Packet<PacketListenerPlayO
     private DataWatcher h;
     private List<DataWatcher.Item<?>> i;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutNamedEntitySpawn(int id, UUID uuid, double xPos, double yPos, double zPos, byte yaw, byte pitch, List<DataWatcher.Item<?>> metadata) {
+        this.a = id;
+        this.b = uuid;
+        this.c = xPos;
+        this.d = yPos;
+        this.e = zPos;
+        this.f = yaw;
+        this.g = pitch;
+        this.i = metadata;
+    }
+    // SportBukkit end
+
     public PacketPlayOutNamedEntitySpawn() {}
 
     public PacketPlayOutNamedEntitySpawn(EntityHuman entityhuman) {
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntity.java
@@ -18,6 +18,23 @@ public class PacketPlayOutSpawnEntity implements Packet<PacketListenerPlayOut> {
     private int k;
     private int l;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutSpawnEntity(int id, UUID uuid, double xPos, double yPos, double zPos, int xVel, int yVel, int zVel, int pitch, int yaw, int type, int data) {
+        this.a = id;
+        this.b = uuid;
+        this.c = xPos;
+        this.d = yPos;
+        this.e = zPos;
+        this.f = xVel;
+        this.g = yVel;
+        this.h = zVel;
+        this.i = pitch;
+        this.j = yaw;
+        this.k = type;
+        this.l = data;
+    }
+    // SportBukkit end
+
     public PacketPlayOutSpawnEntity() {}
 
     public PacketPlayOutSpawnEntity(Entity entity, int i) {
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
--- a/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -21,6 +21,24 @@ public class PacketPlayOutSpawnEntityLiving implements Packet<PacketListenerPlay
     private DataWatcher m;
     private List<DataWatcher.Item<?>> n;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutSpawnEntityLiving(int id, UUID uuid, int type, double xPos, double yPos, double zPos, int xVel, int yVel, int zVel, byte yaw, byte pitch, byte headPitch, List<DataWatcher.Item<?>> metadata) {
+        this.a = id;
+        this.b = uuid;
+        this.c = type;
+        this.d = xPos;
+        this.e = yPos;
+        this.f = zPos;
+        this.g = xVel;
+        this.h = yVel;
+        this.i = zVel;
+        this.j = yaw;
+        this.k = pitch;
+        this.l = headPitch;
+        this.n = metadata;
+    }
+    // SportBukkit end
+
     public PacketPlayOutSpawnEntityLiving() {}
 
     public PacketPlayOutSpawnEntityLiving(EntityLiving entityliving) {
