From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 13 Oct 2016 16:12:27 -0400
Subject: [PATCH] Fix Java 8 bytecode verify error


diff --git a/src/main/java/net/minecraft/server/DataWatcher.java b/src/main/java/net/minecraft/server/DataWatcher.java
--- a/src/main/java/net/minecraft/server/DataWatcher.java
+++ b/src/main/java/net/minecraft/server/DataWatcher.java
@@ -231,7 +231,12 @@ public class DataWatcher {
 
         short short0;
 
-        while ((short0 = packetdataserializer.readUnsignedByte()) != 255) {
+        // SportBukkit start - Java 8 bytecode verifier doesn't like the inline assignment, no clue why
+        for(;;) {
+            short0 = packetdataserializer.readUnsignedByte();
+            if(short0 == 255) break;
+            // SportBukkit end
+
             if (arraylist == null) {
                 arraylist = Lists.newArrayList();
             }
