From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 28 May 2015 04:31:39 -0400
Subject: [PATCH] Improve Permissible interface


diff --git a/src/main/java/org/bukkit/craftbukkit/command/ProxiedNativeCommandSender.java b/src/main/java/org/bukkit/craftbukkit/command/ProxiedNativeCommandSender.java
--- a/src/main/java/org/bukkit/craftbukkit/command/ProxiedNativeCommandSender.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/ProxiedNativeCommandSender.java
@@ -1,6 +1,7 @@
 
 package org.bukkit.craftbukkit.command;
 
+import java.util.Collection;
 import java.util.Set;
 import net.minecraft.server.ICommandListener;
 
@@ -119,6 +120,66 @@ public class ProxiedNativeCommandSender implements ProxiedCommandSender {
     }
 
     @Override
+    public boolean removeAttachments(Plugin plugin) {
+        return getCaller().removeAttachments(plugin);
+    }
+
+    @Override
+    public boolean removeAttachments(String name) {
+        return getCaller().removeAttachments(name);
+    }
+
+    @Override
+    public boolean removeAttachments(Permission permission) {
+        return getCaller().removeAttachments(permission);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, String name) {
+        return getCaller().removeAttachments(plugin, name);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, Permission permission) {
+        return getCaller().removeAttachments(plugin, permission);
+    }
+
+    @Override
+    public PermissionAttachmentInfo getEffectivePermission(String name) {
+        return getCaller().getEffectivePermission(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments() {
+        return getCaller().getAttachments();
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin) {
+        return getCaller().getAttachments(plugin);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(String name) {
+        return getCaller().getAttachments(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Permission permission) {
+        return getCaller().getAttachments(permission);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, String name) {
+        return getCaller().getAttachments(plugin, name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, Permission permission) {
+        return getCaller().getAttachments(plugin, permission);
+    }
+
+    @Override
     public boolean isOp() {
         return getCaller().isOp();
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/command/ServerCommandSender.java b/src/main/java/org/bukkit/craftbukkit/command/ServerCommandSender.java
--- a/src/main/java/org/bukkit/craftbukkit/command/ServerCommandSender.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/ServerCommandSender.java
@@ -9,6 +9,7 @@ import org.bukkit.permissions.PermissionAttachment;
 import org.bukkit.permissions.PermissionAttachmentInfo;
 import org.bukkit.plugin.Plugin;
 
+import java.util.Collection;
 import java.util.Set;
 
 public abstract class ServerCommandSender implements CommandSender {
@@ -61,6 +62,66 @@ public abstract class ServerCommandSender implements CommandSender {
         return perm.getEffectivePermissions();
     }
 
+    @Override
+    public boolean removeAttachments(Plugin plugin) {
+        return perm.removeAttachments(plugin);
+    }
+
+    @Override
+    public boolean removeAttachments(String name) {
+        return perm.removeAttachments(name);
+    }
+
+    @Override
+    public boolean removeAttachments(Permission permission) {
+        return perm.removeAttachments(permission);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, String name) {
+        return perm.removeAttachments(plugin, name);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, Permission permission) {
+        return perm.removeAttachments(plugin, permission);
+    }
+
+    @Override
+    public PermissionAttachmentInfo getEffectivePermission(String name) {
+        return perm.getEffectivePermission(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments() {
+        return perm.getAttachments();
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin) {
+        return perm.getAttachments(plugin);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(String name) {
+        return perm.getAttachments(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Permission permission) {
+        return perm.getAttachments(permission);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, String name) {
+        return perm.getAttachments(plugin, name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, Permission permission) {
+        return perm.getAttachments(plugin, permission);
+    }
+
     public boolean isPlayer() {
         return false;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1,6 +1,7 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Preconditions;
+import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.UUID;
@@ -569,6 +570,66 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     }
 
     @Override
+    public boolean removeAttachments(Plugin plugin) {
+        return perm.removeAttachments(plugin);
+    }
+
+    @Override
+    public boolean removeAttachments(String name) {
+        return perm.removeAttachments(name);
+    }
+
+    @Override
+    public boolean removeAttachments(Permission permission) {
+        return perm.removeAttachments(permission);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, String name) {
+        return perm.removeAttachments(plugin, name);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, Permission permission) {
+        return perm.removeAttachments(plugin, permission);
+    }
+
+    @Override
+    public PermissionAttachmentInfo getEffectivePermission(String name) {
+        return perm.getEffectivePermission(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments() {
+        return perm.getAttachments();
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin) {
+        return perm.getAttachments(plugin);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(String name) {
+        return perm.getAttachments(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Permission permission) {
+        return perm.getAttachments(permission);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, String name) {
+        return perm.getAttachments(plugin, name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, Permission permission) {
+        return perm.getAttachments(plugin, permission);
+    }
+
+    @Override
     public boolean isOp() {
         return getPermissibleBase().isOp();
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -1,6 +1,7 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Preconditions;
+import java.util.Collection;
 import java.util.Set;
 
 import net.minecraft.server.*;
@@ -148,6 +149,66 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         return perm.getEffectivePermissions();
     }
 
+    @Override
+    public boolean removeAttachments(Plugin plugin) {
+        return perm.removeAttachments(plugin);
+    }
+
+    @Override
+    public boolean removeAttachments(String name) {
+        return perm.removeAttachments(name);
+    }
+
+    @Override
+    public boolean removeAttachments(Permission permission) {
+        return perm.removeAttachments(permission);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, String name) {
+        return perm.removeAttachments(plugin, name);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, Permission permission) {
+        return perm.removeAttachments(plugin, permission);
+    }
+
+    @Override
+    public PermissionAttachmentInfo getEffectivePermission(String name) {
+        return perm.getEffectivePermission(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments() {
+        return perm.getAttachments();
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin) {
+        return perm.getAttachments(plugin);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(String name) {
+        return perm.getAttachments(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Permission permission) {
+        return perm.getAttachments(permission);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, String name) {
+        return perm.getAttachments(plugin, name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, Permission permission) {
+        return perm.getAttachments(plugin, permission);
+    }
+
     public GameMode getGameMode() {
         return mode;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartCommand.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartCommand.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartCommand.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartCommand.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import java.util.Collection;
 import java.util.Set;
 
 import net.minecraft.server.EntityMinecartCommandBlock;
@@ -132,6 +133,66 @@ public class CraftMinecartCommand extends CraftMinecart implements CommandMineca
     }
 
     @Override
+    public boolean removeAttachments(Plugin plugin) {
+        return perm.removeAttachments(plugin);
+    }
+
+    @Override
+    public boolean removeAttachments(String name) {
+        return perm.removeAttachments(name);
+    }
+
+    @Override
+    public boolean removeAttachments(Permission permission) {
+        return perm.removeAttachments(permission);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, String name) {
+        return perm.removeAttachments(plugin, name);
+    }
+
+    @Override
+    public boolean removeAttachments(Plugin plugin, Permission permission) {
+        return perm.removeAttachments(plugin, permission);
+    }
+
+    @Override
+    public PermissionAttachmentInfo getEffectivePermission(String name) {
+        return perm.getEffectivePermission(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments() {
+        return perm.getAttachments();
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin) {
+        return perm.getAttachments(plugin);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(String name) {
+        return perm.getAttachments(name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Permission permission) {
+        return perm.getAttachments(permission);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, String name) {
+        return perm.getAttachments(plugin, name);
+    }
+
+    @Override
+    public Collection<PermissionAttachmentInfo> getAttachments(Plugin plugin, Permission permission) {
+        return perm.getAttachments(plugin, permission);
+    }
+
+    @Override
     public Server getServer() {
         return Bukkit.getServer();
     }
