From: mrapple <tony@oc.tc>
Date: Thu, 17 Apr 2014 19:56:31 -0500
Subject: [PATCH] Make ticksFlown in EntityFireworks public


diff --git a/src/main/java/net/minecraft/server/EntityFireworks.java b/src/main/java/net/minecraft/server/EntityFireworks.java
--- a/src/main/java/net/minecraft/server/EntityFireworks.java
+++ b/src/main/java/net/minecraft/server/EntityFireworks.java
@@ -7,7 +7,7 @@ public class EntityFireworks extends Entity {
 
     public static final DataWatcherObject<ItemStack> FIREWORK_ITEM = DataWatcher.a(EntityFireworks.class, DataWatcherRegistry.f);
     private static final DataWatcherObject<Integer> b = DataWatcher.a(EntityFireworks.class, DataWatcherRegistry.b);
-    private int ticksFlown;
+    public int ticksFlown; // SportBukkit - private -> public
     public int expectedLifespan;
     private EntityLiving e;
 
