From: Steve Anton <anxuiz.nx@gmail.com>
Date: Fri, 31 Aug 2012 12:30:14 -0700
Subject: [PATCH] More complex logic when looking for nearby players


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2573,7 +2573,7 @@ public abstract class World implements IBlockAccess {
         for (int i = 0; i < this.players.size(); ++i) {
             EntityHuman entityhuman1 = (EntityHuman) this.players.get(i);
             // CraftBukkit start - Fixed an NPE
-            if (entityhuman1 == null || entityhuman1.dead) {
+            if (entityhuman1 == null || !entityhuman1.isAlive()) { // SportBukkit allow for more complex logic by using the "is alive" method
                 continue;
             }
             // CraftBukkit end
