From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 22 Mar 2016 19:53:00 -0400
Subject: [PATCH] Run data converters on passenger entities


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -141,6 +141,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.b = this.i();
         // this.convertable = new WorldLoaderServer(file); // CraftBukkit - moved to DedicatedServer.init
         this.dataConverterManager = dataconvertermanager;
+        this.dataConverterManager.a(DataConverterTypes.ENTITY, new org.bukkit.craftbukkit.DataInspectorEntityPassengers()); // SportBukkit
         // CraftBukkit start
         this.options = options;
         // Try to see if we're actually running in a terminal, disable jline if not
diff --git a/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java b/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java
new file mode 100644
index 0000000..3afe5b8
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java
@@ -0,0 +1,23 @@
+package org.bukkit.craftbukkit;
+
+import net.minecraft.server.DataConverter;
+import net.minecraft.server.DataConverterTypes;
+import net.minecraft.server.DataInspector;
+import net.minecraft.server.NBTTagCompound;
+import net.minecraft.server.NBTTagList;
+
+/**
+ * Run data converters on passenger entities
+ */
+public class DataInspectorEntityPassengers implements DataInspector {
+    @Override
+    public NBTTagCompound a(DataConverter converter, NBTTagCompound nbt, int version) {
+        if(nbt.hasKeyOfType("Passengers", 9)) {
+            final NBTTagList passengers = nbt.getList("Passengers", 10);
+            for(int i = 0; i < passengers.size(); i++) {
+                passengers.a(i, converter.a(DataConverterTypes.ENTITY, passengers.get(i), version));
+            }
+        }
+        return nbt;
+    }
+}
