From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Jan 2017 07:10:14 -0500
Subject: [PATCH] SportBukkit


diff --git a/README.md b/README.md
--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-CraftBukkit
+SportBukkit
 ======
 An implementation of the [Bukkit](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit) plugin API for [Minecraft](https://minecraft.net/) servers, currently maintained by [SpigotMC](http://www.spigotmc.org/).
 
diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -175,6 +175,11 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
 
             try {
                 this.an().a(inetaddress, this.P());
+                // SportBukkit start
+                if(this.getServerConnection().localAddress().getPort() != this.getConfiguredPort()) {
+                    DedicatedServer.LOGGER.info("Bound to port " + this.getServerConnection().localAddress().getPort());
+                }
+                // SportBukkit end
             } catch (IOException ioexception) {
                 DedicatedServer.LOGGER.warn("**** FAILED TO BIND TO PORT!");
                 DedicatedServer.LOGGER.warn("The exception was: {}", new Object[] { ioexception.toString()});
@@ -454,7 +459,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
     }
 
     public int e_() {
-        return this.P();
+        return getServerConnection().localAddress().getPort(); // SportBukkit - return dynamic port, if set, otherwise configured port
     }
 
     public String f_() {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -295,6 +295,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 
                 world.a(worldsettings);
                 this.server.scoreboardManager = new org.bukkit.craftbukkit.scoreboard.CraftScoreboardManager(this, world.getScoreboard());
+                this.server.overworld = world;
             } else {
                 String dim = "DIM" + dimension;
 
@@ -1187,6 +1188,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         return this.I;
     }
 
+    public int getConfiguredPort() { return this.P(); } // SportBukkit - alias for following method
     public int P() {
         return this.u;
     }
@@ -1417,6 +1419,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 
     }
 
+    public ServerConnection getServerConnection() { return this.an(); } // SportBukkit - alias for following method
     public ServerConnection an() {
         return this.p;
     }
diff --git a/src/main/java/net/minecraft/server/PacketStatusListener.java b/src/main/java/net/minecraft/server/PacketStatusListener.java
--- a/src/main/java/net/minecraft/server/PacketStatusListener.java
+++ b/src/main/java/net/minecraft/server/PacketStatusListener.java
@@ -6,6 +6,8 @@ import io.netty.channel.ChannelFutureListener;
 import java.net.InetSocketAddress;
 import java.util.Iterator;
 
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.protocol.CraftServerPing;
 import org.bukkit.craftbukkit.util.CraftIconCache;
 import org.bukkit.entity.Player;
 
@@ -110,12 +112,13 @@ public class PacketStatusListener implements PacketStatusInListener {
             ServerPing.ServerPingPlayerSample playerSample = new ServerPing.ServerPingPlayerSample(event.getMaxPlayers(), profiles.size());
             playerSample.a(profiles.toArray(new GameProfile[profiles.size()]));
 
-            ServerPing ping = new ServerPing();
+            CraftServerPing ping = new CraftServerPing();
             ping.setFavicon(event.icon.value);
             ping.setMOTD(new ChatComponentText(event.getMotd()));
             ping.setPlayerSample(playerSample);
             int version = minecraftServer.getServerPing().getServerData().getProtocolVersion();
             ping.setServerInfo(new ServerPing.ServerData(minecraftServer.getServerModName() + " " + minecraftServer.getVersion(), version));
+            event.getExtra().forEach((name, value) -> ping.put(name, CraftServer.gson.toJsonTree(value)));
 
             this.networkManager.sendPacket(new PacketStatusOutServerInfo(ping));
         }
diff --git a/src/main/java/net/minecraft/server/PacketStatusOutServerInfo.java b/src/main/java/net/minecraft/server/PacketStatusOutServerInfo.java
--- a/src/main/java/net/minecraft/server/PacketStatusOutServerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketStatusOutServerInfo.java
@@ -2,11 +2,18 @@ package net.minecraft.server;
 
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
+import org.bukkit.craftbukkit.protocol.CraftServerPing;
+
 import java.io.IOException;
 
 public class PacketStatusOutServerInfo implements Packet<PacketStatusOutListener> {
 
-    private static final Gson a = (new GsonBuilder()).registerTypeAdapter(ServerPing.ServerData.class, new ServerPing.ServerData.ServerData$Serializer()).registerTypeAdapter(ServerPing.ServerPingPlayerSample.class, new ServerPing.ServerPingPlayerSample.ServerPingPlayerSample$Serializer()).registerTypeAdapter(ServerPing.class, new ServerPing.Serializer()).registerTypeHierarchyAdapter(IChatBaseComponent.class, new IChatBaseComponent.ChatSerializer()).registerTypeHierarchyAdapter(ChatModifier.class, new ChatModifier.ChatModifierSerializer()).registerTypeAdapterFactory(new ChatTypeAdapterFactory()).create();
+    // SportBukkit start - add CraftServerPing.Serializer
+    private static final Gson a = (new GsonBuilder()).registerTypeAdapter(ServerPing.ServerData.class, new ServerPing.ServerData.Serializer()).registerTypeAdapter(ServerPing.ServerPingPlayerSample.class, new ServerPing.ServerPingPlayerSample.Serializer()).registerTypeAdapter(ServerPing.class, new ServerPing.Serializer()).registerTypeHierarchyAdapter(IChatBaseComponent.class, new IChatBaseComponent.ChatSerializer()).registerTypeHierarchyAdapter(ChatModifier.class, new ChatModifier.ChatModifierSerializer()).registerTypeAdapterFactory(new ChatTypeAdapterFactory())
+                                                     .registerTypeAdapter(CraftServerPing.class, new CraftServerPing.Serializer())
+                                                     .create();
+    // SportBukkit end
+
     private ServerPing b;
 
     public PacketStatusOutServerInfo() {}
diff --git a/src/main/java/net/minecraft/server/ServerConnection.java b/src/main/java/net/minecraft/server/ServerConnection.java
--- a/src/main/java/net/minecraft/server/ServerConnection.java
+++ b/src/main/java/net/minecraft/server/ServerConnection.java
@@ -64,6 +64,20 @@ public class ServerConnection {
     private final List<NetworkManager> h = Collections.synchronizedList(Lists.<NetworkManager>newArrayList());
     // SportBukkit end
 
+    // SportBukkit start
+    private java.net.InetSocketAddress localAddress;
+    public java.net.InetSocketAddress localAddress() {
+        if(localAddress != null) {
+            // If connected, return bound address (with potentially dynamic port)
+            return localAddress;
+        } else {
+            // If not connected yet, return configured address and port (which may be 0)
+            return java.net.InetSocketAddress.createUnresolved(this.f.getServerIp(),
+                                                               this.f.getConfiguredPort());
+        }
+    }
+    // SportBukkit end
+
     public ServerConnection(MinecraftServer minecraftserver) {
         this.f = minecraftserver;
         this.d = true;
@@ -103,6 +117,8 @@ public class ServerConnection {
                     f.server.eventBus().callEvent(new org.bukkit.craftbukkit.event.AsyncClientConnectEvent(channel, networkmanager)); // SportBukkit
                 }
             }).group((EventLoopGroup) lazyinitvar.c()).localAddress(inetaddress, i)).bind().syncUninterruptibly());
+
+            this.localAddress = (java.net.InetSocketAddress) this.g.get(0).channel().localAddress(); // SportBukkit
         }
     }
 
diff --git a/src/main/java/net/minecraft/server/WorldNBTStorage.java b/src/main/java/net/minecraft/server/WorldNBTStorage.java
--- a/src/main/java/net/minecraft/server/WorldNBTStorage.java
+++ b/src/main/java/net/minecraft/server/WorldNBTStorage.java
@@ -191,6 +191,14 @@ public class WorldNBTStorage implements IDataManager, IPlayerFileData {
     }
 
     // CraftBukkit start
+    public File getPlayerFile(UUID id) {
+        return new File(this.playerDir, id + ".dat");
+    }
+
+    public boolean hasPlayerData(UUID id) {
+        return getPlayerFile(id).isFile();
+    }
+
     public NBTTagCompound getPlayerData(String s) {
         try {
             File file1 = new File(this.playerDir, s + ".dat");
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -2,12 +2,13 @@ package org.bukkit.craftbukkit;
 
 import com.mojang.authlib.GameProfile;
 import java.io.File;
+import java.time.Instant;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.UUID;
 
-import net.minecraft.server.EntityPlayer;
 import net.minecraft.server.NBTTagCompound;
 import net.minecraft.server.WorldNBTStorage;
 
@@ -21,18 +22,34 @@ import org.bukkit.configuration.serialization.SerializableAs;
 import org.bukkit.entity.Player;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
+import tc.oc.collection.Optionals;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 @SerializableAs("Player")
 public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializable {
     private final GameProfile profile;
     private final CraftServer server;
     private final WorldNBTStorage storage;
+    private final Optional<Instant> updatedAt;
 
     protected CraftOfflinePlayer(CraftServer server, GameProfile profile) {
+        this(server, profile, Optional.empty());
+    }
+
+    protected CraftOfflinePlayer(CraftServer server, GameProfile profile, Optional<Instant> updatedAt) {
+        checkNotNull(profile.getId(), "OfflinePlayer must have a UUID");
+        checkArgument(!(updatedAt.isPresent() && profile.getName() == null), "OfflinePlayer has an update timestamp, but no username");
         this.server = server;
         this.profile = profile;
-        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
+        this.updatedAt = updatedAt;
+        this.storage = (WorldNBTStorage) (server.overworld.getDataManager());
+    }
 
+    @Override
+    public Optional<Instant> updatedAt() {
+        return Optionals.first(lastPlayedAt(), updatedAt);
     }
 
     public GameProfile getProfile() {
@@ -213,6 +230,12 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         }
     }
 
+    @Override
+    public Optional<Instant> lastPlayedAt() {
+        final long milli = getLastPlayed();
+        return milli == 0 ? Optional.empty() : Optional.of(Instant.ofEpochMilli(milli));
+    }
+
     public long getLastPlayed() {
         Player player = getPlayer();
         if (player != null) return player.getLastPlayed();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayerFactory.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayerFactory.java
new file mode 100644
index 0000000..e9c460e
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayerFactory.java
@@ -0,0 +1,51 @@
+package org.bukkit.craftbukkit;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import com.mojang.authlib.GameProfile;
+import tc.oc.minecraft.api.user.User;
+import tc.oc.minecraft.api.user.UserFactory;
+import tc.oc.minecraft.api.user.UserUtils;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@Singleton
+public class CraftOfflinePlayerFactory implements UserFactory {
+
+    private final Provider<CraftServer> server;
+
+    @Inject CraftOfflinePlayerFactory(Provider<CraftServer> server) {
+        this.server = server;
+    }
+
+    public CraftOfflinePlayer createUser(GameProfile profile, Optional<Instant> updatedAt) {
+        return new CraftOfflinePlayer(server.get(), profile, updatedAt);
+    }
+
+    @Override
+    public CraftOfflinePlayer createUser(UUID id) {
+        return createUser(new GameProfile(checkNotNull(id), null), Optional.empty());
+    }
+
+    @Override
+    public CraftOfflinePlayer createUser(String name) {
+        return createUser(new GameProfile(UserUtils.offlinePlayerId(name), name), Optional.empty());
+    }
+
+    @Override
+    public CraftOfflinePlayer createUser(UUID id, String name, Instant updatedAt) {
+        return createUser(new GameProfile(checkNotNull(id), checkNotNull(name)), Optional.of(updatedAt));
+    }
+
+    @Override
+    public CraftOfflinePlayer copyUser(User user) {
+        return user instanceof CraftOfflinePlayer
+               ? (CraftOfflinePlayer) user
+               : (CraftOfflinePlayer) UserFactory.super.copyUser(user);
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -6,6 +6,9 @@ import java.io.FileInputStream;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStreamReader;
+import java.net.InetSocketAddress;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.ArrayList;
@@ -21,9 +24,10 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.logging.Level;
 import java.util.logging.Logger;
-import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 import javax.imageio.ImageIO;
+import javax.inject.Inject;
+import javax.inject.Provider;
 
 import com.google.common.collect.Maps;
 import com.google.inject.Guice;
@@ -133,7 +137,6 @@ import com.avaje.ebeaninternal.server.lib.sql.TransactionIsolation;
 import com.google.common.base.Charsets;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
-import com.google.common.collect.MapMaker;
 import com.mojang.authlib.GameProfile;
 
 import io.netty.buffer.ByteBuf;
@@ -143,8 +146,10 @@ import io.netty.handler.codec.base64.Base64;
 import jline.console.ConsoleReader;
 import org.bukkit.event.server.TabCompleteEvent;
 import net.md_5.bungee.api.chat.BaseComponent;
+import tc.oc.inject.Providers;
 import tc.oc.minecraft.api.configuration.InvalidConfigurationException;
 import tc.oc.minecraft.api.plugin.PluginFinder;
+import tc.oc.minecraft.api.user.UserFinder;
 
 public final class CraftServer extends CraftBukkitRuntime implements Server {
     private static final Player[] EMPTY_PLAYER_ARRAY = new Player[0];
@@ -153,7 +158,8 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
-    private final CraftScheduler scheduler = new CraftScheduler();
+    private @Inject Provider<CraftScheduler> scheduler = Providers.unavailable(CraftScheduler.class);
+    private @Inject Provider<UserFinder> userFinder = Providers.unavailable(UserFinder.class);
     private final SimpleCommandMap commandMap = new SimpleCommandMap(this);
     private final SimpleHelpMap helpMap = new SimpleHelpMap(this);
     private final StandardMessenger messenger = new StandardMessenger();
@@ -161,6 +167,7 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     private final EventBus eventBus;
     protected final MinecraftServer console;
     protected final DedicatedPlayerList playerList;
+    public WorldServer overworld;
     private final Map<String, World> worlds = new LinkedHashMap<String, World>();
     private final Map<String, World> worldsView = new CaseInsensitiveNameMap<>(worlds.values(), World::getName);
     private final Map<UUID, World> worldsById = new LinkedHashMap<>();
@@ -168,7 +175,6 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     private YamlConfiguration configuration;
     private YamlConfiguration commandsConfiguration;
     private final Yaml yaml = new Yaml(new SafeConstructor());
-    private final Map<UUID, OfflinePlayer> offlinePlayers = new MapMaker().softValues().makeMap();
     private final EntityMetadataStore entityMetadata = new EntityMetadataStore();
     private final PlayerMetadataStore playerMetadata = new PlayerMetadataStore();
     private final WorldMetadataStore worldMetadata = new WorldMetadataStore();
@@ -187,13 +193,13 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     private boolean printSaveWarning;
     private CraftIconCache icon;
     private boolean overrideAllCommandBlockCommands = false;
-    private final Pattern validUserPattern = Pattern.compile("^[a-zA-Z0-9_]{2,16}$");
     private final UUID invalidUserUUID = UUID.nameUUIDFromBytes("InvalidUsername".getBytes(Charsets.UTF_8));
     private final List<CraftPlayer> playerView;
     private final Map<UUID, Player> playersById;
     public int reloadCount;
     public boolean bungee = false;
     public static final com.google.gson.Gson gson = new com.google.gson.Gson();
+    private final Path root;
 
     private @Nullable Instant emptySince;
 
@@ -208,6 +214,7 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
 
     public CraftServer(MinecraftServer console, PlayerList playerList) {
         this.console = console;
+        this.root = Paths.get(".").toAbsolutePath();
 
         this.eventBus = new SimpleEventBus(this.console.primaryThread, task -> {
             if(console.isMainThread()) {
@@ -337,7 +344,11 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
             logger.info("Creating injector in stage " + stage);
 
             try {
-                injector = Guice.createInjector(stage, new ServerInstanceModule(this, Arrays.asList(plugins)));
+                injector = Guice.createInjector(
+                    stage,
+                    new CraftServerModule(),
+                    new ServerInstanceModule(this, Arrays.asList(plugins))
+                );
             } catch(RuntimeException ex) {
                 logger.log(Level.SEVERE, "Injector creation failed, server will shut down", ex);
                 throw ex;
@@ -559,7 +570,7 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     // so if that changes this will need to as well
     @Override
     public int getPort() {
-        return this.getConfigInt("server-port", 25565);
+        return getAddress().getPort();
     }
 
     @Override
@@ -569,7 +580,12 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
 
     @Override
     public String getIp() {
-        return this.getConfigString("server-ip", "");
+        return getAddress().getHostString();
+    }
+
+    @Override
+    public InetSocketAddress getAddress() {
+        return getHandle().getServer().getServerConnection().localAddress();
     }
 
     @Override
@@ -645,6 +661,11 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     // End Temporary calls
 
     @Override
+    public Path getRootPath() {
+        return root;
+    }
+
+    @Override
     public String getUpdateFolder() {
         return this.configuration.getString("settings.update-folder", "update");
     }
@@ -706,7 +727,7 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
 
     @Override
     public CraftScheduler getScheduler() {
-        return scheduler;
+        return scheduler.get();
     }
 
     @Override
@@ -1430,6 +1451,11 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     }
 
     @Override
+    public boolean isStopping() {
+        return !console.isRunning();
+    }
+
+    @Override
     public int broadcast(String message, String permission) {
         int count = 0;
         Set<Permissible> permissibles = getPluginManager().getPermissionSubscriptions(permission);
@@ -1448,53 +1474,16 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     @Override
     @Deprecated
     public OfflinePlayer getOfflinePlayer(String name) {
-        Validate.notNull(name, "Name cannot be null");
-
-        // If the name given cannot ever be a valid username give a dummy return, for scoreboard plugins
-        if (!validUserPattern.matcher(name).matches()) {
-            return new CraftOfflinePlayer(this, new GameProfile(invalidUserUUID, name));
-        }
-
-        OfflinePlayer result = getPlayerExact(name);
-        if (result == null) {
-            // This is potentially blocking :(
-            GameProfile profile = console.getUserCache().getProfile(name);
-            if (profile == null) {
-                // Make an OfflinePlayer using an offline mode UUID since the name has no profile
-                result = getOfflinePlayer(new GameProfile(UUID.nameUUIDFromBytes(("OfflinePlayer:" + name.toLowerCase()).getBytes(Charsets.UTF_8)), name));
-            } else {
-                // Use the GameProfile even when we get a UUID so we ensure we still have a name
-                result = getOfflinePlayer(profile);
-            }
-        } else {
-            offlinePlayers.remove(result.getUniqueId());
-        }
-
-        return result;
+        return (OfflinePlayer) userFinder.get().findUser(name);
     }
 
     @Override
     public OfflinePlayer getOfflinePlayer(UUID id) {
-        Validate.notNull(id, "UUID cannot be null");
-
-        OfflinePlayer result = getPlayer(id);
-        if (result == null) {
-            result = offlinePlayers.get(id);
-            if (result == null) {
-                result = new CraftOfflinePlayer(this, new GameProfile(id, null));
-                offlinePlayers.put(id, result);
-            }
-        } else {
-            offlinePlayers.remove(id);
-        }
-
-        return result;
+        return (OfflinePlayer) userFinder.get().findUser(id);
     }
 
     public OfflinePlayer getOfflinePlayer(GameProfile profile) {
-        OfflinePlayer player = new CraftOfflinePlayer(this, profile);
-        offlinePlayers.put(profile.getId(), player);
-        return player;
+        return new CraftOfflinePlayer(this, profile);
     }
 
     @Override
@@ -1622,21 +1611,22 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     }
 
     @Override
-    public OfflinePlayer[] getOfflinePlayers() {
-        WorldNBTStorage storage = (WorldNBTStorage) console.worlds.get(0).getDataManager();
-        String[] files = storage.getPlayerDir().list(new DatFileFilter());
-        Set<OfflinePlayer> players = new HashSet<OfflinePlayer>();
-
-        for (String file : files) {
+    public Set<OfflinePlayer> getSavedPlayers() {
+        final Set<OfflinePlayer> players = new HashSet<>();
+        for (String file : ((WorldNBTStorage) console.worlds.get(0).getDataManager()).getPlayerDir().list(new DatFileFilter())) {
             try {
                 players.add(getOfflinePlayer(UUID.fromString(file.substring(0, file.length() - 4))));
             } catch (IllegalArgumentException ex) {
                 // Who knows what is in this directory, just ignore invalid files
             }
         }
-
         players.addAll(getOnlinePlayers());
+        return players;
+    }
 
+    @Override
+    public OfflinePlayer[] getOfflinePlayers() {
+        final Set<OfflinePlayer> players = getSavedPlayers();
         return players.toArray(new OfflinePlayer[players.size()]);
     }
 
@@ -1900,6 +1890,11 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     }
 
     @Override
+    public boolean isMainThread() {
+        return getHandle().getServer().isMainThread();
+    }
+
+    @Override
     public void postToMainThread(Plugin plugin, boolean priority, Runnable task) {
         getHandle().getServer().addMainThreadTask(priority, wrapTask(plugin, task));
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServerModule.java b/src/main/java/org/bukkit/craftbukkit/CraftServerModule.java
new file mode 100644
index 0000000..5a9bb0a
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServerModule.java
@@ -0,0 +1,56 @@
+package org.bukkit.craftbukkit;
+
+import javax.inject.Singleton;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Provides;
+import com.google.inject.multibindings.OptionalBinder;
+import com.mojang.authlib.GameProfileRepository;
+import com.mojang.authlib.minecraft.MinecraftSessionService;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Server;
+import org.bukkit.ServerModule;
+import org.bukkit.craftbukkit.event.CraftServerEventRegistry;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
+import org.bukkit.craftbukkit.user.MojangUserSource;
+import org.bukkit.craftbukkit.user.UserUpdateListener;
+import tc.oc.minecraft.api.user.UserFactory;
+import tc.oc.minecraft.api.user.UserSourceBinder;
+
+public class CraftServerModule extends AbstractModule {
+
+    @Override
+    protected void configure() {
+        install(new ServerModule());
+
+        bind(CraftServerEventRegistry.class);
+        bind(CraftScheduler.class).in(Singleton.class);
+        bind(UserUpdateListener.class).asEagerSingleton();
+
+        OptionalBinder.newOptionalBinder(binder(), UserFactory.class)
+                      .setBinding().to(CraftOfflinePlayerFactory.class);
+
+        new UserSourceBinder(binder())
+            .addBinding().to(MojangUserSource.class);
+    }
+
+    @Provides
+    CraftServer craftServer(Server server) {
+        return (CraftServer) server;
+    }
+
+    @Provides
+    MinecraftServer minecraftServer(CraftServer craftServer) {
+        return craftServer.getServer();
+    }
+
+    @Provides
+    MinecraftSessionService minecraftSessionService(MinecraftServer minecraftServer) {
+        return minecraftServer.az();
+    }
+
+    @Provides
+    GameProfileRepository gameProfileRepository(MinecraftServer minecraftServer) {
+        return minecraftServer.getGameProfileRepository();
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftServerEventRegistry.java b/src/main/java/org/bukkit/craftbukkit/event/CraftServerEventRegistry.java
new file mode 100644
index 0000000..5da2c15
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftServerEventRegistry.java
@@ -0,0 +1,16 @@
+package org.bukkit.craftbukkit.event;
+
+import javax.inject.Inject;
+
+import org.bukkit.event.SimpleEventRegistry;
+import tc.oc.minecraft.api.server.ServerExceptionHandler;
+
+/**
+ * Used internally to register event listeners that don't belong to any plugin
+ */
+public class CraftServerEventRegistry extends SimpleEventRegistry {
+
+    @Inject CraftServerEventRegistry(ServerExceptionHandler exceptionHandler) {
+        super(exceptionHandler);
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/protocol/CraftServerPing.java b/src/main/java/org/bukkit/craftbukkit/protocol/CraftServerPing.java
new file mode 100644
index 0000000..1ada7b6
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/protocol/CraftServerPing.java
@@ -0,0 +1,31 @@
+package org.bukkit.craftbukkit.protocol;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import net.minecraft.server.ServerPing;
+
+public class CraftServerPing extends ServerPing {
+
+    private final Map<String, JsonElement> extra = new HashMap<>();
+
+    public void put(String name, JsonElement value) {
+        extra.put(name, value);
+    }
+
+    private static final ServerPing.Serializer parentSerializer = new ServerPing.Serializer();
+
+    public static class Serializer implements JsonSerializer<CraftServerPing> {
+        @Override
+        public JsonElement serialize(CraftServerPing ping, Type type, JsonSerializationContext context) {
+            final JsonObject json = (JsonObject) parentSerializer.serialize(ping, ServerPing.class, context);
+            ping.extra.forEach(json::add);
+            return json;
+        }
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -9,11 +9,11 @@ import java.util.PriorityQueue;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
-import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.logging.Level;
+import javax.inject.Inject;
 
 import org.apache.commons.lang.Validate;
 import org.bukkit.plugin.IllegalPluginAccessException;
@@ -22,6 +22,7 @@ import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.scheduler.BukkitScheduler;
 import org.bukkit.scheduler.BukkitTask;
 import org.bukkit.scheduler.BukkitWorker;
+import tc.oc.minecraft.api.scheduler.Async;
 
 /**
  * The fundamental concepts for this implementation:
@@ -76,7 +77,7 @@ public class CraftScheduler implements BukkitScheduler {
      */
     private final ConcurrentHashMap<Integer, CraftTask> runners = new ConcurrentHashMap<Integer, CraftTask>();
     private volatile int currentTick = -1;
-    private final Executor executor = Executors.newCachedThreadPool(new com.google.common.util.concurrent.ThreadFactoryBuilder().setNameFormat("Craft Scheduler Thread - %1$d").build()); // Spigot
+    private @Inject @Async Executor executor;
     private CraftAsyncDebugger debugHead = new CraftAsyncDebugger(-1, null, null) {@Override StringBuilder debugTo(StringBuilder string) {return string;}};
     private CraftAsyncDebugger debugTail = debugHead;
     private static final int RECENT_TICKS;
diff --git a/src/main/java/org/bukkit/craftbukkit/user/MojangUserSource.java b/src/main/java/org/bukkit/craftbukkit/user/MojangUserSource.java
new file mode 100644
index 0000000..655927f
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/user/MojangUserSource.java
@@ -0,0 +1,68 @@
+package org.bukkit.craftbukkit.user;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.mojang.authlib.Agent;
+import com.mojang.authlib.GameProfile;
+import com.mojang.authlib.GameProfileRepository;
+import com.mojang.authlib.ProfileLookupCallback;
+import com.mojang.authlib.minecraft.MinecraftSessionService;
+import org.bukkit.craftbukkit.CraftOfflinePlayer;
+import org.bukkit.craftbukkit.CraftOfflinePlayerFactory;
+import tc.oc.minecraft.api.user.User;
+import tc.oc.minecraft.api.user.UserSearch;
+import tc.oc.minecraft.api.user.UserSource;
+
+@Singleton
+public class MojangUserSource implements UserSource {
+
+    private final CraftOfflinePlayerFactory userFactory;
+    private final MinecraftSessionService sessions;
+    private final GameProfileRepository profiles;
+
+    @Inject MojangUserSource(CraftOfflinePlayerFactory userFactory, MinecraftSessionService sessions, GameProfileRepository profiles) {
+        this.userFactory = userFactory;
+        this.sessions = sessions;
+        this.profiles = profiles;
+    }
+
+    @Override
+    public User search(UserSearch search, Supplier<User> next) {
+        // Never fetch on the main thread
+        if(search.sync()) return next.get();
+
+        final CraftOfflinePlayer op;
+        if(search.query() instanceof UUID) {
+            op = userFactory.createUser(
+                sessions.fillProfileProperties(new GameProfile((UUID) search.query(), null), true),
+                Optional.of(Instant.now())
+            );
+        } else {
+            // Despite the weird callback, this does appear to be entirely synchronous
+            final NameCallback callback = new NameCallback();
+            profiles.findProfilesByNames(new String[] {(String) search.query()}, Agent.MINECRAFT, callback);
+            if(callback.user == null) return next.get();
+            op = callback.user;
+        }
+        return search.filter().test(op) ? op : next.get();
+    }
+
+    private class NameCallback implements ProfileLookupCallback {
+
+        @Nullable CraftOfflinePlayer user;
+
+        @Override
+        public void onProfileLookupSucceeded(GameProfile profile) {
+            user = userFactory.createUser(profile, Optional.of(Instant.now()));
+        }
+
+        @Override
+        public void onProfileLookupFailed(GameProfile profile, Exception e) {}
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/user/UserUpdateListener.java b/src/main/java/org/bukkit/craftbukkit/user/UserUpdateListener.java
new file mode 100644
index 0000000..c89efb7
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/user/UserUpdateListener.java
@@ -0,0 +1,39 @@
+package org.bukkit.craftbukkit.user;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.inject.Inject;
+
+import org.bukkit.craftbukkit.event.CraftServerEventRegistry;
+import org.bukkit.event.EventException;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.player.AsyncPlayerPreLoginEvent;
+import tc.oc.minecraft.api.user.User;
+import tc.oc.minecraft.api.user.UserFactory;
+import tc.oc.minecraft.api.user.UserSource;
+
+/**
+ * Refreshes all {@link UserSource}s when a player logs in
+ */
+public class UserUpdateListener implements Listener {
+
+    private final List<UserSource> userSources;
+    private final UserFactory userFactory;
+
+    @Inject UserUpdateListener(Set<UserSource> userSources, UserFactory userFactory, CraftServerEventRegistry eventRegistry) {
+        this.userSources = new ArrayList<>(userSources);
+        Collections.reverse(this.userSources); // Update sources in the same order as cache propagates
+        this.userFactory = userFactory;
+        eventRegistry.registerListener(this);
+    }
+
+    @EventHandler
+    public void callEvent(AsyncPlayerPreLoginEvent event) throws EventException {
+        final User user = userFactory.createUser(event.getUniqueId(), event.getName(), Instant.now());
+        userSources.forEach(source -> source.update(user));
+    }
+}
