From: mrapple <tony@oc.tc>
Date: Tue, 15 Apr 2014 01:12:54 -0500
Subject: [PATCH] Stop server on failed port binding


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -568,7 +568,8 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
                 }
                 // SportBukkit end
             } else {
-                this.a((CrashReport) null);
+                // CraftBukkit - if init fails, stop the server
+                // this.a((CrashReport) null);
             }
         } catch (Throwable throwable) {
             MinecraftServer.LOGGER.error("Encountered an unexpected exception", throwable);
