From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -65,6 +65,8 @@ public abstract class EntityHuman extends EntityLiving {
     @Nullable
     public EntityFishingHook hookedFish;
 
+    public int clientTicksSinceLastAttack; // SportBukkit
+
     // SportBukkit start - natural regen API
     public boolean fastNaturalRegeneration = true;
     public boolean slowNaturalRegeneration = true;
@@ -1909,11 +1911,12 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     public float o(float f) {
-        return MathHelper.a(((float) this.aE + f) / this.dg(), 0.0F, 1.0F);
+        return MathHelper.a(((float) Math.max(this.aE, this.clientTicksSinceLastAttack) + f) / this.dg(), 0.0F, 1.0F); // SportBukkit - check client and server ticks
     }
 
     public void dh() {
         this.aE = 0;
+        this.clientTicksSinceLastAttack = 0; // SportBukkit
     }
 
     public ItemCooldown di() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -496,6 +496,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.syncPosition();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // SportBukkit
+
                 // SportBukkit start - velocity prediction
                 this.player.handleClientTick(
                     packetplayinflying.hasPos ? new Vector(packetplayinflying.x,
