--- ../src-base/minecraft/net/minecraft/util/EnumFacing.java
+++ ../src-work/minecraft/net/minecraft/util/EnumFacing.java
@@ -12,8 +12,6 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3i;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public enum EnumFacing implements IStringSerializable
 {
@@ -66,7 +64,6 @@
         return getFront(this.opposite);
     }
 
-    @SideOnly(Side.CLIENT)
     public EnumFacing rotateAround(EnumFacing.Axis axis)
     {
         switch (axis)
@@ -117,7 +114,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateX()
     {
         switch (this)
@@ -137,7 +133,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateZ()
     {
         switch (this)
@@ -199,7 +194,6 @@
     }
 
     @Nullable
-    @SideOnly(Side.CLIENT)
     public static EnumFacing byName(String name)
     {
         return name == null ? null : (EnumFacing)NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
@@ -230,7 +224,6 @@
         return values()[rand.nextInt(values().length)];
     }
 
-    @SideOnly(Side.CLIENT)
     public static EnumFacing getFacingFromVector(float x, float y, float z)
     {
         EnumFacing enumfacing = NORTH;
@@ -293,7 +286,6 @@
         return placer.getHorizontalFacing().getOpposite();
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3i getDirectionVec()
     {
         return this.directionVec;
@@ -330,7 +322,6 @@
         }
 
         @Nullable
-        @SideOnly(Side.CLIENT)
         public static EnumFacing.Axis byName(String name)
         {
             return name == null ? null : (EnumFacing.Axis)NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
