From e3060229ca80a5864986c6858946d388b6d0e274 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Tue, 7 Aug 2018 00:30:15 -0700
Subject: [PATCH] Add invis override API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 735b4fc1..a75b86ca 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1101,6 +1101,10 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public boolean canSee(Player player);
 
+    public boolean canSeeInvisibles();
+
+    public void showInvisibles(boolean see);
+
     /**
      * Checks to see if this player is currently standing on a block. This
      * information may not be reliable, as it is a state provided by the
-- 
2.18.0

