From 3d160f6cdfb5f271ce309280cf1d235ed3f11b0b Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Tue, 7 Aug 2018 02:14:28 -0700
Subject: [PATCH] Relative teleport API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index bfa93c05..2156211f 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1334,6 +1334,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     public void updateVelocity();
 
+    /**
+     * Seamlessly move the player by the given offset and rotation angles.
+     * Because the relative calculation is applied by the client, you cannot
+     * know the player's absolute location immediately after teleporting.
+     */
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause);
+    public boolean teleportRelative(org.bukkit.util.Vector deltaPos, float deltaYaw, float deltaPitch);
+
     // Spigot start
     public class Spigot extends Entity.Spigot
     {
-- 
2.18.0

