From 6a7c8b8b6dde05fc48045650e2994e90e0805a40 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Wed, 8 Aug 2018 14:37:20 -0700
Subject: [PATCH] Custom replacement for eaten items


diff --git a/src/main/java/org/bukkit/event/player/PlayerItemConsumeEvent.java b/src/main/java/org/bukkit/event/player/PlayerItemConsumeEvent.java
index 8ab76b1d..a81b6429 100644
--- a/src/main/java/org/bukkit/event/player/PlayerItemConsumeEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerItemConsumeEvent.java
@@ -20,6 +20,7 @@ public class PlayerItemConsumeEvent extends PlayerEvent implements Cancellable {
     private static final HandlerList handlers = new HandlerList();
     private boolean isCancelled = false;
     private ItemStack item;
+    private ItemStack replacement;
 
     /**
      * @param player the player consuming
@@ -55,6 +56,22 @@ public class PlayerItemConsumeEvent extends PlayerEvent implements Cancellable {
         }
     }
 
+    /**
+     * Return the custom item stack that will replace the consumed item, or null if no
+     * custom replacement has been set (which means the default replacement will be used).
+     */
+    public ItemStack getReplacement() {
+        return replacement;
+    }
+
+    /**
+     * Set a custom item stack to replace the consumed item. Pass null to clear any custom
+     * stack that has been set and use the default replacement.
+     */
+    public void setReplacement(ItemStack replacement) {
+        this.replacement = replacement;
+    }
+
     public boolean isCancelled() {
         return this.isCancelled;
     }
-- 
2.18.0

