From e5553b4f5e909ab454fbb3154d98e4bee2b4bcc2 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Wed, 8 Aug 2018 22:02:36 -0700
Subject: [PATCH] Skull skin API


diff --git a/src/main/java/org/bukkit/block/Skull.java b/src/main/java/org/bukkit/block/Skull.java
index 4f4896f8..ff92c2d3 100644
--- a/src/main/java/org/bukkit/block/Skull.java
+++ b/src/main/java/org/bukkit/block/Skull.java
@@ -1,7 +1,10 @@
 package org.bukkit.block;
 
+import org.bukkit.Skin;
 import org.bukkit.SkullType;
 
+import java.util.UUID;
+
 /**
  * Represents a Skull
  */
@@ -32,6 +35,21 @@ public interface Skull extends BlockState {
      */
     public boolean setOwner(String name);
 
+    /**
+     * Set the owner and appearance of this skull. A skull with this data set
+     * does not need to fetch anything remotely.
+     *
+     * @param name Username of the skull's owner, can be null (appears in item tooltip)
+     * @param uuid UUID of the skull's owner
+     * @param skin Skull owner's skin
+     */
+    void setOwner(String name, UUID uuid, Skin skin);
+
+    /**
+     * Clear any owner data in this skull
+     */
+    void clearOwner();
+
     /**
      * Gets the rotation of the skull in the world
      *
diff --git a/src/main/java/org/bukkit/inventory/meta/SkullMeta.java b/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
index fab31190..cc5497a0 100644
--- a/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
@@ -1,6 +1,9 @@
 package org.bukkit.inventory.meta;
 
 import org.bukkit.Material;
+import org.bukkit.Skin;
+
+import java.util.UUID;
 
 /**
  * Represents a skull ({@link Material#SKULL_ITEM}) that can have an owner.
@@ -33,4 +36,19 @@ public interface SkullMeta extends ItemMeta {
     boolean setOwner(String owner);
 
     SkullMeta clone();
+
+    /**
+     * Set the owner and appearance of this skull. A skull with this data set
+     * does not need to fetch anything remotely.
+     *
+     * @param name Username of the skull's owner, can be null (appears in item tooltip)
+     * @param uuid UUID of the skull's owner
+     * @param skin Skull owner's skin
+     */
+    void setOwner(String name, UUID uuid, Skin skin);
+
+    /**
+     * Clear any owner data in this skull
+     */
+    void clearOwner();
 }
-- 
2.19.0

