From cff260fcd61297cca89ead53ef83ddfd30dc114a Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Sun, 12 Aug 2018 22:32:33 -0700
Subject: [PATCH] Add PlayerUseUnkownEntityEvent


diff --git a/src/main/java/org/github/paperspigot/event/player/PlayerUseUnknownEntityEvent.java b/src/main/java/org/github/paperspigot/event/player/PlayerUseUnknownEntityEvent.java
new file mode 100644
index 00000000..9b1047b5
--- /dev/null
+++ b/src/main/java/org/github/paperspigot/event/player/PlayerUseUnknownEntityEvent.java
@@ -0,0 +1,35 @@
+package org.github.paperspigot.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+
+public class PlayerUseUnknownEntityEvent extends PlayerEvent {
+
+    private static final HandlerList handlers = new HandlerList();
+    private final int entityId;
+    private final boolean attack;
+
+    public PlayerUseUnknownEntityEvent(Player who, int entityId, boolean attack) {
+        super(who);
+        this.entityId = entityId;
+        this.attack = attack;
+    }
+
+    public int getEntityId() {
+        return this.entityId;
+    }
+
+    public boolean isAttack() {
+        return this.attack;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
-- 
2.18.0

