From 3f06997a3fa91018ed159485680fc7dd1fbdd039 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Tue, 14 Aug 2018 15:39:13 -0700
Subject: [PATCH] Add EntityZapEvent


diff --git a/src/main/java/org/bukkit/event/entity/PigZapEvent.java b/src/main/java/org/bukkit/event/entity/PigZapEvent.java
index 9b495562..7a1337b1 100644
--- a/src/main/java/org/bukkit/event/entity/PigZapEvent.java
+++ b/src/main/java/org/bukkit/event/entity/PigZapEvent.java
@@ -6,18 +6,18 @@ import org.bukkit.entity.PigZombie;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.EntityAction;
 import org.bukkit.event.HandlerList;
+import org.github.paperspigot.event.entity.EntityZapEvent;
 
 /**
  * Stores data for pigs being zapped
  */
-public class PigZapEvent extends EntityActionBase implements Cancellable, EntityAction {
-    private static final HandlerList handlers = new HandlerList();
+public class PigZapEvent extends EntityZapEvent implements Cancellable, EntityAction {
     private boolean canceled;
     private final PigZombie pigzombie;
     private final LightningStrike bolt;
 
     public PigZapEvent(final Pig pig, final LightningStrike bolt, final PigZombie pigzombie) {
-        super(pig);
+        super(pig, bolt, pigzombie);
         this.bolt = bolt;
         this.pigzombie = pigzombie;
     }
@@ -58,13 +58,4 @@ public class PigZapEvent extends EntityActionBase implements Cancellable, Entity
     public PigZombie getPigZombie() {
         return pigzombie;
     }
-
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
 }
diff --git a/src/main/java/org/github/paperspigot/event/entity/EntityZapEvent.java b/src/main/java/org/github/paperspigot/event/entity/EntityZapEvent.java
new file mode 100644
index 00000000..247430f2
--- /dev/null
+++ b/src/main/java/org/github/paperspigot/event/entity/EntityZapEvent.java
@@ -0,0 +1,64 @@
+package org.github.paperspigot.event.entity;
+
+import org.apache.commons.lang.Validate;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LightningStrike;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityActionBase;
+
+/**
+ *  Fired when lightning strikes an entity
+ */
+public class EntityZapEvent extends EntityActionBase implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+    private boolean cancelled;
+    private final LightningStrike bolt;
+    private final Entity replacementEntity;
+
+    public EntityZapEvent(final Entity entity, final LightningStrike bolt, final Entity replacementEntity) {
+        super(entity);
+        Validate.notNull(bolt);
+        Validate.notNull(replacementEntity);
+        this.bolt = bolt;
+        this.replacementEntity = replacementEntity;
+    }
+
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+
+    /**
+     * Gets the lightning bolt that is striking the entity.
+     * @return The lightning bolt responsible for this event
+     */
+    public LightningStrike getBolt() {
+        return bolt;
+    }
+
+    /**
+     * Gets the entity that will replace the struck entity.
+     * @return The entity that will replace the struck entity
+     */
+    public Entity getReplacementEntity() {
+        return replacementEntity;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @Override
+    public LightningStrike getActor() {
+        return getBolt();
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
-- 
2.18.0

