From 3d60db2c6cb016e7fb5e811342b5c955466b0148 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Tue, 14 Aug 2018 23:33:36 -0700
Subject: [PATCH] Add API methods to control if armour stands can move


diff --git a/src/main/java/org/bukkit/entity/ArmorStand.java b/src/main/java/org/bukkit/entity/ArmorStand.java
index ad77d590..7693575e 100644
--- a/src/main/java/org/bukkit/entity/ArmorStand.java
+++ b/src/main/java/org/bukkit/entity/ArmorStand.java
@@ -274,4 +274,22 @@ public interface ArmorStand extends LivingEntity {
      * @param marker whether this is a marker
      */
     void setMarker(boolean marker);
+
+    // Paper start
+    /**
+     * Tests if this armor stand can move.
+     *
+     * <p>The default value is {@code true}.</p>
+     *
+     * @return {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    boolean canMove();
+
+    /**
+     * Sets if this armor stand can move.
+     *
+     * @param move {@code true} if this armour stand can move, {@code false} otherwise
+     */
+    void setCanMove(boolean move);
+    // Paper end
 }
-- 
2.18.0

