From e05d8e6c07fce630a822de9a111acf130cdc8d83 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Fri, 17 Aug 2018 23:30:52 -0700
Subject: [PATCH] Add ProtocolVersion API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 3cd5a669..c6d10438 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1397,6 +1397,8 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
      */
     RayBlockIntersection getTargetedBlock(boolean nonSolids, boolean liquids);
 
+    public int getProtocolVersion();
+
     // Spigot start
     public class Spigot extends Entity.Spigot
     {
-- 
2.18.0

