From 7d5357d224bc5dd2c563ef529a666266c876fc5b Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Mon, 27 Aug 2018 02:30:09 -0700
Subject: [PATCH] Per World Custom Crafting API


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 53484fd2..2f1ed771 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -530,7 +530,7 @@ public interface Server extends PluginMessageRecipient {
     public void configureDbConfig(ServerConfig config);
 
     /**
-     * Adds a recipe to the crafting manager.
+     * Adds a recipe to all world crafting managers.
      *
      * @param recipe the recipe to add
      * @return true if the recipe was added, false if it wasn't for some
@@ -539,7 +539,7 @@ public interface Server extends PluginMessageRecipient {
     public boolean addRecipe(Recipe recipe);
 
     /**
-     * Get a list of all recipes for a given item. The stack size is ignored
+     * Get a list of all default recipes for a given item. The stack size is ignored
      * in comparisons. If the durability is -1, it will match any data value.
      *
      * @param result the item to match against recipe results
@@ -548,19 +548,19 @@ public interface Server extends PluginMessageRecipient {
     public List<Recipe> getRecipesFor(ItemStack result);
 
     /**
-     * Get an iterator through the list of crafting recipes.
+     * Get an iterator through the list of default crafting recipes.
      *
      * @return an iterator
      */
     public Iterator<Recipe> recipeIterator();
 
     /**
-     * Clears the list of crafting recipes.
+     * Clears the list of crafting recipes in all worlds.
      */
     public void clearRecipes();
 
     /**
-     * Resets the list of crafting recipes to the default.
+     * Resets the list of crafting recipes in all worlds.
      */
     public void resetRecipes();
 
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 9e9d2d27..49d913a9 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -6,6 +6,7 @@ import org.bukkit.generator.ChunkGenerator;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
@@ -16,6 +17,7 @@ import org.bukkit.block.Block;
 import org.bukkit.entity.*;
 import org.bukkit.generator.BlockPopulator;
 import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.Recipe;
 import org.bukkit.metadata.Metadatable;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
 import org.bukkit.util.RayBlockIntersection;
@@ -1609,6 +1611,41 @@ public interface World extends PluginMessageRecipient, Metadatable, Physical {
      */
     public boolean isGameRule(String rule);
 
+    /**
+     * Adds a recipe to the crafting manager.
+     *
+     * @param recipe the recipe to add
+     * @return true if the recipe was added, false if it wasn't for some
+     *     reason
+     */
+    public boolean addRecipe(Recipe recipe);
+
+    /**
+     * Get a list of all recipes for a given item. The stack size is ignored
+     * in comparisons. If the durability is -1, it will match any data value.
+     *
+     * @param result the item to match against recipe results
+     * @return a list of recipes with the given result
+     */
+    public List<Recipe> getRecipesFor(ItemStack result);
+
+    /**
+     * Get an iterator through the list of crafting recipes.
+     *
+     * @return an iterator
+     */
+    public Iterator<Recipe> recipeIterator();
+
+    /**
+     * Clears the list of crafting recipes.
+     */
+    public void clearRecipes();
+
+    /**
+     * Resets the list of crafting recipes to the default.
+     */
+    public void resetRecipes();
+
     // Spigot start
     public class Spigot
     {
-- 
2.18.0

