From 20b51fb46aea7b502ef6efa2537b87c344e0c5d0 Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Fri, 4 Jan 2019 18:05:36 -0600
Subject: [PATCH] Add command unregistration API


diff --git a/src/main/java/org/bukkit/command/CommandMap.java b/src/main/java/org/bukkit/command/CommandMap.java
index e7e20d89..168d6a90 100644
--- a/src/main/java/org/bukkit/command/CommandMap.java
+++ b/src/main/java/org/bukkit/command/CommandMap.java
@@ -1,6 +1,7 @@
 package org.bukkit.command;
 
 import java.util.List;
+import java.util.function.Predicate;
 
 public interface CommandMap {
 
@@ -65,6 +66,21 @@ public interface CommandMap {
      */
     public boolean register(String fallbackPrefix, Command command);
 
+    /**
+     * Unregister a specific command.
+     *
+     * @param command to unregister
+     * @return if the command was actually registered in the first place
+     */
+    boolean unregister(Command command);
+
+    /**
+     * Unregister all commands which match the supplied filter.
+     *
+     * @param filter to check commands against for un-registration
+     */
+    void unregisterAll(Predicate<? super Command> filter);
+
     /**
      * Looks for the requested command and executes it if found.
      *
diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 10d07114..af311aaa 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -9,6 +9,7 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Predicate;
 import java.util.regex.Pattern;
 
 import org.apache.commons.lang.Validate;
@@ -124,6 +125,23 @@ public class SimpleCommandMap implements CommandMap {
         return registered;
     }
 
+    @Override
+    public boolean unregister(Command command) {
+       command.unregister(this);
+       return knownCommands.values().removeIf(command::equals);
+    }
+
+    @Override
+    public void unregisterAll(Predicate<? super Command> filter) {
+        knownCommands.values().removeIf(command -> {
+            if(filter.test(command)) {
+                command.unregister(this);
+                return true;
+            }
+            return false;
+        });
+    }
+
     /**
      * {@inheritDoc}
      */
-- 
2.19.0

