From cb79efa591755aceef6a39feacdabfd705d593ee Mon Sep 17 00:00:00 2001
From: Ashcon Partovi <ashcon@partovi.net>
Date: Tue, 30 Apr 2019 08:03:51 -0700
Subject: [PATCH] Mutable ChunkSnapshot API


diff --git a/src/main/java/org/bukkit/ChunkSnapshot.java b/src/main/java/org/bukkit/ChunkSnapshot.java
index 83fccc85..37349b39 100644
--- a/src/main/java/org/bukkit/ChunkSnapshot.java
+++ b/src/main/java/org/bukkit/ChunkSnapshot.java
@@ -1,6 +1,9 @@
 package org.bukkit;
 
 import org.bukkit.block.Biome;
+import org.bukkit.block.BlockState;
+import org.bukkit.material.MaterialData;
+import org.bukkit.util.Vector;
 
 /**
  * Represents a static, thread-safe snapshot of chunk of blocks.
@@ -55,6 +58,16 @@ public interface ChunkSnapshot {
     @Deprecated
     int getBlockData(int x, int y, int z);
 
+    /**
+     * Get the material of the block at the given position in the chunk
+     */
+    MaterialData getMaterialData(int x, int y, int z);
+
+    /**
+     * Get the material of the block at the given position in the chunk
+     */
+    MaterialData getMaterialData(Vector pos);
+
     /**
      * Get sky light level for block at corresponding coordinate in the chunk
      *
@@ -126,4 +139,38 @@ public interface ChunkSnapshot {
      * @return true if empty, false if not
      */
     boolean isSectionEmpty(int sy);
+
+    /**
+      * Set the block type at the given position in the chunk.
+      * Note that mutating the snapshot will compromise its thread safety.
+      */
+    void setBlockTypeId(int x, int y, int z, int typeId);
+
+    /**
+     * Set the block data at the given position in the chunk.
+     * Note that mutating the snapshot will compromise its thread safety.
+     */
+    void setBlockData(int x, int y, int z, int data);
+
+    /**
+     * Set the block material (type and data) at the given position in the chunk.
+     * Note that mutating the snapshot will compromise its thread safety.
+     */
+    void setMaterialData(int x, int y, int z, MaterialData material);
+
+    /**
+     * Set the block material (type and data) at the given position in the chunk.
+     * Note that mutating the snapshot will compromise its thread safety.
+     */
+    void setMaterialData(Vector pos, MaterialData material);
+
+    /**
+     * Update the material (type and data) of the given block in the snapshot to
+     * match the given state. The chunk coordinates saved in the snapshot are used
+     * to transform the given block's position into local coordinates. If the block
+     * is not inside this chunk, nothing happens.
+     *
+     * Note that mutating the snapshot will compromise its thread safety.
+     */
+    void updateBlock(BlockState state);
 }
-- 
2.20.1

