--- ../src-base/minecraft/org/bukkit/plugin/Plugin.java
+++ ../src-work/minecraft/org/bukkit/plugin/Plugin.java
@@ -1,16 +1,14 @@
 package org.bukkit.plugin;
 
+import com.avaje.ebean.EbeanServer;
 import java.io.File;
 import java.io.InputStream;
 import java.util.logging.Logger;
-
 import org.bukkit.Server;
 import org.bukkit.command.TabExecutor;
 import org.bukkit.configuration.file.FileConfiguration;
 import org.bukkit.generator.ChunkGenerator;
 
-import com.avaje.ebean.EbeanServer;
-
 /**
  * Represents a Plugin
  * <p>
