--- a/net/minecraft/util/text/LanguageMap.java
+++ b/net/minecraft/util/text/LanguageMap.java
@@ -7,6 +_,7 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import com.mohistmc.util.i18n.i18n;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
@@ -37,10 +_,11 @@
       try (InputStream inputstream = LanguageMap.class.getResourceAsStream("/assets/minecraft/lang/en_us.json")) {
          func_240593_a_(inputstream, biconsumer);
       } catch (JsonParseException | IOException ioexception) {
-         field_201045_a.error("Couldn't read strings from /assets/minecraft/lang/en_us.json", (Throwable)ioexception);
+         field_201045_a.error(i18n.get("languagemap.1", (Throwable)ioexception));
       }
 
-      final Map<String, String> map = builder.build();
+      final Map<String, String> map = new java.util.HashMap<>(builder.build());
+      net.minecraftforge.fml.server.LanguageHook.captureLanguageMap(map);
       return new LanguageMap() {
          public String func_230503_a_(String p_230503_1_) {
             return map.getOrDefault(p_230503_1_, p_230503_1_);
@@ -63,6 +_,11 @@
                }, Style.field_240709_b_).isPresent();
             };
          }
+         
+         @Override
+         public Map<String, String> getLanguageData() {
+            return map;
+         }
       };
    }
 
@@ -84,6 +_,9 @@
    public static void func_240594_a_(LanguageMap p_240594_0_) {
       field_240592_d_ = p_240594_0_;
    }
+   
+   // FORGE START
+   public Map<String, String> getLanguageData() { return ImmutableMap.of(); }
 
    public abstract String func_230503_a_(String p_230503_1_);
 
