--- a/net/minecraft/network/play/server/SChunkDataPacket.java
+++ b/net/minecraft/network/play/server/SChunkDataPacket.java
@@ -130,6 +_,7 @@
       return bytebuf;
    }
 
+   public int writeChunk(PacketBuffer packetDataSerializer, Chunk chunk, int chunkSectionSelector) { return this.func_218708_a(packetDataSerializer, chunk, chunkSectionSelector); } // Paper - OBFHELPER
    public int func_218708_a(PacketBuffer p_218708_1_, Chunk p_218708_2_, int p_218708_3_) {
       int i = 0;
       ChunkSection[] achunksection = p_218708_2_.func_76587_i();
