--- a/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
+++ b/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
@@ -1,6 +_,7 @@
 package net.minecraft.network.play.server;
 
 import java.io.IOException;
+
 import net.minecraft.client.network.play.IClientPlayNetHandler;
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
@@ -9,30 +_,41 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class SPlayerListHeaderFooterPacket implements IPacket<IClientPlayNetHandler> {
-   private ITextComponent field_179703_a;
-   private ITextComponent field_179702_b;
-
-   public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
-      this.field_179703_a = p_148837_1_.func_179258_d();
-      this.field_179702_b = p_148837_1_.func_179258_d();
-   }
-
-   public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
-      p_148840_1_.func_179256_a(this.field_179703_a);
-      p_148840_1_.func_179256_a(this.field_179702_b);
-   }
-
-   public void func_148833_a(IClientPlayNetHandler p_148833_1_) {
-      p_148833_1_.func_175096_a(this);
-   }
-
-   @OnlyIn(Dist.CLIENT)
-   public ITextComponent func_179700_a() {
-      return this.field_179703_a;
-   }
-
-   @OnlyIn(Dist.CLIENT)
-   public ITextComponent func_179701_b() {
-      return this.field_179702_b;
-   }
+    public ITextComponent field_179703_a;
+    public ITextComponent field_179702_b;
+    // Paper start
+    public net.kyori.adventure.text.Component adventure$header;
+    public net.kyori.adventure.text.Component adventure$footer;
+    // Paper end
+
+    public void func_148837_a(PacketBuffer p_148837_1_) throws IOException {
+        this.field_179703_a = p_148837_1_.func_179258_d();
+        this.field_179702_b = p_148837_1_.func_179258_d();
+    }
+
+    public void func_148840_b(PacketBuffer p_148840_1_) throws IOException {
+        // Paper start
+        if (this.adventure$header != null && this.adventure$footer != null) {
+            p_148840_1_.writeComponent(this.adventure$header);
+            p_148840_1_.writeComponent(this.adventure$footer);
+            return;
+        }
+        // Paper end
+        p_148840_1_.func_179256_a(this.field_179703_a);
+        p_148840_1_.func_179256_a(this.field_179702_b);
+    }
+
+    public void func_148833_a(IClientPlayNetHandler p_148833_1_) {
+        p_148833_1_.func_175096_a(this);
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    public ITextComponent func_179700_a() {
+        return this.field_179703_a;
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    public ITextComponent func_179701_b() {
+        return this.field_179702_b;
+    }
 }
