--- a/net/minecraft/network/play/server/SSpawnPositionPacket.java
+++ b/net/minecraft/network/play/server/SSpawnPositionPacket.java
@@ -9,7 +9,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class SSpawnPositionPacket implements IPacket<IClientPlayNetHandler> {
-   private BlockPos field_179801_a;
+   public BlockPos field_179801_a;
 
    public SSpawnPositionPacket() {
    }
@@ -30,7 +30,7 @@
       p_148833_1_.func_147271_a(this);
    }
 
-   @OnlyIn(Dist.CLIENT)
+   //@OnlyIn(Dist.CLIENT)
    public BlockPos func_179800_a() {
       return this.field_179801_a;
    }
