--- ../src-base/minecraft/net/minecraft/profiler/Profiler.java
+++ ../src-work/minecraft/net/minecraft/profiler/Profiler.java
@@ -13,6 +13,7 @@
 
 public class Profiler
 {
+    public static final boolean ENABLED = Boolean.getBoolean("enableDebugMethodProfiler"); // CraftBukkit - disable unless specified in JVM arguments
     private static final Logger LOGGER = LogManager.getLogger();
     private final List<String> sectionList = Lists.<String>newArrayList();
     private final List<Long> timestampList = Lists.<Long>newArrayList();
@@ -22,6 +23,7 @@
 
     public void clearProfiling()
     {
+        if (!ENABLED) return;
         this.profilingMap.clear();
         this.profilingSection = "";
         this.sectionList.clear();
@@ -29,6 +31,7 @@
 
     public void startSection(String name)
     {
+        if (!ENABLED) return;
         if (this.profilingEnabled)
         {
             if (!this.profilingSection.isEmpty())
@@ -44,6 +47,7 @@
 
     public void func_194340_a(Supplier<String> p_194340_1_)
     {
+        if (!ENABLED) return;
         if (this.profilingEnabled)
         {
             this.startSection(p_194340_1_.get());
@@ -52,6 +56,7 @@
 
     public void endSection()
     {
+        if (!ENABLED) return;
         if (this.profilingEnabled)
         {
             long i = System.nanoTime();
@@ -79,7 +84,7 @@
 
     public List<Profiler.Result> getProfilingData(String profilerName)
     {
-        if (!this.profilingEnabled)
+        if (!ENABLED || !this.profilingEnabled)
         {
             return Collections.<Profiler.Result>emptyList();
         }
@@ -146,12 +151,14 @@
 
     public void endStartSection(String name)
     {
+        if (!ENABLED) return;
         this.endSection();
         this.startSection(name);
     }
 
     public String getNameOfLastSection()
     {
+        if (!ENABLED) return "[DISABLED]";
         return this.sectionList.isEmpty() ? "[UNKNOWN]" : (String)this.sectionList.get(this.sectionList.size() - 1);
     }
 
