--- a/net/minecraft/util/math/BlockPos.java
+++ b/net/minecraft/util/math/BlockPos.java
@@ -10,6 +_,7 @@
 import java.util.stream.StreamSupport;
 import javax.annotation.concurrent.Immutable;
 import net.minecraft.dispenser.IPosition;
+import net.minecraft.entity.Entity;
 import net.minecraft.util.AxisRotation;
 import net.minecraft.util.Direction;
 import net.minecraft.util.Rotation;
@@ -40,6 +_,11 @@
    private static final int field_218292_j = field_177989_d;
    private static final int field_218293_k = field_177989_d + field_177991_c;
 
+   // CraftBukkit
+   public BlockPos(Entity p_i46032_1_) {
+      this(p_i46032_1_.func_226277_ct_(), p_i46032_1_.func_226278_cu_(), p_i46032_1_.func_226281_cx_());
+   }
+
    public BlockPos(int p_i46030_1_, int p_i46030_2_, int p_i46030_3_) {
       super(p_i46030_1_, p_i46030_2_, p_i46030_3_);
    }
@@ -107,6 +_,12 @@
       return p_177982_1_ == 0 && p_177982_2_ == 0 && p_177982_3_ == 0 ? this : new BlockPos(this.func_177958_n() + p_177982_1_, this.func_177956_o() + p_177982_2_, this.func_177952_p() + p_177982_3_);
    }
 
+   // Mohist start: MNS compatibility
+   public BlockPos add(Vector3i p_177971_1_) {
+      return this.func_177971_a(p_177971_1_);
+   }
+   // Mohist end
+
    public BlockPos func_177971_a(Vector3i p_177971_1_) {
       return this.func_177982_a(p_177971_1_.func_177958_n(), p_177971_1_.func_177956_o(), p_177971_1_.func_177952_p());
    }
@@ -115,6 +_,24 @@
       return this.func_177982_a(-p_177973_1_.func_177958_n(), -p_177973_1_.func_177956_o(), -p_177973_1_.func_177952_p());
    }
 
+   // Mohist start: MNS compatibility
+   public BlockPos up() {
+      return func_177984_a();
+   }
+
+   public BlockPos up(int p_177981_1_) {
+      return func_177981_b(p_177981_1_);
+   }
+
+   public BlockPos down() {
+      return func_177977_b();
+   }
+
+   public BlockPos down(int p_177979_1_) {
+      return func_177979_c(p_177979_1_);
+   }
+   // Mohist end
+
    public BlockPos func_177984_a() {
       return this.func_177972_a(Direction.UP);
    }
@@ -162,6 +_,12 @@
    public BlockPos func_177965_g(int p_177965_1_) {
       return this.func_177967_a(Direction.EAST, p_177965_1_);
    }
+
+   // Mohist start: MNS compatibility
+   public BlockPos shift(Direction p_177972_1_) {
+      return func_177972_a(p_177972_1_);
+   }
+   //Mohist end
 
    public BlockPos func_177972_a(Direction p_177972_1_) {
       return new BlockPos(this.func_177958_n() + p_177972_1_.func_82601_c(), this.func_177956_o() + p_177972_1_.func_96559_d(), this.func_177952_p() + p_177972_1_.func_82599_e());
