--- a/net/minecraft/world/level/chunk/ChunkStatus.java
+++ b/net/minecraft/world/level/chunk/ChunkStatus.java
@@ -38,7 +_,7 @@
    public static final ChunkStatus f_62314_ = m_62414_("empty", (ChunkStatus)null, -1, f_62327_, ChunkStatus.ChunkType.PROTOCHUNK, (p_156307_, p_156308_, p_156309_, p_156310_, p_156311_) -> {
    });
    public static final ChunkStatus f_62315_ = m_280108_("structure_starts", f_62314_, 0, false, f_62327_, ChunkStatus.ChunkType.PROTOCHUNK, (p_289514_, p_289515_, p_289516_, p_289517_, p_289518_, p_289519_, p_289520_, p_289521_, p_289522_) -> {
-      if (p_289516_.m_7654_().m_129910_().m_246337_().m_247749_()) {
+      if (p_289516_.K.m_246337_().m_247749_()) { // CraftBukkit
          p_289517_.m_255037_(p_289516_.m_9598_(), p_289516_.m_7726_().m_255415_(), p_289516_.m_215010_(), p_289522_, p_289518_);
       }
 
@@ -192,7 +_,7 @@
       return f_62331_.getInt(p_62371_.m_62445_());
    }
 
-   ChunkStatus(@Nullable ChunkStatus p_289640_, int p_289655_, boolean p_289657_, EnumSet<Heightmap.Types> p_289662_, ChunkStatus.ChunkType p_289652_, ChunkStatus.GenerationTask p_289679_, ChunkStatus.LoadingTask p_289646_) {
+   public ChunkStatus(@Nullable ChunkStatus p_289640_, int p_289655_, boolean p_289657_, EnumSet<Heightmap.Types> p_289662_, ChunkStatus.ChunkType p_289652_, ChunkStatus.GenerationTask p_289679_, ChunkStatus.LoadingTask p_289646_) {
       this.f_62334_ = p_289640_ == null ? this : p_289640_;
       this.f_62335_ = p_289679_;
       this.f_62336_ = p_289646_;
