--- a/net/minecraft/world/level/chunk/EmptyLevelChunk.java
+++ b/net/minecraft/world/level/chunk/EmptyLevelChunk.java
@@ -1,6 +_,9 @@
 package net.minecraft.world.level.chunk;
 
 import javax.annotation.Nullable;
+
+import ca.spottedleaf.starlight.common.light.SWMRNibbleArray;
+import ca.spottedleaf.starlight.common.light.StarLightEngine;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.server.level.FullChunkStatus;
@@ -67,4 +_,28 @@
    public Holder<Biome> m_203495_(int p_204426_, int p_204427_, int p_204428_) {
       return this.f_204420_;
    }
+
+   public SWMRNibbleArray[] getBlockNibbles() {
+      return StarLightEngine.getFilledEmptyLight(this.m_62953_());
+   }
+
+   public void setBlockNibbles(final SWMRNibbleArray[] nibbles) {}
+
+   public SWMRNibbleArray[] getSkyNibbles() {
+      return StarLightEngine.getFilledEmptyLight(this.m_62953_());
+   }
+
+   public void setSkyNibbles(final SWMRNibbleArray[] nibbles) {}
+
+   public boolean[] getSkyEmptinessMap() {
+      return null;
+   }
+
+   public void setSkyEmptinessMap(final boolean[] emptinessMap) {}
+
+   public boolean[] getBlockEmptinessMap() {
+      return null;
+   }
+
+   public void setBlockEmptinessMap(final boolean[] emptinessMap) {}
 }
