--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -2,6 +_,7 @@
 
 import java.util.function.Predicate;
 import net.minecraft.core.Holder;
+import net.minecraft.core.QuartPos;
 import net.minecraft.core.Registry;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.world.level.biome.Biome;
@@ -30,6 +_,12 @@
       this.m_63018_();
    }
 
+   public LevelChunkSection(PalettedContainer<BlockState> pStates, PalettedContainer<Holder<Biome>> pBiomes) {
+      this.f_62972_ = pStates;
+      this.f_187995_ = pBiomes;
+      this.m_63018_();
+   }
+
    public LevelChunkSection(Registry<Biome> p_282873_) {
       this.f_62972_ = new PalettedContainer<>(Block.f_49791_, Blocks.f_50016_.m_49966_(), PalettedContainer.Strategy.f_188137_);
       this.f_187995_ = new PalettedContainer<>(p_282873_.m_206115_(), p_282873_.m_246971_(Biomes.f_48202_), PalettedContainer.Strategy.f_188138_);
@@ -192,4 +_,10 @@
 
       this.f_187995_ = palettedcontainer;
    }
+
+   // CraftBukkit start
+   public void setBiome(int i, int j, int k, Holder<Biome> biome) {
+      ((PalettedContainer<Holder<Biome>>) this.f_187995_).m_156470_(i, j, k, biome);
+   }
+   // CraftBukkit end
 }
