--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -30,6 +_,12 @@
         this.recalcBlockCounts();
     }
 
+   public LevelChunkSection(PalettedContainer<BlockState> pStates, PalettedContainer<Holder<Biome>> pBiomes) {
+      this.states = pStates;
+      this.biomes = pBiomes;
+      this.recalcBlockCounts();
+   }
+
     public LevelChunkSection(Registry<Biome> p_282873_) {
         this.states = new PalettedContainer<>(Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES);
         this.biomes = new PalettedContainer<>(p_282873_.asHolderIdMap(), p_282873_.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES);
@@ -191,4 +_,10 @@
 
         this.biomes = palettedcontainer;
     }
+
+   // CraftBukkit start
+   public void setBiome(int i, int j, int k, Holder<Biome> biome) {
+      ((PalettedContainer<Holder<Biome>>) this.biomes).set(i, j, k, biome);
+   }
+   // CraftBukkit end
 }
