--- a/net/minecraft/entity/passive/AnimalEntity.java
+++ b/net/minecraft/entity/passive/AnimalEntity.java
@@ -23,8 +23,9 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public abstract class AnimalEntity extends AgeableEntity {
-   private int field_70881_d;
-   private UUID field_146084_br;
+   public int field_70881_d; // private->public CraftBukkit
+   public UUID field_146084_br; // private->public CraftBukkit
+   public ItemStack breedItem; // CraftBukkit - Add breedItem variable
 
    protected AnimalEntity(EntityType<? extends AnimalEntity> p_i48568_1_, World p_i48568_2_) {
       super(p_i48568_1_, p_i48568_2_);
@@ -56,14 +57,18 @@
 
    }
 
-   public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_) {
-      if (this.func_180431_b(p_70097_1_)) {
+   /* CraftBukkit start
+   // Function disabled as it has no special function anymore after
+   // setSitting is disabled.
+   public boolean attackEntityFrom(DamageSource source, float amount) {
+      if (this.isInvulnerableTo(source)) {
          return false;
       } else {
-         this.field_70881_d = 0;
-         return super.func_70097_a(p_70097_1_, p_70097_2_);
+         this.inLove = 0;
+         return super.attackEntityFrom(source, amount);
       }
    }
+   // CraftBukkit end */
 
    public float func_205022_a(BlockPos p_205022_1_, IWorldReader p_205022_2_) {
       return p_205022_2_.func_180495_p(p_205022_1_.func_177977_b()).func_177230_c() == Blocks.field_196658_i ? 10.0F : p_205022_2_.func_205052_D(p_205022_1_) - 0.5F;
@@ -144,6 +149,7 @@
       if (p_146082_1_ != null) {
          this.field_146084_br = p_146082_1_.func_110124_au();
       }
+      this.breedItem = p_146082_1_.field_71071_by.func_70448_g(); // CraftBukkit
 
       this.field_70170_p.func_72960_a(this, (byte)18);
    }
