--- ../src-base/minecraft/net/minecraft/entity/passive/EntitySquid.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntitySquid.java
@@ -38,7 +38,7 @@
     {
         super(p_i1693_1_);
         this.func_70105_a(0.8F, 0.8F);
-        this.field_70146_Z.setSeed((long)(1 + this.func_145782_y()));
+        this.field_70146_Z.setSeed((long)(1 + this.func_145782_y())); // Paper
         this.field_70864_bA = 1.0F / (this.field_70146_Z.nextFloat() + 1.0F) * 0.2F;
     }
 
@@ -191,7 +191,7 @@
 
     public boolean func_70601_bi()
     {
-        return this.field_70163_u > 45.0D && this.field_70163_u < (double)this.field_70170_p.func_181545_F() && super.func_70601_bi();
+        return this.field_70163_u > this.field_70170_p.spigotConfig.squidSpawnRangeMin && this.field_70163_u < (double)this.field_70170_p.func_181545_F() && super.func_70601_bi();
     }
 
     @SideOnly(Side.CLIENT)
