--- a/net/minecraft/entity/passive/ParrotEntity.java
+++ b/net/minecraft/entity/passive/ParrotEntity.java
@@ -232,7 +232,7 @@
          }
 
          if (!this.field_70170_p.field_72995_K) {
-            if (this.field_70146_Z.nextInt(10) == 0) {
+            if (this.field_70146_Z.nextInt(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_184645_1_) && !org.bukkit.craftbukkit.v1_15_R1.event.CraftEventFactory.callEntityTameEvent(this, p_184645_1_).isCancelled()) { // CraftBukkit
                this.func_193101_c(p_184645_1_);
                this.field_70170_p.func_72960_a(this, (byte)7);
             } else {
@@ -246,7 +246,7 @@
             itemstack.func_190918_g(1);
          }
 
-         this.func_195064_c(new EffectInstance(Effects.field_76436_u, 900));
+         this.addEffect(new EffectInstance(Effects.field_76436_u, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
          if (p_184645_1_.func_184812_l_() || !this.func_190530_aW()) {
             this.func_70097_a(DamageSource.func_76365_a(p_184645_1_), Float.MAX_VALUE);
          }
@@ -365,7 +365,8 @@
          return false;
       } else {
          if (this.field_70911_d != null) {
-            this.field_70911_d.func_75270_a(false);
+            // CraftBukkit - moved into EntityLiving.d(DamageSource, float)
+            // this.sitGoal.setSitting(false);
          }
 
          return super.func_70097_a(p_70097_1_, p_70097_2_);
