--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -341,19 +_,19 @@
         return this.renderState;
     }
 
-    private void setTransformationInterpolationDuration(int p_297488_) {
+    public void setTransformationInterpolationDuration(int p_297488_) {
         this.entityData.set(DATA_TRANSFORMATION_INTERPOLATION_DURATION_ID, p_297488_);
     }
 
-    private int getTransformationInterpolationDuration() {
+    public int getTransformationInterpolationDuration() {
         return this.entityData.get(DATA_TRANSFORMATION_INTERPOLATION_DURATION_ID);
     }
 
-    private void setTransformationInterpolationDelay(int p_300640_) {
+    public void setTransformationInterpolationDelay(int p_300640_) {
         this.entityData.set(DATA_TRANSFORMATION_INTERPOLATION_START_DELTA_TICKS_ID, p_300640_, true);
     }
 
-    private int getTransformationInterpolationDelay() {
+    public int getTransformationInterpolationDelay() {
         return this.entityData.get(DATA_TRANSFORMATION_INTERPOLATION_START_DELTA_TICKS_ID);
     }
 
@@ -369,7 +_,7 @@
         this.entityData.set(DATA_BILLBOARD_RENDER_CONSTRAINTS_ID, p_270345_.getId());
     }
 
-    private Display.BillboardConstraints getBillboardConstraints() {
+    public Display.BillboardConstraints getBillboardConstraints() {
         return Display.BillboardConstraints.BY_ID.apply(this.entityData.get(DATA_BILLBOARD_RENDER_CONSTRAINTS_ID));
     }
 
@@ -557,11 +_,11 @@
             }
         }
 
-        private BlockState getBlockState() {
+        public BlockState getBlockState() {
             return this.entityData.get(DATA_BLOCK_STATE_ID);
         }
 
-        private void setBlockState(BlockState p_270267_) {
+        public void setBlockState(BlockState p_270267_) {
             this.entityData.set(DATA_BLOCK_STATE_ID, p_270267_);
         }
 
@@ -653,7 +_,7 @@
             }
         }
 
-        private ItemStack getItemStack() {
+        public ItemStack getItemStack() {
             return this.entityData.get(DATA_ITEM_STACK_ID);
         }
 
@@ -665,7 +_,7 @@
             this.entityData.set(DATA_ITEM_DISPLAY_ID, p_270370_.getId());
         }
 
-        private ItemDisplayContext getItemTransform() {
+        public ItemDisplayContext getItemTransform() {
             return ItemDisplayContext.BY_ID.apply(this.entityData.get(DATA_ITEM_DISPLAY_ID));
         }
 
@@ -785,8 +_,8 @@
         private static final byte INITIAL_TEXT_OPACITY = -1;
         public static final int INITIAL_BACKGROUND = 1073741824;
         private static final EntityDataAccessor<Component> DATA_TEXT_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.COMPONENT);
-        private static final EntityDataAccessor<Integer> DATA_LINE_WIDTH_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
-        private static final EntityDataAccessor<Integer> DATA_BACKGROUND_COLOR_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
+        public static final EntityDataAccessor<Integer> DATA_LINE_WIDTH_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
+        public static final EntityDataAccessor<Integer> DATA_BACKGROUND_COLOR_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
         private static final EntityDataAccessor<Byte> DATA_TEXT_OPACITY_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.BYTE);
         private static final EntityDataAccessor<Byte> DATA_STYLE_FLAGS_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.BYTE);
         private static final IntSet TEXT_RENDER_STATE_IDS = IntSet.of(
@@ -819,7 +_,7 @@
             }
         }
 
-        private Component getText() {
+        public Component getText() {
             return this.entityData.get(DATA_TEXT_ID);
         }
 
@@ -827,11 +_,11 @@
             this.entityData.set(DATA_TEXT_ID, p_270902_);
         }
 
-        private int getLineWidth() {
+        public int getLineWidth() {
             return this.entityData.get(DATA_LINE_WIDTH_ID);
         }
 
-        public void setLineWidth(int p_270545_) {
+        private void setLineWidth(int p_270545_) {
             this.entityData.set(DATA_LINE_WIDTH_ID, p_270545_);
         }
 
@@ -847,11 +_,11 @@
             return this.entityData.get(DATA_BACKGROUND_COLOR_ID);
         }
 
-        public void setBackgroundColor(int p_270241_) {
+        private void setBackgroundColor(int p_270241_) {
             this.entityData.set(DATA_BACKGROUND_COLOR_ID, p_270241_);
         }
 
-        private byte getFlags() {
+        public byte getFlags() {
             return this.entityData.get(DATA_STYLE_FLAGS_ID);
         }
 
