--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -4,9 +_,12 @@
 import com.mojang.logging.LogUtils;
 import com.mojang.math.Transformation;
 import com.mojang.serialization.Codec;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
 import it.unimi.dsi.fastutil.ints.IntSet;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
+import java.util.Set;
 import java.util.function.IntFunction;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -17,6 +_,9 @@
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ClientGamePacketListener;
+import net.minecraft.network.protocol.game.ClientboundAddEntityPacket;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
@@ -320,19 +_,19 @@
       return this.renderState;
    }
 
-   private void setTransformationInterpolationDuration(int p_297488_) {
+   public void setTransformationInterpolationDuration(int p_297488_) {
       this.entityData.set(DATA_TRANSFORMATION_INTERPOLATION_DURATION_ID, p_297488_);
    }
 
-   private int getTransformationInterpolationDuration() {
+   public int getTransformationInterpolationDuration() {
       return this.entityData.get(DATA_TRANSFORMATION_INTERPOLATION_DURATION_ID);
    }
 
-   private void setTransformationInterpolationDelay(int p_300640_) {
+   public void setTransformationInterpolationDelay(int p_300640_) {
       this.entityData.set(DATA_TRANSFORMATION_INTERPOLATION_START_DELTA_TICKS_ID, p_300640_, true);
    }
 
-   private int getTransformationInterpolationDelay() {
+   public int getTransformationInterpolationDelay() {
       return this.entityData.get(DATA_TRANSFORMATION_INTERPOLATION_START_DELTA_TICKS_ID);
    }
 
@@ -348,7 +_,7 @@
       this.entityData.set(DATA_BILLBOARD_RENDER_CONSTRAINTS_ID, p_270345_.getId());
    }
 
-   private Display.BillboardConstraints getBillboardConstraints() {
+   public Display.BillboardConstraints getBillboardConstraints() {
       return Display.BillboardConstraints.BY_ID.apply(this.entityData.get(DATA_BILLBOARD_RENDER_CONSTRAINTS_ID));
    }
 
@@ -362,7 +_,7 @@
       return i != -1 ? Brightness.unpack(i) : null;
    }
 
-   private int getPackedBrightnessOverride() {
+   public int getPackedBrightnessOverride() {
       return this.entityData.get(DATA_BRIGHTNESS_OVERRIDE_ID);
    }
 
@@ -516,11 +_,11 @@
 
       }
 
-      private BlockState getBlockState() {
+      public BlockState getBlockState() {
          return this.entityData.get(DATA_BLOCK_STATE_ID);
       }
 
-      private void setBlockState(BlockState p_270267_) {
+      public void setBlockState(BlockState p_270267_) {
          this.entityData.set(DATA_BLOCK_STATE_ID, p_270267_);
       }
 
@@ -622,7 +_,7 @@
 
       }
 
-      ItemStack getItemStack() {
+      public ItemStack getItemStack() {
          return this.entityData.get(DATA_ITEM_STACK_ID);
       }
 
@@ -634,7 +_,7 @@
          this.entityData.set(DATA_ITEM_DISPLAY_ID, p_270370_.getId());
       }
 
-      private ItemDisplayContext getItemTransform() {
+      public ItemDisplayContext getItemTransform() {
          return ItemDisplayContext.BY_ID.apply(this.entityData.get(DATA_ITEM_DISPLAY_ID));
       }
 
@@ -733,8 +_,8 @@
       private static final byte INITIAL_TEXT_OPACITY = -1;
       public static final int INITIAL_BACKGROUND = 1073741824;
       private static final EntityDataAccessor<Component> DATA_TEXT_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.COMPONENT);
-      private static final EntityDataAccessor<Integer> DATA_LINE_WIDTH_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
-      private static final EntityDataAccessor<Integer> DATA_BACKGROUND_COLOR_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
+      public static final EntityDataAccessor<Integer> DATA_LINE_WIDTH_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
+      public static final EntityDataAccessor<Integer> DATA_BACKGROUND_COLOR_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.INT);
       private static final EntityDataAccessor<Byte> DATA_TEXT_OPACITY_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.BYTE);
       private static final EntityDataAccessor<Byte> DATA_STYLE_FLAGS_ID = SynchedEntityData.defineId(Display.TextDisplay.class, EntityDataSerializers.BYTE);
       private static final IntSet TEXT_RENDER_STATE_IDS = IntSet.of(DATA_TEXT_ID.getId(), DATA_LINE_WIDTH_ID.getId(), DATA_BACKGROUND_COLOR_ID.getId(), DATA_TEXT_OPACITY_ID.getId(), DATA_STYLE_FLAGS_ID.getId());
@@ -764,7 +_,7 @@
 
       }
 
-      private Component getText() {
+      public Component getText() {
          return this.entityData.get(DATA_TEXT_ID);
       }
 
@@ -772,7 +_,7 @@
          this.entityData.set(DATA_TEXT_ID, p_270902_);
       }
 
-      private int getLineWidth() {
+      public int getLineWidth() {
          return this.entityData.get(DATA_LINE_WIDTH_ID);
       }
 
@@ -796,7 +_,7 @@
          this.entityData.set(DATA_BACKGROUND_COLOR_ID, p_270241_);
       }
 
-      private byte getFlags() {
+      public byte getFlags() {
          return this.entityData.get(DATA_STYLE_FLAGS_ID);
       }
 
