--- a/net/minecraft/world/entity/Interaction.java
+++ b/net/minecraft/world/entity/Interaction.java
@@ -33,9 +_,9 @@
    private static final String f_271476_ = "interaction";
    private static final String f_271508_ = "response";
    @Nullable
-   private Interaction.PlayerAction f_271404_;
+   public Interaction.PlayerAction f_271404_;
    @Nullable
-   private Interaction.PlayerAction f_271193_;
+   public Interaction.PlayerAction f_271193_;
 
    public Interaction(EntityType<?> p_273319_, Level p_272713_) {
       super(p_273319_, p_272713_);
@@ -150,27 +_,27 @@
       return this.f_271193_ != null ? this.f_19853_.m_46003_(this.f_271193_.f_271379_()) : null;
    }
 
-   private void m_272058_(float p_273385_) {
+   public void m_272058_(float p_273385_) {
       this.f_19804_.m_135381_(f_271210_, p_273385_);
    }
 
-   private float m_272023_() {
+   public float m_272023_() {
       return this.f_19804_.m_135370_(f_271210_);
    }
 
-   private void m_271774_(float p_273733_) {
+   public void m_271774_(float p_273733_) {
       this.f_19804_.m_135381_(f_271237_, p_273733_);
    }
 
-   private float m_271858_() {
+   public float m_271858_() {
       return this.f_19804_.m_135370_(f_271237_);
    }
 
-   private void m_271717_(boolean p_273657_) {
+   public void m_271717_(boolean p_273657_) {
       this.f_19804_.m_135381_(f_271389_, p_273657_);
    }
 
-   private boolean m_271819_() {
+   public boolean m_271819_() {
       return this.f_19804_.m_135370_(f_271389_);
    }
 
@@ -186,7 +_,7 @@
       return this.m_271772_().m_20393_(this.m_20182_());
    }
 
-   static record PlayerAction(UUID f_271379_, long f_271492_) {
+   public static record PlayerAction(UUID f_271379_, long f_271492_) {
       public static final Codec<Interaction.PlayerAction> f_271433_ = RecordCodecBuilder.create((p_273237_) -> {
          return p_273237_.group(UUIDUtil.f_235867_.fieldOf("player").forGetter(Interaction.PlayerAction::f_271379_), Codec.LONG.fieldOf("timestamp").forGetter(Interaction.PlayerAction::f_271492_)).apply(p_273237_, Interaction.PlayerAction::new);
       });
