--- a/net/minecraft/world/entity/Interaction.java
+++ b/net/minecraft/world/entity/Interaction.java
@@ -19,10 +_,13 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
+import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityDamageEvent;
 import org.slf4j.Logger;
 
 public class Interaction extends Entity implements Attackable, Targeting {
@@ -36,9 +_,9 @@
     private static final String TAG_INTERACTION = "interaction";
     private static final String TAG_RESPONSE = "response";
     @Nullable
-    public Interaction.PlayerAction attack;
-    @Nullable
-    public Interaction.PlayerAction interaction;
+   	public Interaction.PlayerAction attack;
+   	@Nullable
+   	public Interaction.PlayerAction interaction;
 
     public Interaction(EntityType<?> p_273319_, Level p_272713_) {
         super(p_273319_, p_272713_);
@@ -134,14 +_,21 @@
         if (p_273553_ instanceof Player player) {
             this.attack = new Interaction.PlayerAction(player.getUUID(), this.level().getGameTime());
             if (player instanceof ServerPlayer serverplayer) {
-                CriteriaTriggers.PLAYER_HURT_ENTITY.trigger(serverplayer, this, player.damageSources().generic(), 1.0F, 1.0F, false);
+            // CraftBukkit start
+            DamageSource source = serverplayer.damageSources().playerAttack(serverplayer);
+            EntityDamageEvent event = CraftEventFactory.callNonLivingEntityDamageEvent(this, source, 1.0F, false);
+            if (event.isCancelled()) {
+               return true;
             }
+            // CraftBukkit end
+            CriteriaTriggers.PLAYER_HURT_ENTITY.trigger(serverplayer, this, source, (float) event.getFinalDamage(), 1.0F, false); // CraftBukkit
+         }
 
-            return !this.getResponse();
-        } else {
-            return false;
-        }
-    }
+         return !this.getResponse();
+      } else {
+         return false;
+      }
+   }
 
     @Override
     public InteractionResult interact(Player p_273507_, InteractionHand p_273048_) {
@@ -169,27 +_,27 @@
         return this.interaction != null ? this.level().getPlayerByUUID(this.interaction.player()) : null;
     }
 
-    public void setWidth(float p_273385_) {
+    private void setWidth(float p_273385_) {
         this.entityData.set(DATA_WIDTH_ID, p_273385_);
     }
 
-    public float getWidth() {
+    private float getWidth() {
         return this.entityData.get(DATA_WIDTH_ID);
     }
 
-    public void setHeight(float p_273733_) {
+    private void setHeight(float p_273733_) {
         this.entityData.set(DATA_HEIGHT_ID, p_273733_);
     }
 
-    public float getHeight() {
+    private float getHeight() {
         return this.entityData.get(DATA_HEIGHT_ID);
     }
 
-    public void setResponse(boolean p_273657_) {
+    private void setResponse(boolean p_273657_) {
         this.entityData.set(DATA_RESPONSE_ID, p_273657_);
     }
 
-    public boolean getResponse() {
+    private boolean getResponse() {
         return this.entityData.get(DATA_RESPONSE_ID);
     }
 
