--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1,21 +_,23 @@
 package net.minecraft.world.entity;
 
+import com.google.common.base.Function;
 import com.google.common.base.Objects;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.AtomicDouble;
+import com.mohistmc.plugins.ban.bans.BanItem;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
-import java.util.Collection;
-import java.util.ConcurrentModificationException;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.UUID;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.BlockUtil;
@@ -29,6 +_,8 @@
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.FloatTag;
+import net.minecraft.nbt.IntTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.Tag;
@@ -77,7 +_,6 @@
 import net.minecraft.world.entity.ai.attributes.DefaultAttributes;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.animal.FlyingAnimal;
-import net.minecraft.world.entity.animal.Wolf;
 import net.minecraft.world.entity.boss.wither.WitherBoss;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
@@ -114,17 +_,36 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
+import net.minecraftforge.common.ForgeHooks;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_20_R1.attribute.CraftAttributeMap;
+import org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_20_R1.inventory.CraftItemStack;
+import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+import org.bukkit.event.entity.EntityResurrectEvent;
+import org.bukkit.event.entity.EntityTeleportEvent;
+import org.bukkit.event.player.PlayerItemConsumeEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
 import org.slf4j.Logger;
+import org.spigotmc.AsyncCatcher;
 
-public abstract class LivingEntity extends Entity implements Attackable {
+public abstract class LivingEntity extends Entity implements Attackable, net.minecraftforge.common.extensions.IForgeLivingEntity {
    private static final Logger f_201943_ = LogUtils.getLogger();
    private static final UUID f_20929_ = UUID.fromString("662A6B8D-DA3E-4C1C-8813-96EA6097278D");
    private static final UUID f_20959_ = UUID.fromString("87f46a96-686f-4796-b035-22e16ee9e038");
    private static final UUID f_147184_ = UUID.fromString("1eaf83ff-7207-4596-b37a-d7a07b3ec4ce");
+   private static final UUID SLOW_FALLING_ID = UUID.fromString("A5B6CF2A-2F7C-31EF-9022-7C3E7D5E6ABA");
    private static final AttributeModifier f_20960_ = new AttributeModifier(f_20929_, "Sprinting speed boost", (double)0.3F, AttributeModifier.Operation.MULTIPLY_TOTAL);
+   private static final AttributeModifier SLOW_FALLING = new AttributeModifier(SLOW_FALLING_ID, "Slow falling acceleration reduction", -0.07, AttributeModifier.Operation.ADDITION); // Add -0.07 to 0.08 so we get the vanilla default of 0.01
    public static final int f_147166_ = 2;
    public static final int f_147167_ = 4;
    public static final int f_147168_ = 98;
@@ -157,7 +_,7 @@
    private static final int f_271312_ = 50;
    private final AttributeMap f_20943_;
    public CombatTracker f_20944_ = new CombatTracker(this);
-   public final Map<MobEffect, MobEffectInstance> f_20945_ = Maps.newHashMap();
+   public final Map<MobEffect, MobEffectInstance> f_20945_ = new ConcurrentHashMap<>(); // Mohist
    private final NonNullList<ItemStack> f_20946_ = NonNullList.m_122780_(2, ItemStack.f_41583_);
    private final NonNullList<ItemStack> f_20947_ = NonNullList.m_122780_(4, ItemStack.f_41583_);
    public boolean f_20911_;
@@ -207,7 +_,7 @@
    public boolean f_20948_ = true;
    @Nullable
    public LivingEntity f_20949_;
-   private int f_20950_;
+   public int f_20950_;
    private LivingEntity f_20951_;
    private int f_20952_;
    private float f_20953_;
@@ -226,6 +_,32 @@
    private float f_20932_;
    protected Brain<?> f_20939_;
    private boolean f_217034_;
+   // CraftBukkit start
+   public int expToDrop;
+   public boolean forceDrops;
+   public java.util.ArrayList<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<org.bukkit.inventory.ItemStack>();
+   public final org.bukkit.craftbukkit.v1_20_R1.attribute.CraftAttributeMap craftAttributes;
+   public boolean collides = true;
+   public Set<UUID> collidableExemptions = new HashSet<>();
+   public boolean bukkitPickUpLoot;
+   public org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.v1_20_R1.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
+   public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
+   public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper
+
+   @Override
+   public float getBukkitYaw() {
+      return this.m_6080_();
+   }
+   // CraftBukkit end
+   // Spigot start
+   public void inactiveTick()
+   {
+      super.inactiveTick();
+      ++this.f_20891_; // Above all the floats
+   }
+   // Spigot end
+   // Mohist start
+   public AtomicInteger invulnerableDurationAtom = new AtomicInteger(20);
 
    protected LivingEntity(EntityType<? extends LivingEntity> p_20966_, Level p_20967_) {
       super(p_20966_, p_20967_);
@@ -239,7 +_,9 @@
       this.f_20885_ = this.m_146908_();
       this.m_274367_(0.6F);
       NbtOps nbtops = NbtOps.f_128958_;
-      this.f_20939_ = this.m_8075_(new Dynamic<>(nbtops, nbtops.createMap(ImmutableMap.of(nbtops.createString("memories"), nbtops.emptyMap()))));
+      var dyn = new Dynamic<>(nbtops, nbtops.createMap(ImmutableMap.of(nbtops.createString("memories"), nbtops.emptyMap())));
+      this.f_20939_ = net.minecraftforge.common.ForgeHooks.onLivingMakeBrain(this, this.m_8075_(dyn), dyn);
+      this.craftAttributes = new CraftAttributeMap(f_20943_); // CraftBukkit
    }
 
    public Brain<?> m_6274_() {
@@ -273,7 +_,7 @@
    }
 
    public static AttributeSupplier.Builder m_21183_() {
-      return AttributeSupplier.m_22244_().m_22266_(Attributes.f_22276_).m_22266_(Attributes.f_22278_).m_22266_(Attributes.f_22279_).m_22266_(Attributes.f_22284_).m_22266_(Attributes.f_22285_);
+      return AttributeSupplier.m_22244_().m_22266_(Attributes.f_22276_).m_22266_(Attributes.f_22278_).m_22266_(Attributes.f_22279_).m_22266_(Attributes.f_22284_).m_22266_(Attributes.f_22285_).m_22266_(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).m_22266_(net.minecraftforge.common.ForgeMod.NAMETAG_DISTANCE.get()).m_22266_(net.minecraftforge.common.ForgeMod.ENTITY_GRAVITY.get()).m_22266_(net.minecraftforge.common.ForgeMod.STEP_HEIGHT.get());
    }
 
    protected void m_7840_(double p_20990_, boolean p_20991_, BlockState p_20992_, BlockPos p_20993_) {
@@ -298,20 +_,29 @@
             d0 = (double)p_20993_.m_123341_() + 0.5D + d3 / d6 * 0.5D;
             d2 = (double)p_20993_.m_123343_() + 0.5D + d5 / d6 * 0.5D;
          }
-
          float f = (float)Mth.m_14167_(this.f_19789_ - 3.0F);
-         double d4 = Math.min((double)(0.2F + f / 15.0F), 2.5D);
-         int i = (int)(150.0D * d4);
-         ((ServerLevel)this.m_9236_()).m_8767_(new BlockParticleOption(ParticleTypes.f_123794_, p_20992_), d0, d1, d2, i, 0.0D, 0.0D, 0.0D, (double)0.15F);
+         if (!p_20992_.m_60795_()) {
+            double d4 = Math.min((double)(0.2F + f / 15.0F), 2.5D);
+            int i = (int)(150.0D * d4);
+            if (!p_20992_.addLandingEffects((ServerLevel)this.f_19853_, p_20993_, p_20992_, this, i)) {
+               // CraftBukkit start - visiblity api
+               if (this instanceof ServerPlayer) {
+                  ((ServerLevel) this.f_19853_).sendParticles((ServerPlayer) this, new BlockParticleOption(ParticleTypes.f_123794_, p_20992_).setPos(p_20993_), this.m_20185_(), this.m_20186_(), this.m_20189_(), i, 0.0D, 0.0D, 0.0D, (double) 0.15F, false);
+               } else {
+                  ((ServerLevel) this.m_9236_()).m_8767_(new BlockParticleOption(ParticleTypes.f_123794_, p_20992_).setPos(p_20993_), this.m_20185_(), this.m_20186_(), this.m_20189_(), i, 0.0D, 0.0D, 0.0D, (double) 0.15F);
+               }
+               // CraftBukkit end
+            }
+         }
       }
 
       super.m_7840_(p_20990_, p_20991_, p_20992_, p_20993_);
-      if (p_20991_) {
+	  if (p_20991_) {
          this.f_20957_ = Optional.empty();
       }
-
    }
 
+   @Deprecated //FORGE: Use canDrownInFluidType instead
    public boolean m_6040_() {
       return this.m_6336_() == MobType.f_21641_;
    }
@@ -352,8 +_,11 @@
             }
          }
 
+         int airSupply = this.m_20146_();
+         net.minecraftforge.common.ForgeHooks.onLivingBreathe(this, airSupply - m_7302_(airSupply), m_7305_(airSupply) - airSupply);
+         if (false) // Forge: Handled in ForgeHooks#onLivingBreathe(LivingEntity, int, int)
          if (this.m_204029_(FluidTags.f_13131_) && !this.m_9236_().m_8055_(BlockPos.m_274561_(this.m_20185_(), this.m_20188_(), this.m_20189_())).m_60713_(Blocks.f_50628_)) {
-            boolean flag1 = !this.m_6040_() && !MobEffectUtil.m_19588_(this) && (!flag || !((Player)this).m_150110_().f_35934_);
+            boolean flag1 = this.canDrownInFluidType(this.getEyeInFluidType()) && !MobEffectUtil.m_19588_(this) && (!flag || !((Player)this).m_150110_().f_35934_);
             if (flag1) {
                this.m_20301_(this.m_7302_(this.m_20146_()));
                if (this.m_20146_() == -20) {
@@ -387,7 +_,7 @@
          }
       }
 
-      if (this.m_6084_() && (this.m_20071_() || this.f_146808_)) {
+      if (this.m_6084_() && (this.m_20071_() || this.f_146808_ || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType)))) {
          this.m_252836_();
       }
 
@@ -635,12 +_,14 @@
       return true;
    }
 
+   // Mohist start
+   public AtomicBoolean silent0 = new AtomicBoolean(false);
    public void m_238392_(EquipmentSlot p_238393_, ItemStack p_238394_, ItemStack p_238395_) {
       boolean flag = p_238395_.m_41619_() && p_238394_.m_41619_();
       if (!flag && !ItemStack.m_150942_(p_238394_, p_238395_) && !this.f_19803_) {
          Equipable equipable = Equipable.m_269088_(p_238395_);
          if (equipable != null && !this.m_5833_() && equipable.m_40402_() == p_238393_) {
-            if (!this.m_9236_().m_5776_() && !this.m_20067_()) {
+            if (!this.m_9236_().m_5776_() && !this.m_20067_() && !silent0.getAndSet(false)) {
                this.m_9236_().m_6263_((Player)null, this.m_20185_(), this.m_20186_(), this.m_20189_(), equipable.m_150681_(), this.m_5720_(), 1.0F, 1.0F);
             }
 
@@ -652,6 +_,15 @@
       }
    }
 
+   // CraftBukkit start
+   // Mohist start
+   public void onEquipItem(EquipmentSlot pSlot, ItemStack pOldItem, ItemStack pNewItem, boolean silent) {
+      silent0.set(silent);
+      m_238392_(pSlot, pOldItem, pNewItem);
+   }
+   // CraftBukkit end
+   // Mohist end
+   
    public void m_142687_(Entity.RemovalReason p_276115_) {
       super.m_142687_(p_276115_);
       this.f_20939_.m_276084_();
@@ -704,6 +_,18 @@
          }
       }
 
+      // CraftBukkit start
+      if (p_21096_.m_128441_("Bukkit.MaxHealth")) {
+         Tag nbtbase = p_21096_.m_128423_("Bukkit.MaxHealth");
+         if (nbtbase.m_7060_() == 5) {
+            this.m_21051_(Attributes.f_22276_).m_22100_(((FloatTag) nbtbase).m_7061_());
+         } else if (nbtbase.m_7060_() == 3) {
+            this.m_21051_(Attributes.f_22276_).m_22100_(((IntTag) nbtbase).m_7061_());
+         }
+      }
+      // CraftBukkit end
+
+
       if (p_21096_.m_128425_("Health", 99)) {
          this.m_21153_(p_21096_.m_128457_("Health"));
       }
@@ -734,14 +_,37 @@
       }
 
       if (p_21096_.m_128425_("Brain", 10)) {
-         this.f_20939_ = this.m_8075_(new Dynamic<>(NbtOps.f_128958_, p_21096_.m_128423_("Brain")));
-      }
-
-   }
+         var dyn = new Dynamic<>(NbtOps.f_128958_, p_21096_.m_128423_("Brain"));
+         this.f_20939_ = net.minecraftforge.common.ForgeHooks.onLivingMakeBrain(this, this.m_8075_(dyn), dyn);
+      }
+
+   }
+
+   // CraftBukkit start
+   private boolean isTickingEffects = false;
+   private List<ProcessableEffect> effectsToProcess = Lists.newArrayList();
+
+   private static class ProcessableEffect {
+
+      private MobEffect type;
+      private MobEffectInstance effect;
+      private final EntityPotionEffectEvent.Cause cause;
+
+      private ProcessableEffect(MobEffectInstance effect, EntityPotionEffectEvent.Cause cause) {
+         this.effect = effect;
+         this.cause = cause;
+      }
+
+      private ProcessableEffect(MobEffect type, EntityPotionEffectEvent.Cause cause) {
+         this.type = type;
+         this.cause = cause;
+      }
+   }
+    // CraftBukkit end
 
    protected void m_21217_() {
       Iterator<MobEffect> iterator = this.f_20945_.keySet().iterator();
-
+      isTickingEffects = true; // CraftBukkit
       try {
          while(iterator.hasNext()) {
             MobEffect mobeffect = iterator.next();
@@ -749,7 +_,8 @@
             if (!mobeffectinstance.m_19552_(this, () -> {
                this.m_141973_(mobeffectinstance, true, (Entity)null);
             })) {
-               if (!this.m_9236_().f_46443_) {
+               EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffectinstance, null, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.EXPIRATION);
+               if (!this.f_19853_.f_46443_ && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.MobEffectEvent.Expired(this, mobeffectinstance)) && !event.isCancelled()) {
                   iterator.remove();
                   this.m_7285_(mobeffectinstance);
                }
@@ -760,6 +_,18 @@
       } catch (ConcurrentModificationException concurrentmodificationexception) {
       }
 
+      // CraftBukkit start
+      isTickingEffects = false;
+      for (ProcessableEffect e : effectsToProcess) {
+         if (e.effect != null) {
+            addEffect(e.effect, e.cause);
+         } else {
+            removeEffect(e.type, e.cause);
+         }
+      }
+      effectsToProcess.clear();
+      // CraftBukkit end
+
       if (this.f_20948_) {
          if (!this.m_9236_().f_46443_) {
             this.m_8034_();
@@ -799,8 +_,10 @@
          this.m_6842_(false);
       } else {
          Collection<MobEffectInstance> collection = this.f_20945_.values();
-         this.f_19804_.m_135381_(f_20963_, m_21179_(collection));
-         this.f_19804_.m_135381_(f_20962_, PotionUtils.m_43564_(collection));
+         net.minecraftforge.event.entity.living.PotionColorCalculationEvent event = new net.minecraftforge.event.entity.living.PotionColorCalculationEvent(this, PotionUtils.m_43564_(collection), m_21179_(collection), collection);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         this.f_19804_.m_135381_(f_20963_, event.areParticlesHidden());
+         this.f_19804_.m_135381_(f_20962_, event.getColor());
          this.m_6842_(this.m_21023_(MobEffects.f_19609_));
       }
 
@@ -836,7 +_,7 @@
             d0 *= 0.5D;
          }
       }
-
+      d0 = net.minecraftforge.common.ForgeHooks.getEntityVisibilityMultiplier(this, p_20969_, d0);
       return d0;
    }
 
@@ -871,6 +_,8 @@
       this.f_19804_.m_135381_(f_20962_, 0);
    }
 
+   //Mohist start
+   public AtomicReference<EntityPotionEffectEvent.Cause> removeAllEffectsCause = new AtomicReference<>(EntityPotionEffectEvent.Cause.UNKNOWN);
    public boolean m_21219_() {
       if (this.m_9236_().f_46443_) {
          return false;
@@ -879,13 +_,31 @@
 
          boolean flag;
          for(flag = false; iterator.hasNext(); flag = true) {
-            this.m_7285_(iterator.next());
-            iterator.remove();
+            // CraftBukkit start
+            MobEffectInstance effect = iterator.next();
+            try {
+               if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.MobEffectEvent.Remove(this, effect)))
+                  continue;
+               EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, removeAllEffectsCause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN), EntityPotionEffectEvent.Action.CLEARED);
+               if (event.isCancelled()) {
+                  continue;
+               }
+               this.m_7285_(effect);
+               // CraftBukkit end
+               iterator.remove();
+            } catch (ConcurrentModificationException ignored) {
+            }
          }
 
          return flag;
       }
    }
+
+   public boolean removeAllEffects(EntityPotionEffectEvent.Cause cause) {
+      removeAllEffectsCause.set(cause);
+      return m_21219_();
+   }
+   // Mohist end
 
    public Collection<MobEffectInstance> m_21220_() {
       return this.f_20945_.values();
@@ -908,25 +_,69 @@
       return this.m_147207_(p_21165_, (Entity)null);
    }
 
+   // CraftBukkit start
+   public boolean addEffect(MobEffectInstance pEffectInstance, EntityPotionEffectEvent.Cause cause) {
+      return addEffect(pEffectInstance, (Entity) null, cause);
+   }
+
+   // Mohist start
+   public static AtomicReference<EntityPotionEffectEvent.Cause> addEffectCause = new AtomicReference<>(EntityPotionEffectEvent.Cause.UNKNOWN);
+
+   public void pushEffectCause(EntityPotionEffectEvent.Cause cause) {
+      addEffectCause.set(cause);
+   }
+
    public boolean m_147207_(MobEffectInstance p_147208_, @Nullable Entity p_147209_) {
+      boolean isAsync = AsyncCatcher.catchAsync();
+      if (isTickingEffects) {
+         effectsToProcess.add(new ProcessableEffect(p_147208_, addEffectCause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN)));
+         return true;
+      }
+      // CraftBukkit end
       if (!this.m_7301_(p_147208_)) {
          return false;
       } else {
          MobEffectInstance mobeffectinstance = this.f_20945_.get(p_147208_.m_19544_());
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.MobEffectEvent.Added(this, mobeffectinstance, p_147208_, p_147209_));
+         // CraftBukkit start
+         boolean override = false;
+         if (mobeffectinstance != null) {
+            override = new MobEffectInstance(mobeffectinstance).m_19558_(p_147208_);
+         }
+
+         if (!isAsync) {
+            EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, mobeffectinstance, p_147208_, addEffectCause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN), override);
+            if (event.isCancelled()) {
+               return false;
+            }
+            override = event.isOverride();
+         }
+         // CraftBukkit end
          if (mobeffectinstance == null) {
             this.f_20945_.put(p_147208_.m_19544_(), p_147208_);
             this.m_142540_(p_147208_, p_147209_);
             return true;
-         } else if (mobeffectinstance.m_19558_(p_147208_)) {
+         } else if (override) {
+            mobeffectinstance.m_19558_(p_147208_);
             this.m_141973_(mobeffectinstance, true, p_147209_);
+            // CraftBukkit end
             return true;
          } else {
             return false;
          }
       }
    }
+
+   public boolean addEffect(MobEffectInstance pEffectInstance, @Nullable Entity pEntity, EntityPotionEffectEvent.Cause cause) {
+      addEffectCause.set(cause);
+      return m_147207_(pEffectInstance, pEntity);
+   }
+   // Mohist end
 
    public boolean m_7301_(MobEffectInstance p_21197_) {
+      net.minecraftforge.event.entity.living.MobEffectEvent.Applicable event = new net.minecraftforge.event.entity.living.MobEffectEvent.Applicable(this, p_21197_);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+      if (event.getResult() != net.minecraftforge.eventbus.api.Event.Result.DEFAULT) return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
       if (this.m_6336_() == MobType.f_21641_) {
          MobEffect mobeffect = p_21197_.m_19544_();
          if (mobeffect == MobEffects.f_19605_ || mobeffect == MobEffects.f_19614_) {
@@ -953,12 +_,35 @@
       return this.m_6336_() == MobType.f_21641_;
    }
 
+   // Mohist start
+   public AtomicReference<EntityPotionEffectEvent.Cause> removeEffectNoUpdateCause = new AtomicReference<>(EntityPotionEffectEvent.Cause.UNKNOWN);
    @Nullable
    public MobEffectInstance m_6234_(@Nullable MobEffect p_21164_) {
+      if (isTickingEffects) {
+         effectsToProcess.add(new ProcessableEffect(p_21164_, removeEffectNoUpdateCause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN)));
+         return null;
+      }
+
+      MobEffectInstance effect = this.f_20945_.get(p_21164_);
+      if (effect == null) {
+         return null;
+      }
+
+      EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, removeEffectNoUpdateCause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN));
+      if (event.isCancelled()) {
+         return null;
+      }
       return this.f_20945_.remove(p_21164_);
    }
 
+   @Nullable
+   public MobEffectInstance c(@Nullable MobEffect pEffect, EntityPotionEffectEvent.Cause cause) {
+      removeEffectNoUpdateCause.set(cause);
+      return m_6234_(pEffect);
+   }
+
    public boolean m_21195_(MobEffect p_21196_) {
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.MobEffectEvent.Remove(this, p_21196_))) return false;
       MobEffectInstance mobeffectinstance = this.m_6234_(p_21196_);
       if (mobeffectinstance != null) {
          this.m_7285_(mobeffectinstance);
@@ -968,6 +_,12 @@
       }
    }
 
+   public boolean removeEffect(MobEffect pEffect, EntityPotionEffectEvent.Cause cause) {
+      removeEffectNoUpdateCause.set(cause);
+      return m_21195_(pEffect);
+   }
+   // Mohist end
+
    protected void m_142540_(MobEffectInstance p_147190_, @Nullable Entity p_147191_) {
       this.f_20948_ = true;
       if (!this.m_9236_().f_46443_) {
@@ -1015,19 +_,59 @@
 
    }
 
+   public AtomicReference<RegainReason> regainReason0 = new AtomicReference<>(RegainReason.CUSTOM);
+
+   public void pushHealReason(EntityRegainHealthEvent.RegainReason regainReason) {
+      this.regainReason0.set(regainReason);
+   }
+
    public void m_5634_(float p_21116_) {
+      p_21116_ = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, p_21116_);
+      if (p_21116_ <= 0) return;
       float f = this.m_21223_();
       if (f > 0.0F) {
-         this.m_21153_(f + p_21116_);
+         EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), p_21116_, regainReason0.getAndSet(RegainReason.CUSTOM));
+         // Suppress during worldgen
+         if (this.valid) {
+            this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+         }
+
+         if (!event.isCancelled()) {
+            this.m_21153_((float) (this.m_21223_() + event.getAmount()));
+         }
+         // CraftBukkit end
       }
+   }
 
+   public void heal(float pHealAmount, EntityRegainHealthEvent.RegainReason regainReason) {
+      this.regainReason0.set(regainReason);
+      this.m_5634_(pHealAmount);
    }
 
    public float m_21223_() {
+      if (this instanceof ServerPlayer serverPlayer && serverPlayer.initialized) {
+         return (float) serverPlayer.getBukkitEntity().getHealth();
+      }
       return this.f_19804_.m_135370_(f_20961_);
    }
 
    public void m_21153_(float p_21154_) {
+      // CraftBukkit start - Handle scaled health
+      if (this instanceof ServerPlayer serverPlayer && serverPlayer.initialized) {
+         org.bukkit.craftbukkit.v1_20_R1.entity.CraftPlayer player = serverPlayer.getBukkitEntity();
+         // Squeeze
+         if (p_21154_ < 0.0F) {
+            player.setRealHealth(0.0D);
+         } else if (p_21154_ > player.getMaxHealth()) {
+            player.setRealHealth(player.getMaxHealth());
+         } else {
+            player.setRealHealth(p_21154_);
+         }
+
+         player.updateScaledHealth(false);
+         return;
+      }
+      // CraftBukkit end
       this.f_19804_.m_135381_(f_20961_, Mth.m_14036_(p_21154_, 0.0F, this.m_21233_()));
    }
 
@@ -1036,35 +_,28 @@
    }
 
    public boolean m_6469_(DamageSource p_21016_, float p_21017_) {
+      if (!net.minecraftforge.common.ForgeHooks.onLivingAttack(this, p_21016_, p_21017_)) return false;
       if (this.m_6673_(p_21016_)) {
          return false;
       } else if (this.m_9236_().f_46443_) {
          return false;
-      } else if (this.m_21224_()) {
+      } else if (this.m_213877_() || this.f_20890_ || this.m_21223_() <= 0.0F) { // CraftBukkit - Don't allow entities that got set to dead/killed elsewhere to get damaged and die
+         return false;
+      } else if (this.m_21224_()) { // Mohist
          return false;
       } else if (p_21016_.m_269533_(DamageTypeTags.f_268745_) && this.m_21023_(MobEffects.f_19607_)) {
          return false;
       } else {
-         if (this.m_5803_() && !this.m_9236_().f_46443_) {
+         if (this.m_5803_() && !this.f_19853_.f_46443_) {
             this.m_5796_();
          }
 
          this.f_20891_ = 0;
          float f = p_21017_;
+
          boolean flag = false;
          float f1 = 0.0F;
          if (p_21017_ > 0.0F && this.m_21275_(p_21016_)) {
-            this.m_7909_(p_21017_);
-            f1 = p_21017_;
-            p_21017_ = 0.0F;
-            if (!p_21016_.m_269533_(DamageTypeTags.f_268524_)) {
-               Entity entity = p_21016_.m_7640_();
-               if (entity instanceof LivingEntity) {
-                  LivingEntity livingentity = (LivingEntity)entity;
-                  this.m_6728_(livingentity);
-               }
-            }
-
             flag = true;
          }
 
@@ -1074,23 +_,39 @@
 
          this.f_267362_.m_267771_(1.5F);
          boolean flag1 = true;
-         if ((float)this.f_19802_ > 10.0F && !p_21016_.m_269533_(DamageTypeTags.f_273918_)) {
+         if ((float) this.f_19802_ > (float) this.f_20926_ / 2.0F && !p_21016_.m_269533_(DamageTypeTags.f_273918_)) { // CraftBukkit - restore use of maxNoDamageTicks
             if (p_21017_ <= this.f_20898_) {
                return false;
             }
 
+            // CraftBukkit start
             this.m_6475_(p_21016_, p_21017_ - this.f_20898_);
+            if (!this.canDamage()) {
+               return false;
+            }
+            // CraftBukkit end
             this.f_20898_ = p_21017_;
             flag1 = false;
          } else {
-            this.f_20898_ = p_21017_;
-            this.f_19802_ = 20;
+            // CraftBukkit start
             this.m_6475_(p_21016_, p_21017_);
+            if (!this.canDamage()) {
+               return false;
+            }
+            // CraftBukkit end
+
+            this.f_20898_ = p_21017_;
+            if (this.f_20926_ == invulnerableDurationAtom.get()) {
+               this.f_19802_ = 20; // CraftBukkit - restore use of maxNoDamageTicks
+            } else {
+               this.f_19802_ = this.f_20926_;
+            }
+            // CraftBukkit end
             this.f_20917_ = 10;
             this.f_20916_ = this.f_20917_;
          }
 
-         if (p_21016_.m_269533_(DamageTypeTags.f_268627_) && !this.m_6844_(EquipmentSlot.HEAD).m_41619_()) {
+         if (false && p_21016_.m_269533_(DamageTypeTags.f_268627_) && !this.m_6844_(EquipmentSlot.HEAD).m_41619_()) {
             this.m_142642_(p_21016_, p_21017_);
             p_21017_ *= 0.75F;
          }
@@ -1108,11 +_,10 @@
                Player player1 = (Player)entity1;
                this.f_20889_ = 100;
                this.f_20888_ = player1;
-            } else if (entity1 instanceof Wolf) {
-               Wolf wolf = (Wolf)entity1;
-               if (wolf.m_21824_()) {
+            } else if (entity1 instanceof net.minecraft.world.entity.TamableAnimal tamableEntity) {
+               if (tamableEntity.m_21824_()) {
                   this.f_20889_ = 100;
-                  LivingEntity livingentity2 = wolf.m_269323_();
+                  LivingEntity livingentity2 = tamableEntity.m_269323_();
                   if (livingentity2 instanceof Player) {
                      Player player = (Player)livingentity2;
                      this.f_20888_ = player;
@@ -1171,7 +_,7 @@
          if (this instanceof ServerPlayer) {
             CriteriaTriggers.f_10574_.m_35174_((ServerPlayer)this, p_21016_, f, p_21017_, flag);
             if (f1 > 0.0F && f1 < 3.4028235E37F) {
-               ((ServerPlayer)this).m_36222_(Stats.f_12932_, Math.round(f1 * 10.0F));
+               ((ServerPlayer)this).m_6278_(Stats.f_12988_.m_12902_(Stats.f_12932_), Math.round(f1 * 10.0F));
             }
          }
 
@@ -1197,31 +_,46 @@
       } else {
          ItemStack itemstack = null;
 
+         // CraftBukkit start
+         InteractionHand hand = null;
+         ItemStack itemstack1 = ItemStack.f_41583_;
          for(InteractionHand interactionhand : InteractionHand.values()) {
-            ItemStack itemstack1 = this.m_21120_(interactionhand);
-            if (itemstack1.m_150930_(Items.f_42747_)) {
+            itemstack1 = this.m_21120_(interactionhand);
+            if (itemstack1.m_150930_(Items.f_42747_) && net.minecraftforge.common.ForgeHooks.onLivingUseTotem(this, p_21263_, itemstack1, interactionhand)) {
+               hand = interactionhand; // CraftBukkit
                itemstack = itemstack1.m_41777_();
+               break;
+            }
+         }
+
+         org.bukkit.inventory.EquipmentSlot handSlot = (hand != null) ? org.bukkit.craftbukkit.v1_20_R1.CraftEquipmentSlot.getHand(hand) : null;
+         EntityResurrectEvent event = new EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot);
+         event.setCancelled(itemstack == null);
+         this.f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            if (!itemstack1.m_41619_()) {
                itemstack1.m_41774_(1);
-               break;
             }
-         }
-
-         if (itemstack != null) {
-            if (this instanceof ServerPlayer) {
-               ServerPlayer serverplayer = (ServerPlayer)this;
-               serverplayer.m_36246_(Stats.f_12982_.m_12902_(Items.f_42747_));
+            if (itemstack != null && this instanceof ServerPlayer serverplayer) {
+               // CraftBukkit end
+               serverplayer.m_6278_(Stats.f_12982_.m_12902_(Items.f_42747_), 1);
                CriteriaTriggers.f_10551_.m_74431_(serverplayer, itemstack);
             }
 
             this.m_21153_(1.0F);
+            removeAllEffectsCause.set(EntityPotionEffectEvent.Cause.TOTEM);
             this.m_21219_();
+            addEffectCause.set(EntityPotionEffectEvent.Cause.TOTEM);
             this.m_7292_(new MobEffectInstance(MobEffects.f_19605_, 900, 1));
+            addEffectCause.set(EntityPotionEffectEvent.Cause.TOTEM);
             this.m_7292_(new MobEffectInstance(MobEffects.f_19617_, 100, 1));
+            addEffectCause.set(EntityPotionEffectEvent.Cause.TOTEM);
             this.m_7292_(new MobEffectInstance(MobEffects.f_19607_, 800, 0));
             this.m_9236_().m_7605_(this, (byte)35);
          }
 
-         return itemstack != null;
+         return !event.isCancelled();
       }
    }
 
@@ -1278,36 +_,79 @@
    }
 
    public void m_6667_(DamageSource p_21014_) {
+      if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_21014_)) return;
       if (!this.m_213877_() && !this.f_20890_) {
          Entity entity = p_21014_.m_7639_();
-         LivingEntity livingentity = this.m_21232_();
-         if (this.f_20897_ >= 0 && livingentity != null) {
-            livingentity.m_5993_(this, this.f_20897_, p_21014_);
-         }
-
-         if (this.m_5803_()) {
-            this.m_5796_();
-         }
-
-         if (!this.m_9236_().f_46443_ && this.m_8077_()) {
-            f_201943_.info("Named entity {} died: {}", this, this.m_21231_().m_19293_().getString());
-         }
+         LivingEntity entityliving = this.m_21232_();
+            /* // Paper - move down to make death event cancellable - this is the awardKillScore below
+            if (this.deathScore >= 0 && entityliving != null) {
+                entityliving.awardKillScore(this, this.deathScore, p_21014_);
+            }
+
+            if (this.isSleeping()) {
+                this.stopSleeping();
+            }
+
+            if (!this.level().isClientSide && this.hasCustomName()) {
+                if (org.spigotmc.SpigotConfig.logNamedDeaths) LivingEntity.LOGGER.info("Named entity {} died: {}", this, this.getCombatTracker().getDeathMessage().getString()); // Spigot
+            }
+             */ // Paper - move down to make death event cancellable - this is the awardKillScore below
 
          this.f_20890_ = true;
-         this.m_21231_().m_19296_();
-         Level level = this.m_9236_();
-         if (level instanceof ServerLevel) {
-            ServerLevel serverlevel = (ServerLevel)level;
-            if (entity == null || entity.m_214076_(serverlevel, this)) {
+         // Paper - moved into if below
+         Level world = this.m_9236_();
+
+         if (world instanceof ServerLevel) {
+            ServerLevel worldserver = (ServerLevel) world;
+            // Paper - move below into if for onKill
+
+            // Paper start
+            this.m_6668_(p_21014_);
+            // Mohist start - For better mod compatibility
+            org.bukkit.event.entity.EntityDeathEvent deathEvent = mohist$lastDeathEvent;
+            mohist$lastDeathEvent = null;
+            // Mohist end
+            if (deathEvent == null || !deathEvent.isCancelled()) {
+               // if (this.deathScore >= 0 && entityliving != null) { // Paper moved to be run earlier in #dropAllDeathLoot before destroying the drop items in CraftEventFactory#callEntityDeathEvent
+               //     entityliving.awardKillScore(this, this.deathScore, p_21014_);
+               // }
+               // Paper start - clear equipment if event is not cancelled
+               if (this instanceof Mob) {
+                  for (EquipmentSlot slot : this.clearedEquipmentSlots) {
+                     this.m_8061_(slot, ItemStack.f_41583_);
+                  }
+                  this.clearedEquipmentSlots.clear();
+               }
+               // Paper end
+
+               if (this.m_5803_()) {
+                  this.m_5796_();
+               }
+
+               if (!this.m_9236_().f_46443_ && this.m_8077_()) {
+                  if (org.spigotmc.SpigotConfig.logNamedDeaths) LivingEntity.f_201943_.info("Named entity {} died: {}", this, this.m_21231_().m_19293_().getString()); // Spigot
+               }
+
+               this.m_21231_().m_19296_();
+               if (entity != null) {
+                  entity.m_214076_((ServerLevel) this.m_9236_(), this);
+               }
                this.m_146850_(GameEvent.f_223707_);
-               this.m_6668_(p_21014_);
-               this.m_21268_(livingentity);
+            } else {
+               this.f_20890_ = false;
+               this.m_21153_((float) deathEvent.getReviveHealth());
             }
 
-            this.m_9236_().m_7605_(this, (byte)3);
+            // Paper end
+            this.m_21268_(entityliving);
          }
 
-         this.m_20124_(Pose.DYING);
+         // Paper start
+         if (this.f_20890_) { // Paper
+            this.m_9236_().m_7605_(this, (byte) 3);
+            this.m_20124_(Pose.DYING);
+         }
+         // Paper end
       }
    }
 
@@ -1315,17 +_,26 @@
       if (!this.m_9236_().f_46443_) {
          boolean flag = false;
          if (p_21269_ instanceof WitherBoss) {
-            if (this.m_9236_().m_46469_().m_46207_(GameRules.f_46132_)) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.f_19853_, p_21269_)) {
                BlockPos blockpos = this.m_20183_();
                BlockState blockstate = Blocks.f_50070_.m_49966_();
-               if (this.m_9236_().m_8055_(blockpos).m_60795_() && blockstate.m_60710_(this.m_9236_(), blockpos)) {
-                  this.m_9236_().m_7731_(blockpos, blockstate, 3);
-                  flag = true;
+               if (this.f_19853_.m_46859_(blockpos) && blockstate.m_60710_(this.f_19853_, blockpos)) {
+                  // CraftBukkit start - call EntityBlockFormEvent for Wither Rose
+                  flag = org.bukkit.craftbukkit.v1_20_R1.event.CraftEventFactory.handleBlockFormEvent(this.f_19853_, blockpos, blockstate, 3, this);
+                  // CraftBukkit end
                }
             }
 
             if (!flag) {
                ItemEntity itementity = new ItemEntity(this.m_9236_(), this.m_20185_(), this.m_20186_(), this.m_20189_(), new ItemStack(Items.f_41951_));
+
+               // CraftBukkit start
+               org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+               CraftEventFactory.callEvent(event);
+               if (event.isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                this.m_9236_().m_7967_(itementity);
             }
          }
@@ -1333,31 +_,71 @@
       }
    }
 
+   // Paper start
+   protected boolean clearEquipmentSlots = true;
+   protected Set<EquipmentSlot> clearedEquipmentSlots = new java.util.HashSet<>();
+   protected @org.jetbrains.annotations.Nullable org.bukkit.event.entity.EntityDeathEvent mohist$lastDeathEvent; // Mohist - For better mod compatibility
    protected void m_6668_(DamageSource p_21192_) {
+      // Paper end
       Entity entity = p_21192_.m_7639_();
-      int i;
-      if (entity instanceof Player) {
-         i = EnchantmentHelper.m_44930_((LivingEntity)entity);
-      } else {
-         i = 0;
-      }
+      int i = net.minecraftforge.common.ForgeHooks.getLootingLevel(this, entity, p_21192_);
+      this.captureDrops(new java.util.ArrayList<>());
 
       boolean flag = this.f_20889_ > 0;
+
+      this.m_5907_(); // CraftBukkit - from below
       if (this.m_6125_() && this.m_9236_().m_46469_().m_46207_(GameRules.f_46135_)) {
          this.m_7625_(p_21192_, flag);
+         // Paper start
+         final boolean prev = this.clearEquipmentSlots;
+         this.clearEquipmentSlots = false;
+         this.clearedEquipmentSlots.clear();
+         // Paper end
          this.m_7472_(p_21192_, i, flag);
+         this.clearEquipmentSlots = prev; // Paper
       }
+      // CraftBukkit start - Call death event // Paper start - call advancement triggers with correct entity equipment
+      org.bukkit.event.entity.EntityDeathEvent deathEvent = CraftEventFactory.callEntityDeathEvent(this, this.drops, () -> {
+         final LivingEntity entityliving = this.m_21232_();
+         if (this.f_20897_ >= 0 && entityliving != null) {
+            entityliving.m_5993_(this, this.f_20897_, p_21192_);
+         }
+      }); // Paper end
+      this.postDeathDropItems(deathEvent); // Paper
+      // this.drops = new ArrayList<>(); // Mohist - prevent issues with Forge
+      // CraftBukkit end
 
-      this.m_5907_();
+      // this.dropInventory();// CraftBukkit - moved up
       this.m_21226_();
+
+      // Forge start
+      Collection<ItemEntity> drops = captureDrops(null);
+      if (!net.minecraftforge.common.ForgeHooks.onLivingDrops(this, p_21192_, drops, i, f_20889_ > 0))
+         drops.forEach(e -> m_9236_().m_7967_(e));
+      // Forge end
+      mohist$lastDeathEvent = deathEvent; // Mohist - Use alternative way to pass the event out
    }
 
    protected void m_5907_() {
    }
 
+   protected void postDeathDropItems(org.bukkit.event.entity.EntityDeathEvent event) {} // Paper - method for post death logic that cannot be ran before the event is potentially cancelled
+
+   // CraftBukkit start
+   public int getExpReward() {
+      if (this.m_9236_() instanceof ServerLevel && !this.m_217046_() && (this.m_6124_() || this.f_20889_ > 0 && this.m_6149_() && this.m_9236_().m_46469_().m_46207_(GameRules.f_46135_))) {
+         int i = this.m_213860_();
+         return i;
+      } else {
+         return 0;
+      }
+   }
+   // CraftBukkit end
+
    protected void m_21226_() {
       if (this.m_9236_() instanceof ServerLevel && !this.m_217046_() && (this.m_6124_() || this.f_20889_ > 0 && this.m_6149_() && this.m_9236_().m_46469_().m_46207_(GameRules.f_46135_))) {
-         ExperienceOrb.m_147082_((ServerLevel)this.m_9236_(), this.m_20182_(), this.m_213860_());
+         int reward = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.f_20888_, this.m_213860_());
+         ExperienceOrb.m_147082_((ServerLevel) this.m_9236_(), this.m_20182_(), reward);
       }
 
    }
@@ -1386,6 +_,11 @@
    }
 
    public void m_147240_(double p_147241_, double p_147242_, double p_147243_) {
+      net.minecraftforge.event.entity.living.LivingKnockBackEvent event = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(this, (float) p_147241_, p_147242_, p_147243_);
+      if(event.isCanceled()) return;
+      p_147241_ = event.getStrength();
+      p_147242_ = event.getRatioX();
+      p_147243_ = event.getRatioZ();
       p_147241_ *= 1.0D - this.m_21133_(Attributes.f_22278_);
       if (!(p_147241_ <= 0.0D)) {
          this.f_19812_ = true;
@@ -1445,6 +_,36 @@
       return p_21202_.m_41616_();
    }
 
+   // CraftBukkit start - Add delegate methods
+   public SoundEvent getHurtSound0(DamageSource damagesource) {
+      return m_7975_(damagesource);
+   }
+
+   public SoundEvent getDeathSound0() {
+      return m_5592_();
+   }
+
+   public SoundEvent getFallDamageSound0(int fallHeight) {
+      return m_5896_(fallHeight);
+   }
+
+   public SoundEvent getDrinkingSound0(ItemStack itemstack) {
+      return m_7838_(itemstack);
+   }
+
+   public SoundEvent getEatingSound0(ItemStack itemstack) {
+      return m_7866_(itemstack);
+   }
+   // CraftBukkit end
+
+   public void m_6853_(boolean p_21182_) {
+      super.m_6853_(p_21182_);
+      if (p_21182_) {
+         this.f_20957_ = Optional.empty();
+      }
+
+   }
+
    public Optional<BlockPos> m_21227_() {
       return this.f_20957_;
    }
@@ -1455,15 +_,9 @@
       } else {
          BlockPos blockpos = this.m_20183_();
          BlockState blockstate = this.m_146900_();
-         if (blockstate.m_204336_(BlockTags.f_13082_)) {
-            this.f_20957_ = Optional.of(blockpos);
-            return true;
-         } else if (blockstate.m_60734_() instanceof TrapDoorBlock && this.m_21176_(blockpos, blockstate)) {
-            this.f_20957_ = Optional.of(blockpos);
-            return true;
-         } else {
-            return false;
-         }
+         Optional<BlockPos> ladderPos = net.minecraftforge.common.ForgeHooks.isLivingOnLadder(blockstate, m_9236_(), blockpos, this);
+         if (ladderPos.isPresent()) this.f_20957_ = ladderPos;
+         return ladderPos.isPresent();
       }
    }
 
@@ -1483,12 +_,22 @@
    }
 
    public boolean m_142535_(float p_147187_, float p_147188_, DamageSource p_147189_) {
+      float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, p_147187_, p_147188_);
+      if (ret == null) return false;
+      p_147187_ = ret[0];
+      p_147188_ = ret[1];
+
       boolean flag = super.m_142535_(p_147187_, p_147188_, p_147189_);
       int i = this.m_5639_(p_147187_, p_147188_);
       if (i > 0) {
+         // CraftBukkit start
+         if (!this.m_6469_(p_147189_, (float) i)) {
+            return true;
+         }
+         // CraftBukkit end
          this.m_5496_(this.m_5896_(i), 1.0F, 1.0F);
          this.m_21229_();
-         this.m_6469_(p_147189_, (float)i);
+         // this.hurt(pSource, (float)i); // CraftBukkit - moved up
          return true;
       } else {
          return flag;
@@ -1510,9 +_,10 @@
          int i = Mth.m_14107_(this.m_20185_());
          int j = Mth.m_14107_(this.m_20186_() - (double)0.2F);
          int k = Mth.m_14107_(this.m_20189_());
-         BlockState blockstate = this.m_9236_().m_8055_(new BlockPos(i, j, k));
+         BlockPos pos = new BlockPos(i, j, k);
+         BlockState blockstate = this.m_9236_().m_8055_(pos);
          if (!blockstate.m_60795_()) {
-            SoundType soundtype = blockstate.m_60827_();
+            SoundType soundtype = blockstate.getSoundType(m_9236_(), pos, this);
             this.m_5496_(soundtype.m_56779_(), soundtype.m_56773_() * 0.5F, soundtype.m_56774_() * 0.75F);
          }
 
@@ -1537,9 +_,12 @@
    protected void m_7909_(float p_21316_) {
    }
 
+   // Mohist start
+   protected AtomicBoolean apoli_mixin = new AtomicBoolean(false);
    protected float m_21161_(DamageSource p_21162_, float p_21163_) {
       if (!p_21162_.m_269533_(DamageTypeTags.f_268490_)) {
-         this.m_6472_(p_21162_, p_21163_);
+         apoli_mixin.set(true);
+         this.m_6472_(p_21162_, p_21163_); // Mohist
          p_21163_ = CombatRules.m_19272_(p_21163_, (float)this.m_21230_(), (float)this.m_21133_(Attributes.f_22285_));
       }
 
@@ -1555,15 +_,18 @@
             int j = 25 - i;
             float f = p_21194_ * (float)j;
             float f1 = p_21194_;
-            p_21194_ = Math.max(f / 25.0F, 0.0F);
-            float f2 = f1 - p_21194_;
+            float mixin = Math.max(f / 25.0F, 0.0F);
+            // CraftBukkit - Moved to damageEntity0(DamageSource, float)
+            /*
+            float f2 = f1 - pDamageAmount;
             if (f2 > 0.0F && f2 < 3.4028235E37F) {
                if (this instanceof ServerPlayer) {
-                  ((ServerPlayer)this).m_36222_(Stats.f_12934_, Math.round(f2 * 10.0F));
-               } else if (p_21193_.m_7639_() instanceof ServerPlayer) {
-                  ((ServerPlayer)p_21193_.m_7639_()).m_36222_(Stats.f_12930_, Math.round(f2 * 10.0F));
+                  ((ServerPlayer)this).awardStat(Stats.CUSTOM.get(Stats.DAMAGE_RESISTED), Math.round(f2 * 10.0F));
+               } else if (pDamageSource.getEntity() instanceof ServerPlayer) {
+                  ((ServerPlayer)pDamageSource.getEntity()).awardStat(Stats.CUSTOM.get(Stats.DAMAGE_DEALT_RESISTED), Math.round(f2 * 10.0F));
                }
             }
+            */
          }
 
          if (p_21194_ <= 0.0F) {
@@ -1582,27 +_,250 @@
    }
 
    protected void m_6475_(DamageSource p_21240_, float p_21241_) {
+      if (injectedHurt) {
+         canDamage.set(damageEntity0(p_21240_, p_21241_));
+         if (damage.get() != -999) {
+            this.m_21153_(this.m_21223_() - (float) damage.getAndSet(-999));
+         }
+      } else {
+         if (!this.m_6673_(p_21240_)) {
+            p_21241_ = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, p_21240_, p_21241_);
+            if (p_21241_ <= 0) return;
+            p_21241_ = this.m_21161_(p_21240_, p_21241_);
+            p_21241_ = this.m_6515_(p_21240_, p_21241_);
+            float f1 = Math.max(p_21241_ - this.m_6103_(), 0.0F);
+            this.m_7911_(this.m_6103_() - (p_21241_ - f1));
+            float f = p_21241_ - f1;
+            if (f > 0.0F && f < 3.4028235E37F) {
+               Entity entity = p_21240_.m_7639_();
+               if (entity instanceof ServerPlayer) {
+                  ServerPlayer serverplayer = (ServerPlayer) entity;
+                  serverplayer.m_36222_(Stats.f_12929_, Math.round(f * 10.0F));
+               }
+            }
+
+            f1 = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, p_21240_, f1);
+            if (f1 != 0.0F) {
+               this.m_21231_().m_289194_(p_21240_, f1);
+               this.m_21153_(this.m_21223_() - f1);
+               this.m_7911_(this.m_6103_() - f1);
+               this.m_146850_(GameEvent.f_223706_);
+            }
+         }
+      }
+   }
+
+   protected boolean injectedHurt = true;
+   public AtomicBoolean canDamage = new AtomicBoolean(true);
+   public AtomicDouble damage = new AtomicDouble(-999);
+   public boolean canDamage() {
+      return canDamage.getAndSet(true);
+   }
+   // Mohist end
+
+   // CraftBukkit start
+   protected boolean damageEntity0(final DamageSource p_21240_, float p_21241_) { // void -> boolean, add final
       if (!this.m_6673_(p_21240_)) {
-         p_21241_ = this.m_21161_(p_21240_, p_21241_);
-         p_21241_ = this.m_6515_(p_21240_, p_21241_);
-         float f1 = Math.max(p_21241_ - this.m_6103_(), 0.0F);
-         this.m_7911_(this.m_6103_() - (p_21241_ - f1));
-         float f = p_21241_ - f1;
-         if (f > 0.0F && f < 3.4028235E37F) {
-            Entity entity = p_21240_.m_7639_();
-            if (entity instanceof ServerPlayer) {
-               ServerPlayer serverplayer = (ServerPlayer)entity;
-               serverplayer.m_36222_(Stats.f_12929_, Math.round(f * 10.0F));
-            }
-         }
-
-         if (f1 != 0.0F) {
-            this.m_21231_().m_289194_(p_21240_, f1);
-            this.m_21153_(this.m_21223_() - f1);
-            this.m_7911_(this.m_6103_() - f1);
-            this.m_146850_(GameEvent.f_223706_);
+         // Check if entity is a "human" aka player
+         final boolean human = this instanceof Player;
+         p_21241_ = ForgeHooks.onLivingHurt(this, p_21240_, p_21241_);
+         // If the damage is negative return true
+         if (p_21241_ < 0) {
+            return true;
+         }
+         final float originalDamage = p_21241_;
+
+         Function<Double, Double> hardHat = new Function<Double, Double>() {
+            @Override
+            public Double apply(Double input) {
+               if (p_21240_.m_269533_(DamageTypeTags.f_268627_)  && !LivingEntity.this.m_6844_(EquipmentSlot.HEAD).m_41619_()) {
+                  return -(input - (input * 0.75F));
+               }
+               return -0.0;
+            }
+         };
+         float hardHatModifier = hardHat.apply((double) p_21241_).floatValue();
+         p_21241_ += hardHatModifier;
+
+         Function<Double, Double> blocking;
+         boolean shieldTakesDamage = false;
+         if (this.m_21275_(p_21240_)) {
+            var shieldEvent = ForgeHooks.onShieldBlock(this, p_21240_, p_21241_);
+            if (!shieldEvent.isCanceled()) {
+               var blocked = shieldEvent.getBlockedDamage();
+               shieldTakesDamage = shieldEvent.shieldTakesDamage();
+               blocking = new Function<Double, Double>() {
+                  @Override
+                  public Double apply(Double input) {
+                     return -(double) blocked;
+                  }
+               };
+            } else {
+               blocking = new Function<Double, Double>() {
+                  @Override
+                  public Double apply(Double input) {
+                     return 0d;
+                  }
+               };
+            }
+         } else {
+            blocking = new Function<Double, Double>() {
+               @Override
+               public Double apply(Double input) {
+                  return 0d;
+               }
+            };
+         }
+         float blockingModifier = blocking.apply((double) p_21241_).floatValue();
+         p_21241_ += blockingModifier;
+
+         Function<Double, Double> armor = new Function<Double, Double>() {
+            @Override
+            public Double apply(Double input) {
+               return -(input - LivingEntity.this.m_21161_(p_21240_, input.floatValue()));
+            }
+         };
+         float armorModifier = armor.apply((double) p_21241_).floatValue();
+         p_21241_ += armorModifier;
+
+         Function<Double, Double> resistance = new Function<Double, Double>() {
+            @Override
+            public Double apply(Double input) {
+               if (!p_21240_.m_269533_(DamageTypeTags.f_268437_) && LivingEntity.this.m_21023_(MobEffects.f_19606_) && !p_21240_.m_269533_(DamageTypeTags.f_268630_)) {
+                  int i = (LivingEntity.this.m_21124_(MobEffects.f_19606_).m_19564_() + 1) * 5;
+                  int j = 25 - i;
+                  float f1 = input.floatValue() * (float) j;
+                  return -(input - (f1 / 25.0F));
+               }
+               return -0.0;
+            }
+         };
+         float resistanceModifier = resistance.apply((double) p_21241_).floatValue();
+         p_21241_ += resistanceModifier;
+
+         // Mohist start
+         float mohist = this.m_6515_(p_21240_, p_21241_);
+         float magicModifier = -(p_21241_ - mohist);
+         Function<Double, Double> magic = new Function<Double, Double>() {
+            @Override
+            public Double apply(Double input) {
+               return (double) magicModifier;
+            }
+         };
+         p_21241_ += magicModifier;
+         // Mohist end
+
+         Function<Double, Double> absorption = new Function<Double, Double>() {
+            @Override
+            public Double apply(Double input) {
+               return -(Math.max(input - Math.max(input - LivingEntity.this.m_6103_(), 0.0F), 0.0F));
+            }
+         };
+         float absorptionModifier = absorption.apply((double) p_21241_).floatValue();
+
+         EntityDamageEvent event = CraftEventFactory.handleLivingEntityDamageEvent(this, p_21240_, originalDamage, hardHatModifier, blockingModifier, armorModifier, resistanceModifier, magicModifier, absorptionModifier, hardHat, blocking, armor, resistance, magic, absorption);
+         if (p_21240_.m_7639_() instanceof Player) {
+            ((Player) p_21240_.m_7639_()).m_36334_(); // Moved from Player in order to make the cooldown reset get called after the damage event is fired
+         }
+         if (event.isCancelled()) {
+            this.canDamage.set(false);
+            return false;
+         }
+
+         p_21241_ = (float) event.getFinalDamage();
+
+         // Resistance
+         if (event.getDamage(EntityDamageEvent.DamageModifier.RESISTANCE) < 0) {
+            float f3 = (float) -event.getDamage(EntityDamageEvent.DamageModifier.RESISTANCE);
+            if (f3 > 0.0F && f3 < 3.4028235E37F) {
+               if (this instanceof ServerPlayer) {
+                  ((ServerPlayer) this).m_36222_(Stats.f_12934_, Math.round(f3 * 10.0F));
+               } else if (p_21240_.m_7639_() instanceof ServerPlayer serverPlayer) {
+                  serverPlayer.m_36222_(Stats.f_12930_, Math.round(f3 * 10.0F));
+               }
+            }
+         }
+
+         // Apply damage to helmet
+         if (p_21240_.m_269533_(DamageTypeTags.f_268627_) && !this.m_6844_(EquipmentSlot.HEAD).m_41619_()) {
+            this.m_142642_(p_21240_, p_21241_);
+         }
+
+         // Apply damage to armor
+         if (!p_21240_.m_269533_(DamageTypeTags.f_268490_)) {
+            float armorDamage = (float) (event.getDamage() + event.getDamage(EntityDamageEvent.DamageModifier.BLOCKING) + event.getDamage(EntityDamageEvent.DamageModifier.HARD_HAT));
+            this.m_6472_(p_21240_, armorDamage);
+         }
+
+         // Apply blocking code // PAIL: steal from above
+         if (event.getDamage(EntityDamageEvent.DamageModifier.BLOCKING) < 0) {
+            this.m_9236_().m_7605_(this, (byte) 29); // SPIGOT-4635 - shield damage sound
+            if (shieldTakesDamage) {
+               this.m_7909_((float) -event.getDamage(EntityDamageEvent.DamageModifier.BLOCKING));
+            }
+            Entity entity = p_21240_.m_7640_();
+            if (entity instanceof LivingEntity living) {
+               this.m_6728_(living);
+            }
+         }
+
+         absorptionModifier = (float) -event.getDamage(EntityDamageEvent.DamageModifier.ABSORPTION);
+         this.m_7911_(Math.max(this.m_6103_() - absorptionModifier, 0.0F));
+         float f2 = absorptionModifier;
+
+         if (f2 > 0.0F && f2 < 3.4028235E37F && this instanceof Player) {
+            ((Player) this).m_36222_(Stats.f_12933_, Math.round(f2 * 10.0F));
+         }
+         if (f2 > 0.0F && f2 < 3.4028235E37F && p_21240_.m_7639_() instanceof ServerPlayer) {
+            ((ServerPlayer) p_21240_.m_7639_()).m_36222_(Stats.f_12929_, Math.round(f2 * 10.0F));
+         }
+
+         p_21241_ = net.minecraftforge.common.ForgeHooks.onLivingDamage(this, p_21240_, p_21241_);
+
+         if (p_21241_ > 0 || !human) {
+            if (human) {
+               // PAIL: Be sure to drag all this code from the Player subclass each update.
+               ((Player) this).exhaustionReason(org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.DAMAGED);
+               ((Player) this).m_36399_(p_21240_.m_19377_());
+               if (p_21241_ < 3.4028235E37F) {
+                  ((Player) this).m_36222_(Stats.f_12931_, Math.round(p_21241_ * 10.0F));
+               }
+            }
+            // CraftBukkit end
+            this.m_21231_().m_289194_(p_21240_, p_21241_);
+            // Mohist start
+            this.damage.set(p_21241_);
+            // Mohist end
+            // CraftBukkit start
+            if (!human) {
+               this.m_7911_(this.m_6103_() - p_21241_);
+            }
+            this.m_146852_(GameEvent.f_223706_, p_21240_.m_7639_());
+            this.canDamage.set(true);
+            return true;
+         } else {
+            // Duplicate triggers if blocking
+            if (event.getDamage(EntityDamageEvent.DamageModifier.BLOCKING) < 0) {
+               if (this instanceof ServerPlayer serverPlayer) {
+                  CriteriaTriggers.f_10574_.m_35174_(serverPlayer, p_21240_, p_21241_, originalDamage, true);
+                  f2 = (float) -event.getDamage(EntityDamageEvent.DamageModifier.BLOCKING);
+                  if (f2 > 0.0F && f2 < 3.4028235E37F) {
+                     serverPlayer.m_36222_(Stats.f_12932_, Math.round(originalDamage * 10.0F));
+                  }
+               }
+               if (p_21240_.m_7639_() instanceof ServerPlayer serverPlayer) {
+                  CriteriaTriggers.f_10573_.m_60112_(serverPlayer, this, p_21240_, p_21241_, originalDamage, true);
+               }
+               this.canDamage.set(false);
+               return false ;
+            } else {
+                return originalDamage > 0;
+            }
+            // CraftBukkit end
          }
       }
+      this.canDamage.set(false);
+      return false;
    }
 
    public CombatTracker m_21231_() {
@@ -1627,9 +_,25 @@
    }
 
    public final void m_21317_(int p_21318_) {
+         ArrowBodyCountChangeEvent event = CraftEventFactory.callArrowBodyCountChangeEvent( this, m_21234_(), p_21318_, false);
+         if(event.isCancelled()){
+            return;
+         }
+
+      // CraftBukkit end
       this.f_19804_.m_135381_(f_20940_, p_21318_);
    }
 
+   public final void setArrowCount(int pCount, boolean flag){
+      ArrowBodyCountChangeEvent event = CraftEventFactory.callArrowBodyCountChangeEvent( this, m_21234_(), pCount, flag );
+      if(event.isCancelled()){
+         return;
+      }
+
+      this.f_19804_.m_135381_(f_20940_, event.getNewAmount());
+   }
+   // CraftBukkit end
+
    public final int m_21235_() {
       return this.f_19804_.m_135370_(f_20941_);
    }
@@ -1651,6 +_,8 @@
    }
 
    public void m_21011_(InteractionHand p_21012_, boolean p_21013_) {
+      ItemStack stack = this.m_21120_(p_21012_);
+      if (!stack.m_41619_() && stack.onEntitySwing(this)) return;
       if (!this.f_20911_ || this.f_20913_ >= this.m_21304_() / 2 || this.f_20913_ < 0) {
          this.f_20913_ = -1;
          this.f_20911_ = true;
@@ -1761,8 +_,10 @@
 
    private void m_21312_() {
       ItemStack itemstack = this.m_6844_(EquipmentSlot.OFFHAND);
-      this.m_8061_(EquipmentSlot.OFFHAND, this.m_6844_(EquipmentSlot.MAINHAND));
-      this.m_8061_(EquipmentSlot.MAINHAND, itemstack);
+      var event = net.minecraftforge.common.ForgeHooks.onLivingSwapHandItems(this);
+      if (event.isCanceled()) return;
+      this.m_8061_(EquipmentSlot.OFFHAND, event.getItemSwappedToOffHand());
+      this.m_8061_(EquipmentSlot.MAINHAND, event.getItemSwappedToMainHand());
    }
 
    protected void m_6088_() {
@@ -1770,6 +_,7 @@
    }
 
    protected void m_21203_() {
+      if (!this.f_20911_ && this.f_20913_ == 0) return; // Mirai
       int i = this.m_21304_();
       if (this.f_20911_) {
          ++this.f_20913_;
@@ -1862,6 +_,12 @@
 
    public abstract ItemStack m_6844_(EquipmentSlot p_21127_);
 
+   // CraftBukkit start
+   public void setItemSlot(EquipmentSlot enumitemslot, ItemStack itemstack, boolean silent) {
+      this.m_8061_(enumitemslot, itemstack);
+   }
+   // CraftBukkit end
+
    public abstract void m_8061_(EquipmentSlot p_21036_, ItemStack p_21037_);
 
    protected void m_181122_(ItemStack p_181123_) {
@@ -1905,6 +_,8 @@
       return 1.0F;
    }
 
+   public float getSoundVolume0() { return m_6121_(); } // Mohist - avoid a lot of AT
+
    public float m_6100_() {
       return this.m_6162_() ? (this.f_19796_.m_188501_() - this.f_19796_.m_188501_()) * 0.2F + 1.5F : (this.f_19796_.m_188501_() - this.f_19796_.m_188501_()) * 0.2F + 1.0F;
    }
@@ -1955,14 +_,17 @@
       }
 
       this.f_19812_ = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
+   @Deprecated // FORGE: use sinkInFluid instead
    protected void m_21208_() {
-      this.m_20256_(this.m_20184_().m_82520_(0.0D, (double)-0.04F, 0.0D));
+      this.sinkInFluid(net.minecraftforge.common.ForgeMod.WATER_TYPE.get());
    }
 
+   @Deprecated // FORGE: use jumpInFluid instead
    protected void m_203347_(TagKey<Fluid> p_204043_) {
-      this.m_20256_(this.m_20184_().m_82520_(0.0D, (double)0.04F, 0.0D));
+      this.m_20256_(this.m_20184_().m_82520_(0.0D, (double)0.04F * this.m_21051_(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).m_22135_(), 0.0D));
    }
 
    protected float m_6108_() {
@@ -1976,13 +_,18 @@
    public void m_7023_(Vec3 p_21280_) {
       if (this.m_6109_()) {
          double d0 = 0.08D;
+         AttributeInstance gravity = this.m_21051_(net.minecraftforge.common.ForgeMod.ENTITY_GRAVITY.get());
          boolean flag = this.m_20184_().f_82480_ <= 0.0D;
          if (flag && this.m_21023_(MobEffects.f_19591_)) {
-            d0 = 0.01D;
+            if (!gravity.m_22109_(SLOW_FALLING)) gravity.m_22118_(SLOW_FALLING);
+         } else if (gravity.m_22109_(SLOW_FALLING)) {
+            gravity.m_22130_(SLOW_FALLING);
          }
+         d0 = gravity.m_22135_();
 
          FluidState fluidstate = this.m_9236_().m_6425_(this.m_20183_());
-         if (this.m_20069_() && this.m_6129_() && !this.m_203441_(fluidstate)) {
+         if ((this.m_20069_() || (this.isInFluidType(fluidstate) && fluidstate.getFluidType() != net.minecraftforge.common.ForgeMod.LAVA_TYPE.get())) && this.m_6129_() && !this.m_203441_(fluidstate)) {
+            if (this.m_20069_() || (this.isInFluidType(fluidstate) && !this.moveInFluid(fluidstate, p_21280_, d0))) {
             double d9 = this.m_20186_();
             float f4 = this.m_20142_() ? 0.9F : this.m_6108_();
             float f5 = 0.02F;
@@ -2004,6 +_,7 @@
                f4 = 0.96F;
             }
 
+            f5 *= (float)this.m_21051_(net.minecraftforge.common.ForgeMod.SWIM_SPEED.get()).m_22135_();
             this.m_19920_(f5, p_21280_);
             this.m_6478_(MoverType.SELF, this.m_20184_());
             Vec3 vec36 = this.m_20184_();
@@ -2017,6 +_,7 @@
             if (this.f_19862_ && this.m_20229_(vec32.f_82479_, vec32.f_82480_ + (double)0.6F - this.m_20186_() + d9, vec32.f_82481_)) {
                this.m_20334_(vec32.f_82479_, (double)0.3F, vec32.f_82481_);
             }
+            }
          } else if (this.m_20077_() && this.m_6129_() && !this.m_203441_(fluidstate)) {
             double d8 = this.m_20186_();
             this.m_19920_(0.02F, p_21280_);
@@ -2074,12 +_,12 @@
                }
             }
 
-            if (this.m_20096_() && !this.m_9236_().f_46443_) {
+            if (this.m_20096_() && !this.f_19853_.f_46443_ && m_20291_(7) && !CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) {
                this.m_20115_(7, false);
             }
          } else {
             BlockPos blockpos = this.m_20099_();
-            float f2 = this.m_9236_().m_8055_(blockpos).m_60734_().m_49958_();
+            float f2 = this.f_19853_.m_8055_(this.m_20099_()).getFriction(f_19853_, this.m_20099_(), this);
             float f3 = this.m_20096_() ? f2 * 0.91F : 0.91F;
             Vec3 vec35 = this.m_21074_(p_21280_, f2);
             double d2 = vec35.f_82480_;
@@ -2175,7 +_,7 @@
          double d0 = Mth.m_14008_(p_21298_.f_82479_, (double)-0.15F, (double)0.15F);
          double d1 = Mth.m_14008_(p_21298_.f_82481_, (double)-0.15F, (double)0.15F);
          double d2 = Math.max(p_21298_.f_82480_, (double)-0.15F);
-         if (d2 < 0.0D && !this.m_146900_().m_60713_(Blocks.f_50616_) && this.m_5791_() && this instanceof Player) {
+         if (d2 < 0.0D && !this.m_146900_().isScaffolding(this) && this.m_5791_() && this instanceof Player) {
             d2 = 0.0D;
          }
 
@@ -2207,6 +_,7 @@
    }
 
    public void m_8119_() {
+      if (net.minecraftforge.common.ForgeHooks.onLivingTick(this)) return;
       super.m_8119_();
       this.m_21329_();
       this.m_21333_();
@@ -2328,7 +_,7 @@
 
    }
 
-   private void m_21315_() {
+   public void m_21315_() {
       Map<EquipmentSlot, ItemStack> map = this.m_21319_();
       if (map != null) {
          this.m_21091_(map);
@@ -2358,6 +_,7 @@
 
          ItemStack itemstack1 = this.m_6844_(equipmentslot);
          if (this.m_246525_(itemstack, itemstack1)) {
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.LivingEquipmentChangeEvent(this, equipmentslot, itemstack, itemstack1));
             if (map == null) {
                map = Maps.newEnumMap(EquipmentSlot.class);
             }
@@ -2503,6 +_,9 @@
       this.m_9236_().m_46473_().m_6180_("jump");
       if (this.f_20899_ && this.m_6129_()) {
          double d7;
+         net.minecraftforge.fluids.FluidType fluidType = this.getMaxHeightFluidType();
+         if (!fluidType.isAir()) d7 = this.getFluidTypeHeight(fluidType);
+         else
          if (this.m_20077_()) {
             d7 = this.m_204036_(FluidTags.f_13132_);
          } else {
@@ -2513,15 +_,17 @@
          double d8 = this.m_20204_();
          if (!flag || this.m_20096_() && !(d7 > d8)) {
             if (!this.m_20077_() || this.m_20096_() && !(d7 > d8)) {
+               if (fluidType.isAir() || this.m_20096_() && !(d7 > d8)) {
                if ((this.m_20096_() || flag && d7 <= d8) && this.f_20954_ == 0) {
                   this.m_6135_();
                   this.f_20954_ = 10;
                }
+               } else this.jumpInFluid(fluidType);
             } else {
-               this.m_203347_(FluidTags.f_13132_);
+               this.jumpInFluid(net.minecraftforge.common.ForgeMod.LAVA_TYPE.get());
             }
          } else {
-            this.m_203347_(FluidTags.f_13131_);
+            this.jumpInFluid(net.minecraftforge.common.ForgeMod.WATER_TYPE.get());
          }
       } else {
          this.f_20954_ = 0;
@@ -2587,9 +_,12 @@
    }
 
    private void m_21323_() {
+      if (!this.m_21255_()) return; // Mirai
       boolean flag = this.m_20291_(7);
       if (flag && !this.m_20096_() && !this.m_20159_() && !this.m_21023_(MobEffects.f_19620_)) {
          ItemStack itemstack = this.m_6844_(EquipmentSlot.CHEST);
+         flag = itemstack.canElytraFly(this) && itemstack.elytraFlightTick(this, this.f_20937_);
+         if (false) //Forge: Moved to ElytraItem
          if (itemstack.m_150930_(Items.f_42741_) && ElytraItem.m_41140_(itemstack)) {
             flag = true;
             int i = this.f_20937_ + 1;
@@ -2611,6 +_,7 @@
       }
 
       if (!this.m_9236_().f_46443_) {
+         if (flag != this.m_20291_(7) && !CraftEventFactory.callToggleGlideEvent(this, flag).isCancelled()) // CraftBukkit
          this.m_20115_(7, flag);
       }
 
@@ -2620,12 +_,12 @@
    }
 
    protected void m_6138_() {
-      if (this.m_9236_().m_5776_()) {
-         this.m_9236_().m_142425_(EntityTypeTest.m_156916_(Player.class), this.m_20191_(), EntitySelector.m_20421_(this)).forEach(this::m_7324_);
+      if (this.f_19853_.m_5776_()) {
+         this.f_19853_.m_142425_(EntityTypeTest.m_156916_(Player.class), this.m_20191_(), EntitySelector.m_20421_(this)).forEach(this::m_7324_);
       } else {
-         List<Entity> list = this.m_9236_().m_6249_(this, this.m_20191_(), EntitySelector.m_20421_(this));
+         List<Entity> list = this.f_19853_.m_6249_(this, this.m_20191_(), EntitySelector.m_20421_(this));
          if (!list.isEmpty()) {
-            int i = this.m_9236_().m_46469_().m_46215_(GameRules.f_46149_);
+            int i = this.f_19853_.m_46469_().m_46215_(GameRules.f_46149_);
             if (i > 0 && list.size() > i - 1 && this.f_19796_.m_188503_(4) == 0) {
                int j = 0;
 
@@ -2651,7 +_,7 @@
 
    protected void m_21071_(AABB p_21072_, AABB p_21073_) {
       AABB aabb = p_21072_.m_82367_(p_21073_);
-      List<Entity> list = this.m_9236_().m_45933_(this, aabb);
+      List<Entity> list = this.f_19853_.m_45933_(this, aabb);
       if (!list.isEmpty()) {
          for(int i = 0; i < list.size(); ++i) {
             Entity entity = list.get(i);
@@ -2666,7 +_,7 @@
          this.f_20938_ = 0;
       }
 
-      if (!this.m_9236_().f_46443_ && this.f_20938_ <= 0) {
+      if (!this.f_19853_.f_46443_ && this.f_20938_ <= 0) {
          this.m_21155_(4, false);
       }
 
@@ -2686,7 +_,7 @@
    public void m_8127_() {
       Entity entity = this.m_20202_();
       super.m_8127_();
-      if (entity != null && entity != this.m_20202_() && !this.m_9236_().f_46443_) {
+      if (entity != null && entity != this.m_20202_() && !this.f_19853_.f_46443_) {
          this.m_21028_(entity);
       }
 
@@ -2726,14 +_,14 @@
    }
 
    public void m_7938_(Entity p_21030_, int p_21031_) {
-      if (!p_21030_.m_213877_() && !this.m_9236_().f_46443_ && (p_21030_ instanceof ItemEntity || p_21030_ instanceof AbstractArrow || p_21030_ instanceof ExperienceOrb)) {
-         ((ServerLevel)this.m_9236_()).m_7726_().m_8445_(p_21030_, new ClientboundTakeItemEntityPacket(p_21030_.m_19879_(), this.m_19879_(), p_21031_));
+      if (!p_21030_.m_213877_() && !this.f_19853_.f_46443_ && (p_21030_ instanceof ItemEntity || p_21030_ instanceof AbstractArrow || p_21030_ instanceof ExperienceOrb)) {
+         ((ServerLevel)this.f_19853_).m_7726_().m_8445_(p_21030_, new ClientboundTakeItemEntityPacket(p_21030_.m_19879_(), this.m_19879_(), p_21031_));
       }
 
    }
 
    public boolean m_142582_(Entity p_147185_) {
-      if (p_147185_.m_9236_() != this.m_9236_()) {
+      if (p_147185_.f_19853_ != this.f_19853_) {
          return false;
       } else {
          Vec3 vec3 = new Vec3(this.m_20185_(), this.m_20188_(), this.m_20189_());
@@ -2741,7 +_,7 @@
          if (vec31.m_82554_(vec3) > 128.0D) {
             return false;
          } else {
-            return this.m_9236_().m_45547_(new ClipContext(vec3, vec31, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, this)).m_6662_() == HitResult.Type.MISS;
+            return this.f_19853_.m_45547_(new ClipContext(vec3, vec31, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, this)).m_6662_() == HitResult.Type.MISS;
          }
       }
    }
@@ -2760,12 +_,19 @@
    }
 
    public boolean m_6087_() {
-      return !this.m_213877_();
+      return !this.m_213877_() && this.collides; // CraftBukkit
    }
 
    public boolean m_6094_() {
-      return this.m_6084_() && !this.m_5833_() && !this.m_6147_();
-   }
+      return this.m_6084_() && !this.m_5833_() && !this.m_6147_() && this.collides; // CraftBukkit
+   }
+
+   // CraftBukkit start - collidable API
+   @Override
+   public boolean canCollideWithBukkit(Entity entity) {
+     return this.m_6094_() && this.collides != this.collidableExemptions.contains(entity.m_20148_());
+   }
+   // CraftBukkit end
 
    public float m_6080_() {
       return this.f_20885_;
@@ -2821,23 +_,28 @@
 
    private void m_21329_() {
       if (this.m_6117_()) {
-         if (ItemStack.m_41656_(this.m_21120_(this.m_7655_()), this.f_20935_)) {
-            this.f_20935_ = this.m_21120_(this.m_7655_());
+         ItemStack itemStack = this.m_21120_(this.m_7655_());
+         if (net.minecraftforge.common.ForgeHooks.canContinueUsing(this.f_20935_, itemStack)) {
+            this.f_20935_ = itemStack;
+         }
+         if (itemStack == this.f_20935_) {
             this.m_142106_(this.f_20935_);
          } else {
             this.m_5810_();
          }
       }
-
    }
 
    protected void m_142106_(ItemStack p_147201_) {
-      p_147201_.m_41731_(this.m_9236_(), this, this.m_21212_());
+      if (!p_147201_.m_41619_())
+         this.f_20936_ = net.minecraftforge.event.ForgeEventFactory.onItemUseTick(this, p_147201_, this.m_21212_());
+      if (this.m_21212_() > 0)
+         p_147201_.m_41731_(this.f_19853_, this, this.m_21212_());
       if (this.m_21332_()) {
          this.m_21137_(p_147201_, 5);
       }
 
-      if (--this.f_20936_ == 0 && !this.m_9236_().f_46443_ && !p_147201_.m_41781_()) {
+      if (--this.f_20936_ <= 0 && !this.f_19853_.f_46443_ && !p_147201_.m_41781_()) {
          this.m_8095_();
       }
 
@@ -2845,7 +_,7 @@
 
    private boolean m_21332_() {
       int i = this.m_21212_();
-      FoodProperties foodproperties = this.f_20935_.m_41720_().m_41473_();
+      FoodProperties foodproperties = this.f_20935_.getFoodProperties(this);
       boolean flag = foodproperties != null && foodproperties.m_38748_();
       flag |= i <= this.f_20935_.m_41779_() - 7;
       return flag && i % 4 == 0;
@@ -2875,8 +_,11 @@
    public void m_6672_(InteractionHand p_21159_) {
       ItemStack itemstack = this.m_21120_(p_21159_);
       if (!itemstack.m_41619_() && !this.m_6117_()) {
+         if (BanItem.check(itemstack)) return;
+         int duration = net.minecraftforge.event.ForgeEventFactory.onItemUseStart(this, itemstack, itemstack.m_41779_());
+         if (duration <= 0) return;
          this.f_20935_ = itemstack;
-         this.f_20936_ = itemstack.m_41779_();
+         this.f_20936_ = duration;
          if (!this.m_9236_().f_46443_) {
             this.m_21155_(1, true);
             this.m_21155_(2, p_21159_ == InteractionHand.OFF_HAND);
@@ -2937,6 +_,9 @@
          vec31 = vec31.m_82496_(-this.m_146909_() * ((float)Math.PI / 180F));
          vec31 = vec31.m_82524_(-this.m_146908_() * ((float)Math.PI / 180F));
          vec31 = vec31.m_82520_(this.m_20185_(), this.m_20188_(), this.m_20189_());
+         if (this.m_9236_() instanceof ServerLevel) //Forge: Fix MC-2518 spawnParticle is nooped on server, need to use server specific variant
+             ((ServerLevel) this.m_9236_()).m_8767_(new ItemParticleOption(ParticleTypes.f_123752_, p_21061_), vec31.f_82479_, vec31.f_82480_, vec31.f_82481_, 1, vec3.f_82479_, vec3.f_82480_ + 0.05D, vec3.f_82481_, 0.0D);
+         else
          this.m_9236_().m_7106_(new ItemParticleOption(ParticleTypes.f_123752_, p_21061_), vec31.f_82479_, vec31.f_82480_, vec31.f_82481_, vec3.f_82479_, vec3.f_82480_ + 0.05D, vec3.f_82481_);
       }
 
@@ -2950,9 +_,30 @@
          } else {
             if (!this.f_20935_.m_41619_() && this.m_6117_()) {
                this.m_21137_(this.f_20935_, 16);
-               ItemStack itemstack = this.f_20935_.m_41671_(this.m_9236_(), this);
-               if (itemstack != this.f_20935_) {
-                  this.m_21008_(interactionhand, itemstack);
+               ItemStack copy = this.f_20935_.m_41777_();
+               // CraftBukkit start - fire PlayerItemConsumeEvent
+               ItemStack itemstack;
+               if (this instanceof ServerPlayer) {
+                  org.bukkit.inventory.ItemStack craftItem = CraftItemStack.asBukkitCopy(copy);
+                  org.bukkit.inventory.EquipmentSlot hand = org.bukkit.craftbukkit.v1_20_R1.CraftEquipmentSlot.getHand(interactionhand);
+                  PlayerItemConsumeEvent event = new PlayerItemConsumeEvent((org.bukkit.entity.Player) this.getBukkitEntity(), craftItem, hand);
+                  f_19853_.getCraftServer().getPluginManager().callEvent(event);
+
+                  if (event.isCancelled()) {
+                     // Update client
+                     ((ServerPlayer) this).getBukkitEntity().updateInventory();
+                     ((ServerPlayer) this).getBukkitEntity().updateScaledHealth();
+                     return;
+                  }
+                  itemstack = (craftItem.equals(event.getItem())) ? this.f_20935_.m_41671_(this.m_9236_(), this) : CraftItemStack.asNMSCopy(event.getItem()).m_41671_(this.m_9236_(), this);
+               } else {
+                  itemstack = this.f_20935_.m_41671_(this.f_19853_, this);
+               }
+               // CraftBukkit end
+               ItemStack itemstackForge = net.minecraftforge.event.ForgeEventFactory.onItemUseFinish(this, copy, m_21212_(), itemstack);
+
+               if (itemstackForge != this.f_20935_) {
+                  this.m_21008_(interactionhand, itemstackForge);
                }
 
                this.m_5810_();
@@ -2976,7 +_,11 @@
 
    public void m_21253_() {
       if (!this.f_20935_.m_41619_()) {
+         if (!net.minecraftforge.event.ForgeEventFactory.onUseItemStop(this, f_20935_, this.m_21212_())) {
+            ItemStack copy = this instanceof Player ? f_20935_.m_41777_() : null;
          this.f_20935_.m_41674_(this.m_9236_(), this, this.m_21212_());
+           if (copy != null && f_20935_.m_41619_()) net.minecraftforge.event.ForgeEventFactory.onPlayerDestroyItem((Player)this, copy, m_7655_());
+         }
          if (this.f_20935_.m_41781_()) {
             this.m_21329_();
          }
@@ -2986,6 +_,7 @@
    }
 
    public void m_5810_() {
+      if (this.m_6117_() && !this.f_20935_.m_41619_()) this.f_20935_.onStopUsing(this, f_20936_);
       if (!this.m_9236_().f_46443_) {
          boolean flag = this.m_6117_();
          this.m_21155_(1, false);
@@ -3001,7 +_,7 @@
    public boolean m_21254_() {
       if (this.m_6117_() && !this.f_20935_.m_41619_()) {
          Item item = this.f_20935_.m_41720_();
-         if (item.m_6164_(this.f_20935_) != UseAnim.BLOCK) {
+         if (!this.f_20935_.canPerformAction(net.minecraftforge.common.ToolActions.SHIELD_BLOCK)) {
             return false;
          } else {
             return item.m_8105_(this.f_20935_) - this.f_20936_ >= 5;
@@ -3011,6 +_,51 @@
       }
    }
 
+   public HitResult getRayTrace(int maxDistance, ClipContext.Fluid fluidCollisionOption) {
+      if (maxDistance < 1 || maxDistance > 120) {
+         throw new IllegalArgumentException("maxDistance must be between 1-120");
+      }
+
+      Vec3 start = new Vec3(m_20185_(), m_20186_() + m_20192_(), m_20189_());
+      org.bukkit.util.Vector dir = getBukkitEntity().getLocation().getDirection().multiply(maxDistance);
+      Vec3 end = new Vec3(start.f_82479_ + dir.getX(), start.f_82480_ + dir.getY(), start.f_82481_ + dir.getZ());
+      ClipContext raytrace = new ClipContext(start, end, ClipContext.Block.OUTLINE, fluidCollisionOption, this);
+
+      return this.m_9236_().m_45547_(raytrace);
+   }
+
+   public @Nullable EntityHitResult getTargetEntity(int maxDistance) {
+      if (maxDistance < 1 || maxDistance > 120) {
+         throw new IllegalArgumentException("maxDistance must be between 1-120");
+      }
+
+      Vec3 start = this.m_20299_(1.0F);
+      Vec3 direction = this.m_20154_();
+      Vec3 end = start.m_82520_(direction.f_82479_ * maxDistance, direction.f_82480_ * maxDistance, direction.f_82481_ * maxDistance);
+
+      List<Entity> entityList = this.m_9236_().m_6249_(this, m_20191_().m_82363_(direction.f_82479_ * maxDistance, direction.f_82480_ * maxDistance, direction.f_82481_ * maxDistance).m_82377_(1.0D, 1.0D, 1.0D), EntitySelector.f_20408_.and(Entity::m_6087_));
+
+      double distance = 0.0D;
+      EntityHitResult result = null;
+
+      for (Entity entity : entityList) {
+         final double inflationAmount = (double) entity.m_6143_();
+         AABB aabb = entity.m_20191_().m_82377_(inflationAmount, inflationAmount, inflationAmount);
+         Optional<Vec3> rayTraceResult = aabb.m_82371_(start, end);
+
+         if (rayTraceResult.isPresent()) {
+            Vec3 rayTrace = rayTraceResult.get();
+            double distanceTo = start.m_82557_(rayTrace);
+            if (distanceTo < distance || distance == 0.0D) {
+               result = new EntityHitResult(entity, rayTrace);
+               distance = distanceTo;
+            }
+         }
+      }
+
+      return result;
+   }
+
    public boolean m_5791_() {
       return this.m_6144_();
    }
@@ -3027,6 +_,8 @@
       return this.f_20937_;
    }
 
+   public AtomicReference<PlayerTeleportEvent.TeleportCause> randomTeleportCause = new AtomicReference<>(PlayerTeleportEvent.TeleportCause.UNKNOWN);
+
    public boolean m_20984_(double p_20985_, double p_20986_, double p_20987_, boolean p_20988_) {
       double d0 = this.m_20185_();
       double d1 = this.m_20186_();
@@ -3050,16 +_,37 @@
          }
 
          if (flag1) {
-            this.m_6021_(p_20985_, d3, p_20987_);
+            // CraftBukkit start - Teleport event
+            this.m_6034_(p_20985_, d3, p_20987_);
             if (level.m_45786_(this) && !level.m_46855_(this.m_20191_())) {
                flag = true;
             }
+            // now revert and call event if the teleport place is valid
+            this.m_6034_(d0, d1, d2);
+
+            if (flag1) {
+               if (!(this instanceof ServerPlayer)) {
+                  EntityTeleportEvent teleport = new EntityTeleportEvent(this.getBukkitEntity(), new Location(this.f_19853_.getWorld(), d0, d1, d2), new Location(this.f_19853_.getWorld(), p_20985_, d3, p_20987_));
+                  this.f_19853_.getCraftServer().getPluginManager().callEvent(teleport);
+                  if (!teleport.isCancelled()) {
+                     Location to = teleport.getTo();
+                     this.m_6021_(to.getX(), to.getY(), to.getZ());
+                  } else {
+                     return false;
+                  }
+               } else {
+                  // player teleport event is called in the underlining code
+                  if (((ServerPlayer) this).f_8906_.teleport(p_20985_, d3, p_20987_, this.m_146908_(), this.m_146909_(), java.util.Collections.emptySet(), randomTeleportCause.getAndSet(PlayerTeleportEvent.TeleportCause.UNKNOWN))) {
+                     return false;
+                  }
+               }
+            }
+            // CraftBukkit end
          }
       }
 
       if (!flag) {
-         this.m_6021_(d0, d1, d2);
-         return false;
+         return false; // CraftBukkit
       } else {
          if (p_20988_) {
             level.m_7605_(this, (byte)46);
@@ -3069,8 +_,14 @@
             ((PathfinderMob)this).m_21573_().m_26573_();
          }
 
-         return true;
+         return true; // CraftBukkit
       }
+   }
+
+   public Optional<Boolean> randomTeleport(double pX, double pY, double pZ, boolean pBroadcastTeleport, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause) {
+      randomTeleportCause.set(cause);
+      boolean v = m_20984_(pX, pY, pZ, pBroadcastTeleport);
+      return Optional.of(v);
    }
 
    public boolean m_5801_() {
@@ -3117,7 +_,14 @@
       this.f_19804_.m_135381_(f_20942_, Optional.empty());
    }
 
+   // Mohist start
+   public boolean mohist$fauxSleeping = false;
    public boolean m_5803_() {
+      if (mohist$fauxSleeping && this instanceof ServerPlayer player) {
+         mohist$fauxSleeping = false;
+         return this.m_21257_().isPresent() || player.fauxSleeping;
+      }
+      // Mohist end
       return this.m_21257_().isPresent();
    }
 
@@ -3127,8 +_,8 @@
       }
 
       BlockState blockstate = this.m_9236_().m_8055_(p_21141_);
-      if (blockstate.m_60734_() instanceof BedBlock) {
-         this.m_9236_().m_7731_(p_21141_, blockstate.m_61124_(BedBlock.f_49441_, Boolean.valueOf(true)), 3);
+      if (blockstate.isBed(m_9236_(), p_21141_, this)) {
+         blockstate.setBedOccupied(m_9236_(), p_21141_, this, true);
       }
 
       this.m_20124_(Pose.SLEEPING);
@@ -3144,16 +_,16 @@
 
    private boolean m_21334_() {
       return this.m_21257_().map((p_289310_) -> {
-         return this.m_9236_().m_8055_(p_289310_).m_60734_() instanceof BedBlock;
+         return net.minecraftforge.event.ForgeEventFactory.fireSleepingLocationCheck(this, p_289310_);
       }).orElse(false);
    }
 
    public void m_5796_() {
       this.m_21257_().filter(this.m_9236_()::m_46805_).ifPresent((p_261435_) -> {
          BlockState blockstate = this.m_9236_().m_8055_(p_261435_);
-         if (blockstate.m_60734_() instanceof BedBlock) {
+         if (blockstate.isBed(m_9236_(), p_261435_, this)) {
             Direction direction = blockstate.m_61143_(BedBlock.f_54117_);
-            this.m_9236_().m_7731_(p_261435_, blockstate.m_61124_(BedBlock.f_49441_, Boolean.valueOf(false)), 3);
+            blockstate.setBedOccupied(m_9236_(), p_261435_, this, false);
             Vec3 vec31 = BedBlock.m_260958_(this.m_6095_(), this.m_9236_(), p_261435_, direction, this.m_146908_()).orElseGet(() -> {
                BlockPos blockpos = p_261435_.m_7494_();
                return new Vec3((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.1D, (double)blockpos.m_123343_() + 0.5D);
@@ -3175,7 +_,9 @@
    @Nullable
    public Direction m_21259_() {
       BlockPos blockpos = this.m_21257_().orElse((BlockPos)null);
-      return blockpos != null ? BedBlock.m_49485_(this.m_9236_(), blockpos) : null;
+      if (blockpos == null) return Direction.UP;
+      BlockState state = this.m_9236_().m_8055_(blockpos);
+      return !state.isBed(m_9236_(), blockpos, this) ? Direction.UP : state.getBedDirection(m_9236_(), blockpos);
    }
 
    public boolean m_5830_() {
@@ -3191,7 +_,7 @@
    }
 
    public ItemStack m_6298_(ItemStack p_21272_) {
-      return ItemStack.f_41583_;
+      return net.minecraftforge.common.ForgeHooks.getProjectile(this, p_21272_, ItemStack.f_41583_);
    }
 
    public ItemStack m_5584_(Level p_21067_, ItemStack p_21068_) {
@@ -3211,9 +_,9 @@
    private void m_21063_(ItemStack p_21064_, Level p_21065_, LivingEntity p_21066_) {
       Item item = p_21064_.m_41720_();
       if (item.m_41472_()) {
-         for(Pair<MobEffectInstance, Float> pair : item.m_41473_().m_38749_()) {
+         for(Pair<MobEffectInstance, Float> pair : p_21064_.getFoodProperties(this).m_38749_()) {
             if (!p_21065_.f_46443_ && pair.getFirst() != null && p_21065_.f_46441_.m_188501_() < pair.getSecond()) {
-               p_21066_.m_7292_(new MobEffectInstance(pair.getFirst()));
+               p_21066_.addEffect(new MobEffectInstance(pair.getFirst()), EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
             }
          }
       }
@@ -3247,6 +_,64 @@
       this.m_21166_(p_21191_ == InteractionHand.MAIN_HAND ? EquipmentSlot.MAINHAND : EquipmentSlot.OFFHAND);
    }
 
+   /* ==== FORGE START ==== */
+   /***
+    * Removes all potion effects that have curativeItem as a curative item for its effect
+    * @param curativeItem The itemstack we are using to cure potion effects
+    */
+   public boolean curePotionEffects(ItemStack curativeItem) {
+      if (this.m_9236_().f_46443_)
+         return false;
+      boolean ret = false;
+      Iterator<MobEffectInstance> itr = this.f_20945_.values().iterator();
+      while (itr.hasNext()) {
+         MobEffectInstance effect = itr.next();
+         if (effect.isCurativeItem(curativeItem) && !net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.MobEffectEvent.Remove(this, effect))) {
+            this.m_7285_(effect);
+            itr.remove();
+            ret = true;
+            this.f_20948_ = true;
+         }
+      }
+      return ret;
+   }
+
+   /**
+    * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+    * currently only used in vanilla code by pigs.
+    *
+    * @param player The player who is riding the entity.
+    * @return If the player should orient the same direction as this entity.
+    */
+   public boolean shouldRiderFaceForward(Player player) {
+      return this instanceof net.minecraft.world.entity.animal.Pig;
+   }
+
+   private net.minecraftforge.common.util.LazyOptional<?>[] handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
+
+   @Override
+   public <T> net.minecraftforge.common.util.LazyOptional<T> getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable Direction facing) {
+      if (capability == net.minecraftforge.common.capabilities.ForgeCapabilities.ITEM_HANDLER && this.m_6084_()) {
+         if (facing == null) return handlers[2].cast();
+         else if (facing.m_122434_().m_122478_()) return handlers[0].cast();
+         else if (facing.m_122434_().m_122479_()) return handlers[1].cast();
+      }
+      return super.getCapability(capability, facing);
+   }
+
+   @Override
+   public void invalidateCaps() {
+      super.invalidateCaps();
+      for (int x = 0; x < handlers.length; x++)
+         handlers[x].invalidate();
+   }
+
+   @Override
+   public void reviveCaps() {
+      super.reviveCaps();
+      handlers = net.minecraftforge.items.wrapper.EntityEquipmentInvWrapper.create(this);
+   }
+
    public AABB m_6921_() {
       if (this.m_6844_(EquipmentSlot.HEAD).m_150930_(Items.f_42683_)) {
          float f = 0.5F;
@@ -3257,6 +_,8 @@
    }
 
    public static EquipmentSlot m_147233_(ItemStack p_147234_) {
+      final EquipmentSlot slot = p_147234_.getEquipmentSlot();
+      if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
       Equipable equipable = Equipable.m_269088_(p_147234_);
       return equipable != null ? equipable.m_40402_() : EquipmentSlot.MAINHAND;
    }
@@ -3327,6 +_,11 @@
       return this.m_21205_().m_41720_() instanceof AxeItem;
    }
 
+   /**
+    * Gets the value of the legacy {@link #maxUpStep} field.
+    * @deprecated Use {@link net.minecraftforge.common.extensions.IForgeEntity#getStepHeight()} to get the real step height value.
+    */
+   @Deprecated
    public float m_274421_() {
       float f = super.m_274421_();
       return this.m_6688_() instanceof Player ? Math.max(f, 1.0F) : f;
