--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -10,6 +_,8 @@
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -22,6 +_,8 @@
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.FloatTag;
 import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.NbtUtils;
+import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
 import net.minecraft.network.protocol.game.DebugPackets;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
@@ -29,6 +_,7 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.Mth;
@@ -53,6 +_,8 @@
 import net.minecraft.world.entity.ai.navigation.GroundPathNavigation;
 import net.minecraft.world.entity.ai.navigation.PathNavigation;
 import net.minecraft.world.entity.ai.sensing.Sensing;
+import net.minecraft.world.entity.decoration.HangingEntity;
+import net.minecraft.world.entity.decoration.LeashFenceKnotEntity;
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.monster.Enemy;
 import net.minecraft.world.entity.player.Player;
@@ -85,8 +_,18 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.EntityRemoveEvent.Cause;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.EntityTransformEvent.TransformReason;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
 
-public abstract class Mob extends LivingEntity implements EquipmentUser, Leashable, Targeting {
+public abstract class Mob extends LivingEntity implements EquipmentUser, Targeting {
     private static final EntityDataAccessor<Byte> DATA_MOB_FLAGS_ID = SynchedEntityData.defineId(Mob.class, EntityDataSerializers.BYTE);
     private static final int MOB_FLAG_NO_AI = 1;
     private static final int MOB_FLAG_LEFTHANDED = 2;
@@ -122,7 +_,7 @@
     private ItemStack bodyArmorItem = ItemStack.EMPTY;
     protected float bodyArmorDropChance;
     private boolean canPickUpLoot;
-    public boolean persistenceRequired;
+    private boolean persistenceRequired;
     private final Map<PathType, Float> pathfindingMalus = Maps.newEnumMap(PathType.class);
     @Nullable
     public ResourceKey<LootTable> lootTable;
@@ -131,6 +_,10 @@
     private Leashable.LeashData leashData;
     private BlockPos restrictCenter = BlockPos.ZERO;
     private float restrictRadius = -1.0F;
+    @Nullable
+    private MobSpawnType spawnType;
+    private boolean spawnCancelled = false;
+   	public boolean aware = true; // CraftBukkit
 
     protected Mob(EntityType<? extends Mob> p_21368_, Level p_21369_) {
         super(p_21368_, p_21369_);
@@ -237,8 +_,26 @@
     }
 
     public void setTarget(@Nullable LivingEntity p_21544_) {
-        this.target = p_21544_;
+        var event = net.minecraftforge.event.ForgeEventFactory.onLivingChangeTargetMob(this, p_21544_);
+        if (!event.isCanceled()) {
+            this.target = event.getNewTarget();
+        }
     }
+
+   // Mohist start
+   public boolean setTarget(@Nullable LivingEntity pTarget, EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+      if (getTarget() == pTarget) return false;
+       var event = net.minecraftforge.event.ForgeEventFactory.onLivingChangeTargetMob(this, pTarget);
+       event.setReason(reason);
+       event.setfireCBEvent(fireEvent);
+      if (!event.isCanceled()) {
+         this.target = event.getNewTarget();
+      } else {
+         return false;
+      }
+      return true;
+   }
+   // Mohist end
 
     @Override
     public boolean canAttackType(EntityType<?> p_21399_) {
@@ -367,6 +_,11 @@
         return null;
     }
 
+   // CraftBukkit start - Add delegate method
+   public SoundEvent getAmbientSound0() {
+      return getAmbientSound();
+   }
+   // CraftBukkit end
     @Override
     public void addAdditionalSaveData(CompoundTag p_21484_) {
         super.addAdditionalSaveData(p_21484_);
@@ -425,7 +_,13 @@
         if (this.isNoAi()) {
             p_21484_.putBoolean("NoAI", this.isNoAi());
         }
-    }
+
+        if (this.spawnType != null) {
+            p_21484_.putString("forge:spawn_type", this.spawnType.name());
+        }
+
+      	p_21484_.putBoolean("Bukkit.Aware", this.aware); // CraftBukkit
+   }
 
     @Override
     public void readAdditionalSaveData(CompoundTag p_21450_) {
@@ -484,6 +_,19 @@
         }
 
         this.setNoAi(p_21450_.getBoolean("NoAI"));
+
+        if (p_21450_.contains("forge:spawn_type")) {
+            try {
+                this.spawnType = MobSpawnType.valueOf(p_21450_.getString("forge:spawn_type"));
+            } catch (Exception ex) {
+                p_21450_.remove("forge:spawn_type");
+            }
+        }
+
+      // CraftBukkit start
+      if (p_21450_.contains("Bukkit.Aware")) {
+          this.aware = p_21450_.getBoolean("Bukkit.Aware");
+      }
     }
 
     @Override
@@ -535,7 +_,7 @@
     public void aiStep() {
         super.aiStep();
         this.level().getProfiler().push("looting");
-        if (!this.level().isClientSide && this.canPickUpLoot() && this.isAlive() && !this.dead && this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+        if (!this.level().isClientSide && this.canPickUpLoot() && this.isAlive() && !this.dead && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level(), this)) {
             Vec3i vec3i = this.getPickupReach();
 
             for (ItemEntity itementity : this.level()
@@ -555,40 +_,63 @@
 
     protected void pickUpItem(ItemEntity p_21471_) {
         ItemStack itemstack = p_21471_.getItem();
+      	this.entityitem = p_21471_;
         ItemStack itemstack1 = this.equipItemIfPossible(itemstack.copy());
         if (!itemstack1.isEmpty()) {
             this.onItemPickup(p_21471_);
             this.take(p_21471_, itemstack1.getCount());
             itemstack.shrink(itemstack1.getCount());
             if (itemstack.isEmpty()) {
-                p_21471_.discard();
-            }
-        }
-    }
-
-    public ItemStack equipItemIfPossible(ItemStack p_255842_) {
-        EquipmentSlot equipmentslot = this.getEquipmentSlotForItem(p_255842_);
-        ItemStack itemstack = this.getItemBySlot(equipmentslot);
-        boolean flag = this.canReplaceCurrentItem(p_255842_, itemstack);
-        if (equipmentslot.isArmor() && !flag) {
-            equipmentslot = EquipmentSlot.MAINHAND;
-            itemstack = this.getItemBySlot(equipmentslot);
-            flag = itemstack.isEmpty();
-        }
-
-        if (flag && this.canHoldItem(p_255842_)) {
-            double d0 = (double)this.getEquipmentDropChance(equipmentslot);
-            if (!itemstack.isEmpty() && (double)Math.max(this.random.nextFloat() - 0.1F, 0.0F) < d0) {
-                this.spawnAtLocation(itemstack);
-            }
-
-            ItemStack itemstack1 = equipmentslot.limit(p_255842_);
+            p_21471_.putRemoveCause(Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
+            p_21471_.discard();
+         }
+      }
+
+   }
+
+
+   public ItemEntity entityitem = null;
+   public ItemStack equipItemIfPossible(ItemStack p_255842_) {
+      EquipmentSlot equipmentslot = getEquipmentSlotForItem(p_255842_);
+      ItemStack itemstack = this.getItemBySlot(equipmentslot);
+      boolean flag = this.canReplaceCurrentItem(p_255842_, itemstack);
+      if (equipmentslot.isArmor() && !flag) {
+         equipmentslot = EquipmentSlot.MAINHAND;
+         itemstack = this.getItemBySlot(equipmentslot);
+         flag = this.canReplaceCurrentItem(p_255842_, itemstack);
+      }
+
+      // CraftBukkit start
+      boolean canPickup = flag && this.canHoldItem(itemstack);
+      if (entityitem != null) {
+         canPickup = !CraftEventFactory.callEntityPickupItemEvent(this, entityitem, 0, !canPickup).isCancelled();
+      }
+      if (canPickup) {
+         // CraftBukkit end
+         double d0 = (double)this.getEquipmentDropChance(equipmentslot);
+         if (!itemstack.isEmpty() && (double)Math.max(this.random.nextFloat() - 0.1F, 0.0F) < d0) {
+            this.forceDrops = true; // CraftBukkit
+            this.spawnAtLocation(itemstack);
+            this.forceDrops = false; // CraftBukkit
+         }
+
+         if (equipmentslot.isArmor() && p_255842_.getCount() > 1) {
+            ItemStack itemstack1 = p_255842_.copyWithCount(1);
             this.setItemSlotAndDropWhenKilled(equipmentslot, itemstack1);
             return itemstack1;
-        } else {
-            return ItemStack.EMPTY;
-        }
-    }
+         } else {
+            this.setItemSlotAndDropWhenKilled(equipmentslot, p_255842_);
+            return p_255842_;
+         }
+      } else {
+         return ItemStack.EMPTY;
+      }
+   }
+
+   public ItemStack equipItemIfPossible(ItemStack itemstack, ItemEntity entityitem) {
+      this.entityitem = entityitem;
+      return equipItemIfPossible(itemstack);
+   }
 
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot p_21469_, ItemStack p_21470_) {
         this.setItemSlot(p_21469_, p_21470_);
@@ -699,17 +_,27 @@
             this.discard();
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity entity = this.level().getNearestPlayer(this, -1.0);
+            var result = net.minecraftforge.event.ForgeEventFactory.canEntityDespawn(this, (ServerLevel)this.level());
+            if (result == net.minecraftforge.eventbus.api.Event.Result.DENY) {
+                noActionTime = 0;
+                entity = null;
+            } else if (result == net.minecraftforge.eventbus.api.Event.Result.ALLOW) {
+                this.discard();
+                entity = null;
+            }
             if (entity != null) {
                 double d0 = entity.distanceToSqr(this);
                 int i = this.getType().getCategory().getDespawnDistance();
                 int j = i * i;
                 if (d0 > (double)j && this.removeWhenFarAway(d0)) {
+               	this.putRemoveCause(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     this.discard();
                 }
 
                 int k = this.getType().getCategory().getNoDespawnDistance();
                 int l = k * k;
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double)l && this.removeWhenFarAway(d0)) {
+               		this.putRemoveCause(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     this.discard();
                 } else if (d0 < (double)l) {
                     this.noActionTime = 0;
@@ -1121,6 +_,16 @@
         }
     }
 
+    /**
+     * Forge: Override-Only, call via ForgeEventFactory.onFinalizeSpawn.<br>
+     * Overrides are allowed. Do not wrap super calls within override (as that will cause stack overflows).<br>
+     * Vanilla calls are replaced with a transformer, and are not visible in source.<br>
+     * <p>
+     * Be certain to either call super.finalizeSpawn or set the {@link #spawnType} field from within your override.
+     * @see {@link net.minecraftforge.event.ForgeEventFactory#onFinalizeSpawn onFinalizeSpawn} for additional documentation.
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     @Nullable
     public SpawnGroupData finalizeSpawn(ServerLevelAccessor p_21434_, DifficultyInstance p_21435_, MobSpawnType p_21436_, @Nullable SpawnGroupData p_21437_) {
         RandomSource randomsource = p_21434_.getRandom();
@@ -1132,6 +_,7 @@
         }
 
         this.setLeftHanded(randomsource.nextFloat() < 0.05F);
+        this.spawnType = p_21436_;
         return p_21437_;
     }
 
@@ -1139,6 +_,11 @@
         this.persistenceRequired = true;
     }
 
+   // CraftBukkit start
+   public void setPersistenceRequired(boolean persistenceRequired) {
+      this.persistenceRequired = persistenceRequired;
+   }
+   // CraftBukkit end
     @Override
     public void setDropChance(EquipmentSlot p_21410_, float p_21411_) {
         switch (p_21410_.getType()) {
@@ -1175,6 +_,16 @@
     public final InteractionResult interact(Player p_21420_, InteractionHand p_21421_) {
         if (!this.isAlive()) {
             return InteractionResult.PASS;
+        } else if (this.getLeashHolder() == p_21420_) {
+         // CraftBukkit start - fire PlayerUnleashEntityEvent
+         if (CraftEventFactory.callPlayerUnleashEntityEvent(this, p_21420_, p_21421_).isCancelled()) {
+            ((ServerPlayer) p_21420_).connection.send(new ClientboundSetEntityLinkPacket(this, this.getLeashHolder()));
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
+            this.dropLeash(true, !p_21420_.hasInfiniteMaterials());
+            this.gameEvent(GameEvent.ENTITY_INTERACT, p_21420_);
+            return InteractionResult.sidedSuccess(this.level().isClientSide);
         } else {
             InteractionResult interactionresult = this.checkAndHandleImportantInteractions(p_21420_, p_21421_);
             if (interactionresult.consumesAction()) {
@@ -1199,26 +_,38 @@
 
     private InteractionResult checkAndHandleImportantInteractions(Player p_21500_, InteractionHand p_21501_) {
         ItemStack itemstack = p_21500_.getItemInHand(p_21501_);
-        if (itemstack.is(Items.NAME_TAG)) {
-            InteractionResult interactionresult = itemstack.interactLivingEntity(p_21500_, this, p_21501_);
-            if (interactionresult.consumesAction()) {
-                return interactionresult;
+        if (itemstack.is(Items.LEAD) && this.canBeLeashed(p_21500_)) {
+         // CraftBukkit start - fire PlayerLeashEntityEvent
+         if (CraftEventFactory.callPlayerLeashEntityEvent(this, p_21500_, p_21500_, p_21501_).isCancelled()) {
+            ((ServerPlayer) p_21500_).connection.send(new ClientboundSetEntityLinkPacket(this, this.getLeashHolder()));
+            return InteractionResult.PASS;
+         }
+         // CraftBukkit end
+            this.setLeashedTo(p_21500_, true);
+            itemstack.shrink(1);
+            return InteractionResult.sidedSuccess(this.level().isClientSide);
+        } else {
+            if (itemstack.is(Items.NAME_TAG)) {
+                InteractionResult interactionresult = itemstack.interactLivingEntity(p_21500_, this, p_21501_);
+                if (interactionresult.consumesAction()) {
+                    return interactionresult;
+                }
             }
-        }
 
-        if (itemstack.getItem() instanceof SpawnEggItem) {
-            if (this.level() instanceof ServerLevel) {
-                SpawnEggItem spawneggitem = (SpawnEggItem)itemstack.getItem();
-                Optional<Mob> optional = spawneggitem.spawnOffspringFromSpawnEgg(
-                    p_21500_, this, (EntityType<? extends Mob>)this.getType(), (ServerLevel)this.level(), this.position(), itemstack
-                );
-                optional.ifPresent(p_21476_ -> this.onOffspringSpawnedFromEgg(p_21500_, p_21476_));
-                return optional.isPresent() ? InteractionResult.SUCCESS : InteractionResult.PASS;
+            if (itemstack.getItem() instanceof SpawnEggItem) {
+                if (this.level() instanceof ServerLevel) {
+                    SpawnEggItem spawneggitem = (SpawnEggItem)itemstack.getItem();
+                    Optional<Mob> optional = spawneggitem.spawnOffspringFromSpawnEgg(
+                        p_21500_, this, (EntityType<? extends Mob>)this.getType(), (ServerLevel)this.level(), this.position(), itemstack
+                    );
+                    optional.ifPresent(p_21476_ -> this.onOffspringSpawnedFromEgg(p_21500_, p_21476_));
+                    return optional.isPresent() ? InteractionResult.SUCCESS : InteractionResult.PASS;
+                } else {
+                    return InteractionResult.CONSUME;
+                }
             } else {
-                return InteractionResult.CONSUME;
+                return InteractionResult.PASS;
             }
-        } else {
-            return InteractionResult.PASS;
         }
     }
 
@@ -1292,6 +_,10 @@
                     }
                 }
 
+            	// CraftBukkit start
+            	if (CraftEventFactory.callEntityTransformEvent(this, t, this.transformReason.getAndSet(EntityTransformEvent.TransformReason.UNKNOWN)).isCancelled()) {
+               		return null;
+            	}
                 this.level().addFreshEntity(t);
                 if (this.isPassenger()) {
                     Entity entity = this.getVehicle();
@@ -1335,16 +_,121 @@
         return !(this instanceof Enemy);
     }
 
+   // Mohist start
+   public AtomicReference<TransformReason> transformReason = new AtomicReference<>(EntityTransformEvent.TransformReason.UNKNOWN);
+   public void transformReasonAndspawnReason(EntityTransformEvent.TransformReason transformReason, CreatureSpawnEvent.SpawnReason spawnReason) {
+      this.transformReason.set(transformReason);
+      this.spawnReason(spawnReason);
+   }
+   @Nullable
+   public <T extends Mob> T convertTo(EntityType<T> pEntityType, boolean pTransferInventory, EntityTransformEvent.TransformReason transformReason, CreatureSpawnEvent.SpawnReason spawnReason) {
+      transformReasonAndspawnReason(transformReason, spawnReason);
+      return convertTo(pEntityType, pTransferInventory);
+   }
+   // Mohist end
+
+    protected void tickLeash() {
+        if (this.delayedLeashInfo != null) {
+            this.restoreLeashFromSave();
+        }
+
+        if (this.leashHolder != null) {
+            if (!this.isAlive() || !this.leashHolder.isAlive()) {
+                this.dropLeash(true, true);
+            }
+        }
+    }
+
+    public void dropLeash(boolean p_21456_, boolean p_21457_) {
+        if (this.leashHolder != null) {
+            this.leashHolder = null;
+            this.delayedLeashInfo = null;
+            this.clearRestriction();
+            if (!this.level().isClientSide && p_21457_) {
+            this.forceDrops = true; // CraftBukkit
+            this.spawnAtLocation(Items.LEAD);
+            this.forceDrops = false; // CraftBukkit
+         }
+
+            if (!this.level().isClientSide && p_21456_ && this.level() instanceof ServerLevel) {
+                ((ServerLevel)this.level()).getChunkSource().broadcast(this, new ClientboundSetEntityLinkPacket(this, null));
+            }
+        }
+    }
+
+    public boolean canBeLeashed(Player p_21418_) {
+        return !this.isLeashed() && !(this instanceof Enemy);
+    }
+
+    public boolean isLeashed() {
+        return this.leashHolder != null;
+    }
+
+    public boolean mayBeLeashed() {
+        return this.isLeashed() || this.delayedLeashInfo != null;
+    }
+
+    @Nullable
+    public Entity getLeashHolder() {
+        if (this.leashHolder == null && this.delayedLeashHolderId != 0 && this.level().isClientSide) {
+            this.leashHolder = this.level().getEntity(this.delayedLeashHolderId);
+        }
+
+        return this.leashHolder;
+    }
+
+    public void setLeashedTo(Entity p_21464_, boolean p_21465_) {
+        this.leashHolder = p_21464_;
+        this.delayedLeashInfo = null;
+        if (!this.level().isClientSide && p_21465_ && this.level() instanceof ServerLevel) {
+            ((ServerLevel)this.level()).getChunkSource().broadcast(this, new ClientboundSetEntityLinkPacket(this, this.leashHolder));
+        }
+
+        if (this.isPassenger()) {
+            this.stopRiding();
+        }
+    }
+
+    public void setDelayedLeashHolderId(int p_21507_) {
+        this.delayedLeashHolderId = p_21507_;
+        this.dropLeash(false, false);
+    }
+
     @Override
     public boolean startRiding(Entity p_21396_, boolean p_21397_) {
         boolean flag = super.startRiding(p_21396_, p_21397_);
         if (flag && this.isLeashed()) {
+         this.level.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
             this.dropLeash(true, true);
         }
 
         return flag;
     }
 
+    private void restoreLeashFromSave() {
+        if (this.delayedLeashInfo != null && this.level() instanceof ServerLevel serverlevel) {
+            Optional<UUID> optional1 = this.delayedLeashInfo.left();
+            Optional<BlockPos> optional = this.delayedLeashInfo.right();
+            if (optional1.isPresent()) {
+                Entity entity = serverlevel.getEntity(optional1.get());
+                if (entity != null) {
+                    this.setLeashedTo(entity, true);
+                    return;
+                }
+            } else if (optional.isPresent()) {
+                this.setLeashedTo(LeashFenceKnotEntity.getOrCreateKnot(this.level(), optional.get()), true);
+                return;
+            }
+
+            if (this.tickCount > 100) {
+            	this.forceDrops = true; // CraftBukkit
+            	this.spawnAtLocation(Items.LEAD);
+            	this.forceDrops = false; // CraftBukkit
+            	this.delayedLeashInfo = null;
+            }
+        }
+    }
+
     @Override
     public boolean isEffectiveAi() {
         return super.isEffectiveAi() && !this.isNoAi();
@@ -1412,36 +_,42 @@
 
     @Override
     public boolean doHurtTarget(Entity p_21372_) {
-        float f = (float)this.getAttributeValue(Attributes.ATTACK_DAMAGE);
-        DamageSource damagesource = this.damageSources().mobAttack(this);
-        if (this.level() instanceof ServerLevel serverlevel) {
-            f = EnchantmentHelper.modifyDamage(serverlevel, this.getWeaponItem(), p_21372_, damagesource, f);
-        }
-
-        boolean flag = p_21372_.hurt(damagesource, f);
+        float f = (float) this.getAttributeValue(Attributes.ATTACK_DAMAGE);
+        float f1 = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
+        if (p_21372_ instanceof LivingEntity) {
+            f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), p_21372_.getType());
+            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+        }
+
+        int i = EnchantmentHelper.getFireAspect(this);
+        if (i > 0) {
+            // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+            EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), p_21372_.getBukkitEntity(), i * 4);
+            org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+
+            if (!combustEvent.isCancelled()) {
+                p_21372_.igniteForSeconds(combustEvent.getDuration(), false);
+            }
+            // CraftBukkit end
+        }
+
+        boolean flag = p_21372_.hurt(this.damageSources().mobAttack(this), f);
         if (flag) {
-            float f1 = this.getKnockback(p_21372_, damagesource);
-            if (f1 > 0.0F && p_21372_ instanceof LivingEntity livingentity) {
-                livingentity.knockback(
-                    (double)(f1 * 0.5F),
-                    (double)Mth.sin(this.getYRot() * (float) (Math.PI / 180.0)),
-                    (double)(-Mth.cos(this.getYRot() * (float) (Math.PI / 180.0)))
-                );
+            if (f1 > 0.0F && p_21372_ instanceof LivingEntity) {
+                ((LivingEntity) p_21372_)
+                        .knockback(
+                                (double) (f1 * 0.5F),
+                                (double) Mth.sin(this.getYRot() * (float) (Math.PI / 180.0)),
+                                (double) (-Mth.cos(this.getYRot() * (float) (Math.PI / 180.0)))
+                        );
                 this.setDeltaMovement(this.getDeltaMovement().multiply(0.6, 1.0, 0.6));
             }
 
-            if (this.level() instanceof ServerLevel serverlevel1) {
-                EnchantmentHelper.doPostAttackEffects(serverlevel1, p_21372_, damagesource);
-            }
-
+            this.doEnchantDamageEffects(this, p_21372_);
             this.setLastHurtMob(p_21372_);
-            this.playAttackSound();
         }
 
         return flag;
-    }
-
-    protected void playAttackSound() {
     }
 
     protected boolean isSunBurnTick() {
@@ -1457,15 +_,25 @@
         return false;
     }
 
+    @Deprecated // FORGE: use jumpInFluid instead
     @Override
     protected void jumpInLiquid(TagKey<Fluid> p_204045_) {
+        this.jumpInLiquidInternal(() -> super.jumpInLiquid(p_204045_));
+    }
+
+    private void jumpInLiquidInternal(Runnable onSuper) {
         if (this.getNavigation().canFloat()) {
-            super.jumpInLiquid(p_204045_);
+            onSuper.run();
         } else {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.3, 0.0));
         }
     }
 
+    @Override
+    public void jumpInFluid(net.minecraftforge.fluids.FluidType type) {
+       this.jumpInLiquidInternal(() -> super.jumpInFluid(type));
+    }
+
     @VisibleForTesting
     public void removeFreeWill() {
         this.removeAllGoals(p_341273_ -> true);
@@ -1479,6 +_,8 @@
     @Override
     protected void removeAfterChangingDimensions() {
         super.removeAfterChangingDimensions();
+      this.level.getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
+        this.dropLeash(true, false);
         this.getAllSlots().forEach(p_278936_ -> {
             if (!p_278936_.isEmpty()) {
                 p_278936_.setCount(0);
@@ -1491,5 +_,40 @@
     public ItemStack getPickResult() {
         SpawnEggItem spawneggitem = SpawnEggItem.byId(this.getType());
         return spawneggitem == null ? null : new ItemStack(spawneggitem);
+    }
+
+    /**
+    * Returns the type of spawn that created this mob, if applicable.
+    * If it could not be determined, this will return null.
+    * <p>
+    * This is set via {@link Mob#finalizeSpawn}, so you should not call this from within that method, instead using the parameter.
+    */
+    @Nullable
+    public final MobSpawnType getSpawnType() {
+        return this.spawnType;
+    }
+
+    /**
+     * This method exists so that spawns can be cancelled from the {@link net.minecraftforge.event.entity.living.MobSpawnEvent.FinalizeSpawn FinalizeSpawnEvent}
+     * without needing to hook up an additional handler for the {@link net.minecraftforge.event.entity.EntityJoinLevelEvent EntityJoinLevelEvent}.
+     * @return if this mob will be blocked from spawning during {@link Level#addFreshEntity(Entity)}
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final boolean isSpawnCancelled() {
+        return this.spawnCancelled;
+    }
+
+    /**
+     * Marks this mob as being disallowed to spawn during {@link Level#addFreshEntity(Entity)}.<p>
+     * @throws UnsupportedOperationException if this entity has already been {@link Entity#isAddedToWorld() added to the world}.
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final void setSpawnCancelled(boolean cancel) {
+        if (this.isAddedToWorld()) {
+            throw new UnsupportedOperationException("Late invocations of Mob#setSpawnCancelled are not permitted.");
+        }
+        this.spawnCancelled = cancel;
     }
 }
