--- a/net/minecraft/world/entity/NeutralMob.java
+++ b/net/minecraft/world/entity/NeutralMob.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public interface NeutralMob {
    String TAG_ANGER_TIME = "AngerTime";
@@ -108,7 +_,7 @@
    default void stopBeingAngry() {
       this.setLastHurtByMob((LivingEntity)null);
       this.setPersistentAngerTarget((UUID)null);
-      this.setTarget((LivingEntity)null);
+      this.setTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.setRemainingPersistentAngerTime(0);
    }
 
@@ -120,6 +_,8 @@
    void setLastHurtByPlayer(@Nullable Player p_21680_);
 
    void setTarget(@Nullable LivingEntity p_21681_);
+
+   boolean setTarget(@Nullable LivingEntity pLivingEntity, EntityTargetEvent.TargetReason reason, boolean fireEvent); // CraftBukkit
 
    boolean canAttack(LivingEntity p_181126_);
 
