--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.scores.PlayerTeam;
+import net.minecraft.world.scores.Team;
 
 public abstract class TamableAnimal extends Animal implements OwnableEntity {
    protected static final EntityDataAccessor<Byte> DATA_FLAGS_ID = SynchedEntityData.defineId(TamableAnimal.class, EntityDataSerializers.BYTE);
@@ -186,11 +_,15 @@
    }
 
    public void die(DamageSource p_21809_) {
-      if (!this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES) && this.getOwner() instanceof ServerPlayer) {
-         this.getOwner().sendSystemMessage(this.getCombatTracker().getDeathMessage());
-      }
-
+      // FORGE: Super moved to top so that death message would be cancelled properly
+      net.minecraft.network.chat.Component deathMessage = this.getCombatTracker().getDeathMessage();
       super.die(p_21809_);
+
+      if (this.dead)
+      if (!this.level().isClientSide && this.level().getGameRules().getBoolean(GameRules.RULE_SHOWDEATHMESSAGES) && this.getOwner() instanceof ServerPlayer) {
+         this.getOwner().sendSystemMessage(deathMessage);
+      }
+
    }
 
    public boolean isOrderedToSit() {
