--- ../src-base/minecraft/net/minecraft/command/CommandTime.java
+++ ../src-work/minecraft/net/minecraft/command/CommandTime.java
@@ -105,20 +105,13 @@
         }
     }
 
-    protected void setAllWorldTimes(MinecraftServer server, int time)
+    protected void setAllWorldTimes(ICommandSender sender, int time)
     {
-        for (int i = 0; i < server.worlds.length; ++i)
-        {
-            server.worlds[i].setWorldTime((long)time);
-        }
+        sender.getEntityWorld().setWorldTime((long)time);
     }
 
-    protected void incrementAllWorldTimes(MinecraftServer server, int amount)
+    protected void incrementAllWorldTimes(ICommandSender sender, int amount)
     {
-        for (int i = 0; i < server.worlds.length; ++i)
-        {
-            WorldServer worldserver = server.worlds[i];
-            worldserver.setWorldTime(worldserver.getWorldTime() + (long)amount);
-        }
+        sender.getEntityWorld().setWorldTime(sender.getEntityWorld().getWorldTime() + (long)amount);
     }
 }
