--- ../src-base/minecraft/net/minecraft/command/CommandBase.java
+++ ../src-work/minecraft/net/minecraft/command/CommandBase.java
@@ -40,6 +40,7 @@
     private static ICommandListener commandListener;
     private static final Splitter COMMA_SPLITTER = Splitter.on(',');
     private static final Splitter EQUAL_SPLITTER = Splitter.on('=').limit(2);
+    public String permissionNode = "minecraft";
 
     protected static SyntaxErrorException toSyntaxException(JsonParseException e)
     {
@@ -88,7 +89,7 @@
 
     public boolean checkPermission(MinecraftServer server, ICommandSender sender)
     {
-        return sender.canUseCommand(this.getRequiredPermissionLevel(), this.getName());
+        return sender.canUseCommand(this.getRequiredPermissionLevel(), this.getName()) || sender.canUseCommand(this.getRequiredPermissionLevel(), this.getName(), this.permissionNode);
     }
 
     public List<String> getTabCompletions(MinecraftServer server, ICommandSender sender, String[] args, @Nullable BlockPos targetPos)
