--- a/net/minecraft/world/level/levelgen/feature/featuresize/FeatureSizeType.java
+++ b/net/minecraft/world/level/levelgen/feature/featuresize/FeatureSizeType.java
@@ -13,7 +_,7 @@
         return Registry.register(BuiltInRegistries.FEATURE_SIZE_TYPE, p_68304_, new FeatureSizeType<>(p_335596_));
     }
 
-    private FeatureSizeType(MapCodec<P> p_332315_) {
+    public FeatureSizeType(MapCodec<P> p_332315_) {
         this.codec = p_332315_;
     }
 
