--- ../src-base/minecraft/org/bukkit/entity/Boat.java
+++ ../src-work/minecraft/org/bukkit/entity/Boat.java
@@ -9,14 +9,14 @@
 
     /**
      * Gets the wood type of the boat.
-     * 
+     *
      * @return the wood type
      */
     TreeSpecies getWoodType();
 
     /**
      * Sets the wood type of the boat.
-     * 
+     *
      * @param species the new wood type
      */
     void setWoodType(TreeSpecies species);
