--- ../src-base/minecraft/org/bukkit/entity/EntityType.java
+++ ../src-work/minecraft/org/bukkit/entity/EntityType.java
@@ -2,17 +2,16 @@
 
 import java.util.HashMap;
 import java.util.Map;
-
+import org.bukkit.Location;
+import org.bukkit.World;
 import org.bukkit.entity.minecart.CommandMinecart;
-import org.bukkit.entity.minecart.HopperMinecart;
-import org.bukkit.entity.minecart.SpawnerMinecart;
-import org.bukkit.entity.minecart.RideableMinecart;
 import org.bukkit.entity.minecart.ExplosiveMinecart;
+import org.bukkit.entity.minecart.HopperMinecart;
 import org.bukkit.entity.minecart.PoweredMinecart;
+import org.bukkit.entity.minecart.RideableMinecart;
+import org.bukkit.entity.minecart.SpawnerMinecart;
 import org.bukkit.entity.minecart.StorageMinecart;
 import org.bukkit.inventory.ItemStack;
-import org.bukkit.Location;
-import org.bukkit.World;
 
 public enum EntityType {
 
