--- ../src-base/minecraft/org/bukkit/entity/Slime.java
+++ ../src-work/minecraft/org/bukkit/entity/Slime.java
@@ -16,17 +16,17 @@
     public void setSize(int sz);
 
     /**
-     * Set the {@link LivingEntity} target for this slime. Set to null to clear
-     * the target.
+     * Get the {@link LivingEntity} this slime is currently targeting.
      *
-     * @param target the entity to target
+     * @return the current target, or null if no target exists.
      */
-    public void setTarget(LivingEntity target);
+    public LivingEntity getTarget();
 
     /**
-     * Get the {@link LivingEntity} this slime is currently targeting.
+     * Set the {@link LivingEntity} target for this slime. Set to null to clear
+     * the target.
      *
-     * @return the current target, or null if no target exists.
+     * @param target the entity to target
      */
-    public LivingEntity getTarget();
+    public void setTarget(LivingEntity target);
 }
