--- ../src-base/minecraft/org/bukkit/entity/Zombie.java
+++ ../src-work/minecraft/org/bukkit/entity/Zombie.java
@@ -36,16 +36,16 @@
     public void setVillager(boolean flag);
 
     /**
-     * @param profession
+     * @return profession
      * @see ZombieVillager#getVillagerProfession()
      */
     @Deprecated
-    public void setVillagerProfession(Villager.Profession profession);
+    public Villager.Profession getVillagerProfession();
 
     /**
-     * @return profession
+     * @param profession
      * @see ZombieVillager#getVillagerProfession()
      */
     @Deprecated
-    public Villager.Profession getVillagerProfession();
+    public void setVillagerProfession(Villager.Profession profession);
 }
