--- ../src-base/minecraft/net/minecraftforge/common/ForgeModContainer.java
+++ ../src-work/minecraft/net/minecraftforge/common/ForgeModContainer.java
@@ -500,10 +500,6 @@
         ForgeChunkManager.captureConfig(evt.getModConfigurationDirectory());
         MinecraftForge.EVENT_BUS.register(this);
 
-        if (!ForgeModContainer.disableVersionCheck)
-        {
-            ForgeVersion.startVersionCheck();
-        }
     }
 
     @SubscribeEvent
