--- ../src-base/minecraft/net/minecraft/inventory/ContainerRepair.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerRepair.java
@@ -18,6 +18,8 @@
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView;
 
 public class ContainerRepair extends Container
 {
@@ -31,6 +33,10 @@
     public String field_82857_m;
     private final EntityPlayer field_82855_n;
 
+    private int lastLevelCost;
+    private CraftInventoryView bukkitEntity;
+    private InventoryPlayer playerInventory;
+
     @SideOnly(Side.CLIENT)
     public ContainerRepair(InventoryPlayer p_i45806_1_, World p_i45806_2_, EntityPlayer p_i45806_3_)
     {
@@ -39,6 +45,7 @@
 
     public ContainerRepair(InventoryPlayer p_i45807_1_, final World p_i45807_2_, final BlockPos p_i45807_3_, EntityPlayer p_i45807_4_)
     {
+        this.playerInventory = p_i45807_1_;
         this.field_82852_f = new InventoryCraftResult();
         this.field_82853_g = new InventoryBasic("Repair", true, 2)
         {
@@ -70,6 +77,8 @@
                     p_190901_1_.func_82242_a(-ContainerRepair.this.field_82854_e);
                 }
 
+                float breakChance = net.minecraftforge.common.ForgeHooks.onAnvilRepair(p_190901_1_, p_190901_2_, ContainerRepair.this.field_82853_g.func_70301_a(0), ContainerRepair.this.field_82853_g.func_70301_a(1));
+
                 ContainerRepair.this.field_82853_g.func_70299_a(0, ItemStack.field_190927_a);
 
                 if (ContainerRepair.this.field_82856_l > 0)
@@ -94,7 +103,7 @@
                 ContainerRepair.this.field_82854_e = 0;
                 IBlockState iblockstate = p_i45807_2_.func_180495_p(p_i45807_3_);
 
-                if (!p_190901_1_.field_71075_bZ.field_75098_d && !p_i45807_2_.field_72995_K && iblockstate.func_177230_c() == Blocks.field_150467_bQ && p_190901_1_.func_70681_au().nextFloat() < 0.12F)
+                if (!p_190901_1_.field_71075_bZ.field_75098_d && !p_i45807_2_.field_72995_K && iblockstate.func_177230_c() == Blocks.field_150467_bQ && p_190901_1_.func_70681_au().nextFloat() < breakChance)
                 {
                     int l = ((Integer)iblockstate.func_177229_b(BlockAnvil.field_176505_b)).intValue();
                     ++l;
@@ -153,7 +162,9 @@
 
         if (itemstack.func_190926_b())
         {
-            this.field_82852_f.func_70299_a(0, ItemStack.field_190927_a);
+            if (this.getBukkitView() != null) {
+                CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.field_190927_a);
+            }
             this.field_82854_e = 0;
         }
         else
@@ -163,10 +174,18 @@
             Map<Enchantment, Integer> map = EnchantmentHelper.func_82781_a(itemstack1);
             j = j + itemstack.func_82838_A() + (itemstack2.func_190926_b() ? 0 : itemstack2.func_82838_A());
             this.field_82856_l = 0;
+            boolean flag = false;
 
             if (!itemstack2.func_190926_b())
             {
-                boolean flag = itemstack2.func_77973_b() == Items.field_151134_bR && !ItemEnchantedBook.func_92110_g(itemstack2).func_82582_d();
+                if (!net.minecraftforge.common.ForgeHooks.onAnvilChange(this, itemstack, itemstack2, field_82852_f, field_82857_m, j))
+                {
+                    if (this.getBukkitView() != null) {
+                        CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), this.field_82852_f.func_70301_a(0));
+                    }
+                    return;
+                }
+                flag = itemstack2.func_77973_b() == Items.field_151134_bR && !ItemEnchantedBook.func_92110_g(itemstack2).func_82582_d();
 
                 if (itemstack1.func_77984_f() && itemstack1.func_77973_b().func_82789_a(itemstack, itemstack2))
                 {
@@ -174,8 +193,11 @@
 
                     if (l2 <= 0)
                     {
-                        this.field_82852_f.func_70299_a(0, ItemStack.field_190927_a);
-                        this.field_82854_e = 0;
+						//this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        if (getBukkitView() != null) {
+							org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.field_190927_a);
+                        }
+						this.field_82854_e = 0;
                         return;
                     }
 
@@ -195,7 +217,9 @@
                 {
                     if (!flag && (itemstack1.func_77973_b() != itemstack2.func_77973_b() || !itemstack1.func_77984_f()))
                     {
-                        this.field_82852_f.func_70299_a(0, ItemStack.field_190927_a);
+                        if (this.getBukkitView() != null) {
+                            CraftEventFactory.callPrepareAnvilEvent(this.getBukkitView(), ItemStack.field_190927_a);
+                        }
                         this.field_82854_e = 0;
                         return;
                     }
@@ -213,7 +237,7 @@
                             l1 = 0;
                         }
 
-                        if (l1 < itemstack1.func_77960_j())
+                        if (l1 < itemstack1.func_77952_i()) // vanilla uses metadata here instead of damage.
                         {
                             itemstack1.func_77964_b(l1);
                             i += 2;
@@ -295,8 +319,11 @@
 
                     if (flag3 && !flag2)
                     {
-                        this.field_82852_f.func_70299_a(0, ItemStack.field_190927_a);
-                        this.field_82854_e = 0;
+						//this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        if (getBukkitView() != null) {
+							org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.field_190927_a);
+                        }
+						this.field_82854_e = 0;
                         return;
                     }
                 }
@@ -317,6 +344,7 @@
                 i += k;
                 itemstack1.func_151001_c(this.field_82857_m);
             }
+            if (flag && !itemstack1.func_77973_b().isBookEnchantable(itemstack1, itemstack2)) itemstack1 = ItemStack.field_190927_a;
 
             this.field_82854_e = j + i;
 
@@ -353,8 +381,11 @@
                 EnchantmentHelper.func_82782_a(map, itemstack1);
             }
 
-            this.field_82852_f.func_70299_a(0, itemstack1);
-            this.func_75142_b();
+			//this.outputSlot.setInventorySlotContents(0, itemstack1);
+            if (getBukkitView() != null) {
+				org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), itemstack1);
+            }
+			this.func_75142_b();
         }
     }
 
@@ -385,6 +416,7 @@
 
     public boolean func_75145_c(EntityPlayer p_75145_1_)
     {
+        if (!this.checkReachable) return true;
         if (this.field_82860_h.func_180495_p(this.field_178156_j).func_177230_c() != Blocks.field_150467_bQ)
         {
             return false;
@@ -466,4 +498,29 @@
 
         this.func_82848_d();
     }
+
+    @Override
+    public void func_75142_b() {
+        super.func_75142_b();
+        for (int i = 0; i < this.field_75149_d.size(); ++i) {
+            IContainerListener icrafting = this.field_75149_d.get(i);
+			if (this.lastLevelCost != this.field_82854_e) {
+            	icrafting.func_71112_a(this, 0, this.field_82854_e);
+			}
+        }
+
+        this.lastLevelCost = this.field_82854_e;
+    }
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryAnvil(
+                new org.bukkit.Location(field_82860_h.getWorld(), field_178156_j.func_177958_n(), field_178156_j.func_177956_o(), field_178156_j.func_177952_p()), this.field_82853_g, this.field_82852_f, this);
+        bukkitEntity = new CraftInventoryView(this.playerInventory.field_70458_d.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
 }
