--- a/net/minecraft/inventory/CraftResultInventory.java
+++ b/net/minecraft/inventory/CraftResultInventory.java
@@ -5,11 +5,60 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.IRecipe;
 import net.minecraft.util.NonNullList;
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
 
 public class CraftResultInventory implements IInventory, IRecipeHolder {
    private final NonNullList<ItemStack> field_70467_a = NonNullList.func_191197_a(1, ItemStack.field_190927_a);
    private IRecipe<?> field_193057_b;
 
+   // CraftBukkit start
+   private int maxStack = MAX_STACK;
+
+   @Override
+   public java.util.List<ItemStack> getContents() {
+      return this.field_70467_a;
+   }
+
+   @Override
+   public InventoryHolder getOwner() {
+      return null;
+   }
+
+   @Override
+   public void onOpen(final CraftHumanEntity who) {
+   }
+
+   @Override
+   public void onClose(final CraftHumanEntity who) {
+   }
+
+   @Override
+   public java.util.List<HumanEntity> getViewers() {
+      return new java.util.ArrayList<HumanEntity>();
+   }
+
+   @Override
+   public int func_70297_j_() {
+      return this.maxStack;
+   }
+
+   @Override
+   public void setMaxStackSize(final int size) {
+      this.maxStack = size;
+   }
+
+   @Override
+   public Location getLocation() {
+      return null;
+   }
+
+   // CraftBukkit end
+
    public int func_70302_i_() {
       return 1;
    }
