--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
@@ -33,7 +_,7 @@
    }
 
    public void start() {
-      this.mob.setTarget(this.ownerLastHurtBy);
+      this.mob.setTarget(this.ownerLastHurtBy, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit - reason
       LivingEntity livingentity = this.tameAnimal.getOwner();
       if (livingentity != null) {
          this.timestamp = livingentity.getLastHurtByMobTimestamp();
