--- ../src-base/minecraft/org/bukkit/metadata/MetadataStoreBase.java
+++ ../src-work/minecraft/org/bukkit/metadata/MetadataStoreBase.java
@@ -1,10 +1,15 @@
 package org.bukkit.metadata;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
 import org.apache.commons.lang.Validate;
 import org.bukkit.plugin.Plugin;
 
-import java.util.*;
-
 public abstract class MetadataStoreBase<T> {
     private Map<String, Map<Plugin, MetadataValue>> metadataMap = new HashMap<String, Map<Plugin, MetadataValue>>();
 
