--- a/net/minecraft/data/DataGenerator.java
+++ b/net/minecraft/data/DataGenerator.java
@@ -2,6 +_,7 @@
 
 import com.google.common.base.Stopwatch;
 import com.google.common.collect.Lists;
+import com.mohistmc.util.i18n.i18n;
 import java.io.IOException;
 import java.nio.file.Path;
 import java.util.Collection;
@@ -16,10 +_,11 @@
    private final Collection<Path> field_200394_b;
    private final Path field_200395_c;
    private final List<IDataProvider> field_200396_d = Lists.newArrayList();
+   private final List<IDataProvider> providerView = java.util.Collections.unmodifiableList(field_200396_d);
 
    public DataGenerator(Path p_i48266_1_, Collection<Path> p_i48266_2_) {
       this.field_200395_c = p_i48266_1_;
-      this.field_200394_b = p_i48266_2_;
+      this.field_200394_b = Lists.newArrayList(p_i48266_2_);
    }
 
    public Collection<Path> func_200389_a() {
@@ -37,20 +_,29 @@
       Stopwatch stopwatch1 = Stopwatch.createUnstarted();
 
       for(IDataProvider idataprovider : this.field_200396_d) {
-         field_200393_a.info("Starting provider: {}", (Object)idataprovider.func_200397_b());
+         field_200393_a.info(i18n.get("datagenerator.1", (Object)idataprovider.func_200397_b()));
+         net.minecraftforge.fml.StartupMessageManager.addModMessage("Generating: " + idataprovider.func_200397_b());
          stopwatch1.start();
          idataprovider.func_200398_a(directorycache);
          stopwatch1.stop();
-         field_200393_a.info("{} finished after {} ms", idataprovider.func_200397_b(), stopwatch1.elapsed(TimeUnit.MILLISECONDS));
+         field_200393_a.info(i18n.get("datagenerator.2", idataprovider.func_200397_b(), stopwatch1.elapsed(TimeUnit.MILLISECONDS)));
          stopwatch1.reset();
       }
 
-      field_200393_a.info("All providers took: {} ms", (long)stopwatch.elapsed(TimeUnit.MILLISECONDS));
+      field_200393_a.info(i18n.get("datagenerator.3", (long)stopwatch.elapsed(TimeUnit.MILLISECONDS)));
       directorycache.func_208317_a();
    }
 
    public void func_200390_a(IDataProvider p_200390_1_) {
       this.field_200396_d.add(p_200390_1_);
+   }
+
+   public List<IDataProvider> getProviders() {
+       return this.providerView;
+   }
+
+   public void addInput(Path value) {
+      this.field_200394_b.add(value);
    }
 
    static {
