--- a/net/minecraft/data/RecipeProvider.java
+++ b/net/minecraft/data/RecipeProvider.java
@@ -5,6 +_,7 @@
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
+import com.mohistmc.util.i18n.i18n;
 import java.io.BufferedWriter;
 import java.io.IOException;
 import java.nio.file.Files;
@@ -59,6 +_,7 @@
 
          }
       });
+      if (this.getClass() == RecipeProvider.class) //Forge: Subclasses don't need this.
       func_208310_b(p_200398_1_, Advancement.Builder.func_200278_a().func_200275_a("impossible", new ImpossibleTrigger.Instance()).func_200273_b(), path.resolve("data/minecraft/advancements/recipes/root.json"));
    }
 
@@ -76,12 +_,12 @@
 
          p_208311_0_.func_208316_a(p_208311_2_, s1);
       } catch (IOException ioexception) {
-         field_200411_a.error("Couldn't save recipe {}", p_208311_2_, ioexception);
+         field_200411_a.error(i18n.get("recipeprovider.1", p_208311_2_, ioexception));
       }
 
    }
 
-   protected static void func_208310_b(DirectoryCache p_208310_0_, JsonObject p_208310_1_, Path p_208310_2_) {
+   protected void func_208310_b(DirectoryCache p_208310_0_, JsonObject p_208310_1_, Path p_208310_2_) {
       try {
          String s = field_200412_b.toJson((JsonElement)p_208310_1_);
          String s1 = field_208307_a.hashUnencodedChars(s).toString();
@@ -95,12 +_,12 @@
 
          p_208310_0_.func_208316_a(p_208310_2_, s1);
       } catch (IOException ioexception) {
-         field_200411_a.error("Couldn't save recipe advancement {}", p_208310_2_, ioexception);
+         field_200411_a.error(i18n.get("recipeprovider.2", p_208310_2_, ioexception));
       }
 
    }
 
-   protected static void func_200404_a(Consumer<IFinishedRecipe> p_200404_0_) {
+   protected void func_200404_a(Consumer<IFinishedRecipe> p_200404_0_) {
       func_240470_a_(p_200404_0_, Blocks.field_196670_r, ItemTags.field_203297_q);
       func_240472_b_(p_200404_0_, Blocks.field_196666_p, ItemTags.field_203296_p);
       func_240472_b_(p_200404_0_, Blocks.field_235344_mC_, ItemTags.field_232913_w_);
