--- ../src-base/minecraft/net/minecraft/world/gen/feature/WorldGenShrub.java
+++ ../src-work/minecraft/net/minecraft/world/gen/feature/WorldGenShrub.java
@@ -1,13 +1,12 @@
 package net.minecraft.world.gen.feature;
 
-import java.util.Random;
-import net.minecraft.block.Block;
-import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
+import java.util.Random;
+
 public class WorldGenShrub extends WorldGenTrees
 {
     private final IBlockState leavesMetadata;
@@ -60,7 +59,11 @@
                     }
                 }
             }
+        // CraftBukkit start - Return false if gen was unsuccessful
+        } else {
+            return false;
         }
+        // CraftBukkit end
 
         return true;
     }
