--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -113,6 +_,12 @@
       this.structureRepository.clear();
    }
 
+   // Mohist start Fix modernfix mod
+   public Optional<StructureTemplate> loadFromResource0(ResourceLocation p_230428_) {
+      return loadFromResource(p_230428_);
+   }
+   // Mohist end
+
    private Optional<StructureTemplate> loadFromResource(ResourceLocation p_230428_) {
       ResourceLocation resourcelocation = LISTER.idToFile(p_230428_);
       return this.load(() -> {
