From 0b262d88eb8a65678834a42558d6c29ebfe523cd Mon Sep 17 00:00:00 2001
From: Andrew Steinborn <git@steinborn.me>
Date: Fri, 13 Mar 2020 02:49:01 -0400
Subject: [PATCH] Add Bukkit#getConsoleSender()


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 8438831d..040e5187 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -260,4 +260,10 @@ public final class Bukkit {
     public static void reloadWhitelist() {
         server.reloadWhitelist();
     }
+
+    // Canyon start
+    public static org.bukkit.command.ConsoleCommandSender getConsoleSender() {
+        return server.getConsoleSender();
+    }
+    // Canyon end
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8b21ad20..dbea05bc 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -450,4 +450,12 @@ public interface Server {
      */
     public double[] getTPS();
     // Canyon end
+
+    // Canyon start
+    /**
+     * Returns a command sender representing the console
+     * @return a console command sender
+     */
+    org.bukkit.command.ConsoleCommandSender getConsoleSender();
+    // Canyon end
 }
-- 
2.26.2

