--- ../src-base/minecraft/net/minecraftforge/common/ISpecialArmor.java
+++ ../src-work/minecraft/net/minecraftforge/common/ISpecialArmor.java
@@ -131,6 +131,11 @@
          */
         public static float applyArmor(EntityLivingBase entity, NonNullList<ItemStack> inventory, DamageSource source, double damage)
         {
+            return ApplyArmor(entity, inventory, source, damage, true);
+        }
+
+        public static float ApplyArmor(EntityLivingBase entity, NonNullList<ItemStack> inventory, DamageSource source, double damage, boolean applyDamage)
+        {
             if (DEBUG)
             {
                 System.out.println("Start: " + damage);
@@ -201,9 +206,12 @@
                         int itemDamage = (int)Math.max(1, absorb);
                         if (stack.getItem() instanceof ISpecialArmor)
                         {
+                            if (applyDamage)
+                            {
                             ((ISpecialArmor)stack.getItem()).damageArmor(entity, stack, source, itemDamage, prop.Slot);
                         }
-                        else
+                    }
+                    else if(applyDamage)
                         {
                             if (DEBUG)
                             {
