--- a/net/minecraft/util/datafix/DataFixers.java
+++ b/net/minecraft/util/datafix/DataFixers.java
@@ -406,6 +_,14 @@
       Schema schema44 = p_14514_.addSchema(1456, SAME_NAMESPACED);
       p_14514_.addFixer(new EntityItemFrameDirectionFix(schema44, false));
       Schema schema45 = p_14514_.addSchema(1458, SAME_NAMESPACED);
+      // CraftBukkit start
+      p_14514_.addFixer(new com.mojang.datafixers.DataFix(schema45, false) {
+         @Override
+         protected com.mojang.datafixers.TypeRewriteRule makeRule() {
+            return this.fixTypeEverywhereTyped("Player CustomName", this.getInputSchema().getType(References.PLAYER), (typed) -> typed.update(DSL.remainderFinder(), EntityCustomNameToComponentFix::fixTagCustomName));
+         }
+      });
+      // CraftBukkit end
       p_14514_.addFixer(new EntityCustomNameToComponentFix(schema45, false));
       p_14514_.addFixer(new ItemCustomNameToComponentFix(schema45, false));
       p_14514_.addFixer(new BlockEntityCustomNameToComponentFix(schema45, false));
